{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Relative Abundance and Alpha Diversity Plots\n",
    "\n",
    "Date created: 11/15/2024\n",
    "\n",
    "This notebook plots the following:\n",
    "\n",
    "- 16S V4 relative abundance plots at Genus taxon level\n",
    "- 16S V4 Shannon alpha diversity plots at Genus taxon level\n",
    "- 16S V4 Faith PD alpha diversity plots at Genus taxon level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 215,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import Python packages\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import biom\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from itertools import cycle\n",
    "import os\n",
    "from matplotlib.colors import ListedColormap\n",
    "from matplotlib.colors import to_rgba"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 216,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the taxa level\n",
    "taxa_level = 'Genus'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 217,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "case_type\n",
       "case-nonlesional skin       111\n",
       "case-anterior nares         108\n",
       "case-lesional skin          107\n",
       "control-anterior nares       89\n",
       "control-nonlesional skin     87\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 217,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load the metadata\n",
    "metadata_path = '../Data/Metadata/updated_clean_ant_skin_metadata.tab'\n",
    "metadata = pd.read_csv(metadata_path, sep='\\t')\n",
    "metadata['case_type'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 218,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define paths to the collapsed taxa tables\n",
    "biom_paths = {\n",
    "    '16S_V4': '../Data/Tables/Relative_Abundance_Tables/df_16S_filtered_feature_table_rare_Genus_relative_abundance.biom'\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Predefined color palette for specific families\n",
    "if taxa_level == \"Genus\":\n",
    "    taxa_colors = {\n",
    "        ' g__Cutibacterium': '#ffa505',  # Bright orange\n",
    "        ' g__Staphylococcus': '#92f0f0',      # Fluorescent light blue\n",
    "        ' g__Streptococcus': '#FF0000',    # Red\n",
    "        ' g__Corynebacterium': '#ffe59a',        # Pastel yellow\n",
    "        ' g__Lawsonella': '#70a8dc',         # Light blue\n",
    "        ' g__Veillonella': '#c5bce0',         # Pastel purplish\n",
    "        ' g__Micrococcus':'#f4cccd',           # Pastel yellow\n",
    "        ' g__Alloprevotella': '#bcbcbc',        # Light gray\n",
    "        ' g__Lactobacillus': '#daead3',     # Pale mint green\n",
    "        ' g__Neisseria': '#f6475f',         # Redish pink\n",
    "        'Others': '#ededed'                 # White\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 220,
   "metadata": {},
   "outputs": [],
   "source": [
    "# A list of unique colors to use for taxa not predefined\n",
    "unique_colors = sns.color_palette(\"deep\", n_colors=20).as_hex()\n",
    "unique_color_iter = cycle(unique_colors)  # Iterator to cycle through unique colors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 221,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to load BIOM table, collapse by taxa, sort rows by row sum, remove specified samples, and convert to relative abundance\n",
    "def load_biom_table(biom_path, metadata_path):\n",
    "    # Load metadata as a DataFrame from the file path\n",
    "    metadata = pd.read_csv(metadata_path, sep='\\t')\n",
    "\n",
    "    # Load BIOM table and convert to a DataFrame\n",
    "    table = biom.load_table(biom_path)\n",
    "    df = pd.DataFrame(table.matrix_data.toarray(),\n",
    "                      index=table.ids(axis='observation'),\n",
    "                      columns=table.ids(axis='sample'))\n",
    "    \n",
    "    # Sort rows by row sum in descending order\n",
    "    df['row_sum'] = df.sum(axis=1)\n",
    "    df = df.sort_values(by='row_sum', ascending=False)\n",
    "    \n",
    "    # Drop the 'row_sum' column before proceeding\n",
    "    df = df.drop(columns=['row_sum'])\n",
    "\n",
    "    # Replace ' g__' rows with ' g__Unknown'\n",
    "    df.index = df.index.map(lambda x: ' g__Unknown' if x == ' g__' else x)\n",
    "    \n",
    "    return df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 222,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to determine the top 15 families and collapse the rest as \"Others\"\n",
    "def collapse_top_15(df):\n",
    "    top_genera = df.sum(axis=1).nlargest(15).index  # Select top 15 families\n",
    "    df_top = df.loc[top_genera]\n",
    "    df_top.loc['Others'] = df.loc[~df.index.isin(top_genera)].sum()\n",
    "    return df_top"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 223,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to get or assign colors to families\n",
    "def get_taxa_colors(families, global_taxa_color_map):\n",
    "    for taxa in families:\n",
    "        if taxa not in global_taxa_color_map:\n",
    "            if taxa in taxa_colors:\n",
    "                global_taxa_color_map[taxa] = taxa_colors[taxa]\n",
    "            else:\n",
    "                global_taxa_color_map[taxa] = next(unique_color_iter)  # Assign a new unique color\n",
    "    return global_taxa_color_map"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Relative abundance plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 224,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_relative_abundance(df, metadata, group_column, output_dir, key, taxa_color_map):\n",
    "    # Average by group\n",
    "    df_grouped = df.groupby(metadata[group_column], axis=1).mean()\n",
    "\n",
    "    # Reorder the columns\n",
    "    desired_order = ['control-nonlesional skin','case-nonlesional skin', 'case-lesional skin', 'control-anterior nares', 'case-anterior nares']\n",
    "\n",
    "    df_grouped = df_grouped[desired_order]\n",
    "    \n",
    "    # Create output file paths\n",
    "    output_png_file = os.path.join(output_dir, f'{key}_{taxa_level}_relative_abundance_plot.png')  # Save as png\n",
    "    output_svg_file = os.path.join(output_dir, f'{key}_{taxa_level}_relative_abundance_plot.svg')  # Save as svg\n",
    "    \n",
    "    # Set plot title based on key\n",
    "    if key == '16S_V4':\n",
    "        plot_title = f'16S rRNA (V4) Relative Abundance'\n",
    "\n",
    "    # Plot\n",
    "    ax = df_grouped.T.plot(kind='bar', stacked=True, figsize=(13, 10),\n",
    "                           width=0.8,  # Bars closer together\n",
    "                           color=[taxa_color_map.get(fam, '#ADD8E6') for fam in df_grouped.index])\n",
    "\n",
    "    plt.ylabel('Relative Abundance', fontsize=16)\n",
    "    plt.xlabel(' ')\n",
    "    plt.title(plot_title, fontsize=18)\n",
    "\n",
    "    # Set x-axis labels and sample size retained after rarefaction for V1-V3 and V4\n",
    "    if key == '16S_V4':\n",
    "        new_labels = ['Healthy skin','AD NL skin', 'AD L skin', 'Healthy anterior nares', 'AD anterior nares']\n",
    "\n",
    "    # Set the new x-tick labels\n",
    "    plt.xticks(ticks=range(len(new_labels)), labels=new_labels, rotation=45, ha='center', fontsize=16)\n",
    "\n",
    "    plt.legend(loc='center left', bbox_to_anchor=(1.0, 0.5), fontsize=14, title=taxa_level, title_fontsize=16)\n",
    "    plt.tight_layout()\n",
    "\n",
    "    plt.savefig(output_png_file, format='png', dpi=600)  # Save as png\n",
    "    plt.savefig(output_svg_file, format='svg')  # Save as svg\n",
    "    plt.show()\n",
    "\n",
    "    plt.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 225,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>900221</th>\n",
       "      <th>900570</th>\n",
       "      <th>900092</th>\n",
       "      <th>900466</th>\n",
       "      <th>9003932</th>\n",
       "      <th>900091</th>\n",
       "      <th>900556</th>\n",
       "      <th>900301</th>\n",
       "      <th>900245</th>\n",
       "      <th>900581</th>\n",
       "      <th>...</th>\n",
       "      <th>900547</th>\n",
       "      <th>900263</th>\n",
       "      <th>Ca008HNL</th>\n",
       "      <th>900081</th>\n",
       "      <th>900501</th>\n",
       "      <th>900279</th>\n",
       "      <th>900304</th>\n",
       "      <th>900580</th>\n",
       "      <th>900484</th>\n",
       "      <th>9003972</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>g__Streptococcus</th>\n",
       "      <td>0.000667</td>\n",
       "      <td>0.120454</td>\n",
       "      <td>0.973333</td>\n",
       "      <td>0.971667</td>\n",
       "      <td>0.700333</td>\n",
       "      <td>0.017391</td>\n",
       "      <td>0.632667</td>\n",
       "      <td>0.001667</td>\n",
       "      <td>0.006333</td>\n",
       "      <td>0.049000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.209333</td>\n",
       "      <td>0.153000</td>\n",
       "      <td>0.000333</td>\n",
       "      <td>0.242000</td>\n",
       "      <td>0.539513</td>\n",
       "      <td>0.021667</td>\n",
       "      <td>0.904667</td>\n",
       "      <td>0.934667</td>\n",
       "      <td>0.703000</td>\n",
       "      <td>0.225343</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>g__Staphylococcus</th>\n",
       "      <td>0.058333</td>\n",
       "      <td>0.236904</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.081271</td>\n",
       "      <td>0.001667</td>\n",
       "      <td>0.980667</td>\n",
       "      <td>0.017667</td>\n",
       "      <td>0.517333</td>\n",
       "      <td>...</td>\n",
       "      <td>0.004000</td>\n",
       "      <td>0.103000</td>\n",
       "      <td>0.977667</td>\n",
       "      <td>0.072333</td>\n",
       "      <td>0.058686</td>\n",
       "      <td>0.251667</td>\n",
       "      <td>0.006667</td>\n",
       "      <td>0.043333</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.127048</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>g__Haemophilus_D_734546</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000667</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.283333</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.252000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.008000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.750000</td>\n",
       "      <td>0.001667</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.154333</td>\n",
       "      <td>0.014672</td>\n",
       "      <td>0.015667</td>\n",
       "      <td>0.022667</td>\n",
       "      <td>0.000333</td>\n",
       "      <td>0.132333</td>\n",
       "      <td>0.015714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>g__Unknown</th>\n",
       "      <td>0.002333</td>\n",
       "      <td>0.175843</td>\n",
       "      <td>0.006667</td>\n",
       "      <td>0.000333</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.411371</td>\n",
       "      <td>0.022333</td>\n",
       "      <td>0.003000</td>\n",
       "      <td>0.280667</td>\n",
       "      <td>0.090667</td>\n",
       "      <td>...</td>\n",
       "      <td>0.011667</td>\n",
       "      <td>0.106667</td>\n",
       "      <td>0.003667</td>\n",
       "      <td>0.147000</td>\n",
       "      <td>0.093698</td>\n",
       "      <td>0.093000</td>\n",
       "      <td>0.008000</td>\n",
       "      <td>0.000333</td>\n",
       "      <td>0.007333</td>\n",
       "      <td>0.041792</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>g__Corynebacterium</th>\n",
       "      <td>0.002333</td>\n",
       "      <td>0.012346</td>\n",
       "      <td>0.015333</td>\n",
       "      <td>0.000333</td>\n",
       "      <td>0.006000</td>\n",
       "      <td>0.090301</td>\n",
       "      <td>0.036333</td>\n",
       "      <td>0.012333</td>\n",
       "      <td>0.170000</td>\n",
       "      <td>0.010000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.003000</td>\n",
       "      <td>0.132333</td>\n",
       "      <td>0.016333</td>\n",
       "      <td>0.027667</td>\n",
       "      <td>0.005335</td>\n",
       "      <td>0.182000</td>\n",
       "      <td>0.023333</td>\n",
       "      <td>0.014000</td>\n",
       "      <td>0.039000</td>\n",
       "      <td>0.015045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>g__Pseudomonas_K</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>g__Herbaspirillum</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>g__Marinomonas</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>g__Tetragenococcus</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>g__Burkholderia</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>475 rows × 254 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           900221    900570    900092    900466   9003932  \\\n",
       "g__Streptococcus         0.000667  0.120454  0.973333  0.971667  0.700333   \n",
       "g__Staphylococcus        0.058333  0.236904  0.000000  0.001000  0.000000   \n",
       "g__Haemophilus_D_734546  0.000000  0.000667  0.000000  0.000000  0.283333   \n",
       "g__Unknown               0.002333  0.175843  0.006667  0.000333  0.000000   \n",
       "g__Corynebacterium       0.002333  0.012346  0.015333  0.000333  0.006000   \n",
       "...                           ...       ...       ...       ...       ...   \n",
       "g__Pseudomonas_K         0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "g__Herbaspirillum        0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "g__Marinomonas           0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "g__Tetragenococcus       0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "g__Burkholderia          0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "\n",
       "                           900091    900556    900301    900245    900581  \\\n",
       "g__Streptococcus         0.017391  0.632667  0.001667  0.006333  0.049000   \n",
       "g__Staphylococcus        0.081271  0.001667  0.980667  0.017667  0.517333   \n",
       "g__Haemophilus_D_734546  0.000000  0.252000  0.000000  0.000000  0.008000   \n",
       "g__Unknown               0.411371  0.022333  0.003000  0.280667  0.090667   \n",
       "g__Corynebacterium       0.090301  0.036333  0.012333  0.170000  0.010000   \n",
       "...                           ...       ...       ...       ...       ...   \n",
       "g__Pseudomonas_K         0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "g__Herbaspirillum        0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "g__Marinomonas           0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "g__Tetragenococcus       0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "g__Burkholderia          0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "\n",
       "                         ...    900547    900263  Ca008HNL    900081  \\\n",
       "g__Streptococcus         ...  0.209333  0.153000  0.000333  0.242000   \n",
       "g__Staphylococcus        ...  0.004000  0.103000  0.977667  0.072333   \n",
       "g__Haemophilus_D_734546  ...  0.750000  0.001667  0.000000  0.154333   \n",
       "g__Unknown               ...  0.011667  0.106667  0.003667  0.147000   \n",
       "g__Corynebacterium       ...  0.003000  0.132333  0.016333  0.027667   \n",
       "...                      ...       ...       ...       ...       ...   \n",
       "g__Pseudomonas_K         ...  0.000000  0.000000  0.000000  0.000000   \n",
       "g__Herbaspirillum        ...  0.000000  0.000000  0.000000  0.000000   \n",
       "g__Marinomonas           ...  0.000000  0.000000  0.000000  0.000000   \n",
       "g__Tetragenococcus       ...  0.000000  0.000000  0.000000  0.000000   \n",
       "g__Burkholderia          ...  0.000000  0.000000  0.000000  0.000000   \n",
       "\n",
       "                           900501    900279    900304    900580    900484  \\\n",
       "g__Streptococcus         0.539513  0.021667  0.904667  0.934667  0.703000   \n",
       "g__Staphylococcus        0.058686  0.251667  0.006667  0.043333  0.000000   \n",
       "g__Haemophilus_D_734546  0.014672  0.015667  0.022667  0.000333  0.132333   \n",
       "g__Unknown               0.093698  0.093000  0.008000  0.000333  0.007333   \n",
       "g__Corynebacterium       0.005335  0.182000  0.023333  0.014000  0.039000   \n",
       "...                           ...       ...       ...       ...       ...   \n",
       "g__Pseudomonas_K         0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "g__Herbaspirillum        0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "g__Marinomonas           0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "g__Tetragenococcus       0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "g__Burkholderia          0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "\n",
       "                          9003972  \n",
       "g__Streptococcus         0.225343  \n",
       "g__Staphylococcus        0.127048  \n",
       "g__Haemophilus_D_734546  0.015714  \n",
       "g__Unknown               0.041792  \n",
       "g__Corynebacterium       0.015045  \n",
       "...                           ...  \n",
       "g__Pseudomonas_K         0.000000  \n",
       "g__Herbaspirillum        0.000000  \n",
       "g__Marinomonas           0.000000  \n",
       "g__Tetragenococcus       0.000000  \n",
       "g__Burkholderia          0.000000  \n",
       "\n",
       "[475 rows x 254 columns]"
      ]
     },
     "execution_count": 225,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "load_biom_table('../Data/Tables/Relative_Abundance_Tables/df_16S_filtered_feature_table_rare_Genus_relative_abundance.biom', metadata_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 226,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/gg/tjf8z7kd7351mxr7fbw1s6j00000gp/T/ipykernel_22113/2749318524.py:3: FutureWarning: DataFrame.groupby with axis=1 is deprecated. Do `frame.T.groupby(...)` without axis instead.\n",
      "  df_grouped = df.groupby(metadata[group_column], axis=1).mean()\n",
      "/var/folders/gg/tjf8z7kd7351mxr7fbw1s6j00000gp/T/ipykernel_22113/2749318524.py:39: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown\n",
      "  plt.show()\n"
     ]
    }
   ],
   "source": [
    "# Process both BIOM files\n",
    "global_taxa_color_map = {}  # Store colors assigned to families across both datasets\n",
    "for key, biom_path in biom_paths.items():\n",
    "    # Load and process the BIOM table\n",
    "    df = load_biom_table(biom_path, metadata_path)\n",
    "\n",
    "    # Keep only the columns in df that have matching entries in the metadata\n",
    "    df = df[df.columns.intersection(metadata['#sample-id'])]\n",
    "\n",
    "    df_top_15 = collapse_top_15(df)  # Now using the top 15 function\n",
    "    \n",
    "    # Define the output directory for the current key\n",
    "    output_dir = '../Plots/Analysis_figures/Relative_Abundance/'\n",
    "    os.makedirs(output_dir, exist_ok=True)  # Ensure the output directory exists\n",
    "    \n",
    "    # Merge with metadata\n",
    "    metadata_subset = metadata.set_index('#sample-id').loc[df.columns]  # Assuming #sample-id in metadata matches samples in BIOM table\n",
    "    \n",
    "    # Generate or retrieve colors for the top families in this dataset\n",
    "    global_taxa_color_map = get_taxa_colors(df_top_15.index, global_taxa_color_map)\n",
    "    \n",
    "    # Plot the results and save in the respective folder \n",
    "    plot_relative_abundance(df_top_15, metadata_subset, 'case_type', output_dir, key, global_taxa_color_map)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Shannon alpha diversity plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Additional Python packages for alpha diversity analyses\n",
    "from skbio.diversity import alpha_diversity\n",
    "from scipy.stats import mannwhitneyu"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 228,
   "metadata": {},
   "outputs": [],
   "source": [
    "def darken_color(color, amount=0.3):\n",
    "    \"\"\"\n",
    "    Darken a given color by a certain amount.\n",
    "    \n",
    "    Parameters:\n",
    "    - color: The base color (as a hex string or color name).\n",
    "    - amount: The amount to darken the color by (default: 0.3).\n",
    "    \n",
    "    Returns:\n",
    "    - A darkened color as a hex string.\n",
    "    \"\"\"\n",
    "    c = to_rgba(color)\n",
    "    return (c[0] * (1 - amount), c[1] * (1 - amount), c[2] * (1 - amount), c[3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 231,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_shannon_alpha_diversity_and_plot(biom_path, metadata_path, group_col, title_suffix):\n",
    "    # Load metadata as a DataFrame from the file path\n",
    "    metadata = pd.read_csv(metadata_path, sep='\\t')\n",
    "    metadata = metadata.set_index('#sample-id')\n",
    "    \n",
    "    # Load biom table\n",
    "    feature_table = load_biom_table(biom_path, metadata_path)\n",
    "\n",
    "    # Transpose feature_table so rows are sample IDs\n",
    "    feature_table = feature_table.transpose()\n",
    "\n",
    "    # Calculate Shannon diversity\n",
    "    shannon_values = alpha_diversity('shannon', feature_table.values, ids=feature_table.index)\n",
    "    \n",
    "    # Merge Shannon values with metadata\n",
    "    common_samples = metadata.index.intersection(feature_table.index)\n",
    "    metadata = metadata.loc[common_samples].copy()\n",
    "    metadata['Shannon'] = shannon_values.loc[common_samples]\n",
    "    \n",
    "    # Set the custom order for the groups: Healthy first, then Acne Non-lesional, then Acne Lesional\n",
    "    desired_order = ['control-nonlesional skin','case-nonlesional skin', 'case-lesional skin', 'control-anterior nares', 'case-anterior nares']\n",
    "\n",
    "    # Set the color palette for the groups in the correct order\n",
    "    palette = {\n",
    "        'control-nonlesional skin': '#3333B3',     # Dark Blue color for Healthy\n",
    "        'case-nonlesional skin': '#5cbccb',     # Blue color for AD Non-Lesional\n",
    "        'case-lesional skin': '#f16c52',       # Red color for AD Lesional\n",
    "        'control-anterior nares': '#008000',   # Green for healthy Nasal\n",
    "        'case-anterior nares': '#FFC0CB'        # Pink for AD Nasal\n",
    "    }\n",
    "\n",
    "    # Create a new column for severity category based on local lesion severity\n",
    "    # metadata['severity_category'] = pd.cut(metadata['age_months'],\n",
    "    #                                     bins=[0, 2, 4, 6],  # Low (1-2), Moderate (3-4), High (5-6)\n",
    "    #                                     labels=['low', 'moderate', 'high'])\n",
    "\n",
    "    # Define a custom palette for severity categories within Acne_L group\n",
    "    # severity_palette = {\n",
    "    #     'low': '#F1948A',      # Light red for low severity\n",
    "    #     'moderate': '#EC7063',  # Red for moderate severity\n",
    "    #     'high': '#C0392B'       # Dark red for high severity\n",
    "    # }\n",
    "\n",
    "    # Create a more rectangular plot (e.g., 8 inches wide and 12 inches tall)\n",
    "    plt.figure(figsize=(10, 10))\n",
    "\n",
    "    # Plot the Shannon diversity across the groups using a boxplot with custom colors and order\n",
    "    ax = sns.boxplot(x=group_col, y='Shannon', data=metadata, palette=palette, order=desired_order)\n",
    "\n",
    "    # Darken the colors for the strip plot manually\n",
    "    darker_palette = {key: darken_color(color) for key, color in palette.items()}\n",
    "\n",
    "    # Plot the general stripplot with darkened colors\n",
    "    sns.stripplot(x=group_col, y='Shannon', data=metadata, palette=darker_palette, jitter=True, dodge=False, ax=ax, linewidth=0.6, order=desired_order)\n",
    "\n",
    "    # Add a second stripplot specifically for Acne_L, coloring by severity\n",
    "    # sns.stripplot(x=metadata[metadata[group_col] == 'Acne_L'][group_col], \n",
    "    #               y=metadata[metadata[group_col] == 'Acne_L']['Shannon'], \n",
    "    #               hue=metadata[metadata[group_col] == 'Acne_L']['severity_category'],\n",
    "    #               palette=severity_palette, jitter=True, dodge=False, ax=ax, linewidth=0.6)\n",
    "\n",
    "    # Adjust the legend for the severity categories\n",
    "    handles, labels = ax.get_legend_handles_labels()\n",
    "    severity_handles = handles[-3:]  # Get the last three handles, which correspond to severity levels\n",
    "    severity_labels = ['Low (1-2)', 'Moderate (3-4)', 'High (5-6)']  # Rename the labels\n",
    "\n",
    "    # Move the legend outside the boxplot to the right\n",
    "    # ax.legend(severity_handles, severity_labels, title='Lesional Severity Score', \n",
    "    #         loc='center left', bbox_to_anchor=(1.0, 0.5), ncol=1, frameon=True, title_fontsize=16, fontsize=14)\n",
    "\n",
    "    # Add the title and labels\n",
    "    plt.title(f'16S rRNA ({title_suffix}) Alpha Diversity', fontsize=18)\n",
    "    plt.xlabel(' ')\n",
    "    plt.ylabel('Shannon Index', fontsize=16)\n",
    "\n",
    "    # Set x-axis labels and sample size retained after rarefaction for V1-V3 and V4\n",
    "    if key == '16S_V4':\n",
    "        new_labels = ['Healthy skin','AD NL skin', 'AD L skin', 'Healthy anterior nares', 'AD anterior nares']\n",
    "        \n",
    "    plt.xticks(ticks=range(len(new_labels)), labels=new_labels, rotation=45, ha='center', fontsize=16)\n",
    "\n",
    "    # Pairwise significance testing using Mann-Whitney U test\n",
    "    groups = desired_order\n",
    "    p_values = {}\n",
    "    \n",
    "    # Heights to draw the annotation lines∂\n",
    "    y_max = max(metadata['Shannon']) + 0.1\n",
    "    height_step = 0.15  # Height step between lines\n",
    "    \n",
    "    # Perform pairwise comparisons\n",
    "    for i, group1 in enumerate(groups):\n",
    "        for j, group2 in enumerate(groups):\n",
    "            if i < j:\n",
    "                # Get the Shannon values for each group\n",
    "                group1_values = metadata[metadata[group_col] == group1]['Shannon']\n",
    "                group2_values = metadata[metadata[group_col] == group2]['Shannon']\n",
    "                \n",
    "                # Perform Mann-Whitney U test\n",
    "                stat, p = mannwhitneyu(group1_values, group2_values, alternative='two-sided')\n",
    "                p_values[f'{group1} vs {group2}'] = p\n",
    "                \n",
    "                # Determine the significance label based on p-value thresholds\n",
    "                if p >= 0.05:\n",
    "                    label = f\"{p:.2e}\"\n",
    "                elif p < 0.001:\n",
    "                    label = '***  ' + f\"{p:.2e}\"\n",
    "                elif p < 0.01:\n",
    "                    label = '**  ' + f\"{p:.2e}\"\n",
    "                else:\n",
    "                    label = '*  ' + f\"{p:.2e}\"\n",
    "                \n",
    "                # Get x coordinates of the boxplots\n",
    "                x1, x2 = i, j\n",
    "                y = y_max + height_step  # Vertical position for the horizontal line\n",
    "                \n",
    "                # Draw horizontal line and annotate the significance label\n",
    "                plt.plot([x1, x1, x2, x2], [y, y + 0.01, y + 0.01, y], lw=1, color='black')\n",
    "                plt.text((x1 + x2) * 0.5, y + 0.02, label, ha='center', va='bottom', fontsize=12)\n",
    "                \n",
    "                # Update y_max for the next comparison\n",
    "                y_max += height_step + 0.05\n",
    "\n",
    "        # Save the figure\n",
    "        plt.savefig(f'../Plots/Analysis_figures/Diversity/{key}_{taxa_level}_Shannon_alpha-diversity.png', dpi=600, bbox_inches='tight', pad_inches=0.1)  # Save as png\n",
    "        plt.savefig(f'../Plots/Analysis_figures/Diversity/{key}_{taxa_level}_Shannon_alpha-diversity.svg')  # Save as svg\n",
    "\n",
    "    # Print pairwise p-values in scientific notation\n",
    "    print(\"Pairwise Mann-Whitney U test p-values:\")\n",
    "    for comparison, p_value in p_values.items():\n",
    "        print(f\"{comparison}: p-value = {p_value:.2e}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 232,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/gg/tjf8z7kd7351mxr7fbw1s6j00000gp/T/ipykernel_22113/3597302403.py:54: FutureWarning: Passing `palette` without assigning `hue` is deprecated.\n",
      "  sns.stripplot(x=group_col, y='Shannon', data=metadata, palette=darker_palette, jitter=True, dodge=False, ax=ax, linewidth=0.6, order=desired_order)\n",
      "/Users/yac027/mambaforge3/envs/qiime2-metagenome-2024.5/lib/python3.9/site-packages/seaborn/_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.\n",
      "  with pd.option_context('mode.use_inf_as_na', True):\n",
      "/Users/yac027/mambaforge3/envs/qiime2-metagenome-2024.5/lib/python3.9/site-packages/seaborn/_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.\n",
      "  with pd.option_context('mode.use_inf_as_na', True):\n",
      "/Users/yac027/mambaforge3/envs/qiime2-metagenome-2024.5/lib/python3.9/site-packages/seaborn/_oldcore.py:1075: FutureWarning: When grouping with a length-1 list-like, you will need to pass a length-1 tuple to get_group in a future version of pandas. Pass `(name,)` instead of `name` to silence this warning.\n",
      "  data_subset = grouped_data.get_group(pd_key)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Pairwise Mann-Whitney U test p-values:\n",
      "control-nonlesional skin vs case-nonlesional skin: p-value = 9.49e-01\n",
      "control-nonlesional skin vs case-lesional skin: p-value = 2.84e-01\n",
      "control-nonlesional skin vs control-anterior nares: p-value = 9.83e-09\n",
      "control-nonlesional skin vs case-anterior nares: p-value = 7.21e-08\n",
      "case-nonlesional skin vs case-lesional skin: p-value = 3.86e-01\n",
      "case-nonlesional skin vs control-anterior nares: p-value = 2.11e-08\n",
      "case-nonlesional skin vs case-anterior nares: p-value = 2.60e-07\n",
      "case-lesional skin vs control-anterior nares: p-value = 9.47e-07\n",
      "case-lesional skin vs case-anterior nares: p-value = 3.30e-06\n",
      "control-anterior nares vs case-anterior nares: p-value = 5.16e-01\n"
     ]
    }
   ],
   "source": [
    "# Plot Alpha Diversity plots for both V1-V3 and V4\n",
    "for key, biom_path in biom_paths.items():\n",
    "    calculate_shannon_alpha_diversity_and_plot(\n",
    "        biom_path=biom_path,\n",
    "        metadata_path=metadata_path,\n",
    "        group_col='case_type',\n",
    "        title_suffix='V4' if key == '16S_V4' else ''\n",
    "    )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
