{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "(A) Venn diagram showing the overlap in genera detected in paired nasal and skin microbiome samples across individuals. (B) Bar plot displaying the number of unique amplicon sequence variants (ASVs) assigned to each overlapping genus, highlighting the microbial diversity shared across sites."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import biom\n",
    "from biom import load_table\n",
    "import numpy as np\n",
    "from scipy.stats import pearsonr\n",
    "from matplotlib_venn import venn2\n",
    "import matplotlib.patches as mpatches\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PlateNumber</th>\n",
       "      <th>PlateLocation</th>\n",
       "      <th>i5</th>\n",
       "      <th>i5Sequence</th>\n",
       "      <th>i7</th>\n",
       "      <th>i7Sequence</th>\n",
       "      <th>identifier</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Plate ID</th>\n",
       "      <th>Well location</th>\n",
       "      <th>...</th>\n",
       "      <th>enrolment_date</th>\n",
       "      <th>enrolment_season</th>\n",
       "      <th>hiv_exposure</th>\n",
       "      <th>hiv_status</th>\n",
       "      <th>household_size</th>\n",
       "      <th>o_scorad</th>\n",
       "      <th>FWD_filepath</th>\n",
       "      <th>REV_filepath</th>\n",
       "      <th>group</th>\n",
       "      <th>microbiome_type</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#sample-id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Ca009STL</th>\n",
       "      <td>1</td>\n",
       "      <td>A1</td>\n",
       "      <td>SA501</td>\n",
       "      <td>ATCGTACG</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA501</td>\n",
       "      <td>CGAGAGTT-ATCGTACG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>A1</td>\n",
       "      <td>...</td>\n",
       "      <td>4/16/2015</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>4.0</td>\n",
       "      <td>40</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900221</th>\n",
       "      <td>1</td>\n",
       "      <td>B1</td>\n",
       "      <td>SA502</td>\n",
       "      <td>ACTATCTG</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA502</td>\n",
       "      <td>CGAGAGTT-ACTATCTG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>B1</td>\n",
       "      <td>...</td>\n",
       "      <td>8/11/2015</td>\n",
       "      <td>Winter</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>7.0</td>\n",
       "      <td>34</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca010EBL</th>\n",
       "      <td>1</td>\n",
       "      <td>C1</td>\n",
       "      <td>SA503</td>\n",
       "      <td>TAGCGAGT</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA503</td>\n",
       "      <td>CGAGAGTT-TAGCGAGT</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>C1</td>\n",
       "      <td>...</td>\n",
       "      <td>11/20/2014</td>\n",
       "      <td>Spring</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>7.0</td>\n",
       "      <td>21</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900460</th>\n",
       "      <td>1</td>\n",
       "      <td>D1</td>\n",
       "      <td>SA504</td>\n",
       "      <td>CTGCGTGT</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA504</td>\n",
       "      <td>CGAGAGTT-CTGCGTGT</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>D1</td>\n",
       "      <td>...</td>\n",
       "      <td>9/23/2015</td>\n",
       "      <td>Spring</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>40</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900051</th>\n",
       "      <td>1</td>\n",
       "      <td>E1</td>\n",
       "      <td>SA505</td>\n",
       "      <td>TCATCGAG</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA505</td>\n",
       "      <td>CGAGAGTT-TCATCGAG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>E1</td>\n",
       "      <td>...</td>\n",
       "      <td>4/21/2015</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>7.0</td>\n",
       "      <td>41</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca006ONL2</th>\n",
       "      <td>6</td>\n",
       "      <td>H1</td>\n",
       "      <td>SA508</td>\n",
       "      <td>GACACCGT</td>\n",
       "      <td>SB701</td>\n",
       "      <td>CTCGACTT</td>\n",
       "      <td>SB701SA508</td>\n",
       "      <td>CTCGACTT-GACACCGT</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>H1</td>\n",
       "      <td>...</td>\n",
       "      <td>3/25/2015</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>3.0</td>\n",
       "      <td>34</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca006ONNL</th>\n",
       "      <td>6</td>\n",
       "      <td>F2</td>\n",
       "      <td>SA506</td>\n",
       "      <td>CGTGAGTG</td>\n",
       "      <td>SB702</td>\n",
       "      <td>CGAAGTAT</td>\n",
       "      <td>SB702SA506</td>\n",
       "      <td>CGAAGTAT-CGTGAGTG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>F2</td>\n",
       "      <td>...</td>\n",
       "      <td>3/25/2015</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>3.0</td>\n",
       "      <td>34</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADNL</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca006ONNL2</th>\n",
       "      <td>6</td>\n",
       "      <td>H2</td>\n",
       "      <td>SA508</td>\n",
       "      <td>GACACCGT</td>\n",
       "      <td>SB702</td>\n",
       "      <td>CGAAGTAT</td>\n",
       "      <td>SB702SA508</td>\n",
       "      <td>CGAAGTAT-GACACCGT</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>H2</td>\n",
       "      <td>...</td>\n",
       "      <td>3/25/2015</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>3.0</td>\n",
       "      <td>34</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADNL</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca006ONPN</th>\n",
       "      <td>6</td>\n",
       "      <td>F3</td>\n",
       "      <td>SA506</td>\n",
       "      <td>CGTGAGTG</td>\n",
       "      <td>SB703</td>\n",
       "      <td>TAGCAGCT</td>\n",
       "      <td>SB703SA506</td>\n",
       "      <td>TAGCAGCT-CGTGAGTG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>F3</td>\n",
       "      <td>...</td>\n",
       "      <td>3/25/2015</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>3.0</td>\n",
       "      <td>34</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>nares-AD</td>\n",
       "      <td>nares</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca006ONPN2</th>\n",
       "      <td>6</td>\n",
       "      <td>H3</td>\n",
       "      <td>SA508</td>\n",
       "      <td>GACACCGT</td>\n",
       "      <td>SB703</td>\n",
       "      <td>TAGCAGCT</td>\n",
       "      <td>SB703SA508</td>\n",
       "      <td>TAGCAGCT-GACACCGT</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>H3</td>\n",
       "      <td>...</td>\n",
       "      <td>3/25/2015</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>3.0</td>\n",
       "      <td>34</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>nares-AD</td>\n",
       "      <td>nares</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>502 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            PlateNumber PlateLocation     i5 i5Sequence     i7 i7Sequence  \\\n",
       "#sample-id                                                                  \n",
       "Ca009STL              1            A1  SA501   ATCGTACG  SA701   CGAGAGTT   \n",
       "900221                1            B1  SA502   ACTATCTG  SA701   CGAGAGTT   \n",
       "Ca010EBL              1            C1  SA503   TAGCGAGT  SA701   CGAGAGTT   \n",
       "900460                1            D1  SA504   CTGCGTGT  SA701   CGAGAGTT   \n",
       "900051                1            E1  SA505   TCATCGAG  SA701   CGAGAGTT   \n",
       "...                 ...           ...    ...        ...    ...        ...   \n",
       "Ca006ONL2             6            H1  SA508   GACACCGT  SB701   CTCGACTT   \n",
       "Ca006ONNL             6            F2  SA506   CGTGAGTG  SB702   CGAAGTAT   \n",
       "Ca006ONNL2            6            H2  SA508   GACACCGT  SB702   CGAAGTAT   \n",
       "Ca006ONPN             6            F3  SA506   CGTGAGTG  SB703   TAGCAGCT   \n",
       "Ca006ONPN2            6            H3  SA508   GACACCGT  SB703   TAGCAGCT   \n",
       "\n",
       "            identifier           Sequence      Plate ID Well location  ...  \\\n",
       "#sample-id                                                             ...   \n",
       "Ca009STL    SA701SA501  CGAGAGTT-ATCGTACG  1.010000e+21            A1  ...   \n",
       "900221      SA701SA502  CGAGAGTT-ACTATCTG  1.010000e+21            B1  ...   \n",
       "Ca010EBL    SA701SA503  CGAGAGTT-TAGCGAGT  1.010000e+21            C1  ...   \n",
       "900460      SA701SA504  CGAGAGTT-CTGCGTGT  1.010000e+21            D1  ...   \n",
       "900051      SA701SA505  CGAGAGTT-TCATCGAG  1.010000e+21            E1  ...   \n",
       "...                ...                ...           ...           ...  ...   \n",
       "Ca006ONL2   SB701SA508  CTCGACTT-GACACCGT  1.010000e+21            H1  ...   \n",
       "Ca006ONNL   SB702SA506  CGAAGTAT-CGTGAGTG  1.010000e+21            F2  ...   \n",
       "Ca006ONNL2  SB702SA508  CGAAGTAT-GACACCGT  1.010000e+21            H2  ...   \n",
       "Ca006ONPN   SB703SA506  TAGCAGCT-CGTGAGTG  1.010000e+21            F3  ...   \n",
       "Ca006ONPN2  SB703SA508  TAGCAGCT-GACACCGT  1.010000e+21            H3  ...   \n",
       "\n",
       "            enrolment_date enrolment_season hiv_exposure hiv_status  \\\n",
       "#sample-id                                                            \n",
       "Ca009STL         4/16/2015          Autumn     Unexposed   negative   \n",
       "900221           8/11/2015           Winter    Unexposed   negative   \n",
       "Ca010EBL        11/20/2014           Spring    Unexposed   negative   \n",
       "900460           9/23/2015           Spring    Unexposed        NaN   \n",
       "900051           4/21/2015          Autumn     Unexposed   negative   \n",
       "...                    ...              ...          ...        ...   \n",
       "Ca006ONL2        3/25/2015           Autumn    Unexposed   negative   \n",
       "Ca006ONNL        3/25/2015           Autumn    Unexposed   negative   \n",
       "Ca006ONNL2       3/25/2015           Autumn    Unexposed   negative   \n",
       "Ca006ONPN        3/25/2015           Autumn    Unexposed   negative   \n",
       "Ca006ONPN2       3/25/2015           Autumn    Unexposed   negative   \n",
       "\n",
       "           household_size o_scorad  \\\n",
       "#sample-id                           \n",
       "Ca009STL              4.0       40   \n",
       "900221                7.0       34   \n",
       "Ca010EBL              7.0       21   \n",
       "900460                4.0       40   \n",
       "900051                7.0       41   \n",
       "...                   ...      ...   \n",
       "Ca006ONL2             3.0       34   \n",
       "Ca006ONNL             3.0       34   \n",
       "Ca006ONNL2            3.0       34   \n",
       "Ca006ONPN             3.0       34   \n",
       "Ca006ONPN2            3.0       34   \n",
       "\n",
       "                                                 FWD_filepath  \\\n",
       "#sample-id                                                      \n",
       "Ca009STL    /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "900221      /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "Ca010EBL    /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "900460      /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "900051      /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "...                                                       ...   \n",
       "Ca006ONL2   /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "Ca006ONNL   /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "Ca006ONNL2  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "Ca006ONPN   /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "Ca006ONPN2  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "\n",
       "                                                 REV_filepath      group  \\\n",
       "#sample-id                                                                 \n",
       "Ca009STL    /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   skin-ADL   \n",
       "900221      /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   skin-ADL   \n",
       "Ca010EBL    /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   skin-ADL   \n",
       "900460      /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   skin-ADL   \n",
       "900051      /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   skin-ADL   \n",
       "...                                                       ...        ...   \n",
       "Ca006ONL2   /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   skin-ADL   \n",
       "Ca006ONNL   /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...  skin-ADNL   \n",
       "Ca006ONNL2  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...  skin-ADNL   \n",
       "Ca006ONPN   /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   nares-AD   \n",
       "Ca006ONPN2  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   nares-AD   \n",
       "\n",
       "           microbiome_type  \n",
       "#sample-id                  \n",
       "Ca009STL              skin  \n",
       "900221                skin  \n",
       "Ca010EBL              skin  \n",
       "900460                skin  \n",
       "900051                skin  \n",
       "...                    ...  \n",
       "Ca006ONL2             skin  \n",
       "Ca006ONNL             skin  \n",
       "Ca006ONNL2            skin  \n",
       "Ca006ONPN            nares  \n",
       "Ca006ONPN2           nares  \n",
       "\n",
       "[502 rows x 33 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load the metadata\n",
    "metadata_path = '../Data/Metadata/updated_clean_ant_skin_metadata_microbiome_type.tab'\n",
    "metadata = pd.read_csv(metadata_path, sep='\\t')\n",
    "\n",
    "metadata['#sample-id'] = metadata['#sample-id'].str.replace('_', '')\n",
    "# Set Sample-ID as the index for the metadata dataframe \n",
    "metadata = metadata.set_index('#sample-id')\n",
    "\n",
    "\n",
    "# Create group column based on case_type to simplify group names\n",
    "metadata['group'] = metadata['case_type'].map({\n",
    "    'case-lesional skin': 'skin-ADL',\n",
    "    'case-nonlesional skin': 'skin-ADNL', \n",
    "    'control-nonlesional skin': 'skin-H',\n",
    "    'case-anterior nares': 'nares-AD',\n",
    "    'control-anterior nares': 'nares-H'\n",
    "})\n",
    "\n",
    "metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>g__Streptococcus_ASV-1</th>\n",
       "      <th>g__Streptococcus_ASV-2</th>\n",
       "      <th>g__Corynebacterium_ASV-1</th>\n",
       "      <th>g__Corynebacterium_ASV-3</th>\n",
       "      <th>g___ASV-3</th>\n",
       "      <th>g__Cutibacterium_ASV-1</th>\n",
       "      <th>g___ASV-18</th>\n",
       "      <th>g___ASV-28</th>\n",
       "      <th>g___ASV-26</th>\n",
       "      <th>g__Cutibacterium_ASV-2</th>\n",
       "      <th>...</th>\n",
       "      <th>g__Copromonas_ASV-2</th>\n",
       "      <th>g__Microvirga_ASV-1</th>\n",
       "      <th>g__Leptotrichia_A_993758_ASV-19</th>\n",
       "      <th>g___ASV-104</th>\n",
       "      <th>g__Leptotrichia_A_993758_ASV-14</th>\n",
       "      <th>g__Capnocytophaga_820688_ASV-7</th>\n",
       "      <th>g__UBA6175_ASV-2</th>\n",
       "      <th>g__Comamonas_F_589250_ASV-4</th>\n",
       "      <th>g__Bosea_ASV-2</th>\n",
       "      <th>g___ASV-140</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>900344</th>\n",
       "      <td>188.0</td>\n",
       "      <td>115.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900459</th>\n",
       "      <td>20.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900221</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900570</th>\n",
       "      <td>18.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900092</th>\n",
       "      <td>221.0</td>\n",
       "      <td>116.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9003972</th>\n",
       "      <td>45.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900097</th>\n",
       "      <td>3.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900498</th>\n",
       "      <td>5.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900276</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>63.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900406</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>502 rows × 798 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         g__Streptococcus_ASV-1  g__Streptococcus_ASV-2  \\\n",
       "900344                    188.0                   115.0   \n",
       "900459                     20.0                    37.0   \n",
       "900221                        0                       0   \n",
       "900570                     18.0                       0   \n",
       "900092                    221.0                   116.0   \n",
       "...                         ...                     ...   \n",
       "9003972                    45.0                    20.0   \n",
       "900097                      3.0                       0   \n",
       "900498                      5.0                     4.0   \n",
       "900276                        0                       0   \n",
       "900406                        0                       0   \n",
       "\n",
       "         g__Corynebacterium_ASV-1  g__Corynebacterium_ASV-3  g___ASV-3  \\\n",
       "900344                       23.0                      19.0        2.0   \n",
       "900459                          0                         0          0   \n",
       "900221                          0                         0          0   \n",
       "900570                          0                         0        1.0   \n",
       "900092                        3.0                       1.0          0   \n",
       "...                           ...                       ...        ...   \n",
       "9003972                       1.0                         0        1.0   \n",
       "900097                          0                         0          0   \n",
       "900498                          0                         0          0   \n",
       "900276                       13.0                         0          0   \n",
       "900406                          0                         0          0   \n",
       "\n",
       "         g__Cutibacterium_ASV-1  g___ASV-18  g___ASV-28  g___ASV-26  \\\n",
       "900344                      2.0         1.0           0           0   \n",
       "900459                        0           0           0           0   \n",
       "900221                        0           0           0           0   \n",
       "900570                        0           0           0           0   \n",
       "900092                        0           0           0         1.0   \n",
       "...                         ...         ...         ...         ...   \n",
       "9003972                    21.0           0           0         1.0   \n",
       "900097                      6.0           0           0           0   \n",
       "900498                     11.0           0         6.0           0   \n",
       "900276                     63.0           0           0           0   \n",
       "900406                        0           0           0           0   \n",
       "\n",
       "         g__Cutibacterium_ASV-2  ...  g__Copromonas_ASV-2  \\\n",
       "900344                        0  ...                    0   \n",
       "900459                        0  ...                    0   \n",
       "900221                        0  ...                    0   \n",
       "900570                        0  ...                    0   \n",
       "900092                        0  ...                    0   \n",
       "...                         ...  ...                  ...   \n",
       "9003972                    18.0  ...                    0   \n",
       "900097                      1.0  ...                    0   \n",
       "900498                     13.0  ...                    0   \n",
       "900276                     51.0  ...                    0   \n",
       "900406                        0  ...                    0   \n",
       "\n",
       "         g__Microvirga_ASV-1  g__Leptotrichia_A_993758_ASV-19  g___ASV-104  \\\n",
       "900344                     0                                0            0   \n",
       "900459                     0                                0            0   \n",
       "900221                     0                                0            0   \n",
       "900570                     0                                0            0   \n",
       "900092                     0                                0            0   \n",
       "...                      ...                              ...          ...   \n",
       "9003972                    0                                0            0   \n",
       "900097                     0                                0            0   \n",
       "900498                     0                                0            0   \n",
       "900276                     0                                0            0   \n",
       "900406                     0                                0            0   \n",
       "\n",
       "         g__Leptotrichia_A_993758_ASV-14  g__Capnocytophaga_820688_ASV-7  \\\n",
       "900344                                 0                               0   \n",
       "900459                                 0                               0   \n",
       "900221                                 0                               0   \n",
       "900570                                 0                               0   \n",
       "900092                                 0                               0   \n",
       "...                                  ...                             ...   \n",
       "9003972                                0                               0   \n",
       "900097                                 0                               0   \n",
       "900498                                 0                               0   \n",
       "900276                                 0                               0   \n",
       "900406                                 0                               0   \n",
       "\n",
       "         g__UBA6175_ASV-2  g__Comamonas_F_589250_ASV-4  g__Bosea_ASV-2  \\\n",
       "900344                  0                            0               0   \n",
       "900459                  0                            0               0   \n",
       "900221                  0                            0               0   \n",
       "900570                  0                            0               0   \n",
       "900092                  0                            0               0   \n",
       "...                   ...                          ...             ...   \n",
       "9003972                 0                            0               0   \n",
       "900097                  0                            0               0   \n",
       "900498                  0                            0               0   \n",
       "900276                  0                            0               0   \n",
       "900406                  0                            0               0   \n",
       "\n",
       "         g___ASV-140  \n",
       "900344             0  \n",
       "900459             0  \n",
       "900221             0  \n",
       "900570             0  \n",
       "900092             0  \n",
       "...              ...  \n",
       "9003972          1.0  \n",
       "900097             0  \n",
       "900498             0  \n",
       "900276             0  \n",
       "900406             0  \n",
       "\n",
       "[502 rows x 798 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read in table at collapsed genera  level\n",
    "biom_path = '../Data/Tables/Absolute_Abundance_Tables/209766_filtered_by_prevalence_1pct_rare_Genus-ASV-non-collapse.biom'\n",
    "# biom_path = '../Data/Tables/Absolute_Abundance_Tables/feature_table_with_tax_labels_Genus.biom'\n",
    "# biom_path = '../Data/Tables/Absolute_Abundance_Tables/209766_filtered_by_prevalence_0pct_rare_Genus.biom'\n",
    "biom_tbl = load_table(biom_path)\n",
    "df = pd.DataFrame(biom_tbl.to_dataframe().T)\n",
    "\n",
    "# delete the prefix from the index\n",
    "df.index = df.index.str.replace('15564.', '')\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>g__Streptococcus_ASV-1</th>\n",
       "      <th>g__Streptococcus_ASV-2</th>\n",
       "      <th>g__Corynebacterium_ASV-1</th>\n",
       "      <th>g__Corynebacterium_ASV-3</th>\n",
       "      <th>g___ASV-3</th>\n",
       "      <th>g__Cutibacterium_ASV-1</th>\n",
       "      <th>g___ASV-18</th>\n",
       "      <th>g___ASV-28</th>\n",
       "      <th>g___ASV-26</th>\n",
       "      <th>g__Cutibacterium_ASV-2</th>\n",
       "      <th>...</th>\n",
       "      <th>g___ASV-104</th>\n",
       "      <th>g__Leptotrichia_A_993758_ASV-14</th>\n",
       "      <th>g__Capnocytophaga_820688_ASV-7</th>\n",
       "      <th>g__UBA6175_ASV-2</th>\n",
       "      <th>g__Comamonas_F_589250_ASV-4</th>\n",
       "      <th>g__Bosea_ASV-2</th>\n",
       "      <th>g___ASV-140</th>\n",
       "      <th>pid</th>\n",
       "      <th>group</th>\n",
       "      <th>microbiome_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>900344</th>\n",
       "      <td>188.0</td>\n",
       "      <td>115.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Co-130-MM</td>\n",
       "      <td>skin-H</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900459</th>\n",
       "      <td>20.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Ca-145-LM</td>\n",
       "      <td>nares-AD</td>\n",
       "      <td>nares</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900221</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Ca-101-ID</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900570</th>\n",
       "      <td>18.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Ca-046-MT</td>\n",
       "      <td>skin-ADNL</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900092</th>\n",
       "      <td>221.0</td>\n",
       "      <td>116.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Ca-023-EJ</td>\n",
       "      <td>nares-AD</td>\n",
       "      <td>nares</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9003972</th>\n",
       "      <td>45.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Ca-125-IM</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900097</th>\n",
       "      <td>3.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Ca-025-AC</td>\n",
       "      <td>skin-ADNL</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900498</th>\n",
       "      <td>5.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Ca-157-LM</td>\n",
       "      <td>skin-ADNL</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900276</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>63.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Ca-111-IN</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900406</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Ca-127-OM</td>\n",
       "      <td>skin-ADNL</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>502 rows × 801 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         g__Streptococcus_ASV-1  g__Streptococcus_ASV-2  \\\n",
       "900344                    188.0                   115.0   \n",
       "900459                     20.0                    37.0   \n",
       "900221                        0                       0   \n",
       "900570                     18.0                       0   \n",
       "900092                    221.0                   116.0   \n",
       "...                         ...                     ...   \n",
       "9003972                    45.0                    20.0   \n",
       "900097                      3.0                       0   \n",
       "900498                      5.0                     4.0   \n",
       "900276                        0                       0   \n",
       "900406                        0                       0   \n",
       "\n",
       "         g__Corynebacterium_ASV-1  g__Corynebacterium_ASV-3  g___ASV-3  \\\n",
       "900344                       23.0                      19.0        2.0   \n",
       "900459                          0                         0          0   \n",
       "900221                          0                         0          0   \n",
       "900570                          0                         0        1.0   \n",
       "900092                        3.0                       1.0          0   \n",
       "...                           ...                       ...        ...   \n",
       "9003972                       1.0                         0        1.0   \n",
       "900097                          0                         0          0   \n",
       "900498                          0                         0          0   \n",
       "900276                       13.0                         0          0   \n",
       "900406                          0                         0          0   \n",
       "\n",
       "         g__Cutibacterium_ASV-1  g___ASV-18  g___ASV-28  g___ASV-26  \\\n",
       "900344                      2.0         1.0           0           0   \n",
       "900459                        0           0           0           0   \n",
       "900221                        0           0           0           0   \n",
       "900570                        0           0           0           0   \n",
       "900092                        0           0           0         1.0   \n",
       "...                         ...         ...         ...         ...   \n",
       "9003972                    21.0           0           0         1.0   \n",
       "900097                      6.0           0           0           0   \n",
       "900498                     11.0           0         6.0           0   \n",
       "900276                     63.0           0           0           0   \n",
       "900406                        0           0           0           0   \n",
       "\n",
       "         g__Cutibacterium_ASV-2  ...  g___ASV-104  \\\n",
       "900344                        0  ...            0   \n",
       "900459                        0  ...            0   \n",
       "900221                        0  ...            0   \n",
       "900570                        0  ...            0   \n",
       "900092                        0  ...            0   \n",
       "...                         ...  ...          ...   \n",
       "9003972                    18.0  ...            0   \n",
       "900097                      1.0  ...            0   \n",
       "900498                     13.0  ...            0   \n",
       "900276                     51.0  ...            0   \n",
       "900406                        0  ...            0   \n",
       "\n",
       "         g__Leptotrichia_A_993758_ASV-14  g__Capnocytophaga_820688_ASV-7  \\\n",
       "900344                                 0                               0   \n",
       "900459                                 0                               0   \n",
       "900221                                 0                               0   \n",
       "900570                                 0                               0   \n",
       "900092                                 0                               0   \n",
       "...                                  ...                             ...   \n",
       "9003972                                0                               0   \n",
       "900097                                 0                               0   \n",
       "900498                                 0                               0   \n",
       "900276                                 0                               0   \n",
       "900406                                 0                               0   \n",
       "\n",
       "         g__UBA6175_ASV-2  g__Comamonas_F_589250_ASV-4  g__Bosea_ASV-2  \\\n",
       "900344                  0                            0               0   \n",
       "900459                  0                            0               0   \n",
       "900221                  0                            0               0   \n",
       "900570                  0                            0               0   \n",
       "900092                  0                            0               0   \n",
       "...                   ...                          ...             ...   \n",
       "9003972                 0                            0               0   \n",
       "900097                  0                            0               0   \n",
       "900498                  0                            0               0   \n",
       "900276                  0                            0               0   \n",
       "900406                  0                            0               0   \n",
       "\n",
       "         g___ASV-140        pid      group  microbiome_type  \n",
       "900344             0  Co-130-MM     skin-H             skin  \n",
       "900459             0  Ca-145-LM   nares-AD            nares  \n",
       "900221             0  Ca-101-ID   skin-ADL             skin  \n",
       "900570             0  Ca-046-MT  skin-ADNL             skin  \n",
       "900092             0  Ca-023-EJ   nares-AD            nares  \n",
       "...              ...        ...        ...              ...  \n",
       "9003972          1.0  Ca-125-IM   skin-ADL             skin  \n",
       "900097             0  Ca-025-AC  skin-ADNL             skin  \n",
       "900498             0  Ca-157-LM  skin-ADNL             skin  \n",
       "900276             0  Ca-111-IN   skin-ADL             skin  \n",
       "900406             0  Ca-127-OM  skin-ADNL             skin  \n",
       "\n",
       "[502 rows x 801 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Map pid, group, and microbiome_type from metadata to df based on matching indexes\n",
    "df['pid'] = metadata.loc[df.index, 'pid']\n",
    "df['group'] = metadata.loc[df.index, 'group']\n",
    "df['microbiome_type'] = metadata.loc[df.index, 'microbiome_type']\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of unique individuals: 197\n"
     ]
    }
   ],
   "source": [
    "print(f'Number of unique individuals: ' + str(df['pid'].nunique()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Get taxa columns (all except pid and microbiome_type)\n",
    "# taxa_cols = df.columns[:-2]\n",
    "\n",
    "# # Convert abundance table to presence/absence\n",
    "# presence_absence = df[taxa_cols] > 0\n",
    "# presence_absence\n",
    "# # Calculate percentage of TRUE values for each taxon\n",
    "# presence_percentages = (presence_absence.sum() / len(presence_absence)) * 100\n",
    "\n",
    "# # Get taxa that are present in at least 10% of samples\n",
    "# features_10_prev = presence_percentages[presence_percentages >= .2].index.tolist()\n",
    "# print(f'Number of taxa present in at least 10% of samples: {len(features_10_prev)}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# presence_absence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Calculate percentage of TRUE values for each column\n",
    "# percent_true = (presence_absence.sum() / len(presence_absence)) * 100\n",
    "# percent_true.to_csv('percent_true.csv')\n",
    "# # Count columns with >= 10% TRUE values \n",
    "# # num_cols_10pct = (percent_true >= 10).sum()\n",
    "\n",
    "# # print(f'Number of taxa present in at least 10% of samples: {num_cols_10pct}')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Filter taxa columns to only include features with >=10% prevalence\n",
    "# ### note: would be good to check how confident these taxa assignments were (should be 100% due to gg2 mapping)\n",
    "# ### note: probably need to check confidence of ASV assignments\n",
    "# taxa_cols_filtered = df[features_10_prev]\n",
    "# taxa_cols_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of taxa unique to skin: 266\n",
      "Number of taxa unique to nares: 16\n",
      "Number of taxa shared between skin and nares: 505\n",
      "Number of unique individuals: 187\n"
     ]
    }
   ],
   "source": [
    "# Get taxa columns (all except pid, groupmicrobiome_type)\n",
    "taxa_cols = df.columns[:-3]\n",
    "\n",
    "\n",
    "# # Filter taxa_cols_filtered to keep only prevalent taxa + metadata\n",
    "# df_filtered = taxa_cols_filtered[features_10_prev + ['pid', 'microbiome_type']]\n",
    "\n",
    "# Split into skin and nares samples\n",
    "skin_samples = df[df['microbiome_type'] == 'skin']\n",
    "nares_samples = df[df['microbiome_type'] == 'nares']\n",
    "\n",
    "# Find individuals with both skin and nares samples\n",
    "shared_pids = set(skin_samples['pid']) & set(nares_samples['pid'])\n",
    "\n",
    "# Initialize sets for taxa present in each body site\n",
    "skin_taxa = set()\n",
    "nares_taxa = set()\n",
    "overlapping_taxa = set()\n",
    "\n",
    "# For each individual with both sample types\n",
    "for pid in shared_pids:\n",
    "    # Get samples for this individual\n",
    "    skin_sample = skin_samples[skin_samples['pid'] == pid]\n",
    "    nares_sample = nares_samples[nares_samples['pid'] == pid]\n",
    "    \n",
    "    # Find taxa present (abundance > 0) in each sample type\n",
    "    skin_present = set(taxa_cols[skin_sample[taxa_cols].gt(0).any()])\n",
    "    nares_present = set(taxa_cols[nares_sample[taxa_cols].gt(0).any()])\n",
    "    \n",
    "    # Add to overall sets\n",
    "    skin_taxa.update(skin_present)\n",
    "    nares_taxa.update(nares_present)\n",
    "    # Add taxa present in both sites for this individual\n",
    "    overlapping_taxa.update(skin_present & nares_present)\n",
    "\n",
    "# Plot Venn diagram\n",
    "plt.figure(figsize=(6, 6))\n",
    "v = venn2(\n",
    "    [skin_taxa, nares_taxa],\n",
    "    set_labels=('', ''),\n",
    "    set_colors=('#2a00ff', '#ffa501'),\n",
    "    alpha=0.5\n",
    ")\n",
    "\n",
    "# Adjust font size for subset labels\n",
    "for label in v.subset_labels:\n",
    "    if label:\n",
    "        label.set_fontsize(16)\n",
    "\n",
    "# Customize circle outlines to match fill color\n",
    "region_colors = {\n",
    "    '10': '#2a00ff',    # Skin only\n",
    "    '01': '#ffa501',     # Nares only\n",
    "    '11': '#955280'   # Overlap\n",
    "}\n",
    "for region_id, color in region_colors.items():\n",
    "    patch = v.get_patch_by_id(region_id)\n",
    "    if patch:\n",
    "        patch.set_linewidth(2)\n",
    "        patch.set_edgecolor(color)\n",
    "\n",
    "# Create matching legend\n",
    "# Create matching legend patches (colors match Venn regions)\n",
    "handles = [\n",
    "    mpatches.Patch(color='#2a00ff', alpha=0.5, label='Skin only'),\n",
    "    mpatches.Patch(color='#955280', alpha=0.5, label='Both'), \n",
    "    mpatches.Patch(color='#ffa501', alpha=0.5, label='Nares only')\n",
    "    \n",
    "]\n",
    "\n",
    "# Add legend with specified font size\n",
    "plt.legend(\n",
    "    handles=handles,\n",
    "    loc='lower center',\n",
    "    bbox_to_anchor=(0.5, -0.05),\n",
    "    ncol=3,\n",
    "    frameon=False,\n",
    "    fontsize=16  # Adjust font size here\n",
    ")\n",
    "\n",
    "plt.suptitle('Unique ASV Overlap between Skin and Nares', fontsize=18)\n",
    "plt.title('(Paired Per-Individual and (>1% Sample Prevalence))', fontsize=16)\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.savefig('../Plots/Analysis_figures/Individual_Analyses/skin-nares_Venn_individual.png', dpi=600)\n",
    "\n",
    "# Print stats\n",
    "print(f\"Number of taxa unique to skin: {len(skin_taxa - nares_taxa)}\")\n",
    "print(f\"Number of taxa unique to nares: {len(nares_taxa - skin_taxa)}\")\n",
    "print(f\"Number of taxa shared between skin and nares: {len(skin_taxa & nares_taxa)}\")\n",
    "print(f\"Number of unique individuals: {len(shared_pids)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AD Individuals:\n",
      " - Skin only: 247\n",
      " - Nares only: 52\n",
      " - Shared: 381\n",
      "\n",
      "Healthy Individuals:\n",
      " - Skin only: 343\n",
      " - Nares only: 63\n",
      " - Shared: 260\n",
      "Number of unique individuals: 187\n"
     ]
    }
   ],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from matplotlib_venn import venn2\n",
    "import matplotlib.patches as mpatches\n",
    "\n",
    "# --- STEP 1: Prepare ---\n",
    "\n",
    "# Only keep numeric columns (i.e., ASV features)\n",
    "taxa_cols = df.select_dtypes(include=[np.number]).columns.tolist()\n",
    "\n",
    "\n",
    "# Split into skin and nares samples\n",
    "skin_samples = df[df['microbiome_type'] == 'skin']\n",
    "nares_samples = df[df['microbiome_type'] == 'nares']\n",
    "\n",
    "# --- STEP 2: Separate by disease group ---\n",
    "\n",
    "# AD individuals\n",
    "skin_ADL_samples = skin_samples[skin_samples['group'] == 'skin-ADL']\n",
    "nares_AD_samples = nares_samples[nares_samples['group'] == 'nares-AD']\n",
    "\n",
    "# Healthy individuals\n",
    "skin_H_samples = skin_samples[skin_samples['group'] == 'skin-H']\n",
    "nares_H_samples = nares_samples[nares_samples['group'] == 'nares-H']\n",
    "\n",
    "# --- STEP 3: Find taxa for each group ---\n",
    "\n",
    "# For AD\n",
    "skin_present_AD = skin_ADL_samples[taxa_cols].gt(0).any()\n",
    "nares_present_AD = nares_AD_samples[taxa_cols].gt(0).any()\n",
    "\n",
    "skin_taxa_AD = set(skin_present_AD[skin_present_AD].index)\n",
    "nares_taxa_AD = set(nares_present_AD[nares_present_AD].index)\n",
    "\n",
    "# For Healthy\n",
    "skin_present_H = skin_H_samples[taxa_cols].gt(0).any()\n",
    "nares_present_H = nares_H_samples[taxa_cols].gt(0).any()\n",
    "\n",
    "skin_taxa_H = set(skin_present_H[skin_present_H].index)\n",
    "nares_taxa_H = set(nares_present_H[nares_present_H].index)\n",
    "\n",
    "\n",
    "# --- STEP 4: Plot Venn diagrams ---\n",
    "\n",
    "fig, axes = plt.subplots(1, 2, figsize=(12, 6))\n",
    "\n",
    "# Colors\n",
    "skin_color = '#2a00ff'\n",
    "nares_color = '#ffa501'\n",
    "overlap_color = '#955280'\n",
    "\n",
    "# --- Venn for AD Individuals ---\n",
    "plt.sca(axes[0])  # set current axis\n",
    "v_AD = venn2(\n",
    "    [skin_taxa_AD, nares_taxa_AD],\n",
    "    set_labels=('', ''),\n",
    "    set_colors=(skin_color, nares_color),\n",
    "    alpha=0.5\n",
    ")\n",
    "\n",
    "# Customize font sizes\n",
    "for label in v_AD.subset_labels:\n",
    "    if label:\n",
    "        label.set_fontsize(16)\n",
    "\n",
    "# Customize circle outlines\n",
    "region_colors = {\n",
    "    '10': skin_color,\n",
    "    '01': nares_color,\n",
    "    '11': overlap_color\n",
    "}\n",
    "for region_id, color in region_colors.items():\n",
    "    patch = v_AD.get_patch_by_id(region_id)\n",
    "    if patch:\n",
    "        patch.set_linewidth(2)\n",
    "        patch.set_edgecolor(color)\n",
    "\n",
    "# Title for AD\n",
    "n_AD = skin_ADL_samples['pid'].nunique()  # Number of unique AD individuals\n",
    "# axes[0].set_title(f'AD Individuals (n={n_AD})', fontsize=18)\n",
    "axes[0].set_title(f'AD Individuals (n=107)', fontsize=20, y= 0.94)\n",
    "\n",
    "# --- Venn for Healthy Individuals ---\n",
    "plt.sca(axes[1])  # set current axis\n",
    "v_H = venn2(\n",
    "    [skin_taxa_H, nares_taxa_H],\n",
    "    set_labels=('', ''),\n",
    "    set_colors=(skin_color, nares_color),\n",
    "    alpha=0.5\n",
    ")\n",
    "\n",
    "for label in v_H.subset_labels:\n",
    "    if label:\n",
    "        label.set_fontsize(16)\n",
    "\n",
    "for region_id, color in region_colors.items():\n",
    "    patch = v_H.get_patch_by_id(region_id)\n",
    "    if patch:\n",
    "        patch.set_linewidth(2)\n",
    "        patch.set_edgecolor(color)\n",
    "\n",
    "# Title for Healthy\n",
    "n_H = skin_H_samples['pid'].nunique()     # Number of unique Healthy individuals\n",
    "# axes[1].set_title(f'Healthy Individuals (n={n_H})', fontsize=18)\n",
    "axes[1].set_title(f'Healthy Individuals (n=88)', fontsize=20, y= 0.94)\n",
    "\n",
    "# --- Add Legend ---\n",
    "handles = [\n",
    "    mpatches.Patch(color=skin_color, alpha=0.5, label='Skin only'),\n",
    "    mpatches.Patch(color=overlap_color, alpha=0.5, label='Both'),\n",
    "    mpatches.Patch(color=nares_color, alpha=0.5, label='Nares only')\n",
    "]\n",
    "\n",
    "fig.legend(\n",
    "    handles=handles,\n",
    "    loc='lower center',\n",
    "    bbox_to_anchor=(0.5, 0),\n",
    "    ncol=3,\n",
    "    frameon=False,\n",
    "    fontsize=16\n",
    ")\n",
    "\n",
    "\n",
    "# --- Final layout tweaks ---\n",
    "plt.tight_layout()\n",
    "plt.subplots_adjust(top=0.9, bottom=0.1)  # Shift top lower and bottom higher\n",
    "\n",
    "# --- Supertitle and subtitle ---\n",
    "fig.suptitle('Paired Per-Individual ASV Overlap Between Skin and Nares by Disease Status', fontsize=20, y=0.99)\n",
    "# fig.text(\n",
    "#     0.5, 0.91,  # x=centered, y=slightly lower\n",
    "#     'Paired Per-Individual and (>1% Sample Prevalence)',\n",
    "#     ha='center',\n",
    "#     fontsize=14\n",
    "# )\n",
    "\n",
    "plt.savefig('../Plots/Analysis_figures/Individual_Analyses/skin-nares_Venn_AD_vs_H.png', dpi=600)\n",
    "\n",
    "# --- Print numbers ---\n",
    "print(\"AD Individuals:\")\n",
    "print(f\" - Skin only: {len(skin_taxa_AD - nares_taxa_AD)}\")\n",
    "print(f\" - Nares only: {len(nares_taxa_AD - skin_taxa_AD)}\")\n",
    "print(f\" - Shared: {len(skin_taxa_AD & nares_taxa_AD)}\\n\")\n",
    "\n",
    "print(\"Healthy Individuals:\")\n",
    "print(f\" - Skin only: {len(skin_taxa_H - nares_taxa_H)}\")\n",
    "print(f\" - Nares only: {len(nares_taxa_H - skin_taxa_H)}\")\n",
    "print(f\" - Shared: {len(skin_taxa_H & nares_taxa_H)}\")\n",
    "print(f\"Number of unique individuals: {len(shared_pids)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Taxa shared in skin & nares of AD but not in healthy: 190\n",
      "{'g__Carnobacterium_A_320743_ASV-1', 'g___ASV-92', 'g__Rhodoferax_A_585629_ASV-1', 'g__Pauljensenia_ASV-5', 'g__Leptotrichia_A_993758_ASV-5', 'g___ASV-64', 'g__Alloprevotella_ASV-2', 'g__Bacteroides_H_ASV-3', 'g__Corynebacterium_ASV-18', 'g___ASV-86', 'g__Prevotella_ASV-8', 'g__Pseudomonas_E_647464_ASV-1', 'g__Streptococcus_ASV-10', 'g__Exiguobacterium_A_ASV-3', 'g___ASV-139', 'g__Alloprevotella_ASV-3', 'g__Abiotrophia_ASV-2', 'g__Mesorhizobium_F_498388_ASV-2', 'g__Acinetobacter_ASV-8', 'g__Prevotella_ASV-41', 'g__Corynebacterium_ASV-20', 'g__Sphingomonas_L_486704_ASV-3', 'g__Mucilaginibacter_A_ASV-1', 'g___ASV-24', 'g__Microbacterium_A_383184_ASV-1', 'g__Mediterraneibacter_A_155507_ASV-1', 'g__Dorea_A_ASV-2', 'g___ASV-90', 'g__Brevibacterium_ASV-4', 'g__Veillonella_A_ASV-8', 'g___ASV-114', 'g__Lautropia_ASV-1', 'g__Actinomyces_ASV-4', 'g__Anaerococcus_ASV-9', 'g__Paracoccus_ASV-2', 'g__Paracoccus_ASV-6', 'g__Gemmiger_A_73129_ASV-4', 'g__Lachnoclostridium_A_130679_ASV-1', 'g___ASV-76', 'g___ASV-67', 'g__Staphylococcus_ASV-5', 'g__Faecalibacterium_ASV-6', 'g__Friedmanniella_ASV-1', 'g___ASV-37', 'g__Comamonas_F_589250_ASV-2', 'g___ASV-74', 'g__Oribacterium_ASV-1', 'g__Massilia_ASV-7', 'g__Dietzia_ASV-1', 'g__Shewanella_ASV-2', 'g__Clostridium_P_ASV-1', 'g___ASV-19', 'g__Rothia_ASV-6', 'g__Halomonas_E_640244_ASV-1', 'g__Brevilactibacter_ASV-1', 'g__Roseomonas_A_507160_ASV-1', 'g__Gemmiger_A_73129_ASV-2', 'g__Massilia_ASV-1', 'g__Devosia_A_502124_ASV-1', 'g__Pseudoxanthomonas_A_615337_ASV-1', 'g__Leptotrichia_A_993641_ASV-2', 'g__Blautia_A_141781_ASV-2', 'g__Acinetobacter_ASV-6', 'g__Brevundimonas_ASV-4', 'g__Bariatricus_ASV-1', 'g__Sphingobium_A_485959_ASV-1', 'g__Faecalibacterium_ASV-4', 'g__Cutibacterium_ASV-3', 'g__Pseudomonas_E_647464_ASV-2', 'g__Pauljensenia_ASV-2', 'g__Collinsella_ASV-1', 'g__Alloprevotella_ASV-1', 'g__Corynebacterium_ASV-15', 'g___ASV-77', 'g__Prevotella_ASV-34', 'g__Pseudomonas_A_ASV-1', 'g__Prevotella_ASV-17', 'g___ASV-132', 'g__Bifidobacterium_388775_ASV-4', 'g___ASV-43', 'g___ASV-82', 'g__Skermanella_ASV-1', 'g__Lancefieldella_ASV-2', 'g__Prevotella_ASV-45', 'g__Pauljensenia_ASV-3', 'g___ASV-102', 'g__Prevotella_ASV-27', 'g__Catenibacterium_ASV-2', 'g___ASV-10', 'g__Prevotella_ASV-20', 'g__Capnocytophaga_820688_ASV-7', 'g__Arthrobacter_D_ASV-2', 'g___ASV-44', 'g__Blautia_A_141781_ASV-5', 'g__Aliicoccus_ASV-1', 'g__Luteimonas_C_615545_ASV-1', 'g__Anaerobutyricum_ASV-2', 'g__Clostridium_Q_135822_ASV-1', 'g__Lactobacillus_ASV-1', 'g__Prevotella_ASV-12', 'g___ASV-66', 'g__Lactobacillus_ASV-5', 'g__Fenollaria_ASV-1', 'g___ASV-33', 'g___ASV-45', 'g__Campylobacter_A_ASV-3', 'g___ASV-21', 'g__Prevotella_ASV-33', 'g__Acinetobacter_ASV-16', 'g__Escherichia_710834_ASV-2', 'g__Sphingobacterium_ASV-3', 'g___ASV-110', 'g__Mycobacterium_ASV-2', 'g__Arthrobacter_D_ASV-1', 'g__Streptococcus_ASV-8', 'g__Lancefieldella_ASV-1', 'g__CAG-217_ASV-1', 'g__Leptotrichia_A_993758_ASV-16', 'g__Massilia_ASV-2', 'g__Stenotrophomonas_A_615274_ASV-2', 'g__Leptotrichia_A_993758_ASV-9', 'g__Enterocloster_ASV-2', 'g__Lactococcus_A_346120_ASV-2', 'g__Pseudomonas_E_647464_ASV-6', 'g__Finegoldia_ASV-1', 'g__Pseudomonas_E_647464_ASV-4', 'g__Psychrobacter_ASV-3', 'g__Campylobacter_A_ASV-4', 'g__Acinetobacter_ASV-11', 'g__Streptococcus_ASV-9', 'g__Capnocytophaga_820688_ASV-6', 'g__Capnocytophaga_820688_ASV-5', 'g___ASV-36', 'g__Skermanella_ASV-2', 'g__Aggregatibacter_736122_ASV-1', 'g__Cloacibacterium_ASV-3', 'g__Prevotella_ASV-38', 'g__Acinetobacter_ASV-12', 'g__Stomatobaculum_ASV-1', 'g___ASV-50', 'g__Streptococcus_ASV-7', 'g__Peptoniphilus_A_ASV-1', 'g__Citricoccus_384863_ASV-1', 'g__Stenotrophomonas_A_615274_ASV-1', 'g___ASV-34', 'g__Prevotella_ASV-10', 'g__Anaerococcus_ASV-3', 'g__Acinetobacter_ASV-14', 'g__Lactobacillus_ASV-6', 'g__Centipeda_ASV-1', 'g__Pleomorphomonas_ASV-2', 'g__Leptotrichia_A_993758_ASV-7', 'g__Peptoniphilus_A_ASV-4', 'g__Roseburia_ASV-3', 'g__Gemmiger_A_73129_ASV-1', 'g__Prevotella_ASV-16', 'g___ASV-140', 'g__Agathobacter_164117_ASV-2', 'g__Microbacterium_A_383184_ASV-2', 'g__Leptotrichia_A_993758_ASV-3', 'g__Parvimonas_ASV-2', 'g__Ruminococcus_E_ASV-1', 'g__Leuconostoc_B_ASV-3', 'g__Pleomorphomonas_ASV-1', 'g___ASV-7', 'g__Peptostreptococcus_ASV-2', 'g__Prevotella_ASV-31', 'g__Exiguobacterium_A_ASV-1', 'g__Rothia_ASV-4', 'g__Prevotella_ASV-13', 'g__Corynebacterium_ASV-12', 'g__Staphylococcus_ASV-6', 'g__Acinetobacter_ASV-18', 'g___ASV-78', 'g__Streptococcus_ASV-5', 'g__Ligilactobacillus_ASV-2', 'g__Acetobacter_ASV-2', 'g__Anaerobutyricum_ASV-1', 'g__Collinsella_ASV-2', 'g__UMGS1994_ASV-1', 'g__Phocaeicola_A_858004_ASV-2', 'g__Flaviaesturariibacter_ASV-1', 'g__Acetobacter_ASV-1', 'g__Mycobacterium_ASV-1', 'g__Prevotella_ASV-23', 'g__Dialister_ASV-1', 'g___ASV-88', 'g__Brevibacterium_ASV-1', 'g__Rubellimicrobium_ASV-1', 'g___ASV-28'}\n"
     ]
    }
   ],
   "source": [
    "# Taxa shared between skin and nares in each group\n",
    "shared_AD = skin_taxa_AD & nares_taxa_AD\n",
    "shared_H = skin_taxa_H & nares_taxa_H\n",
    "\n",
    "# Taxa shared in AD but not in healthy\n",
    "unique_to_AD = shared_AD - shared_H\n",
    "\n",
    "# Print number and names\n",
    "print(f\"Taxa shared in skin & nares of AD but not in healthy: {len(unique_to_AD)}\")\n",
    "print(unique_to_AD)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "--- Taxa unique to nares ---\n",
      "g__Bacteroides_H_ASV-5\n",
      "g__Clostridium_P_ASV-2\n",
      "g__Leptotrichia_A_993758_ASV-11\n",
      "g__Leptotrichia_A_993758_ASV-12\n",
      "g__Leptotrichia_A_993758_ASV-17\n",
      "g__Paraprevotella_ASV-1\n",
      "g__Prevotella_ASV-35\n",
      "g__Prevotella_ASV-42\n",
      "g__Prevotella_ASV-44\n",
      "g__Streptococcus_ASV-12\n",
      "g__Streptococcus_ASV-13\n",
      "g__UMGS1994_ASV-2\n",
      "g___ASV-131\n",
      "g___ASV-141\n",
      "g___ASV-145\n",
      "g___ASV-95\n",
      "\n",
      "--- Taxa shared between skin and nares ---\n",
      "g__Acetobacter_ASV-1\n",
      "g__Acinetobacter_ASV-1\n",
      "g__Acinetobacter_ASV-16\n",
      "g__Acinetobacter_ASV-2\n",
      "g__Acinetobacter_ASV-3\n",
      "g__Acinetobacter_ASV-4\n",
      "g__Acinetobacter_ASV-5\n",
      "g__Acinetobacter_ASV-7\n",
      "g__Acinetobacter_ASV-8\n",
      "g__Actinomyces_ASV-1\n",
      "g__Actinomyces_ASV-2\n",
      "g__Actinomyces_ASV-3\n",
      "g__Aeromonas_ASV-1\n",
      "g__Agathobacter_164117_ASV-1\n",
      "g__Aggregatibacter_736122_ASV-1\n",
      "g__Aliicoccus_ASV-1\n",
      "g__Alloprevotella_ASV-1\n",
      "g__Alloprevotella_ASV-2\n",
      "g__Anaerobutyricum_ASV-1\n",
      "g__Anaerostipes_ASV-1\n",
      "g__Bifidobacterium_388775_ASV-1\n",
      "g__Bifidobacterium_388775_ASV-3\n",
      "g__Bifidobacterium_388775_ASV-4\n",
      "g__Bifidobacterium_388775_ASV-6\n",
      "g__Blautia_A_141781_ASV-1\n",
      "g__Brachybacterium_ASV-1\n",
      "g__Brachybacterium_ASV-2\n",
      "g__Brevibacterium_ASV-1\n",
      "g__Brevibacterium_ASV-2\n",
      "g__Brevibacterium_ASV-4\n",
      "g__Brevundimonas_ASV-1\n",
      "g__Capnocytophaga_820688_ASV-1\n",
      "g__Capnocytophaga_820688_ASV-2\n",
      "g__Capnocytophaga_820688_ASV-5\n",
      "g__Capnocytophaga_820690_ASV-1\n",
      "g__Catenibacterium_ASV-1\n",
      "g__Catenibacterium_ASV-2\n",
      "g__Chryseobacterium_796614_ASV-1\n",
      "g__Chryseobacterium_796614_ASV-2\n",
      "g__Cloacibacterium_ASV-1\n",
      "g__Cloacibacterium_ASV-3\n",
      "g__Clostridium_P_ASV-1\n",
      "g__Collinsella_ASV-1\n",
      "g__Comamonas_F_589250_ASV-1\n",
      "g__Coprococcus_A_121497_ASV-1\n",
      "g__Corynebacterium_ASV-1\n",
      "g__Corynebacterium_ASV-11\n",
      "g__Corynebacterium_ASV-13\n",
      "g__Corynebacterium_ASV-15\n",
      "g__Corynebacterium_ASV-16\n",
      "g__Corynebacterium_ASV-2\n",
      "g__Corynebacterium_ASV-3\n",
      "g__Corynebacterium_ASV-4\n",
      "g__Corynebacterium_ASV-5\n",
      "g__Corynebacterium_ASV-6\n",
      "g__Corynebacterium_ASV-7\n",
      "g__Corynebacterium_ASV-8\n",
      "g__Corynebacterium_ASV-9\n",
      "g__Cutibacterium_ASV-1\n",
      "g__Cutibacterium_ASV-2\n",
      "g__Dialister_ASV-1\n",
      "g__Dolosigranulum_ASV-1\n",
      "g__Dolosigranulum_ASV-2\n",
      "g__Empedobacter_790298_ASV-1\n",
      "g__Empedobacter_790298_ASV-2\n",
      "g__Enterenecus_ASV-1\n",
      "g__Enterocloster_ASV-2\n",
      "g__Enterocloster_ASV-3\n",
      "g__Escherichia_710834_ASV-1\n",
      "g__Exiguobacterium_A_ASV-1\n",
      "g__F0422_ASV-1\n",
      "g__F0422_ASV-2\n",
      "g__Faecalibacterium_ASV-1\n",
      "g__Faecalibacterium_ASV-2\n",
      "g__Faecalibacterium_ASV-3\n",
      "g__Faecalibacterium_ASV-5\n",
      "g__Finegoldia_ASV-1\n",
      "g__Fusobacterium_C_ASV-1\n",
      "g__Fusobacterium_C_ASV-2\n",
      "g__Fusobacterium_C_ASV-7\n",
      "g__Gemella_ASV-1\n",
      "g__Gemella_ASV-2\n",
      "g__Gemmiger_A_73129_ASV-1\n",
      "g__Gemmiger_A_73129_ASV-4\n",
      "g__Granulicatella_ASV-1\n",
      "g__Granulicatella_ASV-2\n",
      "g__Granulicatella_ASV-3\n",
      "g__Haemophilus_A_ASV-1\n",
      "g__Haemophilus_D_734546_ASV-1\n",
      "g__Haemophilus_D_734546_ASV-2\n",
      "g__Haemophilus_D_735815_ASV-1\n",
      "g__Halomonas_E_640244_ASV-1\n",
      "g__Hymenobacter_910554_ASV-1\n",
      "g__Jeotgalicoccus_A_310962_ASV-1\n",
      "g__Jeotgalicoccus_A_310962_ASV-2\n",
      "g__Lactococcus_A_346120_ASV-1\n",
      "g__Lactococcus_A_346120_ASV-2\n",
      "g__Lancefieldella_ASV-1\n",
      "g__Lautropia_ASV-1\n",
      "g__Leptotrichia_A_993758_ASV-1\n",
      "g__Leptotrichia_A_993758_ASV-4\n",
      "g__Leptotrichia_A_993758_ASV-5\n",
      "g__Leuconostoc_B_ASV-1\n",
      "g__Leuconostoc_B_ASV-2\n",
      "g__Luteimonas_C_615545_ASV-1\n",
      "g__Massilia_ASV-1\n",
      "g__Mediterraneibacter_A_155507_ASV-1\n",
      "g__Mesorhizobium_F_498388_ASV-3\n",
      "g__Microbacterium_A_383184_ASV-1\n",
      "g__Microbacterium_A_383184_ASV-2\n",
      "g__Micrococcus_ASV-1\n",
      "g__Micrococcus_ASV-2\n",
      "g__Neisseria_563205_ASV-1\n",
      "g__Neisseria_563205_ASV-2\n",
      "g__Neisseria_563205_ASV-3\n",
      "g__Neisseria_563205_ASV-4\n",
      "g__Nocardioides_A_392796_ASV-1\n",
      "g__Nocardioides_A_392796_ASV-3\n",
      "g__Paracoccus_ASV-1\n",
      "g__Paracoccus_ASV-2\n",
      "g__Paracoccus_ASV-3\n",
      "g__Pauljensenia_ASV-1\n",
      "g__Peptoniphilus_A_ASV-1\n",
      "g__Peptostreptococcus_ASV-2\n",
      "g__Peptostreptococcus_ASV-3\n",
      "g__Porphyromonas_A_859423_ASV-1\n",
      "g__Porphyromonas_A_859423_ASV-2\n",
      "g__Prevotella_ASV-1\n",
      "g__Prevotella_ASV-10\n",
      "g__Prevotella_ASV-11\n",
      "g__Prevotella_ASV-13\n",
      "g__Prevotella_ASV-15\n",
      "g__Prevotella_ASV-16\n",
      "g__Prevotella_ASV-2\n",
      "g__Prevotella_ASV-23\n",
      "g__Prevotella_ASV-3\n",
      "g__Prevotella_ASV-31\n",
      "g__Prevotella_ASV-32\n",
      "g__Prevotella_ASV-4\n",
      "g__Prevotella_ASV-5\n",
      "g__Prevotella_ASV-6\n",
      "g__Prevotella_ASV-8\n",
      "g__Prevotella_ASV-9\n",
      "g__Pseudomonas_E_647464_ASV-1\n",
      "g__Pseudomonas_E_647464_ASV-3\n",
      "g__Pseudomonas_E_647464_ASV-4\n",
      "g__Pseudomonas_E_647464_ASV-5\n",
      "g__Rothia_ASV-1\n",
      "g__Rothia_ASV-2\n",
      "g__Rothia_ASV-3\n",
      "g__Rothia_ASV-5\n",
      "g__Rothia_ASV-6\n",
      "g__SIO2C1_ASV-1\n",
      "g__SIO2C1_ASV-2\n",
      "g__SIO2C1_ASV-3\n",
      "g__SIO2C1_ASV-4\n",
      "g__Skermanella_ASV-1\n",
      "g__Sphingobacterium_ASV-1\n",
      "g__Staphylococcus_ASV-1\n",
      "g__Staphylococcus_ASV-2\n",
      "g__Staphylococcus_ASV-3\n",
      "g__Streptococcus_ASV-1\n",
      "g__Streptococcus_ASV-2\n",
      "g__Streptococcus_ASV-3\n",
      "g__Streptococcus_ASV-4\n",
      "g__Streptococcus_ASV-5\n",
      "g__Streptococcus_ASV-7\n",
      "g__Streptococcus_ASV-9\n",
      "g__Veillonella_A_ASV-1\n",
      "g__Veillonella_A_ASV-2\n",
      "g__Veillonella_A_ASV-3\n",
      "g__Veillonella_A_ASV-4\n",
      "g__Veillonella_A_ASV-5\n",
      "g__Veillonella_A_ASV-6\n",
      "g__Veillonella_A_ASV-8\n",
      "g___ASV-1\n",
      "g___ASV-10\n",
      "g___ASV-11\n",
      "g___ASV-114\n",
      "g___ASV-12\n",
      "g___ASV-14\n",
      "g___ASV-15\n",
      "g___ASV-16\n",
      "g___ASV-17\n",
      "g___ASV-18\n",
      "g___ASV-19\n",
      "g___ASV-2\n",
      "g___ASV-20\n",
      "g___ASV-21\n",
      "g___ASV-22\n",
      "g___ASV-23\n",
      "g___ASV-24\n",
      "g___ASV-25\n",
      "g___ASV-26\n",
      "g___ASV-28\n",
      "g___ASV-29\n",
      "g___ASV-3\n",
      "g___ASV-30\n",
      "g___ASV-31\n",
      "g___ASV-35\n",
      "g___ASV-37\n",
      "g___ASV-39\n",
      "g___ASV-4\n",
      "g___ASV-46\n",
      "g___ASV-47\n",
      "g___ASV-5\n",
      "g___ASV-50\n",
      "g___ASV-51\n",
      "g___ASV-58\n",
      "g___ASV-6\n",
      "g___ASV-61\n",
      "g___ASV-64\n",
      "g___ASV-66\n",
      "g___ASV-7\n",
      "g___ASV-76\n",
      "g___ASV-77\n",
      "g___ASV-8\n",
      "g___ASV-9\n"
     ]
    }
   ],
   "source": [
    "# Taxa only in skin (not in nares)\n",
    "skin_only_taxa = skin_taxa - nares_taxa\n",
    "\n",
    "# Taxa only in nares (not in skin)\n",
    "nares_only_taxa = nares_taxa - skin_taxa\n",
    "\n",
    "# Taxa shared between both\n",
    "shared_taxa = overlapping_taxa\n",
    "\n",
    "# Print them\n",
    "# print(\"\\n--- Taxa unique to skin ---\")\n",
    "# for taxon in sorted(skin_only_taxa):\n",
    "#     print(taxon)\n",
    "\n",
    "print(\"\\n--- Taxa unique to nares ---\")\n",
    "for taxon in sorted(nares_only_taxa):\n",
    "    print(taxon)\n",
    "\n",
    "print(\"\\n--- Taxa shared between skin and nares ---\")\n",
    "for taxon in sorted(shared_taxa):\n",
    "    print(taxon)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'pic'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "File \u001b[0;32m~/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/pandas/core/indexes/base.py:3805\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3804\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 3805\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_engine\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcasted_key\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3806\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n",
      "File \u001b[0;32mindex.pyx:167\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mindex.pyx:196\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:7081\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:7089\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'pic'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[29], line 7\u001b[0m\n\u001b[1;32m      4\u001b[0m subject_taxa_dict \u001b[38;5;241m=\u001b[39m defaultdict(\u001b[38;5;28;01mlambda\u001b[39;00m: {\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mskin\u001b[39m\u001b[38;5;124m'\u001b[39m: \u001b[38;5;28mset\u001b[39m(), \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mnares\u001b[39m\u001b[38;5;124m'\u001b[39m: \u001b[38;5;28mset\u001b[39m()})\n\u001b[1;32m      6\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m _, row \u001b[38;5;129;01min\u001b[39;00m df\u001b[38;5;241m.\u001b[39miterrows():\n\u001b[0;32m----> 7\u001b[0m     subject \u001b[38;5;241m=\u001b[39m \u001b[43mrow\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mpic\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\n\u001b[1;32m      8\u001b[0m     site \u001b[38;5;241m=\u001b[39m row[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124msite\u001b[39m\u001b[38;5;124m'\u001b[39m]  \u001b[38;5;66;03m# must be 'skin' or 'nares'\u001b[39;00m\n\u001b[1;32m      9\u001b[0m     taxa_set \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mset\u001b[39m(row[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mtaxa\u001b[39m\u001b[38;5;124m'\u001b[39m])  \u001b[38;5;66;03m# make sure this is already a list or string split into taxa\u001b[39;00m\n",
      "File \u001b[0;32m~/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/pandas/core/series.py:1121\u001b[0m, in \u001b[0;36mSeries.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   1118\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_values[key]\n\u001b[1;32m   1120\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m key_is_scalar:\n\u001b[0;32m-> 1121\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_get_value\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1123\u001b[0m \u001b[38;5;66;03m# Convert generator to list before going through hashable part\u001b[39;00m\n\u001b[1;32m   1124\u001b[0m \u001b[38;5;66;03m# (We will iterate through the generator there to check for slices)\u001b[39;00m\n\u001b[1;32m   1125\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m is_iterator(key):\n",
      "File \u001b[0;32m~/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/pandas/core/series.py:1237\u001b[0m, in \u001b[0;36mSeries._get_value\u001b[0;34m(self, label, takeable)\u001b[0m\n\u001b[1;32m   1234\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_values[label]\n\u001b[1;32m   1236\u001b[0m \u001b[38;5;66;03m# Similar to Index.get_value, but we do not fall back to positional\u001b[39;00m\n\u001b[0;32m-> 1237\u001b[0m loc \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mindex\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mlabel\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1239\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m is_integer(loc):\n\u001b[1;32m   1240\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_values[loc]\n",
      "File \u001b[0;32m~/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/pandas/core/indexes/base.py:3812\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3807\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(casted_key, \u001b[38;5;28mslice\u001b[39m) \u001b[38;5;129;01mor\u001b[39;00m (\n\u001b[1;32m   3808\u001b[0m         \u001b[38;5;28misinstance\u001b[39m(casted_key, abc\u001b[38;5;241m.\u001b[39mIterable)\n\u001b[1;32m   3809\u001b[0m         \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28many\u001b[39m(\u001b[38;5;28misinstance\u001b[39m(x, \u001b[38;5;28mslice\u001b[39m) \u001b[38;5;28;01mfor\u001b[39;00m x \u001b[38;5;129;01min\u001b[39;00m casted_key)\n\u001b[1;32m   3810\u001b[0m     ):\n\u001b[1;32m   3811\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m InvalidIndexError(key)\n\u001b[0;32m-> 3812\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(key) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01merr\u001b[39;00m\n\u001b[1;32m   3813\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m:\n\u001b[1;32m   3814\u001b[0m     \u001b[38;5;66;03m# If we have a listlike key, _check_indexing_error will raise\u001b[39;00m\n\u001b[1;32m   3815\u001b[0m     \u001b[38;5;66;03m#  InvalidIndexError. Otherwise we fall through and re-raise\u001b[39;00m\n\u001b[1;32m   3816\u001b[0m     \u001b[38;5;66;03m#  the TypeError.\u001b[39;00m\n\u001b[1;32m   3817\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_check_indexing_error(key)\n",
      "\u001b[0;31mKeyError\u001b[0m: 'pic'"
     ]
    }
   ],
   "source": [
    "from collections import defaultdict\n",
    "\n",
    "# Initialize structure: subject_taxa_dict[pid]['skin'] or ['nares'] = set of taxa\n",
    "subject_taxa_dict = defaultdict(lambda: {'skin': set(), 'nares': set()})\n",
    "\n",
    "for _, row in df.iterrows():\n",
    "    subject = row['pic']\n",
    "    site = row['site']  # must be 'skin' or 'nares'\n",
    "    taxa_set = set(row['taxa'])  # make sure this is already a list or string split into taxa\n",
    "    subject_taxa_dict[subject][site] |= taxa_set\n",
    "subject_taxa_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define subject lists by group\n",
    "ad_subjects = [s for s in subject_taxa_dict if 'AD' in s]\n",
    "healthy_subjects = [s for s in subject_taxa_dict if 'H' in s]\n",
    "\n",
    "print(f\"AD subjects: {len(ad_subjects)}\")\n",
    "print(f\"Healthy subjects: {len(healthy_subjects)}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "317f4393",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'ad_subjects' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[14], line 9\u001b[0m\n\u001b[1;32m      6\u001b[0m n_iterations \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m1000\u001b[39m\n\u001b[1;32m      8\u001b[0m \u001b[38;5;66;03m# Equalize group sizes\u001b[39;00m\n\u001b[0;32m----> 9\u001b[0m n \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mmin\u001b[39m(\u001b[38;5;28mlen\u001b[39m(\u001b[43mad_subjects\u001b[49m), \u001b[38;5;28mlen\u001b[39m(healthy_subjects))\n\u001b[1;32m     11\u001b[0m bootstrap_counts \u001b[38;5;241m=\u001b[39m []\n\u001b[1;32m     13\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m _ \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mrange\u001b[39m(n_iterations):\n\u001b[1;32m     14\u001b[0m     \u001b[38;5;66;03m# Randomly sample n subjects from each group\u001b[39;00m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'ad_subjects' is not defined"
     ]
    }
   ],
   "source": [
    "\n",
    "import random\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Number of bootstrap iterations\n",
    "n_iterations = 1000\n",
    "\n",
    "# Equalize group sizes\n",
    "n = min(len(ad_subjects), len(healthy_subjects))\n",
    "\n",
    "bootstrap_counts = []\n",
    "\n",
    "for _ in range(n_iterations):\n",
    "    # Randomly sample n subjects from each group\n",
    "    ad_sample = random.sample(ad_subjects, n)\n",
    "    h_sample = random.sample(healthy_subjects, n)\n",
    "\n",
    "    # Get taxa per site and group for sampled individuals\n",
    "    ad_skin_taxa = set.union(*[subject_taxa_dict[s]['skin'] for s in ad_sample])\n",
    "    ad_nares_taxa = set.union(*[subject_taxa_dict[s]['nares'] for s in ad_sample])\n",
    "    h_skin_taxa = set.union(*[subject_taxa_dict[s]['skin'] for s in h_sample])\n",
    "    h_nares_taxa = set.union(*[subject_taxa_dict[s]['nares'] for s in h_sample])\n",
    "\n",
    "    # Shared taxa in each group\n",
    "    shared_ad = ad_skin_taxa & ad_nares_taxa\n",
    "    shared_h = h_skin_taxa & h_nares_taxa\n",
    "\n",
    "    # Taxa shared only in AD\n",
    "    unique_to_ad = shared_ad - shared_h\n",
    "\n",
    "    # Record count\n",
    "    bootstrap_counts.append(len(unique_to_ad))\n",
    "\n",
    "# Summary statistics\n",
    "mean_count = np.mean(bootstrap_counts)\n",
    "ci_low, ci_high = np.percentile(bootstrap_counts, [2.5, 97.5])\n",
    "\n",
    "print(f\"Mean AD-specific shared taxa (bootstrapped): {mean_count:.1f}\")\n",
    "print(f\"95% CI: ({ci_low:.1f}, {ci_high:.1f})\")\n",
    "\n",
    "# Optional: visualize\n",
    "plt.hist(bootstrap_counts, bins=30, color='skyblue', edgecolor='black')\n",
    "plt.axvline(mean_count, color='red', linestyle='--', label='Mean')\n",
    "plt.xlabel('Shared taxa unique to AD')\n",
    "plt.ylabel('Frequency')\n",
    "plt.title('Bootstrapped distribution of AD-only shared taxa')\n",
    "plt.legend()\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"bootstrap_shared_taxa_AD.png\", dpi=300)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "qiime2-metagenome-2024.10",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
