{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Random Forest Analyses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Standard library\n",
    "import warnings\n",
    "import logging\n",
    "from itertools import combinations\n",
    "\n",
    "# Scientific computing\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from numpy import array\n",
    "import scipy\n",
    "import scipy.stats as ss\n",
    "from scipy import interp\n",
    "from scipy.stats import wilcoxon, ttest_rel\n",
    "\n",
    "# Visualization\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.gridspec import GridSpec\n",
    "import seaborn as sns\n",
    "\n",
    "# scikit-bio\n",
    "from skbio.stats.distance import permanova\n",
    "\n",
    "# BIOM format\n",
    "import biom\n",
    "from biom import load_table\n",
    "\n",
    "# Scikit-learn\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.metrics import (\n",
    "    accuracy_score, confusion_matrix, classification_report,\n",
    "    roc_curve, auc, RocCurveDisplay\n",
    ")\n",
    "from sklearn.model_selection import GroupKFold, StratifiedKFold\n",
    "from sklearn.preprocessing import label_binarize\n",
    "\n",
    "# Set overall styling for plots\n",
    "sns.set_context(\"paper\", font_scale=1.5)\n",
    "sns.set_style(\"ticks\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Random Forest Classification Tasks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GTGCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTTATTGGGCGTAAAGCGAGCGCAGGCGGTTAGATAAGTCTGAAGTTAAAGGCTG</th>\n",
       "      <th>GTGCCAGCCGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTTATTGGGCGTAAAGCGAGCGCAGGCGGTTAGATAAGTCTGAAGTTAAAGGCTG</th>\n",
       "      <th>GTGCCAGCAGCCGCGGTAATACGTAGGGTGCAAGCGTTGTCCGGAATTACTGGGCGTAAAGAGCTCGTAGGTGGTTTGTCACGTCGTCTGTGAAATTCCA</th>\n",
       "      <th>GTGCCAGCCGCCGCGGTAATACGTAGGGTGCAAGCGTTGTCCGGAATTACTGGGCGTAAAGAGCTCGTAGGTGGTTTGTCACGTCGTCTGTGAAATTCCA</th>\n",
       "      <th>GTGCCAGCAGCCGCGGTAATACGTAGGGTGCAAGCGTTAATCGGAATTATTGGGCGTAAAGCGAGTGCAGACGGTTACTTAAGCCAGATGTGAAATCCCC</th>\n",
       "      <th>GTGCCAGCAGCCGCGGTAATACGTAGGTGGCAAGCGTTGTCCGGAATTATTGGGCGTAAAGCGCGCGCAGGCGGTTTCTTAAGTCTGATGTGAAAGCCCC</th>\n",
       "      <th>GTGCCAGCAGCCGCGGTGATACGTAGGGTGCGAGCGTTGTCCGGATTTATTGGGCGTAAAGGGCTCGTAGGTGGTTGATCGCGTCGGAAGTGTAATCTTG</th>\n",
       "      <th>GTGCCAGCAGCCGCGGTAATACGTAGGGTCCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGTGCGCAGGCGGTTGTGCAAGACCGATGTGAAATCCCC</th>\n",
       "      <th>GTGCCAGCCGCCGCGGTAATACGTAGGTGGCAAGCGTTGTCCGGATTTATTGGGCGTAAAGGGAGCGCAGGTGGTTTCTTAAGTCTGATGTGAAAGCCCA</th>\n",
       "      <th>GTGCCAGCCGCCGCGGTAATACGGAAGGTCCAGGCGTTATCCGGATTTATTGGGTTTAAAGGGAGCGTAGGCGGATTATTAAGTCAGTGGTGAAAGACGG</th>\n",
       "      <th>...</th>\n",
       "      <th>GTGCCAGCCGCCGCGGTAATACGTAGGGGGCAAGCGTTATCCGGATTTACTGGGTGTAAAGGGAGCGTAGACGGCGCAGCAAGTCTGATGTGAAAGGCAG</th>\n",
       "      <th>GTGCCAGCAGCCGCGGTAAGACAGAGGGTGCAAACGTTGCTCGGAATCACTGGGCGTAAAGGGCGTGTAGGCGGGAGAGAAAGTCGGGCGTGAAATCCCT</th>\n",
       "      <th>GTGCCAGCCGCGGTAATACGTAGGGGGCTAGCGTTGTCCGGAATCACTGGGCGTAAAGGGTTCGCAGGCGGAAATGCAAGTCAGGTGTAAAAGGCAGTAG</th>\n",
       "      <th>GTGCCAGCAGCCGCGGTAATACGTAGGGCGCGAGCGTTGTCCGGAATTATTGGGCGTAAAGAGCTTGTAGGCGGTTTGTTGCGTCTGCTGTGAAAGACCG</th>\n",
       "      <th>GTGCCAGCCGCCGCGGTAATACGTAGGGCGCGAGCGTTGTCCGGAATTATTGGGCGTAAAGAGCTTGTAGGCGGTTTGTTGCGTCTGCTGTGAAAGACCG</th>\n",
       "      <th>GTGCCAGCAGCCGCGGTAATACGGAGGGTGCAAGCGTTATCCGGAATCATTGGGTTTAAAGGGTCCGCAGGCGGATTTATAAGTCAGTGGTGAAAGCCTA</th>\n",
       "      <th>GTGCCAGCAGCCGCGGTAATACGTAGGTGGCGAGCGTTGTCCGGAATTACTGGGTGTAAAGGGCGTGTAGGCGGGAAGGTAAGTCAGATGTGAAATACCG</th>\n",
       "      <th>GTGCCAGCCGCCGCGGTAATACGGAGGATGCGAGCGTTATTCGGAATCATTGGGTTTAAAGGGTCTGTAGGCGGGCTATTAAGTCAGAGGTGAAAGGTTT</th>\n",
       "      <th>GTGCCAGCCGCCGCGGTAAGACGAAGGGGGCTAGCGTTGTTCGGAATTACTGGGCGTAAAGCGCGTGCAGGCGGTTATCCAAGTCGGGTGTGAAAGCCTT</th>\n",
       "      <th>GTCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTTATTGGGCGTAAAGCGAGCGCAGGCGGTTAGATAAGTCTGAAGTTAAAGGCTGT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>900344</th>\n",
       "      <td>984.0</td>\n",
       "      <td>611.0</td>\n",
       "      <td>114.0</td>\n",
       "      <td>82.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900459</th>\n",
       "      <td>118.0</td>\n",
       "      <td>106.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900221</th>\n",
       "      <td>22.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900570</th>\n",
       "      <td>389.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900092</th>\n",
       "      <td>3106.0</td>\n",
       "      <td>1707.0</td>\n",
       "      <td>59.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9003972</th>\n",
       "      <td>1168.0</td>\n",
       "      <td>593.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>0</td>\n",
       "      <td>736.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900097</th>\n",
       "      <td>24.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>8.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900498</th>\n",
       "      <td>15.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900276</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>151.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900406</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>502 rows × 2283 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         GTGCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTTATTGGGCGTAAAGCGAGCGCAGGCGGTTAGATAAGTCTGAAGTTAAAGGCTG  \\\n",
       "900344                                               984.0                                                      \n",
       "900459                                               118.0                                                      \n",
       "900221                                                22.0                                                      \n",
       "900570                                               389.0                                                      \n",
       "900092                                              3106.0                                                      \n",
       "...                                                    ...                                                      \n",
       "9003972                                             1168.0                                                      \n",
       "900097                                                24.0                                                      \n",
       "900498                                                15.0                                                      \n",
       "900276                                                   0                                                      \n",
       "900406                                                   0                                                      \n",
       "\n",
       "         GTGCCAGCCGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTTATTGGGCGTAAAGCGAGCGCAGGCGGTTAGATAAGTCTGAAGTTAAAGGCTG  \\\n",
       "900344                                               611.0                                                      \n",
       "900459                                               106.0                                                      \n",
       "900221                                                   0                                                      \n",
       "900570                                                   0                                                      \n",
       "900092                                              1707.0                                                      \n",
       "...                                                    ...                                                      \n",
       "9003972                                              593.0                                                      \n",
       "900097                                                   0                                                      \n",
       "900498                                                17.0                                                      \n",
       "900276                                                   0                                                      \n",
       "900406                                                   0                                                      \n",
       "\n",
       "         GTGCCAGCAGCCGCGGTAATACGTAGGGTGCAAGCGTTGTCCGGAATTACTGGGCGTAAAGAGCTCGTAGGTGGTTTGTCACGTCGTCTGTGAAATTCCA  \\\n",
       "900344                                               114.0                                                      \n",
       "900459                                                   0                                                      \n",
       "900221                                                   0                                                      \n",
       "900570                                                   0                                                      \n",
       "900092                                                59.0                                                      \n",
       "...                                                    ...                                                      \n",
       "9003972                                               16.0                                                      \n",
       "900097                                                   0                                                      \n",
       "900498                                                   0                                                      \n",
       "900276                                                30.0                                                      \n",
       "900406                                                   0                                                      \n",
       "\n",
       "         GTGCCAGCCGCCGCGGTAATACGTAGGGTGCAAGCGTTGTCCGGAATTACTGGGCGTAAAGAGCTCGTAGGTGGTTTGTCACGTCGTCTGTGAAATTCCA  \\\n",
       "900344                                                82.0                                                      \n",
       "900459                                                   0                                                      \n",
       "900221                                                   0                                                      \n",
       "900570                                                   0                                                      \n",
       "900092                                                32.0                                                      \n",
       "...                                                    ...                                                      \n",
       "9003972                                                  0                                                      \n",
       "900097                                                   0                                                      \n",
       "900498                                                   0                                                      \n",
       "900276                                                   0                                                      \n",
       "900406                                                   0                                                      \n",
       "\n",
       "         GTGCCAGCAGCCGCGGTAATACGTAGGGTGCAAGCGTTAATCGGAATTATTGGGCGTAAAGCGAGTGCAGACGGTTACTTAAGCCAGATGTGAAATCCCC  \\\n",
       "900344                                                22.0                                                      \n",
       "900459                                                   0                                                      \n",
       "900221                                                   0                                                      \n",
       "900570                                                 8.0                                                      \n",
       "900092                                                 3.0                                                      \n",
       "...                                                    ...                                                      \n",
       "9003972                                               28.0                                                      \n",
       "900097                                                   0                                                      \n",
       "900498                                                   0                                                      \n",
       "900276                                                   0                                                      \n",
       "900406                                                   0                                                      \n",
       "\n",
       "         GTGCCAGCAGCCGCGGTAATACGTAGGTGGCAAGCGTTGTCCGGAATTATTGGGCGTAAAGCGCGCGCAGGCGGTTTCTTAAGTCTGATGTGAAAGCCCC  \\\n",
       "900344                                                15.0                                                      \n",
       "900459                                                   0                                                      \n",
       "900221                                                   0                                                      \n",
       "900570                                                   0                                                      \n",
       "900092                                                   0                                                      \n",
       "...                                                    ...                                                      \n",
       "9003972                                                  0                                                      \n",
       "900097                                                   0                                                      \n",
       "900498                                                   0                                                      \n",
       "900276                                                   0                                                      \n",
       "900406                                                   0                                                      \n",
       "\n",
       "         GTGCCAGCAGCCGCGGTGATACGTAGGGTGCGAGCGTTGTCCGGATTTATTGGGCGTAAAGGGCTCGTAGGTGGTTGATCGCGTCGGAAGTGTAATCTTG  \\\n",
       "900344                                                 8.0                                                      \n",
       "900459                                                   0                                                      \n",
       "900221                                                16.0                                                      \n",
       "900570                                                11.0                                                      \n",
       "900092                                                   0                                                      \n",
       "...                                                    ...                                                      \n",
       "9003972                                              736.0                                                      \n",
       "900097                                                33.0                                                      \n",
       "900498                                                34.0                                                      \n",
       "900276                                               151.0                                                      \n",
       "900406                                                   0                                                      \n",
       "\n",
       "         GTGCCAGCAGCCGCGGTAATACGTAGGGTCCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGTGCGCAGGCGGTTGTGCAAGACCGATGTGAAATCCCC  \\\n",
       "900344                                                 8.0                                                      \n",
       "900459                                                   0                                                      \n",
       "900221                                                   0                                                      \n",
       "900570                                                   0                                                      \n",
       "900092                                                   0                                                      \n",
       "...                                                    ...                                                      \n",
       "9003972                                                  0                                                      \n",
       "900097                                                   0                                                      \n",
       "900498                                                   0                                                      \n",
       "900276                                                   0                                                      \n",
       "900406                                                   0                                                      \n",
       "\n",
       "         GTGCCAGCCGCCGCGGTAATACGTAGGTGGCAAGCGTTGTCCGGATTTATTGGGCGTAAAGGGAGCGCAGGTGGTTTCTTAAGTCTGATGTGAAAGCCCA  \\\n",
       "900344                                                 6.0                                                      \n",
       "900459                                                   0                                                      \n",
       "900221                                                   0                                                      \n",
       "900570                                                   0                                                      \n",
       "900092                                                   0                                                      \n",
       "...                                                    ...                                                      \n",
       "9003972                                                  0                                                      \n",
       "900097                                                   0                                                      \n",
       "900498                                                14.0                                                      \n",
       "900276                                                   0                                                      \n",
       "900406                                                   0                                                      \n",
       "\n",
       "         GTGCCAGCCGCCGCGGTAATACGGAAGGTCCAGGCGTTATCCGGATTTATTGGGTTTAAAGGGAGCGTAGGCGGATTATTAAGTCAGTGGTGAAAGACGG  \\\n",
       "900344                                                 3.0                                                      \n",
       "900459                                                   0                                                      \n",
       "900221                                                   0                                                      \n",
       "900570                                                   0                                                      \n",
       "900092                                                 7.0                                                      \n",
       "...                                                    ...                                                      \n",
       "9003972                                               36.0                                                      \n",
       "900097                                                   0                                                      \n",
       "900498                                                   0                                                      \n",
       "900276                                                   0                                                      \n",
       "900406                                                   0                                                      \n",
       "\n",
       "         ...  \\\n",
       "900344   ...   \n",
       "900459   ...   \n",
       "900221   ...   \n",
       "900570   ...   \n",
       "900092   ...   \n",
       "...      ...   \n",
       "9003972  ...   \n",
       "900097   ...   \n",
       "900498   ...   \n",
       "900276   ...   \n",
       "900406   ...   \n",
       "\n",
       "         GTGCCAGCCGCCGCGGTAATACGTAGGGGGCAAGCGTTATCCGGATTTACTGGGTGTAAAGGGAGCGTAGACGGCGCAGCAAGTCTGATGTGAAAGGCAG  \\\n",
       "900344                                                   0                                                      \n",
       "900459                                                   0                                                      \n",
       "900221                                                   0                                                      \n",
       "900570                                                   0                                                      \n",
       "900092                                                   0                                                      \n",
       "...                                                    ...                                                      \n",
       "9003972                                                  0                                                      \n",
       "900097                                                 8.0                                                      \n",
       "900498                                                   0                                                      \n",
       "900276                                                   0                                                      \n",
       "900406                                                   0                                                      \n",
       "\n",
       "         GTGCCAGCAGCCGCGGTAAGACAGAGGGTGCAAACGTTGCTCGGAATCACTGGGCGTAAAGGGCGTGTAGGCGGGAGAGAAAGTCGGGCGTGAAATCCCT  \\\n",
       "900344                                                   0                                                      \n",
       "900459                                                   0                                                      \n",
       "900221                                                   0                                                      \n",
       "900570                                                   0                                                      \n",
       "900092                                                   0                                                      \n",
       "...                                                    ...                                                      \n",
       "9003972                                                  0                                                      \n",
       "900097                                                 5.0                                                      \n",
       "900498                                                   0                                                      \n",
       "900276                                                   0                                                      \n",
       "900406                                                   0                                                      \n",
       "\n",
       "         GTGCCAGCCGCGGTAATACGTAGGGGGCTAGCGTTGTCCGGAATCACTGGGCGTAAAGGGTTCGCAGGCGGAAATGCAAGTCAGGTGTAAAAGGCAGTAG  \\\n",
       "900344                                                   0                                                      \n",
       "900459                                                   0                                                      \n",
       "900221                                                   0                                                      \n",
       "900570                                                   0                                                      \n",
       "900092                                                   0                                                      \n",
       "...                                                    ...                                                      \n",
       "9003972                                                  0                                                      \n",
       "900097                                                 1.0                                                      \n",
       "900498                                                   0                                                      \n",
       "900276                                                   0                                                      \n",
       "900406                                                   0                                                      \n",
       "\n",
       "         GTGCCAGCAGCCGCGGTAATACGTAGGGCGCGAGCGTTGTCCGGAATTATTGGGCGTAAAGAGCTTGTAGGCGGTTTGTTGCGTCTGCTGTGAAAGACCG  \\\n",
       "900344                                                   0                                                      \n",
       "900459                                                   0                                                      \n",
       "900221                                                   0                                                      \n",
       "900570                                                   0                                                      \n",
       "900092                                                   0                                                      \n",
       "...                                                    ...                                                      \n",
       "9003972                                                  0                                                      \n",
       "900097                                                   0                                                      \n",
       "900498                                                15.0                                                      \n",
       "900276                                                   0                                                      \n",
       "900406                                                   0                                                      \n",
       "\n",
       "         GTGCCAGCCGCCGCGGTAATACGTAGGGCGCGAGCGTTGTCCGGAATTATTGGGCGTAAAGAGCTTGTAGGCGGTTTGTTGCGTCTGCTGTGAAAGACCG  \\\n",
       "900344                                                   0                                                      \n",
       "900459                                                   0                                                      \n",
       "900221                                                   0                                                      \n",
       "900570                                                   0                                                      \n",
       "900092                                                   0                                                      \n",
       "...                                                    ...                                                      \n",
       "9003972                                                  0                                                      \n",
       "900097                                                   0                                                      \n",
       "900498                                                10.0                                                      \n",
       "900276                                                   0                                                      \n",
       "900406                                                   0                                                      \n",
       "\n",
       "         GTGCCAGCAGCCGCGGTAATACGGAGGGTGCAAGCGTTATCCGGAATCATTGGGTTTAAAGGGTCCGCAGGCGGATTTATAAGTCAGTGGTGAAAGCCTA  \\\n",
       "900344                                                   0                                                      \n",
       "900459                                                   0                                                      \n",
       "900221                                                   0                                                      \n",
       "900570                                                   0                                                      \n",
       "900092                                                   0                                                      \n",
       "...                                                    ...                                                      \n",
       "9003972                                                  0                                                      \n",
       "900097                                                   0                                                      \n",
       "900498                                                 8.0                                                      \n",
       "900276                                                   0                                                      \n",
       "900406                                                   0                                                      \n",
       "\n",
       "         GTGCCAGCAGCCGCGGTAATACGTAGGTGGCGAGCGTTGTCCGGAATTACTGGGTGTAAAGGGCGTGTAGGCGGGAAGGTAAGTCAGATGTGAAATACCG  \\\n",
       "900344                                                   0                                                      \n",
       "900459                                                   0                                                      \n",
       "900221                                                   0                                                      \n",
       "900570                                                   0                                                      \n",
       "900092                                                   0                                                      \n",
       "...                                                    ...                                                      \n",
       "9003972                                                  0                                                      \n",
       "900097                                                   0                                                      \n",
       "900498                                                   0                                                      \n",
       "900276                                                11.0                                                      \n",
       "900406                                                   0                                                      \n",
       "\n",
       "         GTGCCAGCCGCCGCGGTAATACGGAGGATGCGAGCGTTATTCGGAATCATTGGGTTTAAAGGGTCTGTAGGCGGGCTATTAAGTCAGAGGTGAAAGGTTT  \\\n",
       "900344                                                   0                                                      \n",
       "900459                                                   0                                                      \n",
       "900221                                                   0                                                      \n",
       "900570                                                   0                                                      \n",
       "900092                                                   0                                                      \n",
       "...                                                    ...                                                      \n",
       "9003972                                                  0                                                      \n",
       "900097                                                   0                                                      \n",
       "900498                                                   0                                                      \n",
       "900276                                                 3.0                                                      \n",
       "900406                                                   0                                                      \n",
       "\n",
       "         GTGCCAGCCGCCGCGGTAAGACGAAGGGGGCTAGCGTTGTTCGGAATTACTGGGCGTAAAGCGCGTGCAGGCGGTTATCCAAGTCGGGTGTGAAAGCCTT  \\\n",
       "900344                                                   0                                                      \n",
       "900459                                                   0                                                      \n",
       "900221                                                   0                                                      \n",
       "900570                                                   0                                                      \n",
       "900092                                                   0                                                      \n",
       "...                                                    ...                                                      \n",
       "9003972                                                  0                                                      \n",
       "900097                                                   0                                                      \n",
       "900498                                                   0                                                      \n",
       "900276                                                 2.0                                                      \n",
       "900406                                                   0                                                      \n",
       "\n",
       "         GTCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTTATTGGGCGTAAAGCGAGCGCAGGCGGTTAGATAAGTCTGAAGTTAAAGGCTGT  \n",
       "900344                                                   0                                                     \n",
       "900459                                                   0                                                     \n",
       "900221                                                   0                                                     \n",
       "900570                                                   0                                                     \n",
       "900092                                                   0                                                     \n",
       "...                                                    ...                                                     \n",
       "9003972                                                  0                                                     \n",
       "900097                                                   0                                                     \n",
       "900498                                                   0                                                     \n",
       "900276                                                 1.0                                                     \n",
       "900406                                                   0                                                     \n",
       "\n",
       "[502 rows x 2283 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read in table at ASV level\n",
    "biom_path = '..//Data/Tables/Absolute_Abundance_Tables/209766_filtered_feature_table.biom'\n",
    "# biom_path = '../Data/Tables/Absolute_Abundance_Tables/209766_filtered_by_prevalence_1pct_rare_Genus-ASV-non-collapse.biom'\n",
    "\n",
    "biom_tbl = load_table(biom_path)\n",
    "df = pd.DataFrame(biom_tbl.to_dataframe().T)\n",
    "\n",
    "# delete the prefix from the index\n",
    "df.index = df.index.str.replace('15564.', '')\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PlateNumber</th>\n",
       "      <th>PlateLocation</th>\n",
       "      <th>i5</th>\n",
       "      <th>i5Sequence</th>\n",
       "      <th>i7</th>\n",
       "      <th>i7Sequence</th>\n",
       "      <th>identifier</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Plate ID</th>\n",
       "      <th>Well location</th>\n",
       "      <th>...</th>\n",
       "      <th>sex</th>\n",
       "      <th>enrolment_date</th>\n",
       "      <th>enrolment_season</th>\n",
       "      <th>hiv_exposure</th>\n",
       "      <th>hiv_status</th>\n",
       "      <th>household_size</th>\n",
       "      <th>o_scorad</th>\n",
       "      <th>FWD_filepath</th>\n",
       "      <th>REV_filepath</th>\n",
       "      <th>group</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#sample-id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Ca009STL</th>\n",
       "      <td>1</td>\n",
       "      <td>A1</td>\n",
       "      <td>SA501</td>\n",
       "      <td>ATCGTACG</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA501</td>\n",
       "      <td>CGAGAGTT-ATCGTACG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>A1</td>\n",
       "      <td>...</td>\n",
       "      <td>male</td>\n",
       "      <td>4/16/2015</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>4.0</td>\n",
       "      <td>40</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900221</th>\n",
       "      <td>1</td>\n",
       "      <td>B1</td>\n",
       "      <td>SA502</td>\n",
       "      <td>ACTATCTG</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA502</td>\n",
       "      <td>CGAGAGTT-ACTATCTG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>B1</td>\n",
       "      <td>...</td>\n",
       "      <td>female</td>\n",
       "      <td>8/11/2015</td>\n",
       "      <td>Winter</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>7.0</td>\n",
       "      <td>34</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca010EBL</th>\n",
       "      <td>1</td>\n",
       "      <td>C1</td>\n",
       "      <td>SA503</td>\n",
       "      <td>TAGCGAGT</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA503</td>\n",
       "      <td>CGAGAGTT-TAGCGAGT</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>C1</td>\n",
       "      <td>...</td>\n",
       "      <td>female</td>\n",
       "      <td>11/20/2014</td>\n",
       "      <td>Spring</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>7.0</td>\n",
       "      <td>21</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900460</th>\n",
       "      <td>1</td>\n",
       "      <td>D1</td>\n",
       "      <td>SA504</td>\n",
       "      <td>CTGCGTGT</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA504</td>\n",
       "      <td>CGAGAGTT-CTGCGTGT</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>D1</td>\n",
       "      <td>...</td>\n",
       "      <td>female</td>\n",
       "      <td>9/23/2015</td>\n",
       "      <td>Spring</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>40</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900051</th>\n",
       "      <td>1</td>\n",
       "      <td>E1</td>\n",
       "      <td>SA505</td>\n",
       "      <td>TCATCGAG</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA505</td>\n",
       "      <td>CGAGAGTT-TCATCGAG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>E1</td>\n",
       "      <td>...</td>\n",
       "      <td>male</td>\n",
       "      <td>4/21/2015</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>7.0</td>\n",
       "      <td>41</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca006ONL2</th>\n",
       "      <td>6</td>\n",
       "      <td>H1</td>\n",
       "      <td>SA508</td>\n",
       "      <td>GACACCGT</td>\n",
       "      <td>SB701</td>\n",
       "      <td>CTCGACTT</td>\n",
       "      <td>SB701SA508</td>\n",
       "      <td>CTCGACTT-GACACCGT</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>H1</td>\n",
       "      <td>...</td>\n",
       "      <td>female</td>\n",
       "      <td>3/25/2015</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>3.0</td>\n",
       "      <td>34</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca006ONNL</th>\n",
       "      <td>6</td>\n",
       "      <td>F2</td>\n",
       "      <td>SA506</td>\n",
       "      <td>CGTGAGTG</td>\n",
       "      <td>SB702</td>\n",
       "      <td>CGAAGTAT</td>\n",
       "      <td>SB702SA506</td>\n",
       "      <td>CGAAGTAT-CGTGAGTG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>F2</td>\n",
       "      <td>...</td>\n",
       "      <td>female</td>\n",
       "      <td>3/25/2015</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>3.0</td>\n",
       "      <td>34</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADNL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca006ONNL2</th>\n",
       "      <td>6</td>\n",
       "      <td>H2</td>\n",
       "      <td>SA508</td>\n",
       "      <td>GACACCGT</td>\n",
       "      <td>SB702</td>\n",
       "      <td>CGAAGTAT</td>\n",
       "      <td>SB702SA508</td>\n",
       "      <td>CGAAGTAT-GACACCGT</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>H2</td>\n",
       "      <td>...</td>\n",
       "      <td>female</td>\n",
       "      <td>3/25/2015</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>3.0</td>\n",
       "      <td>34</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADNL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca006ONPN</th>\n",
       "      <td>6</td>\n",
       "      <td>F3</td>\n",
       "      <td>SA506</td>\n",
       "      <td>CGTGAGTG</td>\n",
       "      <td>SB703</td>\n",
       "      <td>TAGCAGCT</td>\n",
       "      <td>SB703SA506</td>\n",
       "      <td>TAGCAGCT-CGTGAGTG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>F3</td>\n",
       "      <td>...</td>\n",
       "      <td>female</td>\n",
       "      <td>3/25/2015</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>3.0</td>\n",
       "      <td>34</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>nares-AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca006ONPN2</th>\n",
       "      <td>6</td>\n",
       "      <td>H3</td>\n",
       "      <td>SA508</td>\n",
       "      <td>GACACCGT</td>\n",
       "      <td>SB703</td>\n",
       "      <td>TAGCAGCT</td>\n",
       "      <td>SB703SA508</td>\n",
       "      <td>TAGCAGCT-GACACCGT</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>H3</td>\n",
       "      <td>...</td>\n",
       "      <td>female</td>\n",
       "      <td>3/25/2015</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>3.0</td>\n",
       "      <td>34</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>nares-AD</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>502 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            PlateNumber PlateLocation     i5 i5Sequence     i7 i7Sequence  \\\n",
       "#sample-id                                                                  \n",
       "Ca009STL              1            A1  SA501   ATCGTACG  SA701   CGAGAGTT   \n",
       "900221                1            B1  SA502   ACTATCTG  SA701   CGAGAGTT   \n",
       "Ca010EBL              1            C1  SA503   TAGCGAGT  SA701   CGAGAGTT   \n",
       "900460                1            D1  SA504   CTGCGTGT  SA701   CGAGAGTT   \n",
       "900051                1            E1  SA505   TCATCGAG  SA701   CGAGAGTT   \n",
       "...                 ...           ...    ...        ...    ...        ...   \n",
       "Ca006ONL2             6            H1  SA508   GACACCGT  SB701   CTCGACTT   \n",
       "Ca006ONNL             6            F2  SA506   CGTGAGTG  SB702   CGAAGTAT   \n",
       "Ca006ONNL2            6            H2  SA508   GACACCGT  SB702   CGAAGTAT   \n",
       "Ca006ONPN             6            F3  SA506   CGTGAGTG  SB703   TAGCAGCT   \n",
       "Ca006ONPN2            6            H3  SA508   GACACCGT  SB703   TAGCAGCT   \n",
       "\n",
       "            identifier           Sequence      Plate ID Well location  ...  \\\n",
       "#sample-id                                                             ...   \n",
       "Ca009STL    SA701SA501  CGAGAGTT-ATCGTACG  1.010000e+21            A1  ...   \n",
       "900221      SA701SA502  CGAGAGTT-ACTATCTG  1.010000e+21            B1  ...   \n",
       "Ca010EBL    SA701SA503  CGAGAGTT-TAGCGAGT  1.010000e+21            C1  ...   \n",
       "900460      SA701SA504  CGAGAGTT-CTGCGTGT  1.010000e+21            D1  ...   \n",
       "900051      SA701SA505  CGAGAGTT-TCATCGAG  1.010000e+21            E1  ...   \n",
       "...                ...                ...           ...           ...  ...   \n",
       "Ca006ONL2   SB701SA508  CTCGACTT-GACACCGT  1.010000e+21            H1  ...   \n",
       "Ca006ONNL   SB702SA506  CGAAGTAT-CGTGAGTG  1.010000e+21            F2  ...   \n",
       "Ca006ONNL2  SB702SA508  CGAAGTAT-GACACCGT  1.010000e+21            H2  ...   \n",
       "Ca006ONPN   SB703SA506  TAGCAGCT-CGTGAGTG  1.010000e+21            F3  ...   \n",
       "Ca006ONPN2  SB703SA508  TAGCAGCT-GACACCGT  1.010000e+21            H3  ...   \n",
       "\n",
       "               sex enrolment_date enrolment_season hiv_exposure hiv_status  \\\n",
       "#sample-id                                                                   \n",
       "Ca009STL      male      4/16/2015          Autumn     Unexposed   negative   \n",
       "900221      female      8/11/2015           Winter    Unexposed   negative   \n",
       "Ca010EBL    female     11/20/2014           Spring    Unexposed   negative   \n",
       "900460      female      9/23/2015           Spring    Unexposed        NaN   \n",
       "900051        male      4/21/2015          Autumn     Unexposed   negative   \n",
       "...            ...            ...              ...          ...        ...   \n",
       "Ca006ONL2   female      3/25/2015           Autumn    Unexposed   negative   \n",
       "Ca006ONNL   female      3/25/2015           Autumn    Unexposed   negative   \n",
       "Ca006ONNL2  female      3/25/2015           Autumn    Unexposed   negative   \n",
       "Ca006ONPN   female      3/25/2015           Autumn    Unexposed   negative   \n",
       "Ca006ONPN2  female      3/25/2015           Autumn    Unexposed   negative   \n",
       "\n",
       "           household_size o_scorad  \\\n",
       "#sample-id                           \n",
       "Ca009STL              4.0       40   \n",
       "900221                7.0       34   \n",
       "Ca010EBL              7.0       21   \n",
       "900460                4.0       40   \n",
       "900051                7.0       41   \n",
       "...                   ...      ...   \n",
       "Ca006ONL2             3.0       34   \n",
       "Ca006ONNL             3.0       34   \n",
       "Ca006ONNL2            3.0       34   \n",
       "Ca006ONPN             3.0       34   \n",
       "Ca006ONPN2            3.0       34   \n",
       "\n",
       "                                                 FWD_filepath  \\\n",
       "#sample-id                                                      \n",
       "Ca009STL    /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "900221      /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "Ca010EBL    /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "900460      /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "900051      /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "...                                                       ...   \n",
       "Ca006ONL2   /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "Ca006ONNL   /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "Ca006ONNL2  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "Ca006ONPN   /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "Ca006ONPN2  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "\n",
       "                                                 REV_filepath      group  \n",
       "#sample-id                                                                \n",
       "Ca009STL    /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   skin-ADL  \n",
       "900221      /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   skin-ADL  \n",
       "Ca010EBL    /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   skin-ADL  \n",
       "900460      /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   skin-ADL  \n",
       "900051      /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   skin-ADL  \n",
       "...                                                       ...        ...  \n",
       "Ca006ONL2   /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   skin-ADL  \n",
       "Ca006ONNL   /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...  skin-ADNL  \n",
       "Ca006ONNL2  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...  skin-ADNL  \n",
       "Ca006ONPN   /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   nares-AD  \n",
       "Ca006ONPN2  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   nares-AD  \n",
       "\n",
       "[502 rows x 32 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load the metadata\n",
    "metadata_path = '../Data/Metadata/updated_clean_ant_skin_metadata.tab'\n",
    "metadata = pd.read_csv(metadata_path, sep='\\t')\n",
    "\n",
    "metadata['#sample-id'] = metadata['#sample-id'].str.replace('_', '')\n",
    "# Set Sample-ID as the index for the metadata dataframe \n",
    "metadata = metadata.set_index('#sample-id')\n",
    "\n",
    "\n",
    "# Create group column based on case_type to simplify group names\n",
    "metadata['group'] = metadata['case_type'].map({\n",
    "    'case-lesional skin': 'skin-ADL',\n",
    "    'case-nonlesional skin': 'skin-ADNL', \n",
    "    'control-nonlesional skin': 'skin-H',\n",
    "    'case-anterior nares': 'nares-AD',\n",
    "    'control-anterior nares': 'nares-H'\n",
    "})\n",
    "\n",
    "metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "O-SCORAD distribution for skin-ADL:\n",
      "o_scorad\n",
      "21    4\n",
      "22    2\n",
      "23    3\n",
      "24    2\n",
      "25    4\n",
      "26    2\n",
      "28    3\n",
      "29    2\n",
      "30    3\n",
      "32    4\n",
      "33    2\n",
      "34    5\n",
      "35    1\n",
      "36    3\n",
      "37    2\n",
      "38    2\n",
      "39    3\n",
      "40    4\n",
      "41    4\n",
      "42    5\n",
      "43    7\n",
      "44    9\n",
      "45    1\n",
      "47    1\n",
      "48    1\n",
      "49    1\n",
      "51    1\n",
      "52    1\n",
      "53    3\n",
      "54    3\n",
      "56    2\n",
      "58    2\n",
      "61    2\n",
      "66    2\n",
      "67    1\n",
      "75    2\n",
      "78    4\n",
      "82    1\n",
      "NR    1\n",
      "Name: count, dtype: int64\n",
      "\n",
      "O-SCORAD distribution for skin-ADNL:\n",
      "o_scorad\n",
      "10    1\n",
      "21    4\n",
      "22    2\n",
      "23    3\n",
      "24    2\n",
      "25    6\n",
      "26    2\n",
      "28    4\n",
      "29    2\n",
      "30    3\n",
      "32    4\n",
      "33    2\n",
      "34    5\n",
      "35    1\n",
      "36    3\n",
      "37    2\n",
      "38    2\n",
      "39    3\n",
      "40    4\n",
      "41    4\n",
      "42    5\n",
      "43    7\n",
      "44    9\n",
      "45    1\n",
      "47    1\n",
      "48    1\n",
      "49    2\n",
      "51    1\n",
      "52    1\n",
      "53    3\n",
      "54    3\n",
      "56    2\n",
      "58    2\n",
      "61    2\n",
      "66    2\n",
      "67    1\n",
      "78    2\n",
      "NR    1\n",
      "Name: count, dtype: int64\n",
      "\n",
      "O-SCORAD distribution for nares-AD:\n",
      "o_scorad\n",
      "21    4\n",
      "22    2\n",
      "23    3\n",
      "24    2\n",
      "25    5\n",
      "26    2\n",
      "28    4\n",
      "29    2\n",
      "30    3\n",
      "32    4\n",
      "33    2\n",
      "34    5\n",
      "35    1\n",
      "36    3\n",
      "37    2\n",
      "38    1\n",
      "39    3\n",
      "40    4\n",
      "41    4\n",
      "42    5\n",
      "43    6\n",
      "44    9\n",
      "45    1\n",
      "47    1\n",
      "49    1\n",
      "51    1\n",
      "52    1\n",
      "53    3\n",
      "54    3\n",
      "56    2\n",
      "58    2\n",
      "61    2\n",
      "66    2\n",
      "67    1\n",
      "75    2\n",
      "78    3\n",
      "82    1\n",
      "Name: count, dtype: int64\n",
      "\n",
      "O-SCORAD distribution for skin-H:\n",
      "o_scorad\n",
      "43    1\n",
      "Name: count, dtype: int64\n",
      "\n",
      "O-SCORAD distribution for nares-H:\n",
      "Series([], Name: count, dtype: int64)\n"
     ]
    }
   ],
   "source": [
    "# Check o_scorad distribution across different groups\n",
    "for group in metadata['group'].unique():\n",
    "    print(f\"\\nO-SCORAD distribution for {group}:\")\n",
    "    print(metadata[metadata['group'] == group]['o_scorad'].value_counts().sort_index())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# split data into training and testing\n",
    "# ------------------------------\n",
    "# 1. Custom Group-Stratified K-Fold Function\n",
    "# ------------------------------\n",
    "def group_stratified_kfold(X, y, groups, n_splits=5, random_state=42):\n",
    "    unique_groups = np.unique(groups)\n",
    "    np.random.seed(random_state)\n",
    "    np.random.shuffle(unique_groups)\n",
    "    \n",
    "    group_label_dist = {}\n",
    "    for group in unique_groups:\n",
    "        group_mask = groups == group\n",
    "        group_y = y[group_mask]\n",
    "        group_label_dist[group] = {label: np.sum(group_y == label) for label in np.unique(y)}\n",
    "    \n",
    "    folds = [[] for _ in range(n_splits)]\n",
    "    fold_label_dist = [{label: 0 for label in np.unique(y)} for _ in range(n_splits)]\n",
    "    \n",
    "    sorted_groups = sorted(unique_groups, key=lambda g: sum(groups == g), reverse=True)\n",
    "    \n",
    "    for group in sorted_groups:\n",
    "        best_fold = 0\n",
    "        min_imbalance = float('inf')\n",
    "        \n",
    "        for fold_idx in range(n_splits):\n",
    "            temp_fold_dist = fold_label_dist[fold_idx].copy()\n",
    "            for label, count in group_label_dist[group].items():\n",
    "                temp_fold_dist[label] += count\n",
    "            fold_size = sum(temp_fold_dist.values())\n",
    "            proportions = [count / fold_size for count in temp_fold_dist.values()] if fold_size else [0]*len(temp_fold_dist)\n",
    "            imbalance = np.var(proportions) + fold_size / (sum(groups.shape) / n_splits)\n",
    "            \n",
    "            if imbalance < min_imbalance:\n",
    "                min_imbalance = imbalance\n",
    "                best_fold = fold_idx\n",
    "        \n",
    "        folds[best_fold].extend(np.where(groups == group)[0])\n",
    "        for label, count in group_label_dist[group].items():\n",
    "            fold_label_dist[best_fold][label] += count\n",
    "    \n",
    "    train_test_indices = []\n",
    "    for i in range(n_splits):\n",
    "        test_idx = np.array(folds[i])\n",
    "        train_idx = np.concatenate([folds[j] for j in range(n_splits) if j != i])\n",
    "        train_test_indices.append((train_idx, test_idx))\n",
    "    \n",
    "    return train_test_indices\n",
    "\n",
    "# ------------------------------\n",
    "# 2. Function to Run CV and Get ROC/Feature Importance\n",
    "# ------------------------------\n",
    "def run_group_stratified_cv(X, y, groups, n_splits=5):\n",
    "    folds = group_stratified_kfold(X, y, groups, n_splits=n_splits)\n",
    "    \n",
    "    cv_results = []\n",
    "    feature_importances = pd.DataFrame(index=X.columns)\n",
    "    fold_aucs = []\n",
    "    \n",
    "    for i, (train_idx, test_idx) in enumerate(folds):\n",
    "        X_train, X_test = X.iloc[train_idx], X.iloc[test_idx]\n",
    "        y_train, y_test = y.iloc[train_idx], y.iloc[test_idx]\n",
    "        \n",
    "        if len(np.unique(y_train)) < 2 or len(np.unique(y_test)) < 2:\n",
    "            print(f\"Skipping fold {i+1} due to insufficient class representation\")\n",
    "            continue\n",
    "            \n",
    "        clf = RandomForestClassifier(n_estimators=1000, random_state=42)\n",
    "        clf.fit(X_train, y_train)\n",
    "        probas = clf.predict_proba(X_test)\n",
    "        \n",
    "        feature_importances[f'fold_{i}'] = clf.feature_importances_\n",
    "        fpr, tpr, _ = roc_curve(y_test, probas[:, 1])\n",
    "        roc_auc = auc(fpr, tpr)\n",
    "        fold_aucs.append(roc_auc)\n",
    "        \n",
    "        cv_results.append({\n",
    "            'y_true': y_test,\n",
    "            'y_proba': probas[:, 1],\n",
    "            'fpr': fpr,\n",
    "            'tpr': tpr,\n",
    "            'auc': roc_auc,\n",
    "            'fold': i\n",
    "        })\n",
    "    \n",
    "    feature_importances['mean_importance'] = feature_importances.mean(axis=1)\n",
    "    feature_importances['std_importance'] = feature_importances.std(axis=1)\n",
    "    feature_importances = feature_importances.sort_values('mean_importance', ascending=False)\n",
    "    \n",
    "    return cv_results, feature_importances, fold_aucs\n",
    "\n",
    "# ------------------------------\n",
    "# 3. Regenerate ROC Results\n",
    "# ------------------------------\n",
    "# Make sure your df (feature table) and metadata DataFrame are in scope.\n",
    "# Here, 'df' is assumed to contain your features (ASV relative abundances) \n",
    "# and 'metadata' is assumed to contain columns: 'group', 'pid', and 'microbiome_type' if applicable.\n",
    "\n",
    "# 3a. Skin vs Nares (Binary Classification)\n",
    "meta_skin_nares = metadata[metadata['group'].str.startswith(('skin', 'nares'))].copy()\n",
    "meta_skin_nares['site_label'] = meta_skin_nares['group'].apply(lambda x: 0 if x.startswith('skin') else 1)\n",
    "common_samples = df.index.intersection(meta_skin_nares.index)\n",
    "X_skin_nares = df.loc[common_samples]\n",
    "y_skin_nares = meta_skin_nares.loc[common_samples, 'site_label']\n",
    "groups_skin_nares = meta_skin_nares.loc[common_samples, 'pid']\n",
    "y_skin_nares = pd.to_numeric(y_skin_nares, errors='coerce')\n",
    "skin_vs_nares_cv_results, _, _ = run_group_stratified_cv(X_skin_nares, y_skin_nares, groups_skin_nares, n_splits=3)\n",
    "\n",
    "# 3b. Skin Comparisons (Pairwise Among Skin Groups)\n",
    "skin_comparisons = [\n",
    "    ('skin-ADL', 'skin-H'),\n",
    "    ('skin-ADNL', 'skin-ADL'),\n",
    "    ('skin-ADNL', 'skin-H')\n",
    "]\n",
    "skin_cv_results_dict = {}\n",
    "for label1, label2 in skin_comparisons:\n",
    "    meta_subset = metadata[metadata['group'].isin([label1, label2])]\n",
    "    common_samples = df.index.intersection(meta_subset.index)\n",
    "    X_skin = df.loc[common_samples]\n",
    "    # Map to binary: 0 for first label, 1 for second label\n",
    "    y_skin = meta_subset.loc[common_samples, 'group'].map({label1: 0, label2: 1})\n",
    "    groups_skin = meta_subset.loc[common_samples, 'pid']\n",
    "    cv_results, _, _ = run_group_stratified_cv(X_skin, y_skin, groups_skin, n_splits=3)\n",
    "    key = f\"{label1}_vs_{label2}\"\n",
    "    skin_cv_results_dict[key] = cv_results\n",
    "\n",
    "# 3c. Nares Comparison (nares-AD vs nares-H)\n",
    "meta_nares = metadata[metadata['group'].isin(['nares-AD', 'nares-H'])]\n",
    "common_samples = df.index.intersection(meta_nares.index)\n",
    "X_nares = df.loc[common_samples]\n",
    "y_nares = meta_nares.loc[common_samples, 'group'].map({'nares-AD': 0, 'nares-H': 1})\n",
    "groups_nares = meta_nares.loc[common_samples, 'pid']\n",
    "nares_ad_vs_h_cv_results, _, _ = run_group_stratified_cv(X_nares, y_nares, groups_nares, n_splits=3)\n",
    "\n",
    "# ------------------------------\n",
    "# 4. Assemble ROC Results Dictionary\n",
    "# ------------------------------\n",
    "roc_results_dict = {\n",
    "    'Prediction of Skin vs. Nares Samples': {\n",
    "        'skin_vs_nares': skin_vs_nares_cv_results\n",
    "    },\n",
    "    'Prediction of AD Status From Skin Samples': skin_cv_results_dict,\n",
    "    'Prediction of AD Status From Nares Samples': {\n",
    "        'nares-AD_vs_nares-H': nares_ad_vs_h_cv_results\n",
    "    }\n",
    "}\n",
    "\n",
    "\n",
    "# ------------------------------\n",
    "# 5. Combined ROC Plotting Function (3-Panel)\n",
    "# ------------------------------\n",
    "def plot_combined_roc_3panel(roc_results_dict, comparison_order, color_map, output_path=None):\n",
    "    \"\"\"\n",
    "    Plots a 3-panel ROC curve figure from grouped ROC results.\n",
    "    \n",
    "    Parameters:\n",
    "    - roc_results_dict: dict\n",
    "        A dictionary mapping panel titles to sub-dicts. Each sub-dict maps sublabels \n",
    "        to a list of cv_results dictionaries.\n",
    "    - comparison_order: list\n",
    "        List of panel titles in order.\n",
    "    - color_map: dict\n",
    "        Dictionary mapping each sublabel to a color.\n",
    "    - output_path: str or None\n",
    "        If provided, path to save the figure.\n",
    "    \"\"\"\n",
    "    fig, axs = plt.subplots(1, 3, figsize=(18, 6), sharey=True)\n",
    "\n",
    "    pretty_labels = {\n",
    "    'skin_vs_nares': 'All Skin vs All Nares',\n",
    "    'skin-ADL_vs_skin-H': 'Skin ADL vs Skin H',\n",
    "    'skin-ADNL_vs_skin-ADL': 'Skin ADNL vs Skin H',\n",
    "    'skin-ADNL_vs_skin-H': 'Skin ADNL vs Skin ADL',\n",
    "    'nares-AD_vs_nares-H': 'Nares AD vs Nares H'\n",
    "}\n",
    "    \n",
    "    for i, panel_title in enumerate(comparison_order):\n",
    "        ax = axs[i]\n",
    "        sub_dict = roc_results_dict[panel_title]\n",
    "        # Loop over each sub-comparison within this panel\n",
    "        for sublabel, curves in sub_dict.items():\n",
    "            # Ensure curves is a list\n",
    "            if not isinstance(curves, list):\n",
    "                curves = [curves]\n",
    "            mean_fpr = np.linspace(0, 1, 100)\n",
    "            tprs = []\n",
    "            aucs = []\n",
    "            for result in curves:\n",
    "                # Ensure the result dict has fpr and tpr keys\n",
    "                if 'fpr' not in result or 'tpr' not in result:\n",
    "                    continue\n",
    "                tpr_interp = np.interp(mean_fpr, result['fpr'], result['tpr'])\n",
    "                tpr_interp[0] = 0.0\n",
    "                tprs.append(tpr_interp)\n",
    "                aucs.append(result['auc'])\n",
    "            if len(tprs) == 0:\n",
    "                continue\n",
    "            mean_tpr = np.mean(tprs, axis=0)\n",
    "            mean_tpr[-1] = 1.0\n",
    "            std_tpr = np.std(tprs, axis=0)\n",
    "            tprs_upper = np.minimum(mean_tpr + std_tpr, 1)\n",
    "            tprs_lower = np.maximum(mean_tpr - std_tpr, 0)\n",
    "            ax.plot(mean_fpr, mean_tpr, lw=3,\n",
    "                    label=f'{pretty_labels.get(sublabel, sublabel)} (AUC = {np.mean(aucs):.2f} ± {np.std(aucs):.2f})',\n",
    "                    color=color_map.get(sublabel, 'gray'))\n",
    "            ax.fill_between(mean_fpr, tprs_lower, tprs_upper, alpha=0.3,\n",
    "                            color=color_map.get(sublabel, 'gray'))\n",
    "    \n",
    "        ax.plot([0, 1], [0, 1], 'k--', lw=1)\n",
    "        ax.set_xlim([0.0, 1.0])\n",
    "        ax.set_ylim([0.0, 1.05])\n",
    "        ax.set_xlabel('False Positive Rate', fontsize=16)\n",
    "        ax.set_title(panel_title, fontsize=18)\n",
    "        ax.grid(True, linestyle='--', alpha=0.7)\n",
    "        ax.tick_params(axis='both', labelsize=14)\n",
    "        ax.legend(loc='lower right', fontsize=12)\n",
    "    \n",
    "    axs[0].set_ylabel('True Positive Rate', fontsize=16)\n",
    "    plt.suptitle('Random Forest Classifications by 16S ASVs', fontsize=26, x=0.5, y=1)\n",
    "    plt.tight_layout()\n",
    "    \n",
    "    if output_path:\n",
    "        plt.savefig(output_path, dpi=600, bbox_inches='tight')\n",
    "\n",
    "# ------------------------------\n",
    "# 6. Define Panel Order and Color Map, Then Plot\n",
    "# ------------------------------\n",
    "comparison_order = [\n",
    "    'Prediction of Skin vs. Nares Samples',\n",
    "    'Prediction of AD Status From Skin Samples',\n",
    "    'Prediction of AD Status From Nares Samples'\n",
    "]\n",
    "color_map = {\n",
    "    'skin_vs_nares': 'black',\n",
    "    'skin-ADL_vs_skin-H': 'blue',\n",
    "    'skin-ADNL_vs_skin-ADL': 'green',\n",
    "    'skin-ADNL_vs_skin-H': 'purple',\n",
    "    'nares-AD_vs_nares-H': 'orange'\n",
    "}\n",
    "\n",
    "plot_combined_roc_3panel(\n",
    "    roc_results_dict,\n",
    "    comparison_order=comparison_order,\n",
    "    color_map=color_map,\n",
    "    # output_path='../Plots/Analysis_figures/Random_Forest/rf_classifications_ROC_curves_filtered_by_prevalence_1pct_rare_Genus-ASV-non-collapse.png'\n",
    "    output_path = '../Plots/Analysis_figures/Random_Forest/rf_classifications_final.png'\n",
    ")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Combined skin and nares classification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/sklearn/utils/validation.py:869: UserWarning: pandas.DataFrame with sparse columns found.It will be converted to a dense numpy array.\n",
      "  warnings.warn(\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/sklearn/utils/validation.py:869: UserWarning: pandas.DataFrame with sparse columns found.It will be converted to a dense numpy array.\n",
      "  warnings.warn(\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/sklearn/utils/validation.py:869: UserWarning: pandas.DataFrame with sparse columns found.It will be converted to a dense numpy array.\n",
      "  warnings.warn(\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/sklearn/utils/validation.py:869: UserWarning: pandas.DataFrame with sparse columns found.It will be converted to a dense numpy array.\n",
      "  warnings.warn(\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/sklearn/utils/validation.py:869: UserWarning: pandas.DataFrame with sparse columns found.It will be converted to a dense numpy array.\n",
      "  warnings.warn(\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/sklearn/utils/validation.py:869: UserWarning: pandas.DataFrame with sparse columns found.It will be converted to a dense numpy array.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[{'y_true': 9004002    1\n",
       "  9004022    1\n",
       "  900400     1\n",
       "  900402     1\n",
       "  9003932    0\n",
       "            ..\n",
       "  900466     1\n",
       "  900463     1\n",
       "  900469     1\n",
       "  900224     1\n",
       "  900258     1\n",
       "  Name: combined_label, Length: 130, dtype: int64,\n",
       "  'y_proba': array([0.733, 0.786, 0.396, 0.552, 0.226, 0.585, 0.432, 0.295, 0.501,\n",
       "         0.491, 0.61 , 0.287, 0.386, 0.596, 0.649, 0.318, 0.527, 0.623,\n",
       "         0.483, 0.404, 0.512, 0.372, 0.5  , 0.524, 0.503, 0.638, 0.35 ,\n",
       "         0.5  , 0.598, 0.449, 0.364, 0.462, 0.721, 0.501, 0.744, 0.216,\n",
       "         0.275, 0.724, 0.293, 0.762, 0.493, 0.472, 0.718, 0.503, 0.726,\n",
       "         0.527, 0.7  , 0.33 , 0.501, 0.663, 0.359, 0.449, 0.294, 0.675,\n",
       "         0.885, 0.702, 0.392, 0.392, 0.324, 0.576, 0.605, 0.272, 0.549,\n",
       "         0.639, 0.445, 0.848, 0.514, 0.438, 0.731, 0.2  , 0.492, 0.735,\n",
       "         0.711, 0.506, 0.565, 0.511, 0.63 , 0.609, 0.674, 0.528, 0.258,\n",
       "         0.256, 0.39 , 0.258, 0.54 , 0.291, 0.501, 0.498, 0.789, 0.526,\n",
       "         0.276, 0.765, 0.814, 0.463, 0.556, 0.735, 0.513, 0.414, 0.33 ,\n",
       "         0.78 , 0.369, 0.416, 0.407, 0.554, 0.621, 0.198, 0.608, 0.568,\n",
       "         0.581, 0.247, 0.395, 0.4  , 0.252, 0.749, 0.371, 0.602, 0.44 ,\n",
       "         0.688, 0.408, 0.339, 0.29 , 0.43 , 0.606, 0.738, 0.837, 0.583,\n",
       "         0.661, 0.771, 0.522, 0.749]),\n",
       "  'fpr': array([0.        , 0.        , 0.01785714, 0.01785714, 0.03571429,\n",
       "         0.05357143, 0.05357143, 0.05357143, 0.05357143, 0.07142857,\n",
       "         0.07142857, 0.08928571, 0.08928571, 0.10714286, 0.10714286,\n",
       "         0.14285714, 0.14285714, 0.17857143, 0.17857143, 0.21428571,\n",
       "         0.21428571, 0.25      , 0.25      , 0.30357143, 0.30357143,\n",
       "         0.32142857, 0.32142857, 0.33928571, 0.33928571, 0.375     ,\n",
       "         0.375     , 0.39285714, 0.39285714, 0.41071429, 0.41071429,\n",
       "         0.42857143, 0.46428571, 0.46428571, 0.46428571, 0.48214286,\n",
       "         0.48214286, 0.51785714, 0.51785714, 0.53571429, 0.53571429,\n",
       "         0.55357143, 0.55357143, 0.57142857, 0.57142857, 0.58928571,\n",
       "         0.60714286, 0.60714286, 0.625     , 0.625     , 0.64285714,\n",
       "         0.64285714, 0.66071429, 0.66071429, 0.69642857, 0.73214286,\n",
       "         0.75      , 0.75      , 0.80357143, 0.80357143, 0.89285714,\n",
       "         0.91071429, 0.91071429, 1.        ]),\n",
       "  'tpr': array([0.        , 0.01351351, 0.01351351, 0.12162162, 0.13513514,\n",
       "         0.13513514, 0.14864865, 0.17567568, 0.21621622, 0.21621622,\n",
       "         0.31081081, 0.31081081, 0.33783784, 0.33783784, 0.35135135,\n",
       "         0.35135135, 0.39189189, 0.39189189, 0.40540541, 0.40540541,\n",
       "         0.41891892, 0.41891892, 0.43243243, 0.43243243, 0.48648649,\n",
       "         0.48648649, 0.5       , 0.5       , 0.52702703, 0.52702703,\n",
       "         0.54054054, 0.54054054, 0.56756757, 0.56756757, 0.58108108,\n",
       "         0.59459459, 0.62162162, 0.64864865, 0.67567568, 0.67567568,\n",
       "         0.74324324, 0.74324324, 0.75675676, 0.75675676, 0.77027027,\n",
       "         0.77027027, 0.7972973 , 0.7972973 , 0.86486486, 0.86486486,\n",
       "         0.87837838, 0.89189189, 0.89189189, 0.90540541, 0.90540541,\n",
       "         0.91891892, 0.91891892, 0.93243243, 0.93243243, 0.93243243,\n",
       "         0.93243243, 0.94594595, 0.94594595, 0.95945946, 0.95945946,\n",
       "         0.97297297, 1.        , 1.        ]),\n",
       "  'auc': 0.6737451737451737,\n",
       "  'fold': 0},\n",
       " {'y_true': Ca006ONPN     1\n",
       "  Ca006ONL      1\n",
       "  Ca006ONL2     1\n",
       "  Ca006ONPN2    1\n",
       "  900380        0\n",
       "               ..\n",
       "  900502        1\n",
       "  900500        1\n",
       "  900507        1\n",
       "  900369        0\n",
       "  900575        1\n",
       "  Name: combined_label, Length: 131, dtype: int64,\n",
       "  'y_proba': array([0.475, 0.657, 0.61 , 0.648, 0.376, 0.575, 0.594, 0.241, 0.677,\n",
       "         0.734, 0.436, 0.413, 0.432, 0.473, 0.513, 0.521, 0.323, 0.498,\n",
       "         0.538, 0.764, 0.663, 0.4  , 0.698, 0.495, 0.424, 0.317, 0.746,\n",
       "         0.523, 0.721, 0.686, 0.486, 0.621, 0.679, 0.646, 0.345, 0.422,\n",
       "         0.452, 0.721, 0.433, 0.512, 0.658, 0.539, 0.461, 0.62 , 0.212,\n",
       "         0.617, 0.707, 0.61 , 0.529, 0.785, 0.723, 0.41 , 0.396, 0.622,\n",
       "         0.348, 0.445, 0.543, 0.599, 0.306, 0.708, 0.54 , 0.593, 0.788,\n",
       "         0.808, 0.684, 0.425, 0.416, 0.357, 0.224, 0.315, 0.368, 0.357,\n",
       "         0.402, 0.689, 0.448, 0.599, 0.763, 0.673, 0.596, 0.739, 0.554,\n",
       "         0.815, 0.596, 0.507, 0.391, 0.189, 0.511, 0.747, 0.295, 0.737,\n",
       "         0.587, 0.682, 0.307, 0.329, 0.632, 0.521, 0.711, 0.637, 0.39 ,\n",
       "         0.339, 0.722, 0.536, 0.72 , 0.514, 0.261, 0.478, 0.506, 0.25 ,\n",
       "         0.702, 0.536, 0.567, 0.671, 0.749, 0.654, 0.423, 0.617, 0.598,\n",
       "         0.644, 0.642, 0.473, 0.725, 0.771, 0.581, 0.545, 0.582, 0.638,\n",
       "         0.792, 0.666, 0.588, 0.269, 0.566]),\n",
       "  'fpr': array([0.        , 0.        , 0.        , 0.01666667, 0.03333333,\n",
       "         0.03333333, 0.05      , 0.05      , 0.08333333, 0.08333333,\n",
       "         0.1       , 0.1       , 0.11666667, 0.11666667, 0.13333333,\n",
       "         0.13333333, 0.15      , 0.15      , 0.16666667, 0.16666667,\n",
       "         0.2       , 0.21666667, 0.21666667, 0.25      , 0.25      ,\n",
       "         0.25      , 0.28333333, 0.28333333, 0.31666667, 0.31666667,\n",
       "         0.36666667, 0.36666667, 0.36666667, 0.36666667, 0.38333333,\n",
       "         0.43333333, 0.43333333, 0.45      , 0.45      , 0.5       ,\n",
       "         0.5       , 0.51666667, 0.56666667, 0.56666667, 0.6       ,\n",
       "         0.6       , 0.66666667, 0.66666667, 0.68333333, 0.68333333,\n",
       "         0.71666667, 0.71666667, 0.78333333, 0.8       , 0.8       ,\n",
       "         0.85      , 0.85      , 0.91666667, 0.91666667, 0.96666667,\n",
       "         0.96666667, 0.98333333, 0.98333333, 1.        ]),\n",
       "  'tpr': array([0.        , 0.01408451, 0.23943662, 0.25352113, 0.25352113,\n",
       "         0.26760563, 0.26760563, 0.29577465, 0.29577465, 0.30985915,\n",
       "         0.30985915, 0.32394366, 0.32394366, 0.38028169, 0.38028169,\n",
       "         0.43661972, 0.43661972, 0.45070423, 0.45070423, 0.50704225,\n",
       "         0.50704225, 0.52112676, 0.54929577, 0.54929577, 0.56338028,\n",
       "         0.5915493 , 0.5915493 , 0.63380282, 0.63380282, 0.67605634,\n",
       "         0.67605634, 0.70422535, 0.73239437, 0.76056338, 0.77464789,\n",
       "         0.77464789, 0.8028169 , 0.8028169 , 0.81690141, 0.81690141,\n",
       "         0.83098592, 0.84507042, 0.84507042, 0.85915493, 0.85915493,\n",
       "         0.87323944, 0.87323944, 0.88732394, 0.88732394, 0.90140845,\n",
       "         0.90140845, 0.91549296, 0.91549296, 0.92957746, 0.94366197,\n",
       "         0.94366197, 0.95774648, 0.95774648, 0.97183099, 0.97183099,\n",
       "         0.98591549, 0.98591549, 1.        , 1.        ]),\n",
       "  'auc': 0.7355633802816901,\n",
       "  'fold': 1},\n",
       " {'y_true': 900397     1\n",
       "  900399     1\n",
       "  9003992    1\n",
       "  9003972    1\n",
       "  900391     0\n",
       "            ..\n",
       "  900366     0\n",
       "  900367     0\n",
       "  900343     0\n",
       "  900126     1\n",
       "  900363     0\n",
       "  Name: combined_label, Length: 130, dtype: int64,\n",
       "  'y_proba': array([0.724, 0.663, 0.804, 0.747, 0.69 , 0.417, 0.314, 0.546, 0.52 ,\n",
       "         0.57 , 0.693, 0.454, 0.639, 0.586, 0.483, 0.497, 0.488, 0.61 ,\n",
       "         0.634, 0.838, 0.459, 0.419, 0.432, 0.426, 0.452, 0.463, 0.747,\n",
       "         0.641, 0.448, 0.341, 0.465, 0.678, 0.669, 0.622, 0.755, 0.463,\n",
       "         0.704, 0.838, 0.782, 0.364, 0.501, 0.645, 0.5  , 0.474, 0.591,\n",
       "         0.676, 0.8  , 0.507, 0.464, 0.326, 0.368, 0.732, 0.565, 0.595,\n",
       "         0.522, 0.408, 0.383, 0.617, 0.569, 0.597, 0.756, 0.587, 0.579,\n",
       "         0.502, 0.616, 0.512, 0.243, 0.655, 0.742, 0.495, 0.554, 0.669,\n",
       "         0.679, 0.735, 0.769, 0.43 , 0.588, 0.576, 0.313, 0.466, 0.608,\n",
       "         0.566, 0.466, 0.357, 0.338, 0.609, 0.426, 0.619, 0.458, 0.59 ,\n",
       "         0.59 , 0.613, 0.544, 0.661, 0.665, 0.548, 0.666, 0.56 , 0.554,\n",
       "         0.658, 0.631, 0.416, 0.374, 0.412, 0.669, 0.476, 0.3  , 0.799,\n",
       "         0.8  , 0.844, 0.735, 0.611, 0.7  , 0.326, 0.527, 0.478, 0.446,\n",
       "         0.377, 0.675, 0.465, 0.719, 0.764, 0.654, 0.382, 0.43 , 0.562,\n",
       "         0.589, 0.355, 0.613, 0.451]),\n",
       "  'fpr': array([0.        , 0.        , 0.        , 0.        , 0.01666667,\n",
       "         0.01666667, 0.01666667, 0.01666667, 0.01666667, 0.01666667,\n",
       "         0.03333333, 0.03333333, 0.06666667, 0.06666667, 0.08333333,\n",
       "         0.08333333, 0.1       , 0.1       , 0.11666667, 0.11666667,\n",
       "         0.13333333, 0.15      , 0.15      , 0.18333333, 0.18333333,\n",
       "         0.2       , 0.2       , 0.21666667, 0.21666667, 0.25      ,\n",
       "         0.25      , 0.26666667, 0.26666667, 0.28333333, 0.28333333,\n",
       "         0.3       , 0.36666667, 0.36666667, 0.4       , 0.4       ,\n",
       "         0.43333333, 0.43333333, 0.45      , 0.45      , 0.5       ,\n",
       "         0.53333333, 0.53333333, 0.53333333, 0.53333333, 0.61666667,\n",
       "         0.63333333, 0.66666667, 0.71666667, 0.71666667, 0.9       ,\n",
       "         0.93333333, 1.        ]),\n",
       "  'tpr': array([0.        , 0.01428571, 0.04285714, 0.05714286, 0.07142857,\n",
       "         0.15714286, 0.18571429, 0.2       , 0.22857143, 0.31428571,\n",
       "         0.31428571, 0.34285714, 0.34285714, 0.38571429, 0.38571429,\n",
       "         0.48571429, 0.48571429, 0.5       , 0.5       , 0.55714286,\n",
       "         0.55714286, 0.57142857, 0.58571429, 0.58571429, 0.62857143,\n",
       "         0.62857143, 0.65714286, 0.65714286, 0.71428571, 0.71428571,\n",
       "         0.72857143, 0.72857143, 0.74285714, 0.74285714, 0.75714286,\n",
       "         0.77142857, 0.77142857, 0.78571429, 0.78571429, 0.8       ,\n",
       "         0.8       , 0.82857143, 0.82857143, 0.85714286, 0.85714286,\n",
       "         0.88571429, 0.9       , 0.92857143, 0.97142857, 0.97142857,\n",
       "         0.98571429, 0.98571429, 0.98571429, 1.        , 1.        ,\n",
       "         1.        , 1.        ]),\n",
       "  'auc': 0.814047619047619,\n",
       "  'fold': 2}]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Define combined dataset (Skin ADL + Nares AD vs Skin H + Nares H)\n",
    "meta_skin_nares_AD = metadata[metadata['group'].isin(['skin-ADL', 'nares-AD'])].copy()\n",
    "meta_skin_nares_H = metadata[metadata['group'].isin(['skin-H', 'nares-H'])].copy()\n",
    "meta_combined = pd.concat([meta_skin_nares_AD, meta_skin_nares_H])\n",
    "\n",
    "# Label as 1 for AD, 0 for Healthy\n",
    "meta_combined['combined_label'] = meta_combined['group'].map({\n",
    "    'skin-ADL': 1,\n",
    "    'nares-AD': 1,\n",
    "    'skin-H': 0,\n",
    "    'nares-H': 0\n",
    "})\n",
    "meta_combined['site'] = meta_combined['group'].apply(lambda x: 'skin' if 'skin' in x else 'nares')\n",
    "\n",
    "# Get matching features\n",
    "common_samples = df.index.intersection(meta_combined.index)\n",
    "X_combined = df.loc[common_samples]\n",
    "y_combined = meta_combined.loc[common_samples, 'combined_label']\n",
    "groups_combined = meta_combined.loc[common_samples, 'pid']\n",
    "\n",
    "# One-hot encode the site and combine with original features\n",
    "site_dummies = pd.get_dummies(meta_combined.loc[common_samples, 'site'], prefix='site')\n",
    "X_combined_with_site = pd.concat([X_combined.reset_index(drop=True), site_dummies.reset_index(drop=True)], axis=1)\n",
    "\n",
    "# Run classification using the enhanced feature set\n",
    "combined_AD_vs_H_results, combined_feature_importance, combined_aucs = run_group_stratified_cv(\n",
    "    X_combined_with_site, y_combined, groups_combined, n_splits=3\n",
    ")\n",
    "\n",
    "combined_AD_vs_H_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_single_roc(cv_results, label, color='black', output_path=None):\n",
    "    \"\"\"\n",
    "    Plot a ROC curve from cross-validation results.\n",
    "\n",
    "    Parameters:\n",
    "    - cv_results: list of dicts from run_group_stratified_cv\n",
    "    - label: str, label for legend\n",
    "    - color: str, line color\n",
    "    - output_path: str, optional path to save plot\n",
    "    \"\"\"\n",
    "    mean_fpr = np.linspace(0, 1, 100)\n",
    "    tprs, aucs = [], []\n",
    "\n",
    "    for result in cv_results:\n",
    "        fpr, tpr, auc_val = result['fpr'], result['tpr'], result['auc']\n",
    "        interp_tpr = np.interp(mean_fpr, fpr, tpr)\n",
    "        interp_tpr[0] = 0.0\n",
    "        tprs.append(interp_tpr)\n",
    "        aucs.append(auc_val)\n",
    "\n",
    "    mean_tpr = np.mean(tprs, axis=0)\n",
    "    mean_tpr[-1] = 1.0\n",
    "    std_tpr = np.std(tprs, axis=0)\n",
    "    tprs_upper = np.minimum(mean_tpr + std_tpr, 1)\n",
    "    tprs_lower = np.maximum(mean_tpr - std_tpr, 0)\n",
    "\n",
    "    plt.figure(figsize=(6, 6))\n",
    "    plt.plot(mean_fpr, mean_tpr, color=color, lw=3,\n",
    "             label=f'{label} (AUC = {np.mean(aucs):.2f} ± {np.std(aucs):.2f})')\n",
    "    plt.fill_between(mean_fpr, tprs_lower, tprs_upper, color=color, alpha=0.3)\n",
    "    plt.plot([0, 1], [0, 1], 'k--', lw=1)\n",
    "\n",
    "    plt.xlabel('False Positive Rate', fontsize=14)\n",
    "    plt.ylabel('True Positive Rate', fontsize=14)\n",
    "    plt.title('ROC Curve: Combined Skin+Nares AD vs H', fontsize=16)\n",
    "    plt.legend(loc='lower right', fontsize=12)\n",
    "    plt.grid(True, linestyle='--', alpha=0.7)\n",
    "    plt.tight_layout()\n",
    "\n",
    "    if output_path:\n",
    "        plt.savefig(output_path, dpi=600, bbox_inches='tight')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_single_roc(\n",
    "    combined_AD_vs_H_results,\n",
    "    label='Skin+Nares AD vs H',\n",
    "    color='darkred',\n",
    "    output_path='../Plots/Analysis_figures/Random_Forest/roc_combined_skin_nares_AD_vs_H.png'\n",
    ")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Weighted skin and nares classification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "skin_weight = 1.18\n",
    "nares_weight = 0.82\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_group_stratified_cv(X, y, groups, site_series, n_splits=5):\n",
    "    \"\"\"\n",
    "    Run group-stratified CV using RandomForestClassifier, with site-based sample weights.\n",
    "\n",
    "    Parameters:\n",
    "    - X: pd.DataFrame of features\n",
    "    - y: pd.Series of binary labels\n",
    "    - groups: pd.Series of group IDs (e.g., subject IDs)\n",
    "    - site_series: pd.Series indicating 'skin' or 'nares' for each sample\n",
    "    - n_splits: number of CV splits\n",
    "\n",
    "    Returns:\n",
    "    - cv_results: list of ROC results per fold\n",
    "    - feature_importances: pd.DataFrame of feature importances\n",
    "    - fold_aucs: list of AUCs per fold\n",
    "    \"\"\"\n",
    "    folds = group_stratified_kfold(X, y, groups, n_splits=n_splits)\n",
    "\n",
    "    cv_results = []\n",
    "    feature_importances = pd.DataFrame(index=X.columns)\n",
    "    fold_aucs = []\n",
    "\n",
    "    # Define weights: based on AUCs of site-specific classifiers\n",
    "    site_weights = {'skin': skin_weight, 'nares': nares_weight}\n",
    "    sample_weights_all = site_series.map(site_weights).values\n",
    "\n",
    "    for i, (train_idx, test_idx) in enumerate(folds):\n",
    "        X_train, X_test = X.iloc[train_idx], X.iloc[test_idx]\n",
    "        y_train, y_test = y.iloc[train_idx], y.iloc[test_idx]\n",
    "\n",
    "        if len(np.unique(y_train)) < 2 or len(np.unique(y_test)) < 2:\n",
    "            print(f\"Skipping fold {i+1} due to insufficient class representation\")\n",
    "            continue\n",
    "\n",
    "        clf = RandomForestClassifier(n_estimators=1000, random_state=42)\n",
    "        clf.fit(X_train, y_train, sample_weight=sample_weights_all[train_idx])\n",
    "        probas = clf.predict_proba(X_test)\n",
    "\n",
    "        feature_importances[f'fold_{i}'] = clf.feature_importances_\n",
    "        fpr, tpr, _ = roc_curve(y_test, probas[:, 1])\n",
    "        roc_auc = auc(fpr, tpr)\n",
    "        fold_aucs.append(roc_auc)\n",
    "\n",
    "        cv_results.append({\n",
    "            'y_true': y_test,\n",
    "            'y_proba': probas[:, 1],\n",
    "            'fpr': fpr,\n",
    "            'tpr': tpr,\n",
    "            'auc': roc_auc,\n",
    "            'fold': i\n",
    "        })\n",
    "\n",
    "    feature_importances['mean_importance'] = feature_importances.mean(axis=1)\n",
    "    feature_importances['std_importance'] = feature_importances.std(axis=1)\n",
    "    feature_importances = feature_importances.sort_values('mean_importance', ascending=False)\n",
    "\n",
    "    return cv_results, feature_importances, fold_aucs\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/sklearn/utils/validation.py:869: UserWarning: pandas.DataFrame with sparse columns found.It will be converted to a dense numpy array.\n",
      "  warnings.warn(\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/sklearn/utils/validation.py:869: UserWarning: pandas.DataFrame with sparse columns found.It will be converted to a dense numpy array.\n",
      "  warnings.warn(\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/sklearn/utils/validation.py:869: UserWarning: pandas.DataFrame with sparse columns found.It will be converted to a dense numpy array.\n",
      "  warnings.warn(\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/sklearn/utils/validation.py:869: UserWarning: pandas.DataFrame with sparse columns found.It will be converted to a dense numpy array.\n",
      "  warnings.warn(\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/sklearn/utils/validation.py:869: UserWarning: pandas.DataFrame with sparse columns found.It will be converted to a dense numpy array.\n",
      "  warnings.warn(\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/sklearn/utils/validation.py:869: UserWarning: pandas.DataFrame with sparse columns found.It will be converted to a dense numpy array.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "# site_series is the 'site' column you created earlier\n",
    "site_series = meta_combined.loc[common_samples, 'site']\n",
    "\n",
    "combined_AD_vs_H_weighted_results, combined_weighted_importance, combined_weighted_aucs = run_group_stratified_cv(\n",
    "    X_combined_with_site,  # include site one-hot encoded features if you want\n",
    "    y_combined,\n",
    "    groups_combined,\n",
    "    site_series,\n",
    "    n_splits=3\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_roc_comparison(\n",
    "    unweighted_results, weighted_results,\n",
    "    labels=('Unweighted', 'Weighted'),\n",
    "    colors=('black', 'darkred'),\n",
    "    output_path=None\n",
    "):\n",
    "    \"\"\"\n",
    "    Plot side-by-side ROC curves for unweighted vs. weighted classifiers.\n",
    "    \n",
    "    Parameters:\n",
    "    - unweighted_results: list of dicts from run_group_stratified_cv (unweighted)\n",
    "    - weighted_results: list of dicts from run_group_stratified_cv (weighted)\n",
    "    - labels: tuple of (label_unweighted, label_weighted)\n",
    "    - colors: tuple of (color_unweighted, color_weighted)\n",
    "    - output_path: path to save image (optional)\n",
    "    \"\"\"\n",
    "    def compute_mean_roc(cv_results):\n",
    "        mean_fpr = np.linspace(0, 1, 100)\n",
    "        tprs, aucs = [], []\n",
    "        for r in cv_results:\n",
    "            interp_tpr = np.interp(mean_fpr, r['fpr'], r['tpr'])\n",
    "            interp_tpr[0] = 0.0\n",
    "            tprs.append(interp_tpr)\n",
    "            aucs.append(r['auc'])\n",
    "        mean_tpr = np.mean(tprs, axis=0)\n",
    "        mean_tpr[-1] = 1.0\n",
    "        std_tpr = np.std(tprs, axis=0)\n",
    "        return mean_fpr, mean_tpr, std_tpr, np.mean(aucs), np.std(aucs)\n",
    "\n",
    "    fig, axs = plt.subplots(1, 2, figsize=(10, 5), sharey=True)\n",
    "\n",
    "    for ax, results, label, color in zip(axs, [unweighted_results, weighted_results], labels, colors):\n",
    "        mean_fpr, mean_tpr, std_tpr, auc_mean, auc_std = compute_mean_roc(results)\n",
    "        ax.plot(mean_fpr, mean_tpr, lw=3, color=color,\n",
    "                label=f'{label} (AUC = {auc_mean:.2f} ± {auc_std:.2f})')\n",
    "        ax.fill_between(mean_fpr, mean_tpr - std_tpr, mean_tpr + std_tpr,\n",
    "                        color=color, alpha=0.3)\n",
    "        ax.plot([0, 1], [0, 1], 'k--', lw=1)\n",
    "        ax.set_xlim([0.0, 1.0])\n",
    "        ax.set_ylim([0.0, 1.05])\n",
    "        ax.set_xlabel('False Positive Rate', fontsize=14)\n",
    "        ax.set_title(f'{label} Classifier', fontsize=16)\n",
    "        ax.grid(True, linestyle='--', alpha=0.7)\n",
    "        ax.tick_params(axis='both', labelsize=12)\n",
    "        ax.legend(loc='lower right', fontsize=12)\n",
    "\n",
    "    axs[0].set_ylabel('True Positive Rate', fontsize=14)\n",
    "    plt.suptitle('ROC Comparison: Skin+Nares AD vs H', fontsize=18)\n",
    "    plt.tight_layout()\n",
    "\n",
    "    if output_path:\n",
    "        plt.savefig(output_path, dpi=600, bbox_inches='tight')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_roc_comparison(\n",
    "    unweighted_results=combined_AD_vs_H_results,\n",
    "    weighted_results=combined_AD_vs_H_weighted_results,\n",
    "    labels=('Unweighted', f'Weighted (skin={skin_weight}, nares={nares_weight})'),\n",
    "    colors=('black', 'darkred'),\n",
    "    output_path='../Plots/Analysis_figures/Random_Forest/roc_comparison_unweighted_vs_weighted.png'\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Separate training for Skin and Nares, then combine"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_group_stratified_cv_unweighted(X, y, groups, n_splits=5):\n",
    "    folds = group_stratified_kfold(X, y, groups, n_splits=n_splits)\n",
    "\n",
    "    cv_results = []\n",
    "    feature_importances = pd.DataFrame(index=X.columns)\n",
    "    fold_aucs = []\n",
    "\n",
    "    for i, (train_idx, test_idx) in enumerate(folds):\n",
    "        X_train, X_test = X.iloc[train_idx], X.iloc[test_idx]\n",
    "        y_train, y_test = y.iloc[train_idx], y.iloc[test_idx]\n",
    "\n",
    "        if len(np.unique(y_train)) < 2 or len(np.unique(y_test)) < 2:\n",
    "            print(f\"Skipping fold {i+1} due to insufficient class representation\")\n",
    "            continue\n",
    "\n",
    "        clf = RandomForestClassifier(n_estimators=1000, random_state=42)\n",
    "        clf.fit(X_train, y_train)\n",
    "        probas = clf.predict_proba(X_test)\n",
    "\n",
    "        feature_importances[f'fold_{i}'] = clf.feature_importances_\n",
    "        fpr, tpr, _ = roc_curve(y_test, probas[:, 1])\n",
    "        roc_auc = auc(fpr, tpr)\n",
    "        fold_aucs.append(roc_auc)\n",
    "\n",
    "        cv_results.append({\n",
    "            'y_true': y_test,\n",
    "            'y_proba': probas[:, 1],\n",
    "            'fpr': fpr,\n",
    "            'tpr': tpr,\n",
    "            'auc': roc_auc,\n",
    "            'fold': i\n",
    "        })\n",
    "\n",
    "    feature_importances['mean_importance'] = feature_importances.mean(axis=1)\n",
    "    feature_importances['std_importance'] = feature_importances.std(axis=1)\n",
    "    feature_importances = feature_importances.sort_values('mean_importance', ascending=False)\n",
    "\n",
    "    return cv_results, feature_importances, fold_aucs\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Subset metadata and assign binary labels\n",
    "meta_skin = metadata[metadata['group'].isin(['skin-ADL', 'skin-H'])].copy()\n",
    "meta_nares = metadata[metadata['group'].isin(['nares-AD', 'nares-H'])].copy()\n",
    "\n",
    "meta_skin['label'] = meta_skin['group'].map({'skin-ADL': 1, 'skin-H': 0})\n",
    "meta_nares['label'] = meta_nares['group'].map({'nares-AD': 1, 'nares-H': 0})\n",
    "\n",
    "# Match features\n",
    "X_skin = df.loc[df.index.intersection(meta_skin.index)]\n",
    "X_nares = df.loc[df.index.intersection(meta_nares.index)]\n",
    "y_skin = meta_skin.loc[X_skin.index, 'label']\n",
    "y_nares = meta_nares.loc[X_nares.index, 'label']\n",
    "groups_skin = meta_skin.loc[X_skin.index, 'pid']\n",
    "groups_nares = meta_nares.loc[X_nares.index, 'pid']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "skin_cv_results, _, _ = run_group_stratified_cv_unweighted(X_skin, y_skin, groups_skin, n_splits=3)\n",
    "nares_cv_results, _, _ = run_group_stratified_cv_unweighted(X_nares, y_nares, groups_nares, n_splits=3)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_fold_predictions(cv_results):\n",
    "    all_y_true, all_y_proba = [], []\n",
    "    for fold in cv_results:\n",
    "        all_y_true.extend(fold['y_true'])\n",
    "        all_y_proba.extend(fold['y_proba'])\n",
    "    return np.array(all_y_true), np.array(all_y_proba)\n",
    "\n",
    "y_true_skin, proba_skin = extract_fold_predictions(skin_cv_results)\n",
    "y_true_nares, proba_nares = extract_fold_predictions(nares_cv_results)\n",
    "\n",
    "# Check that labels match — they should both be 0/1\n",
    "assert np.array_equal(y_true_skin, y_true_nares) == False  # different samples\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Assign weights based on AUCs\n",
    "auc_skin = 0.84\n",
    "auc_nares = 0.69\n",
    "total_auc = auc_skin + auc_nares\n",
    "\n",
    "weight_skin = auc_skin / total_auc\n",
    "weight_nares = auc_nares / total_auc\n",
    "\n",
    "# Combine probs (if matching samples, otherwise normalize separately)\n",
    "# Here, we'll just combine ROC curves separately\n",
    "from sklearn.metrics import roc_curve, auc\n",
    "\n",
    "# Interpolate to uniform fpr grid for ensemble\n",
    "def interpolate_roc(cv_results):\n",
    "    mean_fpr = np.linspace(0, 1, 100)\n",
    "    tprs, aucs = [], []\n",
    "    for r in cv_results:\n",
    "        interp_tpr = np.interp(mean_fpr, r['fpr'], r['tpr'])\n",
    "        interp_tpr[0] = 0.0\n",
    "        tprs.append(interp_tpr)\n",
    "        aucs.append(r['auc'])\n",
    "    mean_tpr = np.mean(tprs, axis=0)\n",
    "    std_tpr = np.std(tprs, axis=0)\n",
    "    return mean_fpr, mean_tpr, std_tpr, np.mean(aucs), np.std(aucs)\n",
    "\n",
    "fpr_skin, tpr_skin, std_skin, auc_skin, _ = interpolate_roc(skin_cv_results)\n",
    "fpr_nares, tpr_nares, std_nares, auc_nares, _ = interpolate_roc(nares_cv_results)\n",
    "\n",
    "# Combine TPRs using weighted average\n",
    "combined_tpr = weight_skin * tpr_skin + weight_nares * tpr_nares\n",
    "combined_auc = auc(fpr_skin, combined_tpr)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize=(6, 6))\n",
    "plt.plot(fpr_skin, tpr_skin, '--', label=f'Skin (AUC={auc_skin:.2f})', color='blue')\n",
    "plt.plot(fpr_nares, tpr_nares, '--', label=f'Nares (AUC={auc_nares:.2f})', color='orange')\n",
    "plt.plot(fpr_skin, combined_tpr, '-', label=f'Ensemble (AUC={combined_auc:.2f})', color='darkred', lw=3)\n",
    "\n",
    "plt.plot([0, 1], [0, 1], 'k--')\n",
    "plt.xlabel('False Positive Rate', fontsize=14)\n",
    "plt.ylabel('True Positive Rate', fontsize=14)\n",
    "plt.title('Ensemble ROC: Skin + Nares AD vs H', fontsize=16)\n",
    "plt.legend(loc='lower right', fontsize=12)\n",
    "plt.grid(True, linestyle='--', alpha=0.7)\n",
    "plt.tight_layout()\n",
    "plt.savefig('../Plots/Analysis_figures/Random_Forest/roc_ensemble_skin_nares_ADL_vs_H.png', dpi=600, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "qiime2-metagenome-2024.10",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
