{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Alpha Diversity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import Python packages\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import biom\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from itertools import cycle\n",
    "import os\n",
    "from matplotlib.colors import ListedColormap\n",
    "from matplotlib.colors import to_rgba\n",
    "from scipy.stats import mannwhitneyu\n",
    "from skbio.diversity import alpha_diversity\n",
    "from matplotlib.patches import Ellipse\n",
    "import matplotlib.transforms as transforms\n",
    "from matplotlib.patches import Circle\n",
    "from matplotlib.colors import to_hex\n",
    "import matplotlib.ticker as ticker\n",
    "import statsmodels.api as sm\n",
    "from biom import load_table\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "from skbio import TreeNode\n",
    "from skbio.diversity.alpha import shannon\n",
    "from skbio.diversity.alpha import faith_pd\n",
    "import statsmodels.formula.api as smf\n",
    "from scipy.stats import chi2_contingency\n",
    "import matplotlib.colors as mc, colorsys"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Define parameters for inputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "prevalence = '1pct'\n",
    "depth = 2000"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load inputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#sample-id</th>\n",
       "      <th>PlateNumber</th>\n",
       "      <th>PlateLocation</th>\n",
       "      <th>i5</th>\n",
       "      <th>i5Sequence</th>\n",
       "      <th>i7</th>\n",
       "      <th>i7Sequence</th>\n",
       "      <th>identifier</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Plate ID</th>\n",
       "      <th>...</th>\n",
       "      <th>age_months</th>\n",
       "      <th>sex</th>\n",
       "      <th>enrolment_date</th>\n",
       "      <th>enrolment_season</th>\n",
       "      <th>hiv_exposure</th>\n",
       "      <th>hiv_status</th>\n",
       "      <th>household_size</th>\n",
       "      <th>o_scorad</th>\n",
       "      <th>o_scorad_adj</th>\n",
       "      <th>individual_case_location</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Ca009ST_L</td>\n",
       "      <td>1</td>\n",
       "      <td>A1</td>\n",
       "      <td>SA501</td>\n",
       "      <td>ATCGTACG</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA501</td>\n",
       "      <td>CGAGAGTT-ATCGTACG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>...</td>\n",
       "      <td>24.0</td>\n",
       "      <td>male</td>\n",
       "      <td>4/16/2015</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>4.0</td>\n",
       "      <td>40</td>\n",
       "      <td>40</td>\n",
       "      <td>case-lesional_skin Cape Town</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>900221</td>\n",
       "      <td>1</td>\n",
       "      <td>B1</td>\n",
       "      <td>SA502</td>\n",
       "      <td>ACTATCTG</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA502</td>\n",
       "      <td>CGAGAGTT-ACTATCTG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>...</td>\n",
       "      <td>9.0</td>\n",
       "      <td>female</td>\n",
       "      <td>8/11/2015</td>\n",
       "      <td>Winter</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>7.0</td>\n",
       "      <td>34</td>\n",
       "      <td>34</td>\n",
       "      <td>case-lesional_skin Umtata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Ca010EB_L</td>\n",
       "      <td>1</td>\n",
       "      <td>C1</td>\n",
       "      <td>SA503</td>\n",
       "      <td>TAGCGAGT</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA503</td>\n",
       "      <td>CGAGAGTT-TAGCGAGT</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>...</td>\n",
       "      <td>24.0</td>\n",
       "      <td>female</td>\n",
       "      <td>11/20/2014</td>\n",
       "      <td>Spring</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>7.0</td>\n",
       "      <td>21</td>\n",
       "      <td>21</td>\n",
       "      <td>case-lesional_skin Cape Town</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>900460</td>\n",
       "      <td>1</td>\n",
       "      <td>D1</td>\n",
       "      <td>SA504</td>\n",
       "      <td>CTGCGTGT</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA504</td>\n",
       "      <td>CGAGAGTT-CTGCGTGT</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>...</td>\n",
       "      <td>18.0</td>\n",
       "      <td>female</td>\n",
       "      <td>9/23/2015</td>\n",
       "      <td>Spring</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>40</td>\n",
       "      <td>40</td>\n",
       "      <td>case-lesional_skin Umtata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>900051</td>\n",
       "      <td>1</td>\n",
       "      <td>E1</td>\n",
       "      <td>SA505</td>\n",
       "      <td>TCATCGAG</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA505</td>\n",
       "      <td>CGAGAGTT-TCATCGAG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>...</td>\n",
       "      <td>31.0</td>\n",
       "      <td>male</td>\n",
       "      <td>4/21/2015</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>7.0</td>\n",
       "      <td>41</td>\n",
       "      <td>41</td>\n",
       "      <td>case-lesional_skin Cape Town</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>454</th>\n",
       "      <td>900398</td>\n",
       "      <td>5</td>\n",
       "      <td>C10</td>\n",
       "      <td>SB503</td>\n",
       "      <td>AGAGTCAC</td>\n",
       "      <td>SB710</td>\n",
       "      <td>TGCTCGTA</td>\n",
       "      <td>SB710SB503</td>\n",
       "      <td>TGCTCGTA-AGAGTCAC</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>...</td>\n",
       "      <td>25.0</td>\n",
       "      <td>female</td>\n",
       "      <td>9/17/2015</td>\n",
       "      <td>Spring</td>\n",
       "      <td>Exposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>5.0</td>\n",
       "      <td>54</td>\n",
       "      <td>54</td>\n",
       "      <td>case-nonlesional_skin Umtata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>456</th>\n",
       "      <td>900400</td>\n",
       "      <td>5</td>\n",
       "      <td>B12</td>\n",
       "      <td>SB502</td>\n",
       "      <td>CGTTACTA</td>\n",
       "      <td>SB712</td>\n",
       "      <td>CGTAGCGA</td>\n",
       "      <td>SB712SB502</td>\n",
       "      <td>CGTAGCGA-CGTTACTA</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>...</td>\n",
       "      <td>21.0</td>\n",
       "      <td>female</td>\n",
       "      <td>9/17/2015</td>\n",
       "      <td>Spring</td>\n",
       "      <td>Exposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>12.0</td>\n",
       "      <td>38</td>\n",
       "      <td>38</td>\n",
       "      <td>case-lesional_skin Umtata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>457</th>\n",
       "      <td>900401</td>\n",
       "      <td>5</td>\n",
       "      <td>C12</td>\n",
       "      <td>SB503</td>\n",
       "      <td>AGAGTCAC</td>\n",
       "      <td>SB712</td>\n",
       "      <td>CGTAGCGA</td>\n",
       "      <td>SB712SB503</td>\n",
       "      <td>CGTAGCGA-AGAGTCAC</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>...</td>\n",
       "      <td>21.0</td>\n",
       "      <td>female</td>\n",
       "      <td>9/17/2015</td>\n",
       "      <td>Spring</td>\n",
       "      <td>Exposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>12.0</td>\n",
       "      <td>38</td>\n",
       "      <td>38</td>\n",
       "      <td>case-nonlesional_skin Umtata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>459</th>\n",
       "      <td>Ca006ON_L</td>\n",
       "      <td>6</td>\n",
       "      <td>F1</td>\n",
       "      <td>SA506</td>\n",
       "      <td>CGTGAGTG</td>\n",
       "      <td>SB701</td>\n",
       "      <td>CTCGACTT</td>\n",
       "      <td>SB701SA506</td>\n",
       "      <td>CTCGACTT-CGTGAGTG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>...</td>\n",
       "      <td>35.0</td>\n",
       "      <td>female</td>\n",
       "      <td>3/25/2015</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>3.0</td>\n",
       "      <td>34</td>\n",
       "      <td>34</td>\n",
       "      <td>case-lesional_skin Cape Town</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>460</th>\n",
       "      <td>Ca006ON_NL</td>\n",
       "      <td>6</td>\n",
       "      <td>F2</td>\n",
       "      <td>SA506</td>\n",
       "      <td>CGTGAGTG</td>\n",
       "      <td>SB702</td>\n",
       "      <td>CGAAGTAT</td>\n",
       "      <td>SB702SA506</td>\n",
       "      <td>CGAAGTAT-CGTGAGTG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>...</td>\n",
       "      <td>35.0</td>\n",
       "      <td>female</td>\n",
       "      <td>3/25/2015</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>3.0</td>\n",
       "      <td>34</td>\n",
       "      <td>34</td>\n",
       "      <td>case-nonlesional_skin Cape Town</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>282 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     #sample-id  PlateNumber PlateLocation     i5 i5Sequence     i7  \\\n",
       "0     Ca009ST_L            1            A1  SA501   ATCGTACG  SA701   \n",
       "1        900221            1            B1  SA502   ACTATCTG  SA701   \n",
       "2     Ca010EB_L            1            C1  SA503   TAGCGAGT  SA701   \n",
       "3        900460            1            D1  SA504   CTGCGTGT  SA701   \n",
       "4        900051            1            E1  SA505   TCATCGAG  SA701   \n",
       "..          ...          ...           ...    ...        ...    ...   \n",
       "454      900398            5           C10  SB503   AGAGTCAC  SB710   \n",
       "456      900400            5           B12  SB502   CGTTACTA  SB712   \n",
       "457      900401            5           C12  SB503   AGAGTCAC  SB712   \n",
       "459   Ca006ON_L            6            F1  SA506   CGTGAGTG  SB701   \n",
       "460  Ca006ON_NL            6            F2  SA506   CGTGAGTG  SB702   \n",
       "\n",
       "    i7Sequence  identifier           Sequence      Plate ID  ... age_months  \\\n",
       "0     CGAGAGTT  SA701SA501  CGAGAGTT-ATCGTACG  1.010000e+21  ...       24.0   \n",
       "1     CGAGAGTT  SA701SA502  CGAGAGTT-ACTATCTG  1.010000e+21  ...        9.0   \n",
       "2     CGAGAGTT  SA701SA503  CGAGAGTT-TAGCGAGT  1.010000e+21  ...       24.0   \n",
       "3     CGAGAGTT  SA701SA504  CGAGAGTT-CTGCGTGT  1.010000e+21  ...       18.0   \n",
       "4     CGAGAGTT  SA701SA505  CGAGAGTT-TCATCGAG  1.010000e+21  ...       31.0   \n",
       "..         ...         ...                ...           ...  ...        ...   \n",
       "454   TGCTCGTA  SB710SB503  TGCTCGTA-AGAGTCAC  1.010000e+21  ...       25.0   \n",
       "456   CGTAGCGA  SB712SB502  CGTAGCGA-CGTTACTA  1.010000e+21  ...       21.0   \n",
       "457   CGTAGCGA  SB712SB503  CGTAGCGA-AGAGTCAC  1.010000e+21  ...       21.0   \n",
       "459   CTCGACTT  SB701SA506  CTCGACTT-CGTGAGTG  1.010000e+21  ...       35.0   \n",
       "460   CGAAGTAT  SB702SA506  CGAAGTAT-CGTGAGTG  1.010000e+21  ...       35.0   \n",
       "\n",
       "        sex enrolment_date enrolment_season hiv_exposure hiv_status  \\\n",
       "0      male      4/16/2015          Autumn     Unexposed   negative   \n",
       "1    female      8/11/2015           Winter    Unexposed   negative   \n",
       "2    female     11/20/2014           Spring    Unexposed   negative   \n",
       "3    female      9/23/2015           Spring    Unexposed        NaN   \n",
       "4      male      4/21/2015          Autumn     Unexposed   negative   \n",
       "..      ...            ...              ...          ...        ...   \n",
       "454  female      9/17/2015           Spring      Exposed   negative   \n",
       "456  female      9/17/2015           Spring      Exposed   negative   \n",
       "457  female      9/17/2015           Spring      Exposed   negative   \n",
       "459  female      3/25/2015           Autumn    Unexposed   negative   \n",
       "460  female      3/25/2015           Autumn    Unexposed   negative   \n",
       "\n",
       "    household_size o_scorad o_scorad_adj         individual_case_location  \n",
       "0              4.0       40           40     case-lesional_skin Cape Town  \n",
       "1              7.0       34           34        case-lesional_skin Umtata  \n",
       "2              7.0       21           21     case-lesional_skin Cape Town  \n",
       "3              4.0       40           40        case-lesional_skin Umtata  \n",
       "4              7.0       41           41     case-lesional_skin Cape Town  \n",
       "..             ...      ...          ...                              ...  \n",
       "454            5.0       54           54     case-nonlesional_skin Umtata  \n",
       "456           12.0       38           38        case-lesional_skin Umtata  \n",
       "457           12.0       38           38     case-nonlesional_skin Umtata  \n",
       "459            3.0       34           34     case-lesional_skin Cape Town  \n",
       "460            3.0       34           34  case-nonlesional_skin Cape Town  \n",
       "\n",
       "[282 rows x 32 columns]"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load the metadata\n",
    "metadata_path = '../Metadata/16S_AD_South-Africa_metadata_subset.tsv'\n",
    "metadata = pd.read_csv(metadata_path, sep='\\t')\n",
    "metadata['o_scorad_adj'] = metadata['o_scorad'].fillna(0)\n",
    "if \"Unnamed: 0\" in metadata.columns:\n",
    "    metadata = metadata.drop(columns=[\"Unnamed: 0\"])\n",
    "\n",
    "metadata['individual_case_location'] = metadata['case_type'] + ' ' + metadata['area']\n",
    "metadata['individual_case_location'].value_counts()     \n",
    "\n",
    "# Filter to only the skin samples\n",
    "metadata = metadata[metadata['specimen'] == 'skin']\n",
    "metadata\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Streptococcus ASV-1</th>\n",
       "      <th>Streptococcus ASV-2</th>\n",
       "      <th>Corynebacterium ASV-1</th>\n",
       "      <th>Corynebacterium ASV-3</th>\n",
       "      <th>ASV-2</th>\n",
       "      <th>Cutibacterium ASV-1</th>\n",
       "      <th>ASV-56</th>\n",
       "      <th>ASV-27</th>\n",
       "      <th>ASV-28</th>\n",
       "      <th>Cutibacterium ASV-2</th>\n",
       "      <th>...</th>\n",
       "      <th>ASV-109</th>\n",
       "      <th>Filifactor ASV-1</th>\n",
       "      <th>Leptotrichia A 993758 ASV-11</th>\n",
       "      <th>Capnocytophaga 820688 ASV-8</th>\n",
       "      <th>Dermacoccus ASV-2</th>\n",
       "      <th>Comamonas F 589250 ASV-4</th>\n",
       "      <th>Bosea ASV-2</th>\n",
       "      <th>Capnocytophaga 820690 ASV-3</th>\n",
       "      <th>Blautia A 141780 ASV-1</th>\n",
       "      <th>ASV-153</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>900221</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900570</th>\n",
       "      <td>126.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900091</th>\n",
       "      <td>14.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900245</th>\n",
       "      <td>9.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900581</th>\n",
       "      <td>95.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900263</th>\n",
       "      <td>165.0</td>\n",
       "      <td>108.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900081</th>\n",
       "      <td>453.0</td>\n",
       "      <td>0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>0</td>\n",
       "      <td>77.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900501</th>\n",
       "      <td>948.0</td>\n",
       "      <td>127.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900279</th>\n",
       "      <td>39.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900097</th>\n",
       "      <td>20.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>151 rows × 847 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Streptococcus ASV-1  Streptococcus ASV-2  Corynebacterium ASV-1  \\\n",
       "900221                    0                    0                      0   \n",
       "900570                126.0                    0                      0   \n",
       "900091                 14.0                  7.0                      0   \n",
       "900245                  9.0                    0                      0   \n",
       "900581                 95.0                    0                      0   \n",
       "...                     ...                  ...                    ...   \n",
       "900263                165.0                108.0                    5.0   \n",
       "900081                453.0                    0                   21.0   \n",
       "900501                948.0                127.0                      0   \n",
       "900279                 39.0                    0                      0   \n",
       "900097                 20.0                    0                      0   \n",
       "\n",
       "        Corynebacterium ASV-3   ASV-2  Cutibacterium ASV-1   ASV-56   ASV-27  \\\n",
       "900221                      0       0                    0        0        0   \n",
       "900570                      0       0                  5.0        0        0   \n",
       "900091                      0       0                    0        0        0   \n",
       "900245                      0       0                  9.0        0        0   \n",
       "900581                      0    13.0                 23.0        0        0   \n",
       "...                       ...     ...                  ...      ...      ...   \n",
       "900263                      0     7.0                 17.0        0        0   \n",
       "900081                      0    77.0                 19.0        0        0   \n",
       "900501                      0     3.0                  8.0        0        0   \n",
       "900279                      0     3.0                 33.0        0      8.0   \n",
       "900097                      0       0                 28.0        0        0   \n",
       "\n",
       "         ASV-28  Cutibacterium ASV-2  ...   ASV-109  Filifactor ASV-1  \\\n",
       "900221        0                    0  ...         0                 0   \n",
       "900570        0                  4.0  ...         0                 0   \n",
       "900091      8.0                    0  ...         0                 0   \n",
       "900245        0                    0  ...         0                 0   \n",
       "900581        0                  9.0  ...         0                 0   \n",
       "...         ...                  ...  ...       ...               ...   \n",
       "900263      4.0                  9.0  ...         0                 0   \n",
       "900081        0                    0  ...         0                 0   \n",
       "900501        0                  1.0  ...         0                 0   \n",
       "900279        0                 35.0  ...         0                 0   \n",
       "900097        0                 16.0  ...         0                 0   \n",
       "\n",
       "        Leptotrichia A 993758 ASV-11  Capnocytophaga 820688 ASV-8  \\\n",
       "900221                             0                            0   \n",
       "900570                             0                            0   \n",
       "900091                             0                            0   \n",
       "900245                             0                            0   \n",
       "900581                             0                            0   \n",
       "...                              ...                          ...   \n",
       "900263                             0                            0   \n",
       "900081                             0                            0   \n",
       "900501                             0                            0   \n",
       "900279                             0                            0   \n",
       "900097                             0                            0   \n",
       "\n",
       "        Dermacoccus ASV-2  Comamonas F 589250 ASV-4  Bosea ASV-2  \\\n",
       "900221                  0                         0            0   \n",
       "900570                  0                         0            0   \n",
       "900091                  0                         0            0   \n",
       "900245                  0                         0            0   \n",
       "900581                  0                         0            0   \n",
       "...                   ...                       ...          ...   \n",
       "900263                  0                         0            0   \n",
       "900081                  0                         0            0   \n",
       "900501                  0                         0            0   \n",
       "900279                  0                         0            0   \n",
       "900097                  0                         0            0   \n",
       "\n",
       "        Capnocytophaga 820690 ASV-3  Blautia A 141780 ASV-1   ASV-153  \n",
       "900221                            0                       0         0  \n",
       "900570                            0                       0         0  \n",
       "900091                            0                       0         0  \n",
       "900245                            0                       0         0  \n",
       "900581                            0                       0         0  \n",
       "...                             ...                     ...       ...  \n",
       "900263                            0                       0         0  \n",
       "900081                            0                       0         0  \n",
       "900501                            0                       0         0  \n",
       "900279                            0                       0         0  \n",
       "900097                            0                       0         0  \n",
       "\n",
       "[151 rows x 847 columns]"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read in table at Genus-ASV level\n",
    "biom_path = f'../Data/Tables/Count_Tables/6_209766_feature_table_dedup_prev-filt-{prevalence}_rare-{depth}_Genus-ASV_skin.biom'\n",
    "biom_tbl = load_table(biom_path)\n",
    "df = pd.DataFrame(biom_tbl.to_dataframe().T)\n",
    "\n",
    "# Clean up column names so they match tree tips\n",
    "df.columns = (\n",
    "    df.columns\n",
    "      .str.replace(r\"^g__\", \"\", regex=True)\n",
    "      .str.replace(\"_\", \" \", regex=False)\n",
    ")\n",
    "\n",
    "# Keep only samples that exist in metadata\n",
    "df = df[df.index.isin(metadata['#sample-id'])]\n",
    "\n",
    "# Subset to skin samples only\n",
    "skin_samples = metadata.loc[metadata['specimen'] == 'skin', '#sample-id']\n",
    "df = df.loc[df.index.intersection(skin_samples)]\n",
    "\n",
    "df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "individual_case_location\n",
       "case-lesional_skin Umtata             41\n",
       "control-nonlesional_skin Umtata       30\n",
       "case-lesional_skin Cape Town          26\n",
       "case-nonlesional_skin Umtata          21\n",
       "case-nonlesional_skin Cape Town       20\n",
       "control-nonlesional_skin Cape Town    13\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Subset metadata to match df (check numbers for groups below)\n",
    "metadata = metadata[metadata['#sample-id'].isin(df.index)]\n",
    "metadata['individual_case_location'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<TreeNode, name: root, internal node count: 845, tips count: 847>"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read in newick phylogenetic tree\n",
    "tree_path = f'../Data/Trees/209766_feature_table_dedup_prev-filt-{prevalence}_rare-{depth}_Genus-ASV_skin_aln.nwk'\n",
    "\n",
    "tree = TreeNode.read(tree_path)\n",
    "\n",
    "# Remove leading 'g  ' from each tip name (if present)\n",
    "for tip in tree.tips():\n",
    "    if tip.name.startswith(\"g  \"):\n",
    "        tip.name = tip.name.replace(\"g  \", \"\", 1)\n",
    "\n",
    "# Root the tree (required for Faith PD) at its midpoint\n",
    "tree = tree.root_at_midpoint()\n",
    "\n",
    "tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "847\n"
     ]
    }
   ],
   "source": [
    "# Check that table and tree features match\n",
    "features_in_tree = {tip.name for tip in tree.tips()}\n",
    "shared_features = df.columns.intersection(features_in_tree)\n",
    "\n",
    "print(len(shared_features))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run Faith Phylogenetic Diversity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Faith PD successfully computed for all skin samples!\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "900221    3.134932\n",
       "900570    8.753575\n",
       "900091    3.333254\n",
       "900245    7.948009\n",
       "900581    8.964010\n",
       "Name: Faith_PD, dtype: float64"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Run Faith PD\n",
    "faith_pd_results = {}\n",
    "\n",
    "for sample in df.index:\n",
    "    counts = df.loc[sample].values\n",
    "    faith_pd_results[sample] = faith_pd(counts, otu_ids=df.columns, tree=tree)\n",
    "\n",
    "faith_pd_series = pd.Series(faith_pd_results, name=\"Faith_PD\")\n",
    "faith_pd_series.to_csv(\"../Data/Alpha_Diversity/faith_pd_results.tsv\", sep=\"\\t\", header=True)\n",
    "\n",
    "print(\"Faith PD successfully computed for all skin samples!\")\n",
    "faith_pd_series.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Counts of samples with Faith PD = 0.0 by individual_case_location:\n",
      "Empty DataFrame\n",
      "Columns: [individual_case_location, num_zero_Faith_PD]\n",
      "Index: []\n",
      "\n",
      "Total samples with Faith PD = 0.0: 0 / 151\n"
     ]
    }
   ],
   "source": [
    "# Merge Faith PD results with metadata\n",
    "metadata = metadata.set_index(\"#sample-id\")  # ensure sample IDs align\n",
    "merged = metadata.join(faith_pd_series, how=\"left\")\n",
    "\n",
    "# Filter only samples with Faith PD = 0.0\n",
    "zero_pd_samples = merged[merged[\"Faith_PD\"] == 0.0]\n",
    "\n",
    "# Count how many in each individual_case_location\n",
    "counts = (\n",
    "    zero_pd_samples\n",
    "    .groupby(\"individual_case_location\")\n",
    "    .size()\n",
    "    .reset_index(name=\"num_zero_Faith_PD\")\n",
    "    .sort_values(\"num_zero_Faith_PD\", ascending=False)\n",
    ")\n",
    "\n",
    "print(\"Counts of samples with Faith PD = 0.0 by individual_case_location:\")\n",
    "print(counts)\n",
    "\n",
    "print(f\"\\nTotal samples with Faith PD = 0.0: {len(zero_pd_samples)} / {len(faith_pd_series)}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plot Faith PD from Umtata and Cape Town"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_faith_region(metadata, group_col, downsample=True):\n",
    "    \"\"\"\n",
    "    Generate Faith PD box + strip plots for Umtata and Cape Town,\n",
    "    including pairwise Mann–Whitney U tests with p-value correction.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    metadata : pd.DataFrame\n",
    "        Metadata table with sample information.\n",
    "    group_col : str\n",
    "        Column name containing group labels (e.g., \"case_type_region\").\n",
    "    downsample : bool, default=True\n",
    "        Whether to perform equal downsampling of groups.\n",
    "    \"\"\"\n",
    "\n",
    "    # ------------------------------------------------------------------\n",
    "    # Helper functions\n",
    "    # ------------------------------------------------------------------\n",
    "    def brighten_color(color, amount=0.3):\n",
    "        \"\"\"Brighten a color by a given amount.\"\"\"\n",
    "        try:\n",
    "            c = mc.cnames[color]\n",
    "        except KeyError:\n",
    "            c = color\n",
    "        h, l, s = colorsys.rgb_to_hls(*mc.to_rgb(c))\n",
    "        return colorsys.hls_to_rgb(h, 1 - amount * (1 - l), s)\n",
    "\n",
    "    def add_pairwise_annotations(ax, groups, df, group_col, y_col):\n",
    "        \"\"\"Perform pairwise Mann–Whitney tests and annotate figure.\"\"\"\n",
    "        comparisons, pvals, stats = [], [], []\n",
    "\n",
    "        for i in range(len(groups)):\n",
    "            for j in range(i + 1, len(groups)):\n",
    "                g1, g2 = groups[i], groups[j]\n",
    "                vals1 = df[df[group_col] == g1][y_col]\n",
    "                vals2 = df[df[group_col] == g2][y_col]\n",
    "                stat, p = mannwhitneyu(vals1, vals2, alternative='two-sided')\n",
    "                comparisons.append((g1, g2))\n",
    "                pvals.append(p)\n",
    "                stats.append(stat)\n",
    "\n",
    "        # Correction (for reporting only)\n",
    "        reject, pvals_corr, _, _ = multipletests(pvals, method='fdr_bh')\n",
    "\n",
    "        # ------------------------------------------------------------------\n",
    "        # Print results\n",
    "        # ------------------------------------------------------------------\n",
    "        print(f\"\\nPairwise Mann–Whitney tests for {y_col}:\")\n",
    "        print(\"------------------------------------------------------------\")\n",
    "        for (g1, g2), stat, p, p_corr, rej in zip(comparisons, stats, pvals, pvals_corr, reject):\n",
    "            sig_marker = \"*\" if rej else \"ns\"\n",
    "            print(f\"{g1} vs {g2} | U={stat:.0f}, p={p:.3g}, p_adj={p_corr:.3g}, {sig_marker}\")\n",
    "        print(\"------------------------------------------------------------\")\n",
    "\n",
    "        # ------------------------------------------------------------------\n",
    "        # Annotate on figure (using raw p-values)\n",
    "        # ------------------------------------------------------------------\n",
    "        y_max = df[y_col].max()\n",
    "        spacing = 0.5\n",
    "        for ((g1, g2), p, stat) in zip(comparisons, pvals, stats):\n",
    "            if p < 0.05:\n",
    "                i, j = groups.index(g1), groups.index(g2)\n",
    "                y = y_max + spacing\n",
    "                ax.plot([i, i, j, j],\n",
    "                        [y, y + 0.015, y + 0.015, y],\n",
    "                        lw=1, color='black')\n",
    "                label = f\"p={p:.2g}, U={stat:.0f}\"   # <-- raw p-value\n",
    "                ax.text((i + j) / 2, y, label,\n",
    "                        ha='center', va='bottom', fontsize=9)\n",
    "                y_max += spacing\n",
    "\n",
    "\n",
    "    # ------------------------------------------------------------------\n",
    "    # Load Faith PD results\n",
    "    # ------------------------------------------------------------------\n",
    "    feature_table = pd.read_csv(\"../Data/Alpha_Diversity/faith_pd_results.tsv\", sep=\"\\t\", index_col=0)\n",
    "    feature_table = feature_table[feature_table['Faith_PD'] != 0]\n",
    "\n",
    "    # Match metadata and Faith PD samples\n",
    "    common_samples = metadata.index.intersection(feature_table.index)\n",
    "    metadata = metadata.loc[common_samples].copy()\n",
    "    metadata['Faith_PD'] = feature_table.loc[common_samples, 'Faith_PD']\n",
    "    metadata['o_scorad'] = pd.to_numeric(metadata['o_scorad'], errors='coerce')\n",
    "\n",
    "    # ------------------------------------------------------------------\n",
    "    # Region-specific plotting function\n",
    "    # ------------------------------------------------------------------\n",
    "    def plot_region(region, groups, palette, output_path, ylim, ytick, downsample_sizes=None):\n",
    "        \"\"\"Generic plotting routine for each region.\"\"\"\n",
    "        region_data = metadata[metadata[group_col].isin(groups)].copy()\n",
    "\n",
    "        if downsample and downsample_sizes:\n",
    "            # Downsample each group to defined size\n",
    "            sampled = []\n",
    "            for g in groups:\n",
    "                n = downsample_sizes.get(g, None)\n",
    "                subset = region_data[region_data[group_col] == g]\n",
    "                if n and len(subset) > n:\n",
    "                    subset = subset.sample(n=n, random_state=42, replace=False)\n",
    "                sampled.append(subset)\n",
    "            region_data = pd.concat(sampled)\n",
    "        else:\n",
    "            # Use all samples (no downsampling)\n",
    "            print(f\"No downsampling applied for {region}. Using all {len(region_data)} samples.\")\n",
    "\n",
    "        # Count for labeling\n",
    "        group_counts = region_data[group_col].value_counts().to_dict()\n",
    "        short_labels = ['H', 'ADNL', 'ADL']\n",
    "        new_labels = [\n",
    "            f\"{lab}\\n(n={group_counts.get(g, 0)})\"\n",
    "            for lab, g in zip(short_labels, groups)\n",
    "        ]\n",
    "\n",
    "        # Plot\n",
    "        fig, ax = plt.subplots(figsize=(2.5, 4.25))\n",
    "        sns.boxplot(\n",
    "            x=group_col, y='Faith_PD', data=region_data,\n",
    "            palette=palette, order=groups, ax=ax\n",
    "        )\n",
    "        sns.stripplot(\n",
    "            x=group_col, y='Faith_PD', data=region_data,\n",
    "            palette={k: brighten_color(v) for k, v in palette.items()},\n",
    "            jitter=True, dodge=False, linewidth=0.6, order=groups, ax=ax\n",
    "        )\n",
    "\n",
    "        ax.set_title(region, fontsize=14, y=1)\n",
    "        ax.set_xlabel('')\n",
    "        ax.set_ylabel('Faith PD', fontsize=13)\n",
    "        ax.set_xticks(range(len(new_labels)))\n",
    "        ax.set_xticklabels(new_labels, ha='center', fontsize=13)\n",
    "        ax.set_ylim(*ylim)\n",
    "        ax.yaxis.set_major_locator(ticker.MultipleLocator(ytick))\n",
    "        ax.yaxis.set_major_formatter(ticker.FuncFormatter(lambda x, _: f'{int(x)}'))\n",
    "\n",
    "        add_pairwise_annotations(ax, groups, region_data, group_col, 'Faith_PD')\n",
    "\n",
    "        fig.savefig(output_path, dpi=600, bbox_inches='tight', pad_inches=0.1)\n",
    "        plt.close(fig)\n",
    "\n",
    "    # ------------------------------------------------------------------\n",
    "    # Umtata\n",
    "    # ------------------------------------------------------------------\n",
    "    umtata_groups = [\n",
    "        'control-nonlesional_skin Umtata',\n",
    "        'case-nonlesional_skin Umtata',\n",
    "        'case-lesional_skin Umtata'\n",
    "    ]\n",
    "    umtata_palette = {\n",
    "        'control-nonlesional_skin Umtata': '#66C2EE',\n",
    "        'case-nonlesional_skin Umtata': '#FAD5A5',\n",
    "        'case-lesional_skin Umtata': '#F0806B',\n",
    "    }\n",
    "\n",
    "    # Downsample only ADL to match ADNL\n",
    "    umtata_downsample = {\n",
    "        'control-nonlesional_skin Umtata': 21,\n",
    "        'case-nonlesional_skin Umtata': 21,\n",
    "        'case-lesional_skin Umtata': 21\n",
    "    } if downsample else None\n",
    "\n",
    "    plot_region(\n",
    "        'Umtata', umtata_groups, umtata_palette,\n",
    "        '../Figures/Supplementary/Suppl_Fig_5A.jpg',\n",
    "        ylim=(0, 12), ytick=2, downsample_sizes=umtata_downsample\n",
    "    )\n",
    "\n",
    "    # ------------------------------------------------------------------\n",
    "    # Cape Town\n",
    "    # ------------------------------------------------------------------\n",
    "    cape_groups = [\n",
    "        'control-nonlesional_skin Cape Town',\n",
    "        'case-nonlesional_skin Cape Town',\n",
    "        'case-lesional_skin Cape Town'\n",
    "    ]\n",
    "    cape_palette = {\n",
    "        'control-nonlesional_skin Cape Town': '#7FBCEB',\n",
    "        'case-nonlesional_skin Cape Town': '#FAD5A5',\n",
    "        'case-lesional_skin Cape Town': '#C9A34F',\n",
    "    }\n",
    "    # Downsample only ADL to match ADNL\n",
    "    cape_downsample = {\n",
    "        'control-nonlesional_skin Cape Town': 23,\n",
    "        'case-nonlesional_skin Cape Town': 20,\n",
    "        'case-lesional_skin Cape Town': 20\n",
    "    } if downsample else None\n",
    "\n",
    "    plot_region(\n",
    "        'Cape Town', cape_groups, cape_palette,\n",
    "        '../Figures/Supplementary/Suppl_Fig_5D.jpg',\n",
    "        ylim=(0, 12), ytick=2, downsample_sizes=cape_downsample\n",
    "    )\n",
    "\n",
    "\n",
    "    print(f\"Done! Plots saved. Downsampling applied: {downsample}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Pairwise Mann–Whitney tests for Faith_PD:\n",
      "------------------------------------------------------------\n",
      "control-nonlesional_skin Umtata vs case-nonlesional_skin Umtata | U=293, p=0.0701, p_adj=0.105, ns\n",
      "control-nonlesional_skin Umtata vs case-lesional_skin Umtata | U=300, p=0.0469, p_adj=0.105, ns\n",
      "case-nonlesional_skin Umtata vs case-lesional_skin Umtata | U=224, p=0.94, p_adj=0.94, ns\n",
      "------------------------------------------------------------\n",
      "\n",
      "Pairwise Mann–Whitney tests for Faith_PD:\n",
      "------------------------------------------------------------\n",
      "control-nonlesional_skin Cape Town vs case-nonlesional_skin Cape Town | U=109, p=0.45, p_adj=0.655, ns\n",
      "control-nonlesional_skin Cape Town vs case-lesional_skin Cape Town | U=99, p=0.261, p_adj=0.655, ns\n",
      "case-nonlesional_skin Cape Town vs case-lesional_skin Cape Town | U=183, p=0.655, p_adj=0.655, ns\n",
      "------------------------------------------------------------\n",
      "Done! Plots saved. Downsampling applied: True\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/22/yck9vwx53w1c38tvj_c0_tz00000gn/T/ipykernel_98321/1344081598.py:120: FutureWarning: Passing `palette` without assigning `hue` is deprecated.\n",
      "  sns.stripplot(\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.\n",
      "  with pd.option_context('mode.use_inf_as_na', True):\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.\n",
      "  with pd.option_context('mode.use_inf_as_na', True):\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1075: FutureWarning: When grouping with a length-1 list-like, you will need to pass a length-1 tuple to get_group in a future version of pandas. Pass `(name,)` instead of `name` to silence this warning.\n",
      "  data_subset = grouped_data.get_group(pd_key)\n",
      "/var/folders/22/yck9vwx53w1c38tvj_c0_tz00000gn/T/ipykernel_98321/1344081598.py:120: FutureWarning: Passing `palette` without assigning `hue` is deprecated.\n",
      "  sns.stripplot(\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.\n",
      "  with pd.option_context('mode.use_inf_as_na', True):\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.\n",
      "  with pd.option_context('mode.use_inf_as_na', True):\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1075: FutureWarning: When grouping with a length-1 list-like, you will need to pass a length-1 tuple to get_group in a future version of pandas. Pass `(name,)` instead of `name` to silence this warning.\n",
      "  data_subset = grouped_data.get_group(pd_key)\n"
     ]
    }
   ],
   "source": [
    "plot_faith_region(\n",
    "    metadata=metadata,\n",
    "    group_col='individual_case_location'\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Faith PD between Umtata healthy and Cape Town healthy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "def prepare_faith_pd_data(metadata, group_col, faith_pd_path=\"../Data/Alpha_Diversity/faith_pd_results.tsv\"):\n",
    "    import pandas as pd\n",
    "\n",
    "    # Load Faith PD results\n",
    "    feature_table = pd.read_csv(faith_pd_path, sep=\"\\t\")\n",
    "    feature_table = feature_table.set_index(feature_table.columns[0])\n",
    "\n",
    "    # Align with metadata\n",
    "    for possible_id in ['#sample-id', '#SampleID', 'sample_id']:\n",
    "        if possible_id in metadata.columns:\n",
    "            metadata = metadata.set_index(possible_id)\n",
    "            break\n",
    "\n",
    "    # Subset to common samples\n",
    "    feature_table = feature_table[feature_table.index.isin(metadata.index)]\n",
    "    common_samples = metadata.index.intersection(feature_table.index)\n",
    "\n",
    "    metadata = metadata.loc[common_samples].copy()\n",
    "    metadata['Faith_PD'] = feature_table.loc[common_samples]\n",
    "\n",
    "    # Define region_group labels\n",
    "    metadata['region_group'] = None\n",
    "    metadata.loc[metadata[group_col] == 'control-nonlesional_skin Cape Town', 'region_group'] = 'H_CapeTown'\n",
    "    metadata.loc[metadata[group_col] == 'control-nonlesional_skin Umtata', 'region_group'] = 'H_Umtata'\n",
    "    \n",
    "    metadata.loc[metadata[group_col].isin([\n",
    "        'case-nonlesional_skin Cape Town',\n",
    "        'case-lesional_skin Cape Town']), 'region_group'] = 'AD_CapeTown'\n",
    "\n",
    "    metadata.loc[metadata[group_col].isin([\n",
    "        'case-nonlesional_skin Umtata',\n",
    "        'case-lesional_skin Umtata']), 'region_group'] = 'AD_Umtata'\n",
    "\n",
    "    return metadata\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_faith_pd_groups(metadata):\n",
    "\n",
    "    # Helper function for plotting one comparison\n",
    "    def plot_box(data, order, palette, title, outfile):\n",
    "        fig, ax = plt.subplots(figsize=(2.5, 4))\n",
    "        sns.boxplot(\n",
    "            x='region_group', y='Faith_PD',\n",
    "            data=data, palette=palette, order=order, ax=ax\n",
    "        )\n",
    "        sns.stripplot(\n",
    "            x='region_group', y='Faith_PD',\n",
    "            data=data, palette=palette,\n",
    "            jitter=True, dodge=False, linewidth=0.6, order=order, ax=ax\n",
    "        )\n",
    "\n",
    "        ax.set_title(title, fontsize=14, y=1.02)\n",
    "        ax.set_xlabel('')\n",
    "        ax.set_ylabel('Faith PD', fontsize=13)\n",
    "\n",
    "        # Label x-axis with group names and n\n",
    "        ax.set_xticklabels([\n",
    "            f\"{order[0].split('_')[1]}\\n(n={len(data[data['region_group']==order[0]])})\",\n",
    "            f\"{order[1].split('_')[1]}\\n(n={len(data[data['region_group']==order[1]])})\"\n",
    "        ], fontsize=12)\n",
    "        ax.yaxis.set_major_locator(ticker.MultipleLocator(2))\n",
    "\n",
    "        # Mann–Whitney U Test\n",
    "        group1 = data[data['region_group'] == order[0]]['Faith_PD'].dropna()\n",
    "        group2 = data[data['region_group'] == order[1]]['Faith_PD'].dropna()\n",
    "        stat, p = mannwhitneyu(group1, group2, alternative='two-sided')\n",
    "\n",
    "        # Annotate with p and U\n",
    "        ax.text(\n",
    "            0.5, data['Faith_PD'].max() * 0.95,\n",
    "            f\"p={p:.3f}\\nU={stat:.0f}\",\n",
    "            ha='center', va='bottom', fontsize=9\n",
    "        )\n",
    "\n",
    "        fig.savefig(outfile, dpi=600, bbox_inches='tight', pad_inches=0.1)\n",
    "        plt.close(fig)\n",
    "\n",
    "    # --- Healthy Cape Town vs Healthy Umtata ---\n",
    "    h_data = metadata[metadata['region_group'].isin(['H_CapeTown', 'H_Umtata'])].copy()\n",
    "\n",
    "    h_palette = {\n",
    "        'H_Umtata': '#66C2EE',   # light blue\n",
    "        'H_CapeTown': '#7FBCEB'  # slightly deeper blue\n",
    "    }\n",
    "\n",
    "    plot_box(\n",
    "        h_data,\n",
    "        ['H_Umtata', 'H_CapeTown'],\n",
    "        h_palette,\n",
    "        'Healthy Skin Diversity',\n",
    "        '../Figures/Main/Fig_4A.jpg'\n",
    "    )\n",
    "\n",
    "    print(\"Done\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/22/yck9vwx53w1c38tvj_c0_tz00000gn/T/ipykernel_98321/4250840665.py:10: FutureWarning: Passing `palette` without assigning `hue` is deprecated.\n",
      "  sns.stripplot(\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.\n",
      "  with pd.option_context('mode.use_inf_as_na', True):\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.\n",
      "  with pd.option_context('mode.use_inf_as_na', True):\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1075: FutureWarning: When grouping with a length-1 list-like, you will need to pass a length-1 tuple to get_group in a future version of pandas. Pass `(name,)` instead of `name` to silence this warning.\n",
      "  data_subset = grouped_data.get_group(pd_key)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done\n"
     ]
    }
   ],
   "source": [
    "# Step 1: Prepare Faith PD data with region_group labels\n",
    "md_faith = prepare_faith_pd_data(\n",
    "    metadata=metadata,\n",
    "    group_col='individual_case_location'\n",
    ")\n",
    "\n",
    "# Step 2: Plot the Faith PD comparisons\n",
    "plot_faith_pd_groups(\n",
    "    metadata=md_faith\n",
    ")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run Shannon Diversity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shannon diversity successfully computed for all skin samples!\n",
      "900221    1.327170\n",
      "900570    4.581213\n",
      "900091    3.732909\n",
      "900245    5.102464\n",
      "900581    4.391689\n",
      "Name: Shannon, dtype: float64\n"
     ]
    }
   ],
   "source": [
    "shannon_results = {}\n",
    "\n",
    "for sample in df.index:\n",
    "    counts = df.loc[sample].values\n",
    "    shannon_results[sample] = shannon(counts)\n",
    "\n",
    "# Convert to pandas Series\n",
    "shannon_series = pd.Series(shannon_results, name=\"Shannon\")\n",
    "\n",
    "# Save results\n",
    "shannon_series.to_csv(\"../Data/Alpha_Diversity/shannon_results.tsv\", sep=\"\\t\", header=True)\n",
    "\n",
    "print(\"Shannon diversity successfully computed for all skin samples!\")\n",
    "print(shannon_series.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plot Shannon from Umtata healthy and Cape Town healthy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "28fdef62",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#sample-id</th>\n",
       "      <th>PlateNumber</th>\n",
       "      <th>PlateLocation</th>\n",
       "      <th>i5</th>\n",
       "      <th>i5Sequence</th>\n",
       "      <th>i7</th>\n",
       "      <th>i7Sequence</th>\n",
       "      <th>identifier</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Plate ID</th>\n",
       "      <th>...</th>\n",
       "      <th>age_months</th>\n",
       "      <th>sex</th>\n",
       "      <th>enrolment_date</th>\n",
       "      <th>enrolment_season</th>\n",
       "      <th>hiv_exposure</th>\n",
       "      <th>hiv_status</th>\n",
       "      <th>household_size</th>\n",
       "      <th>o_scorad</th>\n",
       "      <th>o_scorad_adj</th>\n",
       "      <th>individual_case_location</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Ca009ST_L</td>\n",
       "      <td>1</td>\n",
       "      <td>A1</td>\n",
       "      <td>SA501</td>\n",
       "      <td>ATCGTACG</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA501</td>\n",
       "      <td>CGAGAGTT-ATCGTACG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>...</td>\n",
       "      <td>24.0</td>\n",
       "      <td>male</td>\n",
       "      <td>4/16/2015</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>4.0</td>\n",
       "      <td>40</td>\n",
       "      <td>40</td>\n",
       "      <td>case-lesional_skin Cape Town</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>900221</td>\n",
       "      <td>1</td>\n",
       "      <td>B1</td>\n",
       "      <td>SA502</td>\n",
       "      <td>ACTATCTG</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA502</td>\n",
       "      <td>CGAGAGTT-ACTATCTG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>...</td>\n",
       "      <td>9.0</td>\n",
       "      <td>female</td>\n",
       "      <td>8/11/2015</td>\n",
       "      <td>Winter</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>7.0</td>\n",
       "      <td>34</td>\n",
       "      <td>34</td>\n",
       "      <td>case-lesional_skin Umtata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Ca010EB_L</td>\n",
       "      <td>1</td>\n",
       "      <td>C1</td>\n",
       "      <td>SA503</td>\n",
       "      <td>TAGCGAGT</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA503</td>\n",
       "      <td>CGAGAGTT-TAGCGAGT</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>...</td>\n",
       "      <td>24.0</td>\n",
       "      <td>female</td>\n",
       "      <td>11/20/2014</td>\n",
       "      <td>Spring</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>7.0</td>\n",
       "      <td>21</td>\n",
       "      <td>21</td>\n",
       "      <td>case-lesional_skin Cape Town</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>900460</td>\n",
       "      <td>1</td>\n",
       "      <td>D1</td>\n",
       "      <td>SA504</td>\n",
       "      <td>CTGCGTGT</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA504</td>\n",
       "      <td>CGAGAGTT-CTGCGTGT</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>...</td>\n",
       "      <td>18.0</td>\n",
       "      <td>female</td>\n",
       "      <td>9/23/2015</td>\n",
       "      <td>Spring</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>40</td>\n",
       "      <td>40</td>\n",
       "      <td>case-lesional_skin Umtata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>900051</td>\n",
       "      <td>1</td>\n",
       "      <td>E1</td>\n",
       "      <td>SA505</td>\n",
       "      <td>TCATCGAG</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA505</td>\n",
       "      <td>CGAGAGTT-TCATCGAG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>...</td>\n",
       "      <td>31.0</td>\n",
       "      <td>male</td>\n",
       "      <td>4/21/2015</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>7.0</td>\n",
       "      <td>41</td>\n",
       "      <td>41</td>\n",
       "      <td>case-lesional_skin Cape Town</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>454</th>\n",
       "      <td>900398</td>\n",
       "      <td>5</td>\n",
       "      <td>C10</td>\n",
       "      <td>SB503</td>\n",
       "      <td>AGAGTCAC</td>\n",
       "      <td>SB710</td>\n",
       "      <td>TGCTCGTA</td>\n",
       "      <td>SB710SB503</td>\n",
       "      <td>TGCTCGTA-AGAGTCAC</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>...</td>\n",
       "      <td>25.0</td>\n",
       "      <td>female</td>\n",
       "      <td>9/17/2015</td>\n",
       "      <td>Spring</td>\n",
       "      <td>Exposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>5.0</td>\n",
       "      <td>54</td>\n",
       "      <td>54</td>\n",
       "      <td>case-nonlesional_skin Umtata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>456</th>\n",
       "      <td>900400</td>\n",
       "      <td>5</td>\n",
       "      <td>B12</td>\n",
       "      <td>SB502</td>\n",
       "      <td>CGTTACTA</td>\n",
       "      <td>SB712</td>\n",
       "      <td>CGTAGCGA</td>\n",
       "      <td>SB712SB502</td>\n",
       "      <td>CGTAGCGA-CGTTACTA</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>...</td>\n",
       "      <td>21.0</td>\n",
       "      <td>female</td>\n",
       "      <td>9/17/2015</td>\n",
       "      <td>Spring</td>\n",
       "      <td>Exposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>12.0</td>\n",
       "      <td>38</td>\n",
       "      <td>38</td>\n",
       "      <td>case-lesional_skin Umtata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>457</th>\n",
       "      <td>900401</td>\n",
       "      <td>5</td>\n",
       "      <td>C12</td>\n",
       "      <td>SB503</td>\n",
       "      <td>AGAGTCAC</td>\n",
       "      <td>SB712</td>\n",
       "      <td>CGTAGCGA</td>\n",
       "      <td>SB712SB503</td>\n",
       "      <td>CGTAGCGA-AGAGTCAC</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>...</td>\n",
       "      <td>21.0</td>\n",
       "      <td>female</td>\n",
       "      <td>9/17/2015</td>\n",
       "      <td>Spring</td>\n",
       "      <td>Exposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>12.0</td>\n",
       "      <td>38</td>\n",
       "      <td>38</td>\n",
       "      <td>case-nonlesional_skin Umtata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>459</th>\n",
       "      <td>Ca006ON_L</td>\n",
       "      <td>6</td>\n",
       "      <td>F1</td>\n",
       "      <td>SA506</td>\n",
       "      <td>CGTGAGTG</td>\n",
       "      <td>SB701</td>\n",
       "      <td>CTCGACTT</td>\n",
       "      <td>SB701SA506</td>\n",
       "      <td>CTCGACTT-CGTGAGTG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>...</td>\n",
       "      <td>35.0</td>\n",
       "      <td>female</td>\n",
       "      <td>3/25/2015</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>3.0</td>\n",
       "      <td>34</td>\n",
       "      <td>34</td>\n",
       "      <td>case-lesional_skin Cape Town</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>460</th>\n",
       "      <td>Ca006ON_NL</td>\n",
       "      <td>6</td>\n",
       "      <td>F2</td>\n",
       "      <td>SA506</td>\n",
       "      <td>CGTGAGTG</td>\n",
       "      <td>SB702</td>\n",
       "      <td>CGAAGTAT</td>\n",
       "      <td>SB702SA506</td>\n",
       "      <td>CGAAGTAT-CGTGAGTG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>...</td>\n",
       "      <td>35.0</td>\n",
       "      <td>female</td>\n",
       "      <td>3/25/2015</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>3.0</td>\n",
       "      <td>34</td>\n",
       "      <td>34</td>\n",
       "      <td>case-nonlesional_skin Cape Town</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>282 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     #sample-id  PlateNumber PlateLocation     i5 i5Sequence     i7  \\\n",
       "0     Ca009ST_L            1            A1  SA501   ATCGTACG  SA701   \n",
       "1        900221            1            B1  SA502   ACTATCTG  SA701   \n",
       "2     Ca010EB_L            1            C1  SA503   TAGCGAGT  SA701   \n",
       "3        900460            1            D1  SA504   CTGCGTGT  SA701   \n",
       "4        900051            1            E1  SA505   TCATCGAG  SA701   \n",
       "..          ...          ...           ...    ...        ...    ...   \n",
       "454      900398            5           C10  SB503   AGAGTCAC  SB710   \n",
       "456      900400            5           B12  SB502   CGTTACTA  SB712   \n",
       "457      900401            5           C12  SB503   AGAGTCAC  SB712   \n",
       "459   Ca006ON_L            6            F1  SA506   CGTGAGTG  SB701   \n",
       "460  Ca006ON_NL            6            F2  SA506   CGTGAGTG  SB702   \n",
       "\n",
       "    i7Sequence  identifier           Sequence      Plate ID  ... age_months  \\\n",
       "0     CGAGAGTT  SA701SA501  CGAGAGTT-ATCGTACG  1.010000e+21  ...       24.0   \n",
       "1     CGAGAGTT  SA701SA502  CGAGAGTT-ACTATCTG  1.010000e+21  ...        9.0   \n",
       "2     CGAGAGTT  SA701SA503  CGAGAGTT-TAGCGAGT  1.010000e+21  ...       24.0   \n",
       "3     CGAGAGTT  SA701SA504  CGAGAGTT-CTGCGTGT  1.010000e+21  ...       18.0   \n",
       "4     CGAGAGTT  SA701SA505  CGAGAGTT-TCATCGAG  1.010000e+21  ...       31.0   \n",
       "..         ...         ...                ...           ...  ...        ...   \n",
       "454   TGCTCGTA  SB710SB503  TGCTCGTA-AGAGTCAC  1.010000e+21  ...       25.0   \n",
       "456   CGTAGCGA  SB712SB502  CGTAGCGA-CGTTACTA  1.010000e+21  ...       21.0   \n",
       "457   CGTAGCGA  SB712SB503  CGTAGCGA-AGAGTCAC  1.010000e+21  ...       21.0   \n",
       "459   CTCGACTT  SB701SA506  CTCGACTT-CGTGAGTG  1.010000e+21  ...       35.0   \n",
       "460   CGAAGTAT  SB702SA506  CGAAGTAT-CGTGAGTG  1.010000e+21  ...       35.0   \n",
       "\n",
       "        sex enrolment_date enrolment_season hiv_exposure hiv_status  \\\n",
       "0      male      4/16/2015          Autumn     Unexposed   negative   \n",
       "1    female      8/11/2015           Winter    Unexposed   negative   \n",
       "2    female     11/20/2014           Spring    Unexposed   negative   \n",
       "3    female      9/23/2015           Spring    Unexposed        NaN   \n",
       "4      male      4/21/2015          Autumn     Unexposed   negative   \n",
       "..      ...            ...              ...          ...        ...   \n",
       "454  female      9/17/2015           Spring      Exposed   negative   \n",
       "456  female      9/17/2015           Spring      Exposed   negative   \n",
       "457  female      9/17/2015           Spring      Exposed   negative   \n",
       "459  female      3/25/2015           Autumn    Unexposed   negative   \n",
       "460  female      3/25/2015           Autumn    Unexposed   negative   \n",
       "\n",
       "    household_size o_scorad o_scorad_adj         individual_case_location  \n",
       "0              4.0       40           40     case-lesional_skin Cape Town  \n",
       "1              7.0       34           34        case-lesional_skin Umtata  \n",
       "2              7.0       21           21     case-lesional_skin Cape Town  \n",
       "3              4.0       40           40        case-lesional_skin Umtata  \n",
       "4              7.0       41           41     case-lesional_skin Cape Town  \n",
       "..             ...      ...          ...                              ...  \n",
       "454            5.0       54           54     case-nonlesional_skin Umtata  \n",
       "456           12.0       38           38        case-lesional_skin Umtata  \n",
       "457           12.0       38           38     case-nonlesional_skin Umtata  \n",
       "459            3.0       34           34     case-lesional_skin Cape Town  \n",
       "460            3.0       34           34  case-nonlesional_skin Cape Town  \n",
       "\n",
       "[282 rows x 32 columns]"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load the metadata\n",
    "metadata_path = '../Metadata/16S_AD_South-Africa_metadata_subset.tsv'\n",
    "metadata = pd.read_csv(metadata_path, sep='\\t')\n",
    "metadata['o_scorad_adj'] = metadata['o_scorad'].fillna(0)\n",
    "if \"Unnamed: 0\" in metadata.columns:\n",
    "    metadata = metadata.drop(columns=[\"Unnamed: 0\"])\n",
    "\n",
    "metadata['individual_case_location'] = metadata['case_type'] + ' ' + metadata['area']\n",
    "metadata['individual_case_location'].value_counts()     \n",
    "\n",
    "# Filter to only the skin samples\n",
    "metadata = metadata[metadata['specimen'] == 'skin']\n",
    "metadata\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "def prepare_shannon_data(metadata, group_col, shannon_path=\"../Data/Alpha_Diversity/shannon_results.tsv\"):\n",
    "    import pandas as pd\n",
    "\n",
    "    # Load Shannon results\n",
    "    feature_table = pd.read_csv(shannon_path, sep=\"\\t\")\n",
    "    feature_table = feature_table.set_index(feature_table.columns[0])\n",
    "\n",
    "    # Align with metadata\n",
    "    for possible_id in ['#sample-id']:\n",
    "        if possible_id in metadata.columns:\n",
    "            metadata = metadata.set_index(possible_id)\n",
    "            break\n",
    "\n",
    "    # Subset to common samples\n",
    "    feature_table = feature_table[feature_table.index.isin(metadata.index)]\n",
    "    common_samples = metadata.index.intersection(feature_table.index)\n",
    "\n",
    "    metadata = metadata.loc[common_samples].copy()\n",
    "    metadata['Shannon'] = feature_table.loc[common_samples]\n",
    "\n",
    "    # Define region_group labels\n",
    "    metadata['region_group'] = None\n",
    "    metadata.loc[metadata[group_col] == 'control-nonlesional_skin Cape Town', 'region_group'] = 'H_CapeTown'\n",
    "    metadata.loc[metadata[group_col] == 'control-nonlesional_skin Umtata', 'region_group'] = 'H_Umtata'\n",
    "\n",
    "    metadata.loc[metadata[group_col].isin([\n",
    "        'case-nonlesional_skin Cape Town',\n",
    "        'case-lesional_skin Cape Town']), 'region_group'] = 'AD_CapeTown'\n",
    "\n",
    "    metadata.loc[metadata[group_col].isin([\n",
    "        'case-nonlesional_skin Umtata',\n",
    "        'case-lesional_skin Umtata']), 'region_group'] = 'AD_Umtata'\n",
    "\n",
    "    return metadata\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_shannon_groups(metadata):\n",
    "    import seaborn as sns\n",
    "    import matplotlib.pyplot as plt\n",
    "    import matplotlib.ticker as ticker\n",
    "    from scipy.stats import mannwhitneyu\n",
    "    import pandas as pd\n",
    "\n",
    "    # Helper function for plotting one comparison\n",
    "    def plot_box(data, order, palette, title, outfile):\n",
    "        fig, ax = plt.subplots(figsize=(2.5, 4))\n",
    "        sns.boxplot(x='region_group', y='Shannon', data=data, palette=palette,\n",
    "                    order=order, ax=ax)\n",
    "        sns.stripplot(x='region_group', y='Shannon', data=data, palette=palette,\n",
    "                      jitter=True, dodge=False, linewidth=0.6, order=order, ax=ax)\n",
    "        ax.set_title(title, fontsize=14, y=1.02)\n",
    "        ax.set_xlabel('')\n",
    "        ax.set_ylabel('Shannon Diversity', fontsize=13)\n",
    "        ax.set_xticklabels([\n",
    "            f\"{order[0].split('_')[1]}\\n(n={len(data[data['region_group']==order[0]])})\",\n",
    "            f\"{order[1].split('_')[1]}\\n(n={len(data[data['region_group']==order[1]])})\"\n",
    "        ], fontsize=12)\n",
    "        ax.yaxis.set_major_locator(ticker.MultipleLocator(2))\n",
    "\n",
    "        # Mann–Whitney U Test\n",
    "        stat, p = mannwhitneyu(\n",
    "            data[data['region_group'] == order[0]]['Shannon'].dropna(),\n",
    "            data[data['region_group'] == order[1]]['Shannon'].dropna(),\n",
    "            alternative='two-sided'\n",
    "        )\n",
    "        ax.text(0.55, data['Shannon'].max() - 0.75, f\"p={p:.2}\\nU={stat:.0f}\",\n",
    "                ha='center', va='bottom', fontsize=9)\n",
    "        fig.savefig(outfile, dpi=600, bbox_inches='tight', pad_inches=0.1)\n",
    "        plt.close(fig)\n",
    "\n",
    "    # --- Healthy vs Healthy ---\n",
    "    h_data = metadata[metadata['region_group'].isin(['H_CapeTown', 'H_Umtata'])].copy()\n",
    "\n",
    "    h_umtata = h_data[h_data['region_group'] == 'H_Umtata']\n",
    "    h_cape = h_data[h_data['region_group'] == 'H_CapeTown']\n",
    "\n",
    "    h_data = pd.concat([h_umtata, h_cape])\n",
    "    h_palette = {'H_Umtata': '#66C2EE', 'H_CapeTown': '#7FBCEB'}\n",
    "    plot_box(h_data, ['H_Umtata', 'H_CapeTown'], h_palette,\n",
    "             'Healthy Skin Diversity', '../Figures/Supplementary/Suppl_Fig_4.jpg')\n",
    "\n",
    "    print(\"Done: all group counts printed above and Shannon plots saved.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done: all group counts printed above and Shannon plots saved.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/22/yck9vwx53w1c38tvj_c0_tz00000gn/T/ipykernel_98321/3655841751.py:13: FutureWarning: Passing `palette` without assigning `hue` is deprecated.\n",
      "  sns.stripplot(x='region_group', y='Shannon', data=data, palette=palette,\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.\n",
      "  with pd.option_context('mode.use_inf_as_na', True):\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.\n",
      "  with pd.option_context('mode.use_inf_as_na', True):\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1075: FutureWarning: When grouping with a length-1 list-like, you will need to pass a length-1 tuple to get_group in a future version of pandas. Pass `(name,)` instead of `name` to silence this warning.\n",
      "  data_subset = grouped_data.get_group(pd_key)\n"
     ]
    }
   ],
   "source": [
    "# Step 1: Prepare Shannon data with region_group labels\n",
    "md_shannon = prepare_shannon_data(\n",
    "    metadata=metadata,\n",
    "    group_col='individual_case_location'\n",
    ")\n",
    "\n",
    "# Step 2: Plot the Shannon comparisons\n",
    "plot_shannon_groups(\n",
    "    metadata=md_shannon\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "b6636809",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "individual_case_location\n",
       "case-lesional_skin Umtata             41\n",
       "control-nonlesional_skin Umtata       30\n",
       "case-lesional_skin Cape Town          26\n",
       "case-nonlesional_skin Umtata          21\n",
       "case-nonlesional_skin Cape Town       20\n",
       "control-nonlesional_skin Cape Town    13\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Subset metadata to match df (check numbers for groups below)\n",
    "metadata = metadata[metadata['#sample-id'].isin(df.index)]\n",
    "metadata['individual_case_location'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b206dace",
   "metadata": {},
   "source": [
    "### Plot Shannon Diversity from Umtata and Cape Town"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "f38213b6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Counts of samples with Shannon = 0.0 by individual_case_location:\n",
      "Empty DataFrame\n",
      "Columns: [individual_case_location, num_zero_Shannon]\n",
      "Index: []\n",
      "\n",
      "Total samples with Shannon = 0.0: 0 / 151 total samples\n"
     ]
    }
   ],
   "source": [
    "# Load metadata and Shannon diversity results\n",
    "shannon = pd.read_csv(\"../Data/Alpha_Diversity/shannon_results.tsv\", sep=\"\\t\", index_col=0)\n",
    "\n",
    "# Ensure column name consistency\n",
    "if \"shannon\" not in shannon.columns:\n",
    "    shannon.columns = [\"shannon\"]\n",
    "\n",
    "# Merge Shannon values into metadata\n",
    "merged = metadata.join(shannon[\"shannon\"], how=\"left\")\n",
    "\n",
    "# Identify samples with Shannon = 0.0\n",
    "zero_shannon_samples = merged[merged[\"shannon\"] == 0.0]\n",
    "\n",
    "# Summarize by location\n",
    "counts = (\n",
    "    zero_shannon_samples\n",
    "    .groupby(\"individual_case_location\")\n",
    "    .size()\n",
    "    .reset_index(name=\"num_zero_Shannon\")\n",
    "    .sort_values(\"num_zero_Shannon\", ascending=False)\n",
    ")\n",
    "\n",
    "# Print results\n",
    "print(\"Counts of samples with Shannon = 0.0 by individual_case_location:\")\n",
    "print(counts)\n",
    "\n",
    "print(f\"\\nTotal samples with Shannon = 0.0: {len(zero_shannon_samples)} / {len(shannon)} total samples\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "36835afd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Counts of samples with Shannon = 0.0 by individual_case_location:\n",
      "Empty DataFrame\n",
      "Columns: [individual_case_location, num_zero_Shannon]\n",
      "Index: []\n",
      "\n",
      "Total samples with Shannon = 0.0: 0 / 151 total samples\n"
     ]
    }
   ],
   "source": [
    "# Load in Shannon results\n",
    "shannon = pd.read_csv(\"../Data/Alpha_Diversity/shannon_results.tsv\", sep=\"\\t\", index_col=0)\n",
    "\n",
    "# Ensure column name consistency\n",
    "if \"shannon\" not in shannon.columns:\n",
    "    shannon.columns = [\"shannon\"]\n",
    "\n",
    "# Merge Shannon values into metadata\n",
    "metadata = metadata.set_index(\"#sample-id\")  # ensure alignment\n",
    "merged = metadata.join(shannon[\"shannon\"], how=\"left\")\n",
    "\n",
    "# Identify samples with Shannon = 0.0\n",
    "zero_shannon_samples = merged[merged[\"shannon\"] == 0.0]\n",
    "\n",
    "# Summarize by location\n",
    "counts = (\n",
    "    zero_shannon_samples\n",
    "    .groupby(\"individual_case_location\")\n",
    "    .size()\n",
    "    .reset_index(name=\"num_zero_Shannon\")\n",
    "    .sort_values(\"num_zero_Shannon\", ascending=False)\n",
    ")\n",
    "\n",
    "# Print results\n",
    "print(\"Counts of samples with Shannon = 0.0 by individual_case_location:\")\n",
    "print(counts)\n",
    "\n",
    "print(f\"\\nTotal samples with Shannon = 0.0: {len(zero_shannon_samples)} / {len(shannon)} total samples\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "7176381f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_shannon_region(metadata, group_col, downsample=True):\n",
    "    \"\"\"\n",
    "    Generate Shannon diversity box + strip plots for Umtata and Cape Town,\n",
    "    including pairwise Mann–Whitney U tests with p-value correction.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    metadata : pd.DataFrame\n",
    "        Metadata table with sample information.\n",
    "    group_col : str\n",
    "        Column name containing group labels (e.g., \"case_type_region\").\n",
    "    downsample : bool, default=True\n",
    "        Whether to perform equal downsampling of groups.\n",
    "    \"\"\"\n",
    "\n",
    "    # ------------------------------------------------------------------\n",
    "    # Helper functions\n",
    "    # ------------------------------------------------------------------\n",
    "    def brighten_color(color, amount=0.3):\n",
    "        \"\"\"Brighten a color by a given amount.\"\"\"\n",
    "        try:\n",
    "            c = mc.cnames[color]\n",
    "        except KeyError:\n",
    "            c = color\n",
    "        h, l, s = colorsys.rgb_to_hls(*mc.to_rgb(c))\n",
    "        return colorsys.hls_to_rgb(h, 1 - amount * (1 - l), s)\n",
    "\n",
    "    def add_pairwise_annotations(ax, groups, df, group_col, y_col):\n",
    "        \"\"\"Perform pairwise Mann–Whitney tests and annotate figure.\"\"\"\n",
    "        comparisons, pvals, stats = [], [], []\n",
    "\n",
    "        for i in range(len(groups)):\n",
    "            for j in range(i + 1, len(groups)):\n",
    "                g1, g2 = groups[i], groups[j]\n",
    "                vals1 = df[df[group_col] == g1][y_col]\n",
    "                vals2 = df[df[group_col] == g2][y_col]\n",
    "                stat, p = mannwhitneyu(vals1, vals2, alternative='two-sided')\n",
    "                comparisons.append((g1, g2))\n",
    "                pvals.append(p)\n",
    "                stats.append(stat)\n",
    "\n",
    "        # Benjamini-Hochberg FDR correction\n",
    "        reject, pvals_corr, _, _ = multipletests(pvals, method='fdr_bh')\n",
    "\n",
    "        # Print results to console\n",
    "        print(f\"\\nPairwise Mann–Whitney tests for {y_col}:\")\n",
    "        print(\"------------------------------------------------------------\")\n",
    "        for (g1, g2), stat, p, p_corr, rej in zip(comparisons, stats, pvals, pvals_corr, reject):\n",
    "            sig_marker = \"*\" if rej else \"ns\"\n",
    "            print(f\"{g1} vs {g2} | U={stat:.0f}, p={p:.3g}, p_adj={p_corr:.3g}, {sig_marker}\")\n",
    "        print(\"------------------------------------------------------------\")\n",
    "\n",
    "        # Annotate significant comparisons on plot (raw p-values)\n",
    "        y_max = df[y_col].max()\n",
    "        spacing = 0.2\n",
    "        for ((g1, g2), p, stat) in zip(comparisons, pvals, stats):\n",
    "            if p < 0.05:\n",
    "                i, j = groups.index(g1), groups.index(g2)\n",
    "                y = y_max + spacing\n",
    "                ax.plot([i, i, j, j], [y, y + 0.1, y + 0.1, y], lw=1, color='black')\n",
    "                label = f\"p={p:.2g}, U={stat:.0f}\"\n",
    "                ax.text((i + j) / 2, y + 0.1, label, ha='center', va='bottom', fontsize=9)\n",
    "                y_max += spacing + 0.1\n",
    "\n",
    "    # ------------------------------------------------------------------\n",
    "    # Load Shannon diversity results\n",
    "    # ------------------------------------------------------------------\n",
    "    feature_table = pd.read_csv(\"../Data/Alpha_Diversity/shannon_results.tsv\", sep=\"\\t\", index_col=0)\n",
    "    feature_table = feature_table[feature_table['Shannon'] != 0]\n",
    "\n",
    "    # Match metadata and Shannon samples\n",
    "    common_samples = metadata.index.intersection(feature_table.index)\n",
    "    metadata = metadata.loc[common_samples].copy()\n",
    "    metadata['Shannon'] = feature_table.loc[common_samples, 'Shannon']\n",
    "    metadata['o_scorad'] = pd.to_numeric(metadata['o_scorad'], errors='coerce')\n",
    "\n",
    "    # ------------------------------------------------------------------\n",
    "    # Region-specific plotting function\n",
    "    # ------------------------------------------------------------------\n",
    "    def plot_region(region, groups, palette, output_path, ylim, ytick, downsample_sizes=None):\n",
    "        \"\"\"Generic plotting routine for each region.\"\"\"\n",
    "        region_data = metadata[metadata[group_col].isin(groups)].copy()\n",
    "\n",
    "        if downsample and downsample_sizes:\n",
    "            # Downsample each group to defined size\n",
    "            sampled = []\n",
    "            for g in groups:\n",
    "                n = downsample_sizes.get(g, None)\n",
    "                subset = region_data[region_data[group_col] == g]\n",
    "                if n and len(subset) > n:\n",
    "                    subset = subset.sample(n=n, random_state=42, replace=False)\n",
    "                sampled.append(subset)\n",
    "            region_data = pd.concat(sampled)\n",
    "        else:\n",
    "            print(f\"No downsampling applied for {region}. Using all {len(region_data)} samples.\")\n",
    "\n",
    "        # Count for labeling\n",
    "        group_counts = region_data[group_col].value_counts().to_dict()\n",
    "        short_labels = ['H', 'ADNL', 'ADL']\n",
    "        new_labels = [\n",
    "            f\"{lab}\\n(n={group_counts.get(g, 0)})\"\n",
    "            for lab, g in zip(short_labels, groups)\n",
    "        ]\n",
    "\n",
    "        # Plot\n",
    "        fig, ax = plt.subplots(figsize=(2.5, 4.25))\n",
    "        sns.boxplot(\n",
    "            x=group_col, y='Shannon', data=region_data,\n",
    "            palette=palette, order=groups, ax=ax\n",
    "        )\n",
    "        sns.stripplot(\n",
    "            x=group_col, y='Shannon', data=region_data,\n",
    "            palette={k: brighten_color(v) for k, v in palette.items()},\n",
    "            jitter=True, dodge=False, linewidth=0.6, order=groups, ax=ax\n",
    "        )\n",
    "\n",
    "        ax.set_title(region, fontsize=14, y=1)\n",
    "        ax.set_xlabel('')\n",
    "        ax.set_ylabel('Shannon Diversity', fontsize=13)\n",
    "        ax.set_xticks(range(len(new_labels)))\n",
    "        ax.set_xticklabels(new_labels, ha='center', fontsize=13)\n",
    "        ax.set_ylim(*ylim)\n",
    "        ax.yaxis.set_major_locator(ticker.MultipleLocator(ytick))\n",
    "        ax.yaxis.set_major_formatter(ticker.FuncFormatter(lambda x, _: f'{x:.1f}'))\n",
    "\n",
    "        add_pairwise_annotations(ax, groups, region_data, group_col, 'Shannon')\n",
    "\n",
    "        fig.savefig(output_path, dpi=600, bbox_inches='tight', pad_inches=0.1)\n",
    "        plt.close(fig)\n",
    "\n",
    "    # ------------------------------------------------------------------\n",
    "    # Umtata\n",
    "    # ------------------------------------------------------------------\n",
    "    umtata_groups = [\n",
    "        'control-nonlesional_skin Umtata',\n",
    "        'case-nonlesional_skin Umtata',\n",
    "        'case-lesional_skin Umtata'\n",
    "    ]\n",
    "    umtata_palette = {\n",
    "        'control-nonlesional_skin Umtata': '#66C2EE',\n",
    "        'case-nonlesional_skin Umtata': '#FAD5A5',\n",
    "        'case-lesional_skin Umtata': '#F0806B',\n",
    "    }\n",
    "\n",
    "    umtata_downsample = {\n",
    "        'control-nonlesional_skin Umtata': 21,\n",
    "        'case-nonlesional_skin Umtata': 21,\n",
    "        'case-lesional_skin Umtata': 21\n",
    "    } if downsample else None\n",
    "\n",
    "    plot_region(\n",
    "        'Umtata', umtata_groups, umtata_palette,\n",
    "        '../Figures/Supplementary/Suppl_Fig_5B.jpg',\n",
    "        ylim=(0, 6), ytick=1, downsample_sizes=umtata_downsample\n",
    "    )\n",
    "\n",
    "    # ------------------------------------------------------------------\n",
    "    # Cape Town\n",
    "    # ------------------------------------------------------------------\n",
    "    cape_groups = [\n",
    "        'control-nonlesional_skin Cape Town',\n",
    "        'case-nonlesional_skin Cape Town',\n",
    "        'case-lesional_skin Cape Town'\n",
    "    ]\n",
    "    cape_palette = {\n",
    "        'control-nonlesional_skin Cape Town': '#7FBCEB',\n",
    "        'case-nonlesional_skin Cape Town': '#FAD5A5',\n",
    "        'case-lesional_skin Cape Town': '#C9A34F',\n",
    "    }\n",
    "\n",
    "    cape_downsample = {\n",
    "        'control-nonlesional_skin Cape Town': 23,\n",
    "        'case-nonlesional_skin Cape Town': 20,\n",
    "        'case-lesional_skin Cape Town': 20\n",
    "    } if downsample else None\n",
    "\n",
    "    plot_region(\n",
    "        'Cape Town', cape_groups, cape_palette,\n",
    "        '../Figures/Supplementary/Suppl_Fig_5E.jpg',\n",
    "        ylim=(0, 6), ytick=1, downsample_sizes=cape_downsample\n",
    "    )\n",
    "\n",
    "    print(f\"Done! Shannon plots saved. Downsampling applied: {downsample}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "e8f29607",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Pairwise Mann–Whitney tests for Shannon:\n",
      "------------------------------------------------------------\n",
      "control-nonlesional_skin Umtata vs case-nonlesional_skin Umtata | U=292, p=0.0741, p_adj=0.111, ns\n",
      "control-nonlesional_skin Umtata vs case-lesional_skin Umtata | U=298, p=0.0527, p_adj=0.111, ns\n",
      "case-nonlesional_skin Umtata vs case-lesional_skin Umtata | U=230, p=0.821, p_adj=0.821, ns\n",
      "------------------------------------------------------------\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/22/yck9vwx53w1c38tvj_c0_tz00000gn/T/ipykernel_98321/1348119658.py:111: FutureWarning: Passing `palette` without assigning `hue` is deprecated.\n",
      "  sns.stripplot(\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.\n",
      "  with pd.option_context('mode.use_inf_as_na', True):\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.\n",
      "  with pd.option_context('mode.use_inf_as_na', True):\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1075: FutureWarning: When grouping with a length-1 list-like, you will need to pass a length-1 tuple to get_group in a future version of pandas. Pass `(name,)` instead of `name` to silence this warning.\n",
      "  data_subset = grouped_data.get_group(pd_key)\n",
      "/var/folders/22/yck9vwx53w1c38tvj_c0_tz00000gn/T/ipykernel_98321/1348119658.py:111: FutureWarning: Passing `palette` without assigning `hue` is deprecated.\n",
      "  sns.stripplot(\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.\n",
      "  with pd.option_context('mode.use_inf_as_na', True):\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.\n",
      "  with pd.option_context('mode.use_inf_as_na', True):\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1075: FutureWarning: When grouping with a length-1 list-like, you will need to pass a length-1 tuple to get_group in a future version of pandas. Pass `(name,)` instead of `name` to silence this warning.\n",
      "  data_subset = grouped_data.get_group(pd_key)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Pairwise Mann–Whitney tests for Shannon:\n",
      "------------------------------------------------------------\n",
      "control-nonlesional_skin Cape Town vs case-nonlesional_skin Cape Town | U=82, p=0.0801, p_adj=0.152, ns\n",
      "control-nonlesional_skin Cape Town vs case-lesional_skin Cape Town | U=85, p=0.101, p_adj=0.152, ns\n",
      "case-nonlesional_skin Cape Town vs case-lesional_skin Cape Town | U=195, p=0.903, p_adj=0.903, ns\n",
      "------------------------------------------------------------\n",
      "Done! Shannon plots saved. Downsampling applied: True\n"
     ]
    }
   ],
   "source": [
    "plot_shannon_region(\n",
    "    metadata=metadata,\n",
    "    group_col='individual_case_location'\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "qiime2-metagenome-2024.10",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
