{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# ASV Overlap Venn Diagram"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import Python packages\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import biom\n",
    "from biom import load_table\n",
    "import numpy as np\n",
    "from scipy.stats import pearsonr\n",
    "from matplotlib_venn import venn2\n",
    "import matplotlib.patches as mpatches\n",
    "from scipy.stats import mannwhitneyu\n",
    "from biom import Table\n",
    "from gemelli.rpca import rpca\n",
    "from scipy.spatial.distance import euclidean\n",
    "import re\n",
    "import matplotlib.colors as mcolors\n",
    "from scipy.stats import fisher_exact\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PlateNumber</th>\n",
       "      <th>PlateLocation</th>\n",
       "      <th>i5</th>\n",
       "      <th>i5Sequence</th>\n",
       "      <th>i7</th>\n",
       "      <th>i7Sequence</th>\n",
       "      <th>identifier</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Plate ID</th>\n",
       "      <th>Well location</th>\n",
       "      <th>...</th>\n",
       "      <th>age_months</th>\n",
       "      <th>sex</th>\n",
       "      <th>enrolment_date</th>\n",
       "      <th>enrolment_season</th>\n",
       "      <th>hiv_exposure</th>\n",
       "      <th>hiv_status</th>\n",
       "      <th>household_size</th>\n",
       "      <th>o_scorad</th>\n",
       "      <th>group</th>\n",
       "      <th>ad_status</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#sample-id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Ca009STL</th>\n",
       "      <td>1</td>\n",
       "      <td>A1</td>\n",
       "      <td>SA501</td>\n",
       "      <td>ATCGTACG</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA501</td>\n",
       "      <td>CGAGAGTT-ATCGTACG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>A1</td>\n",
       "      <td>...</td>\n",
       "      <td>24.0</td>\n",
       "      <td>male</td>\n",
       "      <td>4/16/2015</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>4.0</td>\n",
       "      <td>40</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900221</th>\n",
       "      <td>1</td>\n",
       "      <td>B1</td>\n",
       "      <td>SA502</td>\n",
       "      <td>ACTATCTG</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA502</td>\n",
       "      <td>CGAGAGTT-ACTATCTG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>B1</td>\n",
       "      <td>...</td>\n",
       "      <td>9.0</td>\n",
       "      <td>female</td>\n",
       "      <td>8/11/2015</td>\n",
       "      <td>Winter</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>7.0</td>\n",
       "      <td>34</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca010EBL</th>\n",
       "      <td>1</td>\n",
       "      <td>C1</td>\n",
       "      <td>SA503</td>\n",
       "      <td>TAGCGAGT</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA503</td>\n",
       "      <td>CGAGAGTT-TAGCGAGT</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>C1</td>\n",
       "      <td>...</td>\n",
       "      <td>24.0</td>\n",
       "      <td>female</td>\n",
       "      <td>11/20/2014</td>\n",
       "      <td>Spring</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>7.0</td>\n",
       "      <td>21</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900460</th>\n",
       "      <td>1</td>\n",
       "      <td>D1</td>\n",
       "      <td>SA504</td>\n",
       "      <td>CTGCGTGT</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA504</td>\n",
       "      <td>CGAGAGTT-CTGCGTGT</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>D1</td>\n",
       "      <td>...</td>\n",
       "      <td>18.0</td>\n",
       "      <td>female</td>\n",
       "      <td>9/23/2015</td>\n",
       "      <td>Spring</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>40</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900051</th>\n",
       "      <td>1</td>\n",
       "      <td>E1</td>\n",
       "      <td>SA505</td>\n",
       "      <td>TCATCGAG</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA505</td>\n",
       "      <td>CGAGAGTT-TCATCGAG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>E1</td>\n",
       "      <td>...</td>\n",
       "      <td>31.0</td>\n",
       "      <td>male</td>\n",
       "      <td>4/21/2015</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>7.0</td>\n",
       "      <td>41</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900401</th>\n",
       "      <td>5</td>\n",
       "      <td>C12</td>\n",
       "      <td>SB503</td>\n",
       "      <td>AGAGTCAC</td>\n",
       "      <td>SB712</td>\n",
       "      <td>CGTAGCGA</td>\n",
       "      <td>SB712SB503</td>\n",
       "      <td>CGTAGCGA-AGAGTCAC</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>C12</td>\n",
       "      <td>...</td>\n",
       "      <td>21.0</td>\n",
       "      <td>female</td>\n",
       "      <td>9/17/2015</td>\n",
       "      <td>Spring</td>\n",
       "      <td>Exposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>12.0</td>\n",
       "      <td>38</td>\n",
       "      <td>skin-ADNL</td>\n",
       "      <td>AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900402</th>\n",
       "      <td>6</td>\n",
       "      <td>B4</td>\n",
       "      <td>SA502</td>\n",
       "      <td>ACTATCTG</td>\n",
       "      <td>SB704</td>\n",
       "      <td>TCTCTATG</td>\n",
       "      <td>SB704SA502</td>\n",
       "      <td>TCTCTATG-ACTATCTG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>B4</td>\n",
       "      <td>...</td>\n",
       "      <td>21.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nares-AD</td>\n",
       "      <td>AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca006ONL</th>\n",
       "      <td>6</td>\n",
       "      <td>F1</td>\n",
       "      <td>SA506</td>\n",
       "      <td>CGTGAGTG</td>\n",
       "      <td>SB701</td>\n",
       "      <td>CTCGACTT</td>\n",
       "      <td>SB701SA506</td>\n",
       "      <td>CTCGACTT-CGTGAGTG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>F1</td>\n",
       "      <td>...</td>\n",
       "      <td>35.0</td>\n",
       "      <td>female</td>\n",
       "      <td>3/25/2015</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>3.0</td>\n",
       "      <td>34</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca006ONNL</th>\n",
       "      <td>6</td>\n",
       "      <td>F2</td>\n",
       "      <td>SA506</td>\n",
       "      <td>CGTGAGTG</td>\n",
       "      <td>SB702</td>\n",
       "      <td>CGAAGTAT</td>\n",
       "      <td>SB702SA506</td>\n",
       "      <td>CGAAGTAT-CGTGAGTG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>F2</td>\n",
       "      <td>...</td>\n",
       "      <td>35.0</td>\n",
       "      <td>female</td>\n",
       "      <td>3/25/2015</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>3.0</td>\n",
       "      <td>34</td>\n",
       "      <td>skin-ADNL</td>\n",
       "      <td>AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca006ONPN</th>\n",
       "      <td>6</td>\n",
       "      <td>F3</td>\n",
       "      <td>SA506</td>\n",
       "      <td>CGTGAGTG</td>\n",
       "      <td>SB703</td>\n",
       "      <td>TAGCAGCT</td>\n",
       "      <td>SB703SA506</td>\n",
       "      <td>TAGCAGCT-CGTGAGTG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>F3</td>\n",
       "      <td>...</td>\n",
       "      <td>35.0</td>\n",
       "      <td>female</td>\n",
       "      <td>3/25/2015</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>3.0</td>\n",
       "      <td>34</td>\n",
       "      <td>nares-AD</td>\n",
       "      <td>AD</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>462 rows × 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            PlateNumber PlateLocation     i5 i5Sequence     i7 i7Sequence  \\\n",
       "#sample-id                                                                  \n",
       "Ca009STL              1            A1  SA501   ATCGTACG  SA701   CGAGAGTT   \n",
       "900221                1            B1  SA502   ACTATCTG  SA701   CGAGAGTT   \n",
       "Ca010EBL              1            C1  SA503   TAGCGAGT  SA701   CGAGAGTT   \n",
       "900460                1            D1  SA504   CTGCGTGT  SA701   CGAGAGTT   \n",
       "900051                1            E1  SA505   TCATCGAG  SA701   CGAGAGTT   \n",
       "...                 ...           ...    ...        ...    ...        ...   \n",
       "900401                5           C12  SB503   AGAGTCAC  SB712   CGTAGCGA   \n",
       "900402                6            B4  SA502   ACTATCTG  SB704   TCTCTATG   \n",
       "Ca006ONL              6            F1  SA506   CGTGAGTG  SB701   CTCGACTT   \n",
       "Ca006ONNL             6            F2  SA506   CGTGAGTG  SB702   CGAAGTAT   \n",
       "Ca006ONPN             6            F3  SA506   CGTGAGTG  SB703   TAGCAGCT   \n",
       "\n",
       "            identifier           Sequence      Plate ID Well location  ...  \\\n",
       "#sample-id                                                             ...   \n",
       "Ca009STL    SA701SA501  CGAGAGTT-ATCGTACG  1.010000e+21            A1  ...   \n",
       "900221      SA701SA502  CGAGAGTT-ACTATCTG  1.010000e+21            B1  ...   \n",
       "Ca010EBL    SA701SA503  CGAGAGTT-TAGCGAGT  1.010000e+21            C1  ...   \n",
       "900460      SA701SA504  CGAGAGTT-CTGCGTGT  1.010000e+21            D1  ...   \n",
       "900051      SA701SA505  CGAGAGTT-TCATCGAG  1.010000e+21            E1  ...   \n",
       "...                ...                ...           ...           ...  ...   \n",
       "900401      SB712SB503  CGTAGCGA-AGAGTCAC  1.010000e+21           C12  ...   \n",
       "900402      SB704SA502  TCTCTATG-ACTATCTG  1.010000e+21            B4  ...   \n",
       "Ca006ONL    SB701SA506  CTCGACTT-CGTGAGTG  1.010000e+21            F1  ...   \n",
       "Ca006ONNL   SB702SA506  CGAAGTAT-CGTGAGTG  1.010000e+21            F2  ...   \n",
       "Ca006ONPN   SB703SA506  TAGCAGCT-CGTGAGTG  1.010000e+21            F3  ...   \n",
       "\n",
       "            age_months     sex enrolment_date enrolment_season hiv_exposure  \\\n",
       "#sample-id                                                                    \n",
       "Ca009STL          24.0    male      4/16/2015          Autumn     Unexposed   \n",
       "900221             9.0  female      8/11/2015           Winter    Unexposed   \n",
       "Ca010EBL          24.0  female     11/20/2014           Spring    Unexposed   \n",
       "900460            18.0  female      9/23/2015           Spring    Unexposed   \n",
       "900051            31.0    male      4/21/2015          Autumn     Unexposed   \n",
       "...                ...     ...            ...              ...          ...   \n",
       "900401            21.0  female      9/17/2015           Spring      Exposed   \n",
       "900402            21.0     NaN            NaN              NaN          NaN   \n",
       "Ca006ONL          35.0  female      3/25/2015           Autumn    Unexposed   \n",
       "Ca006ONNL         35.0  female      3/25/2015           Autumn    Unexposed   \n",
       "Ca006ONPN         35.0  female      3/25/2015           Autumn    Unexposed   \n",
       "\n",
       "           hiv_status household_size o_scorad      group ad_status  \n",
       "#sample-id                                                          \n",
       "Ca009STL     negative            4.0       40   skin-ADL        AD  \n",
       "900221       negative            7.0       34   skin-ADL        AD  \n",
       "Ca010EBL     negative            7.0       21   skin-ADL        AD  \n",
       "900460            NaN            4.0       40   skin-ADL        AD  \n",
       "900051       negative            7.0       41   skin-ADL        AD  \n",
       "...               ...            ...      ...        ...       ...  \n",
       "900401       negative           12.0       38  skin-ADNL        AD  \n",
       "900402            NaN            NaN      NaN   nares-AD        AD  \n",
       "Ca006ONL     negative            3.0       34   skin-ADL        AD  \n",
       "Ca006ONNL    negative            3.0       34  skin-ADNL        AD  \n",
       "Ca006ONPN    negative            3.0       34   nares-AD        AD  \n",
       "\n",
       "[462 rows x 31 columns]"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load the metadata\n",
    "metadata_path = '../Metadata/16S_AD_South-Africa_metadata_subset.tsv'\n",
    "metadata = pd.read_csv(metadata_path, sep='\\t')\n",
    "\n",
    "metadata['#sample-id'] = metadata['#sample-id'].str.replace('_', '')\n",
    "# Set Sample-ID as the index for the metadata dataframe \n",
    "metadata = metadata.set_index('#sample-id')\n",
    "\n",
    "# Create group column based on case_type to simplify group names\n",
    "metadata['group'] = metadata['case_type'].map({\n",
    "    'case-lesional_skin': 'skin-ADL',\n",
    "    'case-nonlesional_skin': 'skin-ADNL', \n",
    "    'control-nonlesional_skin': 'skin-H',\n",
    "    'case-anterior_nares': 'nares-AD',\n",
    "    'control-anterior_nares': 'nares-H'\n",
    "})\n",
    "\n",
    "# Add AD status column based on group values\n",
    "metadata['ad_status'] = metadata['group'].apply(lambda x: 'AD' if x.split('-')[-1].startswith('AD') else 'H')\n",
    "\n",
    "metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>g__Streptococcus_ASV-1</th>\n",
       "      <th>g__Streptococcus_ASV-2</th>\n",
       "      <th>g__Corynebacterium_ASV-1</th>\n",
       "      <th>g__Corynebacterium_ASV-3</th>\n",
       "      <th>g___ASV-3</th>\n",
       "      <th>g__Cutibacterium_ASV-1</th>\n",
       "      <th>g___ASV-15</th>\n",
       "      <th>g___ASV-29</th>\n",
       "      <th>g___ASV-25</th>\n",
       "      <th>g__Cutibacterium_ASV-2</th>\n",
       "      <th>...</th>\n",
       "      <th>g__UBA6175_ASV-2</th>\n",
       "      <th>g__Dermacoccus_ASV-2</th>\n",
       "      <th>g__Comamonas_F_589250_ASV-4</th>\n",
       "      <th>g__Bosea_ASV-2</th>\n",
       "      <th>g__Capnocytophaga_820690_ASV-3</th>\n",
       "      <th>g__Blautia_A_141780_ASV-1</th>\n",
       "      <th>g___ASV-168</th>\n",
       "      <th>pid</th>\n",
       "      <th>group</th>\n",
       "      <th>microbiome_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>900344</th>\n",
       "      <td>188.0</td>\n",
       "      <td>115.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Co-130-MM</td>\n",
       "      <td>skin-H</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900459</th>\n",
       "      <td>20.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Ca-145-LM</td>\n",
       "      <td>nares-AD</td>\n",
       "      <td>nasal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900221</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Ca-101-ID</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900570</th>\n",
       "      <td>18.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Ca-046-MT</td>\n",
       "      <td>skin-ADNL</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900092</th>\n",
       "      <td>221.0</td>\n",
       "      <td>116.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Ca-023-EJ</td>\n",
       "      <td>nares-AD</td>\n",
       "      <td>nasal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900057</th>\n",
       "      <td>29.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Ca-013-NN</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900294</th>\n",
       "      <td>10.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Ca-113-MN</td>\n",
       "      <td>skin-ADNL</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900097</th>\n",
       "      <td>3.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Ca-025-AC</td>\n",
       "      <td>skin-ADNL</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900498</th>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Ca-157-LM</td>\n",
       "      <td>skin-ADNL</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900276</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>87.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Ca-111-IN</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>431 rows × 840 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        g__Streptococcus_ASV-1  g__Streptococcus_ASV-2  \\\n",
       "900344                   188.0                   115.0   \n",
       "900459                    20.0                    37.0   \n",
       "900221                       0                       0   \n",
       "900570                    18.0                       0   \n",
       "900092                   221.0                   116.0   \n",
       "...                        ...                     ...   \n",
       "900057                    29.0                       0   \n",
       "900294                    10.0                     3.0   \n",
       "900097                     3.0                       0   \n",
       "900498                     5.0                     5.0   \n",
       "900276                       0                       0   \n",
       "\n",
       "        g__Corynebacterium_ASV-1  g__Corynebacterium_ASV-3  g___ASV-3  \\\n",
       "900344                      23.0                      19.0        2.0   \n",
       "900459                         0                         0          0   \n",
       "900221                         0                         0          0   \n",
       "900570                         0                         0        1.0   \n",
       "900092                       3.0                       1.0          0   \n",
       "...                          ...                       ...        ...   \n",
       "900057                         0                         0          0   \n",
       "900294                         0                         0          0   \n",
       "900097                         0                         0          0   \n",
       "900498                         0                         0          0   \n",
       "900276                      15.0                         0          0   \n",
       "\n",
       "        g__Cutibacterium_ASV-1  g___ASV-15  g___ASV-29  g___ASV-25  \\\n",
       "900344                     2.0         1.0           0           0   \n",
       "900459                       0           0           0           0   \n",
       "900221                       0           0           0           0   \n",
       "900570                       0           0           0           0   \n",
       "900092                       0           0           0         1.0   \n",
       "...                        ...         ...         ...         ...   \n",
       "900057                     2.0         4.0           0           0   \n",
       "900294                    41.0           0           0           0   \n",
       "900097                     2.0           0           0           0   \n",
       "900498                    17.0           0         4.0           0   \n",
       "900276                    87.0           0           0           0   \n",
       "\n",
       "        g__Cutibacterium_ASV-2  ...  g__UBA6175_ASV-2  g__Dermacoccus_ASV-2  \\\n",
       "900344                       0  ...                 0                     0   \n",
       "900459                       0  ...                 0                     0   \n",
       "900221                       0  ...                 0                     0   \n",
       "900570                       0  ...                 0                     0   \n",
       "900092                       0  ...                 0                     0   \n",
       "...                        ...  ...               ...                   ...   \n",
       "900057                     3.0  ...                 0                     0   \n",
       "900294                    36.0  ...                 0                     0   \n",
       "900097                     1.0  ...                 0                     0   \n",
       "900498                     8.0  ...                 0                     0   \n",
       "900276                    34.0  ...                 0                     0   \n",
       "\n",
       "        g__Comamonas_F_589250_ASV-4  g__Bosea_ASV-2  \\\n",
       "900344                            0               0   \n",
       "900459                            0               0   \n",
       "900221                            0               0   \n",
       "900570                            0               0   \n",
       "900092                            0               0   \n",
       "...                             ...             ...   \n",
       "900057                            0               0   \n",
       "900294                            0               0   \n",
       "900097                            0               0   \n",
       "900498                            0               0   \n",
       "900276                            0               0   \n",
       "\n",
       "        g__Capnocytophaga_820690_ASV-3  g__Blautia_A_141780_ASV-1  \\\n",
       "900344                               0                          0   \n",
       "900459                               0                          0   \n",
       "900221                               0                          0   \n",
       "900570                               0                          0   \n",
       "900092                               0                          0   \n",
       "...                                ...                        ...   \n",
       "900057                               0                          0   \n",
       "900294                               0                          0   \n",
       "900097                               0                          0   \n",
       "900498                               0                          0   \n",
       "900276                               0                          0   \n",
       "\n",
       "        g___ASV-168        pid      group  microbiome_type  \n",
       "900344            0  Co-130-MM     skin-H             skin  \n",
       "900459            0  Ca-145-LM   nares-AD            nasal  \n",
       "900221            0  Ca-101-ID   skin-ADL             skin  \n",
       "900570            0  Ca-046-MT  skin-ADNL             skin  \n",
       "900092            0  Ca-023-EJ   nares-AD            nasal  \n",
       "...             ...        ...        ...              ...  \n",
       "900057            0  Ca-013-NN   skin-ADL             skin  \n",
       "900294            0  Ca-113-MN  skin-ADNL             skin  \n",
       "900097            0  Ca-025-AC  skin-ADNL             skin  \n",
       "900498            0  Ca-157-LM  skin-ADNL             skin  \n",
       "900276            0  Ca-111-IN   skin-ADL             skin  \n",
       "\n",
       "[431 rows x 840 columns]"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read in table at collapsed genera level\n",
    "biom_path = '../Data/Tables/Count_Tables/6_209766_feature_table_dedup_prev-filt-1pct_rare-350_Genus-ASV_all.biom'\n",
    "\n",
    "biom_tbl = load_table(biom_path)\n",
    "df = pd.DataFrame(biom_tbl.to_dataframe().T)\n",
    "\n",
    "# delete the prefix from the \n",
    "df.index = df.index.str.replace('15564.', '')\n",
    "\n",
    "# Map pid, group, and microbiome_type from metadata to df based on matching indexes\n",
    "df['pid'] = metadata.loc[df.index, 'pid']\n",
    "df['group'] = metadata.loc[df.index, 'group']\n",
    "df['microbiome_type'] = metadata.loc[df.index, 'specimen']\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of unique individuals: 183\n"
     ]
    }
   ],
   "source": [
    "print(f'Number of unique individuals: ' + str(df['pid'].nunique()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [],
   "source": [
    "# --- STEP 1: Prepare ---\n",
    "# Only keep numeric columns (i.e., ASV features)\n",
    "taxa_cols = df.select_dtypes(include=[np.number]).columns.tolist()\n",
    "\n",
    "\n",
    "# Split into skin and nares samples\n",
    "skin_samples = df[df['microbiome_type'] == 'skin']\n",
    "nares_samples = df[df['microbiome_type'] == 'nasal']\n",
    "\n",
    "# --- STEP 2: Separate by disease group ---\n",
    "\n",
    "# AD individuals\n",
    "skin_ADL_samples = skin_samples[skin_samples['group'] == 'skin-ADL']\n",
    "nares_AD_samples = nares_samples[nares_samples['group'] == 'nares-AD']\n",
    "\n",
    "# Healthy individuals\n",
    "skin_H_samples = skin_samples[skin_samples['group'] == 'skin-H']\n",
    "nares_H_samples = nares_samples[nares_samples['group'] == 'nares-H']\n",
    "\n",
    "# --- STEP 3: Find taxa for each group ---\n",
    "\n",
    "# For AD\n",
    "skin_present_AD = skin_ADL_samples[taxa_cols].gt(0).any()\n",
    "nares_present_AD = nares_AD_samples[taxa_cols].gt(0).any()\n",
    "\n",
    "skin_taxa_AD = set(skin_present_AD[skin_present_AD].index)\n",
    "nares_taxa_AD = set(nares_present_AD[nares_present_AD].index)\n",
    "\n",
    "# For Healthy\n",
    "skin_present_H = skin_H_samples[taxa_cols].gt(0).any()\n",
    "nares_present_H = nares_H_samples[taxa_cols].gt(0).any()\n",
    "\n",
    "skin_taxa_H = set(skin_present_H[skin_present_H].index)\n",
    "nares_taxa_H = set(nares_present_H[nares_present_H].index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of AD participants with both paired skin and nasal samples:  86\n"
     ]
    }
   ],
   "source": [
    "# Get unique patient IDs for each group\n",
    "skin_ADL_pids = set(skin_ADL_samples['pid'].unique())\n",
    "nares_AD_pids = set(nares_AD_samples['pid'].unique())\n",
    "\n",
    "# Count unique patients\n",
    "AD_pids_skin_nasal_overlap = skin_ADL_pids.intersection(nares_AD_pids)\n",
    "n_AD_pids_skin_nasal_overlap = len(AD_pids_skin_nasal_overlap)\n",
    "print(f'Number of AD participants with both paired skin and nasal samples: ', n_AD_pids_skin_nasal_overlap)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of Healthy participants with both paired skin and nasal samples:  72\n"
     ]
    }
   ],
   "source": [
    "# Get unique patient IDs for each group\n",
    "skin_H_pids = set(skin_H_samples['pid'].unique())\n",
    "nares_H_pids = set(nares_H_samples['pid'].unique())\n",
    "\n",
    "# Count unique patients\n",
    "H_pids_skin_nasal_overlap = skin_H_pids.intersection(nares_H_pids)\n",
    "n_H_pids_skin_nasal_overlap = len(H_pids_skin_nasal_overlap)\n",
    "\n",
    "print(f'Number of Healthy participants with both paired skin and nasal samples: ', n_H_pids_skin_nasal_overlap)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [],
   "source": [
    "# --- STEP 4: Plot Venn diagrams side by side ---\n",
    "fig, axes = plt.subplots(1, 2, figsize=(12, 6))  # 1 row, 2 columns\n",
    "\n",
    "# Colors\n",
    "skin_color = '#2a00ff'\n",
    "nares_color = '#ffa501'\n",
    "overlap_color = '#955280'\n",
    "\n",
    "# --- Venn for AD Individuals ---\n",
    "plt.sca(axes[0])  # set current axis\n",
    "v_AD = venn2(\n",
    "    [skin_taxa_AD, nares_taxa_AD],\n",
    "    set_labels=('', ''),\n",
    "    set_colors=(skin_color, nares_color),\n",
    "    alpha=0.5\n",
    ")\n",
    "\n",
    "# Customize font sizes and outlines\n",
    "for label in v_AD.subset_labels:\n",
    "    if label:\n",
    "        label.set_fontsize(16)\n",
    "\n",
    "region_colors = {\n",
    "    '10': skin_color,\n",
    "    '01': nares_color,\n",
    "    '11': overlap_color\n",
    "}\n",
    "for region_id, color in region_colors.items():\n",
    "    patch = v_AD.get_patch_by_id(region_id)\n",
    "    if patch:\n",
    "        patch.set_linewidth(2)\n",
    "        patch.set_edgecolor(color)\n",
    "\n",
    "axes[0].set_title(f'AD Individuals (n={n_AD_pids_skin_nasal_overlap})', fontsize=18, y=0.94)\n",
    "\n",
    "# --- Venn for Healthy Individuals ---\n",
    "plt.sca(axes[1])  # set current axis\n",
    "v_H = venn2(\n",
    "    [skin_taxa_H, nares_taxa_H],\n",
    "    set_labels=('', ''),\n",
    "    set_colors=(skin_color, nares_color),\n",
    "    alpha=0.5\n",
    ")\n",
    "\n",
    "for label in v_H.subset_labels:\n",
    "    if label:\n",
    "        label.set_fontsize(16)\n",
    "\n",
    "for region_id, color in region_colors.items():\n",
    "    patch = v_H.get_patch_by_id(region_id)\n",
    "    if patch:\n",
    "        patch.set_linewidth(2)\n",
    "        patch.set_edgecolor(color)\n",
    "\n",
    "axes[1].set_title(f'Healthy Individuals (n={n_H_pids_skin_nasal_overlap})', fontsize=18, y=0.94)\n",
    "\n",
    "# --- Add Legend ---\n",
    "handles = [\n",
    "    mpatches.Patch(color=skin_color, alpha=0.5, label='Skin only'),\n",
    "    mpatches.Patch(color=overlap_color, alpha=0.5, label='Both'),\n",
    "    mpatches.Patch(color=nares_color, alpha=0.5, label='Nares only')\n",
    "]\n",
    "\n",
    "fig.legend(\n",
    "    handles=handles,\n",
    "    loc='lower center',\n",
    "    bbox_to_anchor=(0.5, 0.02),\n",
    "    ncol=3,\n",
    "    frameon=False,\n",
    "    fontsize=16\n",
    ")\n",
    "\n",
    "# --- Final layout tweaks ---\n",
    "plt.tight_layout()\n",
    "plt.subplots_adjust(top=0.88, bottom=0.15)\n",
    "\n",
    "# --- Supertitle ---\n",
    "fig.suptitle('ASV Overlap Between Skin and Nares', fontsize=20, y=0.99)\n",
    "\n",
    "# --- Annotation data ---\n",
    "# You already computed these\n",
    "shared_AD = len(skin_taxa_AD & nares_taxa_AD)\n",
    "total_AD = len(skin_taxa_AD | nares_taxa_AD)\n",
    "\n",
    "shared_H = len(skin_taxa_H & nares_taxa_H)\n",
    "total_H = len(skin_taxa_H | nares_taxa_H)\n",
    "\n",
    "# Run Fisher's test again if needed\n",
    "from scipy.stats import fisher_exact\n",
    "table = [[shared_AD, total_AD - shared_AD], [shared_H, total_H - shared_H]]\n",
    "oddsratio, p_value = fisher_exact(table)\n",
    "\n",
    "# Add p-value and odds ratio centrally below both\n",
    "fig.text(\n",
    "    0.5, 0.13,\n",
    "    f\"Fisher's Exact Test: p = 8.1×10⁻⁷, Odds Ratio = {oddsratio:.2f}\",\n",
    "    ha='center', fontsize=14\n",
    ")\n",
    "\n",
    "plt.savefig('../Figures/Main/Fig_3A.jpg', dpi=600)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fisher's Exact Test results:\n",
      "Contingency Table: [[372, 338], [272, 422]]\n",
      "Odds Ratio: 1.71\n",
      "p-value: 0.00000081\n"
     ]
    }
   ],
   "source": [
    "# --- For AD group ---\n",
    "shared_AD = len(skin_taxa_AD & nares_taxa_AD)\n",
    "total_AD = len(skin_taxa_AD | nares_taxa_AD)\n",
    "\n",
    "# --- For Healthy group ---\n",
    "shared_H = len(skin_taxa_H & nares_taxa_H)\n",
    "total_H = len(skin_taxa_H | nares_taxa_H)\n",
    "\n",
    "# --- Build 2x2 contingency table ---\n",
    "table = [\n",
    "    [shared_AD, total_AD - shared_AD],\n",
    "    [shared_H, total_H - shared_H]\n",
    "]\n",
    "\n",
    "# --- Fisher's Exact Test ---\n",
    "oddsratio, p_value = fisher_exact(table)\n",
    "\n",
    "# --- Output Results ---\n",
    "print(f\"Fisher's Exact Test results:\")\n",
    "print(f\"Contingency Table: {table}\")\n",
    "print(f\"Odds Ratio: {oddsratio:.2f}\")\n",
    "print(f\"p-value: {p_value:.8f}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "qiime2-metagenome-2024.10",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
