{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# ASV Overlap Venn Diagram"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import Python packages\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import biom\n",
    "from biom import load_table\n",
    "import numpy as np\n",
    "from scipy.stats import pearsonr\n",
    "from matplotlib_venn import venn2\n",
    "import matplotlib.patches as mpatches\n",
    "from scipy.stats import mannwhitneyu\n",
    "from biom import Table\n",
    "from gemelli.rpca import rpca\n",
    "from scipy.spatial.distance import euclidean\n",
    "import re\n",
    "import matplotlib.colors as mcolors\n",
    "from scipy.stats import fisher_exact\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PlateNumber</th>\n",
       "      <th>PlateLocation</th>\n",
       "      <th>i5</th>\n",
       "      <th>i5Sequence</th>\n",
       "      <th>i7</th>\n",
       "      <th>i7Sequence</th>\n",
       "      <th>identifier</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Plate ID</th>\n",
       "      <th>Well location</th>\n",
       "      <th>...</th>\n",
       "      <th>enrolment_season</th>\n",
       "      <th>hiv_exposure</th>\n",
       "      <th>hiv_status</th>\n",
       "      <th>household_size</th>\n",
       "      <th>o_scorad</th>\n",
       "      <th>FWD_filepath</th>\n",
       "      <th>REV_filepath</th>\n",
       "      <th>group</th>\n",
       "      <th>microbiome_type</th>\n",
       "      <th>ad_status</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#sample-id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Ca009STL</th>\n",
       "      <td>1</td>\n",
       "      <td>A1</td>\n",
       "      <td>SA501</td>\n",
       "      <td>ATCGTACG</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA501</td>\n",
       "      <td>CGAGAGTT-ATCGTACG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>A1</td>\n",
       "      <td>...</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>4.0</td>\n",
       "      <td>40</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>skin</td>\n",
       "      <td>AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900221</th>\n",
       "      <td>1</td>\n",
       "      <td>B1</td>\n",
       "      <td>SA502</td>\n",
       "      <td>ACTATCTG</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA502</td>\n",
       "      <td>CGAGAGTT-ACTATCTG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>B1</td>\n",
       "      <td>...</td>\n",
       "      <td>Winter</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>7.0</td>\n",
       "      <td>34</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>skin</td>\n",
       "      <td>AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca010EBL</th>\n",
       "      <td>1</td>\n",
       "      <td>C1</td>\n",
       "      <td>SA503</td>\n",
       "      <td>TAGCGAGT</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA503</td>\n",
       "      <td>CGAGAGTT-TAGCGAGT</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>C1</td>\n",
       "      <td>...</td>\n",
       "      <td>Spring</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>7.0</td>\n",
       "      <td>21</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>skin</td>\n",
       "      <td>AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900460</th>\n",
       "      <td>1</td>\n",
       "      <td>D1</td>\n",
       "      <td>SA504</td>\n",
       "      <td>CTGCGTGT</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA504</td>\n",
       "      <td>CGAGAGTT-CTGCGTGT</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>D1</td>\n",
       "      <td>...</td>\n",
       "      <td>Spring</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>40</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>skin</td>\n",
       "      <td>AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900051</th>\n",
       "      <td>1</td>\n",
       "      <td>E1</td>\n",
       "      <td>SA505</td>\n",
       "      <td>TCATCGAG</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA505</td>\n",
       "      <td>CGAGAGTT-TCATCGAG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>E1</td>\n",
       "      <td>...</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>7.0</td>\n",
       "      <td>41</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>skin</td>\n",
       "      <td>AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca006ONL2</th>\n",
       "      <td>6</td>\n",
       "      <td>H1</td>\n",
       "      <td>SA508</td>\n",
       "      <td>GACACCGT</td>\n",
       "      <td>SB701</td>\n",
       "      <td>CTCGACTT</td>\n",
       "      <td>SB701SA508</td>\n",
       "      <td>CTCGACTT-GACACCGT</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>H1</td>\n",
       "      <td>...</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>3.0</td>\n",
       "      <td>34</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>skin</td>\n",
       "      <td>AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca006ONNL</th>\n",
       "      <td>6</td>\n",
       "      <td>F2</td>\n",
       "      <td>SA506</td>\n",
       "      <td>CGTGAGTG</td>\n",
       "      <td>SB702</td>\n",
       "      <td>CGAAGTAT</td>\n",
       "      <td>SB702SA506</td>\n",
       "      <td>CGAAGTAT-CGTGAGTG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>F2</td>\n",
       "      <td>...</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>3.0</td>\n",
       "      <td>34</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADNL</td>\n",
       "      <td>skin</td>\n",
       "      <td>AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca006ONNL2</th>\n",
       "      <td>6</td>\n",
       "      <td>H2</td>\n",
       "      <td>SA508</td>\n",
       "      <td>GACACCGT</td>\n",
       "      <td>SB702</td>\n",
       "      <td>CGAAGTAT</td>\n",
       "      <td>SB702SA508</td>\n",
       "      <td>CGAAGTAT-GACACCGT</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>H2</td>\n",
       "      <td>...</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>3.0</td>\n",
       "      <td>34</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADNL</td>\n",
       "      <td>skin</td>\n",
       "      <td>AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca006ONPN</th>\n",
       "      <td>6</td>\n",
       "      <td>F3</td>\n",
       "      <td>SA506</td>\n",
       "      <td>CGTGAGTG</td>\n",
       "      <td>SB703</td>\n",
       "      <td>TAGCAGCT</td>\n",
       "      <td>SB703SA506</td>\n",
       "      <td>TAGCAGCT-CGTGAGTG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>F3</td>\n",
       "      <td>...</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>3.0</td>\n",
       "      <td>34</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>nares-AD</td>\n",
       "      <td>nares</td>\n",
       "      <td>AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca006ONPN2</th>\n",
       "      <td>6</td>\n",
       "      <td>H3</td>\n",
       "      <td>SA508</td>\n",
       "      <td>GACACCGT</td>\n",
       "      <td>SB703</td>\n",
       "      <td>TAGCAGCT</td>\n",
       "      <td>SB703SA508</td>\n",
       "      <td>TAGCAGCT-GACACCGT</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>H3</td>\n",
       "      <td>...</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>3.0</td>\n",
       "      <td>34</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>nares-AD</td>\n",
       "      <td>nares</td>\n",
       "      <td>AD</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>502 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            PlateNumber PlateLocation     i5 i5Sequence     i7 i7Sequence  \\\n",
       "#sample-id                                                                  \n",
       "Ca009STL              1            A1  SA501   ATCGTACG  SA701   CGAGAGTT   \n",
       "900221                1            B1  SA502   ACTATCTG  SA701   CGAGAGTT   \n",
       "Ca010EBL              1            C1  SA503   TAGCGAGT  SA701   CGAGAGTT   \n",
       "900460                1            D1  SA504   CTGCGTGT  SA701   CGAGAGTT   \n",
       "900051                1            E1  SA505   TCATCGAG  SA701   CGAGAGTT   \n",
       "...                 ...           ...    ...        ...    ...        ...   \n",
       "Ca006ONL2             6            H1  SA508   GACACCGT  SB701   CTCGACTT   \n",
       "Ca006ONNL             6            F2  SA506   CGTGAGTG  SB702   CGAAGTAT   \n",
       "Ca006ONNL2            6            H2  SA508   GACACCGT  SB702   CGAAGTAT   \n",
       "Ca006ONPN             6            F3  SA506   CGTGAGTG  SB703   TAGCAGCT   \n",
       "Ca006ONPN2            6            H3  SA508   GACACCGT  SB703   TAGCAGCT   \n",
       "\n",
       "            identifier           Sequence      Plate ID Well location  ...  \\\n",
       "#sample-id                                                             ...   \n",
       "Ca009STL    SA701SA501  CGAGAGTT-ATCGTACG  1.010000e+21            A1  ...   \n",
       "900221      SA701SA502  CGAGAGTT-ACTATCTG  1.010000e+21            B1  ...   \n",
       "Ca010EBL    SA701SA503  CGAGAGTT-TAGCGAGT  1.010000e+21            C1  ...   \n",
       "900460      SA701SA504  CGAGAGTT-CTGCGTGT  1.010000e+21            D1  ...   \n",
       "900051      SA701SA505  CGAGAGTT-TCATCGAG  1.010000e+21            E1  ...   \n",
       "...                ...                ...           ...           ...  ...   \n",
       "Ca006ONL2   SB701SA508  CTCGACTT-GACACCGT  1.010000e+21            H1  ...   \n",
       "Ca006ONNL   SB702SA506  CGAAGTAT-CGTGAGTG  1.010000e+21            F2  ...   \n",
       "Ca006ONNL2  SB702SA508  CGAAGTAT-GACACCGT  1.010000e+21            H2  ...   \n",
       "Ca006ONPN   SB703SA506  TAGCAGCT-CGTGAGTG  1.010000e+21            F3  ...   \n",
       "Ca006ONPN2  SB703SA508  TAGCAGCT-GACACCGT  1.010000e+21            H3  ...   \n",
       "\n",
       "            enrolment_season hiv_exposure hiv_status household_size o_scorad  \\\n",
       "#sample-id                                                                     \n",
       "Ca009STL             Autumn     Unexposed   negative            4.0       40   \n",
       "900221                Winter    Unexposed   negative            7.0       34   \n",
       "Ca010EBL              Spring    Unexposed   negative            7.0       21   \n",
       "900460                Spring    Unexposed        NaN            4.0       40   \n",
       "900051               Autumn     Unexposed   negative            7.0       41   \n",
       "...                      ...          ...        ...            ...      ...   \n",
       "Ca006ONL2             Autumn    Unexposed   negative            3.0       34   \n",
       "Ca006ONNL             Autumn    Unexposed   negative            3.0       34   \n",
       "Ca006ONNL2            Autumn    Unexposed   negative            3.0       34   \n",
       "Ca006ONPN             Autumn    Unexposed   negative            3.0       34   \n",
       "Ca006ONPN2            Autumn    Unexposed   negative            3.0       34   \n",
       "\n",
       "                                                 FWD_filepath  \\\n",
       "#sample-id                                                      \n",
       "Ca009STL    /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "900221      /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "Ca010EBL    /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "900460      /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "900051      /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "...                                                       ...   \n",
       "Ca006ONL2   /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "Ca006ONNL   /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "Ca006ONNL2  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "Ca006ONPN   /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "Ca006ONPN2  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "\n",
       "                                                 REV_filepath      group  \\\n",
       "#sample-id                                                                 \n",
       "Ca009STL    /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   skin-ADL   \n",
       "900221      /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   skin-ADL   \n",
       "Ca010EBL    /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   skin-ADL   \n",
       "900460      /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   skin-ADL   \n",
       "900051      /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   skin-ADL   \n",
       "...                                                       ...        ...   \n",
       "Ca006ONL2   /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   skin-ADL   \n",
       "Ca006ONNL   /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...  skin-ADNL   \n",
       "Ca006ONNL2  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...  skin-ADNL   \n",
       "Ca006ONPN   /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   nares-AD   \n",
       "Ca006ONPN2  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   nares-AD   \n",
       "\n",
       "           microbiome_type ad_status  \n",
       "#sample-id                            \n",
       "Ca009STL              skin        AD  \n",
       "900221                skin        AD  \n",
       "Ca010EBL              skin        AD  \n",
       "900460                skin        AD  \n",
       "900051                skin        AD  \n",
       "...                    ...       ...  \n",
       "Ca006ONL2             skin        AD  \n",
       "Ca006ONNL             skin        AD  \n",
       "Ca006ONNL2            skin        AD  \n",
       "Ca006ONPN            nares        AD  \n",
       "Ca006ONPN2           nares        AD  \n",
       "\n",
       "[502 rows x 34 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load the metadata\n",
    "metadata_path = '../Data/Metadata/updated_clean_ant_skin_metadata_microbiome_type.tab'\n",
    "metadata = pd.read_csv(metadata_path, sep='\\t')\n",
    "\n",
    "metadata['#sample-id'] = metadata['#sample-id'].str.replace('_', '')\n",
    "# Set Sample-ID as the index for the metadata dataframe \n",
    "metadata = metadata.set_index('#sample-id')\n",
    "\n",
    "# Create group column based on case_type to simplify group names\n",
    "metadata['group'] = metadata['case_type'].map({\n",
    "    'case-lesional skin': 'skin-ADL',\n",
    "    'case-nonlesional skin': 'skin-ADNL', \n",
    "    'control-nonlesional skin': 'skin-H',\n",
    "    'case-anterior nares': 'nares-AD',\n",
    "    'control-anterior nares': 'nares-H'\n",
    "})\n",
    "\n",
    "# Add AD status column based on group values\n",
    "metadata['ad_status'] = metadata['group'].apply(lambda x: 'AD' if x.split('-')[-1].startswith('AD') else 'H')\n",
    "\n",
    "metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>g__Streptococcus_ASV-1</th>\n",
       "      <th>g__Streptococcus_ASV-2</th>\n",
       "      <th>g__Corynebacterium_ASV-1</th>\n",
       "      <th>g__Corynebacterium_ASV-3</th>\n",
       "      <th>g___ASV-3</th>\n",
       "      <th>g__Cutibacterium_ASV-1</th>\n",
       "      <th>g___ASV-16</th>\n",
       "      <th>g___ASV-25</th>\n",
       "      <th>g___ASV-27</th>\n",
       "      <th>g__Cutibacterium_ASV-2</th>\n",
       "      <th>...</th>\n",
       "      <th>g___ASV-126</th>\n",
       "      <th>g__Leptotrichia_A_993758_ASV-16</th>\n",
       "      <th>g__Capnocytophaga_820688_ASV-7</th>\n",
       "      <th>g__UBA6175_ASV-2</th>\n",
       "      <th>g__Comamonas_F_589250_ASV-4</th>\n",
       "      <th>g__Bosea_ASV-2</th>\n",
       "      <th>g___ASV-119</th>\n",
       "      <th>pid</th>\n",
       "      <th>group</th>\n",
       "      <th>microbiome_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>900344</th>\n",
       "      <td>156.0</td>\n",
       "      <td>95.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Co-130-MM</td>\n",
       "      <td>skin-H</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900459</th>\n",
       "      <td>21.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Ca-145-LM</td>\n",
       "      <td>nares-AD</td>\n",
       "      <td>nares</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900221</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Ca-101-ID</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900570</th>\n",
       "      <td>18.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Ca-046-MT</td>\n",
       "      <td>skin-ADNL</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900092</th>\n",
       "      <td>174.0</td>\n",
       "      <td>104.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Ca-023-EJ</td>\n",
       "      <td>nares-AD</td>\n",
       "      <td>nares</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9003972</th>\n",
       "      <td>38.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Ca-125-IM</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900097</th>\n",
       "      <td>3.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Ca-025-AC</td>\n",
       "      <td>skin-ADNL</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900498</th>\n",
       "      <td>4.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Ca-157-LM</td>\n",
       "      <td>skin-ADNL</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900276</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>69.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Ca-111-IN</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900406</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Ca-127-OM</td>\n",
       "      <td>skin-ADNL</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>502 rows × 801 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         g__Streptococcus_ASV-1  g__Streptococcus_ASV-2  \\\n",
       "900344                    156.0                    95.0   \n",
       "900459                     21.0                    30.0   \n",
       "900221                        0                       0   \n",
       "900570                     18.0                       0   \n",
       "900092                    174.0                   104.0   \n",
       "...                         ...                     ...   \n",
       "9003972                    38.0                    21.0   \n",
       "900097                      3.0                       0   \n",
       "900498                      4.0                     6.0   \n",
       "900276                        0                       0   \n",
       "900406                        0                       0   \n",
       "\n",
       "         g__Corynebacterium_ASV-1  g__Corynebacterium_ASV-3  g___ASV-3  \\\n",
       "900344                       23.0                      17.0        1.0   \n",
       "900459                          0                         0          0   \n",
       "900221                          0                         0          0   \n",
       "900570                          0                         0          0   \n",
       "900092                       10.0                       1.0          0   \n",
       "...                           ...                       ...        ...   \n",
       "9003972                         0                         0        1.0   \n",
       "900097                          0                         0          0   \n",
       "900498                          0                         0          0   \n",
       "900276                       10.0                         0          0   \n",
       "900406                          0                         0          0   \n",
       "\n",
       "         g__Cutibacterium_ASV-1  g___ASV-16  g___ASV-25  g___ASV-27  \\\n",
       "900344                      2.0         1.0           0           0   \n",
       "900459                        0           0           0           0   \n",
       "900221                        0           0           0           0   \n",
       "900570                        0           0           0           0   \n",
       "900092                        0           0           0         2.0   \n",
       "...                         ...         ...         ...         ...   \n",
       "9003972                    18.0           0           0           0   \n",
       "900097                      3.0           0           0           0   \n",
       "900498                     13.0           0         6.0           0   \n",
       "900276                     69.0           0           0           0   \n",
       "900406                        0           0           0           0   \n",
       "\n",
       "         g__Cutibacterium_ASV-2  ...  g___ASV-126  \\\n",
       "900344                        0  ...            0   \n",
       "900459                        0  ...            0   \n",
       "900221                        0  ...            0   \n",
       "900570                        0  ...            0   \n",
       "900092                        0  ...            0   \n",
       "...                         ...  ...          ...   \n",
       "9003972                    12.0  ...            0   \n",
       "900097                        0  ...            0   \n",
       "900498                     13.0  ...            0   \n",
       "900276                     26.0  ...            0   \n",
       "900406                        0  ...            0   \n",
       "\n",
       "         g__Leptotrichia_A_993758_ASV-16  g__Capnocytophaga_820688_ASV-7  \\\n",
       "900344                                 0                               0   \n",
       "900459                                 0                               0   \n",
       "900221                                 0                               0   \n",
       "900570                                 0                               0   \n",
       "900092                                 0                               0   \n",
       "...                                  ...                             ...   \n",
       "9003972                                0                               0   \n",
       "900097                                 0                               0   \n",
       "900498                                 0                               0   \n",
       "900276                                 0                               0   \n",
       "900406                                 0                               0   \n",
       "\n",
       "         g__UBA6175_ASV-2  g__Comamonas_F_589250_ASV-4  g__Bosea_ASV-2  \\\n",
       "900344                  0                            0               0   \n",
       "900459                  0                            0               0   \n",
       "900221                  0                            0               0   \n",
       "900570                  0                            0               0   \n",
       "900092                  0                            0               0   \n",
       "...                   ...                          ...             ...   \n",
       "9003972                 0                            0               0   \n",
       "900097                  0                            0               0   \n",
       "900498                  0                            0               0   \n",
       "900276                  0                            0               0   \n",
       "900406                  0                            0               0   \n",
       "\n",
       "         g___ASV-119        pid      group  microbiome_type  \n",
       "900344             0  Co-130-MM     skin-H             skin  \n",
       "900459             0  Ca-145-LM   nares-AD            nares  \n",
       "900221             0  Ca-101-ID   skin-ADL             skin  \n",
       "900570             0  Ca-046-MT  skin-ADNL             skin  \n",
       "900092             0  Ca-023-EJ   nares-AD            nares  \n",
       "...              ...        ...        ...              ...  \n",
       "9003972            0  Ca-125-IM   skin-ADL             skin  \n",
       "900097             0  Ca-025-AC  skin-ADNL             skin  \n",
       "900498             0  Ca-157-LM  skin-ADNL             skin  \n",
       "900276             0  Ca-111-IN   skin-ADL             skin  \n",
       "900406             0  Ca-127-OM  skin-ADNL             skin  \n",
       "\n",
       "[502 rows x 801 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read in table at collapsed genera level\n",
    "taxa_level = 'Genus'\n",
    "biom_path = f'../Data/Tables/Absolute_Abundance_Tables/209766_filtered_by_prevalence_1pct_rare_{taxa_level}-ASV-non-collapse.biom'\n",
    "\n",
    "biom_tbl = load_table(biom_path)\n",
    "df = pd.DataFrame(biom_tbl.to_dataframe().T)\n",
    "\n",
    "# delete the prefix from the \n",
    "df.index = df.index.str.replace('15564.', '')\n",
    "\n",
    "# Map pid, group, and microbiome_type from metadata to df based on matching indexes\n",
    "df['pid'] = metadata.loc[df.index, 'pid']\n",
    "df['group'] = metadata.loc[df.index, 'group']\n",
    "df['microbiome_type'] = metadata.loc[df.index, 'microbiome_type']\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of unique individuals: 197\n"
     ]
    }
   ],
   "source": [
    "print(f'Number of unique individuals: ' + str(df['pid'].nunique()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# --- STEP 1: Prepare ---\n",
    "# Only keep numeric columns (i.e., ASV features)\n",
    "taxa_cols = df.select_dtypes(include=[np.number]).columns.tolist()\n",
    "\n",
    "\n",
    "# Split into skin and nares samples\n",
    "skin_samples = df[df['microbiome_type'] == 'skin']\n",
    "nares_samples = df[df['microbiome_type'] == 'nares']\n",
    "\n",
    "# --- STEP 2: Separate by disease group ---\n",
    "\n",
    "# AD individuals\n",
    "skin_ADL_samples = skin_samples[skin_samples['group'] == 'skin-ADL']\n",
    "nares_AD_samples = nares_samples[nares_samples['group'] == 'nares-AD']\n",
    "\n",
    "# Healthy individuals\n",
    "skin_H_samples = skin_samples[skin_samples['group'] == 'skin-H']\n",
    "nares_H_samples = nares_samples[nares_samples['group'] == 'nares-H']\n",
    "\n",
    "# --- STEP 3: Find taxa for each group ---\n",
    "\n",
    "# For AD\n",
    "skin_present_AD = skin_ADL_samples[taxa_cols].gt(0).any()\n",
    "nares_present_AD = nares_AD_samples[taxa_cols].gt(0).any()\n",
    "\n",
    "skin_taxa_AD = set(skin_present_AD[skin_present_AD].index)\n",
    "nares_taxa_AD = set(nares_present_AD[nares_present_AD].index)\n",
    "\n",
    "# For Healthy\n",
    "skin_present_H = skin_H_samples[taxa_cols].gt(0).any()\n",
    "nares_present_H = nares_H_samples[taxa_cols].gt(0).any()\n",
    "\n",
    "skin_taxa_H = set(skin_present_H[skin_present_H].index)\n",
    "nares_taxa_H = set(nares_present_H[nares_present_H].index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get unique patient IDs for each group\n",
    "skin_ADL_pids = set(skin_ADL_samples['pid'].unique())\n",
    "nares_AD_pids = set(nares_AD_samples['pid'].unique())\n",
    "\n",
    "# Count unique patients\n",
    "n_AD = len(skin_ADL_pids.union(nares_AD_pids))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get unique patient IDs for each group\n",
    "skin_H_pids = set(skin_H_samples['pid'].unique())\n",
    "nares_H_pids = set(nares_H_samples['pid'].unique())\n",
    "\n",
    "# Count unique patients\n",
    "n_H = len(skin_H_pids.union(nares_H_pids))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# --- STEP 4: Plot Venn diagrams side by side ---\n",
    "fig, axes = plt.subplots(1, 2, figsize=(12, 6))  # 1 row, 2 columns\n",
    "\n",
    "# Colors\n",
    "skin_color = '#2a00ff'\n",
    "nares_color = '#ffa501'\n",
    "overlap_color = '#955280'\n",
    "\n",
    "# --- Venn for AD Individuals ---\n",
    "plt.sca(axes[0])  # set current axis\n",
    "v_AD = venn2(\n",
    "    [skin_taxa_AD, nares_taxa_AD],\n",
    "    set_labels=('', ''),\n",
    "    set_colors=(skin_color, nares_color),\n",
    "    alpha=0.5\n",
    ")\n",
    "\n",
    "# Customize font sizes and outlines\n",
    "for label in v_AD.subset_labels:\n",
    "    if label:\n",
    "        label.set_fontsize(16)\n",
    "\n",
    "region_colors = {\n",
    "    '10': skin_color,\n",
    "    '01': nares_color,\n",
    "    '11': overlap_color\n",
    "}\n",
    "for region_id, color in region_colors.items():\n",
    "    patch = v_AD.get_patch_by_id(region_id)\n",
    "    if patch:\n",
    "        patch.set_linewidth(2)\n",
    "        patch.set_edgecolor(color)\n",
    "\n",
    "axes[0].set_title(f'AD Individuals (n=107)', fontsize=18, y=0.94)\n",
    "\n",
    "# --- Venn for Healthy Individuals ---\n",
    "plt.sca(axes[1])  # set current axis\n",
    "v_H = venn2(\n",
    "    [skin_taxa_H, nares_taxa_H],\n",
    "    set_labels=('', ''),\n",
    "    set_colors=(skin_color, nares_color),\n",
    "    alpha=0.5\n",
    ")\n",
    "\n",
    "for label in v_H.subset_labels:\n",
    "    if label:\n",
    "        label.set_fontsize(16)\n",
    "\n",
    "for region_id, color in region_colors.items():\n",
    "    patch = v_H.get_patch_by_id(region_id)\n",
    "    if patch:\n",
    "        patch.set_linewidth(2)\n",
    "        patch.set_edgecolor(color)\n",
    "\n",
    "axes[1].set_title(f'Healthy Individuals (n=88)', fontsize=18, y=0.94)\n",
    "\n",
    "# --- Add Legend ---\n",
    "handles = [\n",
    "    mpatches.Patch(color=skin_color, alpha=0.5, label='Skin only'),\n",
    "    mpatches.Patch(color=overlap_color, alpha=0.5, label='Both'),\n",
    "    mpatches.Patch(color=nares_color, alpha=0.5, label='Nares only')\n",
    "]\n",
    "\n",
    "fig.legend(\n",
    "    handles=handles,\n",
    "    loc='lower center',\n",
    "    bbox_to_anchor=(0.5, 0.02),\n",
    "    ncol=3,\n",
    "    frameon=False,\n",
    "    fontsize=16\n",
    ")\n",
    "\n",
    "# --- Final layout tweaks ---\n",
    "plt.tight_layout()\n",
    "plt.subplots_adjust(top=0.88, bottom=0.15)\n",
    "\n",
    "# --- Supertitle ---\n",
    "fig.suptitle('ASV Overlap Between Skin and Nares', fontsize=20, y=0.99)\n",
    "\n",
    "# --- Annotation data ---\n",
    "# You already computed these\n",
    "shared_AD = len(skin_taxa_AD & nares_taxa_AD)\n",
    "total_AD = len(skin_taxa_AD | nares_taxa_AD)\n",
    "\n",
    "shared_H = len(skin_taxa_H & nares_taxa_H)\n",
    "total_H = len(skin_taxa_H | nares_taxa_H)\n",
    "\n",
    "# Run Fisher's test again if needed\n",
    "from scipy.stats import fisher_exact\n",
    "table = [[shared_AD, total_AD - shared_AD], [shared_H, total_H - shared_H]]\n",
    "oddsratio, p_value = fisher_exact(table)\n",
    "\n",
    "# Add p-value and odds ratio centrally below both\n",
    "fig.text(\n",
    "    0.5, 0.13,\n",
    "    f\"Fisher's Exact Test: p < 0.0001, Odds Ratio = {oddsratio:.2f}\",\n",
    "    ha='center', fontsize=14\n",
    ")\n",
    "\n",
    "# plt.savefig('../Figures/Main/Fig_3A.png', dpi=600)\n",
    "plt.savefig('../Figures/Main/JPG/Fig_3A.jpg', dpi=600)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fisher's Exact Test results:\n",
      "Contingency Table: [[369, 299], [254, 383]]\n",
      "Odds Ratio: 1.86\n",
      "p-value: 0.00000003\n"
     ]
    }
   ],
   "source": [
    "# --- For AD group ---\n",
    "shared_AD = len(skin_taxa_AD & nares_taxa_AD)\n",
    "total_AD = len(skin_taxa_AD | nares_taxa_AD)\n",
    "\n",
    "# --- For Healthy group ---\n",
    "shared_H = len(skin_taxa_H & nares_taxa_H)\n",
    "total_H = len(skin_taxa_H | nares_taxa_H)\n",
    "\n",
    "# --- Build 2x2 contingency table ---\n",
    "table = [\n",
    "    [shared_AD, total_AD - shared_AD],\n",
    "    [shared_H, total_H - shared_H]\n",
    "]\n",
    "\n",
    "# --- Fisher's Exact Test ---\n",
    "oddsratio, p_value = fisher_exact(table)\n",
    "\n",
    "# --- Output Results ---\n",
    "print(f\"Fisher's Exact Test results:\")\n",
    "print(f\"Contingency Table: {table}\")\n",
    "print(f\"Odds Ratio: {oddsratio:.2f}\")\n",
    "print(f\"p-value: {p_value:.8f}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "qiime2-metagenome-2024.10",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
