{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Cape Town vs Umtata Severity Distribution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import Python packages\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import biom\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from itertools import cycle\n",
    "import os\n",
    "from matplotlib.colors import ListedColormap\n",
    "from matplotlib.colors import to_rgba\n",
    "from scipy.stats import mannwhitneyu\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "from scipy.stats import ttest_ind\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PlateNumber</th>\n",
       "      <th>PlateLocation</th>\n",
       "      <th>i5</th>\n",
       "      <th>i5Sequence</th>\n",
       "      <th>i7</th>\n",
       "      <th>i7Sequence</th>\n",
       "      <th>identifier</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Plate ID</th>\n",
       "      <th>Well location</th>\n",
       "      <th>...</th>\n",
       "      <th>enrolment_season</th>\n",
       "      <th>hiv_exposure</th>\n",
       "      <th>hiv_status</th>\n",
       "      <th>household_size</th>\n",
       "      <th>o_scorad</th>\n",
       "      <th>FWD_filepath</th>\n",
       "      <th>REV_filepath</th>\n",
       "      <th>group</th>\n",
       "      <th>microbiome_type</th>\n",
       "      <th>ad_status</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#sample-id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Ca009STL</th>\n",
       "      <td>1</td>\n",
       "      <td>A1</td>\n",
       "      <td>SA501</td>\n",
       "      <td>ATCGTACG</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA501</td>\n",
       "      <td>CGAGAGTT-ATCGTACG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>A1</td>\n",
       "      <td>...</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>4.0</td>\n",
       "      <td>40</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>skin</td>\n",
       "      <td>AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900221</th>\n",
       "      <td>1</td>\n",
       "      <td>B1</td>\n",
       "      <td>SA502</td>\n",
       "      <td>ACTATCTG</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA502</td>\n",
       "      <td>CGAGAGTT-ACTATCTG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>B1</td>\n",
       "      <td>...</td>\n",
       "      <td>Winter</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>7.0</td>\n",
       "      <td>34</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>skin</td>\n",
       "      <td>AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca010EBL</th>\n",
       "      <td>1</td>\n",
       "      <td>C1</td>\n",
       "      <td>SA503</td>\n",
       "      <td>TAGCGAGT</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA503</td>\n",
       "      <td>CGAGAGTT-TAGCGAGT</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>C1</td>\n",
       "      <td>...</td>\n",
       "      <td>Spring</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>7.0</td>\n",
       "      <td>21</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>skin</td>\n",
       "      <td>AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900460</th>\n",
       "      <td>1</td>\n",
       "      <td>D1</td>\n",
       "      <td>SA504</td>\n",
       "      <td>CTGCGTGT</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA504</td>\n",
       "      <td>CGAGAGTT-CTGCGTGT</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>D1</td>\n",
       "      <td>...</td>\n",
       "      <td>Spring</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>40</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>skin</td>\n",
       "      <td>AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900051</th>\n",
       "      <td>1</td>\n",
       "      <td>E1</td>\n",
       "      <td>SA505</td>\n",
       "      <td>TCATCGAG</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA505</td>\n",
       "      <td>CGAGAGTT-TCATCGAG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>E1</td>\n",
       "      <td>...</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>7.0</td>\n",
       "      <td>41</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>skin</td>\n",
       "      <td>AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca006ONL2</th>\n",
       "      <td>6</td>\n",
       "      <td>H1</td>\n",
       "      <td>SA508</td>\n",
       "      <td>GACACCGT</td>\n",
       "      <td>SB701</td>\n",
       "      <td>CTCGACTT</td>\n",
       "      <td>SB701SA508</td>\n",
       "      <td>CTCGACTT-GACACCGT</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>H1</td>\n",
       "      <td>...</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>3.0</td>\n",
       "      <td>34</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>skin</td>\n",
       "      <td>AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca006ONNL</th>\n",
       "      <td>6</td>\n",
       "      <td>F2</td>\n",
       "      <td>SA506</td>\n",
       "      <td>CGTGAGTG</td>\n",
       "      <td>SB702</td>\n",
       "      <td>CGAAGTAT</td>\n",
       "      <td>SB702SA506</td>\n",
       "      <td>CGAAGTAT-CGTGAGTG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>F2</td>\n",
       "      <td>...</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>3.0</td>\n",
       "      <td>34</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADNL</td>\n",
       "      <td>skin</td>\n",
       "      <td>AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca006ONNL2</th>\n",
       "      <td>6</td>\n",
       "      <td>H2</td>\n",
       "      <td>SA508</td>\n",
       "      <td>GACACCGT</td>\n",
       "      <td>SB702</td>\n",
       "      <td>CGAAGTAT</td>\n",
       "      <td>SB702SA508</td>\n",
       "      <td>CGAAGTAT-GACACCGT</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>H2</td>\n",
       "      <td>...</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>3.0</td>\n",
       "      <td>34</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADNL</td>\n",
       "      <td>skin</td>\n",
       "      <td>AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca006ONPN</th>\n",
       "      <td>6</td>\n",
       "      <td>F3</td>\n",
       "      <td>SA506</td>\n",
       "      <td>CGTGAGTG</td>\n",
       "      <td>SB703</td>\n",
       "      <td>TAGCAGCT</td>\n",
       "      <td>SB703SA506</td>\n",
       "      <td>TAGCAGCT-CGTGAGTG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>F3</td>\n",
       "      <td>...</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>3.0</td>\n",
       "      <td>34</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>nares-AD</td>\n",
       "      <td>nares</td>\n",
       "      <td>AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca006ONPN2</th>\n",
       "      <td>6</td>\n",
       "      <td>H3</td>\n",
       "      <td>SA508</td>\n",
       "      <td>GACACCGT</td>\n",
       "      <td>SB703</td>\n",
       "      <td>TAGCAGCT</td>\n",
       "      <td>SB703SA508</td>\n",
       "      <td>TAGCAGCT-GACACCGT</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>H3</td>\n",
       "      <td>...</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>3.0</td>\n",
       "      <td>34</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>nares-AD</td>\n",
       "      <td>nares</td>\n",
       "      <td>AD</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>502 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            PlateNumber PlateLocation     i5 i5Sequence     i7 i7Sequence  \\\n",
       "#sample-id                                                                  \n",
       "Ca009STL              1            A1  SA501   ATCGTACG  SA701   CGAGAGTT   \n",
       "900221                1            B1  SA502   ACTATCTG  SA701   CGAGAGTT   \n",
       "Ca010EBL              1            C1  SA503   TAGCGAGT  SA701   CGAGAGTT   \n",
       "900460                1            D1  SA504   CTGCGTGT  SA701   CGAGAGTT   \n",
       "900051                1            E1  SA505   TCATCGAG  SA701   CGAGAGTT   \n",
       "...                 ...           ...    ...        ...    ...        ...   \n",
       "Ca006ONL2             6            H1  SA508   GACACCGT  SB701   CTCGACTT   \n",
       "Ca006ONNL             6            F2  SA506   CGTGAGTG  SB702   CGAAGTAT   \n",
       "Ca006ONNL2            6            H2  SA508   GACACCGT  SB702   CGAAGTAT   \n",
       "Ca006ONPN             6            F3  SA506   CGTGAGTG  SB703   TAGCAGCT   \n",
       "Ca006ONPN2            6            H3  SA508   GACACCGT  SB703   TAGCAGCT   \n",
       "\n",
       "            identifier           Sequence      Plate ID Well location  ...  \\\n",
       "#sample-id                                                             ...   \n",
       "Ca009STL    SA701SA501  CGAGAGTT-ATCGTACG  1.010000e+21            A1  ...   \n",
       "900221      SA701SA502  CGAGAGTT-ACTATCTG  1.010000e+21            B1  ...   \n",
       "Ca010EBL    SA701SA503  CGAGAGTT-TAGCGAGT  1.010000e+21            C1  ...   \n",
       "900460      SA701SA504  CGAGAGTT-CTGCGTGT  1.010000e+21            D1  ...   \n",
       "900051      SA701SA505  CGAGAGTT-TCATCGAG  1.010000e+21            E1  ...   \n",
       "...                ...                ...           ...           ...  ...   \n",
       "Ca006ONL2   SB701SA508  CTCGACTT-GACACCGT  1.010000e+21            H1  ...   \n",
       "Ca006ONNL   SB702SA506  CGAAGTAT-CGTGAGTG  1.010000e+21            F2  ...   \n",
       "Ca006ONNL2  SB702SA508  CGAAGTAT-GACACCGT  1.010000e+21            H2  ...   \n",
       "Ca006ONPN   SB703SA506  TAGCAGCT-CGTGAGTG  1.010000e+21            F3  ...   \n",
       "Ca006ONPN2  SB703SA508  TAGCAGCT-GACACCGT  1.010000e+21            H3  ...   \n",
       "\n",
       "            enrolment_season hiv_exposure hiv_status household_size o_scorad  \\\n",
       "#sample-id                                                                     \n",
       "Ca009STL             Autumn     Unexposed   negative            4.0       40   \n",
       "900221                Winter    Unexposed   negative            7.0       34   \n",
       "Ca010EBL              Spring    Unexposed   negative            7.0       21   \n",
       "900460                Spring    Unexposed        NaN            4.0       40   \n",
       "900051               Autumn     Unexposed   negative            7.0       41   \n",
       "...                      ...          ...        ...            ...      ...   \n",
       "Ca006ONL2             Autumn    Unexposed   negative            3.0       34   \n",
       "Ca006ONNL             Autumn    Unexposed   negative            3.0       34   \n",
       "Ca006ONNL2            Autumn    Unexposed   negative            3.0       34   \n",
       "Ca006ONPN             Autumn    Unexposed   negative            3.0       34   \n",
       "Ca006ONPN2            Autumn    Unexposed   negative            3.0       34   \n",
       "\n",
       "                                                 FWD_filepath  \\\n",
       "#sample-id                                                      \n",
       "Ca009STL    /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "900221      /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "Ca010EBL    /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "900460      /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "900051      /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "...                                                       ...   \n",
       "Ca006ONL2   /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "Ca006ONNL   /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "Ca006ONNL2  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "Ca006ONPN   /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "Ca006ONPN2  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "\n",
       "                                                 REV_filepath      group  \\\n",
       "#sample-id                                                                 \n",
       "Ca009STL    /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   skin-ADL   \n",
       "900221      /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   skin-ADL   \n",
       "Ca010EBL    /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   skin-ADL   \n",
       "900460      /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   skin-ADL   \n",
       "900051      /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   skin-ADL   \n",
       "...                                                       ...        ...   \n",
       "Ca006ONL2   /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   skin-ADL   \n",
       "Ca006ONNL   /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...  skin-ADNL   \n",
       "Ca006ONNL2  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...  skin-ADNL   \n",
       "Ca006ONPN   /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   nares-AD   \n",
       "Ca006ONPN2  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   nares-AD   \n",
       "\n",
       "           microbiome_type ad_status  \n",
       "#sample-id                            \n",
       "Ca009STL              skin        AD  \n",
       "900221                skin        AD  \n",
       "Ca010EBL              skin        AD  \n",
       "900460                skin        AD  \n",
       "900051                skin        AD  \n",
       "...                    ...       ...  \n",
       "Ca006ONL2             skin        AD  \n",
       "Ca006ONNL             skin        AD  \n",
       "Ca006ONNL2            skin        AD  \n",
       "Ca006ONPN            nares        AD  \n",
       "Ca006ONPN2           nares        AD  \n",
       "\n",
       "[502 rows x 34 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load the metadata\n",
    "metadata_path = '../Data/Metadata/updated_clean_ant_skin_metadata_microbiome_type.tab'\n",
    "metadata = pd.read_csv(metadata_path, sep='\\t')\n",
    "\n",
    "metadata['#sample-id'] = metadata['#sample-id'].str.replace('_', '')\n",
    "# Set Sample-ID as the index for the metadata dataframe \n",
    "metadata = metadata.set_index('#sample-id')\n",
    "\n",
    "\n",
    "# Create group column based on case_type to simplify group names\n",
    "metadata['group'] = metadata['case_type'].map({\n",
    "    'case-lesional skin': 'skin-ADL',\n",
    "    'case-nonlesional skin': 'skin-ADNL', \n",
    "    'control-nonlesional skin': 'skin-H',\n",
    "    'case-anterior nares': 'nares-AD',\n",
    "    'control-anterior nares': 'nares-H'\n",
    "})\n",
    "\n",
    "# Add AD status column based on group values\n",
    "metadata['ad_status'] = metadata['group'].apply(lambda x: 'AD' if x.split('-')[-1].startswith('AD') else 'H')\n",
    "\n",
    "metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Umtata participants by severity:\n",
      "Moderate (oSCORAD 15–40): 27\n",
      "Severe (oSCORAD >40): 32\n"
     ]
    }
   ],
   "source": [
    "# Convert o_scorad to numeric, handling any non-numeric values\n",
    "metadata['o_scorad'] = pd.to_numeric(metadata['o_scorad'], errors='coerce')\n",
    "# Convert to integer, NaN values will remain as NaN\n",
    "metadata['o_scorad'] = metadata['o_scorad'].astype('Int64')\n",
    "\n",
    "# Filter for Umtata participants\n",
    "umtata_metadata = metadata[metadata['area'] == 'Umtata']\n",
    "\n",
    "# Count participants in different severity ranges\n",
    "moderate_scorad = umtata_metadata[\n",
    "    (umtata_metadata['o_scorad'] >= 15) & \n",
    "    (umtata_metadata['o_scorad'] <= 40)\n",
    "]['pid'].nunique()\n",
    "\n",
    "severe_scorad = umtata_metadata[\n",
    "    umtata_metadata['o_scorad'] > 40\n",
    "]['pid'].nunique()\n",
    "\n",
    "print(\"\\nUmtata participants by severity:\")\n",
    "print(f\"Moderate (oSCORAD 15–40): {moderate_scorad}\")\n",
    "print(f\"Severe (oSCORAD >40): {severe_scorad}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Umtata participants by severity:\n",
      "Moderate (oSCORAD 15–40): 24\n",
      "Severe (oSCORAD >40): 21\n"
     ]
    }
   ],
   "source": [
    "# Convert o_scorad to numeric, handling any non-numeric values\n",
    "metadata['o_scorad'] = pd.to_numeric(metadata['o_scorad'], errors='coerce')\n",
    "# Convert to integer, NaN values will remain as NaN\n",
    "metadata['o_scorad'] = metadata['o_scorad'].astype('Int64')\n",
    "\n",
    "# Filter for Umtata participants\n",
    "umtata_metadata = metadata[metadata['area'] == 'Cape Town']\n",
    "\n",
    "# Count participants in different severity ranges\n",
    "moderate_scorad = umtata_metadata[\n",
    "    (umtata_metadata['o_scorad'] >= 15) & \n",
    "    (umtata_metadata['o_scorad'] <= 40)\n",
    "]['pid'].nunique()\n",
    "\n",
    "severe_scorad = umtata_metadata[\n",
    "    umtata_metadata['o_scorad'] > 40\n",
    "]['pid'].nunique()\n",
    "\n",
    "print(\"\\nUmtata participants by severity:\")\n",
    "print(f\"Moderate (oSCORAD 15–40): {moderate_scorad}\")\n",
    "print(f\"Severe (oSCORAD >40): {severe_scorad}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "All participants have consistent o_scorad values across their samples\n"
     ]
    }
   ],
   "source": [
    "# Check if each person has a unique o_scorad\n",
    "pid_scorad = metadata.groupby('pid')['o_scorad'].nunique()\n",
    "pids_with_multiple_scorads = pid_scorad[pid_scorad > 1]\n",
    "\n",
    "if len(pids_with_multiple_scorads) > 0:\n",
    "    print(\"\\nParticipants with multiple different o_scorad values:\")\n",
    "    print(pids_with_multiple_scorads)\n",
    "    \n",
    "    # Show the details for these participants\n",
    "    for pid in pids_with_multiple_scorads.index:\n",
    "        print(f\"\\nDetails for participant {pid}:\")\n",
    "        print(metadata[metadata['pid'] == pid][['o_scorad', 'group']])\n",
    "else:\n",
    "    print(\"\\nAll participants have consistent o_scorad values across their samples\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Sample counts by area and group:\n",
      "group      nares-AD  nares-H  skin-ADL  skin-ADNL  skin-H\n",
      "area                                                     \n",
      "Cape Town        47       34        46         50      35\n",
      "Umtata           61       55        61         61      52\n"
     ]
    }
   ],
   "source": [
    "# Get counts of samples by area and group\n",
    "area_group_counts = pd.crosstab(metadata['area'], metadata['group'])\n",
    "\n",
    "# Display the counts\n",
    "print(\"\\nSample counts by area and group:\")\n",
    "print(area_group_counts)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "All participants' samples are from a single area\n"
     ]
    }
   ],
   "source": [
    "# Check if samples from same pid are from different areas\n",
    "pid_areas = metadata.groupby('pid')['area'].nunique()\n",
    "pids_with_multiple_areas = pid_areas[pid_areas > 1]\n",
    "\n",
    "if len(pids_with_multiple_areas) > 0:\n",
    "    print(\"\\nParticipants with samples from multiple areas:\")\n",
    "    print(pids_with_multiple_areas)\n",
    "    \n",
    "    # Show the details for these participants\n",
    "    for pid in pids_with_multiple_areas.index:\n",
    "        print(f\"\\nDetails for participant {pid}:\")\n",
    "        print(metadata[metadata['pid'] == pid][['area', 'group']])\n",
    "else:\n",
    "    print(\"\\nAll participants' samples are from a single area\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "T-statistic: -0.887, P-value: 0.377\n"
     ]
    }
   ],
   "source": [
    "# Use one o_scorad per individual\n",
    "meta_pid = metadata[['pid', 'area', 'o_scorad']].drop_duplicates(subset='pid')\n",
    "meta_pid['o_scorad'] = pd.to_numeric(meta_pid['o_scorad'], errors='coerce')\n",
    "\n",
    "# Extract and filter data (remove oSCORAD < 15)\n",
    "ct_data = meta_pid[(meta_pid['area'] == 'Cape Town') & (meta_pid['o_scorad'] >= 15)]['o_scorad'].dropna()\n",
    "um_data = meta_pid[(meta_pid['area'] == 'Umtata') & (meta_pid['o_scorad'] >= 15)]['o_scorad'].dropna()\n",
    "\n",
    "\n",
    "# T-test\n",
    "t_stat, p_val = ttest_ind(ct_data, um_data)\n",
    "\n",
    "# Create figure with 2x2 layout (histogram + boxplot)\n",
    "fig, axes = plt.subplots(2, 2, figsize=(15, 6), gridspec_kw={'height_ratios': [3, 1]})\n",
    "\n",
    "# Cape Town histogram\n",
    "axes[0][0].hist(ct_data, bins=20, edgecolor='black', color='#d2b48c')\n",
    "axes[0][0].set_title(f'Cape Town (urban)', fontsize=20)\n",
    "axes[0][0].set_xlabel('oSCORAD Scores', fontsize=18)\n",
    "axes[0][0].set_ylabel('Count', fontsize=18)\n",
    "axes[0][0].tick_params(axis='both', labelsize=14)\n",
    "axes[0][0].set_ylim(0, 12)\n",
    "axes[0][0].set_yticks(range(0, 12, 4))\n",
    "axes[0][0].set_xlim(15, 83)\n",
    "axes[0][0].axvline(x=40, color='gray', linestyle='--', linewidth=2)\n",
    "axes[0][0].text(0.02, 0.95, '15-40: moderate', transform=axes[0][0].transAxes,\n",
    "                fontsize=13, verticalalignment='top',\n",
    "                bbox=dict(facecolor='white', edgecolor='none', alpha=0.8))\n",
    "axes[0][0].text(0.8, 0.95, '>40: severe', transform=axes[0][0].transAxes,\n",
    "                fontsize=13, verticalalignment='top',\n",
    "                bbox=dict(facecolor='white', edgecolor='none', alpha=0.8))\n",
    "\n",
    "# Umtata histogram\n",
    "axes[0][1].hist(um_data, bins=20, edgecolor='black', color='salmon')\n",
    "axes[0][1].set_title(f'Umtata (rural)', fontsize=20)\n",
    "axes[0][1].set_xlabel('oSCORAD Scores', fontsize=18)\n",
    "axes[0][1].set_ylabel('Count', fontsize=18)\n",
    "axes[0][1].tick_params(axis='both', labelsize=14)\n",
    "axes[0][1].set_ylim(0, 12)\n",
    "axes[0][1].set_yticks(range(0, 12, 4))\n",
    "axes[0][1].set_xlim(15, 83)\n",
    "axes[0][1].axvline(x=40, color='gray', linestyle='--', linewidth=2)\n",
    "axes[0][1].text(0.02, 0.95, '15-40: moderate', transform=axes[0][1].transAxes,\n",
    "                fontsize=13, verticalalignment='top',\n",
    "                bbox=dict(facecolor='white', edgecolor='none', alpha=0.8))\n",
    "axes[0][1].text(0.8, 0.95, '>40: severe', transform=axes[0][1].transAxes,\n",
    "                fontsize=13, verticalalignment='top',\n",
    "                bbox=dict(facecolor='white', edgecolor='none', alpha=0.8))\n",
    "\n",
    "# Cape Town boxplot\n",
    "axes[1][0].boxplot(ct_data, vert=False, patch_artist=True, widths=0.4,\n",
    "                   boxprops=dict(facecolor='#d2b48c', color='black'),\n",
    "                   medianprops=dict(color='black'))\n",
    "axes[1][0].set_yticks([])\n",
    "axes[1][0].set_xlim(15, 83)\n",
    "axes[1][0].set_xlabel('oSCORAD', fontsize=16)\n",
    "axes[1][0].tick_params(axis='x', labelsize=13)\n",
    "axes[1][0].text(76, 1.3, f\"n={len(ct_data)}\", ha='left', va='center', fontsize=13)\n",
    "\n",
    "# Umtata boxplot\n",
    "axes[1][1].boxplot(um_data, vert=False, patch_artist=True, widths=0.4,\n",
    "                   boxprops=dict(facecolor='salmon', color='black'),\n",
    "                   medianprops=dict(color='black'))\n",
    "axes[1][1].set_yticks([])\n",
    "axes[1][1].set_xlim(15, 83)\n",
    "axes[1][1].set_xlabel('oSCORAD', fontsize=16)\n",
    "axes[1][1].tick_params(axis='x', labelsize=13)\n",
    "axes[1][1].text(76, 1.3, f\"n={len(um_data)}\", ha='left', va='center', fontsize=13)\n",
    "\n",
    "\n",
    "# Final layout and save\n",
    "plt.suptitle('Distribution of AD Severity (oSCORAD) by Region', fontsize=22, y=0.95)\n",
    "plt.tight_layout(rect=[0, 0, 1, 0.93])\n",
    "plt.subplots_adjust(wspace=0.35, hspace=0.15)\n",
    "\n",
    "plt.savefig('../Figures/Main/Fig_2D,F.png', dpi=600)\n",
    "\n",
    "# Output t-test result\n",
    "print(f\"T-statistic: {t_stat:.3f}, P-value: {p_val:.3f}\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "qiime2-metagenome-2024.10",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
