{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Differential Taxa and Correlation with Severity in Cape Town and Umtata Skin Samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1173,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import Python packages\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import biom\n",
    "from biom import load_table\n",
    "import numpy as np\n",
    "from scipy.stats import spearmanr\n",
    "from scipy.stats import mannwhitneyu\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "import itertools\n",
    "import matplotlib.patches\n",
    "from scipy.stats import pearsonr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1174,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PlateNumber</th>\n",
       "      <th>PlateLocation</th>\n",
       "      <th>i5</th>\n",
       "      <th>i5Sequence</th>\n",
       "      <th>i7</th>\n",
       "      <th>i7Sequence</th>\n",
       "      <th>identifier</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Plate ID</th>\n",
       "      <th>Well location</th>\n",
       "      <th>...</th>\n",
       "      <th>sex</th>\n",
       "      <th>enrolment_date</th>\n",
       "      <th>enrolment_season</th>\n",
       "      <th>hiv_exposure</th>\n",
       "      <th>hiv_status</th>\n",
       "      <th>household_size</th>\n",
       "      <th>o_scorad</th>\n",
       "      <th>FWD_filepath</th>\n",
       "      <th>REV_filepath</th>\n",
       "      <th>group</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#sample-id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Ca009STL</th>\n",
       "      <td>1</td>\n",
       "      <td>A1</td>\n",
       "      <td>SA501</td>\n",
       "      <td>ATCGTACG</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA501</td>\n",
       "      <td>CGAGAGTT-ATCGTACG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>A1</td>\n",
       "      <td>...</td>\n",
       "      <td>male</td>\n",
       "      <td>4/16/2015</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>4.0</td>\n",
       "      <td>40</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900221</th>\n",
       "      <td>1</td>\n",
       "      <td>B1</td>\n",
       "      <td>SA502</td>\n",
       "      <td>ACTATCTG</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA502</td>\n",
       "      <td>CGAGAGTT-ACTATCTG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>B1</td>\n",
       "      <td>...</td>\n",
       "      <td>female</td>\n",
       "      <td>8/11/2015</td>\n",
       "      <td>Winter</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>7.0</td>\n",
       "      <td>34</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca010EBL</th>\n",
       "      <td>1</td>\n",
       "      <td>C1</td>\n",
       "      <td>SA503</td>\n",
       "      <td>TAGCGAGT</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA503</td>\n",
       "      <td>CGAGAGTT-TAGCGAGT</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>C1</td>\n",
       "      <td>...</td>\n",
       "      <td>female</td>\n",
       "      <td>11/20/2014</td>\n",
       "      <td>Spring</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>7.0</td>\n",
       "      <td>21</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900460</th>\n",
       "      <td>1</td>\n",
       "      <td>D1</td>\n",
       "      <td>SA504</td>\n",
       "      <td>CTGCGTGT</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA504</td>\n",
       "      <td>CGAGAGTT-CTGCGTGT</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>D1</td>\n",
       "      <td>...</td>\n",
       "      <td>female</td>\n",
       "      <td>9/23/2015</td>\n",
       "      <td>Spring</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>40</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900051</th>\n",
       "      <td>1</td>\n",
       "      <td>E1</td>\n",
       "      <td>SA505</td>\n",
       "      <td>TCATCGAG</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA505</td>\n",
       "      <td>CGAGAGTT-TCATCGAG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>E1</td>\n",
       "      <td>...</td>\n",
       "      <td>male</td>\n",
       "      <td>4/21/2015</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>7.0</td>\n",
       "      <td>41</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca006ONL2</th>\n",
       "      <td>6</td>\n",
       "      <td>H1</td>\n",
       "      <td>SA508</td>\n",
       "      <td>GACACCGT</td>\n",
       "      <td>SB701</td>\n",
       "      <td>CTCGACTT</td>\n",
       "      <td>SB701SA508</td>\n",
       "      <td>CTCGACTT-GACACCGT</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>H1</td>\n",
       "      <td>...</td>\n",
       "      <td>female</td>\n",
       "      <td>3/25/2015</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>3.0</td>\n",
       "      <td>34</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca006ONNL</th>\n",
       "      <td>6</td>\n",
       "      <td>F2</td>\n",
       "      <td>SA506</td>\n",
       "      <td>CGTGAGTG</td>\n",
       "      <td>SB702</td>\n",
       "      <td>CGAAGTAT</td>\n",
       "      <td>SB702SA506</td>\n",
       "      <td>CGAAGTAT-CGTGAGTG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>F2</td>\n",
       "      <td>...</td>\n",
       "      <td>female</td>\n",
       "      <td>3/25/2015</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>3.0</td>\n",
       "      <td>34</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADNL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca006ONNL2</th>\n",
       "      <td>6</td>\n",
       "      <td>H2</td>\n",
       "      <td>SA508</td>\n",
       "      <td>GACACCGT</td>\n",
       "      <td>SB702</td>\n",
       "      <td>CGAAGTAT</td>\n",
       "      <td>SB702SA508</td>\n",
       "      <td>CGAAGTAT-GACACCGT</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>H2</td>\n",
       "      <td>...</td>\n",
       "      <td>female</td>\n",
       "      <td>3/25/2015</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>3.0</td>\n",
       "      <td>34</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADNL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca006ONPN</th>\n",
       "      <td>6</td>\n",
       "      <td>F3</td>\n",
       "      <td>SA506</td>\n",
       "      <td>CGTGAGTG</td>\n",
       "      <td>SB703</td>\n",
       "      <td>TAGCAGCT</td>\n",
       "      <td>SB703SA506</td>\n",
       "      <td>TAGCAGCT-CGTGAGTG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>F3</td>\n",
       "      <td>...</td>\n",
       "      <td>female</td>\n",
       "      <td>3/25/2015</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>3.0</td>\n",
       "      <td>34</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>nares-AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca006ONPN2</th>\n",
       "      <td>6</td>\n",
       "      <td>H3</td>\n",
       "      <td>SA508</td>\n",
       "      <td>GACACCGT</td>\n",
       "      <td>SB703</td>\n",
       "      <td>TAGCAGCT</td>\n",
       "      <td>SB703SA508</td>\n",
       "      <td>TAGCAGCT-GACACCGT</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>H3</td>\n",
       "      <td>...</td>\n",
       "      <td>female</td>\n",
       "      <td>3/25/2015</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>3.0</td>\n",
       "      <td>34</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>nares-AD</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>502 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            PlateNumber PlateLocation     i5 i5Sequence     i7 i7Sequence  \\\n",
       "#sample-id                                                                  \n",
       "Ca009STL              1            A1  SA501   ATCGTACG  SA701   CGAGAGTT   \n",
       "900221                1            B1  SA502   ACTATCTG  SA701   CGAGAGTT   \n",
       "Ca010EBL              1            C1  SA503   TAGCGAGT  SA701   CGAGAGTT   \n",
       "900460                1            D1  SA504   CTGCGTGT  SA701   CGAGAGTT   \n",
       "900051                1            E1  SA505   TCATCGAG  SA701   CGAGAGTT   \n",
       "...                 ...           ...    ...        ...    ...        ...   \n",
       "Ca006ONL2             6            H1  SA508   GACACCGT  SB701   CTCGACTT   \n",
       "Ca006ONNL             6            F2  SA506   CGTGAGTG  SB702   CGAAGTAT   \n",
       "Ca006ONNL2            6            H2  SA508   GACACCGT  SB702   CGAAGTAT   \n",
       "Ca006ONPN             6            F3  SA506   CGTGAGTG  SB703   TAGCAGCT   \n",
       "Ca006ONPN2            6            H3  SA508   GACACCGT  SB703   TAGCAGCT   \n",
       "\n",
       "            identifier           Sequence      Plate ID Well location  ...  \\\n",
       "#sample-id                                                             ...   \n",
       "Ca009STL    SA701SA501  CGAGAGTT-ATCGTACG  1.010000e+21            A1  ...   \n",
       "900221      SA701SA502  CGAGAGTT-ACTATCTG  1.010000e+21            B1  ...   \n",
       "Ca010EBL    SA701SA503  CGAGAGTT-TAGCGAGT  1.010000e+21            C1  ...   \n",
       "900460      SA701SA504  CGAGAGTT-CTGCGTGT  1.010000e+21            D1  ...   \n",
       "900051      SA701SA505  CGAGAGTT-TCATCGAG  1.010000e+21            E1  ...   \n",
       "...                ...                ...           ...           ...  ...   \n",
       "Ca006ONL2   SB701SA508  CTCGACTT-GACACCGT  1.010000e+21            H1  ...   \n",
       "Ca006ONNL   SB702SA506  CGAAGTAT-CGTGAGTG  1.010000e+21            F2  ...   \n",
       "Ca006ONNL2  SB702SA508  CGAAGTAT-GACACCGT  1.010000e+21            H2  ...   \n",
       "Ca006ONPN   SB703SA506  TAGCAGCT-CGTGAGTG  1.010000e+21            F3  ...   \n",
       "Ca006ONPN2  SB703SA508  TAGCAGCT-GACACCGT  1.010000e+21            H3  ...   \n",
       "\n",
       "               sex enrolment_date enrolment_season hiv_exposure hiv_status  \\\n",
       "#sample-id                                                                   \n",
       "Ca009STL      male      4/16/2015          Autumn     Unexposed   negative   \n",
       "900221      female      8/11/2015           Winter    Unexposed   negative   \n",
       "Ca010EBL    female     11/20/2014           Spring    Unexposed   negative   \n",
       "900460      female      9/23/2015           Spring    Unexposed        NaN   \n",
       "900051        male      4/21/2015          Autumn     Unexposed   negative   \n",
       "...            ...            ...              ...          ...        ...   \n",
       "Ca006ONL2   female      3/25/2015           Autumn    Unexposed   negative   \n",
       "Ca006ONNL   female      3/25/2015           Autumn    Unexposed   negative   \n",
       "Ca006ONNL2  female      3/25/2015           Autumn    Unexposed   negative   \n",
       "Ca006ONPN   female      3/25/2015           Autumn    Unexposed   negative   \n",
       "Ca006ONPN2  female      3/25/2015           Autumn    Unexposed   negative   \n",
       "\n",
       "           household_size o_scorad  \\\n",
       "#sample-id                           \n",
       "Ca009STL              4.0       40   \n",
       "900221                7.0       34   \n",
       "Ca010EBL              7.0       21   \n",
       "900460                4.0       40   \n",
       "900051                7.0       41   \n",
       "...                   ...      ...   \n",
       "Ca006ONL2             3.0       34   \n",
       "Ca006ONNL             3.0       34   \n",
       "Ca006ONNL2            3.0       34   \n",
       "Ca006ONPN             3.0       34   \n",
       "Ca006ONPN2            3.0       34   \n",
       "\n",
       "                                                 FWD_filepath  \\\n",
       "#sample-id                                                      \n",
       "Ca009STL    /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "900221      /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "Ca010EBL    /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "900460      /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "900051      /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "...                                                       ...   \n",
       "Ca006ONL2   /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "Ca006ONNL   /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "Ca006ONNL2  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "Ca006ONPN   /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "Ca006ONPN2  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "\n",
       "                                                 REV_filepath      group  \n",
       "#sample-id                                                                \n",
       "Ca009STL    /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   skin-ADL  \n",
       "900221      /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   skin-ADL  \n",
       "Ca010EBL    /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   skin-ADL  \n",
       "900460      /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   skin-ADL  \n",
       "900051      /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   skin-ADL  \n",
       "...                                                       ...        ...  \n",
       "Ca006ONL2   /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   skin-ADL  \n",
       "Ca006ONNL   /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...  skin-ADNL  \n",
       "Ca006ONNL2  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...  skin-ADNL  \n",
       "Ca006ONPN   /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   nares-AD  \n",
       "Ca006ONPN2  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   nares-AD  \n",
       "\n",
       "[502 rows x 32 columns]"
      ]
     },
     "execution_count": 1174,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load the metadata\n",
    "metadata_path = '../Data/Metadata/updated_clean_ant_skin_metadata.tab'\n",
    "metadata = pd.read_csv(metadata_path, sep='\\t')\n",
    "\n",
    "metadata['#sample-id'] = metadata['#sample-id'].str.replace('_', '')\n",
    "# Set Sample-ID as the index for the metadata dataframe \n",
    "metadata = metadata.set_index('#sample-id')\n",
    "\n",
    "\n",
    "# Create group column based on case_type to simplify group names\n",
    "metadata['group'] = metadata['case_type'].map({\n",
    "    'case-lesional skin': 'skin-ADL',\n",
    "    'case-nonlesional skin': 'skin-ADNL', \n",
    "    'control-nonlesional skin': 'skin-H',\n",
    "    'case-anterior nares': 'nares-AD',\n",
    "    'control-anterior nares': 'nares-H'\n",
    "})\n",
    "\n",
    "metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1175,
   "metadata": {},
   "outputs": [],
   "source": [
    "def rclr_transform(df, pseudocount=1e-5):\n",
    "    \"\"\"\n",
    "    Applies Robust Centered Log-Ratio (RCLR) transformation to a DataFrame.\n",
    "    Zeros are ignored in the geometric mean calculation per sample.\n",
    "    \"\"\"\n",
    "    # Convert sparse to dense if needed\n",
    "    df = df.sparse.to_dense()\n",
    "\n",
    "    # Replace 0 with np.nan to ignore in log and mean\n",
    "    df_masked = df.replace(0, np.nan)\n",
    "\n",
    "    # Apply log (with pseudocount only where needed)\n",
    "    log_df = np.log(df_masked + pseudocount)\n",
    "\n",
    "    # Subtract mean of each row (feature)\n",
    "    rclr_df = log_df.sub(log_df.mean(axis=1, skipna=True), axis=0)\n",
    "\n",
    "    return rclr_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1176,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_rclr_transformed_by_region(biom_path, metadata, regions, skin_groups=None):\n",
    "    \"\"\"\n",
    "    Reads a BIOM table, subsets samples by region and group, applies rclr transformation,\n",
    "    and appends group metadata to each resulting DataFrame.\n",
    "\n",
    "    Parameters:\n",
    "        biom_path (str): Path to the BIOM table.\n",
    "        metadata (pd.DataFrame): Metadata with 'group' and 'area' columns.\n",
    "        regions (list): List of region names (e.g., ['Cape Town', 'Umtata']).\n",
    "        skin_groups (list, optional): List of skin sample group labels to include (e.g., ['skin-ADL', 'skin-ADNL', 'skin-H']).\n",
    "\n",
    "    Returns:\n",
    "        dict: Dictionary of rclr-transformed DataFrames with group labels, keyed by region.\n",
    "    \"\"\"\n",
    "    rclr_dfs = {}\n",
    "\n",
    "    for location in regions:\n",
    "        # Load and format the BIOM table\n",
    "        biom_tbl = load_table(biom_path)\n",
    "        df = pd.DataFrame(biom_tbl.to_dataframe().T)\n",
    "        df.index = df.index.str.replace('15564.', '', regex=False)\n",
    "\n",
    "        # Subset metadata to relevant skin samples in the specified region\n",
    "        skin_samples = metadata[\n",
    "            (metadata['group'].str.startswith('skin')) &\n",
    "            (metadata['area'].str.startswith(location))\n",
    "        ].index\n",
    "\n",
    "        # Filter dataframe to include only relevant samples\n",
    "        df = df.loc[df.index.intersection(skin_samples)]\n",
    "\n",
    "        # Apply rclr transformation\n",
    "        rclr_df = rclr_transform(df)\n",
    "\n",
    "        # Append group information\n",
    "        rclr_df['group'] = metadata.loc[rclr_df.index, 'group']\n",
    "\n",
    "        # Optional: filter for specific skin groups\n",
    "        if skin_groups is not None:\n",
    "            rclr_df = rclr_df[rclr_df['group'].isin(skin_groups)]\n",
    "\n",
    "        # Append o_scorad information\n",
    "        rclr_df['o_scorad'] = metadata.loc[rclr_df.index, 'o_scorad']    \n",
    "\n",
    "        # Save to dictionary\n",
    "        rclr_dfs[location] = rclr_df\n",
    "\n",
    "    return rclr_dfs\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1177,
   "metadata": {},
   "outputs": [],
   "source": [
    "def log_transform(df, pseudocount=1e-6):\n",
    "    \"\"\"\n",
    "    Applies log transformation to relative abundance data.\n",
    "    Converts counts to relative abundance per sample (row), adds pseudocount,\n",
    "    and takes log.\n",
    "    \"\"\"\n",
    "    # Convert sparse to dense if needed\n",
    "    df = df.sparse.to_dense() if hasattr(df, 'sparse') else df\n",
    "\n",
    "    # Convert to relative abundances per sample (row-wise)\n",
    "    rel_abundance_df = df.div(df.sum(axis=1), axis=0)\n",
    "\n",
    "    # Apply log transformation with pseudocount\n",
    "    log_dfs = np.log(rel_abundance_df + pseudocount)\n",
    "\n",
    "    return log_dfs\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1178,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_log_transformed_by_region(biom_path, metadata, regions, skin_groups=None):\n",
    "    \"\"\"\n",
    "    Reads a BIOM table, subsets samples by region and group, applies log transformation\n",
    "    to relative abundance data, and appends group and SCORAD metadata.\n",
    "\n",
    "    Parameters:\n",
    "        biom_path (str): Path to the BIOM table.\n",
    "        metadata (pd.DataFrame): Metadata with 'group' and 'area' columns.\n",
    "        regions (list): List of region names (e.g., ['Cape Town', 'Umtata']).\n",
    "        skin_groups (list, optional): List of skin sample group labels to include.\n",
    "\n",
    "    Returns:\n",
    "        dict: Dictionary of log-transformed DataFrames with group and SCORAD, keyed by region.\n",
    "    \"\"\"\n",
    "    log_dfs = {}\n",
    "\n",
    "    for location in regions:\n",
    "        # Load and format the BIOM table\n",
    "        biom_tbl = load_table(biom_path)\n",
    "        df = pd.DataFrame(biom_tbl.to_dataframe().T)\n",
    "        df.index = df.index.str.replace('15564.', '', regex=False)\n",
    "\n",
    "        # Subset metadata to relevant skin samples in the specified region\n",
    "        skin_samples = metadata[\n",
    "            (metadata['group'].str.startswith('skin')) &\n",
    "            (metadata['area'].str.startswith(location))\n",
    "        ].index\n",
    "\n",
    "        # Filter dataframe to include only relevant samples\n",
    "        df = df.loc[df.index.intersection(skin_samples)]\n",
    "\n",
    "        # Apply log transformation\n",
    "        log_df = log_transform(df)\n",
    "\n",
    "        # Append group information\n",
    "        log_df['group'] = metadata.loc[log_df.index, 'group']\n",
    "\n",
    "        # Optional: filter for specific skin groups\n",
    "        if skin_groups is not None:\n",
    "            log_df = log_df[log_df['group'].isin(skin_groups)]\n",
    "\n",
    "        # Append SCORAD information\n",
    "        log_df['o_scorad'] = metadata.loc[log_df.index, 'o_scorad']\n",
    "\n",
    "        # Save to dictionary\n",
    "        log_dfs[location] = log_df\n",
    "\n",
    "    return log_dfs\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1179,
   "metadata": {},
   "outputs": [],
   "source": [
    "def pairwise_mannwhitney_bh(df, feature, group_col='group'):\n",
    "    groups = df[group_col].unique()\n",
    "    comparisons = list(itertools.combinations(groups, 2))\n",
    "\n",
    "    raw_pvals = []\n",
    "    labels = []\n",
    "    pairs = []\n",
    "\n",
    "    for g1, g2 in comparisons:\n",
    "        x = df[df[group_col] == g1][feature].dropna()\n",
    "        y = df[df[group_col] == g2][feature].dropna()\n",
    "\n",
    "        # Skip comparison if either group has no data\n",
    "        if len(x) == 0 or len(y) == 0:\n",
    "            continue\n",
    "\n",
    "        _, p = mannwhitneyu(x, y, alternative='two-sided')\n",
    "        raw_pvals.append(p)\n",
    "        labels.append(f\"{g1} vs {g2}\")\n",
    "        pairs.append((g1, g2))\n",
    "\n",
    "    if len(raw_pvals) == 0:\n",
    "        return pd.DataFrame(columns=['Comparison', 'Pair', 'Raw p-value', 'BH-corrected p-value'])\n",
    "\n",
    "    _, pvals_corrected, _, _ = multipletests(raw_pvals, method='fdr_bh')\n",
    "\n",
    "    results_df = pd.DataFrame({\n",
    "        'Comparison': labels,\n",
    "        'Pair': pairs,\n",
    "        'Raw p-value': raw_pvals,\n",
    "        'BH-corrected p-value': pvals_corrected\n",
    "    })\n",
    "\n",
    "    return results_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1180,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_combined_box_and_severity(region, rclr_df, log_df, taxa_list, group_palette, strip_palette, title_name, order=['skin-H', 'skin-ADNL', 'skin-ADL']):\n",
    "\n",
    "    n_taxa = len(taxa_list)\n",
    "    fig, axes = plt.subplots(\n",
    "        nrows=n_taxa,\n",
    "        ncols=2,\n",
    "        figsize=(6, 3 * n_taxa),\n",
    "        sharey=False,\n",
    "        gridspec_kw={'width_ratios': [2, 1.5]}\n",
    "    )\n",
    "\n",
    "    if n_taxa == 1:\n",
    "        axes = [axes]\n",
    "\n",
    "    if region == 'Cape Town':\n",
    "        title_y_positions = [0.94, 0.75, 0.56, 0.37, 0.18]\n",
    "    elif region == 'Umtata':    \n",
    "        title_y_positions = [0.94, 0.75, 0.56, 0.37, 0.18]\n",
    "\n",
    "    for i, taxon in enumerate(taxa_list):\n",
    "        ax_box = axes[i][0] if n_taxa > 1 else axes[0]\n",
    "        ax_corr = axes[i][1] if n_taxa > 1 else axes[1]\n",
    "\n",
    "        # Boxplot data (RCLR)\n",
    "        df_box = rclr_df[['group', taxon]].copy()\n",
    "        df_box = df_box[df_box['group'].isin(order)]\n",
    "        df_box['group'] = pd.Categorical(df_box['group'], categories=order, ordered=True)\n",
    "        df_box = df_box.dropna(subset=[taxon])\n",
    "        df_box['taxon'] = taxon.replace('g__', '').replace('_ASV-', ' ASV-')\n",
    "\n",
    "        # Correlation data (Rank SCORAD vs. Rank Abundance)\n",
    "        df_corr = log_df[log_df['group'].isin(['skin-ADNL', 'skin-ADL'])][['o_scorad', taxon]].copy()\n",
    "        df_corr['o_scorad'] = pd.to_numeric(df_corr['o_scorad'], errors='coerce')\n",
    "        df_corr[taxon] = pd.to_numeric(df_corr[taxon], errors='coerce')\n",
    "        df_corr = df_corr.dropna()\n",
    "        df_corr['rank_scorad'] = df_corr['o_scorad'].rank()\n",
    "        df_corr['rank_abundance'] = df_corr[taxon].rank()\n",
    "        df_corr = df_corr[(df_corr['rank_abundance'] >= 25)]\n",
    "        df_corr['taxon'] = df_box['taxon'].iloc[0]\n",
    "\n",
    "        # Set custom palette based on region and taxon\n",
    "        if region == 'Cape Town':\n",
    "            if df_box['taxon'].iloc[0] == 'Streptococcus ASV-1':\n",
    "                custom_palette = {'skin-H': '#ff5050', 'skin-ADNL': '#ff5050', 'skin-ADL': '#ff5050'}\n",
    "            else:\n",
    "                custom_palette = {'skin-H': 'gray', 'skin-ADNL': 'gray', 'skin-ADL': 'gray'}\n",
    "\n",
    "        elif region == 'Umtata':\n",
    "            if df_box['taxon'].iloc[0] == 'Micrococcus ASV-1':\n",
    "                custom_palette = {'skin-H': '#66CCFF', 'skin-ADNL': '#66CCFF', 'skin-ADL': '#66CCFF'}  # light blue\n",
    "            else:\n",
    "                custom_palette = {'skin-H': '#ff5050', 'skin-ADNL': '#ff5050', 'skin-ADL': '#ff5050'}\n",
    "        \n",
    "        # Boxplot (RCLR)\n",
    "        sns.boxplot(\n",
    "            data=df_box,\n",
    "            x='group',\n",
    "            y=taxon,\n",
    "            order=order,\n",
    "            palette=custom_palette,\n",
    "            ax=ax_box,\n",
    "            width=0.5,\n",
    "            fliersize=0\n",
    "        )\n",
    "\n",
    "\n",
    "        sns.stripplot(\n",
    "            data=df_box,\n",
    "            x='group',\n",
    "            y=taxon,\n",
    "            order=order,\n",
    "            palette=strip_palette,\n",
    "            ax=ax_box,\n",
    "            jitter=True,\n",
    "            size=4,\n",
    "            alpha=0.7\n",
    "        )\n",
    "\n",
    "        ax_box.set_ylabel(\"RCLR Rel. Abundance\", fontsize=13)\n",
    "        ax_box.set_xlabel(\"\")\n",
    "        ax_box.set_xticklabels(['H', 'ADNL', 'ADL'], fontsize=13)\n",
    "        ax_box.tick_params(axis='y', labelsize=10)\n",
    "\n",
    "        # Optional stats\n",
    "        stats = pairwise_mannwhitney_bh(rclr_df, taxon)\n",
    "        stats_sorted = stats.sort_values(by='BH-corrected p-value')\n",
    "        y_base = df_box[taxon].max() + 0.5\n",
    "        gap = 1.5\n",
    "\n",
    "        y_positions = []\n",
    "        for _, row in stats_sorted.iterrows():\n",
    "            y = y_base\n",
    "            while any(abs(y - existing_y) < gap for existing_y in y_positions):\n",
    "                y += gap\n",
    "            y_positions.append(y)\n",
    "\n",
    "        if y_positions:\n",
    "            ax_box.set_ylim(top=max(y_positions) + gap * 0.7)\n",
    "\n",
    "        for (y, (_, row)) in zip(y_positions, stats_sorted.iterrows()):\n",
    "            g1, g2 = row['Pair']\n",
    "            pval = row['BH-corrected p-value']\n",
    "            stars = '***' if pval < 0.001 else '**' if pval < 0.01 else '*' if pval < 0.05 else ''\n",
    "\n",
    "            xpos1 = order.index(g1)\n",
    "            xpos2 = order.index(g2)\n",
    "            x_center = (xpos1 + xpos2) / 2\n",
    "\n",
    "            ax_box.plot([xpos1, xpos1, xpos2, xpos2],\n",
    "                        [y, y - 0.05, y - 0.05, y],\n",
    "                        lw=1, color='black')\n",
    "\n",
    "            ax_box.text(x_center, y + 0.05, f\"{pval:.1e} {stars}\",\n",
    "                        ha='center', va='bottom', fontsize=10)\n",
    "\n",
    "        # Correlation plot (rank_scorad vs rank_abundance)\n",
    "        ax_corr.set_visible(True)\n",
    "        if not df_corr.empty:\n",
    "            r, pval = pearsonr(df_corr['rank_scorad'], df_corr['rank_abundance'])\n",
    "            dot_color = 'grey' if pval > 0.05 else '#ff5050'\n",
    "\n",
    "            sns.regplot(\n",
    "                data=df_corr,\n",
    "                x='rank_scorad',\n",
    "                y='rank_abundance',\n",
    "                scatter_kws={'alpha': 0.5, 's': 20, 'color': dot_color},\n",
    "                line_kws={'color': 'black'},\n",
    "                ax=ax_corr\n",
    "            )\n",
    "\n",
    "            r_label = f\"Spearman ρ = {r:.2f}\\np = {pval:.1e}\"\n",
    "            ax_corr.text(0.05, 0.90, r_label, transform=ax_corr.transAxes,\n",
    "                         fontsize=12, va='top', ha='left',\n",
    "                         bbox=dict(facecolor='white', alpha=0.7, edgecolor='none'))\n",
    "\n",
    "        ax_corr.set_xlabel(\"SCORAD (ranked)\", fontsize=13)\n",
    "        ax_corr.set_ylabel(\"Rank Rel. Abundance\", fontsize=13)\n",
    "        ax_corr.tick_params(axis='x', labelsize=10)\n",
    "        ax_corr.tick_params(axis='y', labelsize=10)\n",
    "\n",
    "        if i < len(title_y_positions):\n",
    "            fig.text(0.5, title_y_positions[i], df_box['taxon'].iloc[0], ha='center', fontsize=16)\n",
    "\n",
    "    fig.suptitle(title_name, fontsize=18, y=0.98)\n",
    "    plt.tight_layout(rect=[0, 0, 1, 0.96])\n",
    "    plt.subplots_adjust(hspace=0.5, wspace=0.5)\n",
    "    sns.despine()\n",
    "\n",
    "    return fig\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1181,
   "metadata": {},
   "outputs": [],
   "source": [
    "biom_path = '../Data/Tables/Absolute_Abundance_Tables/209766_filtered_by_prevalence_1pct_rare_Genus-ASV-non-collapse.biom'\n",
    "regions = ['Cape Town', 'Umtata']\n",
    "skin_groups = ['skin-H', 'skin-ADNL', 'skin-ADL']\n",
    "\n",
    "# Run the function\n",
    "rclr_dfs = get_rclr_transformed_by_region(\n",
    "    biom_path=biom_path,\n",
    "    metadata=metadata,\n",
    "    regions=regions,\n",
    "    skin_groups=skin_groups\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1182,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define custom color palette\n",
    "group_palette = {\n",
    "    'skin-H': '#ADD8E6',     # baby blue\n",
    "    'skin-ADNL': '#FFDAB9',  # peach\n",
    "    'skin-ADL': '#E31A1C'    # red\n",
    "}\n",
    "\n",
    "# strip_palette = {\n",
    "#     'skin-H': '#6CA6CD',     # darker baby blue\n",
    "#     'skin-ADNL': '#E6AC8F',  # darker peach\n",
    "#     'skin-ADL': '#A50000'    # darker red\n",
    "# }\n",
    "\n",
    "strip_palette = {\n",
    "    'skin-H': '#3b3b3b',     \n",
    "    'skin-ADNL': '#3b3b3b',  \n",
    "    'skin-ADL': '#3b3b3b'    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1183,
   "metadata": {},
   "outputs": [],
   "source": [
    "log_dfs = get_log_transformed_by_region(\n",
    "    biom_path=biom_path,\n",
    "    metadata=metadata,\n",
    "    regions=['Cape Town', 'Umtata'],  # Or just ['Cape Town'] if that's all you need\n",
    "    skin_groups=['skin-H', 'skin-ADNL', 'skin-ADL']\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1184,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>g__Streptococcus_ASV-1</th>\n",
       "      <th>g__Streptococcus_ASV-2</th>\n",
       "      <th>g__Corynebacterium_ASV-1</th>\n",
       "      <th>g__Corynebacterium_ASV-3</th>\n",
       "      <th>g___ASV-3</th>\n",
       "      <th>g__Cutibacterium_ASV-1</th>\n",
       "      <th>g___ASV-16</th>\n",
       "      <th>g___ASV-25</th>\n",
       "      <th>g___ASV-27</th>\n",
       "      <th>g__Cutibacterium_ASV-2</th>\n",
       "      <th>...</th>\n",
       "      <th>g__Leptotrichia_A_993758_ASV-15</th>\n",
       "      <th>g___ASV-126</th>\n",
       "      <th>g__Leptotrichia_A_993758_ASV-16</th>\n",
       "      <th>g__Capnocytophaga_820688_ASV-7</th>\n",
       "      <th>g__UBA6175_ASV-2</th>\n",
       "      <th>g__Comamonas_F_589250_ASV-4</th>\n",
       "      <th>g__Bosea_ASV-2</th>\n",
       "      <th>g___ASV-119</th>\n",
       "      <th>group</th>\n",
       "      <th>o_scorad</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>900570</th>\n",
       "      <td>-2.796587</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>...</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>skin-ADNL</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900129</th>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-3.489718</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-5.686680</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>...</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900091</th>\n",
       "      <td>-5.686680</td>\n",
       "      <td>-4.993681</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-4.993681</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>...</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>skin-ADNL</td>\n",
       "      <td>53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900581</th>\n",
       "      <td>-3.122003</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-4.993681</td>\n",
       "      <td>-4.300607</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-4.300607</td>\n",
       "      <td>...</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900145</th>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>...</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>skin-H</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900081</th>\n",
       "      <td>-1.330263</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-4.993681</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-4.077478</td>\n",
       "      <td>-4.588265</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>...</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900093</th>\n",
       "      <td>-3.122003</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>...</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-4.588265</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900094</th>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-4.300607</td>\n",
       "      <td>-4.993681</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-5.686680</td>\n",
       "      <td>...</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>skin-ADNL</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900057</th>\n",
       "      <td>-2.272116</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-5.673032</td>\n",
       "      <td>-4.286956</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-4.980031</td>\n",
       "      <td>...</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900097</th>\n",
       "      <td>-4.588265</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-4.588265</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>...</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>-13.815511</td>\n",
       "      <td>skin-ADNL</td>\n",
       "      <td>44</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>131 rows × 800 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        g__Streptococcus_ASV-1  g__Streptococcus_ASV-2  \\\n",
       "900570               -2.796587              -13.815511   \n",
       "900129              -13.815511              -13.815511   \n",
       "900091               -5.686680               -4.993681   \n",
       "900581               -3.122003              -13.815511   \n",
       "900145              -13.815511              -13.815511   \n",
       "...                        ...                     ...   \n",
       "900081               -1.330263              -13.815511   \n",
       "900093               -3.122003              -13.815511   \n",
       "900094              -13.815511              -13.815511   \n",
       "900057               -2.272116              -13.815511   \n",
       "900097               -4.588265              -13.815511   \n",
       "\n",
       "        g__Corynebacterium_ASV-1  g__Corynebacterium_ASV-3  g___ASV-3  \\\n",
       "900570                -13.815511                -13.815511 -13.815511   \n",
       "900129                -13.815511                -13.815511 -13.815511   \n",
       "900091                -13.815511                -13.815511 -13.815511   \n",
       "900581                -13.815511                -13.815511  -4.993681   \n",
       "900145                -13.815511                -13.815511 -13.815511   \n",
       "...                          ...                       ...        ...   \n",
       "900081                 -4.993681                -13.815511  -4.077478   \n",
       "900093                -13.815511                -13.815511 -13.815511   \n",
       "900094                -13.815511                -13.815511  -4.300607   \n",
       "900057                -13.815511                -13.815511 -13.815511   \n",
       "900097                -13.815511                -13.815511 -13.815511   \n",
       "\n",
       "        g__Cutibacterium_ASV-1  g___ASV-16  g___ASV-25  g___ASV-27  \\\n",
       "900570              -13.815511  -13.815511  -13.815511  -13.815511   \n",
       "900129               -3.489718  -13.815511  -13.815511   -5.686680   \n",
       "900091              -13.815511  -13.815511  -13.815511   -4.993681   \n",
       "900581               -4.300607  -13.815511  -13.815511  -13.815511   \n",
       "900145              -13.815511  -13.815511  -13.815511  -13.815511   \n",
       "...                        ...         ...         ...         ...   \n",
       "900081               -4.588265  -13.815511  -13.815511  -13.815511   \n",
       "900093              -13.815511  -13.815511  -13.815511  -13.815511   \n",
       "900094               -4.993681  -13.815511  -13.815511  -13.815511   \n",
       "900057               -5.673032   -4.286956  -13.815511  -13.815511   \n",
       "900097               -4.588265  -13.815511  -13.815511  -13.815511   \n",
       "\n",
       "        g__Cutibacterium_ASV-2  ...  g__Leptotrichia_A_993758_ASV-15  \\\n",
       "900570              -13.815511  ...                       -13.815511   \n",
       "900129              -13.815511  ...                       -13.815511   \n",
       "900091              -13.815511  ...                       -13.815511   \n",
       "900581               -4.300607  ...                       -13.815511   \n",
       "900145              -13.815511  ...                       -13.815511   \n",
       "...                        ...  ...                              ...   \n",
       "900081              -13.815511  ...                       -13.815511   \n",
       "900093              -13.815511  ...                       -13.815511   \n",
       "900094               -5.686680  ...                       -13.815511   \n",
       "900057               -4.980031  ...                       -13.815511   \n",
       "900097              -13.815511  ...                       -13.815511   \n",
       "\n",
       "        g___ASV-126  g__Leptotrichia_A_993758_ASV-16  \\\n",
       "900570   -13.815511                       -13.815511   \n",
       "900129   -13.815511                       -13.815511   \n",
       "900091   -13.815511                       -13.815511   \n",
       "900581   -13.815511                       -13.815511   \n",
       "900145   -13.815511                       -13.815511   \n",
       "...             ...                              ...   \n",
       "900081   -13.815511                       -13.815511   \n",
       "900093   -13.815511                       -13.815511   \n",
       "900094   -13.815511                       -13.815511   \n",
       "900057   -13.815511                       -13.815511   \n",
       "900097   -13.815511                       -13.815511   \n",
       "\n",
       "        g__Capnocytophaga_820688_ASV-7  g__UBA6175_ASV-2  \\\n",
       "900570                      -13.815511        -13.815511   \n",
       "900129                      -13.815511        -13.815511   \n",
       "900091                      -13.815511        -13.815511   \n",
       "900581                      -13.815511        -13.815511   \n",
       "900145                      -13.815511        -13.815511   \n",
       "...                                ...               ...   \n",
       "900081                      -13.815511        -13.815511   \n",
       "900093                      -13.815511         -4.588265   \n",
       "900094                      -13.815511        -13.815511   \n",
       "900057                      -13.815511        -13.815511   \n",
       "900097                      -13.815511        -13.815511   \n",
       "\n",
       "        g__Comamonas_F_589250_ASV-4  g__Bosea_ASV-2  g___ASV-119      group  \\\n",
       "900570                   -13.815511      -13.815511   -13.815511  skin-ADNL   \n",
       "900129                   -13.815511      -13.815511   -13.815511   skin-ADL   \n",
       "900091                   -13.815511      -13.815511   -13.815511  skin-ADNL   \n",
       "900581                   -13.815511      -13.815511   -13.815511   skin-ADL   \n",
       "900145                   -13.815511      -13.815511   -13.815511     skin-H   \n",
       "...                             ...             ...          ...        ...   \n",
       "900081                   -13.815511      -13.815511   -13.815511   skin-ADL   \n",
       "900093                   -13.815511      -13.815511   -13.815511   skin-ADL   \n",
       "900094                   -13.815511      -13.815511   -13.815511  skin-ADNL   \n",
       "900057                   -13.815511      -13.815511   -13.815511   skin-ADL   \n",
       "900097                   -13.815511      -13.815511   -13.815511  skin-ADNL   \n",
       "\n",
       "        o_scorad  \n",
       "900570        36  \n",
       "900129        44  \n",
       "900091        53  \n",
       "900581        67  \n",
       "900145       NaN  \n",
       "...          ...  \n",
       "900081        44  \n",
       "900093        36  \n",
       "900094        36  \n",
       "900057        33  \n",
       "900097        44  \n",
       "\n",
       "[131 rows x 800 columns]"
      ]
     },
     "execution_count": 1184,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "log_dfs['Cape Town']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1185,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/categorical.py:641: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  grouped_vals = vals.groupby(grouper)\n",
      "/var/folders/22/yck9vwx53w1c38tvj_c0_tz00000gn/T/ipykernel_3917/3892701226.py:67: FutureWarning: Passing `palette` without assigning `hue` is deprecated.\n",
      "  sns.stripplot(\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.\n",
      "  with pd.option_context('mode.use_inf_as_na', True):\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.\n",
      "  with pd.option_context('mode.use_inf_as_na', True):\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1075: FutureWarning: When grouping with a length-1 list-like, you will need to pass a length-1 tuple to get_group in a future version of pandas. Pass `(name,)` instead of `name` to silence this warning.\n",
      "  data_subset = grouped_data.get_group(pd_key)\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/categorical.py:641: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  grouped_vals = vals.groupby(grouper)\n",
      "/var/folders/22/yck9vwx53w1c38tvj_c0_tz00000gn/T/ipykernel_3917/3892701226.py:67: FutureWarning: Passing `palette` without assigning `hue` is deprecated.\n",
      "  sns.stripplot(\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.\n",
      "  with pd.option_context('mode.use_inf_as_na', True):\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.\n",
      "  with pd.option_context('mode.use_inf_as_na', True):\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1075: FutureWarning: When grouping with a length-1 list-like, you will need to pass a length-1 tuple to get_group in a future version of pandas. Pass `(name,)` instead of `name` to silence this warning.\n",
      "  data_subset = grouped_data.get_group(pd_key)\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/categorical.py:641: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  grouped_vals = vals.groupby(grouper)\n",
      "/var/folders/22/yck9vwx53w1c38tvj_c0_tz00000gn/T/ipykernel_3917/3892701226.py:67: FutureWarning: Passing `palette` without assigning `hue` is deprecated.\n",
      "  sns.stripplot(\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.\n",
      "  with pd.option_context('mode.use_inf_as_na', True):\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.\n",
      "  with pd.option_context('mode.use_inf_as_na', True):\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1075: FutureWarning: When grouping with a length-1 list-like, you will need to pass a length-1 tuple to get_group in a future version of pandas. Pass `(name,)` instead of `name` to silence this warning.\n",
      "  data_subset = grouped_data.get_group(pd_key)\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/categorical.py:641: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  grouped_vals = vals.groupby(grouper)\n",
      "/var/folders/22/yck9vwx53w1c38tvj_c0_tz00000gn/T/ipykernel_3917/3892701226.py:67: FutureWarning: Passing `palette` without assigning `hue` is deprecated.\n",
      "  sns.stripplot(\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.\n",
      "  with pd.option_context('mode.use_inf_as_na', True):\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.\n",
      "  with pd.option_context('mode.use_inf_as_na', True):\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1075: FutureWarning: When grouping with a length-1 list-like, you will need to pass a length-1 tuple to get_group in a future version of pandas. Pass `(name,)` instead of `name` to silence this warning.\n",
      "  data_subset = grouped_data.get_group(pd_key)\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/categorical.py:641: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  grouped_vals = vals.groupby(grouper)\n",
      "/var/folders/22/yck9vwx53w1c38tvj_c0_tz00000gn/T/ipykernel_3917/3892701226.py:67: FutureWarning: Passing `palette` without assigning `hue` is deprecated.\n",
      "  sns.stripplot(\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.\n",
      "  with pd.option_context('mode.use_inf_as_na', True):\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.\n",
      "  with pd.option_context('mode.use_inf_as_na', True):\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1075: FutureWarning: When grouping with a length-1 list-like, you will need to pass a length-1 tuple to get_group in a future version of pandas. Pass `(name,)` instead of `name` to silence this warning.\n",
      "  data_subset = grouped_data.get_group(pd_key)\n"
     ]
    }
   ],
   "source": [
    "fig = plot_combined_box_and_severity(\n",
    "    region='Cape Town',\n",
    "    rclr_df=rclr_dfs['Cape Town'],\n",
    "    log_df=log_dfs['Cape Town'],\n",
    "    taxa_list=[\n",
    "        'g__Streptococcus_ASV-1',\n",
    "        'g__Streptococcus_ASV-2',\n",
    "        'g__Staphylococcus_ASV-1',\n",
    "        'g__Micrococcus_ASV-1',\n",
    "        'g__Veillonella_A_ASV-1'\n",
    "    ],\n",
    "    group_palette=group_palette,\n",
    "    strip_palette=strip_palette,\n",
    "    title_name='Differential Taxa in Cape Town Children'\n",
    ")\n",
    "fig.savefig('../Figures/Main/Fig_3A.png', dpi=1000)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1186,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/categorical.py:641: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  grouped_vals = vals.groupby(grouper)\n",
      "/var/folders/22/yck9vwx53w1c38tvj_c0_tz00000gn/T/ipykernel_3917/3892701226.py:67: FutureWarning: Passing `palette` without assigning `hue` is deprecated.\n",
      "  sns.stripplot(\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.\n",
      "  with pd.option_context('mode.use_inf_as_na', True):\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.\n",
      "  with pd.option_context('mode.use_inf_as_na', True):\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1075: FutureWarning: When grouping with a length-1 list-like, you will need to pass a length-1 tuple to get_group in a future version of pandas. Pass `(name,)` instead of `name` to silence this warning.\n",
      "  data_subset = grouped_data.get_group(pd_key)\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/categorical.py:641: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  grouped_vals = vals.groupby(grouper)\n",
      "/var/folders/22/yck9vwx53w1c38tvj_c0_tz00000gn/T/ipykernel_3917/3892701226.py:67: FutureWarning: Passing `palette` without assigning `hue` is deprecated.\n",
      "  sns.stripplot(\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.\n",
      "  with pd.option_context('mode.use_inf_as_na', True):\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.\n",
      "  with pd.option_context('mode.use_inf_as_na', True):\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1075: FutureWarning: When grouping with a length-1 list-like, you will need to pass a length-1 tuple to get_group in a future version of pandas. Pass `(name,)` instead of `name` to silence this warning.\n",
      "  data_subset = grouped_data.get_group(pd_key)\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/categorical.py:641: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  grouped_vals = vals.groupby(grouper)\n",
      "/var/folders/22/yck9vwx53w1c38tvj_c0_tz00000gn/T/ipykernel_3917/3892701226.py:67: FutureWarning: Passing `palette` without assigning `hue` is deprecated.\n",
      "  sns.stripplot(\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.\n",
      "  with pd.option_context('mode.use_inf_as_na', True):\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.\n",
      "  with pd.option_context('mode.use_inf_as_na', True):\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1075: FutureWarning: When grouping with a length-1 list-like, you will need to pass a length-1 tuple to get_group in a future version of pandas. Pass `(name,)` instead of `name` to silence this warning.\n",
      "  data_subset = grouped_data.get_group(pd_key)\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/categorical.py:641: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  grouped_vals = vals.groupby(grouper)\n",
      "/var/folders/22/yck9vwx53w1c38tvj_c0_tz00000gn/T/ipykernel_3917/3892701226.py:67: FutureWarning: Passing `palette` without assigning `hue` is deprecated.\n",
      "  sns.stripplot(\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.\n",
      "  with pd.option_context('mode.use_inf_as_na', True):\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.\n",
      "  with pd.option_context('mode.use_inf_as_na', True):\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1075: FutureWarning: When grouping with a length-1 list-like, you will need to pass a length-1 tuple to get_group in a future version of pandas. Pass `(name,)` instead of `name` to silence this warning.\n",
      "  data_subset = grouped_data.get_group(pd_key)\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/categorical.py:641: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  grouped_vals = vals.groupby(grouper)\n",
      "/var/folders/22/yck9vwx53w1c38tvj_c0_tz00000gn/T/ipykernel_3917/3892701226.py:67: FutureWarning: Passing `palette` without assigning `hue` is deprecated.\n",
      "  sns.stripplot(\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.\n",
      "  with pd.option_context('mode.use_inf_as_na', True):\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.\n",
      "  with pd.option_context('mode.use_inf_as_na', True):\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1075: FutureWarning: When grouping with a length-1 list-like, you will need to pass a length-1 tuple to get_group in a future version of pandas. Pass `(name,)` instead of `name` to silence this warning.\n",
      "  data_subset = grouped_data.get_group(pd_key)\n"
     ]
    }
   ],
   "source": [
    "fig = plot_combined_box_and_severity(\n",
    "    region='Umtata',\n",
    "    rclr_df=rclr_dfs['Umtata'],\n",
    "    log_df=log_dfs['Umtata'],\n",
    "    taxa_list=[\n",
    "        'g__Streptococcus_ASV-1',\n",
    "        'g__Streptococcus_ASV-2',\n",
    "        'g__Staphylococcus_ASV-1',\n",
    "        'g__Micrococcus_ASV-1',\n",
    "        'g__Veillonella_A_ASV-1'\n",
    "    ],\n",
    "    group_palette=group_palette,\n",
    "    strip_palette=strip_palette,\n",
    "    title_name='Differential Taxa in Umtata Children'\n",
    ")\n",
    "fig.savefig('../Figures/Main/Fig_3B.png', dpi=1000)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "qiime2-metagenome-2024.10",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
