{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Relative Abundance ASVs between Skin and Nares by Region"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import biom\n",
    "from biom import load_table\n",
    "import numpy as np\n",
    "import scipy.stats as stats\n",
    "from scipy.stats import mannwhitneyu\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "import itertools\n",
    "import matplotlib.patches\n",
    "from scipy.stats import pearsonr\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PlateNumber</th>\n",
       "      <th>PlateLocation</th>\n",
       "      <th>i5</th>\n",
       "      <th>i5Sequence</th>\n",
       "      <th>i7</th>\n",
       "      <th>i7Sequence</th>\n",
       "      <th>identifier</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Plate ID</th>\n",
       "      <th>Well location</th>\n",
       "      <th>...</th>\n",
       "      <th>enrolment_date</th>\n",
       "      <th>enrolment_season</th>\n",
       "      <th>hiv_exposure</th>\n",
       "      <th>hiv_status</th>\n",
       "      <th>household_size</th>\n",
       "      <th>o_scorad</th>\n",
       "      <th>FWD_filepath</th>\n",
       "      <th>REV_filepath</th>\n",
       "      <th>group</th>\n",
       "      <th>microbiome_type</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#sample-id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Ca009STL</th>\n",
       "      <td>1</td>\n",
       "      <td>A1</td>\n",
       "      <td>SA501</td>\n",
       "      <td>ATCGTACG</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA501</td>\n",
       "      <td>CGAGAGTT-ATCGTACG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>A1</td>\n",
       "      <td>...</td>\n",
       "      <td>4/16/2015</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>4.0</td>\n",
       "      <td>40</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900221</th>\n",
       "      <td>1</td>\n",
       "      <td>B1</td>\n",
       "      <td>SA502</td>\n",
       "      <td>ACTATCTG</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA502</td>\n",
       "      <td>CGAGAGTT-ACTATCTG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>B1</td>\n",
       "      <td>...</td>\n",
       "      <td>8/11/2015</td>\n",
       "      <td>Winter</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>7.0</td>\n",
       "      <td>34</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca010EBL</th>\n",
       "      <td>1</td>\n",
       "      <td>C1</td>\n",
       "      <td>SA503</td>\n",
       "      <td>TAGCGAGT</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA503</td>\n",
       "      <td>CGAGAGTT-TAGCGAGT</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>C1</td>\n",
       "      <td>...</td>\n",
       "      <td>11/20/2014</td>\n",
       "      <td>Spring</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>7.0</td>\n",
       "      <td>21</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900460</th>\n",
       "      <td>1</td>\n",
       "      <td>D1</td>\n",
       "      <td>SA504</td>\n",
       "      <td>CTGCGTGT</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA504</td>\n",
       "      <td>CGAGAGTT-CTGCGTGT</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>D1</td>\n",
       "      <td>...</td>\n",
       "      <td>9/23/2015</td>\n",
       "      <td>Spring</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>40</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900051</th>\n",
       "      <td>1</td>\n",
       "      <td>E1</td>\n",
       "      <td>SA505</td>\n",
       "      <td>TCATCGAG</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA505</td>\n",
       "      <td>CGAGAGTT-TCATCGAG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>E1</td>\n",
       "      <td>...</td>\n",
       "      <td>4/21/2015</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>7.0</td>\n",
       "      <td>41</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca006ONL2</th>\n",
       "      <td>6</td>\n",
       "      <td>H1</td>\n",
       "      <td>SA508</td>\n",
       "      <td>GACACCGT</td>\n",
       "      <td>SB701</td>\n",
       "      <td>CTCGACTT</td>\n",
       "      <td>SB701SA508</td>\n",
       "      <td>CTCGACTT-GACACCGT</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>H1</td>\n",
       "      <td>...</td>\n",
       "      <td>3/25/2015</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>3.0</td>\n",
       "      <td>34</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca006ONNL</th>\n",
       "      <td>6</td>\n",
       "      <td>F2</td>\n",
       "      <td>SA506</td>\n",
       "      <td>CGTGAGTG</td>\n",
       "      <td>SB702</td>\n",
       "      <td>CGAAGTAT</td>\n",
       "      <td>SB702SA506</td>\n",
       "      <td>CGAAGTAT-CGTGAGTG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>F2</td>\n",
       "      <td>...</td>\n",
       "      <td>3/25/2015</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>3.0</td>\n",
       "      <td>34</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADNL</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca006ONNL2</th>\n",
       "      <td>6</td>\n",
       "      <td>H2</td>\n",
       "      <td>SA508</td>\n",
       "      <td>GACACCGT</td>\n",
       "      <td>SB702</td>\n",
       "      <td>CGAAGTAT</td>\n",
       "      <td>SB702SA508</td>\n",
       "      <td>CGAAGTAT-GACACCGT</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>H2</td>\n",
       "      <td>...</td>\n",
       "      <td>3/25/2015</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>3.0</td>\n",
       "      <td>34</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADNL</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca006ONPN</th>\n",
       "      <td>6</td>\n",
       "      <td>F3</td>\n",
       "      <td>SA506</td>\n",
       "      <td>CGTGAGTG</td>\n",
       "      <td>SB703</td>\n",
       "      <td>TAGCAGCT</td>\n",
       "      <td>SB703SA506</td>\n",
       "      <td>TAGCAGCT-CGTGAGTG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>F3</td>\n",
       "      <td>...</td>\n",
       "      <td>3/25/2015</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>3.0</td>\n",
       "      <td>34</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>nares-AD</td>\n",
       "      <td>nares</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca006ONPN2</th>\n",
       "      <td>6</td>\n",
       "      <td>H3</td>\n",
       "      <td>SA508</td>\n",
       "      <td>GACACCGT</td>\n",
       "      <td>SB703</td>\n",
       "      <td>TAGCAGCT</td>\n",
       "      <td>SB703SA508</td>\n",
       "      <td>TAGCAGCT-GACACCGT</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>H3</td>\n",
       "      <td>...</td>\n",
       "      <td>3/25/2015</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>3.0</td>\n",
       "      <td>34</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>nares-AD</td>\n",
       "      <td>nares</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>502 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            PlateNumber PlateLocation     i5 i5Sequence     i7 i7Sequence  \\\n",
       "#sample-id                                                                  \n",
       "Ca009STL              1            A1  SA501   ATCGTACG  SA701   CGAGAGTT   \n",
       "900221                1            B1  SA502   ACTATCTG  SA701   CGAGAGTT   \n",
       "Ca010EBL              1            C1  SA503   TAGCGAGT  SA701   CGAGAGTT   \n",
       "900460                1            D1  SA504   CTGCGTGT  SA701   CGAGAGTT   \n",
       "900051                1            E1  SA505   TCATCGAG  SA701   CGAGAGTT   \n",
       "...                 ...           ...    ...        ...    ...        ...   \n",
       "Ca006ONL2             6            H1  SA508   GACACCGT  SB701   CTCGACTT   \n",
       "Ca006ONNL             6            F2  SA506   CGTGAGTG  SB702   CGAAGTAT   \n",
       "Ca006ONNL2            6            H2  SA508   GACACCGT  SB702   CGAAGTAT   \n",
       "Ca006ONPN             6            F3  SA506   CGTGAGTG  SB703   TAGCAGCT   \n",
       "Ca006ONPN2            6            H3  SA508   GACACCGT  SB703   TAGCAGCT   \n",
       "\n",
       "            identifier           Sequence      Plate ID Well location  ...  \\\n",
       "#sample-id                                                             ...   \n",
       "Ca009STL    SA701SA501  CGAGAGTT-ATCGTACG  1.010000e+21            A1  ...   \n",
       "900221      SA701SA502  CGAGAGTT-ACTATCTG  1.010000e+21            B1  ...   \n",
       "Ca010EBL    SA701SA503  CGAGAGTT-TAGCGAGT  1.010000e+21            C1  ...   \n",
       "900460      SA701SA504  CGAGAGTT-CTGCGTGT  1.010000e+21            D1  ...   \n",
       "900051      SA701SA505  CGAGAGTT-TCATCGAG  1.010000e+21            E1  ...   \n",
       "...                ...                ...           ...           ...  ...   \n",
       "Ca006ONL2   SB701SA508  CTCGACTT-GACACCGT  1.010000e+21            H1  ...   \n",
       "Ca006ONNL   SB702SA506  CGAAGTAT-CGTGAGTG  1.010000e+21            F2  ...   \n",
       "Ca006ONNL2  SB702SA508  CGAAGTAT-GACACCGT  1.010000e+21            H2  ...   \n",
       "Ca006ONPN   SB703SA506  TAGCAGCT-CGTGAGTG  1.010000e+21            F3  ...   \n",
       "Ca006ONPN2  SB703SA508  TAGCAGCT-GACACCGT  1.010000e+21            H3  ...   \n",
       "\n",
       "            enrolment_date enrolment_season hiv_exposure hiv_status  \\\n",
       "#sample-id                                                            \n",
       "Ca009STL         4/16/2015          Autumn     Unexposed   negative   \n",
       "900221           8/11/2015           Winter    Unexposed   negative   \n",
       "Ca010EBL        11/20/2014           Spring    Unexposed   negative   \n",
       "900460           9/23/2015           Spring    Unexposed        NaN   \n",
       "900051           4/21/2015          Autumn     Unexposed   negative   \n",
       "...                    ...              ...          ...        ...   \n",
       "Ca006ONL2        3/25/2015           Autumn    Unexposed   negative   \n",
       "Ca006ONNL        3/25/2015           Autumn    Unexposed   negative   \n",
       "Ca006ONNL2       3/25/2015           Autumn    Unexposed   negative   \n",
       "Ca006ONPN        3/25/2015           Autumn    Unexposed   negative   \n",
       "Ca006ONPN2       3/25/2015           Autumn    Unexposed   negative   \n",
       "\n",
       "           household_size o_scorad  \\\n",
       "#sample-id                           \n",
       "Ca009STL              4.0       40   \n",
       "900221                7.0       34   \n",
       "Ca010EBL              7.0       21   \n",
       "900460                4.0       40   \n",
       "900051                7.0       41   \n",
       "...                   ...      ...   \n",
       "Ca006ONL2             3.0       34   \n",
       "Ca006ONNL             3.0       34   \n",
       "Ca006ONNL2            3.0       34   \n",
       "Ca006ONPN             3.0       34   \n",
       "Ca006ONPN2            3.0       34   \n",
       "\n",
       "                                                 FWD_filepath  \\\n",
       "#sample-id                                                      \n",
       "Ca009STL    /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "900221      /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "Ca010EBL    /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "900460      /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "900051      /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "...                                                       ...   \n",
       "Ca006ONL2   /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "Ca006ONNL   /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "Ca006ONNL2  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "Ca006ONPN   /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "Ca006ONPN2  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "\n",
       "                                                 REV_filepath      group  \\\n",
       "#sample-id                                                                 \n",
       "Ca009STL    /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   skin-ADL   \n",
       "900221      /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   skin-ADL   \n",
       "Ca010EBL    /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   skin-ADL   \n",
       "900460      /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   skin-ADL   \n",
       "900051      /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   skin-ADL   \n",
       "...                                                       ...        ...   \n",
       "Ca006ONL2   /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   skin-ADL   \n",
       "Ca006ONNL   /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...  skin-ADNL   \n",
       "Ca006ONNL2  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...  skin-ADNL   \n",
       "Ca006ONPN   /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   nares-AD   \n",
       "Ca006ONPN2  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   nares-AD   \n",
       "\n",
       "           microbiome_type  \n",
       "#sample-id                  \n",
       "Ca009STL              skin  \n",
       "900221                skin  \n",
       "Ca010EBL              skin  \n",
       "900460                skin  \n",
       "900051                skin  \n",
       "...                    ...  \n",
       "Ca006ONL2             skin  \n",
       "Ca006ONNL             skin  \n",
       "Ca006ONNL2            skin  \n",
       "Ca006ONPN            nares  \n",
       "Ca006ONPN2           nares  \n",
       "\n",
       "[502 rows x 33 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load the metadata\n",
    "metadata_path = '../Data/Metadata/updated_clean_ant_skin_metadata_microbiome_type.tab'\n",
    "metadata = pd.read_csv(metadata_path, sep='\\t')\n",
    "\n",
    "metadata['#sample-id'] = metadata['#sample-id'].str.replace('_', '')\n",
    "# Set Sample-ID as the index for the metadata dataframe \n",
    "metadata = metadata.set_index('#sample-id')\n",
    "\n",
    "\n",
    "# Create group column based on case_type to simplify group names\n",
    "metadata['group'] = metadata['case_type'].map({\n",
    "    'case-lesional skin': 'skin-ADL',\n",
    "    'case-nonlesional skin': 'skin-ADNL', \n",
    "    'control-nonlesional skin': 'skin-H',\n",
    "    'case-anterior nares': 'nares-AD',\n",
    "    'control-anterior nares': 'nares-H'\n",
    "})\n",
    "\n",
    "metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of participants with both skin and nares samples: 187\n",
      "\n",
      "Participant IDs with both sample types:\n",
      "['Ca-006-ON', 'Ca-007-NK', 'Ca-008-HN', 'Ca-009-ST', 'Ca-009-ZN', 'Ca-010-EB', 'Ca-011-LQ', 'Ca-013-NN', 'Ca-014-LB', 'Ca-015-AM', 'Ca-016-YT', 'Ca-017-OM', 'Ca-018-AS', 'Ca-019-EC', 'Ca-020-AZ', 'Ca-021-IM', 'Ca-022-MS', 'Ca-023-EJ', 'Ca-024-ZM', 'Ca-025-AC', 'Ca-026-KM', 'Ca-027-IM', 'Ca-028-PN', 'Ca-029-TD', 'Ca-030-LM', 'Ca-031-AN', 'Ca-032-LN', 'Ca-033-UD', 'Ca-034-LS', 'Ca-035-AR', 'Ca-036-TK', 'Ca-037-HM', 'Ca-042-AM', 'Ca-043-SS', 'Ca-045-HN', 'Ca-046-MT', 'Ca-047-MD', 'Ca-049-LM', 'Ca-050-MM', 'Ca-051-AM', 'Ca-052-LM', 'Ca-053-NN', 'Ca-054-RN', 'Ca-055-LN', 'Ca-056-DH', 'Ca-101-ID', 'Ca-103-BT', 'Ca-104-SB', 'Ca-105-LX', 'Ca-106-NM', 'Ca-107-BM', 'Ca-108-EJ', 'Ca-110-LN', 'Ca-111-IN', 'Ca-112-AM', 'Ca-113-MN', 'Ca-114-QT', 'Ca-115-OM', 'Ca-116-AN', 'Ca-117-SM', 'Ca-118-AS', 'Ca-119-AM', 'Ca-120-ST', 'Ca-121-BT', 'Ca-122-UK', 'Ca-123-PG', 'Ca-124-AB', 'Ca-125-IM', 'Ca-126-KB', 'Ca-127-OM', 'Ca-128-SN', 'Ca-129-AJ', 'Ca-130-PM', 'Ca-131-PM', 'Ca-132-LD', 'Ca-133-LT', 'Ca-134-IT', 'Ca-135-NS', 'Ca-136-OM', 'Ca-137-TM', 'Ca-138-SM', 'Ca-139-OS', 'Ca-140-EM', 'Ca-141-TC', 'Ca-142-AM', 'Ca-143-ST', 'Ca-144-HN', 'Ca-145-LM', 'Ca-146-SM', 'Ca-147-LJ', 'Ca-148-AL', 'Ca-149-LM', 'Ca-150-ZS', 'Ca-151-QR', 'Ca-153-MN', 'Ca-154-OS', 'Ca-155-LS', 'Ca-156-MV', 'Ca-157-LM', 'Ca-158-LC', 'Ca-159-AS', 'Ca-160-LB', 'Ca-161-OM', 'Co-001-SM', 'Co-002-MP', 'Co-003-LT', 'Co-004-LN', 'Co-005-SN', 'Co-011-SF', 'Co-012-ON', 'Co-013-AY', 'Co-014-TN', 'Co-015-OJ', 'Co-016-CB', 'Co-026-LT', 'Co-027-IS', 'Co-028-UV', 'Co-029-YL', 'Co-030-AL', 'Co-031-LM', 'Co-032-AR', 'Co-033-LN', 'Co-034-AS', 'Co-036-OM', 'Co-037-LM', 'Co-038-AG', 'Co-039-KG', 'Co-040-ST', 'Co-041-LC', 'Co-042-LN', 'Co-043-ZS', 'Co-044-KM', 'Co-045-OM', 'Co-046-KL', 'Co-047-LV', 'Co-049-LJ', 'Co-050-AH', 'Co-101-AN', 'Co-102-LD', 'Co-103-MM', 'Co-104-IV', 'Co-105-IV', 'Co-106-AB', 'Co-107-TG', 'Co-108-OG', 'Co-109-AG', 'Co-110-SW', 'Co-111-AM', 'Co-112-SN', 'Co-113-MM', 'Co-114-IG', 'Co-115-SG', 'Co-116-MR', 'Co-117-TG', 'Co-118-MG', 'Co-119-MG', 'Co-120-MM', 'Co-121-MM', 'Co-122-IS', 'Co-123-LZ', 'Co-124-UM', 'Co-125-MM', 'Co-126-ID', 'Co-127-MB', 'Co-128-IV', 'Co-130-MM', 'Co-131-AM', 'Co-132-IM', 'Co-133-KT', 'Co-134-KK', 'Co-135-LT', 'Co-136-SM', 'Co-137-AM', 'Co-138-AT', 'Co-140-IM', 'Co-142-IP', 'Co-143-AN', 'Co-144-LM', 'Co-145-HN', 'Co-146-LK', 'Co-147-SV', 'Co-148-IM', 'Co-149-LM', 'Co-150-LN', 'Co-151-AN', 'Co-152-AF', 'Co-153-YM']\n",
      "187\n"
     ]
    }
   ],
   "source": [
    "# Get participants with both skin and nares samples\n",
    "skin_pids = metadata[metadata['microbiome_type'] == 'skin']['pid']\n",
    "nares_pids = metadata[metadata['microbiome_type'] == 'nares']['pid']\n",
    "\n",
    "# Find overlapping participants\n",
    "shared_pids = set(skin_pids).intersection(set(nares_pids))\n",
    "\n",
    "print(f\"Number of participants with both skin and nares samples: {len(shared_pids)}\")\n",
    "\n",
    "# Display the participant IDs\n",
    "print(\"\\nParticipant IDs with both sample types:\")\n",
    "print(sorted(list(shared_pids)))\n",
    "print(len(list(shared_pids)))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>g__Streptococcus_ASV-1</th>\n",
       "      <th>g__Streptococcus_ASV-2</th>\n",
       "      <th>g__Corynebacterium_ASV-1</th>\n",
       "      <th>g__Corynebacterium_ASV-3</th>\n",
       "      <th>g___ASV-3</th>\n",
       "      <th>g__Cutibacterium_ASV-1</th>\n",
       "      <th>g___ASV-16</th>\n",
       "      <th>g___ASV-25</th>\n",
       "      <th>g___ASV-27</th>\n",
       "      <th>g__Cutibacterium_ASV-2</th>\n",
       "      <th>...</th>\n",
       "      <th>g__Capnocytophaga_820688_ASV-7</th>\n",
       "      <th>g__UBA6175_ASV-2</th>\n",
       "      <th>g__Comamonas_F_589250_ASV-4</th>\n",
       "      <th>g__Bosea_ASV-2</th>\n",
       "      <th>g___ASV-119</th>\n",
       "      <th>group</th>\n",
       "      <th>microbiome_type</th>\n",
       "      <th>o_scorad</th>\n",
       "      <th>pid</th>\n",
       "      <th>area</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>900344</th>\n",
       "      <td>156.0</td>\n",
       "      <td>95.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>skin-H</td>\n",
       "      <td>skin</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Co-130-MM</td>\n",
       "      <td>Umtata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900459</th>\n",
       "      <td>21.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>nares-AD</td>\n",
       "      <td>nares</td>\n",
       "      <td>44</td>\n",
       "      <td>Ca-145-LM</td>\n",
       "      <td>Umtata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900221</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>skin</td>\n",
       "      <td>34</td>\n",
       "      <td>Ca-101-ID</td>\n",
       "      <td>Umtata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900570</th>\n",
       "      <td>18.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>skin-ADNL</td>\n",
       "      <td>skin</td>\n",
       "      <td>36</td>\n",
       "      <td>Ca-046-MT</td>\n",
       "      <td>Cape Town</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900092</th>\n",
       "      <td>174.0</td>\n",
       "      <td>104.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>nares-AD</td>\n",
       "      <td>nares</td>\n",
       "      <td>53</td>\n",
       "      <td>Ca-023-EJ</td>\n",
       "      <td>Cape Town</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9003972</th>\n",
       "      <td>38.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>skin</td>\n",
       "      <td>54</td>\n",
       "      <td>Ca-125-IM</td>\n",
       "      <td>Umtata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900097</th>\n",
       "      <td>3.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>skin-ADNL</td>\n",
       "      <td>skin</td>\n",
       "      <td>44</td>\n",
       "      <td>Ca-025-AC</td>\n",
       "      <td>Cape Town</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900498</th>\n",
       "      <td>4.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>skin-ADNL</td>\n",
       "      <td>skin</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ca-157-LM</td>\n",
       "      <td>Umtata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900276</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>69.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>skin</td>\n",
       "      <td>52</td>\n",
       "      <td>Ca-111-IN</td>\n",
       "      <td>Umtata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900406</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>skin-ADNL</td>\n",
       "      <td>skin</td>\n",
       "      <td>28</td>\n",
       "      <td>Ca-127-OM</td>\n",
       "      <td>Umtata</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>502 rows × 803 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         g__Streptococcus_ASV-1  g__Streptococcus_ASV-2  \\\n",
       "900344                    156.0                    95.0   \n",
       "900459                     21.0                    30.0   \n",
       "900221                        0                       0   \n",
       "900570                     18.0                       0   \n",
       "900092                    174.0                   104.0   \n",
       "...                         ...                     ...   \n",
       "9003972                    38.0                    21.0   \n",
       "900097                      3.0                       0   \n",
       "900498                      4.0                     6.0   \n",
       "900276                        0                       0   \n",
       "900406                        0                       0   \n",
       "\n",
       "         g__Corynebacterium_ASV-1  g__Corynebacterium_ASV-3  g___ASV-3  \\\n",
       "900344                       23.0                      17.0        1.0   \n",
       "900459                          0                         0          0   \n",
       "900221                          0                         0          0   \n",
       "900570                          0                         0          0   \n",
       "900092                       10.0                       1.0          0   \n",
       "...                           ...                       ...        ...   \n",
       "9003972                         0                         0        1.0   \n",
       "900097                          0                         0          0   \n",
       "900498                          0                         0          0   \n",
       "900276                       10.0                         0          0   \n",
       "900406                          0                         0          0   \n",
       "\n",
       "         g__Cutibacterium_ASV-1  g___ASV-16  g___ASV-25  g___ASV-27  \\\n",
       "900344                      2.0         1.0           0           0   \n",
       "900459                        0           0           0           0   \n",
       "900221                        0           0           0           0   \n",
       "900570                        0           0           0           0   \n",
       "900092                        0           0           0         2.0   \n",
       "...                         ...         ...         ...         ...   \n",
       "9003972                    18.0           0           0           0   \n",
       "900097                      3.0           0           0           0   \n",
       "900498                     13.0           0         6.0           0   \n",
       "900276                     69.0           0           0           0   \n",
       "900406                        0           0           0           0   \n",
       "\n",
       "         g__Cutibacterium_ASV-2  ...  g__Capnocytophaga_820688_ASV-7  \\\n",
       "900344                        0  ...                               0   \n",
       "900459                        0  ...                               0   \n",
       "900221                        0  ...                               0   \n",
       "900570                        0  ...                               0   \n",
       "900092                        0  ...                               0   \n",
       "...                         ...  ...                             ...   \n",
       "9003972                    12.0  ...                               0   \n",
       "900097                        0  ...                               0   \n",
       "900498                     13.0  ...                               0   \n",
       "900276                     26.0  ...                               0   \n",
       "900406                        0  ...                               0   \n",
       "\n",
       "         g__UBA6175_ASV-2  g__Comamonas_F_589250_ASV-4  g__Bosea_ASV-2  \\\n",
       "900344                  0                            0               0   \n",
       "900459                  0                            0               0   \n",
       "900221                  0                            0               0   \n",
       "900570                  0                            0               0   \n",
       "900092                  0                            0               0   \n",
       "...                   ...                          ...             ...   \n",
       "9003972                 0                            0               0   \n",
       "900097                  0                            0               0   \n",
       "900498                  0                            0               0   \n",
       "900276                  0                            0               0   \n",
       "900406                  0                            0               0   \n",
       "\n",
       "         g___ASV-119      group  microbiome_type  o_scorad        pid  \\\n",
       "900344             0     skin-H             skin       NaN  Co-130-MM   \n",
       "900459             0   nares-AD            nares        44  Ca-145-LM   \n",
       "900221             0   skin-ADL             skin        34  Ca-101-ID   \n",
       "900570             0  skin-ADNL             skin        36  Ca-046-MT   \n",
       "900092             0   nares-AD            nares        53  Ca-023-EJ   \n",
       "...              ...        ...              ...       ...        ...   \n",
       "9003972            0   skin-ADL             skin        54  Ca-125-IM   \n",
       "900097             0  skin-ADNL             skin        44  Ca-025-AC   \n",
       "900498             0  skin-ADNL             skin       NaN  Ca-157-LM   \n",
       "900276             0   skin-ADL             skin        52  Ca-111-IN   \n",
       "900406             0  skin-ADNL             skin        28  Ca-127-OM   \n",
       "\n",
       "              area  \n",
       "900344      Umtata  \n",
       "900459      Umtata  \n",
       "900221      Umtata  \n",
       "900570   Cape Town  \n",
       "900092   Cape Town  \n",
       "...            ...  \n",
       "9003972     Umtata  \n",
       "900097   Cape Town  \n",
       "900498      Umtata  \n",
       "900276      Umtata  \n",
       "900406      Umtata  \n",
       "\n",
       "[502 rows x 803 columns]"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read in table at ASV level\n",
    "# biom_path = '../Data/Tables/Absolute_Abundance_Tables/feature_table_with_tax_labels_Genus.biom'\n",
    "biom_path = '../Data/Tables/Absolute_Abundance_Tables/209766_filtered_by_prevalence_1pct_rare_Genus-ASV-non-collapse.biom'\n",
    "biom_tbl = load_table(biom_path)\n",
    "df = pd.DataFrame(biom_tbl.to_dataframe().T)\n",
    "\n",
    "# delete the prefix from the index\n",
    "df.index = df.index.str.replace('15564.', '')\n",
    "\n",
    "# Map the 'microbiome_type', 'group', 'o_scorad', and 'area' columns from metadata to df based on matching index\n",
    "df['group'] = metadata.loc[df.index, 'group']\n",
    "df['microbiome_type'] = metadata.loc[df.index, 'microbiome_type']\n",
    "df['o_scorad'] = metadata.loc[df.index, 'o_scorad']\n",
    "df['pid'] = metadata.loc[df.index, 'pid']\n",
    "df['area'] = metadata.loc[df.index, 'area']\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_group_data(df, area_val, is_AD_val, microbiome_type):\n",
    "    return df[\n",
    "        (df['area'] == area_val) &\n",
    "        (df['is_AD'] == is_AD_val) &\n",
    "        (df['microbiome_type'] == microbiome_type)\n",
    "    ].drop(columns=['microbiome_type', 'group', 'area', 'is_AD'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "# --- STEP 1: Prepare the data ---\n",
    "\n",
    "# Separate features and metadata\n",
    "features = df.drop(columns=['microbiome_type', 'group'])\n",
    "metadata_cols = df[['microbiome_type', 'group', 'area']]\n",
    "\n",
    "# Coerce all to numeric and drop non-numeric columns\n",
    "features_numeric = features.apply(pd.to_numeric, errors='coerce')\n",
    "features_numeric = features_numeric.dropna(axis=1, how='all')\n",
    "\n",
    "# Normalize to relative abundances\n",
    "features_rel = features_numeric.div(features_numeric.sum(axis=1), axis=0)\n",
    "\n",
    "# Add metadata back\n",
    "features_rel['microbiome_type'] = metadata_cols['microbiome_type']\n",
    "features_rel['group'] = metadata_cols['group']\n",
    "features_rel['area'] = metadata_cols['area']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/pandas/core/arrays/sparse/array.py:1556: RuntimeWarning: invalid value encountered in scalar divide\n",
      "  return sp_sum / ct\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/pandas/core/arrays/sparse/array.py:1707: RuntimeWarning: divide by zero encountered in log10\n",
      "  sp_values = getattr(ufunc, method)(self.sp_values, **kwargs)\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/pandas/core/arrays/sparse/array.py:1707: RuntimeWarning: divide by zero encountered in log10\n",
      "  sp_values = getattr(ufunc, method)(self.sp_values, **kwargs)\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/pandas/core/arrays/sparse/array.py:1556: RuntimeWarning: invalid value encountered in scalar divide\n",
      "  return sp_sum / ct\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/pandas/core/arrays/sparse/array.py:1707: RuntimeWarning: divide by zero encountered in log10\n",
      "  sp_values = getattr(ufunc, method)(self.sp_values, **kwargs)\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/pandas/core/arrays/sparse/array.py:1707: RuntimeWarning: divide by zero encountered in log10\n",
      "  sp_values = getattr(ufunc, method)(self.sp_values, **kwargs)\n"
     ]
    }
   ],
   "source": [
    "# Make sure you have these helper columns\n",
    "features_rel['is_AD'] = features_rel['group'].isin(['skin-ADL', 'nares-AD'])\n",
    "\n",
    "# Define plotting setup for 1x4 layout with custom figure size\n",
    "fig, axes = plt.subplots(1, 4, figsize=(15, 5), sharex=True, sharey=True)\n",
    "\n",
    "# Order: Cape Town Healthy, Cape Town AD+, Umtata Healthy, Umtata AD+\n",
    "panel_info = [\n",
    "    ('Cape Town', False, 'Cape Town Healthy', '#A7C7E7', 0),\n",
    "    ('Cape Town', True,  'Cape Town AD+',     '#d2b48c', 1),\n",
    "    ('Umtata',    False, 'Umtata Healthy',     '#ADD8E6', 2),\n",
    "    ('Umtata',    True,  'Umtata AD+',        '#fa8072', 3) \n",
    "]\n",
    "\n",
    "for area, is_AD, title, color, idx in panel_info:\n",
    "    # Subset skin samples\n",
    "    skin = features_rel[\n",
    "        (features_rel['area'] == area) &\n",
    "        (features_rel['is_AD'] == is_AD) &\n",
    "        (features_rel['microbiome_type'] == 'skin')\n",
    "    ].drop(columns=['microbiome_type', 'group', 'area', 'is_AD'])\n",
    "\n",
    "    # Subset nares samples\n",
    "    nares = features_rel[\n",
    "        (features_rel['area'] == area) &\n",
    "        (features_rel['is_AD'] == is_AD) &\n",
    "        (features_rel['microbiome_type'] == 'nares')\n",
    "    ].drop(columns=['microbiome_type', 'group', 'area', 'is_AD'])\n",
    "\n",
    "    # Average across samples\n",
    "    mean_skin = skin.mean()\n",
    "    mean_nares = nares.mean()\n",
    "\n",
    "    # Log transform\n",
    "    log_skin = np.log10(mean_skin)\n",
    "    log_nares = np.log10(mean_nares)\n",
    "\n",
    "    # Remove NaNs and infs\n",
    "    mask = np.isfinite(log_skin) & np.isfinite(log_nares)\n",
    "    log_skin = log_skin[mask]\n",
    "    log_nares = log_nares[mask]\n",
    "    mean_skin = mean_skin[mask]\n",
    "    mean_nares = mean_nares[mask]\n",
    "\n",
    "    # Correlation\n",
    "    r, pval = stats.pearsonr(log_skin, log_nares)\n",
    "\n",
    "    # Plot\n",
    "    ax = axes[idx]\n",
    "    ax.scatter(\n",
    "        mean_nares,\n",
    "        mean_skin,\n",
    "        alpha=0.7,\n",
    "        edgecolors='k',\n",
    "        linewidths=0.5,\n",
    "        color=color\n",
    "    )\n",
    "\n",
    "    # Identity line\n",
    "    #lims = [1e-6, max(max(mean_nares), max(mean_skin))]\n",
    "    lims = [1e-5, 1e0]\n",
    "    ax.plot(lims, lims, 'k--', alpha=0.75)\n",
    "    ax.set_xscale('log')\n",
    "    ax.set_yscale('log')\n",
    "    ax.set_xlim(lims)\n",
    "    ax.set_ylim(lims)\n",
    "    ax.set_title(f'{title}', fontsize=16)\n",
    "    ax.set_xlabel(f'Nares ({area})', fontsize=14)\n",
    "    ax.set_ylabel(f'Skin ({area})', fontsize=14)\n",
    "\n",
    "    ax.text(\n",
    "        0.05, 0.95,\n",
    "        f'Pearson r = {r:.2f}\\n$p$ = {pval:.1e}',\n",
    "        ha='left', va='top',\n",
    "        transform=ax.transAxes,\n",
    "        fontsize=11\n",
    "    )\n",
    "\n",
    "    # Annotation markers\n",
    "    ax.plot(0.05, 0.80, marker='o', markersize=6, color=color, transform=ax.transAxes)\n",
    "\n",
    "    n_features = mask.sum()\n",
    "    ax.text(\n",
    "        0.07, 0.80,\n",
    "        f'16S ASV (n={n_features})',\n",
    "        ha='left', va='center',\n",
    "        transform=ax.transAxes,\n",
    "        fontsize=10\n",
    "    )\n",
    "\n",
    "    # --- Label selected ASVs with short names and position adjustments ---\n",
    "    label_map = {\n",
    "        0: {  # Cape Town Healthy\n",
    "            'g__Streptococcus_ASV-1': ('Strep', 0.99, 1.6),\n",
    "            'g__Staphylococcus_ASV-1': ('Staph', 0.99, 1.6)\n",
    "        },\n",
    "        1: {  # Cape Town AD+\n",
    "            'g__Streptococcus_ASV-1': ('Strep', 0.99, 0.6),\n",
    "            'g__Staphylococcus_ASV-1': ('Staph', 0.99, 1.6)\n",
    "        },\n",
    "        2: {  # Umtata Healthy\n",
    "            'g__Streptococcus_ASV-1': ('Strep', 0.99, 0.6),\n",
    "            'g__Staphylococcus_ASV-1': ('Staph', 0.99, 1.65)\n",
    "        },\n",
    "        3: {  # Umtata AD+\n",
    "            'g__Streptococcus_ASV-1': ('Strep', 0.99, 0.6),\n",
    "            'g__Staphylococcus_ASV-1': ('Staph', 0.99, 1.6)        }\n",
    "    }\n",
    "\n",
    "    # Get label positions for this panel\n",
    "    panel_labels = label_map.get(idx, {})\n",
    "\n",
    "    for asv, (short_label, x_shift_factor, y_shift_factor) in panel_labels.items():\n",
    "        if asv in mean_skin.index and asv in mean_nares.index:\n",
    "            x = mean_nares[asv]\n",
    "            y = mean_skin[asv]\n",
    "\n",
    "            if np.isfinite(np.log10(x)) and np.isfinite(np.log10(y)):\n",
    "                x_offset = x * x_shift_factor\n",
    "                y_offset = y * y_shift_factor\n",
    "\n",
    "                align = 'right' if x_shift_factor < 1 else 'left'\n",
    "\n",
    "                # Draw thick black outline around the ASV point\n",
    "                ax.scatter(\n",
    "                    [x], [y],\n",
    "                    s=80,\n",
    "                    facecolors='none',\n",
    "                    edgecolors='black',\n",
    "                    linewidths=1\n",
    "                )\n",
    "\n",
    "                # Label\n",
    "                ax.text(\n",
    "                    x_offset,\n",
    "                    y_offset,\n",
    "                    short_label,\n",
    "                    fontsize=10,\n",
    "                    ha=align,\n",
    "                    va='center',\n",
    "                    color='black'\n",
    "                )\n",
    "\n",
    "# Final layout\n",
    "plt.suptitle('Mean Relative Abundance of Taxa between Skin and Nares', fontsize=20, y=0.99)\n",
    "plt.tight_layout(rect=[0, 0.02, 1, 1])\n",
    "plt.savefig('../Figures/Main/Fig_4B.png', dpi=600)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "qiime2-metagenome-2024.10",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
