{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Beta Diversity (RPCA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 652,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import Python packages\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from numpy import array\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from matplotlib.gridspec import GridSpec\n",
    "import scipy\n",
    "import scipy.stats as ss\n",
    "from skbio.stats.distance import permanova\n",
    "import biom\n",
    "from biom import load_table\n",
    "from biom.util import biom_open\n",
    "from gemelli.rpca import rpca\n",
    "from matplotlib.patches import Circle\n",
    "from matplotlib.colors import to_hex\n",
    "import statsmodels.api as sm\n",
    "### See RPCA standalone python tutorial: \n",
    "# https://github.com/biocore/gemelli/blob/master/ipynb/tutorials/RPCA-moving-pictures-standalone-cli-and-api.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 653,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to load BIOM table, collapse by taxa, sort rows by row sum, remove specified samples, and convert to relative abundance\n",
    "def load_biom_table(biom_path, metadata_path):\n",
    "    # Load BIOM table and convert to a DataFrame\n",
    "    table = biom.load_table(biom_path)\n",
    "    df = pd.DataFrame(table.matrix_data.toarray(),\n",
    "                      index=table.ids(axis='observation'),\n",
    "                      columns=table.ids(axis='sample'))\n",
    "    \n",
    "    # Sort rows by row sum in descending order\n",
    "    df['row_sum'] = df.sum(axis=1)\n",
    "    df = df.sort_values(by='row_sum', ascending=False)\n",
    "    df = df.drop(columns=['row_sum'])\n",
    "\n",
    "    # Replace ' g__' rows with ' g__Unknown'\n",
    "    df.index = df.index.map(lambda x: ' g__Unknown' if x == ' g__' else x)\n",
    "\n",
    "    # Remove '15564.' prefix from columns\n",
    "    df.columns = df.columns.str.replace('15564.', '')\n",
    "\n",
    "    # Load metadata\n",
    "    metadata = pd.read_csv(metadata_path, sep='\\t')\n",
    "    metadata['#sample-id'] = metadata['#sample-id'].str.replace('_', '')\n",
    "\n",
    "    # Set Sample-ID as index\n",
    "    metadata = metadata.set_index('#sample-id')\n",
    "\n",
    "    # Subset metadata to only samples in the BIOM table\n",
    "    metadata_sub = metadata.loc[df.columns]\n",
    "\n",
    "    # Get a list of the nares samples\n",
    "    nares_samples = metadata_sub[metadata_sub['case_type'].str.endswith('nares')].index.tolist()\n",
    "\n",
    "    # Drop nares samples from BIOM table\n",
    "    df = df.drop(columns=nares_samples)\n",
    "\n",
    "    # Recompute metadata_sub after nares removal\n",
    "    metadata_sub = metadata.loc[df.columns]\n",
    "\n",
    "    # Compute individual_case_location\n",
    "    metadata['individual_case_location'] = metadata['case_type'] + ' ' + metadata['area']\n",
    "\n",
    "    # Convert SCORAD and filter only AD samples with SCORAD > 50\n",
    "    metadata['o_scorad'] = pd.to_numeric(metadata['o_scorad'], errors='coerce')\n",
    "\n",
    "    # Keep healthy samples (NaN SCORAD) and AD samples with SCORAD â‰¤ 50\n",
    "    metadata = metadata[(metadata['o_scorad'].isna()) | (metadata['o_scorad'] <= 50)]\n",
    "\n",
    "    # Subset both df and metadata to overlapping samples\n",
    "    shared_samples = df.columns.intersection(metadata.index)\n",
    "    df = df[shared_samples]\n",
    "    metadata_sub = metadata.loc[shared_samples]\n",
    "\n",
    "    return df, metadata_sub\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 654,
   "metadata": {},
   "outputs": [],
   "source": [
    "# converts the subsetted skin only BIOM table DataFrame back into a BIOM to run RPCA\n",
    "def convert_df_to_biom(table, biom_output_file):\n",
    "    obs_ids = table.index\n",
    "    samp_ids = table.columns\n",
    "    biom_table = biom.table.Table(table.values, observation_ids=obs_ids, sample_ids=samp_ids) # Convert df back to a biom table\n",
    "    #biom_output_file = f\"../tables/{name}_subset.biom\" # Path to output file\n",
    "\n",
    "    with biom_open(biom_output_file, 'w') as f: # This does the actual saving!\n",
    "        biom_table.to_hdf5(f, generated_by=\"subsetted tables\")\n",
    "\n",
    "    return biom_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 655,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>900344</th>\n",
       "      <th>900221</th>\n",
       "      <th>900570</th>\n",
       "      <th>900129</th>\n",
       "      <th>900321</th>\n",
       "      <th>900245</th>\n",
       "      <th>900423</th>\n",
       "      <th>900145</th>\n",
       "      <th>900364</th>\n",
       "      <th>900329</th>\n",
       "      <th>...</th>\n",
       "      <th>900279</th>\n",
       "      <th>900093</th>\n",
       "      <th>900094</th>\n",
       "      <th>900287</th>\n",
       "      <th>900225</th>\n",
       "      <th>900057</th>\n",
       "      <th>900294</th>\n",
       "      <th>900097</th>\n",
       "      <th>900498</th>\n",
       "      <th>900406</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GTGCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTTATTGGGCGTAAAGCGAGCGCAGGCGGTTAGATAAGTCTGAAGTTAAAGGCTG</th>\n",
       "      <td>984.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>389.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>91.0</td>\n",
       "      <td>...</td>\n",
       "      <td>133.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>57.0</td>\n",
       "      <td>44.0</td>\n",
       "      <td>111.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTGCCAGCAGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAATTATTGGGCGTAAAGCGCGCGTAGGCGGTTTTTTAAGTCTGATGTGAAAGCCCA</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1294.0</td>\n",
       "      <td>182.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>159.0</td>\n",
       "      <td>49.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>159.0</td>\n",
       "      <td>4953.0</td>\n",
       "      <td>...</td>\n",
       "      <td>908.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>108.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>93.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>171.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTGCCAGCAGCCGCGGTAATACGGAGGGTGCGAGCGTTAATCGGAATAACTGGGCGTAAAGGGCACGCAGGCGGTTATTTAAGTGAGGTGTGAAAGCCCC</th>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>...</td>\n",
       "      <td>53.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTGCCAGCCGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAATTATTGGGCGTAAAGCGCGCGTAGGCGGTTTTTTAAGTCTGATGTGAAAGCCCA</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>236.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>57.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1560.0</td>\n",
       "      <td>...</td>\n",
       "      <td>499.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>116.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>94.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTGCCAGCCGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTTATTGGGCGTAAAGCGAGCGCAGGCGGTTAGATAAGTCTGAAGTTAAAGGCTG</th>\n",
       "      <td>611.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>71.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTGCCAGCAGCCGCGGTAATACGGAGGGTCCGAGCGTTATCCGGAATCATTGGGTTTAAAGGGTCCGTAGGCGGCTTTATAAGTCAGTGGTGAAATCCGG</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTGCCAGCAGCCGCGGTAATACGTAGGGTGCAAGCGTTGTCCGGAATTATTGGGCGTAAAGAGCTCGTAGGCGGTTTGTCGCGTCTGCTGTGAAATTCCG</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTGCCAGCCGCCGCGGTAATACGTAGGGCGCAAGCGTTGTCCGGAATTATTGGGCGTAAAGAGCTCGTAGGCGGTTTGTCGCGTCTGGTGTGAAAACTCG</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTGCCAGAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTTATTGGGCGTAAAGCGAGCGCAGGCGGTTAGATAAGTCTGAAGTTAAAGGCTGT</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTTATTGGGCGTAAAGCGAGCGCAGGCGGTTAGATAAGTCTGAAGTTAAAGGCTGT</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2283 rows Ã— 262 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    900344  900221  900570  \\\n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATT...   984.0    22.0   389.0   \n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAAT...     0.0     0.0  1294.0   \n",
       "GTGCCAGCAGCCGCGGTAATACGGAGGGTGCGAGCGTTAATCGGAAT...     0.0     2.0     9.0   \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAAT...     0.0     0.0   236.0   \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATT...   611.0     0.0     0.0   \n",
       "...                                                    ...     ...     ...   \n",
       "GTGCCAGCAGCCGCGGTAATACGGAGGGTCCGAGCGTTATCCGGAAT...     0.0     0.0     0.0   \n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGGTGCAAGCGTTGTCCGGAAT...     0.0     0.0     0.0   \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGGCGCAAGCGTTGTCCGGAAT...     0.0     0.0     0.0   \n",
       "GTGCCAGAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTT...     0.0     0.0     0.0   \n",
       "GTCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTT...     0.0     0.0     0.0   \n",
       "\n",
       "                                                    900129  900321  900245  \\\n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATT...     0.0    26.0    40.0   \n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAAT...   182.0     9.0   159.0   \n",
       "GTGCCAGCAGCCGCGGTAATACGGAGGGTGCGAGCGTTAATCGGAAT...     0.0     0.0     0.0   \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAAT...     0.0     0.0     0.0   \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATT...     0.0     0.0     0.0   \n",
       "...                                                    ...     ...     ...   \n",
       "GTGCCAGCAGCCGCGGTAATACGGAGGGTCCGAGCGTTATCCGGAAT...     0.0     0.0     0.0   \n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGGTGCAAGCGTTGTCCGGAAT...     0.0     0.0     0.0   \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGGCGCAAGCGTTGTCCGGAAT...     0.0     0.0     0.0   \n",
       "GTGCCAGAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTT...     0.0     0.0     0.0   \n",
       "GTCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTT...     0.0     0.0     0.0   \n",
       "\n",
       "                                                    900423  900145  900364  \\\n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATT...     0.0     0.0   100.0   \n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAAT...    49.0    27.0   159.0   \n",
       "GTGCCAGCAGCCGCGGTAATACGGAGGGTGCGAGCGTTAATCGGAAT...     0.0     0.0     0.0   \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAAT...    57.0    37.0     0.0   \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATT...     0.0     6.0    71.0   \n",
       "...                                                    ...     ...     ...   \n",
       "GTGCCAGCAGCCGCGGTAATACGGAGGGTCCGAGCGTTATCCGGAAT...     0.0     0.0     0.0   \n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGGTGCAAGCGTTGTCCGGAAT...     0.0     0.0     0.0   \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGGCGCAAGCGTTGTCCGGAAT...     0.0     0.0     0.0   \n",
       "GTGCCAGAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTT...     0.0     0.0     0.0   \n",
       "GTCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTT...     0.0     0.0     0.0   \n",
       "\n",
       "                                                    900329  ...  900279  \\\n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATT...    91.0  ...   133.0   \n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAAT...  4953.0  ...   908.0   \n",
       "GTGCCAGCAGCCGCGGTAATACGGAGGGTGCGAGCGTTAATCGGAAT...    17.0  ...    53.0   \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAAT...  1560.0  ...   499.0   \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATT...     0.0  ...     0.0   \n",
       "...                                                    ...  ...     ...   \n",
       "GTGCCAGCAGCCGCGGTAATACGGAGGGTCCGAGCGTTATCCGGAAT...     0.0  ...     0.0   \n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGGTGCAAGCGTTGTCCGGAAT...     0.0  ...     0.0   \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGGCGCAAGCGTTGTCCGGAAT...     0.0  ...     0.0   \n",
       "GTGCCAGAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTT...     0.0  ...     0.0   \n",
       "GTCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTT...     0.0  ...     0.0   \n",
       "\n",
       "                                                    900093  900094  900287  \\\n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATT...    18.0     0.0    57.0   \n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAAT...     0.0     0.0   108.0   \n",
       "GTGCCAGCAGCCGCGGTAATACGGAGGGTGCGAGCGTTAATCGGAAT...     0.0     0.0     2.0   \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAAT...     0.0     0.0     0.0   \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATT...     0.0     0.0     0.0   \n",
       "...                                                    ...     ...     ...   \n",
       "GTGCCAGCAGCCGCGGTAATACGGAGGGTCCGAGCGTTATCCGGAAT...     0.0     0.0     0.0   \n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGGTGCAAGCGTTGTCCGGAAT...     0.0     0.0     0.0   \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGGCGCAAGCGTTGTCCGGAAT...     0.0     0.0     0.0   \n",
       "GTGCCAGAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTT...     0.0     0.0     0.0   \n",
       "GTCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTT...     0.0     0.0     0.0   \n",
       "\n",
       "                                                    900225  900057  900294  \\\n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATT...    44.0   111.0    12.0   \n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAAT...    30.0    93.0    13.0   \n",
       "GTGCCAGCAGCCGCGGTAATACGGAGGGTGCGAGCGTTAATCGGAAT...     0.0     0.0     0.0   \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAAT...     4.0   116.0     0.0   \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATT...     0.0     0.0     7.0   \n",
       "...                                                    ...     ...     ...   \n",
       "GTGCCAGCAGCCGCGGTAATACGGAGGGTCCGAGCGTTATCCGGAAT...     0.0     0.0     0.0   \n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGGTGCAAGCGTTGTCCGGAAT...     0.0     0.0     0.0   \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGGCGCAAGCGTTGTCCGGAAT...     0.0     0.0     0.0   \n",
       "GTGCCAGAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTT...     0.0     0.0     0.0   \n",
       "GTCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTT...     0.0     0.0     0.0   \n",
       "\n",
       "                                                    900097  900498  900406  \n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATT...    24.0    15.0     0.0  \n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAAT...   171.0    28.0    12.0  \n",
       "GTGCCAGCAGCCGCGGTAATACGGAGGGTGCGAGCGTTAATCGGAAT...     0.0     0.0     0.0  \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAAT...    94.0    11.0     7.0  \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATT...     0.0    17.0     0.0  \n",
       "...                                                    ...     ...     ...  \n",
       "GTGCCAGCAGCCGCGGTAATACGGAGGGTCCGAGCGTTATCCGGAAT...     0.0     0.0     0.0  \n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGGTGCAAGCGTTGTCCGGAAT...     0.0     0.0     0.0  \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGGCGCAAGCGTTGTCCGGAAT...     0.0     0.0     0.0  \n",
       "GTGCCAGAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTT...     0.0     0.0     0.0  \n",
       "GTCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTT...     0.0     0.0     0.0  \n",
       "\n",
       "[2283 rows x 262 columns]"
      ]
     },
     "execution_count": 655,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# specify the path to the BIOM table and Metadata\n",
    "biom_path = \"../Data/Tables/Absolute_Abundance_Tables/209766_filtered_feature_table.biom\"\n",
    "metadata_path = '../Data/Metadata/updated_clean_ant_skin_metadata_microbiome_type-location.tab'\n",
    "skin_only_table, metadata_sub = load_biom_table(biom_path, metadata_path)\n",
    "\n",
    "skin_only_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 656,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>PlateNumber</th>\n",
       "      <th>PlateLocation</th>\n",
       "      <th>i5</th>\n",
       "      <th>i5Sequence</th>\n",
       "      <th>i7</th>\n",
       "      <th>i7Sequence</th>\n",
       "      <th>identifier</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Plate ID</th>\n",
       "      <th>...</th>\n",
       "      <th>enrolment_season</th>\n",
       "      <th>hiv_exposure</th>\n",
       "      <th>hiv_status</th>\n",
       "      <th>household_size</th>\n",
       "      <th>o_scorad</th>\n",
       "      <th>FWD_filepath</th>\n",
       "      <th>REV_filepath</th>\n",
       "      <th>o_scorad_adj</th>\n",
       "      <th>individual_case</th>\n",
       "      <th>individual_case_location</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>900344</th>\n",
       "      <td>314</td>\n",
       "      <td>4</td>\n",
       "      <td>C10</td>\n",
       "      <td>SA503</td>\n",
       "      <td>TAGCGAGT</td>\n",
       "      <td>SA710</td>\n",
       "      <td>GTCTGCTA</td>\n",
       "      <td>SA710SA503</td>\n",
       "      <td>GTCTGCTA-TAGCGAGT</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>...</td>\n",
       "      <td>Spring</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>0</td>\n",
       "      <td>H skin</td>\n",
       "      <td>control-nonlesional skin Umtata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900221</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>B1</td>\n",
       "      <td>SA502</td>\n",
       "      <td>ACTATCTG</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA502</td>\n",
       "      <td>CGAGAGTT-ACTATCTG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>...</td>\n",
       "      <td>Winter</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>7.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>34</td>\n",
       "      <td>AD skin</td>\n",
       "      <td>case-lesional skin Umtata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900570</th>\n",
       "      <td>301</td>\n",
       "      <td>4</td>\n",
       "      <td>F8</td>\n",
       "      <td>SA506</td>\n",
       "      <td>CGTGAGTG</td>\n",
       "      <td>SA708</td>\n",
       "      <td>CGAGCGAC</td>\n",
       "      <td>SA708SA506</td>\n",
       "      <td>CGAGCGAC-CGTGAGTG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>...</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>4.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>36</td>\n",
       "      <td>AD skin</td>\n",
       "      <td>case-nonlesional skin Cape Town</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900129</th>\n",
       "      <td>250</td>\n",
       "      <td>4</td>\n",
       "      <td>A1</td>\n",
       "      <td>SA501</td>\n",
       "      <td>ATCGTACG</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA501</td>\n",
       "      <td>CGAGAGTT-ATCGTACG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>...</td>\n",
       "      <td>Winter</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>6.0</td>\n",
       "      <td>44.0</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>44</td>\n",
       "      <td>AD skin</td>\n",
       "      <td>case-lesional skin Cape Town</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900321</th>\n",
       "      <td>273</td>\n",
       "      <td>4</td>\n",
       "      <td>B4</td>\n",
       "      <td>SA502</td>\n",
       "      <td>ACTATCTG</td>\n",
       "      <td>SA704</td>\n",
       "      <td>ACTCACTG</td>\n",
       "      <td>SA704SA502</td>\n",
       "      <td>ACTCACTG-ACTATCTG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>...</td>\n",
       "      <td>Spring</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>0</td>\n",
       "      <td>H skin</td>\n",
       "      <td>control-nonlesional skin Umtata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900057</th>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>G1</td>\n",
       "      <td>SA507</td>\n",
       "      <td>GGATATCT</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA507</td>\n",
       "      <td>CGAGAGTT-GGATATCT</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>...</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>9.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>33</td>\n",
       "      <td>AD skin</td>\n",
       "      <td>case-lesional skin Cape Town</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900294</th>\n",
       "      <td>134</td>\n",
       "      <td>2</td>\n",
       "      <td>A8</td>\n",
       "      <td>SB501</td>\n",
       "      <td>CTACTATA</td>\n",
       "      <td>SB708</td>\n",
       "      <td>ATAGTACC</td>\n",
       "      <td>SB708SB501</td>\n",
       "      <td>ATAGTACC-CTACTATA</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>...</td>\n",
       "      <td>Winter</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>4.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>23</td>\n",
       "      <td>AD skin</td>\n",
       "      <td>case-nonlesional skin Umtata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900097</th>\n",
       "      <td>141</td>\n",
       "      <td>2</td>\n",
       "      <td>H8</td>\n",
       "      <td>SB508</td>\n",
       "      <td>GTCAGATA</td>\n",
       "      <td>SB708</td>\n",
       "      <td>ATAGTACC</td>\n",
       "      <td>SB708SB508</td>\n",
       "      <td>ATAGTACC-GTCAGATA</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>...</td>\n",
       "      <td>Winter</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>6.0</td>\n",
       "      <td>44.0</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>44</td>\n",
       "      <td>AD skin</td>\n",
       "      <td>case-nonlesional skin Cape Town</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900498</th>\n",
       "      <td>464</td>\n",
       "      <td>6</td>\n",
       "      <td>H11</td>\n",
       "      <td>SA508</td>\n",
       "      <td>GACACCGT</td>\n",
       "      <td>SB711</td>\n",
       "      <td>AACGCTGA</td>\n",
       "      <td>SB711SA508</td>\n",
       "      <td>AACGCTGA-GACACCGT</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>0</td>\n",
       "      <td>AD skin</td>\n",
       "      <td>case-nonlesional skin Umtata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900406</th>\n",
       "      <td>260</td>\n",
       "      <td>4</td>\n",
       "      <td>D2</td>\n",
       "      <td>SA504</td>\n",
       "      <td>CTGCGTGT</td>\n",
       "      <td>SA702</td>\n",
       "      <td>GACATAGT</td>\n",
       "      <td>SA702SA504</td>\n",
       "      <td>GACATAGT-CTGCGTGT</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>...</td>\n",
       "      <td>Spring</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>6.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>28</td>\n",
       "      <td>AD skin</td>\n",
       "      <td>case-nonlesional skin Umtata</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>262 rows Ã— 35 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Unnamed: 0  PlateNumber PlateLocation     i5 i5Sequence     i7  \\\n",
       "900344         314            4           C10  SA503   TAGCGAGT  SA710   \n",
       "900221           1            1            B1  SA502   ACTATCTG  SA701   \n",
       "900570         301            4            F8  SA506   CGTGAGTG  SA708   \n",
       "900129         250            4            A1  SA501   ATCGTACG  SA701   \n",
       "900321         273            4            B4  SA502   ACTATCTG  SA704   \n",
       "...            ...          ...           ...    ...        ...    ...   \n",
       "900057           6            1            G1  SA507   GGATATCT  SA701   \n",
       "900294         134            2            A8  SB501   CTACTATA  SB708   \n",
       "900097         141            2            H8  SB508   GTCAGATA  SB708   \n",
       "900498         464            6           H11  SA508   GACACCGT  SB711   \n",
       "900406         260            4            D2  SA504   CTGCGTGT  SA702   \n",
       "\n",
       "       i7Sequence  identifier           Sequence      Plate ID  ...  \\\n",
       "900344   GTCTGCTA  SA710SA503  GTCTGCTA-TAGCGAGT  1.010000e+21  ...   \n",
       "900221   CGAGAGTT  SA701SA502  CGAGAGTT-ACTATCTG  1.010000e+21  ...   \n",
       "900570   CGAGCGAC  SA708SA506  CGAGCGAC-CGTGAGTG  1.010000e+21  ...   \n",
       "900129   CGAGAGTT  SA701SA501  CGAGAGTT-ATCGTACG  1.010000e+21  ...   \n",
       "900321   ACTCACTG  SA704SA502  ACTCACTG-ACTATCTG  1.010000e+21  ...   \n",
       "...           ...         ...                ...           ...  ...   \n",
       "900057   CGAGAGTT  SA701SA507  CGAGAGTT-GGATATCT  1.010000e+21  ...   \n",
       "900294   ATAGTACC  SB708SB501  ATAGTACC-CTACTATA  1.010000e+21  ...   \n",
       "900097   ATAGTACC  SB708SB508  ATAGTACC-GTCAGATA  1.010000e+21  ...   \n",
       "900498   AACGCTGA  SB711SA508  AACGCTGA-GACACCGT  1.010000e+21  ...   \n",
       "900406   GACATAGT  SA702SA504  GACATAGT-CTGCGTGT  1.010000e+21  ...   \n",
       "\n",
       "       enrolment_season  hiv_exposure hiv_status household_size o_scorad  \\\n",
       "900344           Spring           NaN        NaN            7.0      NaN   \n",
       "900221           Winter     Unexposed   negative            7.0     34.0   \n",
       "900570          Autumn      Unexposed   negative            4.0     36.0   \n",
       "900129           Winter     Unexposed   negative            6.0     44.0   \n",
       "900321           Spring           NaN        NaN            5.0      NaN   \n",
       "...                 ...           ...        ...            ...      ...   \n",
       "900057          Autumn      Unexposed   negative            9.0     33.0   \n",
       "900294           Winter     Unexposed   negative            4.0     23.0   \n",
       "900097           Winter     Unexposed   negative            6.0     44.0   \n",
       "900498              NaN           NaN        NaN            NaN      NaN   \n",
       "900406           Spring     Unexposed   negative            6.0     28.0   \n",
       "\n",
       "                                             FWD_filepath  \\\n",
       "900344  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "900221  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "900570  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "900129  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "900321  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "...                                                   ...   \n",
       "900057  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "900294  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "900097  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "900498  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "900406  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "\n",
       "                                             REV_filepath o_scorad_adj  \\\n",
       "900344  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...            0   \n",
       "900221  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...           34   \n",
       "900570  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...           36   \n",
       "900129  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...           44   \n",
       "900321  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...            0   \n",
       "...                                                   ...          ...   \n",
       "900057  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...           33   \n",
       "900294  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...           23   \n",
       "900097  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...           44   \n",
       "900498  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...            0   \n",
       "900406  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...           28   \n",
       "\n",
       "       individual_case         individual_case_location  \n",
       "900344          H skin  control-nonlesional skin Umtata  \n",
       "900221         AD skin        case-lesional skin Umtata  \n",
       "900570         AD skin  case-nonlesional skin Cape Town  \n",
       "900129         AD skin     case-lesional skin Cape Town  \n",
       "900321          H skin  control-nonlesional skin Umtata  \n",
       "...                ...                              ...  \n",
       "900057         AD skin     case-lesional skin Cape Town  \n",
       "900294         AD skin     case-nonlesional skin Umtata  \n",
       "900097         AD skin  case-nonlesional skin Cape Town  \n",
       "900498         AD skin     case-nonlesional skin Umtata  \n",
       "900406         AD skin     case-nonlesional skin Umtata  \n",
       "\n",
       "[262 rows x 35 columns]"
      ]
     },
     "execution_count": 656,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata_sub"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 657,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "skin_only_biom_path = '../Data/Tables/Absolute_Abundance_Tables/209766_filtered_feature_table_skin_only.biom'\n"
     ]
    }
   ],
   "source": [
    "# Create the path to the skin BIOM file by using the prefix of the original BIOM file\n",
    "skin_only_biom_path = biom_path.removesuffix('.biom') + '_skin_only.biom'\n",
    "\n",
    "# Print the path that was created\n",
    "print(f'{skin_only_biom_path = }')\n",
    "\n",
    "# Converting df to BIOM table and saving the biom table to the path specified above\n",
    "biom_tbl = convert_df_to_biom(skin_only_table, skin_only_biom_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 658,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PC1</th>\n",
       "      <th>PC2</th>\n",
       "      <th>PC3</th>\n",
       "      <th>individual_case_location</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>900344</th>\n",
       "      <td>-0.042492</td>\n",
       "      <td>-0.030903</td>\n",
       "      <td>0.047629</td>\n",
       "      <td>control-nonlesional skin Umtata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900221</th>\n",
       "      <td>0.044481</td>\n",
       "      <td>0.009170</td>\n",
       "      <td>0.111885</td>\n",
       "      <td>case-lesional skin Umtata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900570</th>\n",
       "      <td>0.047257</td>\n",
       "      <td>-0.107631</td>\n",
       "      <td>-0.033293</td>\n",
       "      <td>case-nonlesional skin Cape Town</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900129</th>\n",
       "      <td>0.065621</td>\n",
       "      <td>-0.016664</td>\n",
       "      <td>-0.012688</td>\n",
       "      <td>case-lesional skin Cape Town</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900321</th>\n",
       "      <td>-0.002474</td>\n",
       "      <td>0.003395</td>\n",
       "      <td>0.042491</td>\n",
       "      <td>control-nonlesional skin Umtata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900057</th>\n",
       "      <td>-0.040740</td>\n",
       "      <td>0.021282</td>\n",
       "      <td>-0.012640</td>\n",
       "      <td>case-lesional skin Cape Town</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900294</th>\n",
       "      <td>-0.036117</td>\n",
       "      <td>0.090884</td>\n",
       "      <td>0.039716</td>\n",
       "      <td>case-nonlesional skin Umtata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900097</th>\n",
       "      <td>0.027780</td>\n",
       "      <td>0.033129</td>\n",
       "      <td>-0.005649</td>\n",
       "      <td>case-nonlesional skin Cape Town</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900498</th>\n",
       "      <td>-0.034955</td>\n",
       "      <td>0.050211</td>\n",
       "      <td>0.028312</td>\n",
       "      <td>case-nonlesional skin Umtata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900406</th>\n",
       "      <td>-0.045348</td>\n",
       "      <td>0.055786</td>\n",
       "      <td>0.032997</td>\n",
       "      <td>case-nonlesional skin Umtata</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>262 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             PC1       PC2       PC3         individual_case_location\n",
       "900344 -0.042492 -0.030903  0.047629  control-nonlesional skin Umtata\n",
       "900221  0.044481  0.009170  0.111885        case-lesional skin Umtata\n",
       "900570  0.047257 -0.107631 -0.033293  case-nonlesional skin Cape Town\n",
       "900129  0.065621 -0.016664 -0.012688     case-lesional skin Cape Town\n",
       "900321 -0.002474  0.003395  0.042491  control-nonlesional skin Umtata\n",
       "...          ...       ...       ...                              ...\n",
       "900057 -0.040740  0.021282 -0.012640     case-lesional skin Cape Town\n",
       "900294 -0.036117  0.090884  0.039716     case-nonlesional skin Umtata\n",
       "900097  0.027780  0.033129 -0.005649  case-nonlesional skin Cape Town\n",
       "900498 -0.034955  0.050211  0.028312     case-nonlesional skin Umtata\n",
       "900406 -0.045348  0.055786  0.032997     case-nonlesional skin Umtata\n",
       "\n",
       "[262 rows x 4 columns]"
      ]
     },
     "execution_count": 658,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# perform RPCA with auto rank estimation\n",
    "np.seterr(divide = 'ignore')\n",
    "ordination, distance = rpca(biom_tbl)\n",
    "\n",
    "# extract and view sample ordinations from RPCA result\n",
    "spca_df = ordination.samples\n",
    "\n",
    "# Add a case type column into the spca_df using the meta_data and matching by indices by the join function\n",
    "spca_df = spca_df.join(metadata_sub['individual_case_location'])\n",
    "\n",
    "spca_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 659,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PC1</th>\n",
       "      <th>PC2</th>\n",
       "      <th>PC3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GTGCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTTATTGGGCGTAAAGCGAGCGCAGGCGGTTAGATAAGTCTGAAGTTAAAGGCTG</th>\n",
       "      <td>-0.180829</td>\n",
       "      <td>-0.367289</td>\n",
       "      <td>0.036368</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTGCCAGCAGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAATTATTGGGCGTAAAGCGCGCGTAGGCGGTTTTTTAAGTCTGATGTGAAAGCCCA</th>\n",
       "      <td>0.021655</td>\n",
       "      <td>-0.172791</td>\n",
       "      <td>-0.630078</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTGCCAGCAGCCGCGGTAATACGGAGGGTGCGAGCGTTAATCGGAATAACTGGGCGTAAAGGGCACGCAGGCGGTTATTTAAGTGAGGTGTGAAAGCCCC</th>\n",
       "      <td>-0.139940</td>\n",
       "      <td>-0.131270</td>\n",
       "      <td>0.078579</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTGCCAGCCGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAATTATTGGGCGTAAAGCGCGCGTAGGCGGTTTTTTAAGTCTGATGTGAAAGCCCA</th>\n",
       "      <td>0.003230</td>\n",
       "      <td>-0.066083</td>\n",
       "      <td>-0.594591</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTGCCAGCCGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTTATTGGGCGTAAAGCGAGCGCAGGCGGTTAGATAAGTCTGAAGTTAAAGGCTG</th>\n",
       "      <td>-0.206663</td>\n",
       "      <td>-0.263790</td>\n",
       "      <td>0.052879</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                         PC1       PC2  \\\n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATT... -0.180829 -0.367289   \n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAAT...  0.021655 -0.172791   \n",
       "GTGCCAGCAGCCGCGGTAATACGGAGGGTGCGAGCGTTAATCGGAAT... -0.139940 -0.131270   \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAAT...  0.003230 -0.066083   \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATT... -0.206663 -0.263790   \n",
       "\n",
       "                                                         PC3  \n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATT...  0.036368  \n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAAT... -0.630078  \n",
       "GTGCCAGCAGCCGCGGTAATACGGAGGGTGCGAGCGTTAATCGGAAT...  0.078579  \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAAT... -0.594591  \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATT...  0.052879  "
      ]
     },
     "execution_count": 659,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# extract and view feature ordinations from RPCA result\n",
    "fpca_df = ordination.features\n",
    "fpca_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 660,
   "metadata": {},
   "outputs": [],
   "source": [
    "def permanova_on_case_type_subset(df, dist_matrix, case_type_subset):\n",
    "    \"\"\"\n",
    "    Perform PERMANOVA on a subset of the data.\n",
    "    \n",
    "    Parameters:\n",
    "    - df: DataFrame with metadata, must include the grouping variable\n",
    "    - dist_matrix: DistanceMatrix object from scikit-bio\n",
    "    - case_type_subset: list of case_type groups to include in the test\n",
    "    \n",
    "    Returns:\n",
    "    - PERMANOVA result (dict-like with p-value, test statistic, etc.)\n",
    "    \"\"\"\n",
    "    # Subset the DataFrame\n",
    "    subset_df = df[df['individual_case_location'].isin(case_type_subset)]\n",
    "    print(subset_df)\n",
    "    group_counts = subset_df['individual_case_location'].value_counts()\n",
    "    print(\"Group counts:\", group_counts)\n",
    "\n",
    "    # Get the matching IDs and subset distance matrix\n",
    "    ids = subset_df.index\n",
    "    sub_dm = dist_matrix.filter(ids, strict=False)\n",
    "    \n",
    "    # Run PERMANOVA\n",
    "    result = permanova(sub_dm, grouping=subset_df['individual_case_location'], permutations=999)\n",
    "    \n",
    "    return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 661,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Subset case_type: ['control-nonlesional skin Cape Town', 'case-lesional skin Cape Town']\n",
      "                PC1       PC2       PC3            individual_case_location\n",
      "900129     0.065621 -0.016664 -0.012688        case-lesional skin Cape Town\n",
      "900145     0.109621  0.046086  0.016735  control-nonlesional skin Cape Town\n",
      "900544     0.060183 -0.041247 -0.058251  control-nonlesional skin Cape Town\n",
      "900110     0.041713  0.038921  0.001057        case-lesional skin Cape Town\n",
      "900600     0.017252  0.069338  0.043534  control-nonlesional skin Cape Town\n",
      "...             ...       ...       ...                                 ...\n",
      "Co005SNNL -0.066133  0.084223  0.032097  control-nonlesional skin Cape Town\n",
      "Ca008HNL  -0.071927  0.105776 -0.157505        case-lesional skin Cape Town\n",
      "900081    -0.056060 -0.031927  0.016984        case-lesional skin Cape Town\n",
      "900093     0.044290 -0.009791  0.037301        case-lesional skin Cape Town\n",
      "900057    -0.040740  0.021282 -0.012640        case-lesional skin Cape Town\n",
      "\n",
      "[73 rows x 4 columns]\n",
      "Group counts: individual_case_location\n",
      "case-lesional skin Cape Town          38\n",
      "control-nonlesional skin Cape Town    35\n",
      "Name: count, dtype: int64\n",
      "Subset case_type: ['control-nonlesional skin Umtata', 'case-lesional skin Umtata']\n",
      "             PC1       PC2       PC3         individual_case_location\n",
      "900344 -0.042492 -0.030903  0.047629  control-nonlesional skin Umtata\n",
      "900221  0.044481  0.009170  0.111885        case-lesional skin Umtata\n",
      "900321 -0.002474  0.003395  0.042491  control-nonlesional skin Umtata\n",
      "900245  0.087403 -0.028954  0.045911  control-nonlesional skin Umtata\n",
      "900423 -0.028452  0.037854  0.001508        case-lesional skin Umtata\n",
      "...          ...       ...       ...                              ...\n",
      "900442  0.063376 -0.090255  0.029798        case-lesional skin Umtata\n",
      "900261 -0.059272 -0.058751  0.027379        case-lesional skin Umtata\n",
      "900263  0.009042 -0.077582 -0.007309  control-nonlesional skin Umtata\n",
      "900279  0.009303  0.010737 -0.089601  control-nonlesional skin Umtata\n",
      "900287 -0.003427  0.015896  0.014158  control-nonlesional skin Umtata\n",
      "\n",
      "[97 rows x 4 columns]\n",
      "Group counts: individual_case_location\n",
      "control-nonlesional skin Umtata    52\n",
      "case-lesional skin Umtata          45\n",
      "Name: count, dtype: int64\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'Skin of Healthy vs. AD Children in Cape Town': {'p': '2.00e-03',\n",
       "  'f': '6.71'},\n",
       " 'Skin of Healthy vs. AD Children in Umtata': {'p': '1.00e-03', 'f': '23.57'}}"
      ]
     },
     "execution_count": 661,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# calculate permanova F-statistic for all combinations between Healthy, case lesional and case non lesional\n",
    "\n",
    "case_type_subsets = [\n",
    "    [\"control-nonlesional skin Cape Town\", \"case-lesional skin Cape Town\"],\n",
    "    [\"control-nonlesional skin Umtata\", \"case-lesional skin Umtata\"]]\n",
    "\n",
    "perma_res = {}\n",
    "\n",
    "for i, case_type_subset in enumerate(case_type_subsets):\n",
    "    print(\"Subset case_type:\", case_type_subset)\n",
    "    # print(\"Available sample IDs:\", spca_df.index.tolist())\n",
    "\n",
    "    result = permanova_on_case_type_subset(spca_df, distance, case_type_subset)\n",
    "    group_label = (\n",
    "        \"Skin of Healthy vs. AD Children in Cape Town\" if i == 0 else\n",
    "        \"Skin of Healthy vs. AD Children in Umtata\"\n",
    "    )\n",
    "    f_val = result[\"test statistic\"]\n",
    "    p_val = result[\"p-value\"]\n",
    "    \n",
    "    perma_res[group_label] = {\n",
    "        \"p\": f\"{p_val:.2e}\",\n",
    "        \"f\": f\"{f_val:.2f}\"\n",
    "    }\n",
    "\n",
    "perma_res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 662,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set the color palette for the groups in the correct order\n",
    "palette = {\n",
    "    'control-nonlesional skin Cape Town': '#A7C7E7',\n",
    "    'control-nonlesional skin Umtata': '#ADD8E6',\n",
    "    'case-lesional skin Cape Town': '#d2b48c',\n",
    "    'case-lesional skin Umtata': '#fa8072'\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 663,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create label map with (n=#)\n",
    "label_map = spca_df[\"individual_case_location\"].value_counts().to_dict()\n",
    "label_map = {k: f\"{k} (n={v})\" for k, v in label_map.items()}\n",
    "\n",
    "# Add labeled group column\n",
    "spca_df[\"group_label\"] = spca_df[\"individual_case_location\"].map(label_map)\n",
    "\n",
    "# Create figure\n",
    "fig, axes = plt.subplots(1, 2, figsize=(8, 5.5), sharex=True, sharey=True)\n",
    "\n",
    "# Define groups per panel\n",
    "split_map = {\n",
    "    \"Cape Town\": [\"control-nonlesional skin Cape Town\", \"case-lesional skin Cape Town\"],\n",
    "    \"Umtata\": [\"control-nonlesional skin Umtata\", \"case-lesional skin Umtata\"]\n",
    "}\n",
    "\n",
    "# Plot loop\n",
    "for ax, (title, groups) in zip(axes, split_map.items()):\n",
    "    subset_df = spca_df[spca_df[\"individual_case_location\"].isin(groups)].copy()\n",
    "    subset_df[\"group_label\"] = subset_df[\"individual_case_location\"].map(label_map)\n",
    "\n",
    "    ax.grid(False)\n",
    "\n",
    "    sns.scatterplot(\n",
    "        data=subset_df,\n",
    "        x=\"PC1\",\n",
    "        y=\"PC2\",\n",
    "        hue=\"group_label\",\n",
    "        hue_order=[label_map[g] for g in groups],\n",
    "        s=50,\n",
    "        edgecolor=\"black\",\n",
    "        linewidth=0.5,\n",
    "        palette={label_map[g]: palette[g] for g in groups},\n",
    "        ax=ax\n",
    "    )\n",
    "\n",
    "    # Draw 90% confidence circles\n",
    "    for case_type, case_type_df in subset_df.groupby(\"individual_case_location\"):\n",
    "        color = palette[case_type]\n",
    "        points = case_type_df[[\"PC1\", \"PC2\"]].values\n",
    "        centroid = points.mean(axis=0)\n",
    "        dists = np.linalg.norm(points - centroid, axis=1)\n",
    "        radius = np.percentile(dists, 90)\n",
    "        circle = Circle(centroid, radius, edgecolor=color, facecolor=color, alpha=0.2, lw=1, zorder=0)\n",
    "        ax.add_patch(circle)\n",
    "\n",
    "    ax.set_title(title, fontsize=16)\n",
    "    ax.set_xlabel(\"\")\n",
    "    ax.set_ylabel(\"\")\n",
    "    ax.legend(frameon=False, fontsize=9, loc='upper right')\n",
    "\n",
    "# === ADD % VARIANCE EXPLAINED TO AXIS LABELS ===\n",
    "pc1_var = ordination.proportion_explained['PC1'] * 100\n",
    "pc2_var = ordination.proportion_explained['PC2'] * 100\n",
    "\n",
    "axes[0].set_ylabel(f\"PC2 ({pc2_var:.1f}%)\", fontsize=14)\n",
    "for ax in axes:\n",
    "    ax.set_xlabel(f\"PC1 ({pc1_var:.1f}%)\", fontsize=14)\n",
    "\n",
    "\n",
    "axes[0].set_ylabel(f\"PC2 ({pc2_var:.1f}%)\", fontsize=14)\n",
    "for ax in axes:\n",
    "    ax.set_xlabel(f\"PC1 ({pc1_var:.1f}%)\", fontsize=14)\n",
    "    ax.set_xticks([-0.2, -0.1, 0.0, 0.1, 0.2, 0.3])\n",
    "    ax.set_xticklabels([-0.2, -0.1, 0.0, 0.1, 0.2, 0.3], fontsize=12)\n",
    "    ax.set_yticks([-0.2, -0.1, 0.0, 0.1, 0.2, 0.3])\n",
    "    ax.set_yticklabels([-0.2, -0.1, 0.0, 0.1, 0.2, 0.3], fontsize=12)\n",
    "\n",
    "# PERMANOVA stats\n",
    "axes[0].text(\n",
    "    0.04, -0.2,\n",
    "    f\"p={float(perma_res['Skin of Healthy vs. AD Children in Cape Town']['p']):.3f}, \"\n",
    "    f\"F={float(perma_res['Skin of Healthy vs. AD Children in Cape Town']['f']):.2f}\",\n",
    "    fontsize=12\n",
    ")\n",
    "axes[1].text(\n",
    "    0.04, -0.2,\n",
    "    f\"p={float(perma_res['Skin of Healthy vs. AD Children in Umtata']['p']):.3f}, \"\n",
    "    f\"F={float(perma_res['Skin of Healthy vs. AD Children in Umtata']['f']):.2f}\",\n",
    "    fontsize=12\n",
    ")\n",
    "\n",
    "# Final touches\n",
    "fig.suptitle(\"Beta Diversity by Region (SCORAD < 50)\", fontsize=18, y=0.93)\n",
    "plt.tight_layout(rect=[0, 0, 1, 0.95])\n",
    "plt.savefig(\"../Figures/Main/Fig_3C.png\", dpi=600)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "qiime2-metagenome-2024.10",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
