{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Beta Diversity (RPCA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import Python packages\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from numpy import array\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from matplotlib.gridspec import GridSpec\n",
    "import scipy\n",
    "import scipy.stats as ss\n",
    "from skbio.stats.distance import permanova\n",
    "import biom\n",
    "from biom import load_table\n",
    "from biom.util import biom_open\n",
    "from gemelli.rpca import rpca\n",
    "from matplotlib.patches import Circle\n",
    "from matplotlib.colors import to_hex\n",
    "import statsmodels.api as sm\n",
    "from skbio import DistanceMatrix\n",
    "\n",
    "### See RPCA standalone python tutorial: \n",
    "# https://github.com/biocore/gemelli/blob/master/ipynb/tutorials/RPCA-moving-pictures-standalone-cli-and-api.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to load BIOM table, collapse by taxa, sort rows by row sum, remove specified samples, and convert to relative abundance\n",
    "def load_biom_table(biom_path, metadata_path):\n",
    "    # Load BIOM table and convert to a DataFrame\n",
    "    table = biom.load_table(biom_path)\n",
    "    df = pd.DataFrame(table.matrix_data.toarray(),\n",
    "                      index=table.ids(axis='observation'),\n",
    "                      columns=table.ids(axis='sample'))\n",
    "    \n",
    "    # Sort rows by row sum in descending order\n",
    "    df['row_sum'] = df.sum(axis=1)\n",
    "    df = df.sort_values(by='row_sum', ascending=False)\n",
    "    df = df.drop(columns=['row_sum'])\n",
    "\n",
    "    # Replace ' g__' rows with ' g__Unknown'\n",
    "    df.index = df.index.map(lambda x: ' g__Unknown' if x == ' g__' else x)\n",
    "\n",
    "    # Remove '15564.' prefix from columns\n",
    "    df.columns = df.columns.str.replace('15564.', '')\n",
    "\n",
    "    # Load metadata\n",
    "    metadata = pd.read_csv(metadata_path, sep='\\t')\n",
    "    metadata['#sample-id'] = metadata['#sample-id'].str.replace('_', '')\n",
    "\n",
    "    # Set Sample-ID as index\n",
    "    metadata = metadata.set_index('#sample-id')\n",
    "\n",
    "    # Subset metadata to only samples in the BIOM table\n",
    "    metadata_sub = metadata.loc[df.columns]\n",
    "\n",
    "    # Get a list of the nares samples\n",
    "    nares_samples = metadata_sub[metadata_sub['case_type'].str.endswith('nares')].index.tolist()\n",
    "\n",
    "    # Drop nares samples from BIOM table\n",
    "    df = df.drop(columns=nares_samples)\n",
    "\n",
    "    # Recompute metadata_sub after nares removal\n",
    "    metadata_sub = metadata.loc[df.columns]\n",
    "\n",
    "    # Compute individual_case_location\n",
    "    metadata['individual_case_location'] = metadata['case_type'] + ' ' + metadata['area']\n",
    "\n",
    "    # Convert SCORAD and filter only AD samples with SCORAD > 40\n",
    "    metadata['o_scorad'] = pd.to_numeric(metadata['o_scorad'], errors='coerce')\n",
    "\n",
    "    # Keep healthy samples (NaN SCORAD) and AD samples with SCORAD ≤ 40\n",
    "    # metadata = metadata[(metadata['o_scorad'].isna()) | (metadata['o_scorad'] <= 40)]\n",
    "    metadata = metadata[(metadata['o_scorad'].isna()) | (metadata['o_scorad'] <= 200)] # ALL SCORAD\n",
    "\n",
    "    # Subset both df and metadata to overlapping samples\n",
    "    shared_samples = df.columns.intersection(metadata.index)\n",
    "    df = df[shared_samples]\n",
    "    metadata_sub = metadata.loc[shared_samples]\n",
    "\n",
    "    return df, metadata_sub\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# converts the subsetted skin only BIOM table DataFrame back into a BIOM to run RPCA\n",
    "def convert_df_to_biom(table, biom_output_file):\n",
    "    obs_ids = table.index\n",
    "    samp_ids = table.columns\n",
    "    biom_table = biom.table.Table(table.values, observation_ids=obs_ids, sample_ids=samp_ids) # Convert df back to a biom table\n",
    "    #biom_output_file = f\"../tables/{name}_subset.biom\" # Path to output file\n",
    "\n",
    "    with biom_open(biom_output_file, 'w') as f: # This does the actual saving!\n",
    "        biom_table.to_hdf5(f, generated_by=\"subsetted tables\")\n",
    "\n",
    "    return biom_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>900344</th>\n",
       "      <th>900221</th>\n",
       "      <th>900570</th>\n",
       "      <th>900129</th>\n",
       "      <th>900321</th>\n",
       "      <th>900091</th>\n",
       "      <th>900245</th>\n",
       "      <th>900423</th>\n",
       "      <th>900581</th>\n",
       "      <th>900145</th>\n",
       "      <th>...</th>\n",
       "      <th>900094</th>\n",
       "      <th>900287</th>\n",
       "      <th>900225</th>\n",
       "      <th>900057</th>\n",
       "      <th>900294</th>\n",
       "      <th>9003972</th>\n",
       "      <th>900097</th>\n",
       "      <th>900498</th>\n",
       "      <th>900276</th>\n",
       "      <th>900406</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GTGCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTTATTGGGCGTAAAGCGAGCGCAGGCGGTTAGATAAGTCTGAAGTTAAAGGCTG</th>\n",
       "      <td>984.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>389.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>198.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>57.0</td>\n",
       "      <td>44.0</td>\n",
       "      <td>111.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>1168.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTGCCAGCAGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAATTATTGGGCGTAAAGCGCGCGTAGGCGGTTTTTTAAGTCTGATGTGAAAGCCCA</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1294.0</td>\n",
       "      <td>182.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>296.0</td>\n",
       "      <td>159.0</td>\n",
       "      <td>49.0</td>\n",
       "      <td>1366.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>108.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>93.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>794.0</td>\n",
       "      <td>171.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTGCCAGCAGCCGCGGTAATACGGAGGGTGCGAGCGTTAATCGGAATAACTGGGCGTAAAGGGCACGCAGGCGGTTATTTAAGTGAGGTGTGAAAGCCCC</th>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTGCCAGCCGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAATTATTGGGCGTAAAGCGCGCGTAGGCGGTTTTTTAAGTCTGATGTGAAAGCCCA</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>236.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>57.0</td>\n",
       "      <td>799.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>116.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>477.0</td>\n",
       "      <td>94.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTGCCAGCCGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTTATTGGGCGTAAAGCGAGCGCAGGCGGTTAGATAAGTCTGAAGTTAAAGGCTG</th>\n",
       "      <td>611.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>593.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTGCCAGCAGCCGCGGTAATACGGAGGGTCCGAGCGTTATCCGGAATCATTGGGTTTAAAGGGTCCGTAGGCGGCTTTATAAGTCAGTGGTGAAATCCGG</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTGCCAGCAGCCGCGGTAATACGTAGGGTGCAAGCGTTGTCCGGAATTATTGGGCGTAAAGAGCTCGTAGGCGGTTTGTCGCGTCTGCTGTGAAATTCCG</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTGCCAGCCGCCGCGGTAATACGTAGGGCGCAAGCGTTGTCCGGAATTATTGGGCGTAAAGAGCTCGTAGGCGGTTTGTCGCGTCTGGTGTGAAAACTCG</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTGCCAGAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTTATTGGGCGTAAAGCGAGCGCAGGCGGTTAGATAAGTCTGAAGTTAAAGGCTGT</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTTATTGGGCGTAAAGCGAGCGCAGGCGGTTAGATAAGTCTGAAGTTAAAGGCTGT</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2283 rows × 305 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    900344  900221  900570  \\\n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATT...   984.0    22.0   389.0   \n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAAT...     0.0     0.0  1294.0   \n",
       "GTGCCAGCAGCCGCGGTAATACGGAGGGTGCGAGCGTTAATCGGAAT...     0.0     2.0     9.0   \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAAT...     0.0     0.0   236.0   \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATT...   611.0     0.0     0.0   \n",
       "...                                                    ...     ...     ...   \n",
       "GTGCCAGCAGCCGCGGTAATACGGAGGGTCCGAGCGTTATCCGGAAT...     0.0     0.0     0.0   \n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGGTGCAAGCGTTGTCCGGAAT...     0.0     0.0     0.0   \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGGCGCAAGCGTTGTCCGGAAT...     0.0     0.0     0.0   \n",
       "GTGCCAGAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTT...     0.0     0.0     0.0   \n",
       "GTCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTT...     0.0     0.0     0.0   \n",
       "\n",
       "                                                    900129  900321  900091  \\\n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATT...     0.0    26.0    27.0   \n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAAT...   182.0     9.0   296.0   \n",
       "GTGCCAGCAGCCGCGGTAATACGGAGGGTGCGAGCGTTAATCGGAAT...     0.0     0.0     0.0   \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAAT...     0.0     0.0     0.0   \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATT...     0.0     0.0    16.0   \n",
       "...                                                    ...     ...     ...   \n",
       "GTGCCAGCAGCCGCGGTAATACGGAGGGTCCGAGCGTTATCCGGAAT...     0.0     0.0     0.0   \n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGGTGCAAGCGTTGTCCGGAAT...     0.0     0.0     0.0   \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGGCGCAAGCGTTGTCCGGAAT...     0.0     0.0     0.0   \n",
       "GTGCCAGAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTT...     0.0     0.0     0.0   \n",
       "GTCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTT...     0.0     0.0     0.0   \n",
       "\n",
       "                                                    900245  900423  900581  \\\n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATT...    40.0     0.0   198.0   \n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAAT...   159.0    49.0  1366.0   \n",
       "GTGCCAGCAGCCGCGGTAATACGGAGGGTGCGAGCGTTAATCGGAAT...     0.0     0.0    22.0   \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAAT...     0.0    57.0   799.0   \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATT...     0.0     0.0     0.0   \n",
       "...                                                    ...     ...     ...   \n",
       "GTGCCAGCAGCCGCGGTAATACGGAGGGTCCGAGCGTTATCCGGAAT...     0.0     0.0     0.0   \n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGGTGCAAGCGTTGTCCGGAAT...     0.0     0.0     0.0   \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGGCGCAAGCGTTGTCCGGAAT...     0.0     0.0     0.0   \n",
       "GTGCCAGAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTT...     0.0     0.0     0.0   \n",
       "GTCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTT...     0.0     0.0     0.0   \n",
       "\n",
       "                                                    900145  ...  900094  \\\n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATT...     0.0  ...     0.0   \n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAAT...    27.0  ...     0.0   \n",
       "GTGCCAGCAGCCGCGGTAATACGGAGGGTGCGAGCGTTAATCGGAAT...     0.0  ...     0.0   \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAAT...    37.0  ...     0.0   \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATT...     6.0  ...     0.0   \n",
       "...                                                    ...  ...     ...   \n",
       "GTGCCAGCAGCCGCGGTAATACGGAGGGTCCGAGCGTTATCCGGAAT...     0.0  ...     0.0   \n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGGTGCAAGCGTTGTCCGGAAT...     0.0  ...     0.0   \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGGCGCAAGCGTTGTCCGGAAT...     0.0  ...     0.0   \n",
       "GTGCCAGAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTT...     0.0  ...     0.0   \n",
       "GTCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTT...     0.0  ...     0.0   \n",
       "\n",
       "                                                    900287  900225  900057  \\\n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATT...    57.0    44.0   111.0   \n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAAT...   108.0    30.0    93.0   \n",
       "GTGCCAGCAGCCGCGGTAATACGGAGGGTGCGAGCGTTAATCGGAAT...     2.0     0.0     0.0   \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAAT...     0.0     4.0   116.0   \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATT...     0.0     0.0     0.0   \n",
       "...                                                    ...     ...     ...   \n",
       "GTGCCAGCAGCCGCGGTAATACGGAGGGTCCGAGCGTTATCCGGAAT...     0.0     0.0     0.0   \n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGGTGCAAGCGTTGTCCGGAAT...     0.0     0.0     0.0   \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGGCGCAAGCGTTGTCCGGAAT...     0.0     0.0     0.0   \n",
       "GTGCCAGAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTT...     0.0     0.0     0.0   \n",
       "GTCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTT...     0.0     0.0     0.0   \n",
       "\n",
       "                                                    900294  9003972  900097  \\\n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATT...    12.0   1168.0    24.0   \n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAAT...    13.0    794.0   171.0   \n",
       "GTGCCAGCAGCCGCGGTAATACGGAGGGTGCGAGCGTTAATCGGAAT...     0.0     74.0     0.0   \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAAT...     0.0    477.0    94.0   \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATT...     7.0    593.0     0.0   \n",
       "...                                                    ...      ...     ...   \n",
       "GTGCCAGCAGCCGCGGTAATACGGAGGGTCCGAGCGTTATCCGGAAT...     0.0      0.0     0.0   \n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGGTGCAAGCGTTGTCCGGAAT...     0.0      0.0     0.0   \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGGCGCAAGCGTTGTCCGGAAT...     0.0      0.0     0.0   \n",
       "GTGCCAGAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTT...     0.0      0.0     0.0   \n",
       "GTCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTT...     0.0      0.0     0.0   \n",
       "\n",
       "                                                    900498  900276  900406  \n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATT...    15.0     0.0     0.0  \n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAAT...    28.0     0.0    12.0  \n",
       "GTGCCAGCAGCCGCGGTAATACGGAGGGTGCGAGCGTTAATCGGAAT...     0.0     0.0     0.0  \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAAT...    11.0     0.0     7.0  \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATT...    17.0     0.0     0.0  \n",
       "...                                                    ...     ...     ...  \n",
       "GTGCCAGCAGCCGCGGTAATACGGAGGGTCCGAGCGTTATCCGGAAT...     0.0     0.0     0.0  \n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGGTGCAAGCGTTGTCCGGAAT...     0.0     0.0     0.0  \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGGCGCAAGCGTTGTCCGGAAT...     0.0     0.0     0.0  \n",
       "GTGCCAGAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTT...     0.0     0.0     0.0  \n",
       "GTCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTT...     0.0     1.0     0.0  \n",
       "\n",
       "[2283 rows x 305 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# specify the path to the BIOM table and Metadata\n",
    "biom_path = \"../Data/Tables/Absolute_Abundance_Tables/209766_filtered_feature_table.biom\"\n",
    "metadata_path = '../Data/Metadata/updated_clean_ant_skin_metadata_microbiome_type-location.tab'\n",
    "skin_only_table, metadata_sub = load_biom_table(biom_path, metadata_path)\n",
    "\n",
    "skin_only_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "skin_only_biom_path = '../Data/Tables/Absolute_Abundance_Tables/209766_filtered_feature_table_skin_only.biom'\n"
     ]
    }
   ],
   "source": [
    "# Create the path to the skin BIOM file by using the prefix of the original BIOM file\n",
    "skin_only_biom_path = biom_path.removesuffix('.biom') + '_skin_only.biom'\n",
    "\n",
    "# Print the path that was created\n",
    "print(f'{skin_only_biom_path = }')\n",
    "\n",
    "# Converting df to BIOM table and saving the biom table to the path specified above\n",
    "biom_tbl = convert_df_to_biom(skin_only_table, skin_only_biom_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PC1</th>\n",
       "      <th>PC2</th>\n",
       "      <th>PC3</th>\n",
       "      <th>individual_case_location</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>900344</th>\n",
       "      <td>-0.031259</td>\n",
       "      <td>0.030585</td>\n",
       "      <td>-0.045341</td>\n",
       "      <td>control-nonlesional skin Umtata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900221</th>\n",
       "      <td>0.034560</td>\n",
       "      <td>-0.014311</td>\n",
       "      <td>-0.101067</td>\n",
       "      <td>case-lesional skin Umtata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900570</th>\n",
       "      <td>0.051177</td>\n",
       "      <td>0.101247</td>\n",
       "      <td>0.020742</td>\n",
       "      <td>case-nonlesional skin Cape Town</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900129</th>\n",
       "      <td>0.061599</td>\n",
       "      <td>0.018120</td>\n",
       "      <td>0.002876</td>\n",
       "      <td>case-lesional skin Cape Town</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900321</th>\n",
       "      <td>0.002420</td>\n",
       "      <td>-0.003792</td>\n",
       "      <td>-0.040013</td>\n",
       "      <td>control-nonlesional skin Umtata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9003972</th>\n",
       "      <td>-0.107095</td>\n",
       "      <td>0.089850</td>\n",
       "      <td>0.019500</td>\n",
       "      <td>case-lesional skin Umtata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900097</th>\n",
       "      <td>0.025972</td>\n",
       "      <td>-0.031322</td>\n",
       "      <td>0.000608</td>\n",
       "      <td>case-nonlesional skin Cape Town</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900498</th>\n",
       "      <td>-0.033348</td>\n",
       "      <td>-0.048009</td>\n",
       "      <td>-0.030825</td>\n",
       "      <td>case-nonlesional skin Umtata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900276</th>\n",
       "      <td>-0.108307</td>\n",
       "      <td>0.102284</td>\n",
       "      <td>-0.005354</td>\n",
       "      <td>case-lesional skin Umtata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900406</th>\n",
       "      <td>-0.041063</td>\n",
       "      <td>-0.053825</td>\n",
       "      <td>-0.033240</td>\n",
       "      <td>case-nonlesional skin Umtata</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>305 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              PC1       PC2       PC3         individual_case_location\n",
       "900344  -0.031259  0.030585 -0.045341  control-nonlesional skin Umtata\n",
       "900221   0.034560 -0.014311 -0.101067        case-lesional skin Umtata\n",
       "900570   0.051177  0.101247  0.020742  case-nonlesional skin Cape Town\n",
       "900129   0.061599  0.018120  0.002876     case-lesional skin Cape Town\n",
       "900321   0.002420 -0.003792 -0.040013  control-nonlesional skin Umtata\n",
       "...           ...       ...       ...                              ...\n",
       "9003972 -0.107095  0.089850  0.019500        case-lesional skin Umtata\n",
       "900097   0.025972 -0.031322  0.000608  case-nonlesional skin Cape Town\n",
       "900498  -0.033348 -0.048009 -0.030825     case-nonlesional skin Umtata\n",
       "900276  -0.108307  0.102284 -0.005354        case-lesional skin Umtata\n",
       "900406  -0.041063 -0.053825 -0.033240     case-nonlesional skin Umtata\n",
       "\n",
       "[305 rows x 4 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# perform RPCA with auto rank estimation\n",
    "np.seterr(divide = 'ignore')\n",
    "ordination, distance = rpca(biom_tbl)\n",
    "\n",
    "# extract and view sample ordinations from RPCA result\n",
    "spca_df = ordination.samples\n",
    "\n",
    "# Add a case type column into the spca_df using the meta_data and matching by indices by the join function\n",
    "spca_df = spca_df.join(metadata_sub['individual_case_location'])\n",
    "\n",
    "spca_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PC1</th>\n",
       "      <th>PC2</th>\n",
       "      <th>PC3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GTGCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTTATTGGGCGTAAAGCGAGCGCAGGCGGTTAGATAAGTCTGAAGTTAAAGGCTG</th>\n",
       "      <td>-0.176233</td>\n",
       "      <td>0.363698</td>\n",
       "      <td>-0.052031</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTGCCAGCAGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAATTATTGGGCGTAAAGCGCGCGTAGGCGGTTTTTTAAGTCTGATGTGAAAGCCCA</th>\n",
       "      <td>0.031293</td>\n",
       "      <td>0.181160</td>\n",
       "      <td>0.635457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTGCCAGCAGCCGCGGTAATACGGAGGGTGCGAGCGTTAATCGGAATAACTGGGCGTAAAGGGCACGCAGGCGGTTATTTAAGTGAGGTGTGAAAGCCCC</th>\n",
       "      <td>-0.137862</td>\n",
       "      <td>0.125219</td>\n",
       "      <td>-0.075513</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTGCCAGCCGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAATTATTGGGCGTAAAGCGCGCGTAGGCGGTTTTTTAAGTCTGATGTGAAAGCCCA</th>\n",
       "      <td>0.018601</td>\n",
       "      <td>0.079648</td>\n",
       "      <td>0.607929</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTGCCAGCCGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTTATTGGGCGTAAAGCGAGCGCAGGCGGTTAGATAAGTCTGAAGTTAAAGGCTG</th>\n",
       "      <td>-0.200034</td>\n",
       "      <td>0.269834</td>\n",
       "      <td>-0.054126</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                         PC1       PC2  \\\n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATT... -0.176233  0.363698   \n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAAT...  0.031293  0.181160   \n",
       "GTGCCAGCAGCCGCGGTAATACGGAGGGTGCGAGCGTTAATCGGAAT... -0.137862  0.125219   \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAAT...  0.018601  0.079648   \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATT... -0.200034  0.269834   \n",
       "\n",
       "                                                         PC3  \n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATT... -0.052031  \n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAAT...  0.635457  \n",
       "GTGCCAGCAGCCGCGGTAATACGGAGGGTGCGAGCGTTAATCGGAAT... -0.075513  \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAAT...  0.607929  \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATT... -0.054126  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# extract and view feature ordinations from RPCA result\n",
    "fpca_df = ordination.features\n",
    "fpca_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def permanova_on_case_type_subset(df, dist_matrix, case_type_subset):\n",
    "    \"\"\"\n",
    "    Perform PERMANOVA on a subset of the data.\n",
    "    \n",
    "    Parameters:\n",
    "    - df: DataFrame with metadata, must include the grouping variable\n",
    "    - dist_matrix: DistanceMatrix object from scikit-bio\n",
    "    - case_type_subset: list of case_type groups to include in the test\n",
    "    \n",
    "    Returns:\n",
    "    - PERMANOVA result (dict-like with p-value, test statistic, etc.)\n",
    "    \"\"\"\n",
    "    # Subset the DataFrame\n",
    "    subset_df = df[df['individual_case_location'].isin(case_type_subset)]\n",
    "    print(subset_df)\n",
    "    group_counts = subset_df['individual_case_location'].value_counts()\n",
    "    print(\"Group counts:\", group_counts)\n",
    "\n",
    "    # Get the matching IDs and subset distance matrix\n",
    "    ids = subset_df.index\n",
    "    sub_dm = dist_matrix.filter(ids, strict=False)\n",
    "    \n",
    "    # Run PERMANOVA\n",
    "    result = permanova(sub_dm, grouping=subset_df['individual_case_location'], permutations=999)\n",
    "    \n",
    "    return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Subset case_type: ['control-nonlesional skin Cape Town', 'case-lesional skin Cape Town']\n",
      "                PC1       PC2       PC3            individual_case_location\n",
      "900129     0.061599  0.018120  0.002876        case-lesional skin Cape Town\n",
      "900581    -0.007051  0.063078  0.054416        case-lesional skin Cape Town\n",
      "900145     0.115958 -0.038398 -0.038217  control-nonlesional skin Cape Town\n",
      "900544     0.053323  0.045220  0.041756  control-nonlesional skin Cape Town\n",
      "900110     0.038821 -0.045479 -0.017278        case-lesional skin Cape Town\n",
      "...             ...       ...       ...                                 ...\n",
      "Co005SNNL -0.056991 -0.078442 -0.030159  control-nonlesional skin Cape Town\n",
      "Ca008HNL  -0.061534 -0.089716  0.146310        case-lesional skin Cape Town\n",
      "900081    -0.050282  0.029005 -0.022752        case-lesional skin Cape Town\n",
      "900093     0.036764  0.008035 -0.035655        case-lesional skin Cape Town\n",
      "900057    -0.037138 -0.021089  0.008198        case-lesional skin Cape Town\n",
      "\n",
      "[81 rows x 4 columns]\n",
      "Group counts: individual_case_location\n",
      "case-lesional skin Cape Town          46\n",
      "control-nonlesional skin Cape Town    35\n",
      "Name: count, dtype: int64\n",
      "Subset case_type: ['control-nonlesional skin Umtata', 'case-lesional skin Umtata']\n",
      "              PC1       PC2       PC3         individual_case_location\n",
      "900344  -0.031259  0.030585 -0.045341  control-nonlesional skin Umtata\n",
      "900221   0.034560 -0.014311 -0.101067        case-lesional skin Umtata\n",
      "900321   0.002420 -0.003792 -0.040013  control-nonlesional skin Umtata\n",
      "900245   0.082882  0.023124 -0.044768  control-nonlesional skin Umtata\n",
      "900423  -0.033048 -0.030774 -0.006261        case-lesional skin Umtata\n",
      "...           ...       ...       ...                              ...\n",
      "900263   0.010123  0.068705 -0.006843  control-nonlesional skin Umtata\n",
      "900279   0.010248 -0.006841  0.075152  control-nonlesional skin Umtata\n",
      "900287  -0.002540 -0.015663 -0.017113  control-nonlesional skin Umtata\n",
      "9003972 -0.107095  0.089850  0.019500        case-lesional skin Umtata\n",
      "900276  -0.108307  0.102284 -0.005354        case-lesional skin Umtata\n",
      "\n",
      "[113 rows x 4 columns]\n",
      "Group counts: individual_case_location\n",
      "case-lesional skin Umtata          61\n",
      "control-nonlesional skin Umtata    52\n",
      "Name: count, dtype: int64\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'Skin of Healthy vs. AD Children in Cape Town': {'p': '2.00e-03',\n",
       "  'f': '6.20'},\n",
       " 'Skin of Healthy vs. AD Children in Umtata': {'p': '1.00e-03', 'f': '25.72'}}"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# calculate permanova F-statistic for all combinations between Healthy, case lesional and case non lesional\n",
    "\n",
    "case_type_subsets = [\n",
    "    [\"control-nonlesional skin Cape Town\", \"case-lesional skin Cape Town\"],\n",
    "    [\"control-nonlesional skin Umtata\", \"case-lesional skin Umtata\"]]\n",
    "\n",
    "perma_res = {}\n",
    "\n",
    "for i, case_type_subset in enumerate(case_type_subsets):\n",
    "    print(\"Subset case_type:\", case_type_subset)\n",
    "    # print(\"Available sample IDs:\", spca_df.index.tolist())\n",
    "\n",
    "    result = permanova_on_case_type_subset(spca_df, distance, case_type_subset)\n",
    "    group_label = (\n",
    "        \"Skin of Healthy vs. AD Children in Cape Town\" if i == 0 else\n",
    "        \"Skin of Healthy vs. AD Children in Umtata\"\n",
    "    )\n",
    "    f_val = result[\"test statistic\"]\n",
    "    p_val = result[\"p-value\"]\n",
    "    \n",
    "    perma_res[group_label] = {\n",
    "        \"p\": f\"{p_val:.2e}\",\n",
    "        \"f\": f\"{f_val:.2f}\"\n",
    "    }\n",
    "\n",
    "perma_res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set the color palette for the groups in the correct order\n",
    "palette = {\n",
    "    'control-nonlesional skin Cape Town': '#7FBCEB',\n",
    "    'control-nonlesional skin Umtata': '#66C2EE',\n",
    "    'case-nonlesional skin Cape Town': '#FAD5A5',\n",
    "    'case-nonlesional skin Umtata': '#FAD5A5',\n",
    "    'case-lesional skin Cape Town': '#cd853f',\n",
    "    'case-lesional skin Umtata': '#fa8072'\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Prepare labels\n",
    "group_short_labels = {\n",
    "    \"case-lesional skin Umtata\": \"UM-ADL\",\n",
    "    \"case-lesional skin Cape Town\": \"CT-ADL\",\n",
    "    \"case-nonlesional skin Umtata\": \"UM-ADNL\",\n",
    "    \"case-nonlesional skin Cape Town\": \"CT-ADNL\",\n",
    "    \"control-nonlesional skin Umtata\": \"UM-H\",\n",
    "    \"control-nonlesional skin Cape Town\": \"CT-H\"\n",
    "}\n",
    "\n",
    "# Add short label column\n",
    "spca_df[\"short_label\"] = spca_df[\"individual_case_location\"].map(group_short_labels)\n",
    "\n",
    "# Create label map with (n=#)\n",
    "label_map = spca_df[\"short_label\"].value_counts().to_dict()\n",
    "label_map = {k: f\"{k} (n={v})\" for k, v in label_map.items()}\n",
    "spca_df[\"group_label\"] = spca_df[\"short_label\"].map(label_map)\n",
    "\n",
    "# Define PC variance\n",
    "pc1_var = ordination.proportion_explained['PC1'] * 100\n",
    "pc2_var = ordination.proportion_explained['PC2'] * 100\n",
    "\n",
    "# ===== CAPE TOWN FIGURE =====\n",
    "ct_full_groups = [\"control-nonlesional skin Cape Town\", \"case-nonlesional skin Cape Town\", \"case-lesional skin Cape Town\"]\n",
    "ct_subset = spca_df[spca_df[\"individual_case_location\"].isin(ct_full_groups)].copy()\n",
    "\n",
    "fig_ct, ax_ct = plt.subplots(figsize=(5.5, 5.5))\n",
    "sns.scatterplot(\n",
    "    data=ct_subset,\n",
    "    x=\"PC1\", y=\"PC2\",\n",
    "    hue=\"group_label\",\n",
    "    hue_order=[label_map[group_short_labels[g]] for g in ct_full_groups],\n",
    "    s=50, edgecolor=\"black\", linewidth=0.5,\n",
    "    palette={label_map[group_short_labels[g]]: palette[g] for g in ct_full_groups},\n",
    "    ax=ax_ct\n",
    ")\n",
    "\n",
    "# Confidence ellipses\n",
    "for group in ct_full_groups:\n",
    "    df = ct_subset[ct_subset[\"individual_case_location\"] == group]\n",
    "    color = palette[group]\n",
    "    pts = df[[\"PC1\", \"PC2\"]].values\n",
    "    center = pts.mean(axis=0)\n",
    "    radius = np.percentile(np.linalg.norm(pts - center, axis=1), 90)\n",
    "    ax_ct.add_patch(Circle(center, radius, edgecolor=color, facecolor=color, alpha=0.2, lw=1, zorder=0))\n",
    "\n",
    "ax_ct.set_xlabel(f\"RPCA PC1 ({pc1_var:.1f}%)\", fontsize=16)\n",
    "ax_ct.set_ylabel(f\"RPCA PC2 ({pc2_var:.1f}%)\", fontsize=16)\n",
    "ax_ct.set_xticks([-0.2, -0.1, 0.0, 0.1, 0.2, 0.3])\n",
    "ax_ct.set_yticks([-0.2, -0.1, 0.0, 0.1, 0.2, 0.3])\n",
    "ax_ct.tick_params(labelsize=12)\n",
    "ax_ct.legend(frameon=False, fontsize=12, loc='upper right')\n",
    "ax_ct.set_title(\"Cape Town (urban)\", fontsize=20)\n",
    "ax_ct.text(\n",
    "    0.04, -0.17,\n",
    "    f\"p={float(perma_res['Skin of Healthy vs. AD Children in Cape Town']['p']):.3f}, \"\n",
    "    f\"F={float(perma_res['Skin of Healthy vs. AD Children in Cape Town']['f']):.2f}\",\n",
    "    fontsize=12\n",
    ")\n",
    "\n",
    "plt.tight_layout()\n",
    "fig_ct.savefig(\"../Figures/Supplementary/Suppl_Fig_3D.png\", dpi=600)\n",
    "fig_ct.savefig(\"../Figures/Main/Fig_2G.png\", dpi=600)\n",
    "\n",
    "# ===== UMTATA FIGURE =====\n",
    "um_full_groups = [\"control-nonlesional skin Umtata\", \"case-nonlesional skin Umtata\", \"case-lesional skin Umtata\"]\n",
    "um_subset = spca_df[spca_df[\"individual_case_location\"].isin(um_full_groups)].copy()\n",
    "\n",
    "fig_um, ax_um = plt.subplots(figsize=(5.5, 5.5))\n",
    "sns.scatterplot(\n",
    "    data=um_subset,\n",
    "    x=\"PC1\", y=\"PC2\",\n",
    "    hue=\"group_label\",\n",
    "    hue_order=[label_map[group_short_labels[g]] for g in um_full_groups],\n",
    "    s=50, edgecolor=\"black\", linewidth=0.5,\n",
    "    palette={label_map[group_short_labels[g]]: palette[g] for g in um_full_groups},\n",
    "    ax=ax_um\n",
    ")\n",
    "\n",
    "# Confidence ellipses\n",
    "for group in um_full_groups:\n",
    "    df = um_subset[um_subset[\"individual_case_location\"] == group]\n",
    "    color = palette[group]\n",
    "    pts = df[[\"PC1\", \"PC2\"]].values\n",
    "    center = pts.mean(axis=0)\n",
    "    radius = np.percentile(np.linalg.norm(pts - center, axis=1), 90)\n",
    "    ax_um.add_patch(Circle(center, radius, edgecolor=color, facecolor=color, alpha=0.2, lw=1, zorder=0))\n",
    "\n",
    "ax_um.set_xlabel(f\"RPCA PC1 ({pc1_var:.1f}%)\", fontsize=16)\n",
    "ax_um.set_ylabel(f\"RPCA PC2 ({pc2_var:.1f}%)\", fontsize=16)\n",
    "ax_um.set_xticks([-0.2, -0.1, 0.0, 0.1, 0.2, 0.3])\n",
    "ax_um.set_yticks([-0.2, -0.1, 0.0, 0.1, 0.2, 0.3])\n",
    "ax_um.tick_params(labelsize=12)\n",
    "ax_um.legend(frameon=False, fontsize=12, loc='upper right')\n",
    "ax_um.set_title(\"Umtata (rural)\", fontsize=20)\n",
    "ax_um.text(\n",
    "    0.02, -0.22,\n",
    "    # 0.02, -0.18,\n",
    "    f\"p={float(perma_res['Skin of Healthy vs. AD Children in Umtata']['p']):.3f}, \"\n",
    "    f\"F={float(perma_res['Skin of Healthy vs. AD Children in Umtata']['f']):.2f}\",\n",
    "    fontsize=12\n",
    ")\n",
    "\n",
    "plt.tight_layout()\n",
    "fig_um.savefig(\"../Figures/Supplementary/Suppl_Fig_3B.png\", dpi=600)\n",
    "# fig_um.savefig(\"../Figures/Main/Fig_2E.png\", dpi=600)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'UM-ADL': 61, 'CT-ADL': 46}\n"
     ]
    }
   ],
   "source": [
    "# Filter only ADL samples from both regions\n",
    "adl_groups = [\n",
    "    \"case-lesional skin Cape Town\",\n",
    "    \"case-lesional skin Umtata\"\n",
    "]\n",
    "adl_subset = spca_df[spca_df[\"individual_case_location\"].isin(adl_groups)].copy()\n",
    "\n",
    "# Prepare labels\n",
    "group_short_labels = {\n",
    "    \"case-lesional skin Umtata\": \"UM-ADL\",\n",
    "    \"case-lesional skin Cape Town\": \"CT-ADL\",\n",
    "    \"case-nonlesional skin Umtata\": \"UM-ADNL\",\n",
    "    \"case-nonlesional skin Cape Town\": \"CT-ADNL\",\n",
    "    \"control-nonlesional skin Umtata\": \"UM-H\",\n",
    "    \"control-nonlesional skin Cape Town\": \"CT-H\"\n",
    "}\n",
    "\n",
    "adl_subset[\"short_label\"] = adl_subset[\"individual_case_location\"].map(group_short_labels)\n",
    "label_map = adl_subset[\"short_label\"].value_counts().to_dict()\n",
    "print(label_map)\n",
    "label_map = {k: f\"{k} (n={v})\" for k, v in label_map.items()}\n",
    "adl_subset[\"group_label\"] = adl_subset[\"short_label\"].map(label_map)\n",
    "\n",
    "# Define PC variance\n",
    "pc1_var = ordination.proportion_explained['PC1'] * 100\n",
    "pc2_var = ordination.proportion_explained['PC2'] * 100\n",
    "\n",
    "# Palette\n",
    "adl_palette = {\n",
    "    \"case-lesional skin Cape Town\": \"#C9A34F\",  # same as CT ADL\n",
    "    \"case-lesional skin Umtata\": \"#F0806B\"      # same as Umtata ADL\n",
    "}\n",
    "plot_palette = {\n",
    "    label_map[\"CT-ADL\"]: adl_palette[\"case-lesional skin Cape Town\"],\n",
    "    label_map[\"UM-ADL\"]: adl_palette[\"case-lesional skin Umtata\"]\n",
    "}\n",
    "\n",
    "# Plot\n",
    "fig, ax = plt.subplots(figsize=(5.5, 5.5))\n",
    "sns.scatterplot(\n",
    "    data=adl_subset,\n",
    "    x=\"PC1\", y=\"PC2\",\n",
    "    hue=\"group_label\",\n",
    "    hue_order=[label_map[\"CT-ADL\"], label_map[\"UM-ADL\"]],\n",
    "    s=50, edgecolor=\"black\", linewidth=0.5,\n",
    "    palette=plot_palette,\n",
    "    ax=ax\n",
    ")\n",
    "\n",
    "# Confidence ellipses\n",
    "for group in adl_groups:\n",
    "    df = adl_subset[adl_subset[\"individual_case_location\"] == group]\n",
    "    color = adl_palette[group]\n",
    "    pts = df[[\"PC1\", \"PC2\"]].values\n",
    "    center = pts.mean(axis=0)\n",
    "    radius = np.percentile(np.linalg.norm(pts - center, axis=1), 90)\n",
    "    ax.add_patch(Circle(center, radius, edgecolor=color, facecolor=color, alpha=0.2, lw=1, zorder=0))\n",
    "\n",
    "# Axis formatting\n",
    "ax.set_xlabel(f\"RPCA PC1 ({pc1_var:.1f}%)\", fontsize=16)\n",
    "ax.set_ylabel(f\"RPCA PC2 ({pc2_var:.1f}%)\", fontsize=16)\n",
    "ax.set_xticks([-0.2, -0.1, 0.0, 0.1, 0.2, 0.3])\n",
    "ax.set_yticks([-0.2, -0.1, 0.0, 0.1, 0.2, 0.3])\n",
    "ax.tick_params(labelsize=12)\n",
    "ax.legend(frameon=False, fontsize=12, loc='upper right')\n",
    "ax.set_title(\"Lesional Skin: Cape Town vs Umtata\", fontsize=18)\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"../Figures/Main/Fig_ADL_CapeTown_vs_Umtata.png\", dpi=600)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "qiime2-metagenome-2024.10",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
