{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Random Forest Analyses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Standard library\n",
    "import os\n",
    "import warnings\n",
    "import logging\n",
    "from itertools import combinations\n",
    "import string\n",
    "import random\n",
    "\n",
    "# Scientific computing\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from numpy import array\n",
    "import scipy\n",
    "import scipy.stats as ss\n",
    "from scipy import interp\n",
    "from scipy.stats import wilcoxon, ttest_rel\n",
    "\n",
    "# Visualization\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.gridspec import GridSpec\n",
    "import seaborn as sns\n",
    "\n",
    "# scikit-bio\n",
    "from skbio.stats.distance import permanova\n",
    "\n",
    "# BIOM format\n",
    "import biom\n",
    "from biom import load_table\n",
    "\n",
    "# Scikit-learn\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.metrics import (\n",
    "    accuracy_score, confusion_matrix, classification_report,\n",
    "    roc_curve, auc, RocCurveDisplay\n",
    ")\n",
    "from sklearn.model_selection import GroupKFold, StratifiedKFold\n",
    "from sklearn.preprocessing import label_binarize\n",
    "\n",
    "# For confusion matrix\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.model_selection import StratifiedKFold, cross_val_predict\n",
    "from sklearn.metrics import classification_report, ConfusionMatrixDisplay\n",
    "\n",
    "# Set overall styling for plots\n",
    "sns.set_context(\"paper\", font_scale=1.5)\n",
    "sns.set_style(\"ticks\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set random seed for reproducibility\n",
    "SEED = 42\n",
    "np.random.seed(SEED)\n",
    "random.seed(SEED)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Random Forest Classification Tasks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GTGCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTTATTGGGCGTAAAGCGAGCGCAGGCGGTTAGATAAGTCTGAAGTTAAAGGCTG</th>\n",
       "      <th>GTGCCAGCCGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTTATTGGGCGTAAAGCGAGCGCAGGCGGTTAGATAAGTCTGAAGTTAAAGGCTG</th>\n",
       "      <th>GTGCCAGCAGCCGCGGTAATACGTAGGGTGCAAGCGTTGTCCGGAATTACTGGGCGTAAAGAGCTCGTAGGTGGTTTGTCACGTCGTCTGTGAAATTCCA</th>\n",
       "      <th>GTGCCAGCCGCCGCGGTAATACGTAGGGTGCAAGCGTTGTCCGGAATTACTGGGCGTAAAGAGCTCGTAGGTGGTTTGTCACGTCGTCTGTGAAATTCCA</th>\n",
       "      <th>GTGCCAGCAGCCGCGGTAATACGTAGGGTGCAAGCGTTAATCGGAATTATTGGGCGTAAAGCGAGTGCAGACGGTTACTTAAGCCAGATGTGAAATCCCC</th>\n",
       "      <th>GTGCCAGCAGCCGCGGTAATACGTAGGTGGCAAGCGTTGTCCGGAATTATTGGGCGTAAAGCGCGCGCAGGCGGTTTCTTAAGTCTGATGTGAAAGCCCC</th>\n",
       "      <th>GTGCCAGCAGCCGCGGTGATACGTAGGGTGCGAGCGTTGTCCGGATTTATTGGGCGTAAAGGGCTCGTAGGTGGTTGATCGCGTCGGAAGTGTAATCTTG</th>\n",
       "      <th>GTGCCAGCAGCCGCGGTAATACGTAGGGTCCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGTGCGCAGGCGGTTGTGCAAGACCGATGTGAAATCCCC</th>\n",
       "      <th>GTGCCAGCCGCCGCGGTAATACGTAGGTGGCAAGCGTTGTCCGGATTTATTGGGCGTAAAGGGAGCGCAGGTGGTTTCTTAAGTCTGATGTGAAAGCCCA</th>\n",
       "      <th>GTGCCAGCCGCCGCGGTAATACGGAAGGTCCAGGCGTTATCCGGATTTATTGGGTTTAAAGGGAGCGTAGGCGGATTATTAAGTCAGTGGTGAAAGACGG</th>\n",
       "      <th>...</th>\n",
       "      <th>GTGCCAGCCGCCGCGGTAATACGTAGGGGGCAAGCGTTATCCGGATTTACTGGGTGTAAAGGGAGCGTAGACGGCGCAGCAAGTCTGATGTGAAAGGCAG</th>\n",
       "      <th>GTGCCAGCAGCCGCGGTAAGACAGAGGGTGCAAACGTTGCTCGGAATCACTGGGCGTAAAGGGCGTGTAGGCGGGAGAGAAAGTCGGGCGTGAAATCCCT</th>\n",
       "      <th>GTGCCAGCCGCGGTAATACGTAGGGGGCTAGCGTTGTCCGGAATCACTGGGCGTAAAGGGTTCGCAGGCGGAAATGCAAGTCAGGTGTAAAAGGCAGTAG</th>\n",
       "      <th>GTGCCAGCAGCCGCGGTAATACGTAGGGCGCGAGCGTTGTCCGGAATTATTGGGCGTAAAGAGCTTGTAGGCGGTTTGTTGCGTCTGCTGTGAAAGACCG</th>\n",
       "      <th>GTGCCAGCCGCCGCGGTAATACGTAGGGCGCGAGCGTTGTCCGGAATTATTGGGCGTAAAGAGCTTGTAGGCGGTTTGTTGCGTCTGCTGTGAAAGACCG</th>\n",
       "      <th>GTGCCAGCAGCCGCGGTAATACGGAGGGTGCAAGCGTTATCCGGAATCATTGGGTTTAAAGGGTCCGCAGGCGGATTTATAAGTCAGTGGTGAAAGCCTA</th>\n",
       "      <th>GTGCCAGCAGCCGCGGTAATACGTAGGTGGCGAGCGTTGTCCGGAATTACTGGGTGTAAAGGGCGTGTAGGCGGGAAGGTAAGTCAGATGTGAAATACCG</th>\n",
       "      <th>GTGCCAGCCGCCGCGGTAATACGGAGGATGCGAGCGTTATTCGGAATCATTGGGTTTAAAGGGTCTGTAGGCGGGCTATTAAGTCAGAGGTGAAAGGTTT</th>\n",
       "      <th>GTGCCAGCCGCCGCGGTAAGACGAAGGGGGCTAGCGTTGTTCGGAATTACTGGGCGTAAAGCGCGTGCAGGCGGTTATCCAAGTCGGGTGTGAAAGCCTT</th>\n",
       "      <th>GTCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTTATTGGGCGTAAAGCGAGCGCAGGCGGTTAGATAAGTCTGAAGTTAAAGGCTGT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>900344</th>\n",
       "      <td>0.529602</td>\n",
       "      <td>0.328848</td>\n",
       "      <td>0.061356</td>\n",
       "      <td>0.044133</td>\n",
       "      <td>0.011841</td>\n",
       "      <td>0.008073</td>\n",
       "      <td>0.004306</td>\n",
       "      <td>0.004306</td>\n",
       "      <td>0.003229</td>\n",
       "      <td>0.001615</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900459</th>\n",
       "      <td>0.068445</td>\n",
       "      <td>0.061485</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900221</th>\n",
       "      <td>0.000744</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000541</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900570</th>\n",
       "      <td>0.058922</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001212</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001666</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900092</th>\n",
       "      <td>0.623945</td>\n",
       "      <td>0.342909</td>\n",
       "      <td>0.011852</td>\n",
       "      <td>0.006428</td>\n",
       "      <td>0.000603</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001406</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9003972</th>\n",
       "      <td>0.115804</td>\n",
       "      <td>0.058794</td>\n",
       "      <td>0.001586</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002776</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.072972</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003569</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900097</th>\n",
       "      <td>0.011594</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.015942</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.003865</td>\n",
       "      <td>0.002415</td>\n",
       "      <td>0.000483</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900498</th>\n",
       "      <td>0.015707</td>\n",
       "      <td>0.017801</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.035602</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.014660</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.015707</td>\n",
       "      <td>0.010471</td>\n",
       "      <td>0.008377</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900276</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.041322</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.207989</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.015152</td>\n",
       "      <td>0.004132</td>\n",
       "      <td>0.002755</td>\n",
       "      <td>0.001377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900406</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>502 rows Ã— 2283 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         GTGCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTTATTGGGCGTAAAGCGAGCGCAGGCGGTTAGATAAGTCTGAAGTTAAAGGCTG  \\\n",
       "900344                                            0.529602                                                      \n",
       "900459                                            0.068445                                                      \n",
       "900221                                            0.000744                                                      \n",
       "900570                                            0.058922                                                      \n",
       "900092                                            0.623945                                                      \n",
       "...                                                    ...                                                      \n",
       "9003972                                           0.115804                                                      \n",
       "900097                                            0.011594                                                      \n",
       "900498                                            0.015707                                                      \n",
       "900276                                            0.000000                                                      \n",
       "900406                                            0.000000                                                      \n",
       "\n",
       "         GTGCCAGCCGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTTATTGGGCGTAAAGCGAGCGCAGGCGGTTAGATAAGTCTGAAGTTAAAGGCTG  \\\n",
       "900344                                            0.328848                                                      \n",
       "900459                                            0.061485                                                      \n",
       "900221                                            0.000000                                                      \n",
       "900570                                            0.000000                                                      \n",
       "900092                                            0.342909                                                      \n",
       "...                                                    ...                                                      \n",
       "9003972                                           0.058794                                                      \n",
       "900097                                            0.000000                                                      \n",
       "900498                                            0.017801                                                      \n",
       "900276                                            0.000000                                                      \n",
       "900406                                            0.000000                                                      \n",
       "\n",
       "         GTGCCAGCAGCCGCGGTAATACGTAGGGTGCAAGCGTTGTCCGGAATTACTGGGCGTAAAGAGCTCGTAGGTGGTTTGTCACGTCGTCTGTGAAATTCCA  \\\n",
       "900344                                            0.061356                                                      \n",
       "900459                                            0.000000                                                      \n",
       "900221                                            0.000000                                                      \n",
       "900570                                            0.000000                                                      \n",
       "900092                                            0.011852                                                      \n",
       "...                                                    ...                                                      \n",
       "9003972                                           0.001586                                                      \n",
       "900097                                            0.000000                                                      \n",
       "900498                                            0.000000                                                      \n",
       "900276                                            0.041322                                                      \n",
       "900406                                            0.000000                                                      \n",
       "\n",
       "         GTGCCAGCCGCCGCGGTAATACGTAGGGTGCAAGCGTTGTCCGGAATTACTGGGCGTAAAGAGCTCGTAGGTGGTTTGTCACGTCGTCTGTGAAATTCCA  \\\n",
       "900344                                            0.044133                                                      \n",
       "900459                                            0.000000                                                      \n",
       "900221                                            0.000000                                                      \n",
       "900570                                            0.000000                                                      \n",
       "900092                                            0.006428                                                      \n",
       "...                                                    ...                                                      \n",
       "9003972                                           0.000000                                                      \n",
       "900097                                            0.000000                                                      \n",
       "900498                                            0.000000                                                      \n",
       "900276                                            0.000000                                                      \n",
       "900406                                            0.000000                                                      \n",
       "\n",
       "         GTGCCAGCAGCCGCGGTAATACGTAGGGTGCAAGCGTTAATCGGAATTATTGGGCGTAAAGCGAGTGCAGACGGTTACTTAAGCCAGATGTGAAATCCCC  \\\n",
       "900344                                            0.011841                                                      \n",
       "900459                                            0.000000                                                      \n",
       "900221                                            0.000000                                                      \n",
       "900570                                            0.001212                                                      \n",
       "900092                                            0.000603                                                      \n",
       "...                                                    ...                                                      \n",
       "9003972                                           0.002776                                                      \n",
       "900097                                            0.000000                                                      \n",
       "900498                                            0.000000                                                      \n",
       "900276                                            0.000000                                                      \n",
       "900406                                            0.000000                                                      \n",
       "\n",
       "         GTGCCAGCAGCCGCGGTAATACGTAGGTGGCAAGCGTTGTCCGGAATTATTGGGCGTAAAGCGCGCGCAGGCGGTTTCTTAAGTCTGATGTGAAAGCCCC  \\\n",
       "900344                                            0.008073                                                      \n",
       "900459                                            0.000000                                                      \n",
       "900221                                            0.000000                                                      \n",
       "900570                                            0.000000                                                      \n",
       "900092                                            0.000000                                                      \n",
       "...                                                    ...                                                      \n",
       "9003972                                           0.000000                                                      \n",
       "900097                                            0.000000                                                      \n",
       "900498                                            0.000000                                                      \n",
       "900276                                            0.000000                                                      \n",
       "900406                                            0.000000                                                      \n",
       "\n",
       "         GTGCCAGCAGCCGCGGTGATACGTAGGGTGCGAGCGTTGTCCGGATTTATTGGGCGTAAAGGGCTCGTAGGTGGTTGATCGCGTCGGAAGTGTAATCTTG  \\\n",
       "900344                                            0.004306                                                      \n",
       "900459                                            0.000000                                                      \n",
       "900221                                            0.000541                                                      \n",
       "900570                                            0.001666                                                      \n",
       "900092                                            0.000000                                                      \n",
       "...                                                    ...                                                      \n",
       "9003972                                           0.072972                                                      \n",
       "900097                                            0.015942                                                      \n",
       "900498                                            0.035602                                                      \n",
       "900276                                            0.207989                                                      \n",
       "900406                                            0.000000                                                      \n",
       "\n",
       "         GTGCCAGCAGCCGCGGTAATACGTAGGGTCCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGTGCGCAGGCGGTTGTGCAAGACCGATGTGAAATCCCC  \\\n",
       "900344                                            0.004306                                                      \n",
       "900459                                            0.000000                                                      \n",
       "900221                                            0.000000                                                      \n",
       "900570                                            0.000000                                                      \n",
       "900092                                            0.000000                                                      \n",
       "...                                                    ...                                                      \n",
       "9003972                                           0.000000                                                      \n",
       "900097                                            0.000000                                                      \n",
       "900498                                            0.000000                                                      \n",
       "900276                                            0.000000                                                      \n",
       "900406                                            0.000000                                                      \n",
       "\n",
       "         GTGCCAGCCGCCGCGGTAATACGTAGGTGGCAAGCGTTGTCCGGATTTATTGGGCGTAAAGGGAGCGCAGGTGGTTTCTTAAGTCTGATGTGAAAGCCCA  \\\n",
       "900344                                            0.003229                                                      \n",
       "900459                                            0.000000                                                      \n",
       "900221                                            0.000000                                                      \n",
       "900570                                            0.000000                                                      \n",
       "900092                                            0.000000                                                      \n",
       "...                                                    ...                                                      \n",
       "9003972                                           0.000000                                                      \n",
       "900097                                            0.000000                                                      \n",
       "900498                                            0.014660                                                      \n",
       "900276                                            0.000000                                                      \n",
       "900406                                            0.000000                                                      \n",
       "\n",
       "         GTGCCAGCCGCCGCGGTAATACGGAAGGTCCAGGCGTTATCCGGATTTATTGGGTTTAAAGGGAGCGTAGGCGGATTATTAAGTCAGTGGTGAAAGACGG  \\\n",
       "900344                                            0.001615                                                      \n",
       "900459                                            0.000000                                                      \n",
       "900221                                            0.000000                                                      \n",
       "900570                                            0.000000                                                      \n",
       "900092                                            0.001406                                                      \n",
       "...                                                    ...                                                      \n",
       "9003972                                           0.003569                                                      \n",
       "900097                                            0.000000                                                      \n",
       "900498                                            0.000000                                                      \n",
       "900276                                            0.000000                                                      \n",
       "900406                                            0.000000                                                      \n",
       "\n",
       "         ...  \\\n",
       "900344   ...   \n",
       "900459   ...   \n",
       "900221   ...   \n",
       "900570   ...   \n",
       "900092   ...   \n",
       "...      ...   \n",
       "9003972  ...   \n",
       "900097   ...   \n",
       "900498   ...   \n",
       "900276   ...   \n",
       "900406   ...   \n",
       "\n",
       "         GTGCCAGCCGCCGCGGTAATACGTAGGGGGCAAGCGTTATCCGGATTTACTGGGTGTAAAGGGAGCGTAGACGGCGCAGCAAGTCTGATGTGAAAGGCAG  \\\n",
       "900344                                            0.000000                                                      \n",
       "900459                                            0.000000                                                      \n",
       "900221                                            0.000000                                                      \n",
       "900570                                            0.000000                                                      \n",
       "900092                                            0.000000                                                      \n",
       "...                                                    ...                                                      \n",
       "9003972                                           0.000000                                                      \n",
       "900097                                            0.003865                                                      \n",
       "900498                                            0.000000                                                      \n",
       "900276                                            0.000000                                                      \n",
       "900406                                            0.000000                                                      \n",
       "\n",
       "         GTGCCAGCAGCCGCGGTAAGACAGAGGGTGCAAACGTTGCTCGGAATCACTGGGCGTAAAGGGCGTGTAGGCGGGAGAGAAAGTCGGGCGTGAAATCCCT  \\\n",
       "900344                                            0.000000                                                      \n",
       "900459                                            0.000000                                                      \n",
       "900221                                            0.000000                                                      \n",
       "900570                                            0.000000                                                      \n",
       "900092                                            0.000000                                                      \n",
       "...                                                    ...                                                      \n",
       "9003972                                           0.000000                                                      \n",
       "900097                                            0.002415                                                      \n",
       "900498                                            0.000000                                                      \n",
       "900276                                            0.000000                                                      \n",
       "900406                                            0.000000                                                      \n",
       "\n",
       "         GTGCCAGCCGCGGTAATACGTAGGGGGCTAGCGTTGTCCGGAATCACTGGGCGTAAAGGGTTCGCAGGCGGAAATGCAAGTCAGGTGTAAAAGGCAGTAG  \\\n",
       "900344                                            0.000000                                                      \n",
       "900459                                            0.000000                                                      \n",
       "900221                                            0.000000                                                      \n",
       "900570                                            0.000000                                                      \n",
       "900092                                            0.000000                                                      \n",
       "...                                                    ...                                                      \n",
       "9003972                                           0.000000                                                      \n",
       "900097                                            0.000483                                                      \n",
       "900498                                            0.000000                                                      \n",
       "900276                                            0.000000                                                      \n",
       "900406                                            0.000000                                                      \n",
       "\n",
       "         GTGCCAGCAGCCGCGGTAATACGTAGGGCGCGAGCGTTGTCCGGAATTATTGGGCGTAAAGAGCTTGTAGGCGGTTTGTTGCGTCTGCTGTGAAAGACCG  \\\n",
       "900344                                            0.000000                                                      \n",
       "900459                                            0.000000                                                      \n",
       "900221                                            0.000000                                                      \n",
       "900570                                            0.000000                                                      \n",
       "900092                                            0.000000                                                      \n",
       "...                                                    ...                                                      \n",
       "9003972                                           0.000000                                                      \n",
       "900097                                            0.000000                                                      \n",
       "900498                                            0.015707                                                      \n",
       "900276                                            0.000000                                                      \n",
       "900406                                            0.000000                                                      \n",
       "\n",
       "         GTGCCAGCCGCCGCGGTAATACGTAGGGCGCGAGCGTTGTCCGGAATTATTGGGCGTAAAGAGCTTGTAGGCGGTTTGTTGCGTCTGCTGTGAAAGACCG  \\\n",
       "900344                                            0.000000                                                      \n",
       "900459                                            0.000000                                                      \n",
       "900221                                            0.000000                                                      \n",
       "900570                                            0.000000                                                      \n",
       "900092                                            0.000000                                                      \n",
       "...                                                    ...                                                      \n",
       "9003972                                           0.000000                                                      \n",
       "900097                                            0.000000                                                      \n",
       "900498                                            0.010471                                                      \n",
       "900276                                            0.000000                                                      \n",
       "900406                                            0.000000                                                      \n",
       "\n",
       "         GTGCCAGCAGCCGCGGTAATACGGAGGGTGCAAGCGTTATCCGGAATCATTGGGTTTAAAGGGTCCGCAGGCGGATTTATAAGTCAGTGGTGAAAGCCTA  \\\n",
       "900344                                            0.000000                                                      \n",
       "900459                                            0.000000                                                      \n",
       "900221                                            0.000000                                                      \n",
       "900570                                            0.000000                                                      \n",
       "900092                                            0.000000                                                      \n",
       "...                                                    ...                                                      \n",
       "9003972                                           0.000000                                                      \n",
       "900097                                            0.000000                                                      \n",
       "900498                                            0.008377                                                      \n",
       "900276                                            0.000000                                                      \n",
       "900406                                            0.000000                                                      \n",
       "\n",
       "         GTGCCAGCAGCCGCGGTAATACGTAGGTGGCGAGCGTTGTCCGGAATTACTGGGTGTAAAGGGCGTGTAGGCGGGAAGGTAAGTCAGATGTGAAATACCG  \\\n",
       "900344                                            0.000000                                                      \n",
       "900459                                            0.000000                                                      \n",
       "900221                                            0.000000                                                      \n",
       "900570                                            0.000000                                                      \n",
       "900092                                            0.000000                                                      \n",
       "...                                                    ...                                                      \n",
       "9003972                                           0.000000                                                      \n",
       "900097                                            0.000000                                                      \n",
       "900498                                            0.000000                                                      \n",
       "900276                                            0.015152                                                      \n",
       "900406                                            0.000000                                                      \n",
       "\n",
       "         GTGCCAGCCGCCGCGGTAATACGGAGGATGCGAGCGTTATTCGGAATCATTGGGTTTAAAGGGTCTGTAGGCGGGCTATTAAGTCAGAGGTGAAAGGTTT  \\\n",
       "900344                                            0.000000                                                      \n",
       "900459                                            0.000000                                                      \n",
       "900221                                            0.000000                                                      \n",
       "900570                                            0.000000                                                      \n",
       "900092                                            0.000000                                                      \n",
       "...                                                    ...                                                      \n",
       "9003972                                           0.000000                                                      \n",
       "900097                                            0.000000                                                      \n",
       "900498                                            0.000000                                                      \n",
       "900276                                            0.004132                                                      \n",
       "900406                                            0.000000                                                      \n",
       "\n",
       "         GTGCCAGCCGCCGCGGTAAGACGAAGGGGGCTAGCGTTGTTCGGAATTACTGGGCGTAAAGCGCGTGCAGGCGGTTATCCAAGTCGGGTGTGAAAGCCTT  \\\n",
       "900344                                            0.000000                                                      \n",
       "900459                                            0.000000                                                      \n",
       "900221                                            0.000000                                                      \n",
       "900570                                            0.000000                                                      \n",
       "900092                                            0.000000                                                      \n",
       "...                                                    ...                                                      \n",
       "9003972                                           0.000000                                                      \n",
       "900097                                            0.000000                                                      \n",
       "900498                                            0.000000                                                      \n",
       "900276                                            0.002755                                                      \n",
       "900406                                            0.000000                                                      \n",
       "\n",
       "         GTCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTTATTGGGCGTAAAGCGAGCGCAGGCGGTTAGATAAGTCTGAAGTTAAAGGCTGT  \n",
       "900344                                            0.000000                                                     \n",
       "900459                                            0.000000                                                     \n",
       "900221                                            0.000000                                                     \n",
       "900570                                            0.000000                                                     \n",
       "900092                                            0.000000                                                     \n",
       "...                                                    ...                                                     \n",
       "9003972                                           0.000000                                                     \n",
       "900097                                            0.000000                                                     \n",
       "900498                                            0.000000                                                     \n",
       "900276                                            0.001377                                                     \n",
       "900406                                            0.000000                                                     \n",
       "\n",
       "[502 rows x 2283 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read in table at ASV level\n",
    "biom_path = '../Data/Tables/Count_Tables/1_209766_feature_table.biom'\n",
    "\n",
    "biom_tbl = load_table(biom_path)\n",
    "df = pd.DataFrame(biom_tbl.to_dataframe().T)\n",
    "\n",
    "# delete the prefix from the index\n",
    "df.index = df.index.str.replace('15564.', '')\n",
    "\n",
    "# Convert to relative abundance by dividing each row by its row sum\n",
    "df_dense = df.div(df.sum(axis=1), axis=0)\n",
    "df = pd.DataFrame(df_dense.values, index=df_dense.index, columns=df_dense.columns)  # Force dense\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PlateNumber</th>\n",
       "      <th>PlateLocation</th>\n",
       "      <th>i5</th>\n",
       "      <th>i5Sequence</th>\n",
       "      <th>i7</th>\n",
       "      <th>i7Sequence</th>\n",
       "      <th>identifier</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Plate ID</th>\n",
       "      <th>Well location</th>\n",
       "      <th>...</th>\n",
       "      <th>specimen</th>\n",
       "      <th>age_months</th>\n",
       "      <th>sex</th>\n",
       "      <th>enrolment_date</th>\n",
       "      <th>enrolment_season</th>\n",
       "      <th>hiv_exposure</th>\n",
       "      <th>hiv_status</th>\n",
       "      <th>household_size</th>\n",
       "      <th>o_scorad</th>\n",
       "      <th>group</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#sample-id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Ca009STL</th>\n",
       "      <td>1</td>\n",
       "      <td>A1</td>\n",
       "      <td>SA501</td>\n",
       "      <td>ATCGTACG</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA501</td>\n",
       "      <td>CGAGAGTT-ATCGTACG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>A1</td>\n",
       "      <td>...</td>\n",
       "      <td>skin</td>\n",
       "      <td>24.0</td>\n",
       "      <td>male</td>\n",
       "      <td>4/16/2015</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>4.0</td>\n",
       "      <td>40</td>\n",
       "      <td>skin-ADL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900221</th>\n",
       "      <td>1</td>\n",
       "      <td>B1</td>\n",
       "      <td>SA502</td>\n",
       "      <td>ACTATCTG</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA502</td>\n",
       "      <td>CGAGAGTT-ACTATCTG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>B1</td>\n",
       "      <td>...</td>\n",
       "      <td>skin</td>\n",
       "      <td>9.0</td>\n",
       "      <td>female</td>\n",
       "      <td>8/11/2015</td>\n",
       "      <td>Winter</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>7.0</td>\n",
       "      <td>34</td>\n",
       "      <td>skin-ADL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca010EBL</th>\n",
       "      <td>1</td>\n",
       "      <td>C1</td>\n",
       "      <td>SA503</td>\n",
       "      <td>TAGCGAGT</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA503</td>\n",
       "      <td>CGAGAGTT-TAGCGAGT</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>C1</td>\n",
       "      <td>...</td>\n",
       "      <td>skin</td>\n",
       "      <td>24.0</td>\n",
       "      <td>female</td>\n",
       "      <td>11/20/2014</td>\n",
       "      <td>Spring</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>7.0</td>\n",
       "      <td>21</td>\n",
       "      <td>skin-ADL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900460</th>\n",
       "      <td>1</td>\n",
       "      <td>D1</td>\n",
       "      <td>SA504</td>\n",
       "      <td>CTGCGTGT</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA504</td>\n",
       "      <td>CGAGAGTT-CTGCGTGT</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>D1</td>\n",
       "      <td>...</td>\n",
       "      <td>skin</td>\n",
       "      <td>18.0</td>\n",
       "      <td>female</td>\n",
       "      <td>9/23/2015</td>\n",
       "      <td>Spring</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>40</td>\n",
       "      <td>skin-ADL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900051</th>\n",
       "      <td>1</td>\n",
       "      <td>E1</td>\n",
       "      <td>SA505</td>\n",
       "      <td>TCATCGAG</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA505</td>\n",
       "      <td>CGAGAGTT-TCATCGAG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>E1</td>\n",
       "      <td>...</td>\n",
       "      <td>skin</td>\n",
       "      <td>31.0</td>\n",
       "      <td>male</td>\n",
       "      <td>4/21/2015</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>7.0</td>\n",
       "      <td>41</td>\n",
       "      <td>skin-ADL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900401</th>\n",
       "      <td>5</td>\n",
       "      <td>C12</td>\n",
       "      <td>SB503</td>\n",
       "      <td>AGAGTCAC</td>\n",
       "      <td>SB712</td>\n",
       "      <td>CGTAGCGA</td>\n",
       "      <td>SB712SB503</td>\n",
       "      <td>CGTAGCGA-AGAGTCAC</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>C12</td>\n",
       "      <td>...</td>\n",
       "      <td>skin</td>\n",
       "      <td>21.0</td>\n",
       "      <td>female</td>\n",
       "      <td>9/17/2015</td>\n",
       "      <td>Spring</td>\n",
       "      <td>Exposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>12.0</td>\n",
       "      <td>38</td>\n",
       "      <td>skin-ADNL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900402</th>\n",
       "      <td>6</td>\n",
       "      <td>B4</td>\n",
       "      <td>SA502</td>\n",
       "      <td>ACTATCTG</td>\n",
       "      <td>SB704</td>\n",
       "      <td>TCTCTATG</td>\n",
       "      <td>SB704SA502</td>\n",
       "      <td>TCTCTATG-ACTATCTG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>B4</td>\n",
       "      <td>...</td>\n",
       "      <td>nasal</td>\n",
       "      <td>21.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nares-AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca006ONL</th>\n",
       "      <td>6</td>\n",
       "      <td>F1</td>\n",
       "      <td>SA506</td>\n",
       "      <td>CGTGAGTG</td>\n",
       "      <td>SB701</td>\n",
       "      <td>CTCGACTT</td>\n",
       "      <td>SB701SA506</td>\n",
       "      <td>CTCGACTT-CGTGAGTG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>F1</td>\n",
       "      <td>...</td>\n",
       "      <td>skin</td>\n",
       "      <td>35.0</td>\n",
       "      <td>female</td>\n",
       "      <td>3/25/2015</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>3.0</td>\n",
       "      <td>34</td>\n",
       "      <td>skin-ADL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca006ONNL</th>\n",
       "      <td>6</td>\n",
       "      <td>F2</td>\n",
       "      <td>SA506</td>\n",
       "      <td>CGTGAGTG</td>\n",
       "      <td>SB702</td>\n",
       "      <td>CGAAGTAT</td>\n",
       "      <td>SB702SA506</td>\n",
       "      <td>CGAAGTAT-CGTGAGTG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>F2</td>\n",
       "      <td>...</td>\n",
       "      <td>skin</td>\n",
       "      <td>35.0</td>\n",
       "      <td>female</td>\n",
       "      <td>3/25/2015</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>3.0</td>\n",
       "      <td>34</td>\n",
       "      <td>skin-ADNL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca006ONPN</th>\n",
       "      <td>6</td>\n",
       "      <td>F3</td>\n",
       "      <td>SA506</td>\n",
       "      <td>CGTGAGTG</td>\n",
       "      <td>SB703</td>\n",
       "      <td>TAGCAGCT</td>\n",
       "      <td>SB703SA506</td>\n",
       "      <td>TAGCAGCT-CGTGAGTG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>F3</td>\n",
       "      <td>...</td>\n",
       "      <td>nasal</td>\n",
       "      <td>35.0</td>\n",
       "      <td>female</td>\n",
       "      <td>3/25/2015</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>3.0</td>\n",
       "      <td>34</td>\n",
       "      <td>nares-AD</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>462 rows Ã— 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            PlateNumber PlateLocation     i5 i5Sequence     i7 i7Sequence  \\\n",
       "#sample-id                                                                  \n",
       "Ca009STL              1            A1  SA501   ATCGTACG  SA701   CGAGAGTT   \n",
       "900221                1            B1  SA502   ACTATCTG  SA701   CGAGAGTT   \n",
       "Ca010EBL              1            C1  SA503   TAGCGAGT  SA701   CGAGAGTT   \n",
       "900460                1            D1  SA504   CTGCGTGT  SA701   CGAGAGTT   \n",
       "900051                1            E1  SA505   TCATCGAG  SA701   CGAGAGTT   \n",
       "...                 ...           ...    ...        ...    ...        ...   \n",
       "900401                5           C12  SB503   AGAGTCAC  SB712   CGTAGCGA   \n",
       "900402                6            B4  SA502   ACTATCTG  SB704   TCTCTATG   \n",
       "Ca006ONL              6            F1  SA506   CGTGAGTG  SB701   CTCGACTT   \n",
       "Ca006ONNL             6            F2  SA506   CGTGAGTG  SB702   CGAAGTAT   \n",
       "Ca006ONPN             6            F3  SA506   CGTGAGTG  SB703   TAGCAGCT   \n",
       "\n",
       "            identifier           Sequence      Plate ID Well location  ...  \\\n",
       "#sample-id                                                             ...   \n",
       "Ca009STL    SA701SA501  CGAGAGTT-ATCGTACG  1.010000e+21            A1  ...   \n",
       "900221      SA701SA502  CGAGAGTT-ACTATCTG  1.010000e+21            B1  ...   \n",
       "Ca010EBL    SA701SA503  CGAGAGTT-TAGCGAGT  1.010000e+21            C1  ...   \n",
       "900460      SA701SA504  CGAGAGTT-CTGCGTGT  1.010000e+21            D1  ...   \n",
       "900051      SA701SA505  CGAGAGTT-TCATCGAG  1.010000e+21            E1  ...   \n",
       "...                ...                ...           ...           ...  ...   \n",
       "900401      SB712SB503  CGTAGCGA-AGAGTCAC  1.010000e+21           C12  ...   \n",
       "900402      SB704SA502  TCTCTATG-ACTATCTG  1.010000e+21            B4  ...   \n",
       "Ca006ONL    SB701SA506  CTCGACTT-CGTGAGTG  1.010000e+21            F1  ...   \n",
       "Ca006ONNL   SB702SA506  CGAAGTAT-CGTGAGTG  1.010000e+21            F2  ...   \n",
       "Ca006ONPN   SB703SA506  TAGCAGCT-CGTGAGTG  1.010000e+21            F3  ...   \n",
       "\n",
       "            specimen age_months     sex enrolment_date enrolment_season  \\\n",
       "#sample-id                                                                \n",
       "Ca009STL        skin       24.0    male      4/16/2015          Autumn    \n",
       "900221          skin        9.0  female      8/11/2015           Winter   \n",
       "Ca010EBL        skin       24.0  female     11/20/2014           Spring   \n",
       "900460          skin       18.0  female      9/23/2015           Spring   \n",
       "900051          skin       31.0    male      4/21/2015          Autumn    \n",
       "...              ...        ...     ...            ...              ...   \n",
       "900401          skin       21.0  female      9/17/2015           Spring   \n",
       "900402         nasal       21.0     NaN            NaN              NaN   \n",
       "Ca006ONL        skin       35.0  female      3/25/2015           Autumn   \n",
       "Ca006ONNL       skin       35.0  female      3/25/2015           Autumn   \n",
       "Ca006ONPN      nasal       35.0  female      3/25/2015           Autumn   \n",
       "\n",
       "           hiv_exposure hiv_status household_size o_scorad      group  \n",
       "#sample-id                                                             \n",
       "Ca009STL      Unexposed   negative            4.0       40   skin-ADL  \n",
       "900221        Unexposed   negative            7.0       34   skin-ADL  \n",
       "Ca010EBL      Unexposed   negative            7.0       21   skin-ADL  \n",
       "900460        Unexposed        NaN            4.0       40   skin-ADL  \n",
       "900051        Unexposed   negative            7.0       41   skin-ADL  \n",
       "...                 ...        ...            ...      ...        ...  \n",
       "900401          Exposed   negative           12.0       38  skin-ADNL  \n",
       "900402              NaN        NaN            NaN      NaN   nares-AD  \n",
       "Ca006ONL      Unexposed   negative            3.0       34   skin-ADL  \n",
       "Ca006ONNL     Unexposed   negative            3.0       34  skin-ADNL  \n",
       "Ca006ONPN     Unexposed   negative            3.0       34   nares-AD  \n",
       "\n",
       "[462 rows x 30 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load the metadata\n",
    "metadata_path = '../Metadata/16S_AD_South-Africa_metadata_subset.tsv'\n",
    "metadata = pd.read_csv(metadata_path, sep='\\t')\n",
    "\n",
    "metadata['#sample-id'] = metadata['#sample-id'].str.replace('_', '')\n",
    "# Set Sample-ID as the index for the metadata dataframe \n",
    "metadata = metadata.set_index('#sample-id')\n",
    "\n",
    "\n",
    "# Create group column based on case_type to simplify group names\n",
    "metadata['group'] = metadata['case_type'].map({\n",
    "    'case-lesional_skin': 'skin-ADL',\n",
    "    'case-nonlesional_skin': 'skin-ADNL', \n",
    "    'control-nonlesional_skin': 'skin-H',\n",
    "    'case-anterior_nares': 'nares-AD',\n",
    "    'control-anterior_nares': 'nares-H'\n",
    "})\n",
    "\n",
    "metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GTGCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTTATTGGGCGTAAAGCGAGCGCAGGCGGTTAGATAAGTCTGAAGTTAAAGGCTG</th>\n",
       "      <th>GTGCCAGCCGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTTATTGGGCGTAAAGCGAGCGCAGGCGGTTAGATAAGTCTGAAGTTAAAGGCTG</th>\n",
       "      <th>GTGCCAGCAGCCGCGGTAATACGTAGGGTGCAAGCGTTGTCCGGAATTACTGGGCGTAAAGAGCTCGTAGGTGGTTTGTCACGTCGTCTGTGAAATTCCA</th>\n",
       "      <th>GTGCCAGCCGCCGCGGTAATACGTAGGGTGCAAGCGTTGTCCGGAATTACTGGGCGTAAAGAGCTCGTAGGTGGTTTGTCACGTCGTCTGTGAAATTCCA</th>\n",
       "      <th>GTGCCAGCAGCCGCGGTAATACGTAGGGTGCAAGCGTTAATCGGAATTATTGGGCGTAAAGCGAGTGCAGACGGTTACTTAAGCCAGATGTGAAATCCCC</th>\n",
       "      <th>GTGCCAGCAGCCGCGGTAATACGTAGGTGGCAAGCGTTGTCCGGAATTATTGGGCGTAAAGCGCGCGCAGGCGGTTTCTTAAGTCTGATGTGAAAGCCCC</th>\n",
       "      <th>GTGCCAGCAGCCGCGGTGATACGTAGGGTGCGAGCGTTGTCCGGATTTATTGGGCGTAAAGGGCTCGTAGGTGGTTGATCGCGTCGGAAGTGTAATCTTG</th>\n",
       "      <th>GTGCCAGCAGCCGCGGTAATACGTAGGGTCCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGTGCGCAGGCGGTTGTGCAAGACCGATGTGAAATCCCC</th>\n",
       "      <th>GTGCCAGCCGCCGCGGTAATACGTAGGTGGCAAGCGTTGTCCGGATTTATTGGGCGTAAAGGGAGCGCAGGTGGTTTCTTAAGTCTGATGTGAAAGCCCA</th>\n",
       "      <th>GTGCCAGCCGCCGCGGTAATACGGAAGGTCCAGGCGTTATCCGGATTTATTGGGTTTAAAGGGAGCGTAGGCGGATTATTAAGTCAGTGGTGAAAGACGG</th>\n",
       "      <th>...</th>\n",
       "      <th>GTGCCAGCCGCCGCGGTAATACGTAGGGGGCAAGCGTTATCCGGATTTACTGGGTGTAAAGGGAGCGTAGACGGCGCAGCAAGTCTGATGTGAAAGGCAG</th>\n",
       "      <th>GTGCCAGCAGCCGCGGTAAGACAGAGGGTGCAAACGTTGCTCGGAATCACTGGGCGTAAAGGGCGTGTAGGCGGGAGAGAAAGTCGGGCGTGAAATCCCT</th>\n",
       "      <th>GTGCCAGCCGCGGTAATACGTAGGGGGCTAGCGTTGTCCGGAATCACTGGGCGTAAAGGGTTCGCAGGCGGAAATGCAAGTCAGGTGTAAAAGGCAGTAG</th>\n",
       "      <th>GTGCCAGCAGCCGCGGTAATACGTAGGGCGCGAGCGTTGTCCGGAATTATTGGGCGTAAAGAGCTTGTAGGCGGTTTGTTGCGTCTGCTGTGAAAGACCG</th>\n",
       "      <th>GTGCCAGCCGCCGCGGTAATACGTAGGGCGCGAGCGTTGTCCGGAATTATTGGGCGTAAAGAGCTTGTAGGCGGTTTGTTGCGTCTGCTGTGAAAGACCG</th>\n",
       "      <th>GTGCCAGCAGCCGCGGTAATACGGAGGGTGCAAGCGTTATCCGGAATCATTGGGTTTAAAGGGTCCGCAGGCGGATTTATAAGTCAGTGGTGAAAGCCTA</th>\n",
       "      <th>GTGCCAGCAGCCGCGGTAATACGTAGGTGGCGAGCGTTGTCCGGAATTACTGGGTGTAAAGGGCGTGTAGGCGGGAAGGTAAGTCAGATGTGAAATACCG</th>\n",
       "      <th>GTGCCAGCCGCCGCGGTAATACGGAGGATGCGAGCGTTATTCGGAATCATTGGGTTTAAAGGGTCTGTAGGCGGGCTATTAAGTCAGAGGTGAAAGGTTT</th>\n",
       "      <th>GTGCCAGCCGCCGCGGTAAGACGAAGGGGGCTAGCGTTGTTCGGAATTACTGGGCGTAAAGCGCGTGCAGGCGGTTATCCAAGTCGGGTGTGAAAGCCTT</th>\n",
       "      <th>GTCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTTATTGGGCGTAAAGCGAGCGCAGGCGGTTAGATAAGTCTGAAGTTAAAGGCTGT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>900344</th>\n",
       "      <td>0.529602</td>\n",
       "      <td>0.328848</td>\n",
       "      <td>0.061356</td>\n",
       "      <td>0.044133</td>\n",
       "      <td>0.011841</td>\n",
       "      <td>0.008073</td>\n",
       "      <td>0.004306</td>\n",
       "      <td>0.004306</td>\n",
       "      <td>0.003229</td>\n",
       "      <td>0.001615</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900459</th>\n",
       "      <td>0.068445</td>\n",
       "      <td>0.061485</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900221</th>\n",
       "      <td>0.000744</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000541</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900570</th>\n",
       "      <td>0.058922</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001212</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001666</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900092</th>\n",
       "      <td>0.623945</td>\n",
       "      <td>0.342909</td>\n",
       "      <td>0.011852</td>\n",
       "      <td>0.006428</td>\n",
       "      <td>0.000603</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001406</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900294</th>\n",
       "      <td>0.017937</td>\n",
       "      <td>0.010463</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.144993</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900097</th>\n",
       "      <td>0.011594</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.015942</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.003865</td>\n",
       "      <td>0.002415</td>\n",
       "      <td>0.000483</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900498</th>\n",
       "      <td>0.015707</td>\n",
       "      <td>0.017801</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.035602</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.014660</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.015707</td>\n",
       "      <td>0.010471</td>\n",
       "      <td>0.008377</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900276</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.041322</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.207989</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.015152</td>\n",
       "      <td>0.004132</td>\n",
       "      <td>0.002755</td>\n",
       "      <td>0.001377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900406</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>462 rows Ã— 2283 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        GTGCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTTATTGGGCGTAAAGCGAGCGCAGGCGGTTAGATAAGTCTGAAGTTAAAGGCTG  \\\n",
       "900344                                           0.529602                                                      \n",
       "900459                                           0.068445                                                      \n",
       "900221                                           0.000744                                                      \n",
       "900570                                           0.058922                                                      \n",
       "900092                                           0.623945                                                      \n",
       "...                                                   ...                                                      \n",
       "900294                                           0.017937                                                      \n",
       "900097                                           0.011594                                                      \n",
       "900498                                           0.015707                                                      \n",
       "900276                                           0.000000                                                      \n",
       "900406                                           0.000000                                                      \n",
       "\n",
       "        GTGCCAGCCGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTTATTGGGCGTAAAGCGAGCGCAGGCGGTTAGATAAGTCTGAAGTTAAAGGCTG  \\\n",
       "900344                                           0.328848                                                      \n",
       "900459                                           0.061485                                                      \n",
       "900221                                           0.000000                                                      \n",
       "900570                                           0.000000                                                      \n",
       "900092                                           0.342909                                                      \n",
       "...                                                   ...                                                      \n",
       "900294                                           0.010463                                                      \n",
       "900097                                           0.000000                                                      \n",
       "900498                                           0.017801                                                      \n",
       "900276                                           0.000000                                                      \n",
       "900406                                           0.000000                                                      \n",
       "\n",
       "        GTGCCAGCAGCCGCGGTAATACGTAGGGTGCAAGCGTTGTCCGGAATTACTGGGCGTAAAGAGCTCGTAGGTGGTTTGTCACGTCGTCTGTGAAATTCCA  \\\n",
       "900344                                           0.061356                                                      \n",
       "900459                                           0.000000                                                      \n",
       "900221                                           0.000000                                                      \n",
       "900570                                           0.000000                                                      \n",
       "900092                                           0.011852                                                      \n",
       "...                                                   ...                                                      \n",
       "900294                                           0.000000                                                      \n",
       "900097                                           0.000000                                                      \n",
       "900498                                           0.000000                                                      \n",
       "900276                                           0.041322                                                      \n",
       "900406                                           0.000000                                                      \n",
       "\n",
       "        GTGCCAGCCGCCGCGGTAATACGTAGGGTGCAAGCGTTGTCCGGAATTACTGGGCGTAAAGAGCTCGTAGGTGGTTTGTCACGTCGTCTGTGAAATTCCA  \\\n",
       "900344                                           0.044133                                                      \n",
       "900459                                           0.000000                                                      \n",
       "900221                                           0.000000                                                      \n",
       "900570                                           0.000000                                                      \n",
       "900092                                           0.006428                                                      \n",
       "...                                                   ...                                                      \n",
       "900294                                           0.000000                                                      \n",
       "900097                                           0.000000                                                      \n",
       "900498                                           0.000000                                                      \n",
       "900276                                           0.000000                                                      \n",
       "900406                                           0.000000                                                      \n",
       "\n",
       "        GTGCCAGCAGCCGCGGTAATACGTAGGGTGCAAGCGTTAATCGGAATTATTGGGCGTAAAGCGAGTGCAGACGGTTACTTAAGCCAGATGTGAAATCCCC  \\\n",
       "900344                                           0.011841                                                      \n",
       "900459                                           0.000000                                                      \n",
       "900221                                           0.000000                                                      \n",
       "900570                                           0.001212                                                      \n",
       "900092                                           0.000603                                                      \n",
       "...                                                   ...                                                      \n",
       "900294                                           0.000000                                                      \n",
       "900097                                           0.000000                                                      \n",
       "900498                                           0.000000                                                      \n",
       "900276                                           0.000000                                                      \n",
       "900406                                           0.000000                                                      \n",
       "\n",
       "        GTGCCAGCAGCCGCGGTAATACGTAGGTGGCAAGCGTTGTCCGGAATTATTGGGCGTAAAGCGCGCGCAGGCGGTTTCTTAAGTCTGATGTGAAAGCCCC  \\\n",
       "900344                                           0.008073                                                      \n",
       "900459                                           0.000000                                                      \n",
       "900221                                           0.000000                                                      \n",
       "900570                                           0.000000                                                      \n",
       "900092                                           0.000000                                                      \n",
       "...                                                   ...                                                      \n",
       "900294                                           0.000000                                                      \n",
       "900097                                           0.000000                                                      \n",
       "900498                                           0.000000                                                      \n",
       "900276                                           0.000000                                                      \n",
       "900406                                           0.000000                                                      \n",
       "\n",
       "        GTGCCAGCAGCCGCGGTGATACGTAGGGTGCGAGCGTTGTCCGGATTTATTGGGCGTAAAGGGCTCGTAGGTGGTTGATCGCGTCGGAAGTGTAATCTTG  \\\n",
       "900344                                           0.004306                                                      \n",
       "900459                                           0.000000                                                      \n",
       "900221                                           0.000541                                                      \n",
       "900570                                           0.001666                                                      \n",
       "900092                                           0.000000                                                      \n",
       "...                                                   ...                                                      \n",
       "900294                                           0.144993                                                      \n",
       "900097                                           0.015942                                                      \n",
       "900498                                           0.035602                                                      \n",
       "900276                                           0.207989                                                      \n",
       "900406                                           0.000000                                                      \n",
       "\n",
       "        GTGCCAGCAGCCGCGGTAATACGTAGGGTCCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGTGCGCAGGCGGTTGTGCAAGACCGATGTGAAATCCCC  \\\n",
       "900344                                           0.004306                                                      \n",
       "900459                                           0.000000                                                      \n",
       "900221                                           0.000000                                                      \n",
       "900570                                           0.000000                                                      \n",
       "900092                                           0.000000                                                      \n",
       "...                                                   ...                                                      \n",
       "900294                                           0.000000                                                      \n",
       "900097                                           0.000000                                                      \n",
       "900498                                           0.000000                                                      \n",
       "900276                                           0.000000                                                      \n",
       "900406                                           0.000000                                                      \n",
       "\n",
       "        GTGCCAGCCGCCGCGGTAATACGTAGGTGGCAAGCGTTGTCCGGATTTATTGGGCGTAAAGGGAGCGCAGGTGGTTTCTTAAGTCTGATGTGAAAGCCCA  \\\n",
       "900344                                           0.003229                                                      \n",
       "900459                                           0.000000                                                      \n",
       "900221                                           0.000000                                                      \n",
       "900570                                           0.000000                                                      \n",
       "900092                                           0.000000                                                      \n",
       "...                                                   ...                                                      \n",
       "900294                                           0.000000                                                      \n",
       "900097                                           0.000000                                                      \n",
       "900498                                           0.014660                                                      \n",
       "900276                                           0.000000                                                      \n",
       "900406                                           0.000000                                                      \n",
       "\n",
       "        GTGCCAGCCGCCGCGGTAATACGGAAGGTCCAGGCGTTATCCGGATTTATTGGGTTTAAAGGGAGCGTAGGCGGATTATTAAGTCAGTGGTGAAAGACGG  \\\n",
       "900344                                           0.001615                                                      \n",
       "900459                                           0.000000                                                      \n",
       "900221                                           0.000000                                                      \n",
       "900570                                           0.000000                                                      \n",
       "900092                                           0.001406                                                      \n",
       "...                                                   ...                                                      \n",
       "900294                                           0.000000                                                      \n",
       "900097                                           0.000000                                                      \n",
       "900498                                           0.000000                                                      \n",
       "900276                                           0.000000                                                      \n",
       "900406                                           0.000000                                                      \n",
       "\n",
       "        ...  \\\n",
       "900344  ...   \n",
       "900459  ...   \n",
       "900221  ...   \n",
       "900570  ...   \n",
       "900092  ...   \n",
       "...     ...   \n",
       "900294  ...   \n",
       "900097  ...   \n",
       "900498  ...   \n",
       "900276  ...   \n",
       "900406  ...   \n",
       "\n",
       "        GTGCCAGCCGCCGCGGTAATACGTAGGGGGCAAGCGTTATCCGGATTTACTGGGTGTAAAGGGAGCGTAGACGGCGCAGCAAGTCTGATGTGAAAGGCAG  \\\n",
       "900344                                           0.000000                                                      \n",
       "900459                                           0.000000                                                      \n",
       "900221                                           0.000000                                                      \n",
       "900570                                           0.000000                                                      \n",
       "900092                                           0.000000                                                      \n",
       "...                                                   ...                                                      \n",
       "900294                                           0.000000                                                      \n",
       "900097                                           0.003865                                                      \n",
       "900498                                           0.000000                                                      \n",
       "900276                                           0.000000                                                      \n",
       "900406                                           0.000000                                                      \n",
       "\n",
       "        GTGCCAGCAGCCGCGGTAAGACAGAGGGTGCAAACGTTGCTCGGAATCACTGGGCGTAAAGGGCGTGTAGGCGGGAGAGAAAGTCGGGCGTGAAATCCCT  \\\n",
       "900344                                           0.000000                                                      \n",
       "900459                                           0.000000                                                      \n",
       "900221                                           0.000000                                                      \n",
       "900570                                           0.000000                                                      \n",
       "900092                                           0.000000                                                      \n",
       "...                                                   ...                                                      \n",
       "900294                                           0.000000                                                      \n",
       "900097                                           0.002415                                                      \n",
       "900498                                           0.000000                                                      \n",
       "900276                                           0.000000                                                      \n",
       "900406                                           0.000000                                                      \n",
       "\n",
       "        GTGCCAGCCGCGGTAATACGTAGGGGGCTAGCGTTGTCCGGAATCACTGGGCGTAAAGGGTTCGCAGGCGGAAATGCAAGTCAGGTGTAAAAGGCAGTAG  \\\n",
       "900344                                           0.000000                                                      \n",
       "900459                                           0.000000                                                      \n",
       "900221                                           0.000000                                                      \n",
       "900570                                           0.000000                                                      \n",
       "900092                                           0.000000                                                      \n",
       "...                                                   ...                                                      \n",
       "900294                                           0.000000                                                      \n",
       "900097                                           0.000483                                                      \n",
       "900498                                           0.000000                                                      \n",
       "900276                                           0.000000                                                      \n",
       "900406                                           0.000000                                                      \n",
       "\n",
       "        GTGCCAGCAGCCGCGGTAATACGTAGGGCGCGAGCGTTGTCCGGAATTATTGGGCGTAAAGAGCTTGTAGGCGGTTTGTTGCGTCTGCTGTGAAAGACCG  \\\n",
       "900344                                           0.000000                                                      \n",
       "900459                                           0.000000                                                      \n",
       "900221                                           0.000000                                                      \n",
       "900570                                           0.000000                                                      \n",
       "900092                                           0.000000                                                      \n",
       "...                                                   ...                                                      \n",
       "900294                                           0.000000                                                      \n",
       "900097                                           0.000000                                                      \n",
       "900498                                           0.015707                                                      \n",
       "900276                                           0.000000                                                      \n",
       "900406                                           0.000000                                                      \n",
       "\n",
       "        GTGCCAGCCGCCGCGGTAATACGTAGGGCGCGAGCGTTGTCCGGAATTATTGGGCGTAAAGAGCTTGTAGGCGGTTTGTTGCGTCTGCTGTGAAAGACCG  \\\n",
       "900344                                           0.000000                                                      \n",
       "900459                                           0.000000                                                      \n",
       "900221                                           0.000000                                                      \n",
       "900570                                           0.000000                                                      \n",
       "900092                                           0.000000                                                      \n",
       "...                                                   ...                                                      \n",
       "900294                                           0.000000                                                      \n",
       "900097                                           0.000000                                                      \n",
       "900498                                           0.010471                                                      \n",
       "900276                                           0.000000                                                      \n",
       "900406                                           0.000000                                                      \n",
       "\n",
       "        GTGCCAGCAGCCGCGGTAATACGGAGGGTGCAAGCGTTATCCGGAATCATTGGGTTTAAAGGGTCCGCAGGCGGATTTATAAGTCAGTGGTGAAAGCCTA  \\\n",
       "900344                                           0.000000                                                      \n",
       "900459                                           0.000000                                                      \n",
       "900221                                           0.000000                                                      \n",
       "900570                                           0.000000                                                      \n",
       "900092                                           0.000000                                                      \n",
       "...                                                   ...                                                      \n",
       "900294                                           0.000000                                                      \n",
       "900097                                           0.000000                                                      \n",
       "900498                                           0.008377                                                      \n",
       "900276                                           0.000000                                                      \n",
       "900406                                           0.000000                                                      \n",
       "\n",
       "        GTGCCAGCAGCCGCGGTAATACGTAGGTGGCGAGCGTTGTCCGGAATTACTGGGTGTAAAGGGCGTGTAGGCGGGAAGGTAAGTCAGATGTGAAATACCG  \\\n",
       "900344                                           0.000000                                                      \n",
       "900459                                           0.000000                                                      \n",
       "900221                                           0.000000                                                      \n",
       "900570                                           0.000000                                                      \n",
       "900092                                           0.000000                                                      \n",
       "...                                                   ...                                                      \n",
       "900294                                           0.000000                                                      \n",
       "900097                                           0.000000                                                      \n",
       "900498                                           0.000000                                                      \n",
       "900276                                           0.015152                                                      \n",
       "900406                                           0.000000                                                      \n",
       "\n",
       "        GTGCCAGCCGCCGCGGTAATACGGAGGATGCGAGCGTTATTCGGAATCATTGGGTTTAAAGGGTCTGTAGGCGGGCTATTAAGTCAGAGGTGAAAGGTTT  \\\n",
       "900344                                           0.000000                                                      \n",
       "900459                                           0.000000                                                      \n",
       "900221                                           0.000000                                                      \n",
       "900570                                           0.000000                                                      \n",
       "900092                                           0.000000                                                      \n",
       "...                                                   ...                                                      \n",
       "900294                                           0.000000                                                      \n",
       "900097                                           0.000000                                                      \n",
       "900498                                           0.000000                                                      \n",
       "900276                                           0.004132                                                      \n",
       "900406                                           0.000000                                                      \n",
       "\n",
       "        GTGCCAGCCGCCGCGGTAAGACGAAGGGGGCTAGCGTTGTTCGGAATTACTGGGCGTAAAGCGCGTGCAGGCGGTTATCCAAGTCGGGTGTGAAAGCCTT  \\\n",
       "900344                                           0.000000                                                      \n",
       "900459                                           0.000000                                                      \n",
       "900221                                           0.000000                                                      \n",
       "900570                                           0.000000                                                      \n",
       "900092                                           0.000000                                                      \n",
       "...                                                   ...                                                      \n",
       "900294                                           0.000000                                                      \n",
       "900097                                           0.000000                                                      \n",
       "900498                                           0.000000                                                      \n",
       "900276                                           0.002755                                                      \n",
       "900406                                           0.000000                                                      \n",
       "\n",
       "        GTCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTTATTGGGCGTAAAGCGAGCGCAGGCGGTTAGATAAGTCTGAAGTTAAAGGCTGT  \n",
       "900344                                           0.000000                                                     \n",
       "900459                                           0.000000                                                     \n",
       "900221                                           0.000000                                                     \n",
       "900570                                           0.000000                                                     \n",
       "900092                                           0.000000                                                     \n",
       "...                                                   ...                                                     \n",
       "900294                                           0.000000                                                     \n",
       "900097                                           0.000000                                                     \n",
       "900498                                           0.000000                                                     \n",
       "900276                                           0.001377                                                     \n",
       "900406                                           0.000000                                                     \n",
       "\n",
       "[462 rows x 2283 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Subset df to only the 462 samples in subsetted metadata\n",
    "df = df[df.index.isin(metadata.index)]\n",
    "# df = df[df.index.isin(metadata.index)].dropna()\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RANDOM FOREST CONFIGURATION\n",
      "RF Configuration: {'n_estimators': 1000, 'random_state': 42, 'n_jobs': -1}\n",
      "Confounders to include: ['age_months', 'sex', 'enrolment_season']\n",
      "================================================================================\n",
      "\n",
      "Available confounders: ['age_months', 'sex', 'enrolment_season']\n",
      "Final confounder features: ['age_months', 'sex_male', 'enrolment_season_Autumn ', 'enrolment_season_Spring', 'enrolment_season_Spring ', 'enrolment_season_Summer', 'enrolment_season_Winter']\n",
      "Categorical variables encoded: ['sex', 'enrolment_season']\n",
      "\n",
      "RUNNING RANDOM FOREST ANALYSES WITH CONFOUNDER ADJUSTMENT\n",
      "\n",
      "1. Analyzing: Skin vs Nares...\n",
      "  Samples: 462\n",
      "  Using confounders: True\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/22/yck9vwx53w1c38tvj_c0_tz00000gn/T/ipykernel_96564/1210162645.py:112: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.\n",
      "The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.\n",
      "\n",
      "For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.\n",
      "\n",
      "\n",
      "  confounders_df[col].fillna(confounders_df[col].median(), inplace=True)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  Mean AUC: 0.952 Â± 0.022\n",
      "\n",
      "2. Analyzing: Skin pairwise comparisons...\n",
      "\n",
      "   skin-ADL vs skin-H...\n",
      "     Samples: 183\n",
      "     Using confounders: True\n",
      "     Mean AUC: 0.820 Â± 0.014\n",
      "\n",
      "   skin-ADNL vs skin-ADL...\n",
      "     Samples: 198\n",
      "     Using confounders: True\n",
      "     Mean AUC: 0.753 Â± 0.017\n",
      "\n",
      "   skin-ADNL vs skin-H...\n",
      "     Samples: 183\n",
      "     Using confounders: True\n",
      "     Mean AUC: 0.764 Â± 0.078\n",
      "\n",
      "3. Analyzing: Nares AD vs H...\n",
      "  Samples: 180\n",
      "  Using confounders: True\n",
      "  Mean AUC: 0.717 Â± 0.028\n",
      "\n",
      "ROC plot saved to: ../Figures/Main/Fig_2A.jpg\n",
      "\n",
      "================================================================================\n",
      "SAVING FEATURE IMPORTANCE TABLES\n",
      "  Saved: feature_importance_skin_vs_nares.csv\n",
      "  Saved: feature_importance_nares_AD_vs_H.csv\n",
      "  Saved: feature_importance_skin-ADL_vs_skin-H.csv\n",
      "  Saved: feature_importance_skin-ADNL_vs_skin-ADL.csv\n",
      "  Saved: feature_importance_skin-ADNL_vs_skin-H.csv\n",
      "ANALYSIS COMPLETE\n",
      "âœ“ Confounders included: ['age_months', 'sex', 'enrolment_season']\n",
      "âœ“ RF Configuration: {'n_estimators': 1000, 'random_state': 42, 'n_jobs': -1}\n",
      "âœ“ ROC curves generated and saved\n",
      "âœ“ Feature importance tables saved\n"
     ]
    }
   ],
   "source": [
    "# =============================================================================\n",
    "# UNIFIED RANDOM FOREST CONFIGURATION\n",
    "# =============================================================================\n",
    "\n",
    "# Define the RF configuration used throughout ALL analyses\n",
    "RF_CONFIG = {\n",
    "    'n_estimators': 1000,\n",
    "    'random_state': 42,\n",
    "    'n_jobs': -1\n",
    "}\n",
    "\n",
    "# Define which confounders to include\n",
    "CONFOUNDER_COLS = ['age_months', 'sex', 'enrolment_season']\n",
    "\n",
    "print(\"RANDOM FOREST CONFIGURATION\")\n",
    "print(f\"RF Configuration: {RF_CONFIG}\")\n",
    "print(f\"Confounders to include: {CONFOUNDER_COLS}\")\n",
    "print(\"=\"*80 + \"\\n\")\n",
    "\n",
    "\n",
    "# =============================================================================\n",
    "# 1. CUSTOM GROUP-STRATIFIED K-FOLD FUNCTION\n",
    "# =============================================================================\n",
    "\n",
    "def group_stratified_kfold(X, y, groups, n_splits=5, random_state=42):\n",
    "    \"\"\"Custom group-stratified k-fold splitting.\"\"\"\n",
    "    unique_groups = np.unique(groups)\n",
    "    np.random.seed(random_state)\n",
    "    np.random.shuffle(unique_groups)\n",
    "    \n",
    "    group_label_dist = {}\n",
    "    for group in unique_groups:\n",
    "        group_mask = groups == group\n",
    "        group_y = y[group_mask]\n",
    "        group_label_dist[group] = {label: np.sum(group_y == label) for label in np.unique(y)}\n",
    "    \n",
    "    folds = [[] for _ in range(n_splits)]\n",
    "    fold_label_dist = [{label: 0 for label in np.unique(y)} for _ in range(n_splits)]\n",
    "    \n",
    "    sorted_groups = sorted(unique_groups, key=lambda g: sum(groups == g), reverse=True)\n",
    "    \n",
    "    for group in sorted_groups:\n",
    "        best_fold = 0\n",
    "        min_imbalance = float('inf')\n",
    "        \n",
    "        for fold_idx in range(n_splits):\n",
    "            temp_fold_dist = fold_label_dist[fold_idx].copy()\n",
    "            for label, count in group_label_dist[group].items():\n",
    "                temp_fold_dist[label] += count\n",
    "            fold_size = sum(temp_fold_dist.values())\n",
    "            proportions = [count / fold_size for count in temp_fold_dist.values()] if fold_size else [0]*len(temp_fold_dist)\n",
    "            imbalance = np.var(proportions) + fold_size / (sum(groups.shape) / n_splits)\n",
    "            \n",
    "            if imbalance < min_imbalance:\n",
    "                min_imbalance = imbalance\n",
    "                best_fold = fold_idx\n",
    "        \n",
    "        folds[best_fold].extend(np.where(groups == group)[0])\n",
    "        for label, count in group_label_dist[group].items():\n",
    "            fold_label_dist[best_fold][label] += count\n",
    "    \n",
    "    train_test_indices = []\n",
    "    for i in range(n_splits):\n",
    "        test_idx = np.array(folds[i])\n",
    "        train_idx = np.concatenate([folds[j] for j in range(n_splits) if j != i])\n",
    "        train_test_indices.append((train_idx, test_idx))\n",
    "    \n",
    "    return train_test_indices\n",
    "\n",
    "\n",
    "# =============================================================================\n",
    "# 2. PREPARE CONFOUNDERS\n",
    "# =============================================================================\n",
    "\n",
    "def prepare_confounders(metadata, confounder_cols):\n",
    "    \"\"\"\n",
    "    Prepare confounder variables for inclusion in models.\n",
    "    \n",
    "    Parameters:\n",
    "    - metadata: DataFrame with metadata\n",
    "    - confounder_cols: List of confounder column names\n",
    "    \n",
    "    Returns:\n",
    "    - confounders_df: DataFrame with encoded and cleaned confounders\n",
    "    - available_confounders: List of available confounder names\n",
    "    \"\"\"\n",
    "    # Check which confounders are available\n",
    "    available_confounders = [col for col in confounder_cols if col in metadata.columns]\n",
    "    \n",
    "    if not available_confounders:\n",
    "        print(\"WARNING: No confounders available in metadata\")\n",
    "        return None, []\n",
    "    \n",
    "    print(f\"Available confounders: {available_confounders}\")\n",
    "    \n",
    "    # Extract confounders\n",
    "    confounders_df = metadata[available_confounders].copy()\n",
    "    \n",
    "    # Encode categorical variables\n",
    "    categorical_cols = []\n",
    "    for col in confounders_df.columns:\n",
    "        if confounders_df[col].dtype == 'object':\n",
    "            categorical_cols.append(col)\n",
    "            # Create dummy variables\n",
    "            dummies = pd.get_dummies(confounders_df[col], prefix=col, drop_first=True)\n",
    "            confounders_df = pd.concat([confounders_df.drop(columns=[col]), dummies], axis=1)\n",
    "    \n",
    "    # Handle missing values (use median for numeric, most common for categorical)\n",
    "    for col in confounders_df.columns:\n",
    "        if confounders_df[col].isna().any():\n",
    "            if confounders_df[col].dtype in ['int64', 'float64']:\n",
    "                confounders_df[col].fillna(confounders_df[col].median(), inplace=True)\n",
    "            else:\n",
    "                confounders_df[col].fillna(confounders_df[col].mode()[0], inplace=True)\n",
    "    \n",
    "    print(f\"Final confounder features: {list(confounders_df.columns)}\")\n",
    "    print(f\"Categorical variables encoded: {categorical_cols}\")\n",
    "    \n",
    "    return confounders_df, available_confounders\n",
    "\n",
    "# Prepare confounders once\n",
    "confounders_full, available_confounder_names = prepare_confounders(metadata, CONFOUNDER_COLS)\n",
    "print()\n",
    "\n",
    "\n",
    "# =============================================================================\n",
    "# 3. FUNCTION TO RUN CV WITH CONFOUNDERS AND GET ROC/FEATURE IMPORTANCE\n",
    "# =============================================================================\n",
    "\n",
    "def run_group_stratified_cv(X, y, groups, confounders=None, n_splits=5, rf_config=None):\n",
    "    \"\"\"\n",
    "    Run group-stratified CV for Random Forest with optional confounder adjustment.\n",
    "    \n",
    "    Parameters:\n",
    "    - X: DataFrame of microbiome features\n",
    "    - y: Series of labels\n",
    "    - groups: Series of group IDs (patient IDs)\n",
    "    - confounders: DataFrame of confounder variables (must have same index as X)\n",
    "    - n_splits: Number of CV folds\n",
    "    - rf_config: Dictionary with RF parameters (uses RF_CONFIG if None)\n",
    "    \n",
    "    Returns:\n",
    "    - cv_results: List of results per fold\n",
    "    - feature_importances: DataFrame of microbiome feature importances only\n",
    "    - fold_aucs: List of AUC values per fold\n",
    "    \"\"\"\n",
    "    if rf_config is None:\n",
    "        rf_config = RF_CONFIG\n",
    "    \n",
    "    folds = group_stratified_kfold(X, y, groups, n_splits=n_splits)\n",
    "    \n",
    "    cv_results = []\n",
    "    feature_importances = pd.DataFrame(index=X.columns)\n",
    "    fold_aucs = []\n",
    "    \n",
    "    for i, (train_idx, test_idx) in enumerate(folds):\n",
    "        X_train, X_test = X.iloc[train_idx], X.iloc[test_idx]\n",
    "        y_train, y_test = y.iloc[train_idx], y.iloc[test_idx]\n",
    "        \n",
    "        if len(np.unique(y_train)) < 2 or len(np.unique(y_test)) < 2:\n",
    "            print(f\"  Skipping fold {i+1} due to insufficient class representation\")\n",
    "            continue\n",
    "        \n",
    "        # Add confounders if provided\n",
    "        if confounders is not None:\n",
    "            conf_train = confounders.iloc[train_idx]\n",
    "            conf_test = confounders.iloc[test_idx]\n",
    "            \n",
    "            # Concatenate microbiome features with confounders\n",
    "            X_train_combined = pd.concat([X_train, conf_train], axis=1)\n",
    "            X_test_combined = pd.concat([X_test, conf_test], axis=1)\n",
    "            \n",
    "            # Train RF on combined features\n",
    "            clf = RandomForestClassifier(**rf_config)\n",
    "            clf.fit(X_train_combined, y_train)\n",
    "            \n",
    "            # Get predictions\n",
    "            probas = clf.predict_proba(X_test_combined)\n",
    "            \n",
    "            # Extract only microbiome feature importances (exclude confounder importances)\n",
    "            n_microbiome_features = X.shape[1]\n",
    "            feature_importances[f'fold_{i}'] = clf.feature_importances_[:n_microbiome_features]\n",
    "            \n",
    "        else:\n",
    "            # Standard RF without confounders\n",
    "            clf = RandomForestClassifier(**rf_config)\n",
    "            clf.fit(X_train, y_train)\n",
    "            probas = clf.predict_proba(X_test)\n",
    "            feature_importances[f'fold_{i}'] = clf.feature_importances_\n",
    "        \n",
    "        # Calculate ROC metrics\n",
    "        fpr, tpr, _ = roc_curve(y_test, probas[:, 1])\n",
    "        roc_auc = auc(fpr, tpr)\n",
    "        fold_aucs.append(roc_auc)\n",
    "        \n",
    "        cv_results.append({\n",
    "            'y_true': y_test,\n",
    "            'y_proba': probas[:, 1],\n",
    "            'fpr': fpr,\n",
    "            'tpr': tpr,\n",
    "            'auc': roc_auc,\n",
    "            'fold': i\n",
    "        })\n",
    "    \n",
    "    feature_importances['mean_importance'] = feature_importances.mean(axis=1)\n",
    "    feature_importances['std_importance'] = feature_importances.std(axis=1)\n",
    "    feature_importances = feature_importances.sort_values('mean_importance', ascending=False)\n",
    "    \n",
    "    return cv_results, feature_importances, fold_aucs\n",
    "\n",
    "\n",
    "# =============================================================================\n",
    "# 4. REGENERATE ROC RESULTS WITH CONFOUNDER ADJUSTMENT\n",
    "# =============================================================================\n",
    "\n",
    "print(\"RUNNING RANDOM FOREST ANALYSES WITH CONFOUNDER ADJUSTMENT\")\n",
    "\n",
    "# 4a. Skin vs Nares (Binary Classification)\n",
    "print(\"\\n1. Analyzing: Skin vs Nares...\")\n",
    "meta_skin_nares = metadata[metadata['group'].str.startswith(('skin', 'nares'))].copy()\n",
    "meta_skin_nares['site_label'] = meta_skin_nares['group'].apply(lambda x: 0 if x.startswith('skin') else 1)\n",
    "common_samples = df.index.intersection(meta_skin_nares.index)\n",
    "\n",
    "X_skin_nares = df.loc[common_samples]\n",
    "X_skin_nares = X_skin_nares.dropna(axis=1)\n",
    "\n",
    "y_skin_nares = meta_skin_nares.loc[common_samples, 'site_label']\n",
    "groups_skin_nares = meta_skin_nares.loc[common_samples, 'pid']\n",
    "y_skin_nares = pd.to_numeric(y_skin_nares, errors='coerce')\n",
    "\n",
    "# Get confounders for this subset\n",
    "conf_skin_nares = confounders_full.loc[common_samples] if confounders_full is not None else None\n",
    "\n",
    "print(f\"  Samples: {len(X_skin_nares)}\")\n",
    "print(f\"  Using confounders: {conf_skin_nares is not None}\")\n",
    "\n",
    "skin_vs_nares_cv_results, skin_vs_nares_fi, skin_vs_nares_aucs = run_group_stratified_cv(\n",
    "    X_skin_nares, y_skin_nares, groups_skin_nares, \n",
    "    confounders=conf_skin_nares, n_splits=3, rf_config=RF_CONFIG\n",
    ")\n",
    "print(f\"  Mean AUC: {np.mean(skin_vs_nares_aucs):.3f} Â± {np.std(skin_vs_nares_aucs):.3f}\")\n",
    "\n",
    "\n",
    "# 4b. Skin Comparisons (Pairwise Among Skin Groups)\n",
    "print(\"\\n2. Analyzing: Skin pairwise comparisons...\")\n",
    "skin_comparisons = [\n",
    "    ('skin-ADL', 'skin-H'),\n",
    "    ('skin-ADNL', 'skin-ADL'),\n",
    "    ('skin-ADNL', 'skin-H')\n",
    "]\n",
    "skin_cv_results_dict = {}\n",
    "skin_feature_importance_dict = {}\n",
    "\n",
    "for label1, label2 in skin_comparisons:\n",
    "    print(f\"\\n   {label1} vs {label2}...\")\n",
    "    meta_subset = metadata[metadata['group'].isin([label1, label2])]\n",
    "    common_samples = df.index.intersection(meta_subset.index)\n",
    "    \n",
    "    X_skin = df.loc[common_samples]\n",
    "    y_skin = meta_subset.loc[common_samples, 'group'].map({label1: 0, label2: 1})\n",
    "    groups_skin = meta_subset.loc[common_samples, 'pid']\n",
    "    \n",
    "    # Get confounders for this subset\n",
    "    conf_skin = confounders_full.loc[common_samples] if confounders_full is not None else None\n",
    "    \n",
    "    print(f\"     Samples: {len(X_skin)}\")\n",
    "    print(f\"     Using confounders: {conf_skin is not None}\")\n",
    "    \n",
    "    cv_results, fi, aucs = run_group_stratified_cv(\n",
    "        X_skin, y_skin, groups_skin, \n",
    "        confounders=conf_skin, n_splits=3, rf_config=RF_CONFIG\n",
    "    )\n",
    "    \n",
    "    key = f\"{label1}_vs_{label2}\"\n",
    "    skin_cv_results_dict[key] = cv_results\n",
    "    skin_feature_importance_dict[key] = fi\n",
    "    print(f\"     Mean AUC: {np.mean(aucs):.3f} Â± {np.std(aucs):.3f}\")\n",
    "\n",
    "\n",
    "# 4c. Nares Comparison (nares-AD vs nares-H)\n",
    "print(\"\\n3. Analyzing: Nares AD vs H...\")\n",
    "meta_nares = metadata[metadata['group'].isin(['nares-AD', 'nares-H'])]\n",
    "common_samples = df.index.intersection(meta_nares.index)\n",
    "\n",
    "X_nares = df.loc[common_samples]\n",
    "y_nares = meta_nares.loc[common_samples, 'group'].map({'nares-AD': 0, 'nares-H': 1})\n",
    "groups_nares = meta_nares.loc[common_samples, 'pid']\n",
    "\n",
    "# Get confounders for this subset\n",
    "conf_nares = confounders_full.loc[common_samples] if confounders_full is not None else None\n",
    "\n",
    "print(f\"  Samples: {len(X_nares)}\")\n",
    "print(f\"  Using confounders: {conf_nares is not None}\")\n",
    "\n",
    "nares_ad_vs_h_cv_results, nares_fi, nares_aucs = run_group_stratified_cv(\n",
    "    X_nares, y_nares, groups_nares, \n",
    "    confounders=conf_nares, n_splits=3, rf_config=RF_CONFIG\n",
    ")\n",
    "print(f\"  Mean AUC: {np.mean(nares_aucs):.3f} Â± {np.std(nares_aucs):.3f}\")\n",
    "\n",
    "\n",
    "# =============================================================================\n",
    "# 5. ASSEMBLE ROC RESULTS DICTIONARY\n",
    "# =============================================================================\n",
    "\n",
    "roc_results_dict = {\n",
    "    'Skin vs. Nares Samples': {\n",
    "        'skin_vs_nares': skin_vs_nares_cv_results\n",
    "    },\n",
    "    'AD Status From Skin Samples': skin_cv_results_dict,\n",
    "    'AD Status From Nares Samples': {\n",
    "        'nares-AD_vs_nares-H': nares_ad_vs_h_cv_results\n",
    "    }\n",
    "}\n",
    "\n",
    "\n",
    "# =============================================================================\n",
    "# 6. COMBINED ROC PLOTTING FUNCTION (3-PANEL)\n",
    "# =============================================================================\n",
    "\n",
    "def plot_combined_roc_3panel(roc_results_dict, comparison_order, color_map, output_path=None):\n",
    "    \"\"\"Plots a 3-panel ROC curve figure from grouped ROC results.\"\"\"\n",
    "    fig, axs = plt.subplots(1, 3, figsize=(18, 6), sharey=True)\n",
    "\n",
    "    pretty_labels = {\n",
    "        'skin_vs_nares': 'All Skin vs All Nares',\n",
    "        'skin-ADL_vs_skin-H': 'Skin ADL vs Skin H',\n",
    "        'skin-ADNL_vs_skin-ADL': 'Skin ADNL vs Skin ADL',\n",
    "        'skin-ADNL_vs_skin-H': 'Skin ADNL vs Skin H',\n",
    "        'nares-AD_vs_nares-H': 'Nares AD vs Nares H'\n",
    "    }\n",
    "    \n",
    "    for i, panel_title in enumerate(comparison_order):\n",
    "        ax = axs[i]\n",
    "        sub_dict = roc_results_dict[panel_title]\n",
    "        for sublabel, curves in sub_dict.items():\n",
    "            if not isinstance(curves, list):\n",
    "                curves = [curves]\n",
    "            mean_fpr = np.linspace(0, 1, 100)\n",
    "            tprs = []\n",
    "            aucs = []\n",
    "            for result in curves:\n",
    "                if 'fpr' not in result or 'tpr' not in result:\n",
    "                    continue\n",
    "                tpr_interp = np.interp(mean_fpr, result['fpr'], result['tpr'])\n",
    "                tpr_interp[0] = 0.0\n",
    "                tprs.append(tpr_interp)\n",
    "                aucs.append(result['auc'])\n",
    "            if len(tprs) == 0:\n",
    "                continue\n",
    "            mean_tpr = np.mean(tprs, axis=0)\n",
    "            mean_tpr[-1] = 1.0\n",
    "            std_tpr = np.std(tprs, axis=0)\n",
    "            tprs_upper = np.minimum(mean_tpr + std_tpr, 1)\n",
    "            tprs_lower = np.maximum(mean_tpr - std_tpr, 0)\n",
    "            ax.plot(mean_fpr, mean_tpr, lw=3,\n",
    "                    label=f'{pretty_labels.get(sublabel, sublabel)} (AUC = {np.mean(aucs):.2f} Â± {np.std(aucs):.2f})',\n",
    "                    color=color_map.get(sublabel, 'gray'))\n",
    "            ax.fill_between(mean_fpr, tprs_lower, tprs_upper, alpha=0.3,\n",
    "                            color=color_map.get(sublabel, 'gray'))\n",
    "    \n",
    "        ax.plot([0, 1], [0, 1], 'k--', lw=1)\n",
    "        ax.set_xlim([0.0, 1.0])\n",
    "        ax.set_ylim([0.0, 1.05])\n",
    "        ax.set_xlabel('False Positive Rate', fontsize=18)\n",
    "        ax.set_title(panel_title, fontsize=20)\n",
    "        ax.grid(True, linestyle='--', alpha=0.7)\n",
    "        ax.tick_params(axis='both', labelsize=14)\n",
    "        ax.legend(loc='lower right', fontsize=12)\n",
    "    \n",
    "    axs[0].set_ylabel('True Positive Rate', fontsize=19)\n",
    "    plt.suptitle(f'Random Forest Classifications by 16S ASVs', \n",
    "                 fontsize=24, x=0.5, y=1.02)\n",
    "    plt.tight_layout()\n",
    "    \n",
    "    if output_path:\n",
    "        os.makedirs(os.path.dirname(output_path), exist_ok=True)\n",
    "        plt.savefig(output_path, dpi=600, bbox_inches='tight')\n",
    "        print(f\"\\nROC plot saved to: {output_path}\")\n",
    "\n",
    "\n",
    "# =============================================================================\n",
    "# 7. DEFINE PANEL ORDER AND COLOR MAP, THEN PLOT\n",
    "# =============================================================================\n",
    "\n",
    "comparison_order = [\n",
    "    'Skin vs. Nares Samples',\n",
    "    'AD Status From Skin Samples',\n",
    "    'AD Status From Nares Samples'\n",
    "]\n",
    "color_map = {\n",
    "    'skin_vs_nares': 'black',\n",
    "    'skin-ADL_vs_skin-H': 'blue',\n",
    "    'skin-ADNL_vs_skin-ADL': 'green',\n",
    "    'skin-ADNL_vs_skin-H': 'purple',\n",
    "    'nares-AD_vs_nares-H': 'orange'\n",
    "}\n",
    "\n",
    "plot_combined_roc_3panel(\n",
    "    roc_results_dict,\n",
    "    comparison_order=comparison_order,\n",
    "    color_map=color_map,\n",
    "    output_path='../Figures/Main/Fig_2A.jpg'\n",
    ")\n",
    "\n",
    "\n",
    "# =============================================================================\n",
    "# 8. SAVE FEATURE IMPORTANCE RESULTS\n",
    "# =============================================================================\n",
    "\n",
    "print(\"\\n\" + \"=\"*80)\n",
    "print(\"SAVING FEATURE IMPORTANCE TABLES\")\n",
    "\n",
    "\n",
    "# Create directory if it doesn't exist\n",
    "os.makedirs('../Data/RF_Feature_Importances', exist_ok=True)\n",
    "\n",
    "# Save feature importance for each comparison\n",
    "skin_vs_nares_fi.to_csv('../Data/RF_Feature_Importances/feature_importance_skin_vs_nares.csv')\n",
    "print(\"  Saved: feature_importance_skin_vs_nares.csv\")\n",
    "\n",
    "nares_fi.to_csv('../Data/RF_Feature_Importances/feature_importance_nares_AD_vs_H.csv')\n",
    "print(\"  Saved: feature_importance_nares_AD_vs_H.csv\")\n",
    "\n",
    "for comparison_key, fi in skin_feature_importance_dict.items():\n",
    "    fi.to_csv(f'../Data/RF_Feature_Importances/feature_importance_{comparison_key}.csv')\n",
    "    print(f\"  Saved: feature_importance_{comparison_key}.csv\")\n",
    "\n",
    "print(\"ANALYSIS COMPLETE\")\n",
    "if confounders_full is not None:\n",
    "    print(f\"âœ“ Confounders included: {available_confounder_names}\")\n",
    "else:\n",
    "    print(\"âœ“ No confounders available/used\")\n",
    "\n",
    "print(f\"âœ“ RF Configuration: {RF_CONFIG}\")\n",
    "print(\"âœ“ ROC curves generated and saved\")\n",
    "print(\"âœ“ Feature importance tables saved\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Comparison of different ML models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.svm import SVC\n",
    "from xgboost import XGBClassifier\n",
    "from lightgbm import LGBMClassifier"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RANDOM FOREST CONFIGURATION\n",
      "Configuration: {'n_estimators': 1000, 'random_state': 42, 'n_jobs': -1}\n",
      "This configuration will be used for:\n",
      "  1. Main RF analysis (ROC curves, feature importance)\n",
      "  2. Model comparison benchmark\n",
      "================================================================================\n",
      "\n",
      "\n",
      "================================================================================\n",
      "PART 1: MAIN RANDOM FOREST ANALYSIS\n",
      "\n",
      "Analyzing: Skin vs Nares...\n",
      "  Mean AUC: 0.949 Â± 0.021\n",
      "\n",
      "Analyzing: Skin pairwise comparisons...\n",
      "  skin-ADL vs skin-H...\n",
      "    Mean AUC: 0.790 Â± 0.018\n",
      "  skin-ADNL vs skin-ADL...\n",
      "    Mean AUC: 0.751 Â± 0.011\n",
      "  skin-ADNL vs skin-H...\n",
      "    Mean AUC: 0.689 Â± 0.076\n",
      "\n",
      "Analyzing: Nares AD vs H...\n",
      "  Mean AUC: 0.646 Â± 0.064\n",
      "Saving feature importance tables...\n",
      "Feature importance tables saved!\n",
      "\n",
      "ROC plot saved to: ../Figures/Main/Fig_2A.jpg\n",
      "PART 2: MODEL COMPARISON\n",
      "Comparing RF (with identical config) against alternative ML methods...\n",
      "MODEL COMPARISON: Skin vs. Nares\n",
      "  Training Random Forest...\n",
      "  Training XGBoost...\n",
      "  Training LightGBM...\n",
      "  Training SVM...\n",
      "\n",
      "Results:\n",
      "        Model  AUC-ROC  AUC_std  Accuracy  Acc_std\n",
      "      XGBoost    0.955    0.019     0.911    0.013\n",
      "     LightGBM    0.951    0.022     0.905    0.008\n",
      "Random Forest    0.949    0.021     0.922    0.023\n",
      "          SVM    0.935    0.007     0.905    0.011\n",
      "MODEL COMPARISON: skin-ADL vs skin-H\n",
      "  Training Random Forest...\n",
      "  Training XGBoost...\n",
      "  Training LightGBM...\n",
      "  Training SVM...\n",
      "\n",
      "Results:\n",
      "        Model  AUC-ROC  AUC_std  Accuracy  Acc_std\n",
      "Random Forest    0.790    0.018     0.694    0.020\n",
      "          SVM    0.771    0.020     0.689    0.084\n",
      "      XGBoost    0.761    0.009     0.689    0.000\n",
      "     LightGBM    0.718    0.022     0.667    0.020\n",
      "MODEL COMPARISON: skin-ADNL vs skin-ADL\n",
      "  Training Random Forest...\n",
      "  Training XGBoost...\n",
      "  Training LightGBM...\n",
      "  Training SVM...\n",
      "\n",
      "Results:\n",
      "        Model  AUC-ROC  AUC_std  Accuracy  Acc_std\n",
      "Random Forest    0.751    0.011     0.697    0.021\n",
      "      XGBoost    0.710    0.053     0.672    0.014\n",
      "          SVM    0.691    0.069     0.652    0.037\n",
      "     LightGBM    0.676    0.063     0.646    0.052\n",
      "MODEL COMPARISON: skin-ADNL vs skin-H\n",
      "  Training Random Forest...\n",
      "  Training XGBoost...\n",
      "  Training LightGBM...\n",
      "  Training SVM...\n",
      "\n",
      "Results:\n",
      "        Model  AUC-ROC  AUC_std  Accuracy  Acc_std\n",
      "          SVM    0.717    0.065     0.656    0.054\n",
      "Random Forest    0.689    0.076     0.667    0.041\n",
      "      XGBoost    0.677    0.088     0.639    0.075\n",
      "     LightGBM    0.662    0.093     0.623    0.084\n",
      "MODEL COMPARISON: Nares AD vs H\n",
      "  Training Random Forest...\n",
      "  Training XGBoost...\n",
      "  Training LightGBM...\n",
      "  Training SVM...\n",
      "\n",
      "Results:\n",
      "        Model  AUC-ROC  AUC_std  Accuracy  Acc_std\n",
      "Random Forest    0.646    0.064     0.594    0.034\n",
      "      XGBoost    0.645    0.029     0.644    0.048\n",
      "     LightGBM    0.612    0.034     0.589    0.008\n",
      "          SVM    0.459    0.075     0.600    0.014\n",
      "COMPREHENSIVE MODEL COMPARISON SUMMARY\n",
      "           Comparison         Model       AUC-ROC      Accuracy\n",
      "        Skin vs Nares       XGBoost 0.955 Â± 0.019 0.911 Â± 0.013\n",
      "        Skin vs Nares      LightGBM 0.951 Â± 0.022 0.905 Â± 0.008\n",
      "        Skin vs Nares Random Forest 0.949 Â± 0.021 0.922 Â± 0.023\n",
      "        Skin vs Nares           SVM 0.935 Â± 0.007 0.905 Â± 0.011\n",
      "   skin-ADL vs skin-H Random Forest 0.790 Â± 0.018 0.694 Â± 0.020\n",
      "   skin-ADL vs skin-H           SVM 0.771 Â± 0.020 0.689 Â± 0.084\n",
      "   skin-ADL vs skin-H       XGBoost 0.761 Â± 0.009 0.689 Â± 0.000\n",
      "   skin-ADL vs skin-H      LightGBM 0.718 Â± 0.022 0.667 Â± 0.020\n",
      "skin-ADNL vs skin-ADL Random Forest 0.751 Â± 0.011 0.697 Â± 0.021\n",
      "skin-ADNL vs skin-ADL       XGBoost 0.710 Â± 0.053 0.672 Â± 0.014\n",
      "skin-ADNL vs skin-ADL           SVM 0.691 Â± 0.069 0.652 Â± 0.037\n",
      "skin-ADNL vs skin-ADL      LightGBM 0.676 Â± 0.063 0.646 Â± 0.052\n",
      "  skin-ADNL vs skin-H           SVM 0.717 Â± 0.065 0.656 Â± 0.054\n",
      "  skin-ADNL vs skin-H Random Forest 0.689 Â± 0.076 0.667 Â± 0.041\n",
      "  skin-ADNL vs skin-H       XGBoost 0.677 Â± 0.088 0.639 Â± 0.075\n",
      "  skin-ADNL vs skin-H      LightGBM 0.662 Â± 0.093 0.623 Â± 0.084\n",
      "        Nares AD vs H Random Forest 0.646 Â± 0.064 0.594 Â± 0.034\n",
      "        Nares AD vs H       XGBoost 0.645 Â± 0.029 0.644 Â± 0.048\n",
      "        Nares AD vs H      LightGBM 0.612 Â± 0.034 0.589 Â± 0.008\n",
      "        Nares AD vs H           SVM 0.459 Â± 0.075 0.600 Â± 0.014\n",
      "\n",
      "Summary saved to: ../Data/ML_comparison/model_comparison_summary.csv\n",
      "\n",
      "Heatmap saved to: ../Figures/Supplementary/Suppl_Fig_2.jpg\n",
      "\n",
      "================================================================================\n",
      "ANALYSIS COMPLETE\n",
      "RF Configuration used throughout: {'n_estimators': 1000, 'random_state': 42, 'n_jobs': -1}\n",
      "âœ“ Main RF analysis complete (ROC curves + feature importance)\n",
      "âœ“ Model comparison complete (RF vs 6 alternatives)\n",
      "âœ“ All results saved\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/22/yck9vwx53w1c38tvj_c0_tz00000gn/T/ipykernel_96564/4044814564.py:533: UserWarning: Matplotlib is currently using agg, which is a non-GUI backend, so cannot show the figure.\n",
      "  plt.show()\n"
     ]
    }
   ],
   "source": [
    "# =============================================================================\n",
    "# UNIFIED RANDOM FOREST CONFIGURATION (this takes about 2 mins)\n",
    "# =============================================================================\n",
    "\n",
    "# Define the RF configuration used throughout ALL analyses\n",
    "RF_CONFIG = {\n",
    "    'n_estimators': 1000,\n",
    "    'random_state': 42,\n",
    "    'n_jobs': -1\n",
    "}\n",
    "\n",
    "\n",
    "print(\"RANDOM FOREST CONFIGURATION\")\n",
    "\n",
    "print(f\"Configuration: {RF_CONFIG}\")\n",
    "print(\"This configuration will be used for:\")\n",
    "print(\"  1. Main RF analysis (ROC curves, feature importance)\")\n",
    "print(\"  2. Model comparison benchmark\")\n",
    "print(\"=\"*80 + \"\\n\")\n",
    "\n",
    "\n",
    "# =============================================================================\n",
    "# PART 1: GROUP-STRATIFIED CV FUNCTIONS\n",
    "# =============================================================================\n",
    "\n",
    "def group_stratified_kfold(X, y, groups, n_splits=5, random_state=42):\n",
    "    \"\"\"Custom group-stratified k-fold splitting.\"\"\"\n",
    "    unique_groups = np.unique(groups)\n",
    "    np.random.seed(random_state)\n",
    "    np.random.shuffle(unique_groups)\n",
    "    \n",
    "    group_label_dist = {}\n",
    "    for group in unique_groups:\n",
    "        group_mask = groups == group\n",
    "        group_y = y[group_mask]\n",
    "        group_label_dist[group] = {label: np.sum(group_y == label) for label in np.unique(y)}\n",
    "    \n",
    "    folds = [[] for _ in range(n_splits)]\n",
    "    fold_label_dist = [{label: 0 for label in np.unique(y)} for _ in range(n_splits)]\n",
    "    \n",
    "    sorted_groups = sorted(unique_groups, key=lambda g: sum(groups == g), reverse=True)\n",
    "    \n",
    "    for group in sorted_groups:\n",
    "        best_fold = 0\n",
    "        min_imbalance = float('inf')\n",
    "        \n",
    "        for fold_idx in range(n_splits):\n",
    "            temp_fold_dist = fold_label_dist[fold_idx].copy()\n",
    "            for label, count in group_label_dist[group].items():\n",
    "                temp_fold_dist[label] += count\n",
    "            fold_size = sum(temp_fold_dist.values())\n",
    "            proportions = [count / fold_size for count in temp_fold_dist.values()] if fold_size else [0]*len(temp_fold_dist)\n",
    "            imbalance = np.var(proportions) + fold_size / (sum(groups.shape) / n_splits)\n",
    "            \n",
    "            if imbalance < min_imbalance:\n",
    "                min_imbalance = imbalance\n",
    "                best_fold = fold_idx\n",
    "        \n",
    "        folds[best_fold].extend(np.where(groups == group)[0])\n",
    "        for label, count in group_label_dist[group].items():\n",
    "            fold_label_dist[best_fold][label] += count\n",
    "    \n",
    "    train_test_indices = []\n",
    "    for i in range(n_splits):\n",
    "        test_idx = np.array(folds[i])\n",
    "        train_idx = np.concatenate([folds[j] for j in range(n_splits) if j != i])\n",
    "        train_test_indices.append((train_idx, test_idx))\n",
    "    \n",
    "    return train_test_indices\n",
    "\n",
    "\n",
    "def run_group_stratified_cv(X, y, groups, n_splits=5, rf_config=None):\n",
    "    \"\"\"\n",
    "    Run group-stratified CV for Random Forest with feature importance tracking.\n",
    "    \n",
    "    Parameters:\n",
    "    - X: DataFrame of features\n",
    "    - y: Series of labels\n",
    "    - groups: Series of group IDs (patient IDs)\n",
    "    - n_splits: Number of CV folds\n",
    "    - rf_config: Dictionary with RF parameters (uses RF_CONFIG if None)\n",
    "    \n",
    "    Returns:\n",
    "    - cv_results: List of results per fold\n",
    "    - feature_importances: DataFrame of feature importances\n",
    "    - fold_aucs: List of AUC values per fold\n",
    "    \"\"\"\n",
    "    if rf_config is None:\n",
    "        rf_config = RF_CONFIG\n",
    "    \n",
    "    folds = group_stratified_kfold(X, y, groups, n_splits=n_splits)\n",
    "    \n",
    "    cv_results = []\n",
    "    feature_importances = pd.DataFrame(index=X.columns)\n",
    "    fold_aucs = []\n",
    "    \n",
    "    for i, (train_idx, test_idx) in enumerate(folds):\n",
    "        X_train, X_test = X.iloc[train_idx], X.iloc[test_idx]\n",
    "        y_train, y_test = y.iloc[train_idx], y.iloc[test_idx]\n",
    "        \n",
    "        if len(np.unique(y_train)) < 2 or len(np.unique(y_test)) < 2:\n",
    "            print(f\"Skipping fold {i+1} due to insufficient class representation\")\n",
    "            continue\n",
    "        \n",
    "        # Use the standardized RF configuration\n",
    "        clf = RandomForestClassifier(**rf_config)\n",
    "        clf.fit(X_train, y_train)\n",
    "        probas = clf.predict_proba(X_test)\n",
    "        \n",
    "        feature_importances[f'fold_{i}'] = clf.feature_importances_\n",
    "        fpr, tpr, _ = roc_curve(y_test, probas[:, 1])\n",
    "        roc_auc = auc(fpr, tpr)\n",
    "        fold_aucs.append(roc_auc)\n",
    "        \n",
    "        cv_results.append({\n",
    "            'y_true': y_test,\n",
    "            'y_proba': probas[:, 1],\n",
    "            'fpr': fpr,\n",
    "            'tpr': tpr,\n",
    "            'auc': roc_auc,\n",
    "            'fold': i\n",
    "        })\n",
    "    \n",
    "    feature_importances['mean_importance'] = feature_importances.mean(axis=1)\n",
    "    feature_importances['std_importance'] = feature_importances.std(axis=1)\n",
    "    feature_importances = feature_importances.sort_values('mean_importance', ascending=False)\n",
    "    \n",
    "    return cv_results, feature_importances, fold_aucs\n",
    "\n",
    "\n",
    "def run_group_stratified_cv_multimodel(X, y, groups, models_dict, n_splits=5):\n",
    "    \"\"\"\n",
    "    Run group-stratified CV for multiple models and return comparative results.\n",
    "    \n",
    "    Parameters:\n",
    "    - X: DataFrame of features\n",
    "    - y: Series of labels\n",
    "    - groups: Series of group IDs (e.g., patient IDs)\n",
    "    - models_dict: Dictionary of model name -> model object\n",
    "    - n_splits: Number of CV folds\n",
    "    \n",
    "    Returns:\n",
    "    - results_summary: DataFrame with performance metrics for each model\n",
    "    - detailed_results: Dictionary with detailed results for each model\n",
    "    \"\"\"\n",
    "    \n",
    "    folds = group_stratified_kfold(X, y, groups, n_splits=n_splits)\n",
    "    \n",
    "    results_summary = []\n",
    "    detailed_results = {}\n",
    "    \n",
    "    for model_name, model in models_dict.items():\n",
    "        print(f\"  Training {model_name}...\")\n",
    "        \n",
    "        fold_aucs = []\n",
    "        fold_accuracies = []\n",
    "        cv_results = []\n",
    "        \n",
    "        for i, (train_idx, test_idx) in enumerate(folds):\n",
    "            X_train, X_test = X.iloc[train_idx], X.iloc[test_idx]\n",
    "            y_train, y_test = y.iloc[train_idx], y.iloc[test_idx]\n",
    "            \n",
    "            # Skip fold if insufficient class representation\n",
    "            if len(np.unique(y_train)) < 2 or len(np.unique(y_test)) < 2:\n",
    "                print(f\"    Skipping fold {i+1} for {model_name} due to insufficient class representation\")\n",
    "                continue\n",
    "            \n",
    "            # Train model\n",
    "            model.fit(X_train, y_train)\n",
    "            \n",
    "            # Get predictions\n",
    "            probas = model.predict_proba(X_test)\n",
    "            y_pred = model.predict(X_test)\n",
    "            \n",
    "            # Calculate metrics\n",
    "            fpr, tpr, _ = roc_curve(y_test, probas[:, 1])\n",
    "            roc_auc = auc(fpr, tpr)\n",
    "            accuracy = accuracy_score(y_test, y_pred)\n",
    "            \n",
    "            fold_aucs.append(roc_auc)\n",
    "            fold_accuracies.append(accuracy)\n",
    "            \n",
    "            cv_results.append({\n",
    "                'y_true': y_test,\n",
    "                'y_proba': probas[:, 1],\n",
    "                'y_pred': y_pred,\n",
    "                'fpr': fpr,\n",
    "                'tpr': tpr,\n",
    "                'auc': roc_auc,\n",
    "                'accuracy': accuracy,\n",
    "            })\n",
    "        \n",
    "        # Calculate summary statistics\n",
    "        if len(cv_results) > 0:\n",
    "            results_summary.append({\n",
    "                'Model': model_name,\n",
    "                'AUC-ROC': np.mean(fold_aucs),\n",
    "                'AUC_std': np.std(fold_aucs),\n",
    "                'Accuracy': np.mean(fold_accuracies),\n",
    "                'Acc_std': np.std(fold_accuracies),\n",
    "            })\n",
    "            \n",
    "            detailed_results[model_name] = cv_results\n",
    "    \n",
    "    results_df = pd.DataFrame(results_summary)\n",
    "    results_df = results_df.sort_values('AUC-ROC', ascending=False)\n",
    "    \n",
    "    return results_df, detailed_results\n",
    "\n",
    "\n",
    "# =============================================================================\n",
    "# PART 2: MAIN RF ANALYSIS (ROC CURVES + FEATURE IMPORTANCE)\n",
    "# =============================================================================\n",
    "\n",
    "print(\"\\n\" + \"=\"*80)\n",
    "print(\"PART 1: MAIN RANDOM FOREST ANALYSIS\")\n",
    "\n",
    "\n",
    "# 3a. Skin vs Nares (Binary Classification)\n",
    "print(\"\\nAnalyzing: Skin vs Nares...\")\n",
    "meta_skin_nares = metadata[metadata['group'].str.startswith(('skin', 'nares'))].copy()\n",
    "meta_skin_nares['site_label'] = meta_skin_nares['group'].apply(lambda x: 0 if x.startswith('skin') else 1)\n",
    "common_samples = df.index.intersection(meta_skin_nares.index)\n",
    "X_skin_nares = df.loc[common_samples]\n",
    "y_skin_nares = meta_skin_nares.loc[common_samples, 'site_label']\n",
    "groups_skin_nares = meta_skin_nares.loc[common_samples, 'pid']\n",
    "y_skin_nares = pd.to_numeric(y_skin_nares, errors='coerce')\n",
    "\n",
    "skin_vs_nares_cv_results, skin_vs_nares_fi, skin_vs_nares_aucs = run_group_stratified_cv(\n",
    "    X_skin_nares, y_skin_nares, groups_skin_nares, n_splits=3, rf_config=RF_CONFIG\n",
    ")\n",
    "print(f\"  Mean AUC: {np.mean(skin_vs_nares_aucs):.3f} Â± {np.std(skin_vs_nares_aucs):.3f}\")\n",
    "\n",
    "# 3b. Skin Comparisons (Pairwise Among Skin Groups)\n",
    "print(\"\\nAnalyzing: Skin pairwise comparisons...\")\n",
    "skin_comparisons = [\n",
    "    ('skin-ADL', 'skin-H'),\n",
    "    ('skin-ADNL', 'skin-ADL'),\n",
    "    ('skin-ADNL', 'skin-H')\n",
    "]\n",
    "skin_cv_results_dict = {}\n",
    "skin_feature_importance_dict = {}\n",
    "\n",
    "for label1, label2 in skin_comparisons:\n",
    "    print(f\"  {label1} vs {label2}...\")\n",
    "    meta_subset = metadata[metadata['group'].isin([label1, label2])]\n",
    "    common_samples = df.index.intersection(meta_subset.index)\n",
    "    X_skin = df.loc[common_samples]\n",
    "    y_skin = meta_subset.loc[common_samples, 'group'].map({label1: 0, label2: 1})\n",
    "    groups_skin = meta_subset.loc[common_samples, 'pid']\n",
    "    \n",
    "    cv_results, fi, aucs = run_group_stratified_cv(\n",
    "        X_skin, y_skin, groups_skin, n_splits=3, rf_config=RF_CONFIG\n",
    "    )\n",
    "    key = f\"{label1}_vs_{label2}\"\n",
    "    skin_cv_results_dict[key] = cv_results\n",
    "    skin_feature_importance_dict[key] = fi\n",
    "    print(f\"    Mean AUC: {np.mean(aucs):.3f} Â± {np.std(aucs):.3f}\")\n",
    "\n",
    "# 3c. Nares Comparison (nares-AD vs nares-H)\n",
    "print(\"\\nAnalyzing: Nares AD vs H...\")\n",
    "meta_nares = metadata[metadata['group'].isin(['nares-AD', 'nares-H'])]\n",
    "common_samples = df.index.intersection(meta_nares.index)\n",
    "X_nares = df.loc[common_samples]\n",
    "y_nares = meta_nares.loc[common_samples, 'group'].map({'nares-AD': 0, 'nares-H': 1})\n",
    "groups_nares = meta_nares.loc[common_samples, 'pid']\n",
    "\n",
    "nares_ad_vs_h_cv_results, nares_fi, nares_aucs = run_group_stratified_cv(\n",
    "    X_nares, y_nares, groups_nares, n_splits=3, rf_config=RF_CONFIG\n",
    ")\n",
    "print(f\"  Mean AUC: {np.mean(nares_aucs):.3f} Â± {np.std(nares_aucs):.3f}\")\n",
    "\n",
    "# Assemble ROC Results Dictionary\n",
    "roc_results_dict = {\n",
    "    'Skin vs. Nares Samples': {\n",
    "        'skin_vs_nares': skin_vs_nares_cv_results\n",
    "    },\n",
    "    'AD Status From Skin Samples': skin_cv_results_dict,\n",
    "    'AD Status From Nares Samples': {\n",
    "        'nares-AD_vs_nares-H': nares_ad_vs_h_cv_results\n",
    "    }\n",
    "}\n",
    "\n",
    "# Save feature importance results\n",
    "print(\"Saving feature importance tables...\")\n",
    "\n",
    "os.makedirs('../Data/RF_Feature_Importances', exist_ok=True)\n",
    "skin_vs_nares_fi.to_csv('../Data/RF_Feature_Importances/feature_importance_skin_vs_nares.csv')\n",
    "nares_fi.to_csv('../Data/RF_Feature_Importances/feature_importance_nares_AD_vs_H.csv')\n",
    "\n",
    "for comparison_key, fi in skin_feature_importance_dict.items():\n",
    "    fi.to_csv(f'../Data/RF_Feature_Importances/feature_importance_{comparison_key}.csv')\n",
    "\n",
    "print(\"Feature importance tables saved!\")\n",
    "\n",
    "\n",
    "# =============================================================================\n",
    "# PART 3: ROC CURVE PLOTTING\n",
    "# =============================================================================\n",
    "\n",
    "def plot_combined_roc_3panel(roc_results_dict, comparison_order, color_map, output_path=None):\n",
    "    \"\"\"Plots a 3-panel ROC curve figure from grouped ROC results.\"\"\"\n",
    "    fig, axs = plt.subplots(1, 3, figsize=(18, 6), sharey=True)\n",
    "\n",
    "    pretty_labels = {\n",
    "        'skin_vs_nares': 'All Skin vs All Nares',\n",
    "        'skin-ADL_vs_skin-H': 'Skin ADL vs Skin H',\n",
    "        'skin-ADNL_vs_skin-ADL': 'Skin ADNL vs Skin ADL',\n",
    "        'skin-ADNL_vs_skin-H': 'Skin ADNL vs Skin H',\n",
    "        'nares-AD_vs_nares-H': 'Nares AD vs Nares H'\n",
    "    }\n",
    "    \n",
    "    for i, panel_title in enumerate(comparison_order):\n",
    "        ax = axs[i]\n",
    "        sub_dict = roc_results_dict[panel_title]\n",
    "        for sublabel, curves in sub_dict.items():\n",
    "            if not isinstance(curves, list):\n",
    "                curves = [curves]\n",
    "            mean_fpr = np.linspace(0, 1, 100)\n",
    "            tprs = []\n",
    "            aucs = []\n",
    "            for result in curves:\n",
    "                if 'fpr' not in result or 'tpr' not in result:\n",
    "                    continue\n",
    "                tpr_interp = np.interp(mean_fpr, result['fpr'], result['tpr'])\n",
    "                tpr_interp[0] = 0.0\n",
    "                tprs.append(tpr_interp)\n",
    "                aucs.append(result['auc'])\n",
    "            if len(tprs) == 0:\n",
    "                continue\n",
    "            mean_tpr = np.mean(tprs, axis=0)\n",
    "            mean_tpr[-1] = 1.0\n",
    "            std_tpr = np.std(tprs, axis=0)\n",
    "            tprs_upper = np.minimum(mean_tpr + std_tpr, 1)\n",
    "            tprs_lower = np.maximum(mean_tpr - std_tpr, 0)\n",
    "            ax.plot(mean_fpr, mean_tpr, lw=3,\n",
    "                    label=f'{pretty_labels.get(sublabel, sublabel)} (AUC = {np.mean(aucs):.2f} Â± {np.std(aucs):.2f})',\n",
    "                    color=color_map.get(sublabel, 'gray'))\n",
    "            ax.fill_between(mean_fpr, tprs_lower, tprs_upper, alpha=0.3,\n",
    "                            color=color_map.get(sublabel, 'gray'))\n",
    "    \n",
    "        ax.plot([0, 1], [0, 1], 'k--', lw=1)\n",
    "        ax.set_xlim([0.0, 1.0])\n",
    "        ax.set_ylim([0.0, 1.05])\n",
    "        ax.set_xlabel('False Positive Rate', fontsize=18)\n",
    "        ax.set_title(panel_title, fontsize=20)\n",
    "        ax.grid(True, linestyle='--', alpha=0.7)\n",
    "        ax.tick_params(axis='both', labelsize=14)\n",
    "        ax.legend(loc='lower right', fontsize=12)\n",
    "    \n",
    "    axs[0].set_ylabel('True Positive Rate', fontsize=19)\n",
    "    plt.suptitle(f'Random Forest Classifications by 16S ASVs', \n",
    "                 fontsize=24, x=0.5, y=1.02)\n",
    "    plt.tight_layout()\n",
    "    \n",
    "    if output_path:\n",
    "        os.makedirs(os.path.dirname(output_path), exist_ok=True)\n",
    "        plt.savefig(output_path, dpi=600, bbox_inches='tight')\n",
    "        print(f\"\\nROC plot saved to: {output_path}\")\n",
    "\n",
    "comparison_order = [\n",
    "    'Skin vs. Nares Samples',\n",
    "    'AD Status From Skin Samples',\n",
    "    'AD Status From Nares Samples'\n",
    "]\n",
    "color_map = {\n",
    "    'skin_vs_nares': 'black',\n",
    "    'skin-ADL_vs_skin-H': 'blue',\n",
    "    'skin-ADNL_vs_skin-ADL': 'green',\n",
    "    'skin-ADNL_vs_skin-H': 'purple',\n",
    "    'nares-AD_vs_nares-H': 'orange'\n",
    "}\n",
    "\n",
    "plot_combined_roc_3panel(\n",
    "    roc_results_dict,\n",
    "    comparison_order=comparison_order,\n",
    "    color_map=color_map,\n",
    "    output_path='../Figures/Main/Fig_2A.jpg'\n",
    ")\n",
    "\n",
    "\n",
    "# =============================================================================\n",
    "# PART 4: MODEL COMPARISON (Using SAME RF configuration)\n",
    "# =============================================================================\n",
    "\n",
    "print(\"PART 2: MODEL COMPARISON\")\n",
    "print(\"Comparing RF (with identical config) against alternative ML methods...\")\n",
    "\n",
    "# Define models to compare - NOTE: RF uses the SAME config as main analysis\n",
    "models_to_compare = {\n",
    "    'Random Forest': RandomForestClassifier(**RF_CONFIG),  # SAME AS MAIN ANALYSIS\n",
    "    \n",
    "    'XGBoost': XGBClassifier(\n",
    "        n_estimators=1000,\n",
    "        random_state=42,\n",
    "        learning_rate=0.1,\n",
    "        eval_metric='logloss',\n",
    "        n_jobs=-1\n",
    "    ),\n",
    "    \n",
    "    'LightGBM': LGBMClassifier(\n",
    "        n_estimators=1000,\n",
    "        random_state=42,\n",
    "        learning_rate=0.1,\n",
    "        verbose=-1,\n",
    "        n_jobs=-1\n",
    "    ),\n",
    "    \n",
    "    \n",
    "    'SVM': SVC(\n",
    "        probability=True,\n",
    "        random_state=42,\n",
    "        kernel='rbf'\n",
    "    )\n",
    "}\n",
    "\n",
    "# Dictionary to store all comparison results\n",
    "all_model_comparisons = {}\n",
    "\n",
    "# 1. Skin vs Nares Comparison\n",
    "print(\"MODEL COMPARISON: Skin vs. Nares\")\n",
    "skin_nares_comparison, _ = run_group_stratified_cv_multimodel(\n",
    "    X_skin_nares, y_skin_nares, groups_skin_nares, \n",
    "    models_to_compare, n_splits=3\n",
    ")\n",
    "all_model_comparisons['Skin vs Nares'] = skin_nares_comparison\n",
    "print(\"\\nResults:\")\n",
    "print(skin_nares_comparison.round(3).to_string(index=False))\n",
    "\n",
    "# 2. Skin Pairwise Comparisons\n",
    "for label1, label2 in skin_comparisons:\n",
    "    print(f\"MODEL COMPARISON: {label1} vs {label2}\")\n",
    "    \n",
    "    meta_subset = metadata[metadata['group'].isin([label1, label2])]\n",
    "    common_samples = df.index.intersection(meta_subset.index)\n",
    "    X_skin = df.loc[common_samples]\n",
    "    y_skin = meta_subset.loc[common_samples, 'group'].map({label1: 0, label2: 1})\n",
    "    groups_skin = meta_subset.loc[common_samples, 'pid']\n",
    "    \n",
    "    comparison_results, _ = run_group_stratified_cv_multimodel(\n",
    "        X_skin, y_skin, groups_skin,\n",
    "        models_to_compare, n_splits=3\n",
    "    )\n",
    "    \n",
    "    key = f\"{label1} vs {label2}\"\n",
    "    all_model_comparisons[key] = comparison_results\n",
    "    print(\"\\nResults:\")\n",
    "    print(comparison_results.round(3).to_string(index=False))\n",
    "\n",
    "# 3. Nares AD vs H Comparison\n",
    "print(\"MODEL COMPARISON: Nares AD vs H\")\n",
    "nares_comparison, _ = run_group_stratified_cv_multimodel(\n",
    "    X_nares, y_nares, groups_nares,\n",
    "    models_to_compare, n_splits=3\n",
    ")\n",
    "all_model_comparisons['Nares AD vs H'] = nares_comparison\n",
    "print(\"\\nResults:\")\n",
    "print(nares_comparison.round(3).to_string(index=False))\n",
    "\n",
    "\n",
    "# =============================================================================\n",
    "# PART 5: SAVE AND VISUALIZE MODEL COMPARISON\n",
    "# =============================================================================\n",
    "\n",
    "def create_comparison_summary_table(all_comparisons):\n",
    "    \"\"\"Create a comprehensive summary table across all comparisons.\"\"\"\n",
    "    summary_data = []\n",
    "    \n",
    "    for comparison_name, results_df in all_comparisons.items():\n",
    "        for _, row in results_df.iterrows():\n",
    "            summary_data.append({\n",
    "                'Comparison': comparison_name,\n",
    "                'Model': row['Model'],\n",
    "                'AUC-ROC': f\"{row['AUC-ROC']:.3f} Â± {row['AUC_std']:.3f}\",\n",
    "                'Accuracy': f\"{row['Accuracy']:.3f} Â± {row['Acc_std']:.3f}\",\n",
    "            })\n",
    "    \n",
    "    summary_df = pd.DataFrame(summary_data)\n",
    "    return summary_df\n",
    "\n",
    "summary_table = create_comparison_summary_table(all_model_comparisons)\n",
    "print(\"COMPREHENSIVE MODEL COMPARISON SUMMARY\")\n",
    "print(summary_table.to_string(index=False))\n",
    "\n",
    "# Save to CSV\n",
    "os.makedirs('../Data/ML_comparison', exist_ok=True)\n",
    "summary_table.to_csv('../Data/ML_comparison/model_comparison_summary.csv', index=False)\n",
    "print(\"\\nSummary saved to: ../Data/ML_comparison/model_comparison_summary.csv\")\n",
    "\n",
    "\n",
    "# Heatmap visualization\n",
    "def plot_model_comparison(all_comparisons, output_path=None):\n",
    "    \"\"\"Create a heatmap showing AUC-ROC for all models across comparisons.\"\"\"\n",
    "    \n",
    "    models = list(models_to_compare.keys())\n",
    "    comparisons = list(all_comparisons.keys())\n",
    "    \n",
    "    data_matrix = np.zeros((len(models), len(comparisons)))\n",
    "    \n",
    "    for j, comp_name in enumerate(comparisons):\n",
    "        comp_df = all_comparisons[comp_name]\n",
    "        for i, model_name in enumerate(models):\n",
    "            model_row = comp_df[comp_df['Model'] == model_name]\n",
    "            if not model_row.empty:\n",
    "                data_matrix[i, j] = model_row['AUC-ROC'].values[0]\n",
    "            else:\n",
    "                data_matrix[i, j] = np.nan\n",
    "    \n",
    "    fig, ax = plt.subplots(figsize=(14, 6))\n",
    "    sns.heatmap(data_matrix, \n",
    "                annot=True, \n",
    "                fmt='.2f',\n",
    "                cmap='Blues',\n",
    "                xticklabels=comparisons,\n",
    "                yticklabels=models,\n",
    "                cbar_kws={'label': 'AUC-ROC'},\n",
    "                vmin=0.5, vmax=1.0,\n",
    "                linewidths=0.5,\n",
    "                ax=ax)\n",
    "    \n",
    "    plt.title('Model Performance Comparison Across All Binary Classifications', \n",
    "              fontsize=20, pad=20)\n",
    "    plt.xlabel('Classification Task', rotation=0, fontsize=16, labelpad=20)\n",
    "    plt.ylabel('Model', fontsize=16)\n",
    "    plt.xticks(rotation=0, ha='center')\n",
    "    plt.yticks(rotation=0)\n",
    "    plt.tight_layout()\n",
    "    \n",
    "    if output_path:\n",
    "        os.makedirs(os.path.dirname(output_path), exist_ok=True)\n",
    "        plt.savefig(output_path, dpi=300, bbox_inches='tight')\n",
    "        print(f\"\\nHeatmap saved to: {output_path}\")\n",
    "    plt.show()\n",
    "\n",
    "plot_model_comparison(all_model_comparisons, \n",
    "                      output_path='../Figures/Supplementary/Suppl_Fig_2.jpg')\n",
    "\n",
    "print(\"\\n\" + \"=\"*80)\n",
    "print(\"ANALYSIS COMPLETE\")\n",
    "\n",
    "print(f\"RF Configuration used throughout: {RF_CONFIG}\")\n",
    "print(\"âœ“ Main RF analysis complete (ROC curves + feature importance)\")\n",
    "print(\"âœ“ Model comparison complete (RF vs 6 alternatives)\")\n",
    "print(\"âœ“ All results saved\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Classification Split by Urban (Cape Town) vs Rural (Umtata) (Supplementary Plots)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Split data by area (Cape Town vs Umtata)\n",
    "cape_town_samples = metadata[metadata['area'] == 'Cape Town'].index\n",
    "umtata_samples = metadata[metadata['area'] == 'Umtata'].index\n",
    "\n",
    "# Filter the ASV table to get samples from each area\n",
    "df_cape_town = df.loc[df.index.isin(cape_town_samples)]\n",
    "df_umtata = df.loc[df.index.isin(umtata_samples)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Area: Capetown\n",
      "Available confounders: ['age_months', 'sex', 'enrolment_season']\n",
      "Final confounder variables: ['age_months', 'sex_male', 'enrolment_season_Autumn ', 'enrolment_season_Spring', 'enrolment_season_Spring ', 'enrolment_season_Summer', 'enrolment_season_Winter']\n",
      "Saved ROC plot for Capetown to ../Figures/Supplementary/Suppl_Fig_1B.jpg\n",
      "\n",
      "Area: Umtata\n",
      "Available confounders: ['age_months', 'sex', 'enrolment_season']\n",
      "Final confounder variables: ['age_months', 'sex_male', 'enrolment_season_Autumn ', 'enrolment_season_Spring', 'enrolment_season_Spring ', 'enrolment_season_Summer', 'enrolment_season_Winter']\n",
      "Saved ROC plot for Umtata to ../Figures/Supplementary/Suppl_Fig_1A.jpg\n"
     ]
    }
   ],
   "source": [
    "# Define area-specific DataFrames\n",
    "\n",
    "area_df_dict = {\n",
    "    'Capetown': df_cape_town,\n",
    "    'Umtata': df_umtata\n",
    "}\n",
    "\n",
    "\n",
    "# Loop through both study areas\n",
    "\n",
    "for area_name, df in area_df_dict.items():   \n",
    "    \n",
    "    # 1. Custom Group-Stratified K-Fold Function (UNCHANGED)\n",
    "    \n",
    "    def group_stratified_kfold(X, y, groups, n_splits=5, random_state=42):\n",
    "        unique_groups = np.unique(groups)\n",
    "        np.random.seed(random_state)\n",
    "        np.random.shuffle(unique_groups)\n",
    "        \n",
    "        group_label_dist = {}\n",
    "        for group in unique_groups:\n",
    "            group_mask = groups == group\n",
    "            group_y = y[group_mask]\n",
    "            group_label_dist[group] = {label: np.sum(group_y == label) for label in np.unique(y)}\n",
    "        \n",
    "        folds = [[] for _ in range(n_splits)]\n",
    "        fold_label_dist = [{label: 0 for label in np.unique(y)} for _ in range(n_splits)]\n",
    "        \n",
    "        sorted_groups = sorted(unique_groups, key=lambda g: sum(groups == g), reverse=True)\n",
    "        \n",
    "        for group in sorted_groups:\n",
    "            best_fold = 0\n",
    "            min_imbalance = float('inf')\n",
    "            \n",
    "            for fold_idx in range(n_splits):\n",
    "                temp_fold_dist = fold_label_dist[fold_idx].copy()\n",
    "                for label, count in group_label_dist[group].items():\n",
    "                    temp_fold_dist[label] += count\n",
    "                fold_size = sum(temp_fold_dist.values())\n",
    "                proportions = [count / fold_size for count in temp_fold_dist.values()] if fold_size else [0]*len(temp_fold_dist)\n",
    "                imbalance = np.var(proportions) + fold_size / (sum(groups.shape) / n_splits)\n",
    "                \n",
    "                if imbalance < min_imbalance:\n",
    "                    min_imbalance = imbalance\n",
    "                    best_fold = fold_idx\n",
    "            \n",
    "            folds[best_fold].extend(np.where(groups == group)[0])\n",
    "            for label, count in group_label_dist[group].items():\n",
    "                fold_label_dist[best_fold][label] += count\n",
    "        \n",
    "        train_test_indices = []\n",
    "        for i in range(n_splits):\n",
    "            test_idx = np.array(folds[i])\n",
    "            train_idx = np.concatenate([folds[j] for j in range(n_splits) if j != i])\n",
    "            train_test_indices.append((train_idx, test_idx))\n",
    "        \n",
    "        return train_test_indices\n",
    "\n",
    "    \n",
    "    # 2. MODIFIED Function to Run CV with Confounders\n",
    "    \n",
    "    def run_group_stratified_cv(X, y, groups, confounders=None, n_splits=5):\n",
    "        \"\"\"\n",
    "        Run group-stratified CV with optional confounder adjustment.\n",
    "        \n",
    "        Parameters:\n",
    "        - X: DataFrame of microbiome features\n",
    "        - y: Series of labels\n",
    "        - groups: Series of group IDs (patient IDs)\n",
    "        - confounders: DataFrame of confounder variable\n",
    "                       Must have same index as X\n",
    "        - n_splits: Number of CV folds\n",
    "        \n",
    "        Returns:\n",
    "        - cv_results: List of results per fold\n",
    "        - feature_importances: DataFrame of feature importances\n",
    "        - fold_aucs: List of AUC values per fold\n",
    "        \"\"\"\n",
    "        folds = group_stratified_kfold(X, y, groups, n_splits=n_splits)\n",
    "\n",
    "        cv_results = []\n",
    "        feature_importances = pd.DataFrame(index=X.columns)\n",
    "        fold_aucs = []\n",
    "\n",
    "        for i, (train_idx, test_idx) in enumerate(folds):\n",
    "            X_train, X_test = X.iloc[train_idx], X.iloc[test_idx]\n",
    "            y_train, y_test = y.iloc[train_idx], y.iloc[test_idx]\n",
    "\n",
    "            if len(np.unique(y_train)) < 2 or len(np.unique(y_test)) < 2:\n",
    "                print(f\"Skipping fold {i+1} due to insufficient class representation\")\n",
    "                continue\n",
    "\n",
    "            # ADD CONFOUNDERS IF PROVIDED\n",
    "            if confounders is not None:\n",
    "                # Ensure confounders are numeric\n",
    "                conf_train = confounders.iloc[train_idx]\n",
    "                conf_test = confounders.iloc[test_idx]\n",
    "                \n",
    "                # Concatenate microbiome features with confounders\n",
    "                X_train_combined = pd.concat([X_train, conf_train], axis=1)\n",
    "                X_test_combined = pd.concat([X_test, conf_test], axis=1)\n",
    "                \n",
    "                # Train RF on combined features\n",
    "                clf = RandomForestClassifier(n_estimators=1000, random_state=42, n_jobs=-1)\n",
    "                clf.fit(X_train_combined, y_train)\n",
    "                \n",
    "                # Get predictions\n",
    "                probas = clf.predict_proba(X_test_combined)\n",
    "                y_pred = clf.predict(X_test_combined)\n",
    "                \n",
    "                # Extract only microbiome feature importances (exclude confounder importances)\n",
    "                n_microbiome_features = X.shape[1]\n",
    "                feature_importances[f'fold_{i}'] = clf.feature_importances_[:n_microbiome_features]\n",
    "                \n",
    "            else:\n",
    "                # Standard RF without confounders\n",
    "                clf = RandomForestClassifier(n_estimators=1000, random_state=42, n_jobs=-1)\n",
    "                clf.fit(X_train, y_train)\n",
    "                probas = clf.predict_proba(X_test)\n",
    "                y_pred = clf.predict(X_test)\n",
    "                feature_importances[f'fold_{i}'] = clf.feature_importances_\n",
    "\n",
    "            fpr, tpr, _ = roc_curve(y_test, probas[:, 1])\n",
    "            roc_auc = auc(fpr, tpr)\n",
    "            fold_aucs.append(roc_auc)\n",
    "\n",
    "            cv_results.append({\n",
    "                'y_true': y_test,\n",
    "                'y_pred': y_pred,\n",
    "                'y_proba': probas[:, 1],\n",
    "                'fpr': fpr,\n",
    "                'tpr': tpr,\n",
    "                'auc': roc_auc,\n",
    "                'fold': i\n",
    "            })\n",
    "\n",
    "        feature_importances['mean_importance'] = feature_importances.mean(axis=1)\n",
    "        feature_importances['std_importance'] = feature_importances.std(axis=1)\n",
    "        feature_importances = feature_importances.sort_values('mean_importance', ascending=False)\n",
    "\n",
    "        return cv_results, feature_importances, fold_aucs\n",
    "\n",
    "    \n",
    "    # 3.Prepare Confounders\n",
    "    \n",
    "    # Define which confounders to include\n",
    "    confounder_cols = ['age_months', 'sex', 'enrolment_season']\n",
    "\n",
    "    # Check which confounders are available\n",
    "    available_confounders = [col for col in confounder_cols if col in metadata.columns]\n",
    "    print(f\"\\nArea: {area_name}\")\n",
    "    print(f\"Available confounders: {available_confounders}\")\n",
    "    \n",
    "    # Prepare confounder dataframe\n",
    "    if available_confounders:\n",
    "        confounders_full = metadata[available_confounders].copy()\n",
    "        \n",
    "        # Encode categorical variables (e.g., age_months)\n",
    "        for col in confounders_full.columns:\n",
    "            if confounders_full[col].dtype == 'object':\n",
    "                # Create dummy variables for categorical\n",
    "                confounders_full = pd.get_dummies(confounders_full, columns=[col], drop_first=True)\n",
    "        \n",
    "        # Handle missing values\n",
    "        confounders_full = confounders_full.fillna(confounders_full.median())\n",
    "        \n",
    "        print(f\"Final confounder variables: {list(confounders_full.columns)}\")\n",
    "    else:\n",
    "        confounders_full = None\n",
    "        print(\"No confounders available - running without confounder adjustment\")\n",
    "    \n",
    "    \n",
    "    # 4. MODIFIED: Regenerate ROC Results with Confounders\n",
    "    \n",
    "    # 4a. Skin vs Nares\n",
    "    meta_skin_nares = metadata[metadata['group'].str.startswith(('skin', 'nares'))].copy()\n",
    "    meta_skin_nares['site_label'] = meta_skin_nares['group'].apply(lambda x: 0 if x.startswith('skin') else 1)\n",
    "    common_samples = df.index.intersection(meta_skin_nares.index)\n",
    "    X_skin_nares = df.loc[common_samples]\n",
    "    y_skin_nares = meta_skin_nares.loc[common_samples, 'site_label']\n",
    "    groups_skin_nares = meta_skin_nares.loc[common_samples, 'pid']\n",
    "    y_skin_nares = pd.to_numeric(y_skin_nares, errors='coerce')\n",
    "    \n",
    "    # Get confounders for this subset\n",
    "    conf_skin_nares = confounders_full.loc[common_samples] if confounders_full is not None else None\n",
    "    \n",
    "    skin_vs_nares_cv_results, _, _ = run_group_stratified_cv(\n",
    "        X_skin_nares, y_skin_nares, groups_skin_nares, \n",
    "        confounders=conf_skin_nares, n_splits=3\n",
    "    )\n",
    "\n",
    "    # 4b. Skin Comparisons\n",
    "    skin_comparisons = [\n",
    "        ('skin-ADL', 'skin-H'),\n",
    "        ('skin-ADNL', 'skin-ADL'),\n",
    "        ('skin-ADNL', 'skin-H')\n",
    "    ]\n",
    "    skin_cv_results_dict = {}\n",
    "    for label1, label2 in skin_comparisons:\n",
    "        meta_subset = metadata[metadata['group'].isin([label1, label2])]\n",
    "        common_samples = df.index.intersection(meta_subset.index)\n",
    "        X_skin = df.loc[common_samples]\n",
    "        y_skin = meta_subset.loc[common_samples, 'group'].map({label1: 0, label2: 1})\n",
    "        groups_skin = meta_subset.loc[common_samples, 'pid']\n",
    "        \n",
    "        # Get confounders for this subset\n",
    "        conf_skin = confounders_full.loc[common_samples] if confounders_full is not None else None\n",
    "        \n",
    "        cv_results, _, _ = run_group_stratified_cv(\n",
    "            X_skin, y_skin, groups_skin, \n",
    "            confounders=conf_skin, n_splits=3\n",
    "        )\n",
    "        key = f\"{label1}_vs_{label2}\"\n",
    "        skin_cv_results_dict[key] = cv_results\n",
    "\n",
    "    # 4c. Nares Comparison\n",
    "    meta_nares = metadata[metadata['group'].isin(['nares-AD', 'nares-H'])]\n",
    "    common_samples = df.index.intersection(meta_nares.index)\n",
    "    X_nares = df.loc[common_samples]\n",
    "    y_nares = meta_nares.loc[common_samples, 'group'].map({'nares-AD': 0, 'nares-H': 1})\n",
    "    groups_nares = meta_nares.loc[common_samples, 'pid']\n",
    "    \n",
    "    # Get confounders for this subset\n",
    "    conf_nares = confounders_full.loc[common_samples] if confounders_full is not None else None\n",
    "    \n",
    "    nares_ad_vs_h_cv_results, _, _ = run_group_stratified_cv(\n",
    "        X_nares, y_nares, groups_nares, \n",
    "        confounders=conf_nares, n_splits=3\n",
    "    )\n",
    "\n",
    "    roc_results_dict = {\n",
    "        'Prediction of Skin vs. Nares Samples': {\n",
    "            'skin_vs_nares': skin_vs_nares_cv_results\n",
    "        },\n",
    "        'Prediction of AD Status From Skin Samples': skin_cv_results_dict,\n",
    "        'Prediction of AD Status From Nares Samples': {\n",
    "            'nares-AD_vs_nares-H': nares_ad_vs_h_cv_results\n",
    "        }\n",
    "    }\n",
    "\n",
    "    \n",
    "    # 5. Combined ROC Plotting Function (UNCHANGED)\n",
    "    \n",
    "    def plot_combined_roc_3panel(roc_results_dict, comparison_order, color_map, area_name, output_path=None):\n",
    "        fig, axs = plt.subplots(1, 3, figsize=(18, 6), sharey=True)\n",
    "\n",
    "        pretty_labels = {\n",
    "            'skin_vs_nares': 'All Skin vs All Nares',\n",
    "            'skin-ADL_vs_skin-H': 'Skin ADL vs Skin H',\n",
    "            'skin-ADNL_vs_skin-ADL': 'Skin ADNL vs Skin ADL',\n",
    "            'skin-ADNL_vs_skin-H': 'Skin ADNL vs Skin H',\n",
    "            'nares-AD_vs_nares-H': 'Nares AD vs Nares H'\n",
    "        }\n",
    "\n",
    "        for i, panel_title in enumerate(comparison_order):\n",
    "            ax = axs[i]\n",
    "            sub_dict = roc_results_dict[panel_title]\n",
    "            for sublabel, curves in sub_dict.items():\n",
    "                if not isinstance(curves, list):\n",
    "                    curves = [curves]\n",
    "                mean_fpr = np.linspace(0, 1, 100)\n",
    "                tprs, aucs = [], []\n",
    "                for result in curves:\n",
    "                    if 'fpr' not in result or 'tpr' not in result:\n",
    "                        continue\n",
    "                    tpr_interp = np.interp(mean_fpr, result['fpr'], result['tpr'])\n",
    "                    tpr_interp[0] = 0.0\n",
    "                    tprs.append(tpr_interp)\n",
    "                    aucs.append(result['auc'])\n",
    "                if not tprs:\n",
    "                    continue\n",
    "                mean_tpr = np.mean(tprs, axis=0)\n",
    "                mean_tpr[-1] = 1.0\n",
    "                std_tpr = np.std(tprs, axis=0)\n",
    "                ax.plot(mean_fpr, mean_tpr, lw=3,\n",
    "                        label=f'{pretty_labels.get(sublabel, sublabel)} (AUC = {np.mean(aucs):.2f} Â± {np.std(aucs):.2f})',\n",
    "                        color=color_map.get(sublabel, 'gray'))\n",
    "                ax.fill_between(mean_fpr,\n",
    "                                np.maximum(mean_tpr - std_tpr, 0),\n",
    "                                np.minimum(mean_tpr + std_tpr, 1),\n",
    "                                color=color_map.get(sublabel, 'gray'),\n",
    "                                alpha=0.3)\n",
    "            ax.plot([0, 1], [0, 1], 'k--', lw=1)\n",
    "            ax.set_xlim([0.0, 1.0])\n",
    "            ax.set_ylim([0.0, 1.05])\n",
    "            ax.set_xlabel('False Positive Rate', fontsize=16)\n",
    "            ax.set_title(panel_title, fontsize=18)\n",
    "            ax.grid(True, linestyle='--', alpha=0.7)\n",
    "            ax.tick_params(axis='both', labelsize=14)\n",
    "            ax.legend(loc='lower right', fontsize=12)\n",
    "\n",
    "        axs[0].set_ylabel('True Positive Rate', fontsize=16)\n",
    "        if area_name == 'Capetown':\n",
    "            plt.suptitle('Random Forest Classifications by 16S ASVs of Cape Town Samples', fontsize=26)\n",
    "        elif area_name == 'Umtata':\n",
    "            plt.suptitle('Random Forest Classifications by 16S ASVs of Umtata Samples', fontsize=26)\n",
    "\n",
    "        plt.tight_layout(rect=[0, 0, 1, 0.95])\n",
    "\n",
    "        if output_path:\n",
    "            os.makedirs(os.path.dirname(output_path), exist_ok=True)\n",
    "            plt.savefig(output_path, dpi=600, bbox_inches='tight')\n",
    "            print(f\"Saved ROC plot for {area_name} to {output_path}\")\n",
    "        else:\n",
    "            plt.show()\n",
    "\n",
    "        plt.close(fig)\n",
    "\n",
    "    \n",
    "    # 6. Define Panel Order and Color Map, Then Plot\n",
    "    \n",
    "    comparison_order = [\n",
    "        'Prediction of Skin vs. Nares Samples',\n",
    "        'Prediction of AD Status From Skin Samples',\n",
    "        'Prediction of AD Status From Nares Samples'\n",
    "    ]\n",
    "    color_map = {\n",
    "        'skin_vs_nares': 'black',\n",
    "        'skin-ADL_vs_skin-H': 'blue',\n",
    "        'skin-ADNL_vs_skin-ADL': 'green',\n",
    "        'skin-ADNL_vs_skin-H': 'purple',\n",
    "        'nares-AD_vs_nares-H': 'orange'\n",
    "    }\n",
    "\n",
    "    if area_name == 'Umtata':\n",
    "        output_path = '../Figures/Supplementary/Suppl_Fig_1A.jpg'\n",
    "    elif area_name == 'Capetown':\n",
    "        output_path = '../Figures/Supplementary/Suppl_Fig_1B.jpg'\n",
    "\n",
    "    plot_combined_roc_3panel(\n",
    "        roc_results_dict,\n",
    "        comparison_order=comparison_order,\n",
    "        color_map=color_map,\n",
    "        area_name=area_name,\n",
    "        output_path=output_path\n",
    "    )"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "qiime2-metagenome-2024.10",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
