{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Abundance of taxa on nose and severity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import biom\n",
    "from biom import load_table\n",
    "import numpy as np\n",
    "from scipy.stats import spearmanr\n",
    "from scipy.stats import mannwhitneyu\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "import itertools\n",
    "import matplotlib.patches\n",
    "from scipy.stats import pearsonr\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 220,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PlateNumber</th>\n",
       "      <th>PlateLocation</th>\n",
       "      <th>i5</th>\n",
       "      <th>i5Sequence</th>\n",
       "      <th>i7</th>\n",
       "      <th>i7Sequence</th>\n",
       "      <th>identifier</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Plate ID</th>\n",
       "      <th>Well location</th>\n",
       "      <th>...</th>\n",
       "      <th>enrolment_date</th>\n",
       "      <th>enrolment_season</th>\n",
       "      <th>hiv_exposure</th>\n",
       "      <th>hiv_status</th>\n",
       "      <th>household_size</th>\n",
       "      <th>o_scorad</th>\n",
       "      <th>FWD_filepath</th>\n",
       "      <th>REV_filepath</th>\n",
       "      <th>group</th>\n",
       "      <th>microbiome_type</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#sample-id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Ca009STL</th>\n",
       "      <td>1</td>\n",
       "      <td>A1</td>\n",
       "      <td>SA501</td>\n",
       "      <td>ATCGTACG</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA501</td>\n",
       "      <td>CGAGAGTT-ATCGTACG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>A1</td>\n",
       "      <td>...</td>\n",
       "      <td>4/16/2015</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>4.0</td>\n",
       "      <td>40</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900221</th>\n",
       "      <td>1</td>\n",
       "      <td>B1</td>\n",
       "      <td>SA502</td>\n",
       "      <td>ACTATCTG</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA502</td>\n",
       "      <td>CGAGAGTT-ACTATCTG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>B1</td>\n",
       "      <td>...</td>\n",
       "      <td>8/11/2015</td>\n",
       "      <td>Winter</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>7.0</td>\n",
       "      <td>34</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca010EBL</th>\n",
       "      <td>1</td>\n",
       "      <td>C1</td>\n",
       "      <td>SA503</td>\n",
       "      <td>TAGCGAGT</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA503</td>\n",
       "      <td>CGAGAGTT-TAGCGAGT</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>C1</td>\n",
       "      <td>...</td>\n",
       "      <td>11/20/2014</td>\n",
       "      <td>Spring</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>7.0</td>\n",
       "      <td>21</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900460</th>\n",
       "      <td>1</td>\n",
       "      <td>D1</td>\n",
       "      <td>SA504</td>\n",
       "      <td>CTGCGTGT</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA504</td>\n",
       "      <td>CGAGAGTT-CTGCGTGT</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>D1</td>\n",
       "      <td>...</td>\n",
       "      <td>9/23/2015</td>\n",
       "      <td>Spring</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>40</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900051</th>\n",
       "      <td>1</td>\n",
       "      <td>E1</td>\n",
       "      <td>SA505</td>\n",
       "      <td>TCATCGAG</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA505</td>\n",
       "      <td>CGAGAGTT-TCATCGAG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>E1</td>\n",
       "      <td>...</td>\n",
       "      <td>4/21/2015</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>7.0</td>\n",
       "      <td>41</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca006ONL2</th>\n",
       "      <td>6</td>\n",
       "      <td>H1</td>\n",
       "      <td>SA508</td>\n",
       "      <td>GACACCGT</td>\n",
       "      <td>SB701</td>\n",
       "      <td>CTCGACTT</td>\n",
       "      <td>SB701SA508</td>\n",
       "      <td>CTCGACTT-GACACCGT</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>H1</td>\n",
       "      <td>...</td>\n",
       "      <td>3/25/2015</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>3.0</td>\n",
       "      <td>34</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca006ONNL</th>\n",
       "      <td>6</td>\n",
       "      <td>F2</td>\n",
       "      <td>SA506</td>\n",
       "      <td>CGTGAGTG</td>\n",
       "      <td>SB702</td>\n",
       "      <td>CGAAGTAT</td>\n",
       "      <td>SB702SA506</td>\n",
       "      <td>CGAAGTAT-CGTGAGTG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>F2</td>\n",
       "      <td>...</td>\n",
       "      <td>3/25/2015</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>3.0</td>\n",
       "      <td>34</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADNL</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca006ONNL2</th>\n",
       "      <td>6</td>\n",
       "      <td>H2</td>\n",
       "      <td>SA508</td>\n",
       "      <td>GACACCGT</td>\n",
       "      <td>SB702</td>\n",
       "      <td>CGAAGTAT</td>\n",
       "      <td>SB702SA508</td>\n",
       "      <td>CGAAGTAT-GACACCGT</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>H2</td>\n",
       "      <td>...</td>\n",
       "      <td>3/25/2015</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>3.0</td>\n",
       "      <td>34</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADNL</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca006ONPN</th>\n",
       "      <td>6</td>\n",
       "      <td>F3</td>\n",
       "      <td>SA506</td>\n",
       "      <td>CGTGAGTG</td>\n",
       "      <td>SB703</td>\n",
       "      <td>TAGCAGCT</td>\n",
       "      <td>SB703SA506</td>\n",
       "      <td>TAGCAGCT-CGTGAGTG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>F3</td>\n",
       "      <td>...</td>\n",
       "      <td>3/25/2015</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>3.0</td>\n",
       "      <td>34</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>nares-AD</td>\n",
       "      <td>nares</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca006ONPN2</th>\n",
       "      <td>6</td>\n",
       "      <td>H3</td>\n",
       "      <td>SA508</td>\n",
       "      <td>GACACCGT</td>\n",
       "      <td>SB703</td>\n",
       "      <td>TAGCAGCT</td>\n",
       "      <td>SB703SA508</td>\n",
       "      <td>TAGCAGCT-GACACCGT</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>H3</td>\n",
       "      <td>...</td>\n",
       "      <td>3/25/2015</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>3.0</td>\n",
       "      <td>34</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>nares-AD</td>\n",
       "      <td>nares</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>502 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            PlateNumber PlateLocation     i5 i5Sequence     i7 i7Sequence  \\\n",
       "#sample-id                                                                  \n",
       "Ca009STL              1            A1  SA501   ATCGTACG  SA701   CGAGAGTT   \n",
       "900221                1            B1  SA502   ACTATCTG  SA701   CGAGAGTT   \n",
       "Ca010EBL              1            C1  SA503   TAGCGAGT  SA701   CGAGAGTT   \n",
       "900460                1            D1  SA504   CTGCGTGT  SA701   CGAGAGTT   \n",
       "900051                1            E1  SA505   TCATCGAG  SA701   CGAGAGTT   \n",
       "...                 ...           ...    ...        ...    ...        ...   \n",
       "Ca006ONL2             6            H1  SA508   GACACCGT  SB701   CTCGACTT   \n",
       "Ca006ONNL             6            F2  SA506   CGTGAGTG  SB702   CGAAGTAT   \n",
       "Ca006ONNL2            6            H2  SA508   GACACCGT  SB702   CGAAGTAT   \n",
       "Ca006ONPN             6            F3  SA506   CGTGAGTG  SB703   TAGCAGCT   \n",
       "Ca006ONPN2            6            H3  SA508   GACACCGT  SB703   TAGCAGCT   \n",
       "\n",
       "            identifier           Sequence      Plate ID Well location  ...  \\\n",
       "#sample-id                                                             ...   \n",
       "Ca009STL    SA701SA501  CGAGAGTT-ATCGTACG  1.010000e+21            A1  ...   \n",
       "900221      SA701SA502  CGAGAGTT-ACTATCTG  1.010000e+21            B1  ...   \n",
       "Ca010EBL    SA701SA503  CGAGAGTT-TAGCGAGT  1.010000e+21            C1  ...   \n",
       "900460      SA701SA504  CGAGAGTT-CTGCGTGT  1.010000e+21            D1  ...   \n",
       "900051      SA701SA505  CGAGAGTT-TCATCGAG  1.010000e+21            E1  ...   \n",
       "...                ...                ...           ...           ...  ...   \n",
       "Ca006ONL2   SB701SA508  CTCGACTT-GACACCGT  1.010000e+21            H1  ...   \n",
       "Ca006ONNL   SB702SA506  CGAAGTAT-CGTGAGTG  1.010000e+21            F2  ...   \n",
       "Ca006ONNL2  SB702SA508  CGAAGTAT-GACACCGT  1.010000e+21            H2  ...   \n",
       "Ca006ONPN   SB703SA506  TAGCAGCT-CGTGAGTG  1.010000e+21            F3  ...   \n",
       "Ca006ONPN2  SB703SA508  TAGCAGCT-GACACCGT  1.010000e+21            H3  ...   \n",
       "\n",
       "            enrolment_date enrolment_season hiv_exposure hiv_status  \\\n",
       "#sample-id                                                            \n",
       "Ca009STL         4/16/2015          Autumn     Unexposed   negative   \n",
       "900221           8/11/2015           Winter    Unexposed   negative   \n",
       "Ca010EBL        11/20/2014           Spring    Unexposed   negative   \n",
       "900460           9/23/2015           Spring    Unexposed        NaN   \n",
       "900051           4/21/2015          Autumn     Unexposed   negative   \n",
       "...                    ...              ...          ...        ...   \n",
       "Ca006ONL2        3/25/2015           Autumn    Unexposed   negative   \n",
       "Ca006ONNL        3/25/2015           Autumn    Unexposed   negative   \n",
       "Ca006ONNL2       3/25/2015           Autumn    Unexposed   negative   \n",
       "Ca006ONPN        3/25/2015           Autumn    Unexposed   negative   \n",
       "Ca006ONPN2       3/25/2015           Autumn    Unexposed   negative   \n",
       "\n",
       "           household_size o_scorad  \\\n",
       "#sample-id                           \n",
       "Ca009STL              4.0       40   \n",
       "900221                7.0       34   \n",
       "Ca010EBL              7.0       21   \n",
       "900460                4.0       40   \n",
       "900051                7.0       41   \n",
       "...                   ...      ...   \n",
       "Ca006ONL2             3.0       34   \n",
       "Ca006ONNL             3.0       34   \n",
       "Ca006ONNL2            3.0       34   \n",
       "Ca006ONPN             3.0       34   \n",
       "Ca006ONPN2            3.0       34   \n",
       "\n",
       "                                                 FWD_filepath  \\\n",
       "#sample-id                                                      \n",
       "Ca009STL    /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "900221      /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "Ca010EBL    /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "900460      /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "900051      /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "...                                                       ...   \n",
       "Ca006ONL2   /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "Ca006ONNL   /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "Ca006ONNL2  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "Ca006ONPN   /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "Ca006ONPN2  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "\n",
       "                                                 REV_filepath      group  \\\n",
       "#sample-id                                                                 \n",
       "Ca009STL    /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   skin-ADL   \n",
       "900221      /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   skin-ADL   \n",
       "Ca010EBL    /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   skin-ADL   \n",
       "900460      /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   skin-ADL   \n",
       "900051      /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   skin-ADL   \n",
       "...                                                       ...        ...   \n",
       "Ca006ONL2   /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   skin-ADL   \n",
       "Ca006ONNL   /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...  skin-ADNL   \n",
       "Ca006ONNL2  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...  skin-ADNL   \n",
       "Ca006ONPN   /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   nares-AD   \n",
       "Ca006ONPN2  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   nares-AD   \n",
       "\n",
       "           microbiome_type  \n",
       "#sample-id                  \n",
       "Ca009STL              skin  \n",
       "900221                skin  \n",
       "Ca010EBL              skin  \n",
       "900460                skin  \n",
       "900051                skin  \n",
       "...                    ...  \n",
       "Ca006ONL2             skin  \n",
       "Ca006ONNL             skin  \n",
       "Ca006ONNL2            skin  \n",
       "Ca006ONPN            nares  \n",
       "Ca006ONPN2           nares  \n",
       "\n",
       "[502 rows x 33 columns]"
      ]
     },
     "execution_count": 220,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load the metadata\n",
    "metadata_path = '../Data/Metadata/updated_clean_ant_skin_metadata_microbiome_type.tab'\n",
    "metadata = pd.read_csv(metadata_path, sep='\\t')\n",
    "\n",
    "metadata['#sample-id'] = metadata['#sample-id'].str.replace('_', '')\n",
    "# Set Sample-ID as the index for the metadata dataframe \n",
    "metadata = metadata.set_index('#sample-id')\n",
    "\n",
    "\n",
    "metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 221,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>g__Streptococcus_ASV-1</th>\n",
       "      <th>g__Streptococcus_ASV-2</th>\n",
       "      <th>g__Corynebacterium_ASV-1</th>\n",
       "      <th>g__Corynebacterium_ASV-4</th>\n",
       "      <th>g___ASV-5</th>\n",
       "      <th>g__Cutibacterium_ASV-1</th>\n",
       "      <th>g___ASV-33</th>\n",
       "      <th>g___ASV-30</th>\n",
       "      <th>g___ASV-22</th>\n",
       "      <th>g__Cutibacterium_ASV-2</th>\n",
       "      <th>...</th>\n",
       "      <th>g__Copromonas_ASV-3</th>\n",
       "      <th>g__Microvirga_ASV-1</th>\n",
       "      <th>g__Leptotrichia_A_993758_ASV-17</th>\n",
       "      <th>g___ASV-103</th>\n",
       "      <th>g__Leptotrichia_A_993758_ASV-13</th>\n",
       "      <th>g__Capnocytophaga_820688_ASV-8</th>\n",
       "      <th>g__UBA6175_ASV-2</th>\n",
       "      <th>g__Comamonas_F_589250_ASV-3</th>\n",
       "      <th>g__Bosea_ASV-2</th>\n",
       "      <th>g___ASV-145</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#sample-id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Ca009STPN</th>\n",
       "      <td>51.0</td>\n",
       "      <td>0</td>\n",
       "      <td>240.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900223</th>\n",
       "      <td>190.0</td>\n",
       "      <td>0</td>\n",
       "      <td>899.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca010EBPN</th>\n",
       "      <td>208.0</td>\n",
       "      <td>0</td>\n",
       "      <td>154.0</td>\n",
       "      <td>0</td>\n",
       "      <td>177.0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900462</th>\n",
       "      <td>532.0</td>\n",
       "      <td>484.0</td>\n",
       "      <td>254.0</td>\n",
       "      <td>200.0</td>\n",
       "      <td>104.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900053</th>\n",
       "      <td>2478.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9003992</th>\n",
       "      <td>328.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1463.0</td>\n",
       "      <td>892.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>59.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900402</th>\n",
       "      <td>156.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9004022</th>\n",
       "      <td>126.0</td>\n",
       "      <td>0</td>\n",
       "      <td>241.0</td>\n",
       "      <td>0</td>\n",
       "      <td>235.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca006ONPN</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>286.0</td>\n",
       "      <td>0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca006ONPN2</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>106.0</td>\n",
       "      <td>0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>43.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>197 rows × 798 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            g__Streptococcus_ASV-1  g__Streptococcus_ASV-2  \\\n",
       "#sample-id                                                   \n",
       "Ca009STPN                     51.0                       0   \n",
       "900223                       190.0                       0   \n",
       "Ca010EBPN                    208.0                       0   \n",
       "900462                       532.0                   484.0   \n",
       "900053                      2478.0                       0   \n",
       "...                            ...                     ...   \n",
       "9003992                      328.0                       0   \n",
       "900402                       156.0                       0   \n",
       "9004022                      126.0                       0   \n",
       "Ca006ONPN                        0                       0   \n",
       "Ca006ONPN2                       0                       0   \n",
       "\n",
       "            g__Corynebacterium_ASV-1  g__Corynebacterium_ASV-4  g___ASV-5  \\\n",
       "#sample-id                                                                  \n",
       "Ca009STPN                      240.0                         0          0   \n",
       "900223                         899.0                         0          0   \n",
       "Ca010EBPN                      154.0                         0      177.0   \n",
       "900462                         254.0                     200.0      104.0   \n",
       "900053                             0                         0        3.0   \n",
       "...                              ...                       ...        ...   \n",
       "9003992                       1463.0                     892.0       15.0   \n",
       "900402                             0                         0          0   \n",
       "9004022                        241.0                         0      235.0   \n",
       "Ca006ONPN                      286.0                         0       13.0   \n",
       "Ca006ONPN2                     106.0                         0       18.0   \n",
       "\n",
       "            g__Cutibacterium_ASV-1  g___ASV-33  g___ASV-30  g___ASV-22  \\\n",
       "#sample-id                                                               \n",
       "Ca009STPN                        0         2.0           0           0   \n",
       "900223                           0         2.0           0           0   \n",
       "Ca010EBPN                     38.0           0           0           0   \n",
       "900462                        11.0           0           0           0   \n",
       "900053                         7.0           0           0           0   \n",
       "...                            ...         ...         ...         ...   \n",
       "9003992                       72.0           0           0           0   \n",
       "900402                        32.0           0           0           0   \n",
       "9004022                       14.0           0           0         4.0   \n",
       "Ca006ONPN                        0           0           0           0   \n",
       "Ca006ONPN2                    43.0           0           0           0   \n",
       "\n",
       "            g__Cutibacterium_ASV-2  ...  g__Copromonas_ASV-3  \\\n",
       "#sample-id                          ...                        \n",
       "Ca009STPN                        0  ...                    0   \n",
       "900223                           0  ...                    0   \n",
       "Ca010EBPN                     19.0  ...                    0   \n",
       "900462                         8.0  ...                    0   \n",
       "900053                         1.0  ...                    0   \n",
       "...                            ...  ...                  ...   \n",
       "9003992                       59.0  ...                    0   \n",
       "900402                        19.0  ...                    0   \n",
       "9004022                        7.0  ...                    0   \n",
       "Ca006ONPN                        0  ...                    0   \n",
       "Ca006ONPN2                    20.0  ...                    0   \n",
       "\n",
       "            g__Microvirga_ASV-1  g__Leptotrichia_A_993758_ASV-17  g___ASV-103  \\\n",
       "#sample-id                                                                      \n",
       "Ca009STPN                     0                                0            0   \n",
       "900223                        0                                0            0   \n",
       "Ca010EBPN                     0                                0            0   \n",
       "900462                        0                                0            0   \n",
       "900053                        0                                0            0   \n",
       "...                         ...                              ...          ...   \n",
       "9003992                       0                                0            0   \n",
       "900402                        0                                0            0   \n",
       "9004022                       0                                0            0   \n",
       "Ca006ONPN                     0                                0            0   \n",
       "Ca006ONPN2                    0                                0            0   \n",
       "\n",
       "            g__Leptotrichia_A_993758_ASV-13  g__Capnocytophaga_820688_ASV-8  \\\n",
       "#sample-id                                                                    \n",
       "Ca009STPN                                 0                               0   \n",
       "900223                                    0                               0   \n",
       "Ca010EBPN                                 0                               0   \n",
       "900462                                    0                               0   \n",
       "900053                                    0                               0   \n",
       "...                                     ...                             ...   \n",
       "9003992                                   0                               0   \n",
       "900402                                    0                               0   \n",
       "9004022                                   0                               0   \n",
       "Ca006ONPN                                 0                               0   \n",
       "Ca006ONPN2                                0                               0   \n",
       "\n",
       "            g__UBA6175_ASV-2  g__Comamonas_F_589250_ASV-3  g__Bosea_ASV-2  \\\n",
       "#sample-id                                                                  \n",
       "Ca009STPN                  0                            0               0   \n",
       "900223                     0                            0               0   \n",
       "Ca010EBPN                  0                            0               0   \n",
       "900462                     0                            0               0   \n",
       "900053                     0                            0               0   \n",
       "...                      ...                          ...             ...   \n",
       "9003992                    0                            0               0   \n",
       "900402                     0                            0               0   \n",
       "9004022                    0                            0               0   \n",
       "Ca006ONPN                  0                            0               0   \n",
       "Ca006ONPN2                 0                            0               0   \n",
       "\n",
       "            g___ASV-145  \n",
       "#sample-id               \n",
       "Ca009STPN             0  \n",
       "900223                0  \n",
       "Ca010EBPN             0  \n",
       "900462                0  \n",
       "900053                0  \n",
       "...                 ...  \n",
       "9003992               0  \n",
       "900402                0  \n",
       "9004022               0  \n",
       "Ca006ONPN             0  \n",
       "Ca006ONPN2            0  \n",
       "\n",
       "[197 rows x 798 columns]"
      ]
     },
     "execution_count": 221,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read in table at ASV level\n",
    "# biom_path = '../Data/Tables/Absolute_Abundance_Tables/feature_table_with_tax_labels_Genus.biom'\n",
    "biom_path = '../Data/Tables/Absolute_Abundance_Tables/209766_filtered_by_prevalence_1pct_rare_Genus-ASV-non-collapse.biom'\n",
    "# biom_path = '../Data/Tables/Absolute_Abundance_Tables/209766_filtered_by_prevalence_1pct_rare_Genus.biom'\n",
    "\n",
    "biom_tbl = load_table(biom_path)\n",
    "df = pd.DataFrame(biom_tbl.to_dataframe().T)\n",
    "\n",
    "# delete the prefix from the index\n",
    "df.index = df.index.str.replace('15564.', '')\n",
    "\n",
    "# Get samples that start with 'nares' in metadata\n",
    "nares_samples = metadata[metadata['group'].str.startswith('nares')].index\n",
    "\n",
    "# Subset to only certain area samples\n",
    "cape_town_samples = metadata[metadata['area'].str.startswith('Umtata')].index\n",
    "umtata_town_samples = metadata[metadata['area'].str.startswith('Umtata')].index\n",
    "\n",
    "# Filter df to keep only nares samples\n",
    "df = df.loc[nares_samples]\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 222,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Extract base taxa names before _ASV\n",
    "# base_taxa = [col.split('_ASV')[0] if '_ASV' in col else col for col in df.columns]\n",
    "\n",
    "# # Create dictionary mapping base taxa to list of original column names\n",
    "# taxa_groups = {}\n",
    "# for base, col in zip(base_taxa, df.columns):\n",
    "#     if base not in taxa_groups:\n",
    "#         taxa_groups[base] = []\n",
    "#     taxa_groups[base].append(col)\n",
    "\n",
    "# # Sum columns within each taxa group\n",
    "# combined_df = pd.DataFrame()\n",
    "# for base_taxon, columns in taxa_groups.items():\n",
    "#     if len(columns) > 1:  # Only combine if multiple ASVs exist\n",
    "#         combined_df[base_taxon] = df[columns].sum(axis=1)\n",
    "#     else:\n",
    "#         combined_df[base_taxon] = df[columns[0]]\n",
    "\n",
    "# # Replace original df with combined version\n",
    "# df = combined_df\n",
    "# df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 223,
   "metadata": {},
   "outputs": [],
   "source": [
    "def rclr_transform(df, pseudocount=1e-6):\n",
    "    \"\"\"\n",
    "    Applies Robust Centered Log-Ratio (RCLR) transformation to a DataFrame.\n",
    "    Zeros are ignored in the geometric mean calculation per sample.\n",
    "    \"\"\"\n",
    "    # Convert sparse to dense if needed\n",
    "    df = df.sparse.to_dense()\n",
    "\n",
    "    # Replace 0 with np.nan to ignore in log and mean\n",
    "    df_masked = df.replace(0, np.nan)\n",
    "\n",
    "    # Apply log (with pseudocount only where needed)\n",
    "    log_df = np.log(df_masked + pseudocount)\n",
    "\n",
    "    # Subtract mean of each row (feature)\n",
    "    rclr_df = log_df.sub(log_df.mean(axis=1, skipna=True), axis=0)\n",
    "\n",
    "    return rclr_df\n",
    "\n",
    "\n",
    "\n",
    "rclr_df = rclr_transform(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 224,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>g__Streptococcus_ASV-1</th>\n",
       "      <th>g__Streptococcus_ASV-2</th>\n",
       "      <th>g__Corynebacterium_ASV-1</th>\n",
       "      <th>g__Corynebacterium_ASV-4</th>\n",
       "      <th>g___ASV-5</th>\n",
       "      <th>g__Cutibacterium_ASV-1</th>\n",
       "      <th>g___ASV-33</th>\n",
       "      <th>g___ASV-30</th>\n",
       "      <th>g___ASV-22</th>\n",
       "      <th>g__Cutibacterium_ASV-2</th>\n",
       "      <th>...</th>\n",
       "      <th>g__Microvirga_ASV-1</th>\n",
       "      <th>g__Leptotrichia_A_993758_ASV-17</th>\n",
       "      <th>g___ASV-103</th>\n",
       "      <th>g__Leptotrichia_A_993758_ASV-13</th>\n",
       "      <th>g__Capnocytophaga_820688_ASV-8</th>\n",
       "      <th>g__UBA6175_ASV-2</th>\n",
       "      <th>g__Comamonas_F_589250_ASV-3</th>\n",
       "      <th>g__Bosea_ASV-2</th>\n",
       "      <th>g___ASV-145</th>\n",
       "      <th>group</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#sample-id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Ca009STPN</th>\n",
       "      <td>2.024339</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.573153</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.214339</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nares-AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900223</th>\n",
       "      <td>0.489140</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.043399</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-4.064737</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nares-AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca010EBPN</th>\n",
       "      <td>2.783217</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.482631</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.621828</td>\n",
       "      <td>1.083265</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.390118</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nares-AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900462</th>\n",
       "      <td>4.109607</td>\n",
       "      <td>4.015048</td>\n",
       "      <td>3.370298</td>\n",
       "      <td>3.131281</td>\n",
       "      <td>2.477354</td>\n",
       "      <td>0.230859</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.087595</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nares-AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900053</th>\n",
       "      <td>5.869346</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.847248</td>\n",
       "      <td>0.000050</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.945860</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nares-AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9003992</th>\n",
       "      <td>3.255158</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.750389</td>\n",
       "      <td>4.255610</td>\n",
       "      <td>0.170195</td>\n",
       "      <td>1.738810</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.539682</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nares-AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900402</th>\n",
       "      <td>2.275817</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.691697</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.170400</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nares-AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9004022</th>\n",
       "      <td>2.428674</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.077189</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.051978</td>\n",
       "      <td>0.231450</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.021313</td>\n",
       "      <td>-0.461697</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nares-AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca006ONPN</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.152846</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.938196</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nares-AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca006ONPN2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.292848</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.480220</td>\n",
       "      <td>0.390609</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.374859</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nares-AD</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>197 rows × 799 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            g__Streptococcus_ASV-1  g__Streptococcus_ASV-2  \\\n",
       "#sample-id                                                   \n",
       "Ca009STPN                 2.024339                     NaN   \n",
       "900223                    0.489140                     NaN   \n",
       "Ca010EBPN                 2.783217                     NaN   \n",
       "900462                    4.109607                4.015048   \n",
       "900053                    5.869346                     NaN   \n",
       "...                            ...                     ...   \n",
       "9003992                   3.255158                     NaN   \n",
       "900402                    2.275817                     NaN   \n",
       "9004022                   2.428674                     NaN   \n",
       "Ca006ONPN                      NaN                     NaN   \n",
       "Ca006ONPN2                     NaN                     NaN   \n",
       "\n",
       "            g__Corynebacterium_ASV-1  g__Corynebacterium_ASV-4  g___ASV-5  \\\n",
       "#sample-id                                                                  \n",
       "Ca009STPN                   3.573153                       NaN        NaN   \n",
       "900223                      2.043399                       NaN        NaN   \n",
       "Ca010EBPN                   2.482631                       NaN   2.621828   \n",
       "900462                      3.370298                  3.131281   2.477354   \n",
       "900053                           NaN                       NaN  -0.847248   \n",
       "...                              ...                       ...        ...   \n",
       "9003992                     4.750389                  4.255610   0.170195   \n",
       "900402                           NaN                       NaN        NaN   \n",
       "9004022                     3.077189                       NaN   3.051978   \n",
       "Ca006ONPN                   2.152846                       NaN  -0.938196   \n",
       "Ca006ONPN2                  1.292848                       NaN  -0.480220   \n",
       "\n",
       "            g__Cutibacterium_ASV-1  g___ASV-33  g___ASV-30  g___ASV-22  \\\n",
       "#sample-id                                                               \n",
       "Ca009STPN                      NaN   -1.214339         NaN         NaN   \n",
       "900223                         NaN   -4.064737         NaN         NaN   \n",
       "Ca010EBPN                 1.083265         NaN         NaN         NaN   \n",
       "900462                    0.230859         NaN         NaN         NaN   \n",
       "900053                    0.000050         NaN         NaN         NaN   \n",
       "...                            ...         ...         ...         ...   \n",
       "9003992                   1.738810         NaN         NaN         NaN   \n",
       "900402                    0.691697         NaN         NaN         NaN   \n",
       "9004022                   0.231450         NaN         NaN   -1.021313   \n",
       "Ca006ONPN                      NaN         NaN         NaN         NaN   \n",
       "Ca006ONPN2                0.390609         NaN         NaN         NaN   \n",
       "\n",
       "            g__Cutibacterium_ASV-2  ...  g__Microvirga_ASV-1  \\\n",
       "#sample-id                          ...                        \n",
       "Ca009STPN                      NaN  ...                  NaN   \n",
       "900223                         NaN  ...                  NaN   \n",
       "Ca010EBPN                 0.390118  ...                  NaN   \n",
       "900462                   -0.087595  ...                  NaN   \n",
       "900053                   -1.945860  ...                  NaN   \n",
       "...                            ...  ...                  ...   \n",
       "9003992                   1.539682  ...                  NaN   \n",
       "900402                    0.170400  ...                  NaN   \n",
       "9004022                  -0.461697  ...                  NaN   \n",
       "Ca006ONPN                      NaN  ...                  NaN   \n",
       "Ca006ONPN2               -0.374859  ...                  NaN   \n",
       "\n",
       "            g__Leptotrichia_A_993758_ASV-17  g___ASV-103  \\\n",
       "#sample-id                                                 \n",
       "Ca009STPN                               NaN          NaN   \n",
       "900223                                  NaN          NaN   \n",
       "Ca010EBPN                               NaN          NaN   \n",
       "900462                                  NaN          NaN   \n",
       "900053                                  NaN          NaN   \n",
       "...                                     ...          ...   \n",
       "9003992                                 NaN          NaN   \n",
       "900402                                  NaN          NaN   \n",
       "9004022                                 NaN          NaN   \n",
       "Ca006ONPN                               NaN          NaN   \n",
       "Ca006ONPN2                              NaN          NaN   \n",
       "\n",
       "            g__Leptotrichia_A_993758_ASV-13  g__Capnocytophaga_820688_ASV-8  \\\n",
       "#sample-id                                                                    \n",
       "Ca009STPN                               NaN                             NaN   \n",
       "900223                                  NaN                             NaN   \n",
       "Ca010EBPN                               NaN                             NaN   \n",
       "900462                                  NaN                             NaN   \n",
       "900053                                  NaN                             NaN   \n",
       "...                                     ...                             ...   \n",
       "9003992                                 NaN                             NaN   \n",
       "900402                                  NaN                             NaN   \n",
       "9004022                                 NaN                             NaN   \n",
       "Ca006ONPN                               NaN                             NaN   \n",
       "Ca006ONPN2                              NaN                             NaN   \n",
       "\n",
       "            g__UBA6175_ASV-2  g__Comamonas_F_589250_ASV-3  g__Bosea_ASV-2  \\\n",
       "#sample-id                                                                  \n",
       "Ca009STPN                NaN                          NaN             NaN   \n",
       "900223                   NaN                          NaN             NaN   \n",
       "Ca010EBPN                NaN                          NaN             NaN   \n",
       "900462                   NaN                          NaN             NaN   \n",
       "900053                   NaN                          NaN             NaN   \n",
       "...                      ...                          ...             ...   \n",
       "9003992                  NaN                          NaN             NaN   \n",
       "900402                   NaN                          NaN             NaN   \n",
       "9004022                  NaN                          NaN             NaN   \n",
       "Ca006ONPN                NaN                          NaN             NaN   \n",
       "Ca006ONPN2               NaN                          NaN             NaN   \n",
       "\n",
       "            g___ASV-145     group  \n",
       "#sample-id                         \n",
       "Ca009STPN           NaN  nares-AD  \n",
       "900223              NaN  nares-AD  \n",
       "Ca010EBPN           NaN  nares-AD  \n",
       "900462              NaN  nares-AD  \n",
       "900053              NaN  nares-AD  \n",
       "...                 ...       ...  \n",
       "9003992             NaN  nares-AD  \n",
       "900402              NaN  nares-AD  \n",
       "9004022             NaN  nares-AD  \n",
       "Ca006ONPN           NaN  nares-AD  \n",
       "Ca006ONPN2          NaN  nares-AD  \n",
       "\n",
       "[197 rows x 799 columns]"
      ]
     },
     "execution_count": 224,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Map the 'group' column from metadata to df based on matching index\n",
    "rclr_df['group'] = metadata.loc[rclr_df.index, 'group']\n",
    "rclr_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 225,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "group\n",
       "nares-AD    108\n",
       "nares-H      89\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 225,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rclr_df['group'].value_counts()\n",
    "\n",
    "# rclr_df = rclr_df.rename(columns={\n",
    "#     col: col.replace(\"g__Haemophilus_D_734546\", \"g__Haemophilus\")\n",
    "#     for col in df.columns\n",
    "#     if col.startswith(\"g__Haemophilus_D_734546\")\n",
    "# })\n",
    "\n",
    "\n",
    "# rclr_df = rclr_df.rename(columns={\n",
    "#     col: col.replace(\"g__Veillonella_A\", \"g__Veillonella\")\n",
    "#     for col in df.columns\n",
    "#     if col.startswith(\"g__Veillonella_A\")\n",
    "# })\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 226,
   "metadata": {},
   "outputs": [],
   "source": [
    "def pairwise_mannwhitney_bh(df, feature, group_col='group'):\n",
    "    groups = df[group_col].unique()\n",
    "    comparisons = list(itertools.combinations(groups, 2))\n",
    "\n",
    "    raw_pvals = []\n",
    "    labels = []\n",
    "    pairs = []\n",
    "\n",
    "    for g1, g2 in comparisons:\n",
    "        x = df[df[group_col] == g1][feature].dropna()\n",
    "        y = df[df[group_col] == g2][feature].dropna()\n",
    "        _, p = mannwhitneyu(x, y, alternative='two-sided')\n",
    "        raw_pvals.append(p)\n",
    "        labels.append(f\"{g1} vs {g2}\")\n",
    "        pairs.append((g1, g2))\n",
    "\n",
    "    _, pvals_corrected, _, _ = multipletests(raw_pvals, method='fdr_bh')\n",
    "\n",
    "    results_df = pd.DataFrame({\n",
    "        'Comparison': labels,\n",
    "        'Pair': pairs,\n",
    "        'Raw p-value': raw_pvals,\n",
    "        'BH-corrected p-value': pvals_corrected\n",
    "    })\n",
    "\n",
    "    return results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>g__Streptococcus_ASV-1</th>\n",
       "      <th>g__Streptococcus_ASV-2</th>\n",
       "      <th>g__Corynebacterium_ASV-1</th>\n",
       "      <th>g__Corynebacterium_ASV-4</th>\n",
       "      <th>g___ASV-5</th>\n",
       "      <th>g__Cutibacterium_ASV-1</th>\n",
       "      <th>g___ASV-33</th>\n",
       "      <th>g___ASV-30</th>\n",
       "      <th>g___ASV-22</th>\n",
       "      <th>g__Cutibacterium_ASV-2</th>\n",
       "      <th>...</th>\n",
       "      <th>g__Microvirga_ASV-1</th>\n",
       "      <th>g__Leptotrichia_A_993758_ASV-17</th>\n",
       "      <th>g___ASV-103</th>\n",
       "      <th>g__Leptotrichia_A_993758_ASV-13</th>\n",
       "      <th>g__Capnocytophaga_820688_ASV-8</th>\n",
       "      <th>g__UBA6175_ASV-2</th>\n",
       "      <th>g__Comamonas_F_589250_ASV-3</th>\n",
       "      <th>g__Bosea_ASV-2</th>\n",
       "      <th>g___ASV-145</th>\n",
       "      <th>group</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#sample-id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Ca009STPN</th>\n",
       "      <td>2.024339</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.573153</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.214339</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nares-AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900223</th>\n",
       "      <td>0.489140</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.043399</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-4.064737</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nares-AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca010EBPN</th>\n",
       "      <td>2.783217</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.482631</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.621828</td>\n",
       "      <td>1.083265</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.390118</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nares-AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900462</th>\n",
       "      <td>4.109607</td>\n",
       "      <td>4.015048</td>\n",
       "      <td>3.370298</td>\n",
       "      <td>3.131281</td>\n",
       "      <td>2.477354</td>\n",
       "      <td>0.230859</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.087595</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nares-AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900053</th>\n",
       "      <td>5.869346</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.847248</td>\n",
       "      <td>0.000050</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.945860</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nares-AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9003992</th>\n",
       "      <td>3.255158</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.750389</td>\n",
       "      <td>4.255610</td>\n",
       "      <td>0.170195</td>\n",
       "      <td>1.738810</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.539682</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nares-AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900402</th>\n",
       "      <td>2.275817</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.691697</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.170400</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nares-AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9004022</th>\n",
       "      <td>2.428674</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.077189</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.051978</td>\n",
       "      <td>0.231450</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.021313</td>\n",
       "      <td>-0.461697</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nares-AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca006ONPN</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.152846</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.938196</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nares-AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca006ONPN2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.292848</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.480220</td>\n",
       "      <td>0.390609</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.374859</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nares-AD</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>197 rows × 799 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            g__Streptococcus_ASV-1  g__Streptococcus_ASV-2  \\\n",
       "#sample-id                                                   \n",
       "Ca009STPN                 2.024339                     NaN   \n",
       "900223                    0.489140                     NaN   \n",
       "Ca010EBPN                 2.783217                     NaN   \n",
       "900462                    4.109607                4.015048   \n",
       "900053                    5.869346                     NaN   \n",
       "...                            ...                     ...   \n",
       "9003992                   3.255158                     NaN   \n",
       "900402                    2.275817                     NaN   \n",
       "9004022                   2.428674                     NaN   \n",
       "Ca006ONPN                      NaN                     NaN   \n",
       "Ca006ONPN2                     NaN                     NaN   \n",
       "\n",
       "            g__Corynebacterium_ASV-1  g__Corynebacterium_ASV-4  g___ASV-5  \\\n",
       "#sample-id                                                                  \n",
       "Ca009STPN                   3.573153                       NaN        NaN   \n",
       "900223                      2.043399                       NaN        NaN   \n",
       "Ca010EBPN                   2.482631                       NaN   2.621828   \n",
       "900462                      3.370298                  3.131281   2.477354   \n",
       "900053                           NaN                       NaN  -0.847248   \n",
       "...                              ...                       ...        ...   \n",
       "9003992                     4.750389                  4.255610   0.170195   \n",
       "900402                           NaN                       NaN        NaN   \n",
       "9004022                     3.077189                       NaN   3.051978   \n",
       "Ca006ONPN                   2.152846                       NaN  -0.938196   \n",
       "Ca006ONPN2                  1.292848                       NaN  -0.480220   \n",
       "\n",
       "            g__Cutibacterium_ASV-1  g___ASV-33  g___ASV-30  g___ASV-22  \\\n",
       "#sample-id                                                               \n",
       "Ca009STPN                      NaN   -1.214339         NaN         NaN   \n",
       "900223                         NaN   -4.064737         NaN         NaN   \n",
       "Ca010EBPN                 1.083265         NaN         NaN         NaN   \n",
       "900462                    0.230859         NaN         NaN         NaN   \n",
       "900053                    0.000050         NaN         NaN         NaN   \n",
       "...                            ...         ...         ...         ...   \n",
       "9003992                   1.738810         NaN         NaN         NaN   \n",
       "900402                    0.691697         NaN         NaN         NaN   \n",
       "9004022                   0.231450         NaN         NaN   -1.021313   \n",
       "Ca006ONPN                      NaN         NaN         NaN         NaN   \n",
       "Ca006ONPN2                0.390609         NaN         NaN         NaN   \n",
       "\n",
       "            g__Cutibacterium_ASV-2  ...  g__Microvirga_ASV-1  \\\n",
       "#sample-id                          ...                        \n",
       "Ca009STPN                      NaN  ...                  NaN   \n",
       "900223                         NaN  ...                  NaN   \n",
       "Ca010EBPN                 0.390118  ...                  NaN   \n",
       "900462                   -0.087595  ...                  NaN   \n",
       "900053                   -1.945860  ...                  NaN   \n",
       "...                            ...  ...                  ...   \n",
       "9003992                   1.539682  ...                  NaN   \n",
       "900402                    0.170400  ...                  NaN   \n",
       "9004022                  -0.461697  ...                  NaN   \n",
       "Ca006ONPN                      NaN  ...                  NaN   \n",
       "Ca006ONPN2               -0.374859  ...                  NaN   \n",
       "\n",
       "            g__Leptotrichia_A_993758_ASV-17  g___ASV-103  \\\n",
       "#sample-id                                                 \n",
       "Ca009STPN                               NaN          NaN   \n",
       "900223                                  NaN          NaN   \n",
       "Ca010EBPN                               NaN          NaN   \n",
       "900462                                  NaN          NaN   \n",
       "900053                                  NaN          NaN   \n",
       "...                                     ...          ...   \n",
       "9003992                                 NaN          NaN   \n",
       "900402                                  NaN          NaN   \n",
       "9004022                                 NaN          NaN   \n",
       "Ca006ONPN                               NaN          NaN   \n",
       "Ca006ONPN2                              NaN          NaN   \n",
       "\n",
       "            g__Leptotrichia_A_993758_ASV-13  g__Capnocytophaga_820688_ASV-8  \\\n",
       "#sample-id                                                                    \n",
       "Ca009STPN                               NaN                             NaN   \n",
       "900223                                  NaN                             NaN   \n",
       "Ca010EBPN                               NaN                             NaN   \n",
       "900462                                  NaN                             NaN   \n",
       "900053                                  NaN                             NaN   \n",
       "...                                     ...                             ...   \n",
       "9003992                                 NaN                             NaN   \n",
       "900402                                  NaN                             NaN   \n",
       "9004022                                 NaN                             NaN   \n",
       "Ca006ONPN                               NaN                             NaN   \n",
       "Ca006ONPN2                              NaN                             NaN   \n",
       "\n",
       "            g__UBA6175_ASV-2  g__Comamonas_F_589250_ASV-3  g__Bosea_ASV-2  \\\n",
       "#sample-id                                                                  \n",
       "Ca009STPN                NaN                          NaN             NaN   \n",
       "900223                   NaN                          NaN             NaN   \n",
       "Ca010EBPN                NaN                          NaN             NaN   \n",
       "900462                   NaN                          NaN             NaN   \n",
       "900053                   NaN                          NaN             NaN   \n",
       "...                      ...                          ...             ...   \n",
       "9003992                  NaN                          NaN             NaN   \n",
       "900402                   NaN                          NaN             NaN   \n",
       "9004022                  NaN                          NaN             NaN   \n",
       "Ca006ONPN                NaN                          NaN             NaN   \n",
       "Ca006ONPN2               NaN                          NaN             NaN   \n",
       "\n",
       "            g___ASV-145     group  \n",
       "#sample-id                         \n",
       "Ca009STPN           NaN  nares-AD  \n",
       "900223              NaN  nares-AD  \n",
       "Ca010EBPN           NaN  nares-AD  \n",
       "900462              NaN  nares-AD  \n",
       "900053              NaN  nares-AD  \n",
       "...                 ...       ...  \n",
       "9003992             NaN  nares-AD  \n",
       "900402              NaN  nares-AD  \n",
       "9004022             NaN  nares-AD  \n",
       "Ca006ONPN           NaN  nares-AD  \n",
       "Ca006ONPN2          NaN  nares-AD  \n",
       "\n",
       "[197 rows x 799 columns]"
      ]
     },
     "execution_count": 227,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filter for nares samples\n",
    "nares_samples = rclr_df[rclr_df['group'].isin(['nares-AD', 'nares-H'])]\n",
    "nares_samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 228,
   "metadata": {},
   "outputs": [],
   "source": [
    "def pairwise_mannwhitney_bh(df, feature, group_col='group'):\n",
    "    groups = df[group_col].unique()\n",
    "    comparisons = list(itertools.combinations(groups, 2))\n",
    "\n",
    "    raw_pvals = []\n",
    "    labels = []\n",
    "    pairs = []\n",
    "\n",
    "    for g1, g2 in comparisons:\n",
    "        x = df[df[group_col] == g1][feature].dropna()\n",
    "        y = df[df[group_col] == g2][feature].dropna()\n",
    "\n",
    "        # Skip comparison if either group has no data\n",
    "        if len(x) == 0 or len(y) == 0:\n",
    "            continue\n",
    "\n",
    "        _, p = mannwhitneyu(x, y, alternative='two-sided')\n",
    "        raw_pvals.append(p)\n",
    "        labels.append(f\"{g1} vs {g2}\")\n",
    "        pairs.append((g1, g2))\n",
    "\n",
    "    if len(raw_pvals) == 0:\n",
    "        return pd.DataFrame(columns=['Comparison', 'Pair', 'Raw p-value', 'BH-corrected p-value'])\n",
    "\n",
    "    _, pvals_corrected, _, _ = multipletests(raw_pvals, method='fdr_bh')\n",
    "\n",
    "    results_df = pd.DataFrame({\n",
    "        'Comparison': labels,\n",
    "        'Pair': pairs,\n",
    "        'Raw p-value': raw_pvals,\n",
    "        'BH-corrected p-value': pvals_corrected\n",
    "    })\n",
    "\n",
    "    return results_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 229,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define custom color palette\n",
    "group_palette = {\n",
    "    'nares-H': '#ADD8E6',     # baby blue\n",
    "    'nares-AD': '#E31A1C'    # red\n",
    "}\n",
    "\n",
    "strip_palette = {\n",
    "    'nares-H': '#6CA6CD',     # darker baby blue\n",
    "    'nares-AD': '#A50000'    # darker red\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 230,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_multi_taxa_boxplots(data, taxa_list, group_palette, strip_palette, title_name, order=['nares-H', 'nares-AD']):\n",
    "    n_taxa = len(taxa_list)\n",
    "    group_count = len(order)\n",
    "\n",
    "    # Prepare long format dataframe\n",
    "    plot_data = []\n",
    "    for i, taxon in enumerate(taxa_list):\n",
    "        for j, group in enumerate(order):\n",
    "            xpos = i * group_count + j\n",
    "            values = data.loc[data['group'] == group, taxon]\n",
    "            for v in values:\n",
    "                plot_data.append({\n",
    "                    'x': xpos,\n",
    "                    'taxon': taxon.replace(' g__', '').split('_')[0],\n",
    "                    'value': v,\n",
    "                    'group': group\n",
    "                })\n",
    "\n",
    "    plot_df = pd.DataFrame(plot_data)\n",
    "\n",
    "    # Plot\n",
    "    fig, ax = plt.subplots(figsize=(n_taxa * 2.5, 5))\n",
    "\n",
    "    sns.boxplot(\n",
    "        data=plot_df,\n",
    "        x='x',\n",
    "        y='value',\n",
    "        hue='group',\n",
    "        palette=group_palette,\n",
    "        width=0.5,\n",
    "        fliersize=0,\n",
    "        dodge=False,\n",
    "        ax=ax\n",
    "    )\n",
    "\n",
    "    sns.stripplot(\n",
    "        data=plot_df,\n",
    "        x='x',\n",
    "        y='value',\n",
    "        hue='group',\n",
    "        palette=strip_palette,\n",
    "        dodge=False,\n",
    "        jitter=True,\n",
    "        size=4,\n",
    "        alpha=0.75,\n",
    "        ax=ax\n",
    "    )\n",
    "\n",
    "    # Remove duplicated legends\n",
    "    handles, labels = ax.get_legend_handles_labels()\n",
    "\n",
    "    # Count samples per group\n",
    "    group_counts = {g: data[data['group'] == g].shape[0] for g in order}\n",
    "\n",
    "    # Create updated legend labels\n",
    "    legend_labels = [f\"{g.replace('skin-', '')} (n={group_counts[g]})\" for g in order]\n",
    "\n",
    "    # Add updated legend — horizontally at bottom-left\n",
    "    ax.legend(\n",
    "        handles[:3],\n",
    "        legend_labels,\n",
    "        # title='Sample',\n",
    "        loc='lower left',\n",
    "        bbox_to_anchor=(0.35, -0.35),  # adjust y for spacing below plot\n",
    "        ncol=len(order),            # show horizontally\n",
    "        fontsize=12,\n",
    "        title_fontsize=10,\n",
    "        frameon=True,\n",
    "        borderaxespad=0\n",
    "    )\n",
    "\n",
    "\n",
    "    # X-tick labels per taxon group\n",
    "    xtick_positions = [i * group_count + 1 for i in range(n_taxa)]\n",
    "    # xtick_labels = [taxon.replace('g__', '').split('_')[0] for taxon in taxa_list]\n",
    "    xtick_labels = [taxon.replace('g__', '').replace('_ASV-', '\\nASV-') for taxon in taxa_list]\n",
    "    ax.set_xticks(xtick_positions)\n",
    "    ax.set_xticklabels(xtick_labels, fontsize=14)\n",
    "\n",
    "    ax.set_xlim(-0.5, n_taxa * group_count - 0.5)\n",
    "    ax.set_ylabel(\"RCLR-transformed abundance\", fontsize=14)\n",
    "    ax.tick_params(axis='y', labelsize=14)\n",
    "    ax.set_xlabel(\" \")\n",
    "\n",
    "    ax.set_title(title_name, fontsize=18, pad=14)\n",
    "\n",
    "    # ------------------------\n",
    "    # Annotate p-values\n",
    "    # ------------------------\n",
    "    for i, taxon in enumerate(taxa_list):\n",
    "        stats = pairwise_mannwhitney_bh(data, taxon)\n",
    "        y_max = data[taxon].max()\n",
    "        offset = y_max * 0.15\n",
    "\n",
    "        for k, row in stats.iterrows():\n",
    "            g1, g2 = row['Pair']\n",
    "            pval = row['BH-corrected p-value']\n",
    "            xpos1 = i * group_count + order.index(g1)\n",
    "            xpos2 = i * group_count + order.index(g2)\n",
    "            x = (xpos1 + xpos2) / 2\n",
    "            y = y_max + offset * k\n",
    "\n",
    "            stars = '***' if pval < 0.001 else '**' if pval < 0.01 else '*' if pval < 0.05 else ''\n",
    "            label = f\"{pval:.1e} {stars}\"\n",
    "\n",
    "            ax.plot([xpos1, xpos1, xpos2, xpos2], [y - 0.01, y, y, y - 0.01], lw=1, color='black')\n",
    "            ax.text(x, y + 0.01, label, ha='center', va='bottom', fontsize=12)\n",
    "\n",
    "    sns.despine()\n",
    "    fig.tight_layout()\n",
    "    return fig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 231,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.\n",
      "  with pd.option_context('mode.use_inf_as_na', True):\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.\n",
      "  with pd.option_context('mode.use_inf_as_na', True):\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1075: FutureWarning: When grouping with a length-1 list-like, you will need to pass a length-1 tuple to get_group in a future version of pandas. Pass `(name,)` instead of `name` to silence this warning.\n",
      "  data_subset = grouped_data.get_group(pd_key)\n"
     ]
    }
   ],
   "source": [
    "taxa_list = ['g__Streptococcus_ASV-1', 'g__Streptococcus_ASV-2', 'g__Staphylococcus_ASV-1',\n",
    "            'g__Micrococcus_ASV-1', 'g__Veillonella_A_ASV-1', 'g__Haemophilus_D_734546_ASV-1']\n",
    "\n",
    "# taxa_list = [' g__Staphylococcus', ' g__Streptococcus',\n",
    "#             ' g__Micrococcus', ' g__Veillonella_A', ' g__Haemophilus_D_734546']\n",
    "\n",
    "# taxa_list = ['g__Streptococcus_ASV-1']\n",
    "\n",
    "fig = plot_multi_taxa_boxplots(nares_samples, taxa_list, group_palette, strip_palette, 'Differential Nares Taxa of Umtata Samples')\n",
    "fig.savefig('../Plots/Analysis_figures/RCLR_Abundance/rclr_multi_taxa_boxplot_all_taxa_ASV_nares_Umtata.png', dpi=600)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Show correlation with lesion severity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 232,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>g__Streptococcus_ASV-1</th>\n",
       "      <th>g__Streptococcus_ASV-2</th>\n",
       "      <th>g__Corynebacterium_ASV-1</th>\n",
       "      <th>g__Corynebacterium_ASV-4</th>\n",
       "      <th>g___ASV-5</th>\n",
       "      <th>g__Cutibacterium_ASV-1</th>\n",
       "      <th>g___ASV-33</th>\n",
       "      <th>g___ASV-30</th>\n",
       "      <th>g___ASV-22</th>\n",
       "      <th>g__Cutibacterium_ASV-2</th>\n",
       "      <th>...</th>\n",
       "      <th>g___ASV-103</th>\n",
       "      <th>g__Leptotrichia_A_993758_ASV-13</th>\n",
       "      <th>g__Capnocytophaga_820688_ASV-8</th>\n",
       "      <th>g__UBA6175_ASV-2</th>\n",
       "      <th>g__Comamonas_F_589250_ASV-3</th>\n",
       "      <th>g__Bosea_ASV-2</th>\n",
       "      <th>g___ASV-145</th>\n",
       "      <th>group</th>\n",
       "      <th>o_scorad</th>\n",
       "      <th>area</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#sample-id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>900223</th>\n",
       "      <td>0.489140</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.043399</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-4.064737</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nares-AD</td>\n",
       "      <td>34</td>\n",
       "      <td>Umtata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900462</th>\n",
       "      <td>4.109607</td>\n",
       "      <td>4.015048</td>\n",
       "      <td>3.370298</td>\n",
       "      <td>3.131281</td>\n",
       "      <td>2.477354</td>\n",
       "      <td>0.230859</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.087595</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nares-AD</td>\n",
       "      <td>40</td>\n",
       "      <td>Umtata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900228</th>\n",
       "      <td>5.928436</td>\n",
       "      <td>4.161521</td>\n",
       "      <td>4.350885</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.483926</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nares-AD</td>\n",
       "      <td>34</td>\n",
       "      <td>Umtata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900231</th>\n",
       "      <td>2.717672</td>\n",
       "      <td>2.194084</td>\n",
       "      <td>1.846542</td>\n",
       "      <td>1.309399</td>\n",
       "      <td>1.453980</td>\n",
       "      <td>-1.398651</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.804116</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nares-AD</td>\n",
       "      <td>23</td>\n",
       "      <td>Umtata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900242</th>\n",
       "      <td>1.392800</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.082003</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-2.352774</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nares-H</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Umtata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9003962</th>\n",
       "      <td>1.564364</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.630157</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-2.242297</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nares-AD</td>\n",
       "      <td>28</td>\n",
       "      <td>Umtata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900399</th>\n",
       "      <td>0.250958</td>\n",
       "      <td>0.056802</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.502814</td>\n",
       "      <td>3.381324</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.993694</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nares-AD</td>\n",
       "      <td>54</td>\n",
       "      <td>Umtata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9003992</th>\n",
       "      <td>3.255158</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.750389</td>\n",
       "      <td>4.255610</td>\n",
       "      <td>0.170195</td>\n",
       "      <td>1.738810</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.539682</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nares-AD</td>\n",
       "      <td>54</td>\n",
       "      <td>Umtata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900402</th>\n",
       "      <td>2.275817</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.691697</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.170400</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nares-AD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Umtata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9004022</th>\n",
       "      <td>2.428674</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.077189</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.051978</td>\n",
       "      <td>0.231450</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.021313</td>\n",
       "      <td>-0.461697</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nares-AD</td>\n",
       "      <td>38</td>\n",
       "      <td>Umtata</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>116 rows × 801 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            g__Streptococcus_ASV-1  g__Streptococcus_ASV-2  \\\n",
       "#sample-id                                                   \n",
       "900223                    0.489140                     NaN   \n",
       "900462                    4.109607                4.015048   \n",
       "900228                    5.928436                4.161521   \n",
       "900231                    2.717672                2.194084   \n",
       "900242                    1.392800                     NaN   \n",
       "...                            ...                     ...   \n",
       "9003962                   1.564364                     NaN   \n",
       "900399                    0.250958                0.056802   \n",
       "9003992                   3.255158                     NaN   \n",
       "900402                    2.275817                     NaN   \n",
       "9004022                   2.428674                     NaN   \n",
       "\n",
       "            g__Corynebacterium_ASV-1  g__Corynebacterium_ASV-4  g___ASV-5  \\\n",
       "#sample-id                                                                  \n",
       "900223                      2.043399                       NaN        NaN   \n",
       "900462                      3.370298                  3.131281   2.477354   \n",
       "900228                      4.350885                       NaN        NaN   \n",
       "900231                      1.846542                  1.309399   1.453980   \n",
       "900242                      2.082003                       NaN        NaN   \n",
       "...                              ...                       ...        ...   \n",
       "9003962                     5.630157                       NaN  -2.242297   \n",
       "900399                           NaN                       NaN  -0.502814   \n",
       "9003992                     4.750389                  4.255610   0.170195   \n",
       "900402                           NaN                       NaN        NaN   \n",
       "9004022                     3.077189                       NaN   3.051978   \n",
       "\n",
       "            g__Cutibacterium_ASV-1  g___ASV-33  g___ASV-30  g___ASV-22  \\\n",
       "#sample-id                                                               \n",
       "900223                         NaN   -4.064737         NaN         NaN   \n",
       "900462                    0.230859         NaN         NaN         NaN   \n",
       "900228                   -1.483926         NaN         NaN         NaN   \n",
       "900231                   -1.398651         NaN         NaN         NaN   \n",
       "900242                         NaN   -2.352774         NaN         NaN   \n",
       "...                            ...         ...         ...         ...   \n",
       "9003962                        NaN         NaN         NaN         NaN   \n",
       "900399                    3.381324         NaN         NaN         NaN   \n",
       "9003992                   1.738810         NaN         NaN         NaN   \n",
       "900402                    0.691697         NaN         NaN         NaN   \n",
       "9004022                   0.231450         NaN         NaN   -1.021313   \n",
       "\n",
       "            g__Cutibacterium_ASV-2  ...  g___ASV-103  \\\n",
       "#sample-id                          ...                \n",
       "900223                         NaN  ...          NaN   \n",
       "900462                   -0.087595  ...          NaN   \n",
       "900228                         NaN  ...          NaN   \n",
       "900231                   -1.804116  ...          NaN   \n",
       "900242                         NaN  ...          NaN   \n",
       "...                            ...  ...          ...   \n",
       "9003962                        NaN  ...          NaN   \n",
       "900399                    2.993694  ...          NaN   \n",
       "9003992                   1.539682  ...          NaN   \n",
       "900402                    0.170400  ...          NaN   \n",
       "9004022                  -0.461697  ...          NaN   \n",
       "\n",
       "            g__Leptotrichia_A_993758_ASV-13  g__Capnocytophaga_820688_ASV-8  \\\n",
       "#sample-id                                                                    \n",
       "900223                                  NaN                             NaN   \n",
       "900462                                  NaN                             NaN   \n",
       "900228                                  NaN                             NaN   \n",
       "900231                                  NaN                             NaN   \n",
       "900242                                  NaN                             NaN   \n",
       "...                                     ...                             ...   \n",
       "9003962                                 NaN                             NaN   \n",
       "900399                                  NaN                             NaN   \n",
       "9003992                                 NaN                             NaN   \n",
       "900402                                  NaN                             NaN   \n",
       "9004022                                 NaN                             NaN   \n",
       "\n",
       "            g__UBA6175_ASV-2  g__Comamonas_F_589250_ASV-3  g__Bosea_ASV-2  \\\n",
       "#sample-id                                                                  \n",
       "900223                   NaN                          NaN             NaN   \n",
       "900462                   NaN                          NaN             NaN   \n",
       "900228                   NaN                          NaN             NaN   \n",
       "900231                   NaN                          NaN             NaN   \n",
       "900242                   NaN                          NaN             NaN   \n",
       "...                      ...                          ...             ...   \n",
       "9003962                  NaN                          NaN             NaN   \n",
       "900399                   NaN                          NaN             NaN   \n",
       "9003992                  NaN                          NaN             NaN   \n",
       "900402                   NaN                          NaN             NaN   \n",
       "9004022                  NaN                          NaN             NaN   \n",
       "\n",
       "            g___ASV-145     group  o_scorad    area  \n",
       "#sample-id                                           \n",
       "900223              NaN  nares-AD        34  Umtata  \n",
       "900462              NaN  nares-AD        40  Umtata  \n",
       "900228              NaN  nares-AD        34  Umtata  \n",
       "900231              NaN  nares-AD        23  Umtata  \n",
       "900242              NaN   nares-H       NaN  Umtata  \n",
       "...                 ...       ...       ...     ...  \n",
       "9003962             NaN  nares-AD        28  Umtata  \n",
       "900399              NaN  nares-AD        54  Umtata  \n",
       "9003992             NaN  nares-AD        54  Umtata  \n",
       "900402              NaN  nares-AD       NaN  Umtata  \n",
       "9004022             NaN  nares-AD        38  Umtata  \n",
       "\n",
       "[116 rows x 801 columns]"
      ]
     },
     "execution_count": 232,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Map o_scorad from metadata to nares_samples based on matching indexes\n",
    "nares_samples = nares_samples.merge(metadata[['o_scorad']], left_index=True, right_index=True)\n",
    "# Map area from metadata to nares_samples based on matching indexes\n",
    "nares_samples = nares_samples.merge(metadata[['area']], left_index=True, right_index=True)\n",
    "\n",
    "# Filter to only include samples from Umtata\n",
    "nares_samples = nares_samples[nares_samples['area'] == 'Umtata']\n",
    "\n",
    "nares_samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 233,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>g__Streptococcus_ASV-1</th>\n",
       "      <th>g__Streptococcus_ASV-2</th>\n",
       "      <th>g__Corynebacterium_ASV-1</th>\n",
       "      <th>g__Corynebacterium_ASV-4</th>\n",
       "      <th>g___ASV-5</th>\n",
       "      <th>g__Cutibacterium_ASV-1</th>\n",
       "      <th>g___ASV-33</th>\n",
       "      <th>g___ASV-30</th>\n",
       "      <th>g___ASV-22</th>\n",
       "      <th>g__Cutibacterium_ASV-2</th>\n",
       "      <th>...</th>\n",
       "      <th>g__Leptotrichia_A_993758_ASV-17</th>\n",
       "      <th>g___ASV-103</th>\n",
       "      <th>g__Leptotrichia_A_993758_ASV-13</th>\n",
       "      <th>g__Capnocytophaga_820688_ASV-8</th>\n",
       "      <th>g__UBA6175_ASV-2</th>\n",
       "      <th>g__Comamonas_F_589250_ASV-3</th>\n",
       "      <th>g__Bosea_ASV-2</th>\n",
       "      <th>g___ASV-145</th>\n",
       "      <th>group</th>\n",
       "      <th>o_scorad</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#sample-id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Ca009STPN</th>\n",
       "      <td>2.024339</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.573153</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.214339</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nares-AD</td>\n",
       "      <td>40.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900223</th>\n",
       "      <td>0.489140</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.043399</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-4.064737</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nares-AD</td>\n",
       "      <td>34.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca010EBPN</th>\n",
       "      <td>2.783217</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.482631</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.621828</td>\n",
       "      <td>1.083265</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.390118</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nares-AD</td>\n",
       "      <td>21.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900462</th>\n",
       "      <td>4.109607</td>\n",
       "      <td>4.015048</td>\n",
       "      <td>3.370298</td>\n",
       "      <td>3.131281</td>\n",
       "      <td>2.477354</td>\n",
       "      <td>0.230859</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.087595</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nares-AD</td>\n",
       "      <td>40.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900053</th>\n",
       "      <td>5.869346</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.847248</td>\n",
       "      <td>0.000050</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.945860</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nares-AD</td>\n",
       "      <td>41.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900399</th>\n",
       "      <td>0.250958</td>\n",
       "      <td>0.056802</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.502814</td>\n",
       "      <td>3.381324</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.993694</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nares-AD</td>\n",
       "      <td>54.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9003992</th>\n",
       "      <td>3.255158</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.750389</td>\n",
       "      <td>4.255610</td>\n",
       "      <td>0.170195</td>\n",
       "      <td>1.738810</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.539682</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nares-AD</td>\n",
       "      <td>54.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9004022</th>\n",
       "      <td>2.428674</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.077189</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.051978</td>\n",
       "      <td>0.231450</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.021313</td>\n",
       "      <td>-0.461697</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nares-AD</td>\n",
       "      <td>38.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca006ONPN</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.152846</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.938196</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nares-AD</td>\n",
       "      <td>34.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca006ONPN2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.292848</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.480220</td>\n",
       "      <td>0.390609</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.374859</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nares-AD</td>\n",
       "      <td>34.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>102 rows × 800 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            g__Streptococcus_ASV-1  g__Streptococcus_ASV-2  \\\n",
       "#sample-id                                                   \n",
       "Ca009STPN                 2.024339                     NaN   \n",
       "900223                    0.489140                     NaN   \n",
       "Ca010EBPN                 2.783217                     NaN   \n",
       "900462                    4.109607                4.015048   \n",
       "900053                    5.869346                     NaN   \n",
       "...                            ...                     ...   \n",
       "900399                    0.250958                0.056802   \n",
       "9003992                   3.255158                     NaN   \n",
       "9004022                   2.428674                     NaN   \n",
       "Ca006ONPN                      NaN                     NaN   \n",
       "Ca006ONPN2                     NaN                     NaN   \n",
       "\n",
       "            g__Corynebacterium_ASV-1  g__Corynebacterium_ASV-4  g___ASV-5  \\\n",
       "#sample-id                                                                  \n",
       "Ca009STPN                   3.573153                       NaN        NaN   \n",
       "900223                      2.043399                       NaN        NaN   \n",
       "Ca010EBPN                   2.482631                       NaN   2.621828   \n",
       "900462                      3.370298                  3.131281   2.477354   \n",
       "900053                           NaN                       NaN  -0.847248   \n",
       "...                              ...                       ...        ...   \n",
       "900399                           NaN                       NaN  -0.502814   \n",
       "9003992                     4.750389                  4.255610   0.170195   \n",
       "9004022                     3.077189                       NaN   3.051978   \n",
       "Ca006ONPN                   2.152846                       NaN  -0.938196   \n",
       "Ca006ONPN2                  1.292848                       NaN  -0.480220   \n",
       "\n",
       "            g__Cutibacterium_ASV-1  g___ASV-33  g___ASV-30  g___ASV-22  \\\n",
       "#sample-id                                                               \n",
       "Ca009STPN                      NaN   -1.214339         NaN         NaN   \n",
       "900223                         NaN   -4.064737         NaN         NaN   \n",
       "Ca010EBPN                 1.083265         NaN         NaN         NaN   \n",
       "900462                    0.230859         NaN         NaN         NaN   \n",
       "900053                    0.000050         NaN         NaN         NaN   \n",
       "...                            ...         ...         ...         ...   \n",
       "900399                    3.381324         NaN         NaN         NaN   \n",
       "9003992                   1.738810         NaN         NaN         NaN   \n",
       "9004022                   0.231450         NaN         NaN   -1.021313   \n",
       "Ca006ONPN                      NaN         NaN         NaN         NaN   \n",
       "Ca006ONPN2                0.390609         NaN         NaN         NaN   \n",
       "\n",
       "            g__Cutibacterium_ASV-2  ...  g__Leptotrichia_A_993758_ASV-17  \\\n",
       "#sample-id                          ...                                    \n",
       "Ca009STPN                      NaN  ...                              NaN   \n",
       "900223                         NaN  ...                              NaN   \n",
       "Ca010EBPN                 0.390118  ...                              NaN   \n",
       "900462                   -0.087595  ...                              NaN   \n",
       "900053                   -1.945860  ...                              NaN   \n",
       "...                            ...  ...                              ...   \n",
       "900399                    2.993694  ...                              NaN   \n",
       "9003992                   1.539682  ...                              NaN   \n",
       "9004022                  -0.461697  ...                              NaN   \n",
       "Ca006ONPN                      NaN  ...                              NaN   \n",
       "Ca006ONPN2               -0.374859  ...                              NaN   \n",
       "\n",
       "            g___ASV-103  g__Leptotrichia_A_993758_ASV-13  \\\n",
       "#sample-id                                                 \n",
       "Ca009STPN           NaN                              NaN   \n",
       "900223              NaN                              NaN   \n",
       "Ca010EBPN           NaN                              NaN   \n",
       "900462              NaN                              NaN   \n",
       "900053              NaN                              NaN   \n",
       "...                 ...                              ...   \n",
       "900399              NaN                              NaN   \n",
       "9003992             NaN                              NaN   \n",
       "9004022             NaN                              NaN   \n",
       "Ca006ONPN           NaN                              NaN   \n",
       "Ca006ONPN2          NaN                              NaN   \n",
       "\n",
       "            g__Capnocytophaga_820688_ASV-8  g__UBA6175_ASV-2  \\\n",
       "#sample-id                                                     \n",
       "Ca009STPN                              NaN               NaN   \n",
       "900223                                 NaN               NaN   \n",
       "Ca010EBPN                              NaN               NaN   \n",
       "900462                                 NaN               NaN   \n",
       "900053                                 NaN               NaN   \n",
       "...                                    ...               ...   \n",
       "900399                                 NaN               NaN   \n",
       "9003992                                NaN               NaN   \n",
       "9004022                                NaN               NaN   \n",
       "Ca006ONPN                              NaN               NaN   \n",
       "Ca006ONPN2                             NaN               NaN   \n",
       "\n",
       "            g__Comamonas_F_589250_ASV-3  g__Bosea_ASV-2  g___ASV-145  \\\n",
       "#sample-id                                                             \n",
       "Ca009STPN                           NaN             NaN          NaN   \n",
       "900223                              NaN             NaN          NaN   \n",
       "Ca010EBPN                           NaN             NaN          NaN   \n",
       "900462                              NaN             NaN          NaN   \n",
       "900053                              NaN             NaN          NaN   \n",
       "...                                 ...             ...          ...   \n",
       "900399                              NaN             NaN          NaN   \n",
       "9003992                             NaN             NaN          NaN   \n",
       "9004022                             NaN             NaN          NaN   \n",
       "Ca006ONPN                           NaN             NaN          NaN   \n",
       "Ca006ONPN2                          NaN             NaN          NaN   \n",
       "\n",
       "               group  o_scorad  \n",
       "#sample-id                      \n",
       "Ca009STPN   nares-AD      40.0  \n",
       "900223      nares-AD      34.0  \n",
       "Ca010EBPN   nares-AD      21.0  \n",
       "900462      nares-AD      40.0  \n",
       "900053      nares-AD      41.0  \n",
       "...              ...       ...  \n",
       "900399      nares-AD      54.0  \n",
       "9003992     nares-AD      54.0  \n",
       "9004022     nares-AD      38.0  \n",
       "Ca006ONPN   nares-AD      34.0  \n",
       "Ca006ONPN2  nares-AD      34.0  \n",
       "\n",
       "[102 rows x 800 columns]"
      ]
     },
     "execution_count": 233,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Convert 'o_scorad' to numeric (coerce errors to NaN)\n",
    "rclr_df['o_scorad'] = pd.to_numeric(metadata['o_scorad'], errors='coerce')\n",
    "\n",
    "# Drop rows with missing SCORAD values\n",
    "rclr_df = rclr_df.dropna(subset=['o_scorad'])\n",
    "rclr_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 234,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>g__Streptococcus_ASV-1</th>\n",
       "      <th>g__Streptococcus_ASV-2</th>\n",
       "      <th>g__Corynebacterium_ASV-1</th>\n",
       "      <th>g__Corynebacterium_ASV-4</th>\n",
       "      <th>g___ASV-5</th>\n",
       "      <th>g__Cutibacterium_ASV-1</th>\n",
       "      <th>g___ASV-33</th>\n",
       "      <th>g___ASV-30</th>\n",
       "      <th>g___ASV-22</th>\n",
       "      <th>g__Cutibacterium_ASV-2</th>\n",
       "      <th>...</th>\n",
       "      <th>g__Leptotrichia_A_993758_ASV-17</th>\n",
       "      <th>g___ASV-103</th>\n",
       "      <th>g__Leptotrichia_A_993758_ASV-13</th>\n",
       "      <th>g__Capnocytophaga_820688_ASV-8</th>\n",
       "      <th>g__UBA6175_ASV-2</th>\n",
       "      <th>g__Comamonas_F_589250_ASV-3</th>\n",
       "      <th>g__Bosea_ASV-2</th>\n",
       "      <th>g___ASV-145</th>\n",
       "      <th>group</th>\n",
       "      <th>o_scorad</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#sample-id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Ca009STPN</th>\n",
       "      <td>2.024339</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.573153</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.214339</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nares-AD</td>\n",
       "      <td>40.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900223</th>\n",
       "      <td>0.489140</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.043399</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-4.064737</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nares-AD</td>\n",
       "      <td>34.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca010EBPN</th>\n",
       "      <td>2.783217</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.482631</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.621828</td>\n",
       "      <td>1.083265</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.390118</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nares-AD</td>\n",
       "      <td>21.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900462</th>\n",
       "      <td>4.109607</td>\n",
       "      <td>4.015048</td>\n",
       "      <td>3.370298</td>\n",
       "      <td>3.131281</td>\n",
       "      <td>2.477354</td>\n",
       "      <td>0.230859</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.087595</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nares-AD</td>\n",
       "      <td>40.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900053</th>\n",
       "      <td>5.869346</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.847248</td>\n",
       "      <td>0.000050</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.945860</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nares-AD</td>\n",
       "      <td>41.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900399</th>\n",
       "      <td>0.250958</td>\n",
       "      <td>0.056802</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.502814</td>\n",
       "      <td>3.381324</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.993694</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nares-AD</td>\n",
       "      <td>54.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9003992</th>\n",
       "      <td>3.255158</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.750389</td>\n",
       "      <td>4.255610</td>\n",
       "      <td>0.170195</td>\n",
       "      <td>1.738810</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.539682</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nares-AD</td>\n",
       "      <td>54.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9004022</th>\n",
       "      <td>2.428674</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.077189</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.051978</td>\n",
       "      <td>0.231450</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.021313</td>\n",
       "      <td>-0.461697</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nares-AD</td>\n",
       "      <td>38.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca006ONPN</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.152846</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.938196</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nares-AD</td>\n",
       "      <td>34.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca006ONPN2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.292848</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.480220</td>\n",
       "      <td>0.390609</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.374859</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nares-AD</td>\n",
       "      <td>34.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>102 rows × 800 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            g__Streptococcus_ASV-1  g__Streptococcus_ASV-2  \\\n",
       "#sample-id                                                   \n",
       "Ca009STPN                 2.024339                     NaN   \n",
       "900223                    0.489140                     NaN   \n",
       "Ca010EBPN                 2.783217                     NaN   \n",
       "900462                    4.109607                4.015048   \n",
       "900053                    5.869346                     NaN   \n",
       "...                            ...                     ...   \n",
       "900399                    0.250958                0.056802   \n",
       "9003992                   3.255158                     NaN   \n",
       "9004022                   2.428674                     NaN   \n",
       "Ca006ONPN                      NaN                     NaN   \n",
       "Ca006ONPN2                     NaN                     NaN   \n",
       "\n",
       "            g__Corynebacterium_ASV-1  g__Corynebacterium_ASV-4  g___ASV-5  \\\n",
       "#sample-id                                                                  \n",
       "Ca009STPN                   3.573153                       NaN        NaN   \n",
       "900223                      2.043399                       NaN        NaN   \n",
       "Ca010EBPN                   2.482631                       NaN   2.621828   \n",
       "900462                      3.370298                  3.131281   2.477354   \n",
       "900053                           NaN                       NaN  -0.847248   \n",
       "...                              ...                       ...        ...   \n",
       "900399                           NaN                       NaN  -0.502814   \n",
       "9003992                     4.750389                  4.255610   0.170195   \n",
       "9004022                     3.077189                       NaN   3.051978   \n",
       "Ca006ONPN                   2.152846                       NaN  -0.938196   \n",
       "Ca006ONPN2                  1.292848                       NaN  -0.480220   \n",
       "\n",
       "            g__Cutibacterium_ASV-1  g___ASV-33  g___ASV-30  g___ASV-22  \\\n",
       "#sample-id                                                               \n",
       "Ca009STPN                      NaN   -1.214339         NaN         NaN   \n",
       "900223                         NaN   -4.064737         NaN         NaN   \n",
       "Ca010EBPN                 1.083265         NaN         NaN         NaN   \n",
       "900462                    0.230859         NaN         NaN         NaN   \n",
       "900053                    0.000050         NaN         NaN         NaN   \n",
       "...                            ...         ...         ...         ...   \n",
       "900399                    3.381324         NaN         NaN         NaN   \n",
       "9003992                   1.738810         NaN         NaN         NaN   \n",
       "9004022                   0.231450         NaN         NaN   -1.021313   \n",
       "Ca006ONPN                      NaN         NaN         NaN         NaN   \n",
       "Ca006ONPN2                0.390609         NaN         NaN         NaN   \n",
       "\n",
       "            g__Cutibacterium_ASV-2  ...  g__Leptotrichia_A_993758_ASV-17  \\\n",
       "#sample-id                          ...                                    \n",
       "Ca009STPN                      NaN  ...                              NaN   \n",
       "900223                         NaN  ...                              NaN   \n",
       "Ca010EBPN                 0.390118  ...                              NaN   \n",
       "900462                   -0.087595  ...                              NaN   \n",
       "900053                   -1.945860  ...                              NaN   \n",
       "...                            ...  ...                              ...   \n",
       "900399                    2.993694  ...                              NaN   \n",
       "9003992                   1.539682  ...                              NaN   \n",
       "9004022                  -0.461697  ...                              NaN   \n",
       "Ca006ONPN                      NaN  ...                              NaN   \n",
       "Ca006ONPN2               -0.374859  ...                              NaN   \n",
       "\n",
       "            g___ASV-103  g__Leptotrichia_A_993758_ASV-13  \\\n",
       "#sample-id                                                 \n",
       "Ca009STPN           NaN                              NaN   \n",
       "900223              NaN                              NaN   \n",
       "Ca010EBPN           NaN                              NaN   \n",
       "900462              NaN                              NaN   \n",
       "900053              NaN                              NaN   \n",
       "...                 ...                              ...   \n",
       "900399              NaN                              NaN   \n",
       "9003992             NaN                              NaN   \n",
       "9004022             NaN                              NaN   \n",
       "Ca006ONPN           NaN                              NaN   \n",
       "Ca006ONPN2          NaN                              NaN   \n",
       "\n",
       "            g__Capnocytophaga_820688_ASV-8  g__UBA6175_ASV-2  \\\n",
       "#sample-id                                                     \n",
       "Ca009STPN                              NaN               NaN   \n",
       "900223                                 NaN               NaN   \n",
       "Ca010EBPN                              NaN               NaN   \n",
       "900462                                 NaN               NaN   \n",
       "900053                                 NaN               NaN   \n",
       "...                                    ...               ...   \n",
       "900399                                 NaN               NaN   \n",
       "9003992                                NaN               NaN   \n",
       "9004022                                NaN               NaN   \n",
       "Ca006ONPN                              NaN               NaN   \n",
       "Ca006ONPN2                             NaN               NaN   \n",
       "\n",
       "            g__Comamonas_F_589250_ASV-3  g__Bosea_ASV-2  g___ASV-145  \\\n",
       "#sample-id                                                             \n",
       "Ca009STPN                           NaN             NaN          NaN   \n",
       "900223                              NaN             NaN          NaN   \n",
       "Ca010EBPN                           NaN             NaN          NaN   \n",
       "900462                              NaN             NaN          NaN   \n",
       "900053                              NaN             NaN          NaN   \n",
       "...                                 ...             ...          ...   \n",
       "900399                              NaN             NaN          NaN   \n",
       "9003992                             NaN             NaN          NaN   \n",
       "9004022                             NaN             NaN          NaN   \n",
       "Ca006ONPN                           NaN             NaN          NaN   \n",
       "Ca006ONPN2                          NaN             NaN          NaN   \n",
       "\n",
       "               group  o_scorad  \n",
       "#sample-id                      \n",
       "Ca009STPN   nares-AD      40.0  \n",
       "900223      nares-AD      34.0  \n",
       "Ca010EBPN   nares-AD      21.0  \n",
       "900462      nares-AD      40.0  \n",
       "900053      nares-AD      41.0  \n",
       "...              ...       ...  \n",
       "900399      nares-AD      54.0  \n",
       "9003992     nares-AD      54.0  \n",
       "9004022     nares-AD      38.0  \n",
       "Ca006ONPN   nares-AD      34.0  \n",
       "Ca006ONPN2  nares-AD      34.0  \n",
       "\n",
       "[102 rows x 800 columns]"
      ]
     },
     "execution_count": 234,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filter to only include skin-ADL samples for severity correlation analysis\n",
    "rclr_df = rclr_df[rclr_df['group'] == 'nares-AD']\n",
    "rclr_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 235,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_combined_box_and_severity(data, taxa_list, group_palette, strip_palette, title_name, order=['nares-H', 'nares-AD']):\n",
    "    n_taxa = len(taxa_list)\n",
    "    fig, axes = plt.subplots(\n",
    "        nrows=n_taxa,\n",
    "        ncols=2,\n",
    "        figsize=(6, 3 * n_taxa),\n",
    "        sharey='row',\n",
    "        gridspec_kw={'width_ratios': [2, 1.5]}\n",
    "    )\n",
    "\n",
    "    if n_taxa == 1:\n",
    "        axes = np.array([axes])\n",
    "\n",
    "    title_y_positions = [0.91, 0.73, 0.55, 0.36, 0.175]\n",
    "\n",
    "    for i, taxon in enumerate(taxa_list):\n",
    "        ax_box = axes[i, 0] if n_taxa > 1 else axes[0]\n",
    "        ax_corr = axes[i, 1] if n_taxa > 1 else axes[1]\n",
    "\n",
    "        # --- Boxplot data ---\n",
    "        df_box = data[['group', taxon]].copy()\n",
    "        df_box = df_box[df_box['group'].isin(order)]\n",
    "        df_box['group'] = pd.Categorical(df_box['group'], categories=order, ordered=True)\n",
    "        df_box = df_box.dropna(subset=[taxon])\n",
    "\n",
    "        if df_box.empty:\n",
    "            print(f\"⚠️ Skipping {taxon} — no non-missing data.\")\n",
    "            ax_box.set_visible(False)\n",
    "            ax_corr.set_visible(False)\n",
    "            continue\n",
    "\n",
    "        df_box['taxon'] = taxon.replace('g__', '').replace('_ASV-', ' ASV-')\n",
    "\n",
    "        # --- Correlation data ---\n",
    "        df_corr = data[data['group'].isin(['nares-H', 'nares-AD'])][['o_scorad', taxon]].copy()\n",
    "        df_corr[taxon] = pd.to_numeric(df_corr[taxon], errors='coerce')\n",
    "        df_corr['o_scorad'] = pd.to_numeric(df_corr['o_scorad'], errors='coerce')\n",
    "        df_corr = df_corr.dropna()\n",
    "\n",
    "        # --- Left: Boxplot ---\n",
    "        sns.boxplot(\n",
    "            data=df_box,\n",
    "            x='group',\n",
    "            y=taxon,\n",
    "            order=order,\n",
    "            palette=group_palette,\n",
    "            ax=ax_box,\n",
    "            width=0.5,\n",
    "            fliersize=0\n",
    "        )\n",
    "        sns.stripplot(\n",
    "            data=df_box,\n",
    "            x='group',\n",
    "            y=taxon,\n",
    "            order=order,\n",
    "            palette=strip_palette,\n",
    "            ax=ax_box,\n",
    "            jitter=True,\n",
    "            size=4,\n",
    "            alpha=0.7\n",
    "        )\n",
    "        ax_box.set_ylabel(\"RCLR Abundance\", fontsize=14)\n",
    "        ax_box.set_xlabel(\"\")\n",
    "        ax_box.set_xticklabels(['H', 'AD'], fontsize=14)\n",
    "        ax_box.tick_params(axis='y', labelsize=10)\n",
    "\n",
    "        # --- Add Mann-Whitney U p-value ---\n",
    "        if len(order) == 2:\n",
    "            group1_vals = df_box[df_box['group'] == order[0]][taxon]\n",
    "            group2_vals = df_box[df_box['group'] == order[1]][taxon]\n",
    "            if not group1_vals.empty and not group2_vals.empty:\n",
    "                stat, pval = mannwhitneyu(group1_vals, group2_vals, alternative='two-sided')\n",
    "                \n",
    "                # Use minimum value instead of max\n",
    "                y_line = 6.2  # 👈 your manual y-position for the bracket line\n",
    "                y_text = 7  # 👈 your manual y-position for the p-value text\n",
    "\n",
    "                # Horizontal line (bracket)\n",
    "                ax_box.plot([0, 0, 1, 1],\n",
    "                            [y_line, y_line - 0.03, y_line - 0.03, y_line],\n",
    "                            color='black')\n",
    "\n",
    "                # P-value text\n",
    "                stars = '***' if pval < 0.001 else '**' if pval < 0.01 else '*' if pval < 0.05 else ''\n",
    "                ax_box.text(0.5, y_text, f\"{pval:.1e} {stars}\", ha='center', va='top', fontsize=12)\n",
    "\n",
    "\n",
    "        # --- Right: Correlation with severity ---\n",
    "        ax_corr.set_visible(True)\n",
    "        if not df_corr.empty:\n",
    "            # Calculate correlation\n",
    "            r, pval = pearsonr(df_corr['o_scorad'], df_corr[taxon])\n",
    "\n",
    "            # Choose dot color\n",
    "            if pval > 0.05:\n",
    "                dot_color = 'grey'\n",
    "            elif taxon == 'g__Staphylococcus_ASV-1':\n",
    "                dot_color = 'salmon'\n",
    "            else:\n",
    "                dot_color = None\n",
    "\n",
    "            # Plot\n",
    "            sns.regplot(\n",
    "                data=df_corr,\n",
    "                x='o_scorad',\n",
    "                y=taxon,\n",
    "                scatter_kws={'alpha': 0.5, 's': 20, 'color': dot_color} if dot_color else {'alpha': 0.5, 's': 20},\n",
    "                line_kws={'color': 'black'},\n",
    "                ax=ax_corr\n",
    "            )\n",
    "\n",
    "            # Add correlation label\n",
    "            r_label = f\"Pearson r = {r:.2f}\\np = {pval:.1e}\"\n",
    "            ax_corr.text(0.05, 0.90, r_label, transform=ax_corr.transAxes,\n",
    "                        fontsize=12, va='top', ha='left',\n",
    "                        bbox=dict(facecolor='white', alpha=0.7, edgecolor='none'))\n",
    "\n",
    "            ax_corr.set_xlabel(\"SCORAD\", fontsize=12)\n",
    "\n",
    "        # --- Title per taxon ---\n",
    "        if i < len(title_y_positions):\n",
    "            fig.text(0.5, title_y_positions[i], df_box['taxon'].iloc[0], ha='center', fontsize=16)\n",
    "\n",
    "    fig.suptitle(title_name, fontsize=18, y=0.97)\n",
    "    plt.tight_layout(rect=[0, 0, 1, 0.96])\n",
    "    plt.subplots_adjust(hspace=0.45)\n",
    "    sns.despine()\n",
    "\n",
    "    return fig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 236,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/categorical.py:641: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  grouped_vals = vals.groupby(grouper)\n",
      "/var/folders/22/yck9vwx53w1c38tvj_c0_tz00000gn/T/ipykernel_7373/1981148617.py:51: FutureWarning: Passing `palette` without assigning `hue` is deprecated.\n",
      "  sns.stripplot(\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.\n",
      "  with pd.option_context('mode.use_inf_as_na', True):\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.\n",
      "  with pd.option_context('mode.use_inf_as_na', True):\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1075: FutureWarning: When grouping with a length-1 list-like, you will need to pass a length-1 tuple to get_group in a future version of pandas. Pass `(name,)` instead of `name` to silence this warning.\n",
      "  data_subset = grouped_data.get_group(pd_key)\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/categorical.py:641: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  grouped_vals = vals.groupby(grouper)\n",
      "/var/folders/22/yck9vwx53w1c38tvj_c0_tz00000gn/T/ipykernel_7373/1981148617.py:51: FutureWarning: Passing `palette` without assigning `hue` is deprecated.\n",
      "  sns.stripplot(\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.\n",
      "  with pd.option_context('mode.use_inf_as_na', True):\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.\n",
      "  with pd.option_context('mode.use_inf_as_na', True):\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1075: FutureWarning: When grouping with a length-1 list-like, you will need to pass a length-1 tuple to get_group in a future version of pandas. Pass `(name,)` instead of `name` to silence this warning.\n",
      "  data_subset = grouped_data.get_group(pd_key)\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/categorical.py:641: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  grouped_vals = vals.groupby(grouper)\n",
      "/var/folders/22/yck9vwx53w1c38tvj_c0_tz00000gn/T/ipykernel_7373/1981148617.py:51: FutureWarning: Passing `palette` without assigning `hue` is deprecated.\n",
      "  sns.stripplot(\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.\n",
      "  with pd.option_context('mode.use_inf_as_na', True):\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.\n",
      "  with pd.option_context('mode.use_inf_as_na', True):\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1075: FutureWarning: When grouping with a length-1 list-like, you will need to pass a length-1 tuple to get_group in a future version of pandas. Pass `(name,)` instead of `name` to silence this warning.\n",
      "  data_subset = grouped_data.get_group(pd_key)\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/categorical.py:641: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  grouped_vals = vals.groupby(grouper)\n",
      "/var/folders/22/yck9vwx53w1c38tvj_c0_tz00000gn/T/ipykernel_7373/1981148617.py:51: FutureWarning: Passing `palette` without assigning `hue` is deprecated.\n",
      "  sns.stripplot(\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.\n",
      "  with pd.option_context('mode.use_inf_as_na', True):\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.\n",
      "  with pd.option_context('mode.use_inf_as_na', True):\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1075: FutureWarning: When grouping with a length-1 list-like, you will need to pass a length-1 tuple to get_group in a future version of pandas. Pass `(name,)` instead of `name` to silence this warning.\n",
      "  data_subset = grouped_data.get_group(pd_key)\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/categorical.py:641: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  grouped_vals = vals.groupby(grouper)\n",
      "/var/folders/22/yck9vwx53w1c38tvj_c0_tz00000gn/T/ipykernel_7373/1981148617.py:51: FutureWarning: Passing `palette` without assigning `hue` is deprecated.\n",
      "  sns.stripplot(\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.\n",
      "  with pd.option_context('mode.use_inf_as_na', True):\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.\n",
      "  with pd.option_context('mode.use_inf_as_na', True):\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1075: FutureWarning: When grouping with a length-1 list-like, you will need to pass a length-1 tuple to get_group in a future version of pandas. Pass `(name,)` instead of `name` to silence this warning.\n",
      "  data_subset = grouped_data.get_group(pd_key)\n"
     ]
    }
   ],
   "source": [
    "# taxa_list = ['g__Staphylococcus_ASV-1', 'g__Streptococcus_ASV-1', 'g__Streptococcus_ASV-2', 'g__Micrococcus_ASV-1', 'g__Veillonella_A_ASV-1']\n",
    "taxa_list = ['g__Staphylococcus_ASV-1', 'g__Streptococcus_ASV-1', 'g__Streptococcus_ASV-2', 'g__Micrococcus_ASV-1', 'g__Veillonella_A_ASV-1']\n",
    "\n",
    "# taxa_list = [' g__Staphylococcus', ' g__Streptococcus', ' g__Micrococcus', ' g__Veillonella_A', ' g__Haemophilus_D_734546']\n",
    "\n",
    "fig = plot_combined_box_and_severity(\n",
    "    data=nares_samples,\n",
    "    taxa_list=taxa_list,\n",
    "    group_palette=group_palette,\n",
    "    strip_palette=strip_palette,\n",
    "    title_name='\\nNares of Umtata Children'\n",
    ")\n",
    "fig.savefig('../Plots/Analysis_figures/Severity_Correlations/combined_box_scatter_Umtata_nares.png', dpi=600)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "qiime2-metagenome-2024.10",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
