{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<!-- # ASV overlap Venn Diagram -->"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import Python packages\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import biom\n",
    "from biom import load_table\n",
    "import numpy as np\n",
    "from scipy.stats import pearsonr\n",
    "from matplotlib_venn import venn2\n",
    "import matplotlib.patches as mpatches\n",
    "from scipy.stats import mannwhitneyu\n",
    "from biom import Table\n",
    "from gemelli.rpca import rpca\n",
    "from scipy.spatial.distance import euclidean\n",
    "import re\n",
    "import matplotlib.colors as mcolors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PlateNumber</th>\n",
       "      <th>PlateLocation</th>\n",
       "      <th>i5</th>\n",
       "      <th>i5Sequence</th>\n",
       "      <th>i7</th>\n",
       "      <th>i7Sequence</th>\n",
       "      <th>identifier</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Plate ID</th>\n",
       "      <th>Well location</th>\n",
       "      <th>...</th>\n",
       "      <th>enrolment_season</th>\n",
       "      <th>hiv_exposure</th>\n",
       "      <th>hiv_status</th>\n",
       "      <th>household_size</th>\n",
       "      <th>o_scorad</th>\n",
       "      <th>FWD_filepath</th>\n",
       "      <th>REV_filepath</th>\n",
       "      <th>group</th>\n",
       "      <th>microbiome_type</th>\n",
       "      <th>ad_status</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#sample-id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Ca009STL</th>\n",
       "      <td>1</td>\n",
       "      <td>A1</td>\n",
       "      <td>SA501</td>\n",
       "      <td>ATCGTACG</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA501</td>\n",
       "      <td>CGAGAGTT-ATCGTACG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>A1</td>\n",
       "      <td>...</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>4.0</td>\n",
       "      <td>40</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>skin</td>\n",
       "      <td>AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900221</th>\n",
       "      <td>1</td>\n",
       "      <td>B1</td>\n",
       "      <td>SA502</td>\n",
       "      <td>ACTATCTG</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA502</td>\n",
       "      <td>CGAGAGTT-ACTATCTG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>B1</td>\n",
       "      <td>...</td>\n",
       "      <td>Winter</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>7.0</td>\n",
       "      <td>34</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>skin</td>\n",
       "      <td>AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca010EBL</th>\n",
       "      <td>1</td>\n",
       "      <td>C1</td>\n",
       "      <td>SA503</td>\n",
       "      <td>TAGCGAGT</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA503</td>\n",
       "      <td>CGAGAGTT-TAGCGAGT</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>C1</td>\n",
       "      <td>...</td>\n",
       "      <td>Spring</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>7.0</td>\n",
       "      <td>21</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>skin</td>\n",
       "      <td>AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900460</th>\n",
       "      <td>1</td>\n",
       "      <td>D1</td>\n",
       "      <td>SA504</td>\n",
       "      <td>CTGCGTGT</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA504</td>\n",
       "      <td>CGAGAGTT-CTGCGTGT</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>D1</td>\n",
       "      <td>...</td>\n",
       "      <td>Spring</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>40</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>skin</td>\n",
       "      <td>AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900051</th>\n",
       "      <td>1</td>\n",
       "      <td>E1</td>\n",
       "      <td>SA505</td>\n",
       "      <td>TCATCGAG</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA505</td>\n",
       "      <td>CGAGAGTT-TCATCGAG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>E1</td>\n",
       "      <td>...</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>7.0</td>\n",
       "      <td>41</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>skin</td>\n",
       "      <td>AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca006ONL2</th>\n",
       "      <td>6</td>\n",
       "      <td>H1</td>\n",
       "      <td>SA508</td>\n",
       "      <td>GACACCGT</td>\n",
       "      <td>SB701</td>\n",
       "      <td>CTCGACTT</td>\n",
       "      <td>SB701SA508</td>\n",
       "      <td>CTCGACTT-GACACCGT</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>H1</td>\n",
       "      <td>...</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>3.0</td>\n",
       "      <td>34</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>skin</td>\n",
       "      <td>AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca006ONNL</th>\n",
       "      <td>6</td>\n",
       "      <td>F2</td>\n",
       "      <td>SA506</td>\n",
       "      <td>CGTGAGTG</td>\n",
       "      <td>SB702</td>\n",
       "      <td>CGAAGTAT</td>\n",
       "      <td>SB702SA506</td>\n",
       "      <td>CGAAGTAT-CGTGAGTG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>F2</td>\n",
       "      <td>...</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>3.0</td>\n",
       "      <td>34</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADNL</td>\n",
       "      <td>skin</td>\n",
       "      <td>AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca006ONNL2</th>\n",
       "      <td>6</td>\n",
       "      <td>H2</td>\n",
       "      <td>SA508</td>\n",
       "      <td>GACACCGT</td>\n",
       "      <td>SB702</td>\n",
       "      <td>CGAAGTAT</td>\n",
       "      <td>SB702SA508</td>\n",
       "      <td>CGAAGTAT-GACACCGT</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>H2</td>\n",
       "      <td>...</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>3.0</td>\n",
       "      <td>34</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>skin-ADNL</td>\n",
       "      <td>skin</td>\n",
       "      <td>AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca006ONPN</th>\n",
       "      <td>6</td>\n",
       "      <td>F3</td>\n",
       "      <td>SA506</td>\n",
       "      <td>CGTGAGTG</td>\n",
       "      <td>SB703</td>\n",
       "      <td>TAGCAGCT</td>\n",
       "      <td>SB703SA506</td>\n",
       "      <td>TAGCAGCT-CGTGAGTG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>F3</td>\n",
       "      <td>...</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>3.0</td>\n",
       "      <td>34</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>nares-AD</td>\n",
       "      <td>nares</td>\n",
       "      <td>AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca006ONPN2</th>\n",
       "      <td>6</td>\n",
       "      <td>H3</td>\n",
       "      <td>SA508</td>\n",
       "      <td>GACACCGT</td>\n",
       "      <td>SB703</td>\n",
       "      <td>TAGCAGCT</td>\n",
       "      <td>SB703SA508</td>\n",
       "      <td>TAGCAGCT-GACACCGT</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>H3</td>\n",
       "      <td>...</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>3.0</td>\n",
       "      <td>34</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>nares-AD</td>\n",
       "      <td>nares</td>\n",
       "      <td>AD</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>502 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            PlateNumber PlateLocation     i5 i5Sequence     i7 i7Sequence  \\\n",
       "#sample-id                                                                  \n",
       "Ca009STL              1            A1  SA501   ATCGTACG  SA701   CGAGAGTT   \n",
       "900221                1            B1  SA502   ACTATCTG  SA701   CGAGAGTT   \n",
       "Ca010EBL              1            C1  SA503   TAGCGAGT  SA701   CGAGAGTT   \n",
       "900460                1            D1  SA504   CTGCGTGT  SA701   CGAGAGTT   \n",
       "900051                1            E1  SA505   TCATCGAG  SA701   CGAGAGTT   \n",
       "...                 ...           ...    ...        ...    ...        ...   \n",
       "Ca006ONL2             6            H1  SA508   GACACCGT  SB701   CTCGACTT   \n",
       "Ca006ONNL             6            F2  SA506   CGTGAGTG  SB702   CGAAGTAT   \n",
       "Ca006ONNL2            6            H2  SA508   GACACCGT  SB702   CGAAGTAT   \n",
       "Ca006ONPN             6            F3  SA506   CGTGAGTG  SB703   TAGCAGCT   \n",
       "Ca006ONPN2            6            H3  SA508   GACACCGT  SB703   TAGCAGCT   \n",
       "\n",
       "            identifier           Sequence      Plate ID Well location  ...  \\\n",
       "#sample-id                                                             ...   \n",
       "Ca009STL    SA701SA501  CGAGAGTT-ATCGTACG  1.010000e+21            A1  ...   \n",
       "900221      SA701SA502  CGAGAGTT-ACTATCTG  1.010000e+21            B1  ...   \n",
       "Ca010EBL    SA701SA503  CGAGAGTT-TAGCGAGT  1.010000e+21            C1  ...   \n",
       "900460      SA701SA504  CGAGAGTT-CTGCGTGT  1.010000e+21            D1  ...   \n",
       "900051      SA701SA505  CGAGAGTT-TCATCGAG  1.010000e+21            E1  ...   \n",
       "...                ...                ...           ...           ...  ...   \n",
       "Ca006ONL2   SB701SA508  CTCGACTT-GACACCGT  1.010000e+21            H1  ...   \n",
       "Ca006ONNL   SB702SA506  CGAAGTAT-CGTGAGTG  1.010000e+21            F2  ...   \n",
       "Ca006ONNL2  SB702SA508  CGAAGTAT-GACACCGT  1.010000e+21            H2  ...   \n",
       "Ca006ONPN   SB703SA506  TAGCAGCT-CGTGAGTG  1.010000e+21            F3  ...   \n",
       "Ca006ONPN2  SB703SA508  TAGCAGCT-GACACCGT  1.010000e+21            H3  ...   \n",
       "\n",
       "            enrolment_season hiv_exposure hiv_status household_size o_scorad  \\\n",
       "#sample-id                                                                     \n",
       "Ca009STL             Autumn     Unexposed   negative            4.0       40   \n",
       "900221                Winter    Unexposed   negative            7.0       34   \n",
       "Ca010EBL              Spring    Unexposed   negative            7.0       21   \n",
       "900460                Spring    Unexposed        NaN            4.0       40   \n",
       "900051               Autumn     Unexposed   negative            7.0       41   \n",
       "...                      ...          ...        ...            ...      ...   \n",
       "Ca006ONL2             Autumn    Unexposed   negative            3.0       34   \n",
       "Ca006ONNL             Autumn    Unexposed   negative            3.0       34   \n",
       "Ca006ONNL2            Autumn    Unexposed   negative            3.0       34   \n",
       "Ca006ONPN             Autumn    Unexposed   negative            3.0       34   \n",
       "Ca006ONPN2            Autumn    Unexposed   negative            3.0       34   \n",
       "\n",
       "                                                 FWD_filepath  \\\n",
       "#sample-id                                                      \n",
       "Ca009STL    /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "900221      /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "Ca010EBL    /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "900460      /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "900051      /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "...                                                       ...   \n",
       "Ca006ONL2   /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "Ca006ONNL   /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "Ca006ONNL2  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "Ca006ONPN   /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "Ca006ONPN2  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "\n",
       "                                                 REV_filepath      group  \\\n",
       "#sample-id                                                                 \n",
       "Ca009STL    /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   skin-ADL   \n",
       "900221      /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   skin-ADL   \n",
       "Ca010EBL    /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   skin-ADL   \n",
       "900460      /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   skin-ADL   \n",
       "900051      /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   skin-ADL   \n",
       "...                                                       ...        ...   \n",
       "Ca006ONL2   /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   skin-ADL   \n",
       "Ca006ONNL   /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...  skin-ADNL   \n",
       "Ca006ONNL2  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...  skin-ADNL   \n",
       "Ca006ONPN   /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   nares-AD   \n",
       "Ca006ONPN2  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   nares-AD   \n",
       "\n",
       "           microbiome_type ad_status  \n",
       "#sample-id                            \n",
       "Ca009STL              skin        AD  \n",
       "900221                skin        AD  \n",
       "Ca010EBL              skin        AD  \n",
       "900460                skin        AD  \n",
       "900051                skin        AD  \n",
       "...                    ...       ...  \n",
       "Ca006ONL2             skin        AD  \n",
       "Ca006ONNL             skin        AD  \n",
       "Ca006ONNL2            skin        AD  \n",
       "Ca006ONPN            nares        AD  \n",
       "Ca006ONPN2           nares        AD  \n",
       "\n",
       "[502 rows x 34 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load the metadata\n",
    "metadata_path = '../Data/Metadata/updated_clean_ant_skin_metadata_microbiome_type.tab'\n",
    "metadata = pd.read_csv(metadata_path, sep='\\t')\n",
    "\n",
    "metadata['#sample-id'] = metadata['#sample-id'].str.replace('_', '')\n",
    "# Set Sample-ID as the index for the metadata dataframe \n",
    "metadata = metadata.set_index('#sample-id')\n",
    "\n",
    "# Create group column based on case_type to simplify group names\n",
    "metadata['group'] = metadata['case_type'].map({\n",
    "    'case-lesional skin': 'skin-ADL',\n",
    "    'case-nonlesional skin': 'skin-ADNL', \n",
    "    'control-nonlesional skin': 'skin-H',\n",
    "    'case-anterior nares': 'nares-AD',\n",
    "    'control-anterior nares': 'nares-H'\n",
    "})\n",
    "\n",
    "# Add AD status column based on group values\n",
    "metadata['ad_status'] = metadata['group'].apply(lambda x: 'AD' if x.split('-')[-1].startswith('AD') else 'H')\n",
    "\n",
    "metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>g__Streptococcus_ASV-1</th>\n",
       "      <th>g__Streptococcus_ASV-2</th>\n",
       "      <th>g__Corynebacterium_ASV-1</th>\n",
       "      <th>g__Corynebacterium_ASV-4</th>\n",
       "      <th>g___ASV-5</th>\n",
       "      <th>g__Cutibacterium_ASV-1</th>\n",
       "      <th>g___ASV-33</th>\n",
       "      <th>g___ASV-30</th>\n",
       "      <th>g___ASV-22</th>\n",
       "      <th>g__Cutibacterium_ASV-2</th>\n",
       "      <th>...</th>\n",
       "      <th>g__Copromonas_ASV-3</th>\n",
       "      <th>g__Microvirga_ASV-1</th>\n",
       "      <th>g__Leptotrichia_A_993758_ASV-17</th>\n",
       "      <th>g___ASV-103</th>\n",
       "      <th>g__Leptotrichia_A_993758_ASV-13</th>\n",
       "      <th>g__Capnocytophaga_820688_ASV-8</th>\n",
       "      <th>g__UBA6175_ASV-2</th>\n",
       "      <th>g__Comamonas_F_589250_ASV-3</th>\n",
       "      <th>g__Bosea_ASV-2</th>\n",
       "      <th>g___ASV-145</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>900344</th>\n",
       "      <td>984.0</td>\n",
       "      <td>611.0</td>\n",
       "      <td>114.0</td>\n",
       "      <td>82.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900459</th>\n",
       "      <td>118.0</td>\n",
       "      <td>106.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900221</th>\n",
       "      <td>22.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900570</th>\n",
       "      <td>389.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900092</th>\n",
       "      <td>3106.0</td>\n",
       "      <td>1707.0</td>\n",
       "      <td>59.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9003972</th>\n",
       "      <td>1168.0</td>\n",
       "      <td>593.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>736.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>388.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>17.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900097</th>\n",
       "      <td>24.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900498</th>\n",
       "      <td>15.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900276</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>151.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>79.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900406</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>502 rows × 798 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         g__Streptococcus_ASV-1  g__Streptococcus_ASV-2  \\\n",
       "900344                    984.0                   611.0   \n",
       "900459                    118.0                   106.0   \n",
       "900221                     22.0                       0   \n",
       "900570                    389.0                       0   \n",
       "900092                   3106.0                  1707.0   \n",
       "...                         ...                     ...   \n",
       "9003972                  1168.0                   593.0   \n",
       "900097                     24.0                       0   \n",
       "900498                     15.0                    17.0   \n",
       "900276                        0                       0   \n",
       "900406                        0                       0   \n",
       "\n",
       "         g__Corynebacterium_ASV-1  g__Corynebacterium_ASV-4  g___ASV-5  \\\n",
       "900344                      114.0                      82.0       22.0   \n",
       "900459                          0                         0          0   \n",
       "900221                          0                         0          0   \n",
       "900570                          0                         0        8.0   \n",
       "900092                       59.0                      32.0        3.0   \n",
       "...                           ...                       ...        ...   \n",
       "9003972                      16.0                         0       28.0   \n",
       "900097                          0                         0          0   \n",
       "900498                          0                         0          0   \n",
       "900276                       30.0                         0          0   \n",
       "900406                          0                         0          0   \n",
       "\n",
       "         g__Cutibacterium_ASV-1  g___ASV-33  g___ASV-30  g___ASV-22  \\\n",
       "900344                      8.0         8.0         6.0         3.0   \n",
       "900459                        0           0           0           0   \n",
       "900221                     16.0           0           0           0   \n",
       "900570                     11.0           0           0           0   \n",
       "900092                        0           0           0         7.0   \n",
       "...                         ...         ...         ...         ...   \n",
       "9003972                   736.0           0           0        36.0   \n",
       "900097                     33.0           0           0           0   \n",
       "900498                     34.0           0        14.0           0   \n",
       "900276                    151.0           0           0           0   \n",
       "900406                        0           0           0           0   \n",
       "\n",
       "         g__Cutibacterium_ASV-2  ...  g__Copromonas_ASV-3  \\\n",
       "900344                      2.0  ...                    0   \n",
       "900459                      3.0  ...                    0   \n",
       "900221                        0  ...                    0   \n",
       "900570                      5.0  ...                    0   \n",
       "900092                        0  ...                    0   \n",
       "...                         ...  ...                  ...   \n",
       "9003972                   388.0  ...                    0   \n",
       "900097                     12.0  ...                    0   \n",
       "900498                     25.0  ...                    0   \n",
       "900276                     79.0  ...                    0   \n",
       "900406                        0  ...                    0   \n",
       "\n",
       "         g__Microvirga_ASV-1  g__Leptotrichia_A_993758_ASV-17  g___ASV-103  \\\n",
       "900344                     0                                0            0   \n",
       "900459                     0                                0            0   \n",
       "900221                     0                                0            0   \n",
       "900570                     0                                0            0   \n",
       "900092                     0                                0            0   \n",
       "...                      ...                              ...          ...   \n",
       "9003972                    0                                0            0   \n",
       "900097                     0                                0            0   \n",
       "900498                     0                                0            0   \n",
       "900276                     0                                0            0   \n",
       "900406                     0                                0            0   \n",
       "\n",
       "         g__Leptotrichia_A_993758_ASV-13  g__Capnocytophaga_820688_ASV-8  \\\n",
       "900344                                 0                               0   \n",
       "900459                                 0                               0   \n",
       "900221                                 0                               0   \n",
       "900570                                 0                               0   \n",
       "900092                                 0                               0   \n",
       "...                                  ...                             ...   \n",
       "9003972                                0                               0   \n",
       "900097                                 0                               0   \n",
       "900498                                 0                               0   \n",
       "900276                                 0                               0   \n",
       "900406                                 0                               0   \n",
       "\n",
       "         g__UBA6175_ASV-2  g__Comamonas_F_589250_ASV-3  g__Bosea_ASV-2  \\\n",
       "900344                  0                            0               0   \n",
       "900459                  0                            0               0   \n",
       "900221                  0                            0               0   \n",
       "900570                  0                            0               0   \n",
       "900092                  0                            0               0   \n",
       "...                   ...                          ...             ...   \n",
       "9003972                 0                            0               0   \n",
       "900097                  0                            0               0   \n",
       "900498                  0                            0               0   \n",
       "900276                  0                            0               0   \n",
       "900406                  0                            0               0   \n",
       "\n",
       "         g___ASV-145  \n",
       "900344             0  \n",
       "900459             0  \n",
       "900221             0  \n",
       "900570             0  \n",
       "900092             0  \n",
       "...              ...  \n",
       "9003972         17.0  \n",
       "900097             0  \n",
       "900498             0  \n",
       "900276             0  \n",
       "900406             0  \n",
       "\n",
       "[502 rows x 798 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read in table at collapsed genera  level\n",
    "\n",
    "taxa_level = 'Genus'\n",
    "biom_path = f'../Data/Tables/Absolute_Abundance_Tables/209766_filtered_by_prevalence_1pct_rare_{taxa_level}-ASV-non-collapse.biom'\n",
    "\n",
    "biom_tbl = load_table(biom_path)\n",
    "df = pd.DataFrame(biom_tbl.to_dataframe().T)\n",
    "\n",
    "# delete the prefix from the \n",
    "df.index = df.index.str.replace('15564.', '')\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>g__Streptococcus_ASV-1</th>\n",
       "      <th>g__Streptococcus_ASV-2</th>\n",
       "      <th>g__Corynebacterium_ASV-1</th>\n",
       "      <th>g__Corynebacterium_ASV-4</th>\n",
       "      <th>g___ASV-5</th>\n",
       "      <th>g__Cutibacterium_ASV-1</th>\n",
       "      <th>g___ASV-33</th>\n",
       "      <th>g___ASV-30</th>\n",
       "      <th>g___ASV-22</th>\n",
       "      <th>g__Cutibacterium_ASV-2</th>\n",
       "      <th>...</th>\n",
       "      <th>g___ASV-103</th>\n",
       "      <th>g__Leptotrichia_A_993758_ASV-13</th>\n",
       "      <th>g__Capnocytophaga_820688_ASV-8</th>\n",
       "      <th>g__UBA6175_ASV-2</th>\n",
       "      <th>g__Comamonas_F_589250_ASV-3</th>\n",
       "      <th>g__Bosea_ASV-2</th>\n",
       "      <th>g___ASV-145</th>\n",
       "      <th>pid</th>\n",
       "      <th>group</th>\n",
       "      <th>microbiome_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>900344</th>\n",
       "      <td>984.0</td>\n",
       "      <td>611.0</td>\n",
       "      <td>114.0</td>\n",
       "      <td>82.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Co-130-MM</td>\n",
       "      <td>skin-H</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900459</th>\n",
       "      <td>118.0</td>\n",
       "      <td>106.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Ca-145-LM</td>\n",
       "      <td>nares-AD</td>\n",
       "      <td>nares</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900221</th>\n",
       "      <td>22.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Ca-101-ID</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900570</th>\n",
       "      <td>389.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Ca-046-MT</td>\n",
       "      <td>skin-ADNL</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900092</th>\n",
       "      <td>3106.0</td>\n",
       "      <td>1707.0</td>\n",
       "      <td>59.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Ca-023-EJ</td>\n",
       "      <td>nares-AD</td>\n",
       "      <td>nares</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9003972</th>\n",
       "      <td>1168.0</td>\n",
       "      <td>593.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>736.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>388.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>Ca-125-IM</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900097</th>\n",
       "      <td>24.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Ca-025-AC</td>\n",
       "      <td>skin-ADNL</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900498</th>\n",
       "      <td>15.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Ca-157-LM</td>\n",
       "      <td>skin-ADNL</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900276</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>151.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>79.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Ca-111-IN</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900406</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Ca-127-OM</td>\n",
       "      <td>skin-ADNL</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>502 rows × 801 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         g__Streptococcus_ASV-1  g__Streptococcus_ASV-2  \\\n",
       "900344                    984.0                   611.0   \n",
       "900459                    118.0                   106.0   \n",
       "900221                     22.0                       0   \n",
       "900570                    389.0                       0   \n",
       "900092                   3106.0                  1707.0   \n",
       "...                         ...                     ...   \n",
       "9003972                  1168.0                   593.0   \n",
       "900097                     24.0                       0   \n",
       "900498                     15.0                    17.0   \n",
       "900276                        0                       0   \n",
       "900406                        0                       0   \n",
       "\n",
       "         g__Corynebacterium_ASV-1  g__Corynebacterium_ASV-4  g___ASV-5  \\\n",
       "900344                      114.0                      82.0       22.0   \n",
       "900459                          0                         0          0   \n",
       "900221                          0                         0          0   \n",
       "900570                          0                         0        8.0   \n",
       "900092                       59.0                      32.0        3.0   \n",
       "...                           ...                       ...        ...   \n",
       "9003972                      16.0                         0       28.0   \n",
       "900097                          0                         0          0   \n",
       "900498                          0                         0          0   \n",
       "900276                       30.0                         0          0   \n",
       "900406                          0                         0          0   \n",
       "\n",
       "         g__Cutibacterium_ASV-1  g___ASV-33  g___ASV-30  g___ASV-22  \\\n",
       "900344                      8.0         8.0         6.0         3.0   \n",
       "900459                        0           0           0           0   \n",
       "900221                     16.0           0           0           0   \n",
       "900570                     11.0           0           0           0   \n",
       "900092                        0           0           0         7.0   \n",
       "...                         ...         ...         ...         ...   \n",
       "9003972                   736.0           0           0        36.0   \n",
       "900097                     33.0           0           0           0   \n",
       "900498                     34.0           0        14.0           0   \n",
       "900276                    151.0           0           0           0   \n",
       "900406                        0           0           0           0   \n",
       "\n",
       "         g__Cutibacterium_ASV-2  ...  g___ASV-103  \\\n",
       "900344                      2.0  ...            0   \n",
       "900459                      3.0  ...            0   \n",
       "900221                        0  ...            0   \n",
       "900570                      5.0  ...            0   \n",
       "900092                        0  ...            0   \n",
       "...                         ...  ...          ...   \n",
       "9003972                   388.0  ...            0   \n",
       "900097                     12.0  ...            0   \n",
       "900498                     25.0  ...            0   \n",
       "900276                     79.0  ...            0   \n",
       "900406                        0  ...            0   \n",
       "\n",
       "         g__Leptotrichia_A_993758_ASV-13  g__Capnocytophaga_820688_ASV-8  \\\n",
       "900344                                 0                               0   \n",
       "900459                                 0                               0   \n",
       "900221                                 0                               0   \n",
       "900570                                 0                               0   \n",
       "900092                                 0                               0   \n",
       "...                                  ...                             ...   \n",
       "9003972                                0                               0   \n",
       "900097                                 0                               0   \n",
       "900498                                 0                               0   \n",
       "900276                                 0                               0   \n",
       "900406                                 0                               0   \n",
       "\n",
       "         g__UBA6175_ASV-2  g__Comamonas_F_589250_ASV-3  g__Bosea_ASV-2  \\\n",
       "900344                  0                            0               0   \n",
       "900459                  0                            0               0   \n",
       "900221                  0                            0               0   \n",
       "900570                  0                            0               0   \n",
       "900092                  0                            0               0   \n",
       "...                   ...                          ...             ...   \n",
       "9003972                 0                            0               0   \n",
       "900097                  0                            0               0   \n",
       "900498                  0                            0               0   \n",
       "900276                  0                            0               0   \n",
       "900406                  0                            0               0   \n",
       "\n",
       "         g___ASV-145        pid      group  microbiome_type  \n",
       "900344             0  Co-130-MM     skin-H             skin  \n",
       "900459             0  Ca-145-LM   nares-AD            nares  \n",
       "900221             0  Ca-101-ID   skin-ADL             skin  \n",
       "900570             0  Ca-046-MT  skin-ADNL             skin  \n",
       "900092             0  Ca-023-EJ   nares-AD            nares  \n",
       "...              ...        ...        ...              ...  \n",
       "9003972         17.0  Ca-125-IM   skin-ADL             skin  \n",
       "900097             0  Ca-025-AC  skin-ADNL             skin  \n",
       "900498             0  Ca-157-LM  skin-ADNL             skin  \n",
       "900276             0  Ca-111-IN   skin-ADL             skin  \n",
       "900406             0  Ca-127-OM  skin-ADNL             skin  \n",
       "\n",
       "[502 rows x 801 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Map pid, group, and microbiome_type from metadata to df based on matching indexes\n",
    "df['pid'] = metadata.loc[df.index, 'pid']\n",
    "df['group'] = metadata.loc[df.index, 'group']\n",
    "df['microbiome_type'] = metadata.loc[df.index, 'microbiome_type']\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of unique individuals: 197\n"
     ]
    }
   ],
   "source": [
    "print(f'Number of unique individuals: ' + str(df['pid'].nunique()))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<!-- ## Combined Venn Diagram with all samples, all individuals -->"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of taxa unique to skin: 98\n",
      "Number of taxa unique to nares: 1\n",
      "Number of taxa shared between skin and nares: 699\n",
      "Number of unique individuals: 187\n"
     ]
    }
   ],
   "source": [
    "# Get taxa columns (all except pid, group, microbiome_type)\n",
    "taxa_cols = df.columns[:-3]\n",
    "\n",
    "\n",
    "# # Filter taxa_cols_filtered to keep only prevalent taxa + metadata\n",
    "# df_filtered = taxa_cols_filtered[features_10_prev + ['pid', 'microbiome_type']]\n",
    "\n",
    "# Split into skin and nares samples\n",
    "skin_samples = df[df['microbiome_type'] == 'skin']\n",
    "nares_samples = df[df['microbiome_type'] == 'nares']\n",
    "\n",
    "# Find individuals with both skin and nares samples\n",
    "shared_pids = set(skin_samples['pid']) & set(nares_samples['pid'])\n",
    "\n",
    "# Initialize sets for taxa present in each body site\n",
    "skin_taxa = set()\n",
    "nares_taxa = set()\n",
    "overlapping_taxa = set()\n",
    "\n",
    "# For each individual with both sample types\n",
    "for pid in shared_pids:\n",
    "    # Get samples for this individual\n",
    "    skin_sample = skin_samples[skin_samples['pid'] == pid]\n",
    "    nares_sample = nares_samples[nares_samples['pid'] == pid]\n",
    "    \n",
    "    # Find taxa present (abundance > 0) in each sample type\n",
    "    skin_present = set(taxa_cols[skin_sample[taxa_cols].gt(0).any()])\n",
    "    nares_present = set(taxa_cols[nares_sample[taxa_cols].gt(0).any()])\n",
    "    \n",
    "    # Add to overall sets\n",
    "    skin_taxa.update(skin_present)\n",
    "    nares_taxa.update(nares_present)\n",
    "    # Add taxa present in both sites for this individual\n",
    "    overlapping_taxa.update(skin_present & nares_present)\n",
    "\n",
    "# Plot Venn diagram\n",
    "plt.figure(figsize=(6, 6))\n",
    "v = venn2(\n",
    "    [skin_taxa, nares_taxa],\n",
    "    set_labels=('', ''),\n",
    "    set_colors=('#2a00ff', '#ffa501'),\n",
    "    alpha=0.5\n",
    ")\n",
    "\n",
    "# Adjust font size for subset labels\n",
    "for label in v.subset_labels:\n",
    "    if label:\n",
    "        label.set_fontsize(16)\n",
    "\n",
    "# Customize circle outlines to match fill color\n",
    "region_colors = {\n",
    "    '10': '#2a00ff',    # Skin only\n",
    "    '01': '#ffa501',     # Nares only\n",
    "    '11': '#955280'   # Overlap\n",
    "}\n",
    "for region_id, color in region_colors.items():\n",
    "    patch = v.get_patch_by_id(region_id)\n",
    "    if patch:\n",
    "        patch.set_linewidth(2)\n",
    "        patch.set_edgecolor(color)\n",
    "\n",
    "# Create matching legend\n",
    "# Create matching legend patches (colors match Venn regions)\n",
    "handles = [\n",
    "    mpatches.Patch(color='#2a00ff', alpha=0.5, label='Skin only'),\n",
    "    mpatches.Patch(color='#955280', alpha=0.5, label='Both'), \n",
    "    mpatches.Patch(color='#ffa501', alpha=0.5, label='Nares only')\n",
    "    \n",
    "]\n",
    "\n",
    "# Add legend with specified font size\n",
    "plt.legend(\n",
    "    handles=handles,\n",
    "    loc='lower center',\n",
    "    bbox_to_anchor=(0.5, -0.05),\n",
    "    ncol=3,\n",
    "    frameon=False,\n",
    "    fontsize=16  # Adjust font size here\n",
    ")\n",
    "\n",
    "plt.suptitle('Unique ASV Overlap between Skin and Nares', fontsize=18)\n",
    "plt.title('(Paired Per-Individual and (>1% Sample Prevalence))', fontsize=16)\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.savefig('../Plots/Analysis_figures/Co-occurrence/skin-nares_Venn_individual.png', dpi=600)\n",
    "\n",
    "# Print stats\n",
    "print(f\"Number of taxa unique to skin: {len(skin_taxa - nares_taxa)}\")\n",
    "print(f\"Number of taxa unique to nares: {len(nares_taxa - skin_taxa)}\")\n",
    "print(f\"Number of taxa shared between skin and nares: {len(skin_taxa & nares_taxa)}\")\n",
    "print(f\"Number of unique individuals: {len(shared_pids)}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<!-- ## Separate Venn Diagram by AD status -->"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>g__Streptococcus_ASV-1</th>\n",
       "      <th>g__Streptococcus_ASV-2</th>\n",
       "      <th>g__Corynebacterium_ASV-1</th>\n",
       "      <th>g__Corynebacterium_ASV-4</th>\n",
       "      <th>g___ASV-5</th>\n",
       "      <th>g__Cutibacterium_ASV-1</th>\n",
       "      <th>g___ASV-33</th>\n",
       "      <th>g___ASV-30</th>\n",
       "      <th>g___ASV-22</th>\n",
       "      <th>g__Cutibacterium_ASV-2</th>\n",
       "      <th>...</th>\n",
       "      <th>g___ASV-103</th>\n",
       "      <th>g__Leptotrichia_A_993758_ASV-13</th>\n",
       "      <th>g__Capnocytophaga_820688_ASV-8</th>\n",
       "      <th>g__UBA6175_ASV-2</th>\n",
       "      <th>g__Comamonas_F_589250_ASV-3</th>\n",
       "      <th>g__Bosea_ASV-2</th>\n",
       "      <th>g___ASV-145</th>\n",
       "      <th>pid</th>\n",
       "      <th>group</th>\n",
       "      <th>microbiome_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>900221</th>\n",
       "      <td>22.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Ca-101-ID</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900129</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Ca-035-AR</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900423</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Ca-133-LT</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900581</th>\n",
       "      <td>198.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>46.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Ca-049-LM</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900460</th>\n",
       "      <td>722.0</td>\n",
       "      <td>493.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>200.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Ca-146-SM</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900081</th>\n",
       "      <td>1006.0</td>\n",
       "      <td>0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>0</td>\n",
       "      <td>137.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Ca-020-AZ</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900093</th>\n",
       "      <td>18.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Ca-024-ZM</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900057</th>\n",
       "      <td>111.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Ca-013-NN</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9003972</th>\n",
       "      <td>1168.0</td>\n",
       "      <td>593.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>736.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>388.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>Ca-125-IM</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900276</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>151.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>79.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Ca-111-IN</td>\n",
       "      <td>skin-ADL</td>\n",
       "      <td>skin</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>107 rows × 801 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         g__Streptococcus_ASV-1  g__Streptococcus_ASV-2  \\\n",
       "900221                     22.0                       0   \n",
       "900129                        0                       0   \n",
       "900423                        0                       0   \n",
       "900581                    198.0                       0   \n",
       "900460                    722.0                   493.0   \n",
       "...                         ...                     ...   \n",
       "900081                   1006.0                       0   \n",
       "900093                     18.0                       0   \n",
       "900057                    111.0                       0   \n",
       "9003972                  1168.0                   593.0   \n",
       "900276                        0                       0   \n",
       "\n",
       "         g__Corynebacterium_ASV-1  g__Corynebacterium_ASV-4  g___ASV-5  \\\n",
       "900221                          0                         0          0   \n",
       "900129                          0                         0          0   \n",
       "900423                          0                         0          0   \n",
       "900581                          0                         0       32.0   \n",
       "900460                          0                         0      200.0   \n",
       "...                           ...                       ...        ...   \n",
       "900081                       37.0                         0      137.0   \n",
       "900093                          0                         0          0   \n",
       "900057                          0                         0          0   \n",
       "9003972                      16.0                         0       28.0   \n",
       "900276                       30.0                         0          0   \n",
       "\n",
       "         g__Cutibacterium_ASV-1  g___ASV-33  g___ASV-30  g___ASV-22  \\\n",
       "900221                     16.0           0           0           0   \n",
       "900129                     25.0         2.0           0         2.0   \n",
       "900423                     40.0        19.0           0           0   \n",
       "900581                     46.0           0           0           0   \n",
       "900460                     40.0           0           0        50.0   \n",
       "...                         ...         ...         ...         ...   \n",
       "900081                     40.0           0           0           0   \n",
       "900093                        0           0           0           0   \n",
       "900057                      3.0        17.0           0           0   \n",
       "9003972                   736.0           0           0        36.0   \n",
       "900276                    151.0           0           0           0   \n",
       "\n",
       "         g__Cutibacterium_ASV-2  ...  g___ASV-103  \\\n",
       "900221                        0  ...            0   \n",
       "900129                      3.0  ...            0   \n",
       "900423                     22.0  ...            0   \n",
       "900581                     23.0  ...            0   \n",
       "900460                     28.0  ...            0   \n",
       "...                         ...  ...          ...   \n",
       "900081                        0  ...            0   \n",
       "900093                        0  ...            0   \n",
       "900057                      4.0  ...            0   \n",
       "9003972                   388.0  ...            0   \n",
       "900276                     79.0  ...            0   \n",
       "\n",
       "         g__Leptotrichia_A_993758_ASV-13  g__Capnocytophaga_820688_ASV-8  \\\n",
       "900221                                 0                               0   \n",
       "900129                                 0                               0   \n",
       "900423                                 0                               0   \n",
       "900581                                 0                               0   \n",
       "900460                                 0                               0   \n",
       "...                                  ...                             ...   \n",
       "900081                                 0                               0   \n",
       "900093                                 0                               0   \n",
       "900057                                 0                               0   \n",
       "9003972                                0                               0   \n",
       "900276                                 0                               0   \n",
       "\n",
       "         g__UBA6175_ASV-2  g__Comamonas_F_589250_ASV-3  g__Bosea_ASV-2  \\\n",
       "900221                  0                            0               0   \n",
       "900129                  0                            0               0   \n",
       "900423                  0                            0               0   \n",
       "900581                  0                            0               0   \n",
       "900460                  0                            0               0   \n",
       "...                   ...                          ...             ...   \n",
       "900081                  0                            0               0   \n",
       "900093                2.0                            0               0   \n",
       "900057                  0                            0               0   \n",
       "9003972                 0                            0               0   \n",
       "900276                  0                            0               0   \n",
       "\n",
       "         g___ASV-145        pid     group  microbiome_type  \n",
       "900221             0  Ca-101-ID  skin-ADL             skin  \n",
       "900129             0  Ca-035-AR  skin-ADL             skin  \n",
       "900423             0  Ca-133-LT  skin-ADL             skin  \n",
       "900581             0  Ca-049-LM  skin-ADL             skin  \n",
       "900460             0  Ca-146-SM  skin-ADL             skin  \n",
       "...              ...        ...       ...              ...  \n",
       "900081             0  Ca-020-AZ  skin-ADL             skin  \n",
       "900093             0  Ca-024-ZM  skin-ADL             skin  \n",
       "900057             0  Ca-013-NN  skin-ADL             skin  \n",
       "9003972         17.0  Ca-125-IM  skin-ADL             skin  \n",
       "900276             0  Ca-111-IN  skin-ADL             skin  \n",
       "\n",
       "[107 rows x 801 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# --- STEP 1: Prepare ---\n",
    "# Only keep numeric columns (i.e., ASV features)\n",
    "taxa_cols = df.select_dtypes(include=[np.number]).columns.tolist()\n",
    "\n",
    "\n",
    "# Split into skin and nares samples\n",
    "skin_samples = df[df['microbiome_type'] == 'skin']\n",
    "nares_samples = df[df['microbiome_type'] == 'nares']\n",
    "\n",
    "# --- STEP 2: Separate by disease group ---\n",
    "\n",
    "# AD individuals\n",
    "skin_ADL_samples = skin_samples[skin_samples['group'] == 'skin-ADL']\n",
    "nares_AD_samples = nares_samples[nares_samples['group'] == 'nares-AD']\n",
    "\n",
    "# Healthy individuals\n",
    "skin_H_samples = skin_samples[skin_samples['group'] == 'skin-H']\n",
    "nares_H_samples = nares_samples[nares_samples['group'] == 'nares-H']\n",
    "\n",
    "# --- STEP 3: Find taxa for each group ---\n",
    "\n",
    "# For AD\n",
    "skin_present_AD = skin_ADL_samples[taxa_cols].gt(0).any()\n",
    "nares_present_AD = nares_AD_samples[taxa_cols].gt(0).any()\n",
    "\n",
    "skin_taxa_AD = set(skin_present_AD[skin_present_AD].index)\n",
    "nares_taxa_AD = set(nares_present_AD[nares_present_AD].index)\n",
    "\n",
    "# For Healthy\n",
    "skin_present_H = skin_H_samples[taxa_cols].gt(0).any()\n",
    "nares_present_H = nares_H_samples[taxa_cols].gt(0).any()\n",
    "\n",
    "skin_taxa_H = set(skin_present_H[skin_present_H].index)\n",
    "nares_taxa_H = set(nares_present_H[nares_present_H].index)\n",
    "\n",
    "skin_ADL_samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "106"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get unique patient IDs for each group\n",
    "skin_ADL_pids = set(skin_ADL_samples['pid'].unique())\n",
    "nares_AD_pids = set(nares_AD_samples['pid'].unique())\n",
    "\n",
    "# Count unique patients\n",
    "n_AD = len(skin_ADL_pids.union(nares_AD_pids))\n",
    "n_AD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "87"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get unique patient IDs for each group\n",
    "skin_H_pids = set(skin_H_samples['pid'].unique())\n",
    "nares_H_pids = set(nares_H_samples['pid'].unique())\n",
    "\n",
    "# Count unique patients\n",
    "n_H = len(skin_H_pids.union(nares_H_pids))\n",
    "n_H"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AD Individuals:\n",
      " - Skin only: 168\n",
      " - Nares only: 19\n",
      " - Shared: 602\n",
      "\n",
      "Healthy Individuals:\n",
      " - Skin only: 250\n",
      " - Nares only: 43\n",
      " - Shared: 479\n"
     ]
    }
   ],
   "source": [
    "# --- STEP 4: Plot Venn diagrams ---\n",
    "\n",
    "fig, axes = plt.subplots(2, 1, figsize=(6, 12))\n",
    "\n",
    "# Colors\n",
    "skin_color = '#2a00ff'\n",
    "nares_color = '#ffa501'\n",
    "overlap_color = '#955280'\n",
    "\n",
    "# --- Venn for AD Individuals ---\n",
    "plt.sca(axes[0])  # set current axis\n",
    "v_AD = venn2(\n",
    "    [skin_taxa_AD, nares_taxa_AD],\n",
    "    set_labels=('', ''),\n",
    "    set_colors=(skin_color, nares_color),\n",
    "    alpha=0.5\n",
    ")\n",
    "\n",
    "# Customize font sizes\n",
    "for label in v_AD.subset_labels:\n",
    "    if label:\n",
    "        label.set_fontsize(16)\n",
    "\n",
    "# Customize circle outlines\n",
    "region_colors = {\n",
    "    '10': skin_color,\n",
    "    '01': nares_color,\n",
    "    '11': overlap_color\n",
    "}\n",
    "for region_id, color in region_colors.items():\n",
    "    patch = v_AD.get_patch_by_id(region_id)\n",
    "    if patch:\n",
    "        patch.set_linewidth(2)\n",
    "        patch.set_edgecolor(color)\n",
    "\n",
    "# Title for AD\n",
    "\n",
    "# axes[0].set_title(f'AD Individuals (n={n_AD})', fontsize=18)\n",
    "axes[0].set_title(f'AD Individuals (n=107)', fontsize=20, y= 0.94)\n",
    "\n",
    "# --- Venn for Healthy Individuals ---\n",
    "plt.sca(axes[1])  # set current axis\n",
    "v_H = venn2(\n",
    "    [skin_taxa_H, nares_taxa_H],\n",
    "    set_labels=('', ''),\n",
    "    set_colors=(skin_color, nares_color),\n",
    "    alpha=0.5\n",
    ")\n",
    "\n",
    "for label in v_H.subset_labels:\n",
    "    if label:\n",
    "        label.set_fontsize(16)\n",
    "\n",
    "for region_id, color in region_colors.items():\n",
    "    patch = v_H.get_patch_by_id(region_id)\n",
    "    if patch:\n",
    "        patch.set_linewidth(2)\n",
    "        patch.set_edgecolor(color)\n",
    "\n",
    "# Title for Healthy\n",
    "n_H = skin_H_samples['pid'].nunique()     # Number of unique Healthy individuals\n",
    "axes[1].set_title(f'Healthy Individuals (n=88)', fontsize=20, y= 0.94)\n",
    "\n",
    "# --- Add Legend ---\n",
    "handles = [\n",
    "    mpatches.Patch(color=skin_color, alpha=0.5, label='Skin only'),\n",
    "    mpatches.Patch(color=overlap_color, alpha=0.5, label='Both'),\n",
    "    mpatches.Patch(color=nares_color, alpha=0.5, label='Nares only')\n",
    "]\n",
    "\n",
    "fig.legend(\n",
    "    handles=handles,\n",
    "    loc='lower center',\n",
    "    bbox_to_anchor=(0.5, 0.04),\n",
    "    ncol=3,\n",
    "    frameon=False,\n",
    "    fontsize=16\n",
    ")\n",
    "\n",
    "\n",
    "# --- Final layout tweaks ---\n",
    "plt.tight_layout()\n",
    "plt.subplots_adjust(top=0.9, bottom=0.1)  # Shift top lower and bottom higher\n",
    "\n",
    "# --- Supertitle and subtitle ---\n",
    "fig.suptitle('ASV Overlap Skin and Nares', fontsize=26, y=0.97)\n",
    "# fig.text(\n",
    "#     0.5, 0.91,  # x=centered, y=slightly lower\n",
    "#     'Paired Per-Individual and (>1% Sample Prevalence)',\n",
    "#     ha='center',\n",
    "#     fontsize=14\n",
    "# )\n",
    "\n",
    "plt.savefig('../Plots/Analysis_figures/Co-occurrence/skin-nares_Venn_AD_vs_H.png', dpi=600)\n",
    "\n",
    "# --- Print numbers ---\n",
    "print(\"AD Individuals:\")\n",
    "print(f\" - Skin only: {len(skin_taxa_AD - nares_taxa_AD)}\")\n",
    "print(f\" - Nares only: {len(nares_taxa_AD - skin_taxa_AD)}\")\n",
    "print(f\" - Shared: {len(skin_taxa_AD & nares_taxa_AD)}\\n\")\n",
    "\n",
    "print(\"Healthy Individuals:\")\n",
    "print(f\" - Skin only: {len(skin_taxa_H - nares_taxa_H)}\")\n",
    "print(f\" - Nares only: {len(nares_taxa_H - skin_taxa_H)}\")\n",
    "print(f\" - Shared: {len(skin_taxa_H & nares_taxa_H)}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "# AD\n",
    "ad_skin_only_asvs = skin_taxa_AD - nares_taxa_AD\n",
    "ad_nose_only_asvs = nares_taxa_AD - skin_taxa_AD\n",
    "ad_shared_asvs = skin_taxa_AD & nares_taxa_AD\n",
    "\n",
    "# Healthy\n",
    "h_skin_only_asvs = skin_taxa_H - nares_taxa_H\n",
    "h_nose_only_asvs = nares_taxa_H - skin_taxa_H\n",
    "h_shared_asvs = skin_taxa_H & nares_taxa_H\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define group-to-color mappings\n",
    "group_colors = {\n",
    "    \"AD_skin_only\": \"#947fff\",\n",
    "    \"AD_nose_only\": \"#ffd17f\",\n",
    "    \"AD_shared\": \"#e7b9d8\",\n",
    "}\n",
    "\n",
    "\n",
    "# Create a list of (ASV, color, label) tuples\n",
    "itol_entries = []\n",
    "\n",
    "for asv in ad_skin_only_asvs:\n",
    "    itol_entries.append((asv, group_colors[\"AD_skin_only\"], \"AD_skin_only\"))\n",
    "for asv in ad_nose_only_asvs:\n",
    "    itol_entries.append((asv, group_colors[\"AD_nose_only\"], \"AD_nose_only\"))\n",
    "for asv in ad_shared_asvs:\n",
    "    itol_entries.append((asv, group_colors[\"AD_shared\"], \"AD_shared\"))\n",
    "\n",
    "\n",
    "# Write to iTOL file\n",
    "with open(\"../Data/Trees/itol_AD_asv_group_colors.txt\", \"w\") as f:\n",
    "    f.write(\"DATASET_COLORSTRIP\\n\")\n",
    "    f.write(\"SEPARATOR TAB\\n\")\n",
    "    f.write(\"DATASET_LABEL\\tASV Groups\\n\")\n",
    "    f.write(\"COLOR\\t#ff0000\\n\\n\")\n",
    "\n",
    "    f.write(\"LEGEND_TITLE\\tASV Group\\n\")\n",
    "    f.write(\"LEGEND_SHAPES\\t1\\t1\\t1\\t1\\t1\\t1\\n\")\n",
    "    f.write(\"LEGEND_COLORS\\t\" + \"\\t\".join(group_colors.values()) + \"\\n\")\n",
    "    f.write(\"LEGEND_LABELS\\t\" + \"\\t\".join(group_colors.keys()) + \"\\n\\n\")\n",
    "\n",
    "    f.write(\"DATA\\n\")\n",
    "    for entry in itol_entries:\n",
    "        f.write(f\"{entry[0]}\\t{entry[1]}\\t{entry[2]}\\n\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define group-to-color mappings\n",
    "group_colors = {\n",
    "    \"H_skin_only\": \"#947fff\",\n",
    "    \"H_nose_only\": \"#ffd17f\",\n",
    "    \"H_shared\": \"#e7b9d8\",\n",
    "}\n",
    "\n",
    "\n",
    "# Create a list of (ASV, color, label) tuples\n",
    "itol_entries = []\n",
    "\n",
    "for asv in h_skin_only_asvs:\n",
    "    itol_entries.append((asv, group_colors[\"H_skin_only\"], \"H_skin_only\"))\n",
    "for asv in h_nose_only_asvs:\n",
    "    itol_entries.append((asv, group_colors[\"H_nose_only\"], \"H_nose_only\"))\n",
    "for asv in h_shared_asvs:\n",
    "    itol_entries.append((asv, group_colors[\"H_shared\"], \"H_shared\"))\n",
    "\n",
    "# Write to iTOL file\n",
    "with open(\"../Data/Trees/itol_H_asv_group_colors.txt\", \"w\") as f:\n",
    "    f.write(\"DATASET_COLORSTRIP\\n\")\n",
    "    f.write(\"SEPARATOR TAB\\n\")\n",
    "    f.write(\"DATASET_LABEL\\tASV Groups\\n\")\n",
    "    f.write(\"COLOR\\t#ff0000\\n\\n\")\n",
    "\n",
    "    f.write(\"LEGEND_TITLE\\tASV Group\\n\")\n",
    "    f.write(\"LEGEND_SHAPES\\t1\\t1\\t1\\t1\\t1\\t1\\n\")\n",
    "    f.write(\"LEGEND_COLORS\\t\" + \"\\t\".join(group_colors.values()) + \"\\n\")\n",
    "    f.write(\"LEGEND_LABELS\\t\" + \"\\t\".join(group_colors.keys()) + \"\\n\\n\")\n",
    "\n",
    "    f.write(\"DATA\\n\")\n",
    "    for entry in itol_entries:\n",
    "        f.write(f\"{entry[0]}\\t{entry[1]}\\t{entry[2]}\\n\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total unique ASVs in AD individuals: 789\n"
     ]
    }
   ],
   "source": [
    "all_asvs_AD = skin_taxa_AD | nares_taxa_AD\n",
    "print(f\"Total unique ASVs in AD individuals: {len(all_asvs_AD)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtered FASTA written to ../Data/Fasta/209766_filtered_by_prevalence_1pct_AD_only.fasta with 789 AD ASVs.\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "# Path to your full FASTA file\n",
    "input_fasta = \"../Data/Fasta/209766_filtered_by_prevalence_1pct.fasta\"\n",
    "\n",
    "# Output FASTA with only AD ASVs\n",
    "output_fasta = \"../Data/Fasta/209766_filtered_by_prevalence_1pct_AD_only.fasta\"\n",
    "\n",
    "# Convert to set for fast lookup\n",
    "asv_ids_to_keep = set(all_asvs_AD)\n",
    "\n",
    "# Filter and write new FASTA\n",
    "with open(output_fasta, \"w\") as out_f:\n",
    "    for record in SeqIO.parse(input_fasta, \"fasta\"):\n",
    "        if record.id in asv_ids_to_keep:\n",
    "            SeqIO.write(record, out_f, \"fasta\")\n",
    "\n",
    "print(f\"Filtered FASTA written to {output_fasta} with {len(asv_ids_to_keep)} AD ASVs.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total unique ASVs in Healthy individuals: 772\n"
     ]
    }
   ],
   "source": [
    "all_asvs_Healthy = skin_taxa_H | nares_taxa_H\n",
    "print(f\"Total unique ASVs in Healthy individuals: {len(all_asvs_Healthy)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtered FASTA written to ../Data/Fasta/209766_filtered_by_prevalence_1pct_Healthy_only.fasta with 772 Healthy ASVs.\n"
     ]
    }
   ],
   "source": [
    "# Path to your full FASTA file\n",
    "input_fasta = \"../Data/Fasta/209766_filtered_by_prevalence_1pct.fasta\"\n",
    "\n",
    "# Output FASTA with only AD ASVs\n",
    "output_fasta = \"../Data/Fasta/209766_filtered_by_prevalence_1pct_Healthy_only.fasta\"\n",
    "\n",
    "# Convert to set for fast lookup\n",
    "asv_ids_to_keep = set(all_asvs_Healthy)\n",
    "\n",
    "# Filter and write new FASTA\n",
    "with open(output_fasta, \"w\") as out_f:\n",
    "    for record in SeqIO.parse(input_fasta, \"fasta\"):\n",
    "        if record.id in asv_ids_to_keep:\n",
    "            SeqIO.write(record, out_f, \"fasta\")\n",
    "\n",
    "print(f\"Filtered FASTA written to {output_fasta} with {len(asv_ids_to_keep)} Healthy ASVs.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<!-- ## Bootstrapped plot -->"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bootstrapping 100 times with 84 paired individuals\n"
     ]
    }
   ],
   "source": [
    "# 1. Find common subjects with both skin and nares samples\n",
    "def get_paired_subjects(skin_df, nares_df):\n",
    "    return set(skin_df['pid']) & set(nares_df['pid'])\n",
    "\n",
    "ad_subjects = list(get_paired_subjects(skin_ADL_samples, nares_AD_samples))\n",
    "healthy_subjects = list(get_paired_subjects(skin_H_samples, nares_H_samples))\n",
    "\n",
    "n_iterations = 100\n",
    "n_paired = min(len(ad_subjects), len(healthy_subjects))  # smallest group size\n",
    "print(f'Bootstrapping {n_iterations} times with {n_paired} paired individuals')\n",
    "\n",
    "ad_overlap = np.zeros(n_iterations)\n",
    "healthy_overlap = np.zeros(n_iterations)\n",
    "\n",
    "taxa_cols = df.select_dtypes(include=[np.number]).columns\n",
    "\n",
    "for i in range(n_iterations):\n",
    "    # 2. Bootstrap same number of paired individuals\n",
    "    sampled_ad_subjects = np.random.choice(ad_subjects, n_paired, replace=True)\n",
    "    sampled_healthy_subjects = np.random.choice(healthy_subjects, n_paired, replace=True)\n",
    "\n",
    "    # 3. Get skin and nares samples for each subject and stack\n",
    "    ad_skin = skin_ADL_samples[skin_ADL_samples['pid'].isin(sampled_ad_subjects)][taxa_cols]\n",
    "    ad_nares = nares_AD_samples[nares_AD_samples['pid'].isin(sampled_ad_subjects)][taxa_cols]\n",
    "\n",
    "    h_skin = skin_H_samples[skin_H_samples['pid'].isin(sampled_healthy_subjects)][taxa_cols]\n",
    "    h_nares = nares_H_samples[nares_H_samples['pid'].isin(sampled_healthy_subjects)][taxa_cols]\n",
    "\n",
    "    # 4. Determine taxa present in each body site\n",
    "    present_ad_skin = (ad_skin > 0).any(axis=0)\n",
    "    present_ad_nares = (ad_nares > 0).any(axis=0)\n",
    "    present_h_skin = (h_skin > 0).any(axis=0)\n",
    "    present_h_nares = (h_nares > 0).any(axis=0)\n",
    "\n",
    "    # 5. Store overlap\n",
    "    ad_overlap[i] = np.sum(present_ad_skin & present_ad_nares)\n",
    "    healthy_overlap[i] = np.sum(present_h_skin & present_h_nares)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AD overlap (mean ± SD): 474.6 ± 23.4\n",
      "Healthy overlap (mean ± SD): 382.3 ± 20.6\n"
     ]
    }
   ],
   "source": [
    "print(f\"AD overlap (mean ± SD): {np.mean(ad_overlap):.1f} ± {np.std(ad_overlap):.1f}\")\n",
    "print(f\"Healthy overlap (mean ± SD): {np.mean(healthy_overlap):.1f} ± {np.std(healthy_overlap):.1f}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.\n",
      "  with pd.option_context('mode.use_inf_as_na', True):\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.\n",
      "  with pd.option_context('mode.use_inf_as_na', True):\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.\n",
      "  with pd.option_context('mode.use_inf_as_na', True):\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.\n",
      "  with pd.option_context('mode.use_inf_as_na', True):\n"
     ]
    }
   ],
   "source": [
    "# Define fixed-width bins\n",
    "all_vals = np.concatenate([ad_overlap, healthy_overlap])\n",
    "bin_min = int(np.floor(all_vals.min()) - 2)\n",
    "bin_max = int(np.ceil(all_vals.max()) + 2)\n",
    "bins = np.arange(bin_min, bin_max + 2, 2)  # Finer binning if values are close\n",
    "\n",
    "# Create figure\n",
    "plt.figure(figsize=(8, 5))\n",
    "\n",
    "# Plot AD histogram and KDE\n",
    "sns.histplot(ad_overlap, bins=bins, color='#e31a1c', stat='density', alpha=0.6, label='AD')\n",
    "sns.kdeplot(ad_overlap, color='#e31a1c', linewidth=2, alpha=1.0)\n",
    "\n",
    "# Plot Healthy histogram and KDE\n",
    "sns.histplot(healthy_overlap, bins=bins, color='#78a4c1', stat='density', alpha=0.6, label='Healthy')\n",
    "sns.kdeplot(healthy_overlap, color='#78a4c1', linewidth=2, alpha=1.0)\n",
    "\n",
    "# Customize axis and layout\n",
    "plt.xlabel(\"Shared Taxa (Skin ∩ Nares)\", fontsize=14)\n",
    "plt.ylabel(\"Probability Density\", fontsize=14)\n",
    "plt.title(\"Overlap of Shared ASVs Between Skin and Nares Per Child\", fontsize=16)\n",
    "# plt.xlim(150, 320)\n",
    "plt.legend(title=\"Skin Status\")\n",
    "\n",
    "stat, p = mannwhitneyu(ad_overlap, healthy_overlap)\n",
    "plt.gca().text(\n",
    "    0.02, 0.97,  # Adjust Y value as needed to position under legend\n",
    "    f\"Mann–Whitney U\\np={p:.1e}\",\n",
    "    transform=plt.gca().transAxes,\n",
    "    ha='left', va='top', fontsize=10,\n",
    "    bbox=dict(boxstyle='round', facecolor='white', edgecolor='none', alpha=0.7)\n",
    ")\n",
    "\n",
    "plt.tight_layout()\n",
    "\n",
    "# Save figure\n",
    "plt.savefig(\"../Plots/Analysis_figures/Co-occurrence/bootstrap_overlap_taxa.png\", dpi=600)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "qiime2-metagenome-2024.10",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
