{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Alpha Diversity by Individual AD Status and Region"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import Python packages\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import biom\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from itertools import cycle\n",
    "import os\n",
    "from matplotlib.colors import ListedColormap\n",
    "from matplotlib.colors import to_rgba\n",
    "from scipy.stats import mannwhitneyu\n",
    "from skbio.diversity import alpha_diversity\n",
    "from matplotlib.patches import Ellipse\n",
    "import matplotlib.transforms as transforms\n",
    "from matplotlib.patches import Circle\n",
    "from matplotlib.colors import to_hex\n",
    "import statsmodels.api as sm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "case_type\n",
       "case-nonlesional skin       111\n",
       "case-anterior nares         108\n",
       "case-lesional skin          107\n",
       "control-anterior nares       89\n",
       "control-nonlesional skin     87\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load the metadata\n",
    "metadata_path = '../Data/Metadata/updated_clean_ant_skin_metadata.tab'\n",
    "metadata = pd.read_csv(metadata_path, sep='\\t')\n",
    "metadata['case_type'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#sample-id</th>\n",
       "      <th>PlateNumber</th>\n",
       "      <th>PlateLocation</th>\n",
       "      <th>i5</th>\n",
       "      <th>i5Sequence</th>\n",
       "      <th>i7</th>\n",
       "      <th>i7Sequence</th>\n",
       "      <th>identifier</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Plate ID</th>\n",
       "      <th>...</th>\n",
       "      <th>sex</th>\n",
       "      <th>enrolment_date</th>\n",
       "      <th>enrolment_season</th>\n",
       "      <th>hiv_exposure</th>\n",
       "      <th>hiv_status</th>\n",
       "      <th>household_size</th>\n",
       "      <th>o_scorad</th>\n",
       "      <th>FWD_filepath</th>\n",
       "      <th>REV_filepath</th>\n",
       "      <th>o_scorad_adj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Ca009ST_L</td>\n",
       "      <td>1</td>\n",
       "      <td>A1</td>\n",
       "      <td>SA501</td>\n",
       "      <td>ATCGTACG</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA501</td>\n",
       "      <td>CGAGAGTT-ATCGTACG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>...</td>\n",
       "      <td>male</td>\n",
       "      <td>4/16/2015</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>4.0</td>\n",
       "      <td>40</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>900221</td>\n",
       "      <td>1</td>\n",
       "      <td>B1</td>\n",
       "      <td>SA502</td>\n",
       "      <td>ACTATCTG</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA502</td>\n",
       "      <td>CGAGAGTT-ACTATCTG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>...</td>\n",
       "      <td>female</td>\n",
       "      <td>8/11/2015</td>\n",
       "      <td>Winter</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>7.0</td>\n",
       "      <td>34</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Ca010EB_L</td>\n",
       "      <td>1</td>\n",
       "      <td>C1</td>\n",
       "      <td>SA503</td>\n",
       "      <td>TAGCGAGT</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA503</td>\n",
       "      <td>CGAGAGTT-TAGCGAGT</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>...</td>\n",
       "      <td>female</td>\n",
       "      <td>11/20/2014</td>\n",
       "      <td>Spring</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>7.0</td>\n",
       "      <td>21</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>900460</td>\n",
       "      <td>1</td>\n",
       "      <td>D1</td>\n",
       "      <td>SA504</td>\n",
       "      <td>CTGCGTGT</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA504</td>\n",
       "      <td>CGAGAGTT-CTGCGTGT</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>...</td>\n",
       "      <td>female</td>\n",
       "      <td>9/23/2015</td>\n",
       "      <td>Spring</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>40</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>900051</td>\n",
       "      <td>1</td>\n",
       "      <td>E1</td>\n",
       "      <td>SA505</td>\n",
       "      <td>TCATCGAG</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA505</td>\n",
       "      <td>CGAGAGTT-TCATCGAG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>...</td>\n",
       "      <td>male</td>\n",
       "      <td>4/21/2015</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>7.0</td>\n",
       "      <td>41</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>497</th>\n",
       "      <td>Ca006ON_L_2</td>\n",
       "      <td>6</td>\n",
       "      <td>H1</td>\n",
       "      <td>SA508</td>\n",
       "      <td>GACACCGT</td>\n",
       "      <td>SB701</td>\n",
       "      <td>CTCGACTT</td>\n",
       "      <td>SB701SA508</td>\n",
       "      <td>CTCGACTT-GACACCGT</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>...</td>\n",
       "      <td>female</td>\n",
       "      <td>3/25/2015</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>3.0</td>\n",
       "      <td>34</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>498</th>\n",
       "      <td>Ca006ON_NL</td>\n",
       "      <td>6</td>\n",
       "      <td>F2</td>\n",
       "      <td>SA506</td>\n",
       "      <td>CGTGAGTG</td>\n",
       "      <td>SB702</td>\n",
       "      <td>CGAAGTAT</td>\n",
       "      <td>SB702SA506</td>\n",
       "      <td>CGAAGTAT-CGTGAGTG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>...</td>\n",
       "      <td>female</td>\n",
       "      <td>3/25/2015</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>3.0</td>\n",
       "      <td>34</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>499</th>\n",
       "      <td>Ca006ON_NL_2</td>\n",
       "      <td>6</td>\n",
       "      <td>H2</td>\n",
       "      <td>SA508</td>\n",
       "      <td>GACACCGT</td>\n",
       "      <td>SB702</td>\n",
       "      <td>CGAAGTAT</td>\n",
       "      <td>SB702SA508</td>\n",
       "      <td>CGAAGTAT-GACACCGT</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>...</td>\n",
       "      <td>female</td>\n",
       "      <td>3/25/2015</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>3.0</td>\n",
       "      <td>34</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>500</th>\n",
       "      <td>Ca006ON_PN</td>\n",
       "      <td>6</td>\n",
       "      <td>F3</td>\n",
       "      <td>SA506</td>\n",
       "      <td>CGTGAGTG</td>\n",
       "      <td>SB703</td>\n",
       "      <td>TAGCAGCT</td>\n",
       "      <td>SB703SA506</td>\n",
       "      <td>TAGCAGCT-CGTGAGTG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>...</td>\n",
       "      <td>female</td>\n",
       "      <td>3/25/2015</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>3.0</td>\n",
       "      <td>34</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>501</th>\n",
       "      <td>Ca006ON_PN_2</td>\n",
       "      <td>6</td>\n",
       "      <td>H3</td>\n",
       "      <td>SA508</td>\n",
       "      <td>GACACCGT</td>\n",
       "      <td>SB703</td>\n",
       "      <td>TAGCAGCT</td>\n",
       "      <td>SB703SA508</td>\n",
       "      <td>TAGCAGCT-GACACCGT</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>...</td>\n",
       "      <td>female</td>\n",
       "      <td>3/25/2015</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>3.0</td>\n",
       "      <td>34</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>502 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       #sample-id  PlateNumber PlateLocation     i5 i5Sequence     i7  \\\n",
       "0       Ca009ST_L            1            A1  SA501   ATCGTACG  SA701   \n",
       "1          900221            1            B1  SA502   ACTATCTG  SA701   \n",
       "2       Ca010EB_L            1            C1  SA503   TAGCGAGT  SA701   \n",
       "3          900460            1            D1  SA504   CTGCGTGT  SA701   \n",
       "4          900051            1            E1  SA505   TCATCGAG  SA701   \n",
       "..            ...          ...           ...    ...        ...    ...   \n",
       "497   Ca006ON_L_2            6            H1  SA508   GACACCGT  SB701   \n",
       "498    Ca006ON_NL            6            F2  SA506   CGTGAGTG  SB702   \n",
       "499  Ca006ON_NL_2            6            H2  SA508   GACACCGT  SB702   \n",
       "500    Ca006ON_PN            6            F3  SA506   CGTGAGTG  SB703   \n",
       "501  Ca006ON_PN_2            6            H3  SA508   GACACCGT  SB703   \n",
       "\n",
       "    i7Sequence  identifier           Sequence      Plate ID  ...     sex  \\\n",
       "0     CGAGAGTT  SA701SA501  CGAGAGTT-ATCGTACG  1.010000e+21  ...    male   \n",
       "1     CGAGAGTT  SA701SA502  CGAGAGTT-ACTATCTG  1.010000e+21  ...  female   \n",
       "2     CGAGAGTT  SA701SA503  CGAGAGTT-TAGCGAGT  1.010000e+21  ...  female   \n",
       "3     CGAGAGTT  SA701SA504  CGAGAGTT-CTGCGTGT  1.010000e+21  ...  female   \n",
       "4     CGAGAGTT  SA701SA505  CGAGAGTT-TCATCGAG  1.010000e+21  ...    male   \n",
       "..         ...         ...                ...           ...  ...     ...   \n",
       "497   CTCGACTT  SB701SA508  CTCGACTT-GACACCGT  1.010000e+21  ...  female   \n",
       "498   CGAAGTAT  SB702SA506  CGAAGTAT-CGTGAGTG  1.010000e+21  ...  female   \n",
       "499   CGAAGTAT  SB702SA508  CGAAGTAT-GACACCGT  1.010000e+21  ...  female   \n",
       "500   TAGCAGCT  SB703SA506  TAGCAGCT-CGTGAGTG  1.010000e+21  ...  female   \n",
       "501   TAGCAGCT  SB703SA508  TAGCAGCT-GACACCGT  1.010000e+21  ...  female   \n",
       "\n",
       "     enrolment_date enrolment_season hiv_exposure hiv_status household_size  \\\n",
       "0         4/16/2015          Autumn     Unexposed   negative            4.0   \n",
       "1         8/11/2015           Winter    Unexposed   negative            7.0   \n",
       "2        11/20/2014           Spring    Unexposed   negative            7.0   \n",
       "3         9/23/2015           Spring    Unexposed        NaN            4.0   \n",
       "4         4/21/2015          Autumn     Unexposed   negative            7.0   \n",
       "..              ...              ...          ...        ...            ...   \n",
       "497       3/25/2015           Autumn    Unexposed   negative            3.0   \n",
       "498       3/25/2015           Autumn    Unexposed   negative            3.0   \n",
       "499       3/25/2015           Autumn    Unexposed   negative            3.0   \n",
       "500       3/25/2015           Autumn    Unexposed   negative            3.0   \n",
       "501       3/25/2015           Autumn    Unexposed   negative            3.0   \n",
       "\n",
       "    o_scorad                                       FWD_filepath  \\\n",
       "0         40  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "1         34  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "2         21  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "3         40  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "4         41  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "..       ...                                                ...   \n",
       "497       34  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "498       34  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "499       34  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "500       34  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "501       34  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "\n",
       "                                          REV_filepath o_scorad_adj  \n",
       "0    /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...           40  \n",
       "1    /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...           34  \n",
       "2    /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...           21  \n",
       "3    /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...           40  \n",
       "4    /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...           41  \n",
       "..                                                 ...          ...  \n",
       "497  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...           34  \n",
       "498  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...           34  \n",
       "499  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...           34  \n",
       "500  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...           34  \n",
       "501  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...           34  \n",
       "\n",
       "[502 rows x 33 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata['o_scorad_adj'] = metadata['o_scorad'].fillna(0)\n",
    "metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "individual_case_location\n",
       "AD skin Umtata        122\n",
       "AD skin Cape Town      96\n",
       "AD nares Umtata        61\n",
       "H nares Umtata         55\n",
       "H skin Umtata          52\n",
       "AD nares Cape Town     47\n",
       "H skin Cape Town       35\n",
       "H nares Cape Town      34\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create mapping dictionary for case types to individual cases\n",
    "case_type_mapping = {\n",
    "    'case-nonlesional skin': 'AD skin',\n",
    "    'case-anterior nares': 'AD nares', \n",
    "    'case-lesional skin': 'AD skin',\n",
    "    'control-anterior nares': 'H nares',\n",
    "    'control-nonlesional skin': 'H skin'\n",
    "}\n",
    "\n",
    "# Create new column using the mapping\n",
    "metadata['individual_case'] = metadata['case_type'].map(case_type_mapping)\n",
    "\n",
    "metadata['individual_case_location'] = metadata['individual_case'] + ' ' + metadata['area']\n",
    "metadata['individual_case_location'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata.to_csv('../Data/Metadata/updated_clean_ant_skin_metadata_microbiome_type-location.tab', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def darken_color(color, amount=0.3):\n",
    "    \"\"\"\n",
    "    Darken a given color by a certain amount.\n",
    "    \n",
    "    Parameters:\n",
    "    - color: The base color (as a hex string or color name).\n",
    "    - amount: The amount to darken the color by (default: 0.3).\n",
    "    \n",
    "    Returns:\n",
    "    - A darkened color as a hex string.\n",
    "    \"\"\"\n",
    "    c = to_rgba(color)\n",
    "    return (c[0] * (1 - amount), c[1] * (1 - amount), c[2] * (1 - amount), c[3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def brighten_color(color, amount=0.3):\n",
    "    \"\"\"\n",
    "    Brighten a color by increasing its RGB intensity without blending with white.\n",
    "    \n",
    "    Parameters:\n",
    "    - color: color name or hex code\n",
    "    - amount: brightness boost factor (0 = no change, 1 = full brightness)\n",
    "\n",
    "    Returns:\n",
    "    - Hex string of the brightened color.\n",
    "    \"\"\"\n",
    "    rgba = to_rgba(color)\n",
    "    r = min(rgba[0] * (1 + amount), 1.0)\n",
    "    g = min(rgba[1] * (1 + amount), 1.0)\n",
    "    b = min(rgba[2] * (1 + amount), 1.0)\n",
    "    return to_hex((r, g, b, rgba[3]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_faith_pd_histo(metadata, group_col):\n",
    "    metadata = metadata.set_index('#sample-id')\n",
    "\n",
    "    # Load Faith PD data\n",
    "    feature_table = pd.read_csv(\n",
    "        \"../Analyses/Xebec/Output/results/alpha_div/phylo/faith_pd/vector.tsv\", sep=\"\\t\"\n",
    "    )\n",
    "    feature_table = feature_table.set_index(feature_table.columns[0])\n",
    "\n",
    "\n",
    "    # Keep only overlapping samples\n",
    "    common_samples = metadata.index.intersection(feature_table.index)\n",
    "    metadata = metadata.loc[common_samples].copy()\n",
    "    metadata['Faith_PD'] = feature_table.loc[common_samples]\n",
    "\n",
    "    # Desired group order\n",
    "    desired_order = ['H skin Cape Town', 'AD skin Cape Town', 'H skin Umtata', 'AD skin Umtata']\n",
    "\n",
    "    # Color palette\n",
    "    # palette = {\n",
    "    #     'H skin Cape Town': '#A7C7E7',\n",
    "    #     'H skin Umtata': '#ADD8E6',\n",
    "    #     'AD skin Cape Town': '#d2b48c',\n",
    "    #     'AD skin Umtata': '#fa8072'\n",
    "    # }\n",
    "\n",
    "    palette = {\n",
    "        'H skin Cape Town': '#7FBCEB',   # Soft but clearer blue\n",
    "        'H skin Umtata': '#66C2EE',      # Light turquoise-blue with more saturation\n",
    "        'AD skin Cape Town': '#C9A34F',  # Warm golden tan, more vibrant than tan\n",
    "        'AD skin Umtata': '#F0806B'      # Coral-salmon tone, richer than original salmon\n",
    "    }\n",
    "\n",
    "\n",
    "\n",
    "    # Add group counts for x-axis labels\n",
    "    group_counts = metadata[group_col].value_counts().to_dict()\n",
    "    new_labels = [\n",
    "        f\"H child\\n(n={group_counts.get('H skin Cape Town', 0)})\",\n",
    "        f\"AD child\\n(n={group_counts.get('AD skin Cape Town', 0)})\",\n",
    "        f\"H child\\n(n={group_counts.get('H skin Umtata', 0)})\",\n",
    "        f\"AD child\\n(n={group_counts.get('AD skin Umtata', 0)})\"\n",
    "    ]\n",
    "\n",
    "    # Start figure\n",
    "    plt.figure(figsize=(8, 4.5))\n",
    "    ax = sns.boxplot(x=group_col, y='Faith_PD', data=metadata, palette=palette, order=desired_order)\n",
    "\n",
    "    # Darker dots on top\n",
    "    # darker_palette = {k: darken_color(v) for k, v in palette.items()}\n",
    "    brighter_palette = {k: brighten_color(v, amount=0.3) for k, v in palette.items()}\n",
    "\n",
    "    sns.stripplot(x=group_col, y='Faith_PD', data=metadata, palette=brighter_palette,\n",
    "                  jitter=True, dodge=False, ax=ax, linewidth=0.6, order=desired_order)\n",
    "\n",
    "    # Customize axes and labels\n",
    "    plt.suptitle(f'Alpha Diversity by Region', fontsize=20, y = 1.03)\n",
    "    plt.title(f'Cape Town                      Umtata', fontsize=16, y = 1)\n",
    "\n",
    "    plt.xlabel(' ')\n",
    "    plt.ylabel('Faith PD', fontsize=16)\n",
    "    plt.xticks(ticks=range(len(new_labels)), labels=new_labels, ha='center', fontsize=16)\n",
    "\n",
    "    # Pairwise significance testing (only p <= 0.05 shown)\n",
    "    groups = desired_order\n",
    "    p_values = {}\n",
    "    y_max = metadata['Faith_PD'].max()\n",
    "    height_step = 1.2\n",
    "\n",
    "    for i, group1 in enumerate(groups):\n",
    "        for j, group2 in enumerate(groups):\n",
    "            if i < j:\n",
    "                vals1 = metadata[metadata[group_col] == group1]['Faith_PD']\n",
    "                vals2 = metadata[metadata[group_col] == group2]['Faith_PD']\n",
    "                stat, p = mannwhitneyu(vals1, vals2, alternative='two-sided')\n",
    "                p_values[f'{group1} vs {group2}'] = p\n",
    "\n",
    "                # Define special comparison to always annotate\n",
    "                always_annotate = ('H skin Cape Town', 'AD skin Cape Town')\n",
    "\n",
    "                # Annotate significant comparisons or the specified non-significant one\n",
    "                if p <= 0.05 or (group1, group2) == always_annotate or (group2, group1) == always_annotate:\n",
    "                    if p < 0.001:\n",
    "                        label = '***  ' + f\"{p:.2}\"\n",
    "                    elif p < 0.01:\n",
    "                        label = '**  ' + f\"{p:.2}\"\n",
    "                    elif p < 0.05:\n",
    "                        label = '*  ' + f\"{p:.2}\"\n",
    "                    else:\n",
    "                        label = f\"{p:.2}\"\n",
    "\n",
    "                    x1, x2 = i, j\n",
    "                    y = y_max + height_step\n",
    "                    plt.plot([x1, x1, x2, x2], [y, y + 0.1, y + 0.1, y], lw=1, color='black')\n",
    "                    plt.text((x1 + x2) * 0.5, y + 0.1, label, ha='center', va='bottom', fontsize=10)\n",
    "                    y_max += height_step + 0.8\n",
    "\n",
    "\n",
    "    # Save the figure\n",
    "    plt.savefig('../Plots/Analysis_figures/Diversity/CapeTown_Umtata_FaithPD.png', dpi=600, bbox_inches='tight', pad_inches=0.1)\n",
    "    plt.savefig('../Plots/Analysis_figures/Diversity/CapeTown_Umtata_FaithPD.svg')\n",
    "\n",
    "    # Print p-values\n",
    "    print(\"Pairwise Mann-Whitney U test p-values:\")\n",
    "    for comparison, p_value in p_values.items():\n",
    "        print(f\"{comparison}: p-value = {p_value:.2e}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_faith_pd_histo(metadata, group_col):\n",
    "    import statsmodels.api as sm\n",
    "    from scipy.stats import mannwhitneyu\n",
    "\n",
    "    metadata = metadata.set_index('#sample-id')\n",
    "\n",
    "    # Load Faith PD data\n",
    "    feature_table = pd.read_csv(\n",
    "        \"../Analyses/Xebec/Output/results/alpha_div/phylo/faith_pd/vector.tsv\", sep=\"\\t\"\n",
    "    )\n",
    "    feature_table = feature_table.set_index(feature_table.columns[0])\n",
    "\n",
    "    # Keep only overlapping samples\n",
    "    common_samples = metadata.index.intersection(feature_table.index)\n",
    "    metadata = metadata.loc[common_samples].copy()\n",
    "    metadata['Faith_PD'] = feature_table.loc[common_samples]\n",
    "\n",
    "    # 🔧 SCORAD adjustment via linear regression\n",
    "\n",
    "    metadata['o_scorad_adj'] = pd.to_numeric(metadata['o_scorad_adj'], errors='coerce')\n",
    "\n",
    "    # Drop rows with missing values in o_scorad_adj or Faith_PD\n",
    "    metadata = metadata.dropna(subset=['Faith_PD', 'o_scorad_adj'])\n",
    "    \n",
    "    model = sm.OLS(metadata['Faith_PD'], sm.add_constant(metadata['o_scorad_adj'])).fit()\n",
    "    metadata['Faith_PD_resid'] = model.resid  # <-- SCORAD-adjusted Faith PD\n",
    "\n",
    "    # Desired group order\n",
    "    desired_order = ['H skin Cape Town', 'AD skin Cape Town', 'H skin Umtata', 'AD skin Umtata']\n",
    "    # desired_order = ['AD skin Cape Town', 'AD skin Umtata']\n",
    "\n",
    "    # Color palette\n",
    "    palette = {\n",
    "        'H skin Cape Town': '#7FBCEB',\n",
    "        'H skin Umtata': '#66C2EE',\n",
    "        'AD skin Cape Town': '#C9A34F',\n",
    "        'AD skin Umtata': '#F0806B'\n",
    "    }\n",
    "\n",
    "    # Add group counts for x-axis labels\n",
    "    group_counts = metadata[group_col].value_counts().to_dict()\n",
    "    new_labels = [\n",
    "        f\"H child\\n(n={group_counts.get('H skin Cape Town', 0)})\",\n",
    "        f\"AD child\\n(n={group_counts.get('AD skin Cape Town', 0)})\",\n",
    "        f\"H child\\n(n={group_counts.get('H skin Umtata', 0)})\",\n",
    "        f\"AD child\\n(n={group_counts.get('AD skin Umtata', 0)})\"\n",
    "    ]\n",
    "\n",
    "    # Start figure\n",
    "    plt.figure(figsize=(4.5, 4.5))\n",
    "    ax = sns.boxplot(x=group_col, y='Faith_PD_resid', data=metadata, palette=palette, order=desired_order)\n",
    "\n",
    "    # Brighter dots for scatter\n",
    "    brighter_palette = {k: brighten_color(v, amount=0.3) for k, v in palette.items()}\n",
    "    sns.stripplot(x=group_col, y='Faith_PD_resid', data=metadata, palette=brighter_palette,\n",
    "                  jitter=True, dodge=False, ax=ax, linewidth=0.6, order=desired_order)\n",
    "\n",
    "    ax.grid(False)\n",
    "\n",
    "    # Customize axes and labels\n",
    "    plt.suptitle(f'SCORAD-Adjusted Alpha Diversity', fontsize=18, y=1.03)\n",
    "    plt.title(f'Cape Town              Umtata', fontsize=16, y=1)\n",
    "\n",
    "    plt.xlabel(' ')\n",
    "    plt.ylabel('Faith PD (residual)', fontsize=16)\n",
    "    plt.xticks(ticks=range(len(new_labels)), labels=new_labels, ha='center', fontsize=16)\n",
    "\n",
    "    # Pairwise significance testing on adjusted values\n",
    "    groups = desired_order\n",
    "    p_values = {}\n",
    "    y_max = metadata['Faith_PD_resid'].max()\n",
    "    height_step = 1.2\n",
    "\n",
    "    for i, group1 in enumerate(groups):\n",
    "        for j, group2 in enumerate(groups):\n",
    "            if i < j:\n",
    "                vals1 = metadata[metadata[group_col] == group1]['Faith_PD_resid']\n",
    "                vals2 = metadata[metadata[group_col] == group2]['Faith_PD_resid']\n",
    "\n",
    "                # ✅ Skip test if either group has no data\n",
    "                if len(vals1) == 0 or len(vals2) == 0:\n",
    "                    print(f\"Skipping comparison {group1} vs {group2} (one group is empty)\")\n",
    "                    continue\n",
    "\n",
    "                stat, p = mannwhitneyu(vals1, vals2, alternative='two-sided')\n",
    "                p_values[f'{group1} vs {group2}'] = p\n",
    "\n",
    "                # Define special comparison to always annotate\n",
    "                always_annotate = ('H skin Cape Town', 'AD skin Cape Town')\n",
    "\n",
    "                if p <= 0.05 or (group1, group2) == always_annotate or (group2, group1) == always_annotate:\n",
    "                    if p < 0.001:\n",
    "                        label = '***  ' + f\"{p:.2}\"\n",
    "                    elif p < 0.01:\n",
    "                        label = '**  ' + f\"{p:.2}\"\n",
    "                    elif p < 0.05:\n",
    "                        label = '*  ' + f\"{p:.2}\"\n",
    "                    else:\n",
    "                        label = f\"{p:.2}\"\n",
    "\n",
    "                    x1, x2 = i, j\n",
    "                    y = y_max + height_step\n",
    "                    plt.plot([x1, x1, x2, x2], [y, y + 0.1, y + 0.1, y], lw=1, color='black')\n",
    "                    plt.text((x1 + x2) * 0.5, y + 0.1, label, ha='center', va='bottom', fontsize=10)\n",
    "                    y_max += height_step + 0.8\n",
    "\n",
    "    # Optional: display R² from regression\n",
    "    # r_squared = model.rsquared\n",
    "    # ax.text(0.98, 0.98, f\"R² = {r_squared:.2f}\", transform=ax.transAxes,\n",
    "    #         ha='right', va='top', fontsize=10, color='black')\n",
    "\n",
    "    # Save the figure\n",
    "    plt.savefig('../Plots/Analysis_figures/Diversity/CapeTown_Umtata_FaithPD_SCORADadjusted.png', dpi=600, bbox_inches='tight', pad_inches=0.1)\n",
    "    plt.savefig('../Plots/Analysis_figures/Diversity/CapeTown_Umtata_FaithPD_SCORADadjusted.svg')\n",
    "\n",
    "    # Print p-values\n",
    "    print(\"Pairwise Mann-Whitney U test p-values (on SCORAD-adjusted Faith PD):\")\n",
    "    for comparison, p_value in p_values.items():\n",
    "        print(f\"{comparison}: p-value = {p_value:.2e}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/22/yck9vwx53w1c38tvj_c0_tz00000gn/T/ipykernel_23319/2389643814.py:55: FutureWarning: Passing `palette` without assigning `hue` is deprecated.\n",
      "  sns.stripplot(x=group_col, y='Faith_PD_resid', data=metadata, palette=brighter_palette,\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.\n",
      "  with pd.option_context('mode.use_inf_as_na', True):\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.\n",
      "  with pd.option_context('mode.use_inf_as_na', True):\n",
      "/Users/yangchen/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/seaborn/_oldcore.py:1075: FutureWarning: When grouping with a length-1 list-like, you will need to pass a length-1 tuple to get_group in a future version of pandas. Pass `(name,)` instead of `name` to silence this warning.\n",
      "  data_subset = grouped_data.get_group(pd_key)\n"
     ]
    },
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: '/Users/yangchen/PhD/Gallo_lab/16S_AD_South-Africa/Plots/Analysis_figures/Diversity/CapeTown_Umtata_FaithPD_SCORADadjusted.png'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[10], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m# Plot Alpha Diversity plots for both V1-V3 and V4\u001b[39;00m\n\u001b[0;32m----> 2\u001b[0m \u001b[43mplot_faith_pd_histo\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m      3\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmetadata\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmetadata\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m      4\u001b[0m \u001b[43m    \u001b[49m\u001b[43mgroup_col\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mindividual_case_location\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\n\u001b[1;32m      5\u001b[0m \u001b[43m)\u001b[49m\n",
      "Cell \u001b[0;32mIn[9], line 113\u001b[0m, in \u001b[0;36mplot_faith_pd_histo\u001b[0;34m(metadata, group_col)\u001b[0m\n\u001b[1;32m    105\u001b[0m                 y_max \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m height_step \u001b[38;5;241m+\u001b[39m \u001b[38;5;241m0.8\u001b[39m\n\u001b[1;32m    107\u001b[0m \u001b[38;5;66;03m# Optional: display R² from regression\u001b[39;00m\n\u001b[1;32m    108\u001b[0m \u001b[38;5;66;03m# r_squared = model.rsquared\u001b[39;00m\n\u001b[1;32m    109\u001b[0m \u001b[38;5;66;03m# ax.text(0.98, 0.98, f\"R² = {r_squared:.2f}\", transform=ax.transAxes,\u001b[39;00m\n\u001b[1;32m    110\u001b[0m \u001b[38;5;66;03m#         ha='right', va='top', fontsize=10, color='black')\u001b[39;00m\n\u001b[1;32m    111\u001b[0m \n\u001b[1;32m    112\u001b[0m \u001b[38;5;66;03m# Save the figure\u001b[39;00m\n\u001b[0;32m--> 113\u001b[0m \u001b[43mplt\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msavefig\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m../Plots/Analysis_figures/Diversity/CapeTown_Umtata_FaithPD_SCORADadjusted.png\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdpi\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m600\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mbbox_inches\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mtight\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mpad_inches\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m0.1\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m    114\u001b[0m plt\u001b[38;5;241m.\u001b[39msavefig(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m../Plots/Analysis_figures/Diversity/CapeTown_Umtata_FaithPD_SCORADadjusted.svg\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m    116\u001b[0m \u001b[38;5;66;03m# Print p-values\u001b[39;00m\n",
      "File \u001b[0;32m~/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/matplotlib/pyplot.py:1023\u001b[0m, in \u001b[0;36msavefig\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m   1020\u001b[0m \u001b[38;5;129m@_copy_docstring_and_deprecators\u001b[39m(Figure\u001b[38;5;241m.\u001b[39msavefig)\n\u001b[1;32m   1021\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21msavefig\u001b[39m(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs):\n\u001b[1;32m   1022\u001b[0m     fig \u001b[38;5;241m=\u001b[39m gcf()\n\u001b[0;32m-> 1023\u001b[0m     res \u001b[38;5;241m=\u001b[39m \u001b[43mfig\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msavefig\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1024\u001b[0m     fig\u001b[38;5;241m.\u001b[39mcanvas\u001b[38;5;241m.\u001b[39mdraw_idle()  \u001b[38;5;66;03m# Need this if 'transparent=True', to reset colors.\u001b[39;00m\n\u001b[1;32m   1025\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m res\n",
      "File \u001b[0;32m~/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/matplotlib/figure.py:3378\u001b[0m, in \u001b[0;36mFigure.savefig\u001b[0;34m(self, fname, transparent, **kwargs)\u001b[0m\n\u001b[1;32m   3374\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m ax \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39maxes:\n\u001b[1;32m   3375\u001b[0m         stack\u001b[38;5;241m.\u001b[39menter_context(\n\u001b[1;32m   3376\u001b[0m             ax\u001b[38;5;241m.\u001b[39mpatch\u001b[38;5;241m.\u001b[39m_cm_set(facecolor\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mnone\u001b[39m\u001b[38;5;124m'\u001b[39m, edgecolor\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mnone\u001b[39m\u001b[38;5;124m'\u001b[39m))\n\u001b[0;32m-> 3378\u001b[0m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcanvas\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mprint_figure\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfname\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/matplotlib/backend_bases.py:2366\u001b[0m, in \u001b[0;36mFigureCanvasBase.print_figure\u001b[0;34m(self, filename, dpi, facecolor, edgecolor, orientation, format, bbox_inches, pad_inches, bbox_extra_artists, backend, **kwargs)\u001b[0m\n\u001b[1;32m   2362\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m   2363\u001b[0m     \u001b[38;5;66;03m# _get_renderer may change the figure dpi (as vector formats\u001b[39;00m\n\u001b[1;32m   2364\u001b[0m     \u001b[38;5;66;03m# force the figure dpi to 72), so we need to set it again here.\u001b[39;00m\n\u001b[1;32m   2365\u001b[0m     \u001b[38;5;28;01mwith\u001b[39;00m cbook\u001b[38;5;241m.\u001b[39m_setattr_cm(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mfigure, dpi\u001b[38;5;241m=\u001b[39mdpi):\n\u001b[0;32m-> 2366\u001b[0m         result \u001b[38;5;241m=\u001b[39m \u001b[43mprint_method\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   2367\u001b[0m \u001b[43m            \u001b[49m\u001b[43mfilename\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   2368\u001b[0m \u001b[43m            \u001b[49m\u001b[43mfacecolor\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mfacecolor\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   2369\u001b[0m \u001b[43m            \u001b[49m\u001b[43medgecolor\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43medgecolor\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   2370\u001b[0m \u001b[43m            \u001b[49m\u001b[43morientation\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43morientation\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   2371\u001b[0m \u001b[43m            \u001b[49m\u001b[43mbbox_inches_restore\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m_bbox_inches_restore\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   2372\u001b[0m \u001b[43m            \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   2373\u001b[0m \u001b[38;5;28;01mfinally\u001b[39;00m:\n\u001b[1;32m   2374\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m bbox_inches \u001b[38;5;129;01mand\u001b[39;00m restore_bbox:\n",
      "File \u001b[0;32m~/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/matplotlib/backend_bases.py:2232\u001b[0m, in \u001b[0;36mFigureCanvasBase._switch_canvas_and_return_print_method.<locals>.<lambda>\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m   2228\u001b[0m     optional_kws \u001b[38;5;241m=\u001b[39m {  \u001b[38;5;66;03m# Passed by print_figure for other renderers.\u001b[39;00m\n\u001b[1;32m   2229\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdpi\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mfacecolor\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124medgecolor\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124morientation\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m   2230\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mbbox_inches_restore\u001b[39m\u001b[38;5;124m\"\u001b[39m}\n\u001b[1;32m   2231\u001b[0m     skip \u001b[38;5;241m=\u001b[39m optional_kws \u001b[38;5;241m-\u001b[39m {\u001b[38;5;241m*\u001b[39minspect\u001b[38;5;241m.\u001b[39msignature(meth)\u001b[38;5;241m.\u001b[39mparameters}\n\u001b[0;32m-> 2232\u001b[0m     print_method \u001b[38;5;241m=\u001b[39m functools\u001b[38;5;241m.\u001b[39mwraps(meth)(\u001b[38;5;28;01mlambda\u001b[39;00m \u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs: \u001b[43mmeth\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   2233\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43m{\u001b[49m\u001b[43mk\u001b[49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mv\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mfor\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mk\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mv\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01min\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mkwargs\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mitems\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mif\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mk\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01mnot\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;129;43;01min\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mskip\u001b[49m\u001b[43m}\u001b[49m\u001b[43m)\u001b[49m)\n\u001b[1;32m   2234\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:  \u001b[38;5;66;03m# Let third-parties do as they see fit.\u001b[39;00m\n\u001b[1;32m   2235\u001b[0m     print_method \u001b[38;5;241m=\u001b[39m meth\n",
      "File \u001b[0;32m~/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/matplotlib/backends/backend_agg.py:509\u001b[0m, in \u001b[0;36mFigureCanvasAgg.print_png\u001b[0;34m(self, filename_or_obj, metadata, pil_kwargs)\u001b[0m\n\u001b[1;32m    462\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mprint_png\u001b[39m(\u001b[38;5;28mself\u001b[39m, filename_or_obj, \u001b[38;5;241m*\u001b[39m, metadata\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m, pil_kwargs\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m):\n\u001b[1;32m    463\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m    464\u001b[0m \u001b[38;5;124;03m    Write the figure to a PNG file.\u001b[39;00m\n\u001b[1;32m    465\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    507\u001b[0m \u001b[38;5;124;03m        *metadata*, including the default 'Software' key.\u001b[39;00m\n\u001b[1;32m    508\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m--> 509\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_print_pil\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilename_or_obj\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mpng\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mpil_kwargs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmetadata\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/matplotlib/backends/backend_agg.py:458\u001b[0m, in \u001b[0;36mFigureCanvasAgg._print_pil\u001b[0;34m(self, filename_or_obj, fmt, pil_kwargs, metadata)\u001b[0m\n\u001b[1;32m    453\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m    454\u001b[0m \u001b[38;5;124;03mDraw the canvas, then save it using `.image.imsave` (to which\u001b[39;00m\n\u001b[1;32m    455\u001b[0m \u001b[38;5;124;03m*pil_kwargs* and *metadata* are forwarded).\u001b[39;00m\n\u001b[1;32m    456\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m    457\u001b[0m FigureCanvasAgg\u001b[38;5;241m.\u001b[39mdraw(\u001b[38;5;28mself\u001b[39m)\n\u001b[0;32m--> 458\u001b[0m \u001b[43mmpl\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mimage\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mimsave\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    459\u001b[0m \u001b[43m    \u001b[49m\u001b[43mfilename_or_obj\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mbuffer_rgba\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mformat\u001b[39;49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mfmt\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43morigin\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mupper\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    460\u001b[0m \u001b[43m    \u001b[49m\u001b[43mdpi\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfigure\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdpi\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmetadata\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmetadata\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mpil_kwargs\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mpil_kwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/matplotlib/image.py:1689\u001b[0m, in \u001b[0;36mimsave\u001b[0;34m(fname, arr, vmin, vmax, cmap, format, origin, dpi, metadata, pil_kwargs)\u001b[0m\n\u001b[1;32m   1687\u001b[0m pil_kwargs\u001b[38;5;241m.\u001b[39msetdefault(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mformat\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28mformat\u001b[39m)\n\u001b[1;32m   1688\u001b[0m pil_kwargs\u001b[38;5;241m.\u001b[39msetdefault(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdpi\u001b[39m\u001b[38;5;124m\"\u001b[39m, (dpi, dpi))\n\u001b[0;32m-> 1689\u001b[0m \u001b[43mimage\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msave\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfname\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mpil_kwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/PIL/Image.py:2456\u001b[0m, in \u001b[0;36mImage.save\u001b[0;34m(self, fp, format, **params)\u001b[0m\n\u001b[1;32m   2454\u001b[0m         fp \u001b[38;5;241m=\u001b[39m builtins\u001b[38;5;241m.\u001b[39mopen(filename, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mr+b\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m   2455\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m-> 2456\u001b[0m         fp \u001b[38;5;241m=\u001b[39m \u001b[43mbuiltins\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mopen\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilename\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mw+b\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m   2458\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m   2459\u001b[0m     save_handler(\u001b[38;5;28mself\u001b[39m, fp, filename)\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: '/Users/yangchen/PhD/Gallo_lab/16S_AD_South-Africa/Plots/Analysis_figures/Diversity/CapeTown_Umtata_FaithPD_SCORADadjusted.png'"
     ]
    }
   ],
   "source": [
    "# Plot Alpha Diversity plots for both V1-V3 and V4\n",
    "plot_faith_pd_histo(\n",
    "    metadata=metadata,\n",
    "    group_col='individual_case_location'\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1571,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Load Faith PD\n",
    "faith_df = pd.read_csv(\n",
    "    \"../Analyses/Xebec/Output/results/alpha_div/phylo/faith_pd/vector.tsv\", sep=\"\\t\"\n",
    ")\n",
    "faith_df = faith_df.set_index(faith_df.columns[0])  # Set sample ID as index\n",
    "faith_df.columns = ['Faith_PD']  # Rename value column\n",
    "\n",
    "# Make sure metadata is indexed by sample ID\n",
    "metadata = metadata.set_index('#sample-id') if '#sample-id' in metadata.columns else metadata\n",
    "\n",
    "# Join Faith PD values to metadata\n",
    "metadata = metadata.join(faith_df, how='inner')\n",
    "\n",
    "# Ensure SCORAD and Faith PD are numeric\n",
    "metadata['o_scorad'] = pd.to_numeric(metadata['o_scorad'], errors='coerce')\n",
    "metadata['Faith_PD'] = pd.to_numeric(metadata['Faith_PD'], errors='coerce')\n",
    "\n",
    "# Drop samples with missing data\n",
    "plot_df = metadata.dropna(subset=['o_scorad', 'Faith_PD', 'area'])  # 'area' should be 'Cape Town' or 'Umtata'\n",
    "\n",
    "# Plot regression: SCORAD vs Faith PD, colored by region\n",
    "sns.set(style=\"whitegrid\")\n",
    "plot = sns.lmplot(\n",
    "    data=plot_df,\n",
    "    x='o_scorad',\n",
    "    y='Faith_PD',\n",
    "    hue='area',\n",
    "    palette={'Cape Town': '#C9A34F', 'Umtata': '#F0806B'},\n",
    "    height=5,\n",
    "    aspect=1.2,\n",
    "    scatter_kws={'alpha': 0.7, 's': 40},\n",
    "    line_kws={'linewidth': 2}\n",
    ")\n",
    "\n",
    "# Title and labels\n",
    "plt.title(\"Faith PD vs SCORAD by Region\", fontsize=16)\n",
    "plt.xlabel(\"SCORAD\", fontsize=14)\n",
    "plt.ylabel(\"Faith Phylogenetic Diversity\", fontsize=14)\n",
    "plt.tight_layout()\n",
    "\n",
    "# Save plot\n",
    "plt.savefig(\"../Plots/Analysis_figures/Diversity/FaithPD_vs_SCORAD_by_region.png\", dpi=600)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Beta Diversity by Individual AD Status and Region"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1572,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import Python packages\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from numpy import array\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from matplotlib.gridspec import GridSpec\n",
    "import scipy\n",
    "import scipy.stats as ss\n",
    "from skbio.stats.distance import permanova\n",
    "import biom\n",
    "from biom import load_table\n",
    "from biom.util import biom_open\n",
    "from gemelli.rpca import rpca\n",
    "\n",
    "### See RPCA standalone python tutorial: \n",
    "# https://github.com/biocore/gemelli/blob/master/ipynb/tutorials/RPCA-moving-pictures-standalone-cli-and-api.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1573,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # read in biom table\n",
    "# biom_tbl = biom.load_table(\"/Users/annanguyen/16S_AD_South-Africa/Data/Tables/Relative_Abundance_Tables/df_16S_filtered_feature_table_rare_Genus_relative_abundance.biom\")\n",
    "\n",
    "# print(biom_tbl.ids(axis= 'sample'))\n",
    "# len(biom_tbl.ids(axis= 'sample'))\n",
    "# type(biom_tbl.ids(axis = 'sample'))\n",
    "\n",
    "# Function to load BIOM table, collapse by taxa, sort rows by row sum, remove specified samples, and convert to relative abundance\n",
    "def load_biom_table(biom_path, metadata_path):\n",
    "    # Load BIOM table and convert to a DataFrame\n",
    "    table = biom.load_table(biom_path)\n",
    "    df = pd.DataFrame(table.matrix_data.toarray(),\n",
    "                      index=table.ids(axis='observation'),\n",
    "                      columns=table.ids(axis='sample'))\n",
    "    \n",
    "    # df = df.T\n",
    "    # Sort rows by row sum in descending order\n",
    "    df['row_sum'] = df.sum(axis=1)\n",
    "    df = df.sort_values(by='row_sum', ascending=False)\n",
    "    \n",
    "    # Drop the 'row_sum' column before proceeding\n",
    "    df = df.drop(columns=['row_sum'])\n",
    "\n",
    "    # Replace ' g__' rows with ' g__Unknown'\n",
    "    df.index = df.index.map(lambda x: ' g__Unknown' if x == ' g__' else x)\n",
    "\n",
    "    # Remove '15564.' prefix from columns\n",
    "    df.columns = df.columns.str.replace('15564.', '')\n",
    "\n",
    "     # Load metadata as a DataFrame from the file path\n",
    "    metadata = pd.read_csv(metadata_path, sep='\\t')\n",
    "    metadata['#sample-id'] = metadata['#sample-id'].str.replace('_', '')\n",
    "\n",
    "    # Set Sample-ID as the index for the metadata dataframe \n",
    "    metadata = metadata.set_index('#sample-id')\n",
    "\n",
    "    # Subset metadata to only contain the same samples as in the BIOM df\n",
    "    metadata_sub = metadata.loc[df.columns]\n",
    "\n",
    "    # Get a list of the sample ids that are just skin\n",
    "    metadataindex = metadata_sub[metadata_sub['case_type'].str.endswith('nares')].index.tolist()\n",
    "    \n",
    "    \n",
    "    # Drop all sample-ids that correspond to nares in the BIOM table DataFrame\n",
    "    df = df.drop(columns= metadataindex)\n",
    "\n",
    "    # Subsets the Metadata to only contain the sample ids that are in the newly subsetted biom table DataFrame with only Skin\n",
    "    metadata_sub = metadata.loc[df.columns]\n",
    "\n",
    "    \n",
    "    #returns BIOM table DataFrame and the metadata that has been subsetted to only contain skin\n",
    "    return df, metadata_sub\n",
    "\n",
    "\n",
    "# converts the subsetted skin only BIOM table DataFrame back into a BIOM to run RPCA\n",
    "def convert_df_to_biom(table, biom_output_file):\n",
    "    obs_ids = table.index\n",
    "    samp_ids = table.columns\n",
    "    biom_table = biom.table.Table(table.values, observation_ids=obs_ids, sample_ids=samp_ids) # Convert df back to a biom table\n",
    "    #biom_output_file = f\"../tables/{name}_subset.biom\" # Path to output file\n",
    "\n",
    "    with biom_open(biom_output_file, 'w') as f: # This does the actual saving!\n",
    "        biom_table.to_hdf5(f, generated_by=\"subsetted tables\")\n",
    "\n",
    "    return biom_table\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1574,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>900344</th>\n",
       "      <th>900221</th>\n",
       "      <th>900570</th>\n",
       "      <th>900129</th>\n",
       "      <th>900321</th>\n",
       "      <th>900091</th>\n",
       "      <th>900245</th>\n",
       "      <th>900423</th>\n",
       "      <th>900581</th>\n",
       "      <th>900145</th>\n",
       "      <th>...</th>\n",
       "      <th>900094</th>\n",
       "      <th>900287</th>\n",
       "      <th>900225</th>\n",
       "      <th>900057</th>\n",
       "      <th>900294</th>\n",
       "      <th>9003972</th>\n",
       "      <th>900097</th>\n",
       "      <th>900498</th>\n",
       "      <th>900276</th>\n",
       "      <th>900406</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GTGCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTTATTGGGCGTAAAGCGAGCGCAGGCGGTTAGATAAGTCTGAAGTTAAAGGCTG</th>\n",
       "      <td>984.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>389.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>198.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>57.0</td>\n",
       "      <td>44.0</td>\n",
       "      <td>111.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>1168.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTGCCAGCAGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAATTATTGGGCGTAAAGCGCGCGTAGGCGGTTTTTTAAGTCTGATGTGAAAGCCCA</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1294.0</td>\n",
       "      <td>182.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>296.0</td>\n",
       "      <td>159.0</td>\n",
       "      <td>49.0</td>\n",
       "      <td>1366.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>108.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>93.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>794.0</td>\n",
       "      <td>171.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTGCCAGCAGCCGCGGTAATACGGAGGGTGCGAGCGTTAATCGGAATAACTGGGCGTAAAGGGCACGCAGGCGGTTATTTAAGTGAGGTGTGAAAGCCCC</th>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTGCCAGCCGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAATTATTGGGCGTAAAGCGCGCGTAGGCGGTTTTTTAAGTCTGATGTGAAAGCCCA</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>236.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>57.0</td>\n",
       "      <td>799.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>116.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>477.0</td>\n",
       "      <td>94.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTGCCAGCCGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTTATTGGGCGTAAAGCGAGCGCAGGCGGTTAGATAAGTCTGAAGTTAAAGGCTG</th>\n",
       "      <td>611.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>593.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTGCCAGCAGCCGCGGTAATACGGAGGGTCCGAGCGTTATCCGGAATCATTGGGTTTAAAGGGTCCGTAGGCGGCTTTATAAGTCAGTGGTGAAATCCGG</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTGCCAGCAGCCGCGGTAATACGTAGGGTGCAAGCGTTGTCCGGAATTATTGGGCGTAAAGAGCTCGTAGGCGGTTTGTCGCGTCTGCTGTGAAATTCCG</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTGCCAGCCGCCGCGGTAATACGTAGGGCGCAAGCGTTGTCCGGAATTATTGGGCGTAAAGAGCTCGTAGGCGGTTTGTCGCGTCTGGTGTGAAAACTCG</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTGCCAGAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTTATTGGGCGTAAAGCGAGCGCAGGCGGTTAGATAAGTCTGAAGTTAAAGGCTGT</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTTATTGGGCGTAAAGCGAGCGCAGGCGGTTAGATAAGTCTGAAGTTAAAGGCTGT</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2283 rows × 305 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    900344  900221  900570  \\\n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATT...   984.0    22.0   389.0   \n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAAT...     0.0     0.0  1294.0   \n",
       "GTGCCAGCAGCCGCGGTAATACGGAGGGTGCGAGCGTTAATCGGAAT...     0.0     2.0     9.0   \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAAT...     0.0     0.0   236.0   \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATT...   611.0     0.0     0.0   \n",
       "...                                                    ...     ...     ...   \n",
       "GTGCCAGCAGCCGCGGTAATACGGAGGGTCCGAGCGTTATCCGGAAT...     0.0     0.0     0.0   \n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGGTGCAAGCGTTGTCCGGAAT...     0.0     0.0     0.0   \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGGCGCAAGCGTTGTCCGGAAT...     0.0     0.0     0.0   \n",
       "GTGCCAGAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTT...     0.0     0.0     0.0   \n",
       "GTCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTT...     0.0     0.0     0.0   \n",
       "\n",
       "                                                    900129  900321  900091  \\\n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATT...     0.0    26.0    27.0   \n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAAT...   182.0     9.0   296.0   \n",
       "GTGCCAGCAGCCGCGGTAATACGGAGGGTGCGAGCGTTAATCGGAAT...     0.0     0.0     0.0   \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAAT...     0.0     0.0     0.0   \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATT...     0.0     0.0    16.0   \n",
       "...                                                    ...     ...     ...   \n",
       "GTGCCAGCAGCCGCGGTAATACGGAGGGTCCGAGCGTTATCCGGAAT...     0.0     0.0     0.0   \n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGGTGCAAGCGTTGTCCGGAAT...     0.0     0.0     0.0   \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGGCGCAAGCGTTGTCCGGAAT...     0.0     0.0     0.0   \n",
       "GTGCCAGAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTT...     0.0     0.0     0.0   \n",
       "GTCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTT...     0.0     0.0     0.0   \n",
       "\n",
       "                                                    900245  900423  900581  \\\n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATT...    40.0     0.0   198.0   \n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAAT...   159.0    49.0  1366.0   \n",
       "GTGCCAGCAGCCGCGGTAATACGGAGGGTGCGAGCGTTAATCGGAAT...     0.0     0.0    22.0   \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAAT...     0.0    57.0   799.0   \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATT...     0.0     0.0     0.0   \n",
       "...                                                    ...     ...     ...   \n",
       "GTGCCAGCAGCCGCGGTAATACGGAGGGTCCGAGCGTTATCCGGAAT...     0.0     0.0     0.0   \n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGGTGCAAGCGTTGTCCGGAAT...     0.0     0.0     0.0   \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGGCGCAAGCGTTGTCCGGAAT...     0.0     0.0     0.0   \n",
       "GTGCCAGAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTT...     0.0     0.0     0.0   \n",
       "GTCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTT...     0.0     0.0     0.0   \n",
       "\n",
       "                                                    900145  ...  900094  \\\n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATT...     0.0  ...     0.0   \n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAAT...    27.0  ...     0.0   \n",
       "GTGCCAGCAGCCGCGGTAATACGGAGGGTGCGAGCGTTAATCGGAAT...     0.0  ...     0.0   \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAAT...    37.0  ...     0.0   \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATT...     6.0  ...     0.0   \n",
       "...                                                    ...  ...     ...   \n",
       "GTGCCAGCAGCCGCGGTAATACGGAGGGTCCGAGCGTTATCCGGAAT...     0.0  ...     0.0   \n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGGTGCAAGCGTTGTCCGGAAT...     0.0  ...     0.0   \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGGCGCAAGCGTTGTCCGGAAT...     0.0  ...     0.0   \n",
       "GTGCCAGAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTT...     0.0  ...     0.0   \n",
       "GTCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTT...     0.0  ...     0.0   \n",
       "\n",
       "                                                    900287  900225  900057  \\\n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATT...    57.0    44.0   111.0   \n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAAT...   108.0    30.0    93.0   \n",
       "GTGCCAGCAGCCGCGGTAATACGGAGGGTGCGAGCGTTAATCGGAAT...     2.0     0.0     0.0   \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAAT...     0.0     4.0   116.0   \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATT...     0.0     0.0     0.0   \n",
       "...                                                    ...     ...     ...   \n",
       "GTGCCAGCAGCCGCGGTAATACGGAGGGTCCGAGCGTTATCCGGAAT...     0.0     0.0     0.0   \n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGGTGCAAGCGTTGTCCGGAAT...     0.0     0.0     0.0   \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGGCGCAAGCGTTGTCCGGAAT...     0.0     0.0     0.0   \n",
       "GTGCCAGAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTT...     0.0     0.0     0.0   \n",
       "GTCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTT...     0.0     0.0     0.0   \n",
       "\n",
       "                                                    900294  9003972  900097  \\\n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATT...    12.0   1168.0    24.0   \n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAAT...    13.0    794.0   171.0   \n",
       "GTGCCAGCAGCCGCGGTAATACGGAGGGTGCGAGCGTTAATCGGAAT...     0.0     74.0     0.0   \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAAT...     0.0    477.0    94.0   \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATT...     7.0    593.0     0.0   \n",
       "...                                                    ...      ...     ...   \n",
       "GTGCCAGCAGCCGCGGTAATACGGAGGGTCCGAGCGTTATCCGGAAT...     0.0      0.0     0.0   \n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGGTGCAAGCGTTGTCCGGAAT...     0.0      0.0     0.0   \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGGCGCAAGCGTTGTCCGGAAT...     0.0      0.0     0.0   \n",
       "GTGCCAGAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTT...     0.0      0.0     0.0   \n",
       "GTCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTT...     0.0      0.0     0.0   \n",
       "\n",
       "                                                    900498  900276  900406  \n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATT...    15.0     0.0     0.0  \n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAAT...    28.0     0.0    12.0  \n",
       "GTGCCAGCAGCCGCGGTAATACGGAGGGTGCGAGCGTTAATCGGAAT...     0.0     0.0     0.0  \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAAT...    11.0     0.0     7.0  \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATT...    17.0     0.0     0.0  \n",
       "...                                                    ...     ...     ...  \n",
       "GTGCCAGCAGCCGCGGTAATACGGAGGGTCCGAGCGTTATCCGGAAT...     0.0     0.0     0.0  \n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGGTGCAAGCGTTGTCCGGAAT...     0.0     0.0     0.0  \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGGCGCAAGCGTTGTCCGGAAT...     0.0     0.0     0.0  \n",
       "GTGCCAGAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTT...     0.0     0.0     0.0  \n",
       "GTCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTT...     0.0     1.0     0.0  \n",
       "\n",
       "[2283 rows x 305 columns]"
      ]
     },
     "execution_count": 1574,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# specify the path to the BIOM table and Metadata\n",
    "# biom_path = \"../Data/Tables/Relative_Abundance_Tables/df_16S_filtered_feature_table_rare_Genus_relative_abundance.biom\"\n",
    "biom_path = \"../Data/Tables/Absolute_Abundance_Tables/209766_filtered_feature_table.biom\"\n",
    "metadata_path = '../Data/Metadata/updated_clean_ant_skin_metadata_microbiome_type-location.tab'\n",
    "skin_only_table, metadata_sub = load_biom_table(biom_path, metadata_path)\n",
    "\n",
    "skin_only_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1575,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "skin_only_biom_path = '../Data/Tables/Absolute_Abundance_Tables/209766_filtered_feature_table_skin_only.biom'\n"
     ]
    }
   ],
   "source": [
    "# Create the path to the skin BIOM file by using the prefix of the original BIOM file\n",
    "skin_only_biom_path = biom_path.removesuffix('.biom') + '_skin_only.biom'\n",
    "\n",
    "# Print the path that was created\n",
    "print(f'{skin_only_biom_path = }')\n",
    "\n",
    "# Converting df to BIOM table and saving the biom table to the path specified above\n",
    "biom_tbl = convert_df_to_biom(skin_only_table, skin_only_biom_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1576,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PC1</th>\n",
       "      <th>PC2</th>\n",
       "      <th>PC3</th>\n",
       "      <th>individual_case_location</th>\n",
       "      <th>o_scorad</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>900344</th>\n",
       "      <td>-0.031259</td>\n",
       "      <td>-0.030585</td>\n",
       "      <td>0.045341</td>\n",
       "      <td>H skin Umtata</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900221</th>\n",
       "      <td>0.034560</td>\n",
       "      <td>0.014311</td>\n",
       "      <td>0.101067</td>\n",
       "      <td>AD skin Umtata</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900570</th>\n",
       "      <td>0.051177</td>\n",
       "      <td>-0.101247</td>\n",
       "      <td>-0.020742</td>\n",
       "      <td>AD skin Cape Town</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900129</th>\n",
       "      <td>0.061599</td>\n",
       "      <td>-0.018120</td>\n",
       "      <td>-0.002876</td>\n",
       "      <td>AD skin Cape Town</td>\n",
       "      <td>44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900321</th>\n",
       "      <td>0.002420</td>\n",
       "      <td>0.003792</td>\n",
       "      <td>0.040013</td>\n",
       "      <td>H skin Umtata</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9003972</th>\n",
       "      <td>-0.107095</td>\n",
       "      <td>-0.089850</td>\n",
       "      <td>-0.019500</td>\n",
       "      <td>AD skin Umtata</td>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900097</th>\n",
       "      <td>0.025972</td>\n",
       "      <td>0.031322</td>\n",
       "      <td>-0.000608</td>\n",
       "      <td>AD skin Cape Town</td>\n",
       "      <td>44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900498</th>\n",
       "      <td>-0.033348</td>\n",
       "      <td>0.048009</td>\n",
       "      <td>0.030825</td>\n",
       "      <td>AD skin Umtata</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900276</th>\n",
       "      <td>-0.108307</td>\n",
       "      <td>-0.102284</td>\n",
       "      <td>0.005354</td>\n",
       "      <td>AD skin Umtata</td>\n",
       "      <td>52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900406</th>\n",
       "      <td>-0.041063</td>\n",
       "      <td>0.053825</td>\n",
       "      <td>0.033240</td>\n",
       "      <td>AD skin Umtata</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>305 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              PC1       PC2       PC3 individual_case_location o_scorad\n",
       "900344  -0.031259 -0.030585  0.045341            H skin Umtata      NaN\n",
       "900221   0.034560  0.014311  0.101067           AD skin Umtata       34\n",
       "900570   0.051177 -0.101247 -0.020742        AD skin Cape Town       36\n",
       "900129   0.061599 -0.018120 -0.002876        AD skin Cape Town       44\n",
       "900321   0.002420  0.003792  0.040013            H skin Umtata      NaN\n",
       "...           ...       ...       ...                      ...      ...\n",
       "9003972 -0.107095 -0.089850 -0.019500           AD skin Umtata       54\n",
       "900097   0.025972  0.031322 -0.000608        AD skin Cape Town       44\n",
       "900498  -0.033348  0.048009  0.030825           AD skin Umtata      NaN\n",
       "900276  -0.108307 -0.102284  0.005354           AD skin Umtata       52\n",
       "900406  -0.041063  0.053825  0.033240           AD skin Umtata       28\n",
       "\n",
       "[305 rows x 5 columns]"
      ]
     },
     "execution_count": 1576,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# perform RPCA with auto rank estimation\n",
    "np.seterr(divide = 'ignore')\n",
    "ordination, distance = rpca(biom_tbl)\n",
    "\n",
    "# extract and view sample ordinations from RPCA result\n",
    "spca_df = ordination.samples\n",
    "\n",
    "# Add a case type column into the spca_df using the meta_data and matching by indices by the join function\n",
    "spca_df = spca_df.join(metadata_sub['individual_case_location'])\n",
    "\n",
    "# Map o_scorad column from metadata to spca_df\n",
    "spca_df = spca_df.join(metadata_sub['o_scorad'])\n",
    "\n",
    "\n",
    "spca_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1577,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PC1</th>\n",
       "      <th>PC2</th>\n",
       "      <th>PC3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GTGCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTTATTGGGCGTAAAGCGAGCGCAGGCGGTTAGATAAGTCTGAAGTTAAAGGCTG</th>\n",
       "      <td>-0.176233</td>\n",
       "      <td>-0.363698</td>\n",
       "      <td>0.052031</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTGCCAGCAGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAATTATTGGGCGTAAAGCGCGCGTAGGCGGTTTTTTAAGTCTGATGTGAAAGCCCA</th>\n",
       "      <td>0.031293</td>\n",
       "      <td>-0.181160</td>\n",
       "      <td>-0.635457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTGCCAGCAGCCGCGGTAATACGGAGGGTGCGAGCGTTAATCGGAATAACTGGGCGTAAAGGGCACGCAGGCGGTTATTTAAGTGAGGTGTGAAAGCCCC</th>\n",
       "      <td>-0.137862</td>\n",
       "      <td>-0.125219</td>\n",
       "      <td>0.075513</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTGCCAGCCGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAATTATTGGGCGTAAAGCGCGCGTAGGCGGTTTTTTAAGTCTGATGTGAAAGCCCA</th>\n",
       "      <td>0.018601</td>\n",
       "      <td>-0.079648</td>\n",
       "      <td>-0.607929</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTGCCAGCCGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATTTATTGGGCGTAAAGCGAGCGCAGGCGGTTAGATAAGTCTGAAGTTAAAGGCTG</th>\n",
       "      <td>-0.200034</td>\n",
       "      <td>-0.269834</td>\n",
       "      <td>0.054126</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                         PC1       PC2  \\\n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATT... -0.176233 -0.363698   \n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAAT...  0.031293 -0.181160   \n",
       "GTGCCAGCAGCCGCGGTAATACGGAGGGTGCGAGCGTTAATCGGAAT... -0.137862 -0.125219   \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAAT...  0.018601 -0.079648   \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATT... -0.200034 -0.269834   \n",
       "\n",
       "                                                         PC3  \n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATT...  0.052031  \n",
       "GTGCCAGCAGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAAT... -0.635457  \n",
       "GTGCCAGCAGCCGCGGTAATACGGAGGGTGCGAGCGTTAATCGGAAT...  0.075513  \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGAAT... -0.607929  \n",
       "GTGCCAGCCGCCGCGGTAATACGTAGGTCCCGAGCGTTGTCCGGATT...  0.054126  "
      ]
     },
     "execution_count": 1577,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# extract and view feature ordinations from RPCA result\n",
    "fpca_df = ordination.features\n",
    "fpca_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1578,
   "metadata": {},
   "outputs": [],
   "source": [
    "def permanova_on_case_type_subset(df, dist_matrix, case_type_subset):\n",
    "    \"\"\"\n",
    "    Perform PERMANOVA on a subset of the data.\n",
    "    \n",
    "    Parameters:\n",
    "    - df: DataFrame with metadata, must include the grouping variable\n",
    "    - dist_matrix: DistanceMatrix object from scikit-bio\n",
    "    - case_type_subset: list of case_type groups to include in the test\n",
    "    \n",
    "    Returns:\n",
    "    - PERMANOVA result (dict-like with p-value, test statistic, etc.)\n",
    "    \"\"\"\n",
    "    # Subset the DataFrame\n",
    "    \n",
    "    subset_df = df[df['individual_case_location'].isin(case_type_subset)]\n",
    "    print(subset_df)\n",
    "    \n",
    "    # Get the matching IDs and subset distance matrix\n",
    "    ids = subset_df.index\n",
    "    print(ids)\n",
    "    sub_dm = dist_matrix.filter(ids, strict=False)\n",
    "    \n",
    "    # Run PERMANOVA\n",
    "    result = permanova(sub_dm, grouping=subset_df['individual_case_location'], permutations=999)\n",
    "    \n",
    "    return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1579,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "              PC1       PC2       PC3 individual_case_location o_scorad\n",
      "900344  -0.031259 -0.030585  0.045341            H skin Umtata      NaN\n",
      "900221   0.034560  0.014311  0.101067           AD skin Umtata       34\n",
      "900570   0.051177 -0.101247 -0.020742        AD skin Cape Town       36\n",
      "900129   0.061599 -0.018120 -0.002876        AD skin Cape Town       44\n",
      "900321   0.002420  0.003792  0.040013            H skin Umtata      NaN\n",
      "...           ...       ...       ...                      ...      ...\n",
      "9003972 -0.107095 -0.089850 -0.019500           AD skin Umtata       54\n",
      "900097   0.025972  0.031322 -0.000608        AD skin Cape Town       44\n",
      "900498  -0.033348  0.048009  0.030825           AD skin Umtata      NaN\n",
      "900276  -0.108307 -0.102284  0.005354           AD skin Umtata       52\n",
      "900406  -0.041063  0.053825  0.033240           AD skin Umtata       28\n",
      "\n",
      "[305 rows x 5 columns]\n",
      "Index(['900570', '900129', '900091', '900581', '900145', '900130', '900544',\n",
      "       '900110', '900105', '900600',\n",
      "       ...\n",
      "       '900604', '900063', '900619', 'Co005SNNL', 'Ca008HNL', '900081',\n",
      "       '900093', '900094', '900057', '900097'],\n",
      "      dtype='object', length=131)\n",
      "              PC1       PC2       PC3 individual_case_location o_scorad\n",
      "900344  -0.031259 -0.030585  0.045341            H skin Umtata      NaN\n",
      "900221   0.034560  0.014311  0.101067           AD skin Umtata       34\n",
      "900570   0.051177 -0.101247 -0.020742        AD skin Cape Town       36\n",
      "900129   0.061599 -0.018120 -0.002876        AD skin Cape Town       44\n",
      "900321   0.002420  0.003792  0.040013            H skin Umtata      NaN\n",
      "...           ...       ...       ...                      ...      ...\n",
      "9003972 -0.107095 -0.089850 -0.019500           AD skin Umtata       54\n",
      "900097   0.025972  0.031322 -0.000608        AD skin Cape Town       44\n",
      "900498  -0.033348  0.048009  0.030825           AD skin Umtata      NaN\n",
      "900276  -0.108307 -0.102284  0.005354           AD skin Umtata       52\n",
      "900406  -0.041063  0.053825  0.033240           AD skin Umtata       28\n",
      "\n",
      "[305 rows x 5 columns]\n",
      "Index(['900344', '900221', '900321', '900245', '900423', '900364', '900329',\n",
      "       '900460', '900378', '900293',\n",
      "       ...\n",
      "       '900263', '900501', '900279', '900287', '900225', '900294', '9003972',\n",
      "       '900498', '900276', '900406'],\n",
      "      dtype='object', length=174)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'Skin of Healthy vs. AD Children in Cape Town': {'p': '2.00e-02',\n",
       "  'f': '3.50'},\n",
       " 'Skin of Healthy vs. AD Children in Umtata': {'p': '1.00e-03', 'f': '25.50'}}"
      ]
     },
     "execution_count": 1579,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# calculate permanova F-statistic for all combinations between Healthy, case lesional and case non lesional\n",
    "\n",
    "case_type_subsets = [\n",
    "    [\"H skin Cape Town\", \"AD skin Cape Town\"],\n",
    "    [\"H skin Umtata\", \"AD skin Umtata\"]]\n",
    "\n",
    "perma_res = {}\n",
    "\n",
    "for i, case_type_subset in enumerate(case_type_subsets):\n",
    "    # print(\"Subset case_type:\", case_type_subset)\n",
    "    # print(\"Available sample IDs:\", spca_df.index.tolist())\n",
    "\n",
    "    result = permanova_on_case_type_subset(spca_df, distance, case_type_subset)\n",
    "    group_label = (\n",
    "        \"Skin of Healthy vs. AD Children in Cape Town\" if i == 0 else\n",
    "        \"Skin of Healthy vs. AD Children in Umtata\"\n",
    "    )\n",
    "    f_val = result[\"test statistic\"]\n",
    "    p_val = result[\"p-value\"]\n",
    "    \n",
    "    perma_res[group_label] = {\n",
    "        \"p\": f\"{p_val:.2e}\",\n",
    "        \"f\": f\"{f_val:.2f}\"\n",
    "    }\n",
    "\n",
    "perma_res\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1580,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PC1</th>\n",
       "      <th>PC2</th>\n",
       "      <th>PC3</th>\n",
       "      <th>individual_case_location</th>\n",
       "      <th>o_scorad</th>\n",
       "      <th>o_scorad_adj</th>\n",
       "      <th>severity_group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>900344</th>\n",
       "      <td>-0.031259</td>\n",
       "      <td>-0.030585</td>\n",
       "      <td>0.045341</td>\n",
       "      <td>H Child UM</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>0</td>\n",
       "      <td>mild/moderate</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900221</th>\n",
       "      <td>0.034560</td>\n",
       "      <td>0.014311</td>\n",
       "      <td>0.101067</td>\n",
       "      <td>AD Child UM</td>\n",
       "      <td>34</td>\n",
       "      <td>34</td>\n",
       "      <td>mild/moderate</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900570</th>\n",
       "      <td>0.051177</td>\n",
       "      <td>-0.101247</td>\n",
       "      <td>-0.020742</td>\n",
       "      <td>AD Child CT</td>\n",
       "      <td>36</td>\n",
       "      <td>36</td>\n",
       "      <td>mild/moderate</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900129</th>\n",
       "      <td>0.061599</td>\n",
       "      <td>-0.018120</td>\n",
       "      <td>-0.002876</td>\n",
       "      <td>AD Child CT</td>\n",
       "      <td>44</td>\n",
       "      <td>44</td>\n",
       "      <td>mild/moderate</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900321</th>\n",
       "      <td>0.002420</td>\n",
       "      <td>0.003792</td>\n",
       "      <td>0.040013</td>\n",
       "      <td>H Child UM</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>0</td>\n",
       "      <td>mild/moderate</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9003972</th>\n",
       "      <td>-0.107095</td>\n",
       "      <td>-0.089850</td>\n",
       "      <td>-0.019500</td>\n",
       "      <td>AD Child UM</td>\n",
       "      <td>54</td>\n",
       "      <td>54</td>\n",
       "      <td>severe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900097</th>\n",
       "      <td>0.025972</td>\n",
       "      <td>0.031322</td>\n",
       "      <td>-0.000608</td>\n",
       "      <td>AD Child CT</td>\n",
       "      <td>44</td>\n",
       "      <td>44</td>\n",
       "      <td>mild/moderate</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900498</th>\n",
       "      <td>-0.033348</td>\n",
       "      <td>0.048009</td>\n",
       "      <td>0.030825</td>\n",
       "      <td>AD Child UM</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>0</td>\n",
       "      <td>mild/moderate</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900276</th>\n",
       "      <td>-0.108307</td>\n",
       "      <td>-0.102284</td>\n",
       "      <td>0.005354</td>\n",
       "      <td>AD Child UM</td>\n",
       "      <td>52</td>\n",
       "      <td>52</td>\n",
       "      <td>severe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900406</th>\n",
       "      <td>-0.041063</td>\n",
       "      <td>0.053825</td>\n",
       "      <td>0.033240</td>\n",
       "      <td>AD Child UM</td>\n",
       "      <td>28</td>\n",
       "      <td>28</td>\n",
       "      <td>mild/moderate</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>305 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              PC1       PC2       PC3 individual_case_location  o_scorad  \\\n",
       "900344  -0.031259 -0.030585  0.045341               H Child UM      <NA>   \n",
       "900221   0.034560  0.014311  0.101067              AD Child UM        34   \n",
       "900570   0.051177 -0.101247 -0.020742              AD Child CT        36   \n",
       "900129   0.061599 -0.018120 -0.002876              AD Child CT        44   \n",
       "900321   0.002420  0.003792  0.040013               H Child UM      <NA>   \n",
       "...           ...       ...       ...                      ...       ...   \n",
       "9003972 -0.107095 -0.089850 -0.019500              AD Child UM        54   \n",
       "900097   0.025972  0.031322 -0.000608              AD Child CT        44   \n",
       "900498  -0.033348  0.048009  0.030825              AD Child UM      <NA>   \n",
       "900276  -0.108307 -0.102284  0.005354              AD Child UM        52   \n",
       "900406  -0.041063  0.053825  0.033240              AD Child UM        28   \n",
       "\n",
       "         o_scorad_adj severity_group  \n",
       "900344              0  mild/moderate  \n",
       "900221             34  mild/moderate  \n",
       "900570             36  mild/moderate  \n",
       "900129             44  mild/moderate  \n",
       "900321              0  mild/moderate  \n",
       "...               ...            ...  \n",
       "9003972            54         severe  \n",
       "900097             44  mild/moderate  \n",
       "900498              0  mild/moderate  \n",
       "900276             52         severe  \n",
       "900406             28  mild/moderate  \n",
       "\n",
       "[305 rows x 7 columns]"
      ]
     },
     "execution_count": 1580,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "spca_df['individual_case_location'] = spca_df['individual_case_location'].replace({\n",
    "    'H skin Cape Town': 'H Child CT',\n",
    "    'AD skin Cape Town': 'AD Child CT',\n",
    "    'H skin Umtata': 'H Child UM',\n",
    "    'AD skin Umtata': 'AD Child UM'\n",
    "})\n",
    "\n",
    "spca_df['o_scorad'] = pd.to_numeric(spca_df['o_scorad'], errors='coerce').astype('Int64')\n",
    "\n",
    "spca_df['o_scorad_adj'] = spca_df['o_scorad'].fillna(0)\n",
    "\n",
    "spca_df['severity_group'] = spca_df['o_scorad'].apply(lambda x: 'severe' if x > 50 else 'mild/moderate')\n",
    "\n",
    "spca_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1581,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Columns in spca_df: Index(['PC1', 'PC2', 'PC3', 'individual_case_location', 'o_scorad',\n",
      "       'o_scorad_adj', 'severity_group'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "print(\"Columns in spca_df:\", spca_df.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1582,
   "metadata": {},
   "outputs": [],
   "source": [
    "spca_df[\"simplified_label\"] = spca_df[\"individual_case_location\"].replace({\n",
    "    \"H Child CT\": \"H Child\",\n",
    "    \"AD Child CT\": \"AD Child\",\n",
    "    \"H Child UM\": \"H Child\",\n",
    "    \"AD Child UM\": \"AD Child\"\n",
    "})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1583,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set the color palette for the groups in the correct order\n",
    "palette = {\n",
    "    'H Child CT': '#A7C7E7',\n",
    "    'H Child UM': '#ADD8E6',\n",
    "    'AD Child CT': '#d2b48c',\n",
    "    'AD Child UM': '#fa8072'\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1584,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import numpy as np\n",
    "# import pandas as pd\n",
    "# import seaborn as sns\n",
    "# import matplotlib.pyplot as plt\n",
    "# from matplotlib.patches import Circle\n",
    "\n",
    "# # Create label map with (n=#)\n",
    "# label_map = spca_df[\"individual_case_location\"].value_counts().to_dict()\n",
    "# label_map = {k: f\"{k} (n={v})\" for k, v in label_map.items()}\n",
    "\n",
    "# # Add labeled group column\n",
    "# spca_df[\"group_label\"] = spca_df[\"individual_case_location\"].map(label_map)\n",
    "\n",
    "# # Create figure\n",
    "# fig, axes = plt.subplots(1, 2, figsize=(9, 5), sharex=True, sharey=True)\n",
    "\n",
    "# # Rename ordination axes\n",
    "# fpca_df.columns = [f\"PC{i+1}\" for i in range(fpca_df.shape[1])]\n",
    "\n",
    "# split_map = {\n",
    "#     \"Cape Town\": [\"H Child CT\", \"AD Child CT\"],\n",
    "#     \"Umtata\": [\"H Child UM\", \"AD Child UM\"]\n",
    "# }\n",
    "\n",
    "# for ax, (title, groups) in zip(axes, split_map.items()):\n",
    "#     subset_df = spca_df[spca_df[\"individual_case_location\"].isin(groups)].copy()\n",
    "#     subset_df[\"group_label\"] = subset_df[\"individual_case_location\"].map(label_map)\n",
    "    \n",
    "#     sns.scatterplot(\n",
    "#         data=subset_df,\n",
    "#         x=\"PC1\",\n",
    "#         y=\"PC2\",\n",
    "#         hue=\"group_label\",\n",
    "#         hue_order=[label_map[g] for g in groups],\n",
    "#         s=50,\n",
    "#         edgecolor=\"black\",\n",
    "#         linewidth=0.5,\n",
    "#         palette={label_map[g]: palette[g] for g in groups},\n",
    "#         ax=ax\n",
    "#     )\n",
    "    \n",
    "#     for case_type, case_type_df in subset_df.groupby(\"individual_case_location\"):\n",
    "#         color = palette[case_type]\n",
    "#         points = case_type_df[[\"PC1\", \"PC2\"]].values\n",
    "\n",
    "#         centroid = points.mean(axis=0)\n",
    "#         dists = np.linalg.norm(points - centroid, axis=1)\n",
    "#         radius = np.percentile(dists, 90)\n",
    "\n",
    "#         circle = Circle(centroid, radius, edgecolor=color, facecolor=color, alpha=0.2, lw=1, zorder=0)\n",
    "#         ax.add_patch(circle)\n",
    "\n",
    "#     ax.set_title(title, fontsize=16)\n",
    "#     ax.set_xlabel(\"\")\n",
    "#     ax.set_ylabel(\"\")\n",
    "#     ax.legend(frameon=False, fontsize=9, loc='upper right')\n",
    "\n",
    "# # Shared axis labels and ticks\n",
    "# pc1_pct, pc2_pct, _ = [f\"RPCA PC{i+1} ({x*100:.2f}%)\" for i, x in enumerate(ordination.proportion_explained)]\n",
    "# axes[0].set_ylabel(pc2_pct, fontsize=14)\n",
    "# for ax in axes:\n",
    "#     ax.set_xlabel(pc1_pct, fontsize=14)\n",
    "#     ax.set_xticks([-0.2, -0.1, 0.0, 0.1, 0.2, 0.3])\n",
    "#     ax.set_xticklabels([-0.2, -0.1, 0.0, 0.1, 0.2, 0.3], fontsize=12)\n",
    "#     ax.set_yticks([-0.2, -0.1, 0.0, 0.1, 0.2, 0.3])\n",
    "#     ax.set_yticklabels([-0.2, -0.1, 0.0, 0.1, 0.2, 0.3], fontsize=12)\n",
    "\n",
    "# # PERMANOVA stats\n",
    "# axes[0].text(\n",
    "#     -0.2, 0.27,\n",
    "#     f\"p={float(perma_res['Skin of Healthy vs. AD Children in Cape Town']['p']):.3f}, \"\n",
    "#     f\"F={float(perma_res['Skin of Healthy vs. AD Children in Cape Town']['f']):.2f}\",\n",
    "#     fontsize=12\n",
    "# )\n",
    "# axes[1].text(\n",
    "#     -0.2, 0.27,\n",
    "#     f\"p={float(perma_res['Skin of Healthy vs. AD Children in Umtata']['p']):.3f}, \"\n",
    "#     f\"F={float(perma_res['Skin of Healthy vs. AD Children in Umtata']['f']):.2f}\",\n",
    "#     fontsize=12\n",
    "# )\n",
    "\n",
    "# # Final touches\n",
    "# fig.suptitle(\"Beta Diversity by Region\", fontsize=20, y=0.92)\n",
    "# plt.tight_layout(rect=[0, 0, 1, 0.95])\n",
    "# plt.savefig(\"../Plots/Analysis_figures/Diversity/16S_Beta_Diversity_RPCA_skin_by-location_split.png\", dpi=600)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1585,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PC1</th>\n",
       "      <th>PC2</th>\n",
       "      <th>PC3</th>\n",
       "      <th>individual_case_location</th>\n",
       "      <th>o_scorad</th>\n",
       "      <th>o_scorad_adj</th>\n",
       "      <th>severity_group</th>\n",
       "      <th>simplified_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>900344</th>\n",
       "      <td>-0.031259</td>\n",
       "      <td>-0.030585</td>\n",
       "      <td>0.045341</td>\n",
       "      <td>H Child UM</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>0</td>\n",
       "      <td>mild/moderate</td>\n",
       "      <td>H Child</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900221</th>\n",
       "      <td>0.034560</td>\n",
       "      <td>0.014311</td>\n",
       "      <td>0.101067</td>\n",
       "      <td>AD Child UM</td>\n",
       "      <td>34</td>\n",
       "      <td>34</td>\n",
       "      <td>mild/moderate</td>\n",
       "      <td>AD Child</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900570</th>\n",
       "      <td>0.051177</td>\n",
       "      <td>-0.101247</td>\n",
       "      <td>-0.020742</td>\n",
       "      <td>AD Child CT</td>\n",
       "      <td>36</td>\n",
       "      <td>36</td>\n",
       "      <td>mild/moderate</td>\n",
       "      <td>AD Child</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900129</th>\n",
       "      <td>0.061599</td>\n",
       "      <td>-0.018120</td>\n",
       "      <td>-0.002876</td>\n",
       "      <td>AD Child CT</td>\n",
       "      <td>44</td>\n",
       "      <td>44</td>\n",
       "      <td>mild/moderate</td>\n",
       "      <td>AD Child</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900321</th>\n",
       "      <td>0.002420</td>\n",
       "      <td>0.003792</td>\n",
       "      <td>0.040013</td>\n",
       "      <td>H Child UM</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>0</td>\n",
       "      <td>mild/moderate</td>\n",
       "      <td>H Child</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9003972</th>\n",
       "      <td>-0.107095</td>\n",
       "      <td>-0.089850</td>\n",
       "      <td>-0.019500</td>\n",
       "      <td>AD Child UM</td>\n",
       "      <td>54</td>\n",
       "      <td>54</td>\n",
       "      <td>severe</td>\n",
       "      <td>AD Child</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900097</th>\n",
       "      <td>0.025972</td>\n",
       "      <td>0.031322</td>\n",
       "      <td>-0.000608</td>\n",
       "      <td>AD Child CT</td>\n",
       "      <td>44</td>\n",
       "      <td>44</td>\n",
       "      <td>mild/moderate</td>\n",
       "      <td>AD Child</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900498</th>\n",
       "      <td>-0.033348</td>\n",
       "      <td>0.048009</td>\n",
       "      <td>0.030825</td>\n",
       "      <td>AD Child UM</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>0</td>\n",
       "      <td>mild/moderate</td>\n",
       "      <td>AD Child</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900276</th>\n",
       "      <td>-0.108307</td>\n",
       "      <td>-0.102284</td>\n",
       "      <td>0.005354</td>\n",
       "      <td>AD Child UM</td>\n",
       "      <td>52</td>\n",
       "      <td>52</td>\n",
       "      <td>severe</td>\n",
       "      <td>AD Child</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900406</th>\n",
       "      <td>-0.041063</td>\n",
       "      <td>0.053825</td>\n",
       "      <td>0.033240</td>\n",
       "      <td>AD Child UM</td>\n",
       "      <td>28</td>\n",
       "      <td>28</td>\n",
       "      <td>mild/moderate</td>\n",
       "      <td>AD Child</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>305 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              PC1       PC2       PC3 individual_case_location  o_scorad  \\\n",
       "900344  -0.031259 -0.030585  0.045341               H Child UM      <NA>   \n",
       "900221   0.034560  0.014311  0.101067              AD Child UM        34   \n",
       "900570   0.051177 -0.101247 -0.020742              AD Child CT        36   \n",
       "900129   0.061599 -0.018120 -0.002876              AD Child CT        44   \n",
       "900321   0.002420  0.003792  0.040013               H Child UM      <NA>   \n",
       "...           ...       ...       ...                      ...       ...   \n",
       "9003972 -0.107095 -0.089850 -0.019500              AD Child UM        54   \n",
       "900097   0.025972  0.031322 -0.000608              AD Child CT        44   \n",
       "900498  -0.033348  0.048009  0.030825              AD Child UM      <NA>   \n",
       "900276  -0.108307 -0.102284  0.005354              AD Child UM        52   \n",
       "900406  -0.041063  0.053825  0.033240              AD Child UM        28   \n",
       "\n",
       "         o_scorad_adj severity_group simplified_label  \n",
       "900344              0  mild/moderate          H Child  \n",
       "900221             34  mild/moderate         AD Child  \n",
       "900570             36  mild/moderate         AD Child  \n",
       "900129             44  mild/moderate         AD Child  \n",
       "900321              0  mild/moderate          H Child  \n",
       "...               ...            ...              ...  \n",
       "9003972            54         severe         AD Child  \n",
       "900097             44  mild/moderate         AD Child  \n",
       "900498              0  mild/moderate         AD Child  \n",
       "900276             52         severe         AD Child  \n",
       "900406             28  mild/moderate         AD Child  \n",
       "\n",
       "[305 rows x 8 columns]"
      ]
     },
     "execution_count": 1585,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "spca_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1586,
   "metadata": {},
   "outputs": [],
   "source": [
    "spca_df['o_scorad_adj'] = spca_df['o_scorad_adj'].astype(float)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1587,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dtype('float64')"
      ]
     },
     "execution_count": 1587,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "spca_df['PC1'].dtype\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1588,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dtype('float64')"
      ]
     },
     "execution_count": 1588,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "spca_df['PC2'].dtype\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1589,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dtype('float64')"
      ]
     },
     "execution_count": 1589,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "spca_df['o_scorad_adj'].dtype\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1590,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PC1</th>\n",
       "      <th>PC2</th>\n",
       "      <th>PC3</th>\n",
       "      <th>individual_case_location</th>\n",
       "      <th>o_scorad</th>\n",
       "      <th>o_scorad_adj</th>\n",
       "      <th>severity_group</th>\n",
       "      <th>simplified_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>900344</th>\n",
       "      <td>-0.031259</td>\n",
       "      <td>-0.030585</td>\n",
       "      <td>0.045341</td>\n",
       "      <td>H Child UM</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>0.0</td>\n",
       "      <td>mild/moderate</td>\n",
       "      <td>H Child</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900221</th>\n",
       "      <td>0.034560</td>\n",
       "      <td>0.014311</td>\n",
       "      <td>0.101067</td>\n",
       "      <td>AD Child UM</td>\n",
       "      <td>34</td>\n",
       "      <td>34.0</td>\n",
       "      <td>mild/moderate</td>\n",
       "      <td>AD Child</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900570</th>\n",
       "      <td>0.051177</td>\n",
       "      <td>-0.101247</td>\n",
       "      <td>-0.020742</td>\n",
       "      <td>AD Child CT</td>\n",
       "      <td>36</td>\n",
       "      <td>36.0</td>\n",
       "      <td>mild/moderate</td>\n",
       "      <td>AD Child</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900129</th>\n",
       "      <td>0.061599</td>\n",
       "      <td>-0.018120</td>\n",
       "      <td>-0.002876</td>\n",
       "      <td>AD Child CT</td>\n",
       "      <td>44</td>\n",
       "      <td>44.0</td>\n",
       "      <td>mild/moderate</td>\n",
       "      <td>AD Child</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900321</th>\n",
       "      <td>0.002420</td>\n",
       "      <td>0.003792</td>\n",
       "      <td>0.040013</td>\n",
       "      <td>H Child UM</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>0.0</td>\n",
       "      <td>mild/moderate</td>\n",
       "      <td>H Child</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9003972</th>\n",
       "      <td>-0.107095</td>\n",
       "      <td>-0.089850</td>\n",
       "      <td>-0.019500</td>\n",
       "      <td>AD Child UM</td>\n",
       "      <td>54</td>\n",
       "      <td>54.0</td>\n",
       "      <td>severe</td>\n",
       "      <td>AD Child</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900097</th>\n",
       "      <td>0.025972</td>\n",
       "      <td>0.031322</td>\n",
       "      <td>-0.000608</td>\n",
       "      <td>AD Child CT</td>\n",
       "      <td>44</td>\n",
       "      <td>44.0</td>\n",
       "      <td>mild/moderate</td>\n",
       "      <td>AD Child</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900498</th>\n",
       "      <td>-0.033348</td>\n",
       "      <td>0.048009</td>\n",
       "      <td>0.030825</td>\n",
       "      <td>AD Child UM</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>0.0</td>\n",
       "      <td>mild/moderate</td>\n",
       "      <td>AD Child</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900276</th>\n",
       "      <td>-0.108307</td>\n",
       "      <td>-0.102284</td>\n",
       "      <td>0.005354</td>\n",
       "      <td>AD Child UM</td>\n",
       "      <td>52</td>\n",
       "      <td>52.0</td>\n",
       "      <td>severe</td>\n",
       "      <td>AD Child</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900406</th>\n",
       "      <td>-0.041063</td>\n",
       "      <td>0.053825</td>\n",
       "      <td>0.033240</td>\n",
       "      <td>AD Child UM</td>\n",
       "      <td>28</td>\n",
       "      <td>28.0</td>\n",
       "      <td>mild/moderate</td>\n",
       "      <td>AD Child</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>305 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              PC1       PC2       PC3 individual_case_location  o_scorad  \\\n",
       "900344  -0.031259 -0.030585  0.045341               H Child UM      <NA>   \n",
       "900221   0.034560  0.014311  0.101067              AD Child UM        34   \n",
       "900570   0.051177 -0.101247 -0.020742              AD Child CT        36   \n",
       "900129   0.061599 -0.018120 -0.002876              AD Child CT        44   \n",
       "900321   0.002420  0.003792  0.040013               H Child UM      <NA>   \n",
       "...           ...       ...       ...                      ...       ...   \n",
       "9003972 -0.107095 -0.089850 -0.019500              AD Child UM        54   \n",
       "900097   0.025972  0.031322 -0.000608              AD Child CT        44   \n",
       "900498  -0.033348  0.048009  0.030825              AD Child UM      <NA>   \n",
       "900276  -0.108307 -0.102284  0.005354              AD Child UM        52   \n",
       "900406  -0.041063  0.053825  0.033240              AD Child UM        28   \n",
       "\n",
       "         o_scorad_adj severity_group simplified_label  \n",
       "900344            0.0  mild/moderate          H Child  \n",
       "900221           34.0  mild/moderate         AD Child  \n",
       "900570           36.0  mild/moderate         AD Child  \n",
       "900129           44.0  mild/moderate         AD Child  \n",
       "900321            0.0  mild/moderate          H Child  \n",
       "...               ...            ...              ...  \n",
       "9003972          54.0         severe         AD Child  \n",
       "900097           44.0  mild/moderate         AD Child  \n",
       "900498            0.0  mild/moderate         AD Child  \n",
       "900276           52.0         severe         AD Child  \n",
       "900406           28.0  mild/moderate         AD Child  \n",
       "\n",
       "[305 rows x 8 columns]"
      ]
     },
     "execution_count": 1590,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "spca_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1591,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Rename ordination axes\n",
    "# fpca_df.columns = [f\"PC{i+1}\" for i in range(fpca_df.shape[1])]\n",
    "\n",
    "# # Merge ordination PCs into spca_df\n",
    "# spca_df[['PC1', 'PC2']] = fpca_df[['PC1', 'PC2']]\n",
    "# # spca_df = spca_df.dropna(subset=['PC1', 'PC2', 'o_scorad_adj']).copy()\n",
    "# spca_df\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1592,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Residualize PC1 and PC2\n",
    "X = sm.add_constant(spca_df['o_scorad_adj'])\n",
    "model_pc1 = sm.OLS(spca_df['PC1'], X).fit()\n",
    "model_pc2 = sm.OLS(spca_df['PC2'], X).fit()\n",
    "spca_df['PC1_resid'] = model_pc1.resid\n",
    "spca_df['PC2_resid'] = model_pc2.resid\n",
    "\n",
    "# Create label map with (n=#)\n",
    "label_map = spca_df[\"individual_case_location\"].value_counts().to_dict()\n",
    "label_map = {k: f\"{k} (n={v})\" for k, v in label_map.items()}\n",
    "\n",
    "# Add labeled group column\n",
    "spca_df[\"group_label\"] = spca_df[\"individual_case_location\"].map(label_map)\n",
    "\n",
    "# Create figure\n",
    "fig, axes = plt.subplots(1, 2, figsize=(8, 5.5), sharex=True, sharey=True)\n",
    "\n",
    "# Define groups per panel\n",
    "split_map = {\n",
    "    \"Cape Town\": [\"H Child CT\", \"AD Child CT\"],\n",
    "    \"Umtata\": [\"H Child UM\", \"AD Child UM\"]\n",
    "}\n",
    "\n",
    "\n",
    "# Plot loop\n",
    "for ax, (title, groups) in zip(axes, split_map.items()):\n",
    "    subset_df = spca_df[spca_df[\"individual_case_location\"].isin(groups)].copy()\n",
    "    subset_df[\"group_label\"] = subset_df[\"individual_case_location\"].map(label_map)\n",
    "\n",
    "    ax.grid(False)\n",
    "\n",
    "    sns.scatterplot(\n",
    "        data=subset_df,\n",
    "        x=\"PC1_resid\",\n",
    "        y=\"PC2_resid\",\n",
    "        hue=\"group_label\",\n",
    "        hue_order=[label_map[g] for g in groups],\n",
    "        s=50,\n",
    "        edgecolor=\"black\",\n",
    "        linewidth=0.5,\n",
    "        palette={label_map[g]: palette[g] for g in groups},\n",
    "        ax=ax\n",
    "    )\n",
    "\n",
    "    # Draw 90% confidence circles\n",
    "    for case_type, case_type_df in subset_df.groupby(\"individual_case_location\"):\n",
    "        color = palette[case_type]\n",
    "        points = case_type_df[[\"PC1_resid\", \"PC2_resid\"]].values\n",
    "        centroid = points.mean(axis=0)\n",
    "        dists = np.linalg.norm(points - centroid, axis=1)\n",
    "        radius = np.percentile(dists, 90)\n",
    "        circle = Circle(centroid, radius, edgecolor=color, facecolor=color, alpha=0.2, lw=1, zorder=0)\n",
    "        ax.add_patch(circle)\n",
    "\n",
    "    ax.set_title(title, fontsize=16)\n",
    "    ax.set_xlabel(\"\")\n",
    "    ax.set_ylabel(\"\")\n",
    "    ax.legend(frameon=False, fontsize=9, loc='upper right')\n",
    "\n",
    "# Axis labels (now residualized)\n",
    "axes[0].set_ylabel(\"Residual PC2 (adj. for SCORAD)\", fontsize=14)\n",
    "for ax in axes:\n",
    "    ax.set_xlabel(\"Residual PC1 (adj. for SCORAD)\", fontsize=14)\n",
    "    ax.set_xticks([-0.2, -0.1, 0.0, 0.1, 0.2, 0.3])\n",
    "    ax.set_xticklabels([-0.2, -0.1, 0.0, 0.1, 0.2, 0.3], fontsize=12)\n",
    "    ax.set_yticks([-0.2, -0.1, 0.0, 0.1, 0.2, 0.3])\n",
    "    ax.set_yticklabels([-0.2, -0.1, 0.0, 0.1, 0.2, 0.3], fontsize=12)\n",
    "\n",
    "# PERMANOVA stats\n",
    "axes[0].text(\n",
    "    -0.2, 0.27,\n",
    "    f\"p={float(perma_res['Skin of Healthy vs. AD Children in Cape Town']['p']):.3f}, \"\n",
    "    f\"F={float(perma_res['Skin of Healthy vs. AD Children in Cape Town']['f']):.2f}\",\n",
    "    fontsize=12\n",
    ")\n",
    "axes[1].text(\n",
    "    -0.2, 0.27,\n",
    "    f\"p={float(perma_res['Skin of Healthy vs. AD Children in Umtata']['p']):.3f}, \"\n",
    "    f\"F={float(perma_res['Skin of Healthy vs. AD Children in Umtata']['f']):.2f}\",\n",
    "    fontsize=12\n",
    ")\n",
    "\n",
    "# Final touches\n",
    "fig.suptitle(\"SCORAD-Adjusted Beta Diversity\", fontsize=18, y=0.93)\n",
    "plt.tight_layout(rect=[0, 0, 1, 0.95])\n",
    "plt.savefig(\"../Plots/Analysis_figures/Diversity/16S_Beta_Diversity_RPCA_skin_by-location_split_residualized.png\", dpi=600)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "qiime2-metagenome-2024.10",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
