{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get phlylum mapping for itol"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import Python packages\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import biom\n",
    "from biom import load_table\n",
    "import numpy as np\n",
    "from scipy.stats import pearsonr\n",
    "from matplotlib_venn import venn2\n",
    "import matplotlib.patches as mpatches\n",
    "from scipy.stats import mannwhitneyu\n",
    "from biom import Table\n",
    "from gemelli.rpca import rpca\n",
    "from scipy.spatial.distance import euclidean\n",
    "import re\n",
    "import matplotlib.colors as mcolors\n",
    "import qiime2 as q2\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Taxon</th>\n",
       "      <th>Confidence</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Feature ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GB-GCA-003568775.1-MWMI01000008.1</th>\n",
       "      <td>d__Bacteria; p__; c__; o__; f__; g__; s__</td>\n",
       "      <td>0.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GB-GCA-001552015.1-CP010514.1</th>\n",
       "      <td>d__Bacteria; p__; c__; o__; f__; g__; s__</td>\n",
       "      <td>0.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GB-GCA-000008085.1-AE017199.1</th>\n",
       "      <td>d__Bacteria; p__; c__; o__; f__; g__; s__</td>\n",
       "      <td>0.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TAGCCGCACCCCAAGTGGTAGTCATTATTATTGGGCTTAAAGTGTTCGTAGCCGGGCCTGAAAGTCCGCTGTGAAATCCAAGCGCTCAAC</th>\n",
       "      <td>d__Bacteria; p__; c__; o__; f__; g__; s__</td>\n",
       "      <td>0.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TACCCGCGCCACGAGTGGTGATCGCGATTATTGGGCCTAAAGGGTTCGTAGCCGGTTTGGCAAGTTCCTGGTGAAATCTTTCAGCTAACTGAAAGGCGTG</th>\n",
       "      <td>d__Bacteria; p__; c__; o__; f__; g__; s__</td>\n",
       "      <td>0.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TACCCGCGCTGCGAGTGGTCACCACGATTATTGGGCTTAGAGCGTTCGTAGCCGGCTTTGCAAGTCCCCGGTGAAATCATCTGGCAAACC</th>\n",
       "      <td>d__Archaea; p__; c__; o__; f__; g__; s__</td>\n",
       "      <td>0.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TACCCGCGCCACGAGTGGTGATCGCGATTATTGGGCCTAAAGGGTTCGTAGCCGGCTCGGCAAGTTCCTGGTGAAATCTTCCAGCTAACTGGAAGGCGTG</th>\n",
       "      <td>d__Archaea; p__; c__; o__; f__; g__; s__</td>\n",
       "      <td>0.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CACTGGCAGTTCGGGTGGCAGTCGGTTCTATTGAGCCTAAAGCGTCCGTAGCCGGTTTGATCAGTCCTCGGTGAAATCTTTGGGCCTAACTCAAAGGCTT</th>\n",
       "      <td>d__Archaea; p__; c__; o__; f__; g__; s__</td>\n",
       "      <td>0.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TACCCGCGCTGCAAGTGGTCACCACGATTATTGGGCTTAGAGCGTTCGTAGCCGGTTTTGGTAGTCCTCGGTGAAATCGCCTGGTTGACC</th>\n",
       "      <td>d__Archaea; p__; c__; o__; f__; g__; s__</td>\n",
       "      <td>0.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TACAGAGGAGGCCAGCGTTGTCCGGACTTACTGGGCGTAAAGGGTCCGCAGGTGGTCGGGCAAGCAGGGTGTGGAATCCGCCGGCTCAACTGGCGGTCGC</th>\n",
       "      <td>d__Archaea; p__; c__; o__; f__; g__; s__</td>\n",
       "      <td>0.3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21074441 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                        Taxon  \\\n",
       "Feature ID                                                                                      \n",
       "GB-GCA-003568775.1-MWMI01000008.1                   d__Bacteria; p__; c__; o__; f__; g__; s__   \n",
       "GB-GCA-001552015.1-CP010514.1                       d__Bacteria; p__; c__; o__; f__; g__; s__   \n",
       "GB-GCA-000008085.1-AE017199.1                       d__Bacteria; p__; c__; o__; f__; g__; s__   \n",
       "TAGCCGCACCCCAAGTGGTAGTCATTATTATTGGGCTTAAAGTGTTC...  d__Bacteria; p__; c__; o__; f__; g__; s__   \n",
       "TACCCGCGCCACGAGTGGTGATCGCGATTATTGGGCCTAAAGGGTTC...  d__Bacteria; p__; c__; o__; f__; g__; s__   \n",
       "...                                                                                       ...   \n",
       "TACCCGCGCTGCGAGTGGTCACCACGATTATTGGGCTTAGAGCGTTC...   d__Archaea; p__; c__; o__; f__; g__; s__   \n",
       "TACCCGCGCCACGAGTGGTGATCGCGATTATTGGGCCTAAAGGGTTC...   d__Archaea; p__; c__; o__; f__; g__; s__   \n",
       "CACTGGCAGTTCGGGTGGCAGTCGGTTCTATTGAGCCTAAAGCGTCC...   d__Archaea; p__; c__; o__; f__; g__; s__   \n",
       "TACCCGCGCTGCAAGTGGTCACCACGATTATTGGGCTTAGAGCGTTC...   d__Archaea; p__; c__; o__; f__; g__; s__   \n",
       "TACAGAGGAGGCCAGCGTTGTCCGGACTTACTGGGCGTAAAGGGTCC...   d__Archaea; p__; c__; o__; f__; g__; s__   \n",
       "\n",
       "                                                   Confidence  \n",
       "Feature ID                                                     \n",
       "GB-GCA-003568775.1-MWMI01000008.1                         0.3  \n",
       "GB-GCA-001552015.1-CP010514.1                             0.3  \n",
       "GB-GCA-000008085.1-AE017199.1                             0.3  \n",
       "TAGCCGCACCCCAAGTGGTAGTCATTATTATTGGGCTTAAAGTGTTC...        0.3  \n",
       "TACCCGCGCCACGAGTGGTGATCGCGATTATTGGGCCTAAAGGGTTC...        0.3  \n",
       "...                                                       ...  \n",
       "TACCCGCGCTGCGAGTGGTCACCACGATTATTGGGCTTAGAGCGTTC...        0.3  \n",
       "TACCCGCGCCACGAGTGGTGATCGCGATTATTGGGCCTAAAGGGTTC...        0.3  \n",
       "CACTGGCAGTTCGGGTGGCAGTCGGTTCTATTGAGCCTAAAGCGTCC...        0.3  \n",
       "TACCCGCGCTGCAAGTGGTCACCACGATTATTGGGCTTAGAGCGTTC...        0.3  \n",
       "TACAGAGGAGGCCAGCGTTGTCCGGACTTACTGGGCGTAAAGGGTCC...        0.3  \n",
       "\n",
       "[21074441 rows x 2 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gg_taxonomy = q2.Artifact.load('../Reference/2022.10.taxonomy.asv.tsv.qza').view(pd.DataFrame)\n",
    "gg_taxonomy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Taxon</th>\n",
       "      <th>Confidence</th>\n",
       "      <th>Kingdom</th>\n",
       "      <th>Phylum</th>\n",
       "      <th>Class</th>\n",
       "      <th>Order</th>\n",
       "      <th>Family</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Species</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Feature ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GB-GCA-003568775.1-MWMI01000008.1</th>\n",
       "      <td>d__Bacteria; p__; c__; o__; f__; g__; s__</td>\n",
       "      <td>0.3</td>\n",
       "      <td>d__Bacteria</td>\n",
       "      <td>p__</td>\n",
       "      <td>c__</td>\n",
       "      <td>o__</td>\n",
       "      <td>f__</td>\n",
       "      <td>g__</td>\n",
       "      <td>s__</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GB-GCA-001552015.1-CP010514.1</th>\n",
       "      <td>d__Bacteria; p__; c__; o__; f__; g__; s__</td>\n",
       "      <td>0.3</td>\n",
       "      <td>d__Bacteria</td>\n",
       "      <td>p__</td>\n",
       "      <td>c__</td>\n",
       "      <td>o__</td>\n",
       "      <td>f__</td>\n",
       "      <td>g__</td>\n",
       "      <td>s__</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GB-GCA-000008085.1-AE017199.1</th>\n",
       "      <td>d__Bacteria; p__; c__; o__; f__; g__; s__</td>\n",
       "      <td>0.3</td>\n",
       "      <td>d__Bacteria</td>\n",
       "      <td>p__</td>\n",
       "      <td>c__</td>\n",
       "      <td>o__</td>\n",
       "      <td>f__</td>\n",
       "      <td>g__</td>\n",
       "      <td>s__</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TAGCCGCACCCCAAGTGGTAGTCATTATTATTGGGCTTAAAGTGTTCGTAGCCGGGCCTGAAAGTCCGCTGTGAAATCCAAGCGCTCAAC</th>\n",
       "      <td>d__Bacteria; p__; c__; o__; f__; g__; s__</td>\n",
       "      <td>0.3</td>\n",
       "      <td>d__Bacteria</td>\n",
       "      <td>p__</td>\n",
       "      <td>c__</td>\n",
       "      <td>o__</td>\n",
       "      <td>f__</td>\n",
       "      <td>g__</td>\n",
       "      <td>s__</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TACCCGCGCCACGAGTGGTGATCGCGATTATTGGGCCTAAAGGGTTCGTAGCCGGTTTGGCAAGTTCCTGGTGAAATCTTTCAGCTAACTGAAAGGCGTG</th>\n",
       "      <td>d__Bacteria; p__; c__; o__; f__; g__; s__</td>\n",
       "      <td>0.3</td>\n",
       "      <td>d__Bacteria</td>\n",
       "      <td>p__</td>\n",
       "      <td>c__</td>\n",
       "      <td>o__</td>\n",
       "      <td>f__</td>\n",
       "      <td>g__</td>\n",
       "      <td>s__</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TACCCGCGCTGCGAGTGGTCACCACGATTATTGGGCTTAGAGCGTTCGTAGCCGGCTTTGCAAGTCCCCGGTGAAATCATCTGGCAAACC</th>\n",
       "      <td>d__Archaea; p__; c__; o__; f__; g__; s__</td>\n",
       "      <td>0.3</td>\n",
       "      <td>d__Archaea</td>\n",
       "      <td>p__</td>\n",
       "      <td>c__</td>\n",
       "      <td>o__</td>\n",
       "      <td>f__</td>\n",
       "      <td>g__</td>\n",
       "      <td>s__</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TACCCGCGCCACGAGTGGTGATCGCGATTATTGGGCCTAAAGGGTTCGTAGCCGGCTCGGCAAGTTCCTGGTGAAATCTTCCAGCTAACTGGAAGGCGTG</th>\n",
       "      <td>d__Archaea; p__; c__; o__; f__; g__; s__</td>\n",
       "      <td>0.3</td>\n",
       "      <td>d__Archaea</td>\n",
       "      <td>p__</td>\n",
       "      <td>c__</td>\n",
       "      <td>o__</td>\n",
       "      <td>f__</td>\n",
       "      <td>g__</td>\n",
       "      <td>s__</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CACTGGCAGTTCGGGTGGCAGTCGGTTCTATTGAGCCTAAAGCGTCCGTAGCCGGTTTGATCAGTCCTCGGTGAAATCTTTGGGCCTAACTCAAAGGCTT</th>\n",
       "      <td>d__Archaea; p__; c__; o__; f__; g__; s__</td>\n",
       "      <td>0.3</td>\n",
       "      <td>d__Archaea</td>\n",
       "      <td>p__</td>\n",
       "      <td>c__</td>\n",
       "      <td>o__</td>\n",
       "      <td>f__</td>\n",
       "      <td>g__</td>\n",
       "      <td>s__</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TACCCGCGCTGCAAGTGGTCACCACGATTATTGGGCTTAGAGCGTTCGTAGCCGGTTTTGGTAGTCCTCGGTGAAATCGCCTGGTTGACC</th>\n",
       "      <td>d__Archaea; p__; c__; o__; f__; g__; s__</td>\n",
       "      <td>0.3</td>\n",
       "      <td>d__Archaea</td>\n",
       "      <td>p__</td>\n",
       "      <td>c__</td>\n",
       "      <td>o__</td>\n",
       "      <td>f__</td>\n",
       "      <td>g__</td>\n",
       "      <td>s__</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TACAGAGGAGGCCAGCGTTGTCCGGACTTACTGGGCGTAAAGGGTCCGCAGGTGGTCGGGCAAGCAGGGTGTGGAATCCGCCGGCTCAACTGGCGGTCGC</th>\n",
       "      <td>d__Archaea; p__; c__; o__; f__; g__; s__</td>\n",
       "      <td>0.3</td>\n",
       "      <td>d__Archaea</td>\n",
       "      <td>p__</td>\n",
       "      <td>c__</td>\n",
       "      <td>o__</td>\n",
       "      <td>f__</td>\n",
       "      <td>g__</td>\n",
       "      <td>s__</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21074441 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                        Taxon  \\\n",
       "Feature ID                                                                                      \n",
       "GB-GCA-003568775.1-MWMI01000008.1                   d__Bacteria; p__; c__; o__; f__; g__; s__   \n",
       "GB-GCA-001552015.1-CP010514.1                       d__Bacteria; p__; c__; o__; f__; g__; s__   \n",
       "GB-GCA-000008085.1-AE017199.1                       d__Bacteria; p__; c__; o__; f__; g__; s__   \n",
       "TAGCCGCACCCCAAGTGGTAGTCATTATTATTGGGCTTAAAGTGTTC...  d__Bacteria; p__; c__; o__; f__; g__; s__   \n",
       "TACCCGCGCCACGAGTGGTGATCGCGATTATTGGGCCTAAAGGGTTC...  d__Bacteria; p__; c__; o__; f__; g__; s__   \n",
       "...                                                                                       ...   \n",
       "TACCCGCGCTGCGAGTGGTCACCACGATTATTGGGCTTAGAGCGTTC...   d__Archaea; p__; c__; o__; f__; g__; s__   \n",
       "TACCCGCGCCACGAGTGGTGATCGCGATTATTGGGCCTAAAGGGTTC...   d__Archaea; p__; c__; o__; f__; g__; s__   \n",
       "CACTGGCAGTTCGGGTGGCAGTCGGTTCTATTGAGCCTAAAGCGTCC...   d__Archaea; p__; c__; o__; f__; g__; s__   \n",
       "TACCCGCGCTGCAAGTGGTCACCACGATTATTGGGCTTAGAGCGTTC...   d__Archaea; p__; c__; o__; f__; g__; s__   \n",
       "TACAGAGGAGGCCAGCGTTGTCCGGACTTACTGGGCGTAAAGGGTCC...   d__Archaea; p__; c__; o__; f__; g__; s__   \n",
       "\n",
       "                                                   Confidence      Kingdom  \\\n",
       "Feature ID                                                                   \n",
       "GB-GCA-003568775.1-MWMI01000008.1                         0.3  d__Bacteria   \n",
       "GB-GCA-001552015.1-CP010514.1                             0.3  d__Bacteria   \n",
       "GB-GCA-000008085.1-AE017199.1                             0.3  d__Bacteria   \n",
       "TAGCCGCACCCCAAGTGGTAGTCATTATTATTGGGCTTAAAGTGTTC...        0.3  d__Bacteria   \n",
       "TACCCGCGCCACGAGTGGTGATCGCGATTATTGGGCCTAAAGGGTTC...        0.3  d__Bacteria   \n",
       "...                                                       ...          ...   \n",
       "TACCCGCGCTGCGAGTGGTCACCACGATTATTGGGCTTAGAGCGTTC...        0.3   d__Archaea   \n",
       "TACCCGCGCCACGAGTGGTGATCGCGATTATTGGGCCTAAAGGGTTC...        0.3   d__Archaea   \n",
       "CACTGGCAGTTCGGGTGGCAGTCGGTTCTATTGAGCCTAAAGCGTCC...        0.3   d__Archaea   \n",
       "TACCCGCGCTGCAAGTGGTCACCACGATTATTGGGCTTAGAGCGTTC...        0.3   d__Archaea   \n",
       "TACAGAGGAGGCCAGCGTTGTCCGGACTTACTGGGCGTAAAGGGTCC...        0.3   d__Archaea   \n",
       "\n",
       "                                                   Phylum Class Order Family  \\\n",
       "Feature ID                                                                     \n",
       "GB-GCA-003568775.1-MWMI01000008.1                     p__   c__   o__    f__   \n",
       "GB-GCA-001552015.1-CP010514.1                         p__   c__   o__    f__   \n",
       "GB-GCA-000008085.1-AE017199.1                         p__   c__   o__    f__   \n",
       "TAGCCGCACCCCAAGTGGTAGTCATTATTATTGGGCTTAAAGTGTTC...    p__   c__   o__    f__   \n",
       "TACCCGCGCCACGAGTGGTGATCGCGATTATTGGGCCTAAAGGGTTC...    p__   c__   o__    f__   \n",
       "...                                                   ...   ...   ...    ...   \n",
       "TACCCGCGCTGCGAGTGGTCACCACGATTATTGGGCTTAGAGCGTTC...    p__   c__   o__    f__   \n",
       "TACCCGCGCCACGAGTGGTGATCGCGATTATTGGGCCTAAAGGGTTC...    p__   c__   o__    f__   \n",
       "CACTGGCAGTTCGGGTGGCAGTCGGTTCTATTGAGCCTAAAGCGTCC...    p__   c__   o__    f__   \n",
       "TACCCGCGCTGCAAGTGGTCACCACGATTATTGGGCTTAGAGCGTTC...    p__   c__   o__    f__   \n",
       "TACAGAGGAGGCCAGCGTTGTCCGGACTTACTGGGCGTAAAGGGTCC...    p__   c__   o__    f__   \n",
       "\n",
       "                                                   Genus Species  \n",
       "Feature ID                                                        \n",
       "GB-GCA-003568775.1-MWMI01000008.1                    g__     s__  \n",
       "GB-GCA-001552015.1-CP010514.1                        g__     s__  \n",
       "GB-GCA-000008085.1-AE017199.1                        g__     s__  \n",
       "TAGCCGCACCCCAAGTGGTAGTCATTATTATTGGGCTTAAAGTGTTC...   g__     s__  \n",
       "TACCCGCGCCACGAGTGGTGATCGCGATTATTGGGCCTAAAGGGTTC...   g__     s__  \n",
       "...                                                  ...     ...  \n",
       "TACCCGCGCTGCGAGTGGTCACCACGATTATTGGGCTTAGAGCGTTC...   g__     s__  \n",
       "TACCCGCGCCACGAGTGGTGATCGCGATTATTGGGCCTAAAGGGTTC...   g__     s__  \n",
       "CACTGGCAGTTCGGGTGGCAGTCGGTTCTATTGAGCCTAAAGCGTCC...   g__     s__  \n",
       "TACCCGCGCTGCAAGTGGTCACCACGATTATTGGGCTTAGAGCGTTC...   g__     s__  \n",
       "TACAGAGGAGGCCAGCGTTGTCCGGACTTACTGGGCGTAAAGGGTCC...   g__     s__  \n",
       "\n",
       "[21074441 rows x 9 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Split taxonomy into levels\n",
    "taxonomy_levels = ['Kingdom', 'Phylum', 'Class', 'Order', 'Family', 'Genus', 'Species']\n",
    "gg_taxonomy[taxonomy_levels] = gg_taxonomy['Taxon'].str.split(';', expand=True)\n",
    "gg_taxonomy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: '../Data/Trees/itol_H_asv_group_colors_edit.txt'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[24], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m# txt_file = pd.read_csv('../Data/Trees/itol_AD_asv_group_colors_edit.txt', sep='\\t', header=None)\u001b[39;00m\n\u001b[0;32m----> 2\u001b[0m txt_file \u001b[38;5;241m=\u001b[39m \u001b[43mpd\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread_csv\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m../Data/Trees/itol_H_asv_group_colors_edit.txt\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msep\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;130;43;01m\\t\u001b[39;49;00m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mheader\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\n\u001b[1;32m      4\u001b[0m txt_file\n",
      "File \u001b[0;32m~/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/pandas/io/parsers/readers.py:1026\u001b[0m, in \u001b[0;36mread_csv\u001b[0;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, date_format, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options, dtype_backend)\u001b[0m\n\u001b[1;32m   1013\u001b[0m kwds_defaults \u001b[38;5;241m=\u001b[39m _refine_defaults_read(\n\u001b[1;32m   1014\u001b[0m     dialect,\n\u001b[1;32m   1015\u001b[0m     delimiter,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1022\u001b[0m     dtype_backend\u001b[38;5;241m=\u001b[39mdtype_backend,\n\u001b[1;32m   1023\u001b[0m )\n\u001b[1;32m   1024\u001b[0m kwds\u001b[38;5;241m.\u001b[39mupdate(kwds_defaults)\n\u001b[0;32m-> 1026\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_read\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/pandas/io/parsers/readers.py:620\u001b[0m, in \u001b[0;36m_read\u001b[0;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[1;32m    617\u001b[0m _validate_names(kwds\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnames\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m))\n\u001b[1;32m    619\u001b[0m \u001b[38;5;66;03m# Create the parser.\u001b[39;00m\n\u001b[0;32m--> 620\u001b[0m parser \u001b[38;5;241m=\u001b[39m \u001b[43mTextFileReader\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    622\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m chunksize \u001b[38;5;129;01mor\u001b[39;00m iterator:\n\u001b[1;32m    623\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m parser\n",
      "File \u001b[0;32m~/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/pandas/io/parsers/readers.py:1620\u001b[0m, in \u001b[0;36mTextFileReader.__init__\u001b[0;34m(self, f, engine, **kwds)\u001b[0m\n\u001b[1;32m   1617\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39moptions[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhas_index_names\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m kwds[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhas_index_names\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n\u001b[1;32m   1619\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles: IOHandles \u001b[38;5;241m|\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m-> 1620\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_engine \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_make_engine\u001b[49m\u001b[43m(\u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mengine\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/pandas/io/parsers/readers.py:1880\u001b[0m, in \u001b[0;36mTextFileReader._make_engine\u001b[0;34m(self, f, engine)\u001b[0m\n\u001b[1;32m   1878\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m mode:\n\u001b[1;32m   1879\u001b[0m         mode \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m-> 1880\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles \u001b[38;5;241m=\u001b[39m \u001b[43mget_handle\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1881\u001b[0m \u001b[43m    \u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1882\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1883\u001b[0m \u001b[43m    \u001b[49m\u001b[43mencoding\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mencoding\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1884\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcompression\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mcompression\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1885\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmemory_map\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mmemory_map\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1886\u001b[0m \u001b[43m    \u001b[49m\u001b[43mis_text\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mis_text\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1887\u001b[0m \u001b[43m    \u001b[49m\u001b[43merrors\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mencoding_errors\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mstrict\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1888\u001b[0m \u001b[43m    \u001b[49m\u001b[43mstorage_options\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mstorage_options\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1889\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1890\u001b[0m \u001b[38;5;28;01massert\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m   1891\u001b[0m f \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles\u001b[38;5;241m.\u001b[39mhandle\n",
      "File \u001b[0;32m~/miniforge3/envs/qiime2-metagenome-2024.10/lib/python3.10/site-packages/pandas/io/common.py:873\u001b[0m, in \u001b[0;36mget_handle\u001b[0;34m(path_or_buf, mode, encoding, compression, memory_map, is_text, errors, storage_options)\u001b[0m\n\u001b[1;32m    868\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(handle, \u001b[38;5;28mstr\u001b[39m):\n\u001b[1;32m    869\u001b[0m     \u001b[38;5;66;03m# Check whether the filename is to be opened in binary mode.\u001b[39;00m\n\u001b[1;32m    870\u001b[0m     \u001b[38;5;66;03m# Binary mode does not support 'encoding' and 'newline'.\u001b[39;00m\n\u001b[1;32m    871\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m ioargs\u001b[38;5;241m.\u001b[39mencoding \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m ioargs\u001b[38;5;241m.\u001b[39mmode:\n\u001b[1;32m    872\u001b[0m         \u001b[38;5;66;03m# Encoding\u001b[39;00m\n\u001b[0;32m--> 873\u001b[0m         handle \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mopen\u001b[39;49m\u001b[43m(\u001b[49m\n\u001b[1;32m    874\u001b[0m \u001b[43m            \u001b[49m\u001b[43mhandle\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    875\u001b[0m \u001b[43m            \u001b[49m\u001b[43mioargs\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    876\u001b[0m \u001b[43m            \u001b[49m\u001b[43mencoding\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mioargs\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mencoding\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    877\u001b[0m \u001b[43m            \u001b[49m\u001b[43merrors\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43merrors\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    878\u001b[0m \u001b[43m            \u001b[49m\u001b[43mnewline\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    879\u001b[0m \u001b[43m        \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    880\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    881\u001b[0m         \u001b[38;5;66;03m# Binary mode\u001b[39;00m\n\u001b[1;32m    882\u001b[0m         handle \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mopen\u001b[39m(handle, ioargs\u001b[38;5;241m.\u001b[39mmode)\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: '../Data/Trees/itol_H_asv_group_colors_edit.txt'"
     ]
    }
   ],
   "source": [
    "# txt_file = pd.read_csv('../Data/Trees/itol_AD_asv_group_colors_edit.txt', sep='\\t', header=None)\n",
    "txt_file = pd.read_csv('../Data/Trees/itol_H_asv_group_colors_edit.txt', sep='\\t', header=None)\n",
    "\n",
    "txt_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>g__Acinetobacter</td>\n",
       "      <td>#947fff</td>\n",
       "      <td>AD_skin_only</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>g__Leptotrichia_A_993758</td>\n",
       "      <td>#947fff</td>\n",
       "      <td>AD_skin_only</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>g__Alistipes_A_871400</td>\n",
       "      <td>#947fff</td>\n",
       "      <td>AD_skin_only</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>g__Sphingomonas_L_486704</td>\n",
       "      <td>#947fff</td>\n",
       "      <td>AD_skin_only</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>g__Clostridium_T</td>\n",
       "      <td>#947fff</td>\n",
       "      <td>AD_skin_only</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>784</th>\n",
       "      <td>g__Streptococcus</td>\n",
       "      <td>#e7b9d8</td>\n",
       "      <td>AD_shared</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>785</th>\n",
       "      <td>g__</td>\n",
       "      <td>#e7b9d8</td>\n",
       "      <td>AD_shared</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>786</th>\n",
       "      <td>g__Berryella</td>\n",
       "      <td>#e7b9d8</td>\n",
       "      <td>AD_shared</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>787</th>\n",
       "      <td>g__Prevotella</td>\n",
       "      <td>#e7b9d8</td>\n",
       "      <td>AD_shared</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>788</th>\n",
       "      <td>g__SIO2C1</td>\n",
       "      <td>#e7b9d8</td>\n",
       "      <td>AD_shared</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>789 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            0        1             2\n",
       "0            g__Acinetobacter  #947fff  AD_skin_only\n",
       "1    g__Leptotrichia_A_993758  #947fff  AD_skin_only\n",
       "2       g__Alistipes_A_871400  #947fff  AD_skin_only\n",
       "3    g__Sphingomonas_L_486704  #947fff  AD_skin_only\n",
       "4            g__Clostridium_T  #947fff  AD_skin_only\n",
       "..                        ...      ...           ...\n",
       "784          g__Streptococcus  #e7b9d8     AD_shared\n",
       "785                       g__  #e7b9d8     AD_shared\n",
       "786              g__Berryella  #e7b9d8     AD_shared\n",
       "787             g__Prevotella  #e7b9d8     AD_shared\n",
       "788                 g__SIO2C1  #e7b9d8     AD_shared\n",
       "\n",
       "[789 rows x 3 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Extract taxonomy names by removing _ASV-* suffix\n",
    "txt_file[0] = txt_file[0].str.replace(r'_ASV-\\d+$', '', regex=True)\n",
    "txt_file\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>g__Acinetobacter</td>\n",
       "      <td>#947fff</td>\n",
       "      <td>p__Proteobacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>g__Leptotrichia_A_993758</td>\n",
       "      <td>#947fff</td>\n",
       "      <td>p__Fusobacteriota</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>g__Alistipes_A_871400</td>\n",
       "      <td>#947fff</td>\n",
       "      <td>p__Bacteroidota</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>g__Sphingomonas_L_486704</td>\n",
       "      <td>#947fff</td>\n",
       "      <td>p__Proteobacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>g__Clostridium_T</td>\n",
       "      <td>#947fff</td>\n",
       "      <td>p__Firmicutes_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>784</th>\n",
       "      <td>g__Streptococcus</td>\n",
       "      <td>#e7b9d8</td>\n",
       "      <td>p__Firmicutes_D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>785</th>\n",
       "      <td>g__</td>\n",
       "      <td>#e7b9d8</td>\n",
       "      <td>p__</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>786</th>\n",
       "      <td>g__Berryella</td>\n",
       "      <td>#e7b9d8</td>\n",
       "      <td>p__Actinobacteriota</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>787</th>\n",
       "      <td>g__Prevotella</td>\n",
       "      <td>#e7b9d8</td>\n",
       "      <td>p__Bacteroidota</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>788</th>\n",
       "      <td>g__SIO2C1</td>\n",
       "      <td>#e7b9d8</td>\n",
       "      <td>p__Cyanobacteria</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>789 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            0        1                    2\n",
       "0            g__Acinetobacter  #947fff    p__Proteobacteria\n",
       "1    g__Leptotrichia_A_993758  #947fff    p__Fusobacteriota\n",
       "2       g__Alistipes_A_871400  #947fff      p__Bacteroidota\n",
       "3    g__Sphingomonas_L_486704  #947fff    p__Proteobacteria\n",
       "4            g__Clostridium_T  #947fff      p__Firmicutes_A\n",
       "..                        ...      ...                  ...\n",
       "784          g__Streptococcus  #e7b9d8      p__Firmicutes_D\n",
       "785                       g__  #e7b9d8                  p__\n",
       "786              g__Berryella  #e7b9d8  p__Actinobacteriota\n",
       "787             g__Prevotella  #e7b9d8      p__Bacteroidota\n",
       "788                 g__SIO2C1  #e7b9d8     p__Cyanobacteria\n",
       "\n",
       "[789 rows x 3 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create a mapping from genus to phylum\n",
    "genus_to_phylum = {}\n",
    "for idx, row in gg_taxonomy.iterrows():\n",
    "    genus = row['Genus'].strip()\n",
    "    phylum = row['Phylum'].strip()\n",
    "    if genus and phylum:  # Only add if both values exist\n",
    "        genus_to_phylum[genus] = phylum\n",
    "\n",
    "# Map genera to phyla and update txt_file\n",
    "for idx, genus in enumerate(txt_file[0]):\n",
    "    if genus in genus_to_phylum:\n",
    "        txt_file.at[idx, 2] = genus_to_phylum[genus]\n",
    "\n",
    "txt_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "qiime2-metagenome-2024.10",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
