{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 541,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import Python packages\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import biom\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from itertools import cycle\n",
    "import os\n",
    "from matplotlib.colors import ListedColormap\n",
    "from matplotlib.colors import to_rgba\n",
    "import itertools\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 542,
   "metadata": {},
   "outputs": [],
   "source": [
    "# A list of unique colors to use for taxa not predefined\n",
    "unique_colors = sns.color_palette(\"deep\", n_colors=20).as_hex()\n",
    "unique_color_iter = cycle(unique_colors)  # Iterator to cycle through unique colors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 543,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_biom_table(biom_path, metadata_path):\n",
    "    \"\"\"\n",
    "    Load a BIOM table and corresponding metadata. Filter to skin samples only,\n",
    "    align data, and return a relative abundance table and metadata DataFrame.\n",
    "\n",
    "    Returns:\n",
    "    - df_rel_abund: taxa x sample DataFrame (relative abundance, taxa as rows)\n",
    "    - metadata_filtered: metadata DataFrame (indexed by sample ID)\n",
    "    \"\"\"\n",
    "\n",
    "    # Load metadata\n",
    "    metadata = pd.read_csv(metadata_path, sep='\\t')\n",
    "    metadata = metadata.set_index('#sample-id')\n",
    "\n",
    "    # Load BIOM table\n",
    "    table = biom.load_table(biom_path)\n",
    "    df = pd.DataFrame(table.matrix_data.toarray(),\n",
    "                      index=table.ids(axis='observation'),\n",
    "                      columns=table.ids(axis='sample'))\n",
    "    \n",
    "    df.columns = df.columns.str.replace('15564.', '')\n",
    "\n",
    "    df = df.T  # samples as rows\n",
    "\n",
    "    # Clean taxon labels\n",
    "    df.columns = df.columns.map(lambda x: 'g__Unknown' if x.strip() == 'g__' else x.strip())\n",
    "\n",
    "    # Join metadata\n",
    "    df = df.join(metadata[['area', 'case_type']], how='left')\n",
    "\n",
    "    # Filter for skin samples only\n",
    "    df = df[df['case_type'].str.endswith('skin', na=False)]\n",
    "\n",
    "    # Separate metadata from abundance data\n",
    "    metadata_filtered = df[['area', 'case_type']]\n",
    "    df = df.drop(columns=['area', 'case_type'])\n",
    "\n",
    "    # Sort rows by total abundance\n",
    "    df['row_sum'] = df.sum(axis=1)\n",
    "    df = df.sort_values(by='row_sum', ascending=False).drop(columns=['row_sum'])\n",
    "\n",
    "    # Convert to relative abundance\n",
    "    df_rel_abund = df.div(df.sum(axis=1), axis=0)\n",
    "\n",
    "    return df_rel_abund.T, metadata_filtered\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 544,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to determine the top 15 families and collapse the rest as \"Others\"\n",
    "def collapse_top_15(df):\n",
    "    top_genera = df.sum(axis=1).nlargest(15).index  # Select top 15 families\n",
    "    df_top = df.loc[top_genera]\n",
    "    df_top.loc['Others'] = df.loc[~df.index.isin(top_genera)].sum()\n",
    "    return df_top"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 545,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_taxa_colors(taxa_list, global_taxa_color_map, taxa_colors=None, unique_color_iter=None):\n",
    "    \"\"\"\n",
    "    Assign colors to each taxon in the list, using a global color map.\n",
    "\n",
    "    Parameters:\n",
    "    - taxa_list: List of taxa names to color.\n",
    "    - global_taxa_color_map: Dictionary to store and reuse assigned colors.\n",
    "    - taxa_colors: (Optional) predefined color dictionary for known taxa.\n",
    "    - unique_color_iter: (Optional) iterator for generating new unique colors.\n",
    "\n",
    "    Returns:\n",
    "    - Updated global_taxa_color_map with all taxa in taxa_list assigned a color.\n",
    "    \"\"\"\n",
    "    for taxa in taxa_list:\n",
    "        if taxa not in global_taxa_color_map:\n",
    "            if taxa_colors and taxa in taxa_colors:\n",
    "                global_taxa_color_map[taxa] = taxa_colors[taxa]\n",
    "            elif unique_color_iter:\n",
    "                global_taxa_color_map[taxa] = next(unique_color_iter)\n",
    "            else:\n",
    "                global_taxa_color_map[taxa] = '#cccccc'  # fallback color\n",
    "    return global_taxa_color_map\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 546,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_relative_abundance_two_panels(df, metadata, output_dir, taxa_color_map):\n",
    "    \"\"\"\n",
    "    Create a two-panel stacked bar plot of relative abundances by region (Cape Town and Umtata),\n",
    "    with sample counts (n=) in x-tick labels.\n",
    "    \"\"\"\n",
    "\n",
    "    # Make sure sample order matches between df and metadata\n",
    "    # metadata = metadata.loc[df.index]\n",
    "    # print(df.index)\n",
    "    # print(metadata.index)\n",
    "    \n",
    "    shared_samples = df.index.intersection(metadata.index)\n",
    "    # print(shared_samples)\n",
    "\n",
    "    df = df.loc[shared_samples]\n",
    "    metadata = metadata.loc[shared_samples]\n",
    "\n",
    "    # Ensure no missing values\n",
    "    if metadata[['case_type', 'area']].isnull().any().any():\n",
    "        raise ValueError(\"Missing values in 'area' or group column in metadata.\")\n",
    "\n",
    "    # Create composite label for grouping\n",
    "    metadata['site_group'] = metadata['area'] + ' | ' + metadata['case_type']\n",
    "    print(metadata['site_group'].value_counts())\n",
    "    \n",
    "    # Count number of samples per group for labeling\n",
    "    group_counts = metadata['site_group'].value_counts().to_dict()\n",
    "    print(group_counts)\n",
    "\n",
    "    # Group and average by site_group\n",
    "    df_grouped = df.groupby(metadata['site_group']).mean().T\n",
    "    # print(df_grouped)\n",
    "\n",
    "    # Define plotting order\n",
    "    ct_order = ['Cape Town | control-nonlesional skin', 'Cape Town | case-nonlesional skin', 'Cape Town | case-lesional skin']\n",
    "    um_order = ['Umtata | control-nonlesional skin', 'Umtata | case-nonlesional skin', 'Umtata | case-lesional skin']\n",
    "\n",
    "    missing = [col for col in ct_order + um_order if col not in df_grouped.columns]\n",
    "    if missing:\n",
    "        raise KeyError(f\"Missing expected site_group columns in df_grouped: {missing}\")\n",
    "\n",
    "    df_ct = df_grouped[ct_order]\n",
    "    df_um = df_grouped[um_order]\n",
    "\n",
    "    # Create output directory\n",
    "    os.makedirs(output_dir, exist_ok=True)\n",
    "    output_png = os.path.join(output_dir, 'Fig_3A.png')\n",
    "\n",
    "    # Make figure wider to accommodate legend\n",
    "    fig, (ax1, ax2) = plt.subplots(2, 1, figsize=(7.5, 8.5), sharex=False)\n",
    "    # fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(16, 6), sharex=False)\n",
    "\n",
    "    # Define x-axis labels with n counts\n",
    "    def get_labels(order_list):\n",
    "        short_labels = ['H', 'ADNL', 'ADL']\n",
    "        return [f'{short} (n={group_counts.get(group, 0)})' for short, group in zip(short_labels, order_list)]\n",
    "\n",
    "    ct_labels = get_labels(ct_order)\n",
    "    um_labels = get_labels(um_order)\n",
    "\n",
    "    # Plot Cape Town\n",
    "    df_ct.T.plot(kind='bar', stacked=True, ax=ax1, width=0.6,\n",
    "                 color=[taxa_color_map.get(taxon, '#ADD8E6') for taxon in df_ct.index], legend=False)\n",
    "    ax1.set_title('Cape Town', fontsize=18)\n",
    "    ax1.set_ylabel('Relative Abundance', fontsize=14)\n",
    "    ax1.set_xlim(-0.4, 2.4)\n",
    "    ax1.set_xticks(range(3))\n",
    "    ax1.set_xlabel(' ')\n",
    "    ax1.set_xticklabels(ct_labels, rotation=0, fontsize=14)\n",
    "\n",
    "    # Plot Umtata\n",
    "    df_um.T.plot(kind='bar', stacked=True, ax=ax2, width=0.6,\n",
    "                 color=[taxa_color_map.get(taxon, '#ADD8E6') for taxon in df_um.index], legend=False)\n",
    "    ax2.set_title('Umtata', fontsize=18)\n",
    "    ax2.set_ylabel('Relative Abundance', fontsize=14)\n",
    "    ax2.set_xlim(-0.4, 2.4)\n",
    "    ax2.set_xticks(range(3))\n",
    "    ax2.set_xticklabels(um_labels, rotation=0, fontsize=14)\n",
    "\n",
    "    # Shared legend\n",
    "    handles, labels = ax1.get_legend_handles_labels()\n",
    "    fig.legend(\n",
    "        handles, labels,\n",
    "        loc='center left',\n",
    "        bbox_to_anchor=(0.65, 0.5),\n",
    "        fontsize=12,\n",
    "        title_fontsize=14,\n",
    "        frameon=True\n",
    "    )\n",
    "\n",
    "    fig.suptitle('Relative Abundance Skin', fontsize=20, y=1.05)\n",
    "    plt.subplots_adjust(left=0.15, right=0.65, top=0.93, bottom=0.08)\n",
    "    plt.xlabel('')\n",
    "\n",
    "    plt.savefig(output_png, dpi=600, bbox_inches='tight')\n",
    "    plt.close()\n",
    "\n",
    "    print(f\"sFigure saved to: {output_png}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 547,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>g__Streptococcus</th>\n",
       "      <th>g__Staphylococcus</th>\n",
       "      <th>g__Corynebacterium</th>\n",
       "      <th>g__Acinetobacter</th>\n",
       "      <th>g__Prevotella</th>\n",
       "      <th>g__Micrococcus</th>\n",
       "      <th>g__Haemophilus_D_734546</th>\n",
       "      <th>g__Veillonella_A</th>\n",
       "      <th>g__Psychrobacter</th>\n",
       "      <th>g__Neisseria_563205</th>\n",
       "      <th>g__SIO2C1</th>\n",
       "      <th>g__Cutibacterium</th>\n",
       "      <th>g__Chryseobacterium_796614</th>\n",
       "      <th>g__Pseudomonas_E_647464</th>\n",
       "      <th>Others</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>900360</th>\n",
       "      <td>0.128946</td>\n",
       "      <td>0.069174</td>\n",
       "      <td>0.069510</td>\n",
       "      <td>0.046004</td>\n",
       "      <td>0.032908</td>\n",
       "      <td>0.145064</td>\n",
       "      <td>0.051377</td>\n",
       "      <td>0.003694</td>\n",
       "      <td>0.034923</td>\n",
       "      <td>0.038281</td>\n",
       "      <td>0.011081</td>\n",
       "      <td>0.008059</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.360981</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900262</th>\n",
       "      <td>0.066533</td>\n",
       "      <td>0.774657</td>\n",
       "      <td>0.018723</td>\n",
       "      <td>0.014376</td>\n",
       "      <td>0.011033</td>\n",
       "      <td>0.009027</td>\n",
       "      <td>0.001337</td>\n",
       "      <td>0.019057</td>\n",
       "      <td>0.001672</td>\n",
       "      <td>0.001672</td>\n",
       "      <td>0.007021</td>\n",
       "      <td>0.010699</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.064193</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900446</th>\n",
       "      <td>0.107333</td>\n",
       "      <td>0.089333</td>\n",
       "      <td>0.108333</td>\n",
       "      <td>0.013000</td>\n",
       "      <td>0.024333</td>\n",
       "      <td>0.020667</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.020000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.005667</td>\n",
       "      <td>0.360000</td>\n",
       "      <td>0.000333</td>\n",
       "      <td>0.249000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900263</th>\n",
       "      <td>0.153000</td>\n",
       "      <td>0.103000</td>\n",
       "      <td>0.132333</td>\n",
       "      <td>0.050000</td>\n",
       "      <td>0.071667</td>\n",
       "      <td>0.022000</td>\n",
       "      <td>0.001667</td>\n",
       "      <td>0.013667</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.081000</td>\n",
       "      <td>0.011333</td>\n",
       "      <td>0.010333</td>\n",
       "      <td>0.001000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.349000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900555</th>\n",
       "      <td>0.152451</td>\n",
       "      <td>0.119208</td>\n",
       "      <td>0.201478</td>\n",
       "      <td>0.048355</td>\n",
       "      <td>0.044661</td>\n",
       "      <td>0.114171</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.065144</td>\n",
       "      <td>0.005037</td>\n",
       "      <td>0.010745</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.238751</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900426</th>\n",
       "      <td>0.037679</td>\n",
       "      <td>0.094365</td>\n",
       "      <td>0.001000</td>\n",
       "      <td>0.030343</td>\n",
       "      <td>0.396465</td>\n",
       "      <td>0.012671</td>\n",
       "      <td>0.001667</td>\n",
       "      <td>0.000667</td>\n",
       "      <td>0.006002</td>\n",
       "      <td>0.001000</td>\n",
       "      <td>0.006669</td>\n",
       "      <td>0.016339</td>\n",
       "      <td>0.038680</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.356452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900243</th>\n",
       "      <td>0.058333</td>\n",
       "      <td>0.639333</td>\n",
       "      <td>0.045667</td>\n",
       "      <td>0.010667</td>\n",
       "      <td>0.018000</td>\n",
       "      <td>0.021000</td>\n",
       "      <td>0.000333</td>\n",
       "      <td>0.021667</td>\n",
       "      <td>0.002333</td>\n",
       "      <td>0.005333</td>\n",
       "      <td>0.003333</td>\n",
       "      <td>0.004667</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000667</td>\n",
       "      <td>0.168667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900570</th>\n",
       "      <td>0.120454</td>\n",
       "      <td>0.236904</td>\n",
       "      <td>0.012346</td>\n",
       "      <td>0.018352</td>\n",
       "      <td>0.215215</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000667</td>\n",
       "      <td>0.028695</td>\n",
       "      <td>0.002002</td>\n",
       "      <td>0.010010</td>\n",
       "      <td>0.000667</td>\n",
       "      <td>0.003337</td>\n",
       "      <td>0.001001</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.350350</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900572</th>\n",
       "      <td>0.486667</td>\n",
       "      <td>0.039333</td>\n",
       "      <td>0.011333</td>\n",
       "      <td>0.008333</td>\n",
       "      <td>0.012667</td>\n",
       "      <td>0.004333</td>\n",
       "      <td>0.007000</td>\n",
       "      <td>0.018667</td>\n",
       "      <td>0.000333</td>\n",
       "      <td>0.068667</td>\n",
       "      <td>0.001000</td>\n",
       "      <td>0.041000</td>\n",
       "      <td>0.018000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.282667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900323</th>\n",
       "      <td>0.005339</td>\n",
       "      <td>0.021021</td>\n",
       "      <td>0.001335</td>\n",
       "      <td>0.530864</td>\n",
       "      <td>0.003337</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.110110</td>\n",
       "      <td>0.001668</td>\n",
       "      <td>0.001668</td>\n",
       "      <td>0.000667</td>\n",
       "      <td>0.000334</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.323657</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>140 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        g__Streptococcus  g__Staphylococcus  g__Corynebacterium  \\\n",
       "900360          0.128946           0.069174            0.069510   \n",
       "900262          0.066533           0.774657            0.018723   \n",
       "900446          0.107333           0.089333            0.108333   \n",
       "900263          0.153000           0.103000            0.132333   \n",
       "900555          0.152451           0.119208            0.201478   \n",
       "...                  ...                ...                 ...   \n",
       "900426          0.037679           0.094365            0.001000   \n",
       "900243          0.058333           0.639333            0.045667   \n",
       "900570          0.120454           0.236904            0.012346   \n",
       "900572          0.486667           0.039333            0.011333   \n",
       "900323          0.005339           0.021021            0.001335   \n",
       "\n",
       "        g__Acinetobacter  g__Prevotella  g__Micrococcus  \\\n",
       "900360          0.046004       0.032908        0.145064   \n",
       "900262          0.014376       0.011033        0.009027   \n",
       "900446          0.013000       0.024333        0.020667   \n",
       "900263          0.050000       0.071667        0.022000   \n",
       "900555          0.048355       0.044661        0.114171   \n",
       "...                  ...            ...             ...   \n",
       "900426          0.030343       0.396465        0.012671   \n",
       "900243          0.010667       0.018000        0.021000   \n",
       "900570          0.018352       0.215215        0.000000   \n",
       "900572          0.008333       0.012667        0.004333   \n",
       "900323          0.530864       0.003337        0.000000   \n",
       "\n",
       "        g__Haemophilus_D_734546  g__Veillonella_A  g__Psychrobacter  \\\n",
       "900360                 0.051377          0.003694          0.034923   \n",
       "900262                 0.001337          0.019057          0.001672   \n",
       "900446                 0.000000          0.020000          0.000000   \n",
       "900263                 0.001667          0.013667          0.000000   \n",
       "900555                 0.000000          0.065144          0.005037   \n",
       "...                         ...               ...               ...   \n",
       "900426                 0.001667          0.000667          0.006002   \n",
       "900243                 0.000333          0.021667          0.002333   \n",
       "900570                 0.000667          0.028695          0.002002   \n",
       "900572                 0.007000          0.018667          0.000333   \n",
       "900323                 0.000000          0.000000          0.110110   \n",
       "\n",
       "        g__Neisseria_563205  g__SIO2C1  g__Cutibacterium  \\\n",
       "900360             0.038281   0.011081          0.008059   \n",
       "900262             0.001672   0.007021          0.010699   \n",
       "900446             0.002000   0.000000          0.005667   \n",
       "900263             0.081000   0.011333          0.010333   \n",
       "900555             0.010745   0.000000          0.000000   \n",
       "...                     ...        ...               ...   \n",
       "900426             0.001000   0.006669          0.016339   \n",
       "900243             0.005333   0.003333          0.004667   \n",
       "900570             0.010010   0.000667          0.003337   \n",
       "900572             0.068667   0.001000          0.041000   \n",
       "900323             0.001668   0.001668          0.000667   \n",
       "\n",
       "        g__Chryseobacterium_796614  g__Pseudomonas_E_647464    Others  \n",
       "900360                    0.000000                 0.000000  0.360981  \n",
       "900262                    0.000000                 0.000000  0.064193  \n",
       "900446                    0.360000                 0.000333  0.249000  \n",
       "900263                    0.001000                 0.000000  0.349000  \n",
       "900555                    0.000000                 0.000000  0.238751  \n",
       "...                            ...                      ...       ...  \n",
       "900426                    0.038680                 0.000000  0.356452  \n",
       "900243                    0.000000                 0.000667  0.168667  \n",
       "900570                    0.001001                 0.000000  0.350350  \n",
       "900572                    0.018000                 0.000000  0.282667  \n",
       "900323                    0.000334                 0.000000  0.323657  \n",
       "\n",
       "[140 rows x 15 columns]"
      ]
     },
     "execution_count": 547,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Paths to input files\n",
    "biom_path = '../Data/Tables/Relative_Abundance_Tables/df_16S_filtered_feature_table_rare_Genus_relative_abundance.biom'\n",
    "# biom_path = '../Data/Tables/Absolute_Abundance_Tables/209766_filtered_by_prevalence_5pct_rare_Genus-ASV-non-collapse.biom'\n",
    "metadata_path = '../Data/Metadata/updated_clean_ant_skin_metadata_microbiome_type-location.tab'\n",
    "\n",
    "df, metadata = load_biom_table(biom_path, metadata_path)\n",
    "df = collapse_top_15(df)\n",
    "\n",
    "# Sum g__Unknown into Others row\n",
    "if 'g__Unknown' in df.index:\n",
    "    # Add g__Unknown values to Others\n",
    "    df.loc['Others'] = df.loc['Others'] + df.loc['g__Unknown']\n",
    "    # Drop g__Unknown row\n",
    "    df = df.drop('g__Unknown')\n",
    "\n",
    "# Switch order of first and second row\n",
    "# Get first two row names\n",
    "first_row = df.index[0]\n",
    "second_row = df.index[1]\n",
    "\n",
    "# Create new df with swapped rows\n",
    "df = df.reindex([second_row, first_row] + list(df.index[2:]))\n",
    "\n",
    "df = df.T\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 548,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PlateNumber</th>\n",
       "      <th>PlateLocation</th>\n",
       "      <th>i5</th>\n",
       "      <th>i5Sequence</th>\n",
       "      <th>i7</th>\n",
       "      <th>i7Sequence</th>\n",
       "      <th>identifier</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Plate ID</th>\n",
       "      <th>Well location</th>\n",
       "      <th>...</th>\n",
       "      <th>enrolment_season</th>\n",
       "      <th>hiv_exposure</th>\n",
       "      <th>hiv_status</th>\n",
       "      <th>household_size</th>\n",
       "      <th>o_scorad</th>\n",
       "      <th>FWD_filepath</th>\n",
       "      <th>REV_filepath</th>\n",
       "      <th>o_scorad_adj</th>\n",
       "      <th>individual_case</th>\n",
       "      <th>individual_case_location</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Ca009ST_L</th>\n",
       "      <td>1</td>\n",
       "      <td>A1</td>\n",
       "      <td>SA501</td>\n",
       "      <td>ATCGTACG</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA501</td>\n",
       "      <td>CGAGAGTT-ATCGTACG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>A1</td>\n",
       "      <td>...</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>4.0</td>\n",
       "      <td>40</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>40</td>\n",
       "      <td>AD skin</td>\n",
       "      <td>AD skin Cape Town</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900221</th>\n",
       "      <td>1</td>\n",
       "      <td>B1</td>\n",
       "      <td>SA502</td>\n",
       "      <td>ACTATCTG</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA502</td>\n",
       "      <td>CGAGAGTT-ACTATCTG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>B1</td>\n",
       "      <td>...</td>\n",
       "      <td>Winter</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>7.0</td>\n",
       "      <td>34</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>34</td>\n",
       "      <td>AD skin</td>\n",
       "      <td>AD skin Umtata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca010EB_L</th>\n",
       "      <td>1</td>\n",
       "      <td>C1</td>\n",
       "      <td>SA503</td>\n",
       "      <td>TAGCGAGT</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA503</td>\n",
       "      <td>CGAGAGTT-TAGCGAGT</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>C1</td>\n",
       "      <td>...</td>\n",
       "      <td>Spring</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>7.0</td>\n",
       "      <td>21</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>21</td>\n",
       "      <td>AD skin</td>\n",
       "      <td>AD skin Cape Town</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900460</th>\n",
       "      <td>1</td>\n",
       "      <td>D1</td>\n",
       "      <td>SA504</td>\n",
       "      <td>CTGCGTGT</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA504</td>\n",
       "      <td>CGAGAGTT-CTGCGTGT</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>D1</td>\n",
       "      <td>...</td>\n",
       "      <td>Spring</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>40</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>40</td>\n",
       "      <td>AD skin</td>\n",
       "      <td>AD skin Umtata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900051</th>\n",
       "      <td>1</td>\n",
       "      <td>E1</td>\n",
       "      <td>SA505</td>\n",
       "      <td>TCATCGAG</td>\n",
       "      <td>SA701</td>\n",
       "      <td>CGAGAGTT</td>\n",
       "      <td>SA701SA505</td>\n",
       "      <td>CGAGAGTT-TCATCGAG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>E1</td>\n",
       "      <td>...</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>7.0</td>\n",
       "      <td>41</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>41</td>\n",
       "      <td>AD skin</td>\n",
       "      <td>AD skin Cape Town</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca006ON_L_2</th>\n",
       "      <td>6</td>\n",
       "      <td>H1</td>\n",
       "      <td>SA508</td>\n",
       "      <td>GACACCGT</td>\n",
       "      <td>SB701</td>\n",
       "      <td>CTCGACTT</td>\n",
       "      <td>SB701SA508</td>\n",
       "      <td>CTCGACTT-GACACCGT</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>H1</td>\n",
       "      <td>...</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>3.0</td>\n",
       "      <td>34</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>34</td>\n",
       "      <td>AD skin</td>\n",
       "      <td>AD skin Cape Town</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca006ON_NL</th>\n",
       "      <td>6</td>\n",
       "      <td>F2</td>\n",
       "      <td>SA506</td>\n",
       "      <td>CGTGAGTG</td>\n",
       "      <td>SB702</td>\n",
       "      <td>CGAAGTAT</td>\n",
       "      <td>SB702SA506</td>\n",
       "      <td>CGAAGTAT-CGTGAGTG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>F2</td>\n",
       "      <td>...</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>3.0</td>\n",
       "      <td>34</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>34</td>\n",
       "      <td>AD skin</td>\n",
       "      <td>AD skin Cape Town</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca006ON_NL_2</th>\n",
       "      <td>6</td>\n",
       "      <td>H2</td>\n",
       "      <td>SA508</td>\n",
       "      <td>GACACCGT</td>\n",
       "      <td>SB702</td>\n",
       "      <td>CGAAGTAT</td>\n",
       "      <td>SB702SA508</td>\n",
       "      <td>CGAAGTAT-GACACCGT</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>H2</td>\n",
       "      <td>...</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>3.0</td>\n",
       "      <td>34</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>34</td>\n",
       "      <td>AD skin</td>\n",
       "      <td>AD skin Cape Town</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca006ON_PN</th>\n",
       "      <td>6</td>\n",
       "      <td>F3</td>\n",
       "      <td>SA506</td>\n",
       "      <td>CGTGAGTG</td>\n",
       "      <td>SB703</td>\n",
       "      <td>TAGCAGCT</td>\n",
       "      <td>SB703SA506</td>\n",
       "      <td>TAGCAGCT-CGTGAGTG</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>F3</td>\n",
       "      <td>...</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>3.0</td>\n",
       "      <td>34</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>34</td>\n",
       "      <td>AD nares</td>\n",
       "      <td>AD nares Cape Town</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ca006ON_PN_2</th>\n",
       "      <td>6</td>\n",
       "      <td>H3</td>\n",
       "      <td>SA508</td>\n",
       "      <td>GACACCGT</td>\n",
       "      <td>SB703</td>\n",
       "      <td>TAGCAGCT</td>\n",
       "      <td>SB703SA508</td>\n",
       "      <td>TAGCAGCT-GACACCGT</td>\n",
       "      <td>1.010000e+21</td>\n",
       "      <td>H3</td>\n",
       "      <td>...</td>\n",
       "      <td>Autumn</td>\n",
       "      <td>Unexposed</td>\n",
       "      <td>negative</td>\n",
       "      <td>3.0</td>\n",
       "      <td>34</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>/Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...</td>\n",
       "      <td>34</td>\n",
       "      <td>AD nares</td>\n",
       "      <td>AD nares Cape Town</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>502 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              PlateNumber PlateLocation     i5 i5Sequence     i7 i7Sequence  \\\n",
       "Ca009ST_L               1            A1  SA501   ATCGTACG  SA701   CGAGAGTT   \n",
       "900221                  1            B1  SA502   ACTATCTG  SA701   CGAGAGTT   \n",
       "Ca010EB_L               1            C1  SA503   TAGCGAGT  SA701   CGAGAGTT   \n",
       "900460                  1            D1  SA504   CTGCGTGT  SA701   CGAGAGTT   \n",
       "900051                  1            E1  SA505   TCATCGAG  SA701   CGAGAGTT   \n",
       "...                   ...           ...    ...        ...    ...        ...   \n",
       "Ca006ON_L_2             6            H1  SA508   GACACCGT  SB701   CTCGACTT   \n",
       "Ca006ON_NL              6            F2  SA506   CGTGAGTG  SB702   CGAAGTAT   \n",
       "Ca006ON_NL_2            6            H2  SA508   GACACCGT  SB702   CGAAGTAT   \n",
       "Ca006ON_PN              6            F3  SA506   CGTGAGTG  SB703   TAGCAGCT   \n",
       "Ca006ON_PN_2            6            H3  SA508   GACACCGT  SB703   TAGCAGCT   \n",
       "\n",
       "              identifier           Sequence      Plate ID Well location  ...  \\\n",
       "Ca009ST_L     SA701SA501  CGAGAGTT-ATCGTACG  1.010000e+21            A1  ...   \n",
       "900221        SA701SA502  CGAGAGTT-ACTATCTG  1.010000e+21            B1  ...   \n",
       "Ca010EB_L     SA701SA503  CGAGAGTT-TAGCGAGT  1.010000e+21            C1  ...   \n",
       "900460        SA701SA504  CGAGAGTT-CTGCGTGT  1.010000e+21            D1  ...   \n",
       "900051        SA701SA505  CGAGAGTT-TCATCGAG  1.010000e+21            E1  ...   \n",
       "...                  ...                ...           ...           ...  ...   \n",
       "Ca006ON_L_2   SB701SA508  CTCGACTT-GACACCGT  1.010000e+21            H1  ...   \n",
       "Ca006ON_NL    SB702SA506  CGAAGTAT-CGTGAGTG  1.010000e+21            F2  ...   \n",
       "Ca006ON_NL_2  SB702SA508  CGAAGTAT-GACACCGT  1.010000e+21            H2  ...   \n",
       "Ca006ON_PN    SB703SA506  TAGCAGCT-CGTGAGTG  1.010000e+21            F3  ...   \n",
       "Ca006ON_PN_2  SB703SA508  TAGCAGCT-GACACCGT  1.010000e+21            H3  ...   \n",
       "\n",
       "              enrolment_season hiv_exposure hiv_status household_size  \\\n",
       "Ca009ST_L              Autumn     Unexposed   negative            4.0   \n",
       "900221                  Winter    Unexposed   negative            7.0   \n",
       "Ca010EB_L               Spring    Unexposed   negative            7.0   \n",
       "900460                  Spring    Unexposed        NaN            4.0   \n",
       "900051                 Autumn     Unexposed   negative            7.0   \n",
       "...                        ...          ...        ...            ...   \n",
       "Ca006ON_L_2             Autumn    Unexposed   negative            3.0   \n",
       "Ca006ON_NL              Autumn    Unexposed   negative            3.0   \n",
       "Ca006ON_NL_2            Autumn    Unexposed   negative            3.0   \n",
       "Ca006ON_PN              Autumn    Unexposed   negative            3.0   \n",
       "Ca006ON_PN_2            Autumn    Unexposed   negative            3.0   \n",
       "\n",
       "             o_scorad                                       FWD_filepath  \\\n",
       "Ca009ST_L          40  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "900221             34  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "Ca010EB_L          21  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "900460             40  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "900051             41  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "...               ...                                                ...   \n",
       "Ca006ON_L_2        34  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "Ca006ON_NL         34  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "Ca006ON_NL_2       34  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "Ca006ON_PN         34  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "Ca006ON_PN_2       34  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...   \n",
       "\n",
       "                                                   REV_filepath o_scorad_adj  \\\n",
       "Ca009ST_L     /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...           40   \n",
       "900221        /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...           34   \n",
       "Ca010EB_L     /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...           21   \n",
       "900460        /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...           40   \n",
       "900051        /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...           41   \n",
       "...                                                         ...          ...   \n",
       "Ca006ON_L_2   /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...           34   \n",
       "Ca006ON_NL    /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...           34   \n",
       "Ca006ON_NL_2  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...           34   \n",
       "Ca006ON_PN    /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...           34   \n",
       "Ca006ON_PN_2  /Users/yac027/Gallo_lab/16S_AD_Dube_Dupont/ato...           34   \n",
       "\n",
       "             individual_case individual_case_location  \n",
       "Ca009ST_L            AD skin        AD skin Cape Town  \n",
       "900221               AD skin           AD skin Umtata  \n",
       "Ca010EB_L            AD skin        AD skin Cape Town  \n",
       "900460               AD skin           AD skin Umtata  \n",
       "900051               AD skin        AD skin Cape Town  \n",
       "...                      ...                      ...  \n",
       "Ca006ON_L_2          AD skin        AD skin Cape Town  \n",
       "Ca006ON_NL           AD skin        AD skin Cape Town  \n",
       "Ca006ON_NL_2         AD skin        AD skin Cape Town  \n",
       "Ca006ON_PN          AD nares       AD nares Cape Town  \n",
       "Ca006ON_PN_2        AD nares       AD nares Cape Town  \n",
       "\n",
       "[502 rows x 34 columns]"
      ]
     },
     "execution_count": 548,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata = pd.read_csv('../Data/Metadata/updated_clean_ant_skin_metadata_microbiome_type-location.tab', sep='\\t')\n",
    "# Drop first column and set #sample-id as index\n",
    "metadata = metadata.drop('Unnamed: 0', axis=1).set_index('#sample-id')\n",
    "# Remove name of index column \n",
    "metadata.index.name = None\n",
    "\n",
    "metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 549,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "case_type\n",
       "case-nonlesional skin       111\n",
       "case-anterior nares         108\n",
       "case-lesional skin          107\n",
       "control-anterior nares       89\n",
       "control-nonlesional skin     87\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 549,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata['case_type'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 550,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "site_group\n",
      "Umtata | case-lesional skin             37\n",
      "Umtata | control-nonlesional skin       27\n",
      "Cape Town | case-lesional skin          24\n",
      "Umtata | case-nonlesional skin          19\n",
      "Cape Town | case-nonlesional skin       19\n",
      "Cape Town | control-nonlesional skin    14\n",
      "Name: count, dtype: int64\n",
      "{'Umtata | case-lesional skin': 37, 'Umtata | control-nonlesional skin': 27, 'Cape Town | case-lesional skin': 24, 'Umtata | case-nonlesional skin': 19, 'Cape Town | case-nonlesional skin': 19, 'Cape Town | control-nonlesional skin': 14}\n",
      "sFigure saved to: ../Figures/Main/Fig_3A.png\n"
     ]
    }
   ],
   "source": [
    "output_dir = '../Figures/Main/'\n",
    "\n",
    "\n",
    "taxa_colors = {\n",
    "    'g__Streptococcus': '#ff0000',          # bright red\n",
    "    'g__Corynebacterium': '#a5d8ff',        # light sky blue\n",
    "    'g__Acinetobacter': '#ffe59a',          # soft pastel yellow\n",
    "    'g__Prevotella': '#a5c9a1',             # muted greenish mint\n",
    "    'g__Micrococcus': '#0000FF',            # blue\n",
    "    'g__Haemophilus_D_734546': '#E6E6FA',   # pale lavender\n",
    "    'g__Staphylococcus': '#000000',         # black\n",
    "    'g__Veillonella_A': '#FA5F55',          # sunset orange\n",
    "    'g__Psychrobacter': '#accbe1',          # cool steel blue\n",
    "    'g__Neisseria_563205': '#b0d9b1',       # light sage green\n",
    "    'g__SIO2C1': '#f7c59f',                 # warm peach\n",
    "    'g__Cutibacterium': '#FADADD',          # pale pink\n",
    "    'g__Chryseobacterium_7966': '#dfc5a5',  # muted tan\n",
    "    'g__Pseudomonas_E_647464': '#d4a5a5',   # muted rose\n",
    "    'Others': '#d9d9d9'                     # neutral gray\n",
    "}\n",
    "\n",
    "\n",
    "# Run the plotting function\n",
    "plot_relative_abundance_two_panels(\n",
    "    df=df,\n",
    "    metadata=metadata,\n",
    "    output_dir=output_dir,\n",
    "    taxa_color_map=taxa_colors)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "qiime2-metagenome-2024.10",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
